--- lib/Language/Haskell/Stylish/Step/Imports.hs.bak	2018-04-16 12:58:57.741846300 +0200
+++ lib/Language/Haskell/Stylish/Step/Imports.hs	2018-04-16 13:00:21.139709671 +0200
@@ -168,10 +168,13 @@
   = ImportSome [H.CName l]  -- ^ @A(x, y, z)@
   | ImportAll               -- ^ @A(..)@
 
+instance Ord l => Semigroup (ImportPortion l) where
+  (ImportSome a) <> (ImportSome b) = ImportSome (setUnion a b)
+  _ <> _ = ImportAll
+
 instance Ord l => Monoid (ImportPortion l) where
   mempty = ImportSome []
-  mappend (ImportSome a) (ImportSome b) = ImportSome (setUnion a b)
-  mappend _ _                           = ImportAll
+  mappend = (<>)
 
 -- | O(n log n) union.
 setUnion :: Ord a => [a] -> [a] -> [a]
