--- a/libfreerdp/codec/h264_ffmpeg.c	2019-11-10 20:28:20.415018315 +0100
+++ b/libfreerdp/codec/h264_ffmpeg.c	2019-11-10 20:37:32.842070000 +0100
@@ -28,12 +28,12 @@
 #include <libavcodec/avcodec.h>
 #include <libavutil/opt.h>
 
-#ifdef WITH_VAAPI
+#ifdef WITH_HWDECODE
 #if LIBAVUTIL_VERSION_INT >= AV_VERSION_INT(55, 9, 0)
 #include <libavutil/hwcontext.h>
 #else
 #pragma warning You have asked for VA-API decoding, but your version of libavutil is too old! Disabling.
-#undef WITH_VAAPI
+#undef WITH_HWDECODE
 #endif
 #endif
 
@@ -65,8 +65,17 @@
 	error_string((char[64]){0}, 64, errnum)
 #endif
 
-#ifdef WITH_VAAPI
-#define VAAPI_DEVICE "/dev/dri/renderD128"
+#if defined(WITH_HWDECODE)
+#if WITH_HWDECODE == VAAPI
+#define HWDECODE_DEVICE "/dev/dri/renderD128"
+#else
+#define HWDECODE_DEVICE NULL
+#endif
+
+#define PPCAT_NX(A, B) A ## B
+#define PPCAT(A, B) PPCAT_NX(A, B)
+#define AV_HWDEVICE_TYPE PPCAT(AV_HWDEVICE_TYPE_, WITH_HWDECODE)
+#define AV_PIX_FMT PPCAT(AV_PIX_FMT_, WITH_HWDECODE)
 #endif
 
 struct _H264_CONTEXT_LIBAVCODEC
@@ -78,7 +87,7 @@
 	AVCodecParserContext* codecParser;
 	AVFrame* videoFrame;
 	AVPacket packet;
-#ifdef WITH_VAAPI
+#ifdef WITH_HWDECODE
 	AVBufferRef* hwctx;
 	AVFrame* hwVideoFrame;
 	enum AVPixelFormat hw_pix_fmt;
@@ -210,7 +219,7 @@
 
 	do
 	{
-#ifdef WITH_VAAPI
+#ifdef WITH_HWDECODE
 		status = avcodec_receive_frame(sys->codecDecoderContext,
 		                               sys->hwctx ? sys->hwVideoFrame : sys->videoFrame);
 #else
@@ -221,7 +230,7 @@
 
 	gotFrame = (status == 0);
 #else
-#ifdef WITH_VAAPI
+#ifdef WITH_HWDECODE
 	status = avcodec_decode_video2(sys->codecDecoderContext,
 	                               sys->hwctx ? sys->hwVideoFrame : sys->videoFrame, &gotFrame,
 	                               &sys->packet);
@@ -237,7 +246,7 @@
 		return -1;
 	}
 
-#ifdef WITH_VAAPI
+#ifdef WITH_HWDECODE
 
 	if (sys->hwctx)
 	{
@@ -406,7 +415,7 @@
 #endif
 	}
 
-#ifdef WITH_VAAPI
+#ifdef WITH_HWDECODE
 
 	if (sys->hwVideoFrame)
 	{
@@ -450,7 +459,7 @@
 	h264->pSystemData = NULL;
 }
 
-#ifdef WITH_VAAPI
+#ifdef WITH_HWDECODE
 static enum AVPixelFormat libavcodec_get_format(struct AVCodecContext* ctx,
         const enum AVPixelFormat* fmts)
 {
@@ -536,11 +545,11 @@
 			sys->codecDecoderContext->flags |= AV_CODEC_FLAG_TRUNCATED;
 		}
 
-#ifdef WITH_VAAPI
+#ifdef WITH_HWDECODE
 
 		if (!sys->hwctx)
 		{
-			int ret = av_hwdevice_ctx_create(&sys->hwctx, AV_HWDEVICE_TYPE_VAAPI, VAAPI_DEVICE, NULL, 0);
+			int ret = av_hwdevice_ctx_create(&sys->hwctx, AV_HWDEVICE_TYPE, HWDECODE_DEVICE, NULL, 0);
 
 			if (ret < 0)
 			{
@@ -553,7 +562,7 @@
 		}
 
 		sys->codecDecoderContext->get_format = libavcodec_get_format;
-		sys->hw_pix_fmt = AV_PIX_FMT_VAAPI;
+		sys->hw_pix_fmt = AV_PIX_FMT;
 #if LIBAVCODEC_VERSION_INT >= AV_VERSION_INT(57, 80, 100)
 		sys->codecDecoderContext->hw_device_ctx = av_buffer_ref(sys->hwctx);
 #endif
@@ -578,7 +587,7 @@
 
 #if LIBAVCODEC_VERSION_INT >= AV_VERSION_INT(55, 18, 102)
 	sys->videoFrame = av_frame_alloc();
-#ifdef WITH_VAAPI
+#ifdef WITH_HWDECODE
 	sys->hwVideoFrame = av_frame_alloc();
 #endif
 #else
@@ -591,7 +600,7 @@
 		goto EXCEPTION;
 	}
 
-#ifdef WITH_VAAPI
+#ifdef WITH_HWDECODE
 
 	if (!sys->hwVideoFrame)
 	{
--- a/CMakeLists.txt	2019-11-10 20:26:20.606875053 +0100
+++ b/CMakeLists.txt	2019-11-10 20:27:29.729266185 +0100
@@ -726,9 +726,9 @@
 set(FFMPEG_FEATURE_PURPOSE "multimedia")
 set(FFMPEG_FEATURE_DESCRIPTION "multimedia redirection, audio and video playback")
 
-set(VAAPI_FEATURE_TYPE "OPTIONAL")
-set(VAAPI_FEATURE_PURPOSE "multimedia")
-set(VAAPI_FEATURE_DESCRIPTION "VA-API hardware acceleration for video playback")
+set(HWDECODE_FEATURE_TYPE "OPTIONAL")
+set(HWDECODE_FEATURE_PURPOSE "multimedia")
+set(HWDECODE_FEATURE_DESCRIPTION "hardware acceleration for video playback")
 
 set(GSTREAMER_0_10_FEATURE_TYPE "OPTIONAL")
 set(GSTREAMER_0_10_FEATURE_PURPOSE "multimedia")
@@ -789,7 +789,7 @@
 	set(CUPS_FEATURE_TYPE "DISABLED")
 	set(PCSC_FEATURE_TYPE "DISABLED")
 	set(FFMPEG_FEATURE_TYPE "DISABLED")
-	set(VAAPI_FEATURE_TYPE "DISABLED")
+	set(HWDECODE_FEATURE_TYPE "DISABLED")
 	set(GSTREAMER_1_0_FEATURE_TYPE "DISABLED")
 	set(GSTREAMER_0_10_FEATURE_TYPE "OPTIONAL")
 	set(OPENSLES_FEATURE_TYPE "DISABLED")
@@ -798,7 +798,7 @@
 if(APPLE)
 	set(DIRECTFB_FEATURE_TYPE "DISABLED")
 	set(FFMPEG_FEATURE_TYPE "OPTIONAL")
-	set(VAAPI_FEATURE_TYPE "DISABLED")
+	set(HWDECODE_FEATURE_TYPE "DISABLED")
 	set(GSTREAMER_1_0_FEATURE_TYPE "OPTIONAL")
 	set(X11_FEATURE_TYPE "OPTIONAL")
 	set(WAYLAND_FEATURE_TYPE "DISABLED")
@@ -840,7 +840,7 @@
 	set(CUPS_FEATURE_TYPE "DISABLED")
 	set(PCSC_FEATURE_TYPE "DISABLED")
 	set(FFMPEG_FEATURE_TYPE "DISABLED")
-	set(VAAPI_FEATURE_TYPE "DISABLED")
+	set(HWDECODE_FEATURE_TYPE "DISABLED")
 	set(GSTREAMER_1_0_FEATURE_TYPE "DISABLED")
 	set(GSTREAMER_0_10_FEATURE_TYPE "DISABLED")
 	set(OPENSLES_FEATURE_TYPE "REQUIRED")
--- a/config.h.in	2018-11-20 11:43:51.000000000 +0100
+++ b/config.h.in	2019-11-10 20:26:12.326588475 +0100
@@ -66,7 +66,7 @@
 #cmakedefine WITH_X264
 #cmakedefine WITH_MEDIA_FOUNDATION
 
-#cmakedefine WITH_VAAPI
+#cmakedefine WITH_HWDECODE "@WITH_HWDECODE@"
 
 /* Plugins */
 #cmakedefine BUILTIN_CHANNELS
--- libfreerdp/codec/h264_ffmpeg.c	2018-11-20 11:43:51.000000000 +0100
+++ /h264_ffmpeg.c	2019-11-10 21:05:17.619771050 +0100
