--- a/libfreerdp/locale/keyboard_xkbfile.c	2020-11-11 14:18:51.597253682 +0100
+++ b/libfreerdp/locale/keyboard_xkbfile.c	2020-11-11 14:20:28.838325495 +0100
@@ -197,8 +197,8 @@
 	// { "I188", RDP_SCANCODE_ },	// evdev 188 KEY_KPRIGHTPAREN
 	// { "I189", RDP_SCANCODE_ },	// evdev 189 KEY_NEW
 	// { "I190", RDP_SCANCODE_ },	// evdev 190 KEY_REDO
-	// { "I208", RDP_SCANCODE_ },	// evdev 208 KEY_PLAYCD
-	// { "I209", RDP_SCANCODE_ },	// evdev 209 KEY_PAUSECD
+	{ "I208", RDP_SCANCODE_MEDIA_PLAY_PAUSE },	// evdev 208 KEY_PLAYCD
+	{ "I209", RDP_SCANCODE_MEDIA_PLAY_PAUSE },	// evdev 209 KEY_PAUSECD
 	// { "I210", RDP_SCANCODE_ },	// evdev 210 KEY_PROG3
 	// { "I211", RDP_SCANCODE_ },	// evdev 211 KEY_PROG4
 	// { "I212", RDP_SCANCODE_ },	// evdev 212 KEY_DASHBOARD
