--- a/src/Ide/Plugin/Rename.hs	2021-12-05 12:32:16.690989736 +0100
+++ b/src/Ide/Plugin/Rename.hs	2021-12-05 12:32:34.995214423 +0100
@@ -125,7 +125,7 @@
                 True
                 (nameOccName name)
                 (Just $ moduleName mod)
-                (Just $ moduleUnitId mod)
+                (Just $ moduleUnit mod)
                 [fromNormalizedFilePath nfp]
     pure $ nubOrd $ astRefs ++ dbRefs
 
