--- a/eng/native/configurecompiler.cmake	2020-07-21 22:56:08.261609755 +0200
+++ b/eng/native/configurecompiler.cmake	2020-07-21 22:56:39.246841586 +0200
@@ -583,7 +583,8 @@
         ERROR_QUIET
         OUTPUT_VARIABLE ldVersionOutput)
 
-    if("${ldVersionOutput}" MATCHES "GNU ld" OR "${ldVersionOutput}" MATCHES "GNU gold")
+    if("${ldVersionOutput}" MATCHES "GNU ld" OR "${ldVersionOutput}" MATCHES "GNU gold"
+       OR "${ldVersionOutput}" MATCHES "LLD")
         set(LD_GNU 1)
     elseif("${ldVersionOutput}" MATCHES "Solaris Link")
         set(LD_SOLARIS 1)
