diff --git a/plugins/CMakeLists.txt b/plugins/CMakeLists.txt
index dda19398bd3d1187c36d912844ff3c3aaef107e5..1e304b88e1d7801537bcef13cc00773d889ab3a3 100644
--- a/plugins/CMakeLists.txt
+++ b/plugins/CMakeLists.txt
@@ -39,7 +39,10 @@ if(NOT SAILFISHOS)
     add_subdirectory(virtualmonitor)
 
     if(${CMAKE_SYSTEM_NAME} STREQUAL "Linux")
-    	add_subdirectory(mmtelephony)
+        find_package(KF5ModemManagerQt ${KF5_MIN_VERSION})
+        if (TARGET KF5::ModemManagerQt)
+            add_subdirectory(mmtelephony)
+        endif()
     endif()
 
     if(NOT WIN32)
diff --git a/plugins/mmtelephony/CMakeLists.txt b/plugins/mmtelephony/CMakeLists.txt
index 1fd9afe1ea1562a81dccf1b1a96df251a978a223..c792b4c7c0c0bc3c360be885ac1c8dfc420139fa 100644
--- a/plugins/mmtelephony/CMakeLists.txt
+++ b/plugins/mmtelephony/CMakeLists.txt
@@ -1,4 +1,3 @@
-find_package(KF5 ${KF5_MIN_VERSION} REQUIRED COMPONENTS ModemManagerQt)
 set(debug_file_SRCS)
 ecm_qt_declare_logging_category(
     debug_file_SRCS HEADER plugin_mmtelephony_debug.h
