--- a/tests/gpgscm/scheme-private.h	2017-01-23 19:39:13.000000000 +0100
+++ b/tests/gpgscm/scheme-private.h	2017-02-04 17:49:33.292658611 +0100
@@ -162,9 +162,7 @@
 long    fcells;          /* # of free cells */
 size_t  inhibit_gc;      /* nesting of gc_disable */
 size_t  reserved_cells;  /* # of reserved cells */
-#ifndef NDEBUG
 int     reserved_lineno;   /* location of last reservation */
-#endif
 
 pointer inport;
 pointer outport;
--- a/tests/gpgscm/scheme.c	2017-02-04 17:51:07.100756094 +0100
+++ b/tests/gpgscm/scheme.c	2017-02-04 17:51:25.923575245 +0100
@@ -774,7 +774,7 @@
 {
 #ifdef NDEBUG
   fprintf(stderr,
-	  "insufficient reservation\n")
+	  "insufficient reservation\n");
 #else
   fprintf(stderr,
 	  "insufficient reservation in line %d\n",
