diff --git a/GhcMod/Gap.hs b/GhcMod/Gap.hs
index 5b65898..4ba5ab4 100644
--- a/GhcMod/Gap.hs
+++ b/GhcMod/Gap.hs
@@ -234,6 +234,7 @@ renderGm = Pretty.fullRender Pretty.PageMode 80 1.2 string_txt ""
    string_txt (Pretty.PStr s1) s2 = unpackFS s1 ++ s2
 #if __GLASGOW_HASKELL__ >= 806
    string_txt (Pretty.LStr s1) s2 = unpackLitString s1 ++ s2
+   string_txt (Pretty.RStr n c) s2 = replicate n c ++ s2
 #else
    string_txt (Pretty.LStr s1 _) s2 = unpackLitString s1 ++ s2
 #endif
