diff --git a/Haskell/Ide/Engine/ArtifactMap.hs b/Haskell/Ide/Engine/ArtifactMap.hs
index ad1d060..12f9333 100644
--- a/Haskell/Ide/Engine/ArtifactMap.hs
+++ b/Haskell/Ide/Engine/ArtifactMap.hs
@@ -57,15 +57,15 @@ genLocMap tm = names
     checker _                             = IM.empty
 
 #if __GLASGOW_HASKELL__ > 710
-    fieldOcc :: GHC.FieldOcc GHC.Name -> LocMap
+    fieldOcc :: GHC.FieldOcc GHC.GhcRn -> LocMap
     fieldOcc (GHC.FieldOcc (GHC.L (GHC.RealSrcSpan r) _) n) = IM.singleton (rspToInt r) n
     fieldOcc _ = IM.empty
 
-    hsRecFieldN :: GHC.LHsExpr GHC.Name -> LocMap
+    hsRecFieldN :: GHC.LHsExpr GHC.GhcRn -> LocMap
     hsRecFieldN (GHC.L _ (GHC.HsRecFld (GHC.Unambiguous (GHC.L (GHC.RealSrcSpan r) _) n) )) = IM.singleton (rspToInt r) n
     hsRecFieldN _ = IM.empty
 
-    hsRecFieldT :: GHC.LHsExpr GHC.Id -> LocMap
+    hsRecFieldT :: GHC.LHsExpr GHC.GhcTc -> LocMap
     hsRecFieldT (GHC.L _ (GHC.HsRecFld (GHC.Ambiguous (GHC.L (GHC.RealSrcSpan r) _) n) )) = IM.singleton (rspToInt r) (Var.varName n)
     hsRecFieldT _ = IM.empty
 #endif
