--- a/CMakeLists.txt	2018-02-17 12:19:38.988447609 +0100
+++ b/CMakeLists.txt	2018-02-17 12:19:55.788408049 +0100
@@ -1,12 +1,12 @@
 cmake_minimum_required(VERSION 3.0)
 
-set(KF5_VERSION "5.42.0") # handled by release scripts
-set(KF5_DEP_VERSION "5.42.0") # handled by release scripts
+set(KF5_VERSION "5.43.0") # handled by release scripts
+set(KF5_DEP_VERSION "5.43.0") # handled by release scripts
 
 project(KIO VERSION ${KF5_VERSION})
 
 include(FeatureSummary)
-find_package(ECM 5.42.0 NO_MODULE)
+find_package(ECM 5.43.0 NO_MODULE)
 set_package_properties(ECM PROPERTIES TYPE REQUIRED DESCRIPTION "Extra CMake Modules." URL "https://projects.kde.org/projects/kdesupport/extra-cmake-modules")
 feature_summary(WHAT REQUIRED_PACKAGES_NOT_FOUND FATAL_ON_MISSING_REQUIRED_PACKAGES)
 
