--- a/src/jdk.jpackage/share/native/common/SysInfo.h	2026-09-02 17:34:57.370539307 +0200
+++ b/src/jdk.jpackage/share/native/common/SysInfo.h	2026-09-02 17:26:35.365071405 +0200
@@ -27,6 +27,7 @@
 #ifndef SYSINFO_H
 #define SYSINFO_H
 
+#include <new>
 #include "tstrings.h"
 
 
--- a/src/jdk.jpackage/linux/native/libapplauncher/Package.cpp	2026-09-02 17:34:44.498208018 +0200
+++ b/src/jdk.jpackage/linux/native/libapplauncher/Package.cpp	2026-09-02 17:27:30.386620077 +0200
@@ -23,6 +23,8 @@
  * questions.
  */
 
+#include <new>
+
 #include "Package.h"
 #include "Executor.h"
 #include "AppLauncher.h"
