--- a/flang-rt/lib/runtime/CMakeLists.txt	2026-08-12 06:58:33.000000000 +0200
+++ b/flang-rt/lib/runtime/CMakeLists.txt	2026-08-13 20:48:46.059871299 +0200
@@ -283,6 +283,7 @@
   add_flangrt_library(flang-rt-mod OBJECT
     ${module_sources}
   )
+  target_compile_options(flang-rt-mod PRIVATE -nobuiltininc)
   flang_module_target(flang-rt-mod PUBLIC)
   add_module_barrier(flang-rt-mod.barrier flang-rt-mod)
 
