--- a/CMakeLists.txt	2022-02-03 21:17:30.291523020 +0100
+++ b/CMakeLists.txt	2022-02-03 21:18:13.516998011 +0100
@@ -27,9 +27,10 @@
 find_package(Qt5 ${QT_MIN_VERSION} CONFIG REQUIRED COMPONENTS DBus)
 
 
-find_package(X11)
-if(X11_FOUND)
-   set(HAVE_X11 1)
+option(WITH_X11 "Build with X11 integration" ON)
+if (WITH_X11)
+    find_package(X11 REQUIRED)
+    set(HAVE_X11 TRUE)
 endif()
 
 add_definitions(
