--- a/magick/locale.c	2016-04-13 20:02:48.526164468 +0200
+++ b/magick/locale.c	2016-04-13 20:04:09.500168455 +0200
@@ -474,12 +474,12 @@
   ssize_t
     n;
 
-  va_list
-    operands;
+//  va_list
+//    operands;
 
-  va_start(operands,format);
-  n=FormatLocaleStringList(string,length,format,operands);
-  va_end(operands);
+//  va_start(operands,format);
+//  n=FormatLocaleStringList(string,length,format,operands);
+//  va_end(operands);
   return(n);
 }
 
