diff --git a/runtime/libpgmath/lib/common/fltfenv.c b/runtime/libpgmath/lib/common/fltfenv.c
index 09090cc1..1ce04dd9 100644
--- a/runtime/libpgmath/lib/common/fltfenv.c
+++ b/runtime/libpgmath/lib/common/fltfenv.c
@@ -511,19 +511,19 @@ __fenv_feraiseexcept(int exc)
 int
 __fenv_feenableexcept(int exc)
 {
-  return feenableexcept(exc);
+  return 0;
 }
 
 int
 __fenv_fedisableexcept(int exc)
 {
-  return fedisableexcept(exc);
+  return 0;
 }
 
 int
 __fenv_fegetexcept(void)
 {
-  return fegetexcept();
+  return 0;
 }
 
 int
@@ -560,8 +560,6 @@ __fenv_feupdateenv(fenv_t *env)
  * __fenv_mask_fz() and __fenv_restore_fz().
  */
 
-#include <fpu_control.h>
-
 /** \brief Set (flush to zero) underflow mode
  *
  * \param uflow zero to allow denorm numbers,
@@ -570,14 +568,14 @@ __fenv_feupdateenv(fenv_t *env)
 int
 __fenv_fesetzerodenorm(int uflow)
 {
-  uint64_t cw;
+  fenv_t fenv;
 
-  _FPU_GETCW(cw);
+  fegetenv(&fenv);
   if (uflow)
-    cw |= (1ULL << 24);
+    fenv.__fpcr |= (1ULL << 24);
   else
-    cw &= ~(1ULL << 24);
-  _FPU_SETCW(cw);
+    fenv.__fpcr &= ~(1ULL << 24);
+  fesetenv(&fenv);
   return 0;
 }
 
@@ -588,10 +586,11 @@ __fenv_fesetzerodenorm(int uflow)
 int
 __fenv_fegetzerodenorm(void)
 {
-  uint64_t cw;
+  fenv_t fenv;
+
+  fegetenv(&fenv);
 
-  _FPU_GETCW(cw);
-  return (cw & (1ULL << 24)) ? 1 : 0;
+  return (fenv.__fpcr & (1ULL << 24)) ? 1 : 0;
 }
 
 /** \brief
@@ -605,16 +604,16 @@ __fenv_fegetzerodenorm(void)
 void
 __fenv_mask_fz(int mask, int *psv)
 {
-  uint64_t tmp;
+  fenv_t fenv;
 
-  _FPU_GETCW(tmp);
+  fegetenv(&fenv);
   if (psv)
-    *psv = ((tmp & (1ULL << 24)) ? 1 : 0);
+    *psv = ((fenv.__fpcr & (1ULL << 24)) ? 1 : 0);
   if (mask)
-    tmp |= (1ULL << 24);
+   fenv.__fpcr |= (1ULL << 24);
   else
-    tmp &= ~(1ULL << 24);
-  _FPU_SETCW(tmp);
+    fenv.__fpcr &= ~(1ULL << 24);
+  fesetenv(&fenv);
 }
 
 /** \brief
@@ -623,14 +622,15 @@ __fenv_mask_fz(int mask, int *psv)
 void
 __fenv_restore_fz(int sv)
 {
-  uint64_t tmp;
+  fenv_t fenv;
+
+  fegetenv(&fenv);
 
-  _FPU_GETCW(tmp);
   if (sv)
-    tmp |= (1ULL << 24);
+    fenv.__fpcr |= (1ULL << 24);
   else
-    tmp &= ~(1ULL << 24);
-  _FPU_SETCW(tmp);
+    fenv.__fpcr &= ~(1ULL << 24);
+  fesetenv(&fenv);
 }
 
 #else
