diff --git a/runtime/libpgmath/lib/common/dceil.c b/runtime/libpgmath/lib/common/dceil.c
index b1f8675e..028b301e 100644
--- a/runtime/libpgmath/lib/common/dceil.c
+++ b/runtime/libpgmath/lib/common/dceil.c
@@ -12,13 +12,13 @@
 
 #if     defined(__AVX__)
 double
-__mth_i_dceil_avx(double x)
+__mth_i_dceil(double x)
 {
   return _mm_cvtsd_f64(_mm_ceil_sd(_mm_set1_pd(x), _mm_set1_pd(x)));
 }
 #elif   defined(__SSE4_1__)
 double
-__mth_i_dceil_sse(double x)
+__mth_i_dceil(double x)
 {
   return _mm_cvtsd_f64(_mm_ceil_sd(_mm_set1_pd(x), _mm_set1_pd(x)));
 }
diff --git a/runtime/libpgmath/lib/common/dfloor.c b/runtime/libpgmath/lib/common/dfloor.c
index db5c54cb..5e78e3d0 100644
--- a/runtime/libpgmath/lib/common/dfloor.c
+++ b/runtime/libpgmath/lib/common/dfloor.c
@@ -12,13 +12,13 @@
 
 #if     defined(__AVX__)
 double
-__mth_i_dfloor_avx(double x)
+__mth_i_dfloor(double x)
 {
   return _mm_cvtsd_f64(_mm_floor_sd(_mm_set1_pd(x), _mm_set1_pd(x)));
 }
 #elif   defined(__SSE4_1__)
 double
-__mth_i_dfloor_sse(double x)
+__mth_i_dfloor(double x)
 {
   return _mm_cvtsd_f64(_mm_floor_sd(_mm_set1_pd(x), _mm_set1_pd(x)));
 }
diff --git a/runtime/libpgmath/lib/common/floor.c b/runtime/libpgmath/lib/common/floor.c
index 68956719..1bd55a8d 100644
--- a/runtime/libpgmath/lib/common/floor.c
+++ b/runtime/libpgmath/lib/common/floor.c
@@ -12,13 +12,13 @@
 
 #if     defined(__AVX__)
 float
-__mth_i_floor_avx(float x)
+__mth_i_floor(float x)
 {
   return _mm_cvtss_f32(_mm_floor_ss(_mm_set1_ps(x), _mm_set1_ps(x)));
 }
 #elif   defined(__SSE4_1__)
 float
-__mth_i_floor_sse(float x)
+__mth_i_floor(float x)
 {
   return _mm_cvtss_f32(_mm_floor_ss(_mm_set1_ps(x), _mm_set1_ps(x)));
 }
