--- a/runtime/flang/hand.c
+++ b/runtime/flang/hand.c
@@ -15,7 +15,7 @@
  */
 
 #ifndef _WIN64
-#include <sys/signal.h>
+#include <signal.h>
 #include "stdioInterf.h"
 #include "fioMacros.h"
 
--- a/runtime/flangrti/ktrap.c
+++ b/runtime/flangrti/ktrap.c
@@ -35,6 +35,7 @@ extern int __ktrapval __attribute__ ((weak));
 void
 __ktrap(void)
 {
+#ifdef __GLIBC__
   if (&__ktrapval != 0) {
     int bv = __ktrapval;
     if (bv != 0) {
@@ -63,6 +64,7 @@ __ktrap(void)
       feenableexcept(excepts);  /* glibc 2.2 extension to fenv.h */
     }
   }
+#endif
 }
 
 #else
--- a/runtime/flangrti/trace_lin.c
+++ b/runtime/flangrti/trace_lin.c
@@ -18,7 +18,9 @@
 #ifdef _WIN64
 #include <windows.h>
 #else
+#ifdef __GLIBC__
 #include <execinfo.h>
+#endif
 #include <unistd.h>
 #include <linux/limits.h>
 #endif
@@ -186,10 +188,13 @@ __abort_trace(int skip)
     dumpregs(regs);
   }
 
+#ifdef __GLIBC__
   size = backtrace(array, MAXTRACE);
   if (skip + 1 >= size) {
+#endif
     fprintf(__io_stderr(), "  --- traceback not available\n");
     return;
+#ifdef __GLIBC__
   }
   strings = backtrace_symbols(array, size);
   if (size < 100) {
@@ -203,6 +208,7 @@ __abort_trace(int skip)
       print_back_trace_line(strings[i], array[i]);
   }
   free(strings);
+#endif
 }
 
 /*
--- a/tools/flang2/flang2exe/gbldefs.h
+++ b/tools/flang2/flang2exe/gbldefs.h
@@ -25,6 +25,7 @@
 #define BE_GBLDEFS_H_
 
 #include <stdint.h>
+#include <alloca.h>
 #include "universal.h"
 #include "platform.h"
 #include "pgifeat.h"
--- a/runtime/flang/norm2.h
+++ b/runtime/flang/norm2.h
@@ -11,16 +11,10 @@
 #include "stdioInterf.h"
 #include "fioMacros.h"
 
-#if defined(__AVX2__) && defined(DESC_I8)
-#define NORM2_REAL4 norm2_avx2_real4_i8_
-#define NORM2_REAL8 norm2_avx2_real8_i8_
-#elif defined(__AVX2__) // implies !defined(DESC_I8)
-#define NORM2_REAL4 norm2_avx2_real4_
-#define NORM2_REAL8 norm2_avx2_real8_
-#elif defined(DESC_I8) // implies !defined(__AVX2__)
+#if defined(DESC_I8)
 #define NORM2_REAL4 norm2_real4_i8_
 #define NORM2_REAL8 norm2_real8_i8_
-#else // implies !defined(__AVX2__) && !defined(DESC_I8)
+#else // !defined(DESC_I8)
 #define NORM2_REAL4 norm2_real4_
 #define NORM2_REAL8 norm2_real8_
 #endif
