diff --git a/runtime/flang/CMakeLists.txt b/runtime/flang/CMakeLists.txt
index 448b7173..5c109985 100644
--- a/runtime/flang/CMakeLists.txt
+++ b/runtime/flang/CMakeLists.txt
@@ -414,7 +415,7 @@ separate_arguments(SEPARATED_CMAKE_Fortran_FLAGS NATIVE_COMMAND ${CMAKE_Fortran_
 # consider macros which results in wrong dependencies.
 add_custom_command(
   OUTPUT "${I8_FILES_DIR}/ieee_arithmetic.F95"
-  COMMAND "${CMAKE_Fortran_COMPILER}" -E -cpp ${SEPARATED_CMAKE_Fortran_FLAGS}
+  COMMAND "${CMAKE_Fortran_COMPILER}" -E -fsyntax-only -cpp ${SEPARATED_CMAKE_Fortran_FLAGS}
   "${CMAKE_CURRENT_SOURCE_DIR}/ieee_arithmetic.F95" -DDESC_I8 
   > "${I8_FILES_DIR}/ieee_arithmetic.F95"
   COMMENT "Preprocessing ieee_arithmetic.F95"
@@ -424,7 +425,7 @@ add_custom_command(
 
 add_custom_command(
   OUTPUT "${I8_FILES_DIR}/ieee_exceptions.F95"
-  COMMAND "${CMAKE_Fortran_COMPILER}" -E -cpp ${SEPARATED_CMAKE_Fortran_FLAGS}
+  COMMAND "${CMAKE_Fortran_COMPILER}" -E -fsyntax-only -cpp ${SEPARATED_CMAKE_Fortran_FLAGS}
   "${CMAKE_CURRENT_SOURCE_DIR}/ieee_exceptions.F95" -DDESC_I8 
   > "${I8_FILES_DIR}/ieee_exceptions.F95"
   COMMENT "Preprocessing ieee_exceptions.F95"
@@ -434,9 +435,9 @@ add_custom_command(
 
 add_custom_command(
   OUTPUT ${I8_FILES_DIR}/norm2_1.F95 norm2_1.F95
-  COMMAND "${CMAKE_Fortran_COMPILER}" -E -DDESC_I8 -DPGFLANG -cpp ${SEPARATED_CMAKE_Fortran_FLAGS}
+  COMMAND "${CMAKE_Fortran_COMPILER}" -E -fsyntax-only -DDESC_I8 -DPGFLANG -cpp ${SEPARATED_CMAKE_Fortran_FLAGS}
   ${CMAKE_CURRENT_SOURCE_DIR}/norm2.F95 > "${I8_FILES_DIR}/norm2_1.F95"
-  COMMAND "${CMAKE_Fortran_COMPILER}" -E -DPGFLANG -cpp ${SEPARATED_CMAKE_Fortran_FLAGS}
+  COMMAND "${CMAKE_Fortran_COMPILER}" -E -fsyntax-only -DPGFLANG -cpp ${SEPARATED_CMAKE_Fortran_FLAGS}
   ${CMAKE_CURRENT_SOURCE_DIR}/norm2.F95 > "norm2_1.F95"
   VERBATIM
   DEPENDS flang1 flang2
