diff --git a/tools/flang2/flang2exe/ll_write.cpp b/tools/flang2/flang2exe/ll_write.cpp
index 69ea30b..5e4716b 100644
--- a/tools/flang2/flang2exe/ll_write.cpp
+++ b/tools/flang2/flang2exe/ll_write.cpp
@@ -963,7 +963,7 @@ static const MDTemplate Tmpl_DISubprogram_39[] = {
     {"templateParams", NodeField},
     {"declaration", NodeField},
     {"unit", NodeField},
-    {"variables", NodeField},
+    {"retainedNodes", NodeField},
     {"scopeLine", UnsignedField}};
 
 static const MDTemplate Tmpl_DILexicalBlock[] = {
