diff --git a/runtime/libpgmath/CMakeLists.txt b/runtime/libpgmath/CMakeLists.txt
index fe2d4ee1..ca3d367f 100644
--- a/runtime/libpgmath/CMakeLists.txt
+++ b/runtime/libpgmath/CMakeLists.txt
@@ -25,12 +25,8 @@ set(LIBPGMATH_WITH_GENERIC FALSE CACHE BOOL "Build using generic?")
 set(PACKAGE_NAME libpgmath)
 
 # Standalone build or part of LLVM?
-set(LIBPGMATH_STANDALONE_BUILD FALSE)
-if (CMAKE_SOURCE_DIR STREQUAL CMAKE_CURRENT_SOURCE_DIR)
-  project(pgmath C CXX ASM)
-
-  set(LIBPGMATH_STANDALONE_BUILD TRUE)
-endif()
+project(pgmath C CXX ASM)
+set(LIBPGMATH_STANDALONE_BUILD TRUE)
 
 if (LIBPGMATH_STANDALONE_BUILD)
   include(FindPythonInterp)
diff --git a/runtime/CMakeLists.txt b/runtime/CMakeLists.txt
index 33dc89b8..5f78b7f0 100644
--- a/runtime/CMakeLists.txt
+++ b/runtime/CMakeLists.txt
@@ -44,3 +44,4 @@ add_subdirectory(ompstub)
 add_subdirectory(flangrti)
 add_subdirectory(flang)
 add_subdirectory(flangmain)
+add_subdirectory(libpgmath)
