--- a/modules/codec/webvtt/CSSGrammar.c	2018-02-14 11:53:55.023195112 +0100
+++ b/modules/codec/webvtt/CSSGrammar.c	2018-02-14 11:54:22.197125950 +0100
@@ -315,8 +315,7 @@
 /* Suppress an incorrect diagnostic about yylval being uninitialized.  */
 # define YY_IGNORE_MAYBE_UNINITIALIZED_BEGIN \
     _Pragma ("GCC diagnostic push") \
-    _Pragma ("GCC diagnostic ignored \"-Wuninitialized\"")\
-    _Pragma ("GCC diagnostic ignored \"-Wmaybe-uninitialized\"")
+    _Pragma ("GCC diagnostic ignored \"-Wuninitialized\"")
 # define YY_IGNORE_MAYBE_UNINITIALIZED_END \
     _Pragma ("GCC diagnostic pop")
 #else
