--- a/modules/Makefile.am	2016-03-24 12:45:42.000000000 +0100
+++ b/modules/Makefile.am	2016-07-17 13:41:41.786484760 +0200
@@ -7,7 +7,7 @@
 	pam_group pam_issue pam_keyinit pam_lastlog pam_limits \
 	pam_listfile pam_localuser pam_loginuid pam_mail \
 	pam_mkhomedir pam_motd pam_namespace pam_nologin \
-	pam_permit pam_pwhistory pam_rhosts pam_rootok pam_securetty \
+	pam_permit pam_pwhistory pam_rootok pam_securetty \
 	pam_selinux pam_sepermit pam_shells pam_stress \
 	pam_succeed_if pam_tally pam_tally2 pam_time pam_timestamp \
 	pam_tty_audit pam_umask \
