--- a/src/HolSat/sat_solvers/minisat/Main.C	2018-07-21 11:03:03.227964547 +0200
+++ b/src/HolSat/sat_solvers/minisat/Main.C	2018-07-21 11:03:46.594097689 +0200
@@ -72,11 +72,11 @@
 {
     cpu_time = cpuTime();
     mem_used = memUsed();
-    reportf("restarts              : %"I64_fmt"\n", stats.starts);
-    reportf("conflicts             : %-12"I64_fmt"   (%.0f /sec)\n", stats.conflicts   , stats.conflicts   /cpu_time);
-    reportf("decisions             : %-12"I64_fmt"   (%.0f /sec)\n", stats.decisions   , stats.decisions   /cpu_time);
-    reportf("propagations          : %-12"I64_fmt"   (%.0f /sec)\n", stats.propagations, stats.propagations/cpu_time);
-    reportf("conflict literals     : %-12"I64_fmt"   (%4.2f %% deleted)\n", stats.tot_literals, (stats.max_literals - stats.tot_literals)*100 / (double)stats.max_literals);
+    reportf("restarts              : %" I64_fmt "\n", stats.starts);
+    reportf("conflicts             : %-12" I64_fmt "   (%.0f /sec)\n", stats.conflicts   , stats.conflicts   /cpu_time);
+    reportf("decisions             : %-12" I64_fmt "   (%.0f /sec)\n", stats.decisions   , stats.decisions   /cpu_time);
+    reportf("propagations          : %-12" I64_fmt "   (%.0f /sec)\n", stats.propagations, stats.propagations/cpu_time);
+    reportf("conflict literals     : %-12" I64_fmt "   (%4.2f %% deleted)\n", stats.tot_literals, (stats.max_literals - stats.tot_literals)*100 / (double)stats.max_literals);
     if (mem_used != 0) reportf("Memory used           : %.2f MB\n", mem_used / 1048576.0);
     reportf("CPU time              : %g s\n", cpu_time);
 }
--- a/src/HolSat/sat_solvers/minisat/Solver.C	2018-07-21 11:02:25.421712403 +0200
+++ b/src/HolSat/sat_solvers/minisat/Solver.C	2018-07-21 11:02:42.755370469 +0200
@@ -832,7 +832,7 @@
         if (var(conflict[i]) >= 100){
             Lit p = conflict[i];
             Var x = var(p);
-            printf("confl[%d] = "L_LIT"\n", i, L_lit(p));
+            printf("confl[%d] = " L_LIT "\n", i, L_lit(p));
             printf("level     = %d\n", level[x]);
             printf("reason    = %p\n", reason[x]);
             printf("unit_id   = %d\n", unit_id[x]);
