--- tools-poly/configure.sml.bak	2018-07-08 17:49:41.573998734 +0200
+++ tools-poly/configure.sml	2018-07-08 17:50:01.827963112 +0200
@@ -215,7 +215,7 @@
          case pkgconfig_info of
              SOME list => list
            | _ => ["-L" ^ polymllibdir, "-lpolymain", "-lpolyml", "-lpthread",
-                   "-lm", "-ldl", "-lstdc++", "-lgcc_s", "-lgcc"]
+                   "-lm", "-ldl", "-lstdc++", "-lgcc_s", "-lgcc" , "-Wl,-nopie"]
        else []
 end;
 
