--- a/src/systemd/src/basic/util.h
+++ b/src/systemd/src/basic/util.h
@@ -94,7 +94,7 @@ void *xbsearch_r(const void *key, const void *base, size_t nmemb, size_t size,
  * Normal qsort requires base to be nonnull. Here were require
  * that only if nmemb > 0.
  */
-static inline void qsort_safe(void *base, size_t nmemb, size_t size, comparison_fn_t compar) {
+static inline void qsort_safe(void *base, size_t nmemb, size_t size, int (*compar)(const void *, const void *)) {
         if (nmemb <= 1)
                 return;
 
