--- a/configure.ac	2025-08-25 01:03:20.000000000 +0200
+++ b/configure.ac	2025-09-27 14:28:51.197366976 +0200
@@ -1250,7 +1250,7 @@
 	       select poll \
 	       readlink faccessx fchdir ftruncate \
 	       fstat lstat lchown fchown fchmod \
-	       fpurge fseeko ftello \
+	       fseeko ftello \
 	       mkfifo _mktemp mkstemp \
 	       waitpid wait3 \
 	       sigqueue \
@@ -1292,6 +1292,16 @@
 	       setutxent getutxent endutxent getutent)
 AC_FUNC_STRCOLL
 
+AH_TEMPLATE([HAVE_FPURGE],
+  [Define to 1 if you have the `fpurge' function.])
+AC_MSG_CHECKING([for fpurge])
+AC_LINK_IFELSE([AC_LANG_SOURCE(
+[[#include <stdio.h>
+int main () { return fpurge(stdout); }]])],
+  [AC_MSG_RESULT([yes])
+   AC_DEFINE([HAVE_FPURGE])],
+  [AC_MSG_RESULT([no])])
+
 # isinf() and isnan() can exist as either functions or macros.
 AH_TEMPLATE([HAVE_ISINF],
   [Define to 1 if you have the `isinf' macro or function.])
