diff --git a/GhcMod/Exe/CaseSplit.hs b/GhcMod/Exe/CaseSplit.hs
index 98f76bf..c6473fd 100644
--- a/GhcMod/Exe/CaseSplit.hs
+++ b/GhcMod/Exe/CaseSplit.hs
@@ -127,11 +127,17 @@ getSrcSpanTypeForFnSplit tcm@G.TypecheckedModule{G.tm_typechecked_source = tcs}
           _ -> return Nothing
 
 isPatternVar :: G.LPat Gap.GhcTc -> Bool
+#if __GLASGOW_HASKELL__ >= 808
+isPatternVar (G.dL->G.L _ (G.VarPat {})) = True
+#else
 isPatternVar (G.L _ (G.VarPat {})) = True
+#endif
 isPatternVar _                     = False
 
 getPatternVarName :: G.LPat Gap.GhcTc -> G.Name
-#if __GLASGOW_HASKELL__ >= 806
+#if __GLASGOW_HASKELL__ >= 808
+getPatternVarName (G.dL->G.L _ (G.VarPat _ (G.L _ vName))) = G.getName vName
+#elif __GLASGOW_HASKELL__ >= 806
 getPatternVarName (G.L _ (G.VarPat _ (G.L _ vName))) = G.getName vName
 #elif __GLASGOW_HASKELL__ >= 800
 getPatternVarName (G.L _ (G.VarPat (G.L _ vName))) = G.getName vName
