--- a/modules.d/90lvm/module-setup.sh	2017-04-03 10:26:25.000000000 +0200
+++ b/modules.d/90lvm/module-setup.sh	2017-11-12 16:16:19.442043700 +0100
@@ -83,6 +83,10 @@
         unset LVM_SUPPRESS_FD_WARNINGS
     fi
 
+    # Also install llvm-profile
+    profiledir="/etc/lvm/profile"
+    $DRACUT_CP -r --parents "${profiledir}" "${initdir}"
+
     if ! [[ -e ${initdir}/etc/lvm/lvm.conf ]]; then
         mkdir -p "${initdir}/etc/lvm"
         {
