--- a/libcxxabi/src/CMakeLists.txt	2023-06-27 18:56:38.198918499 +0200
+++ b/libcxxabi/src/CMakeLists.txt	2023-06-27 18:59:52.687171034 +0200
@@ -277,6 +277,10 @@
       _LIBCPP_DISABLE_VISIBILITY_ANNOTATIONS=)
 endif()
 
+if (CMAKE_POSITION_INDEPENDENT_CODE OR NOT DEFINED CMAKE_POSITION_INDEPENDENT_CODE)
+  set_target_properties(cxxabi_static_objects PROPERTIES POSITION_INDEPENDENT_CODE ON) # must set manually because it's an object library
+endif()
+
 if (LIBCXXABI_ENABLE_STATIC)
   add_library(cxxabi_static STATIC)
   if (LIBCXXABI_USE_LLVM_UNWINDER AND NOT LIBCXXABI_STATICALLY_LINK_UNWINDER_IN_STATIC_LIBRARY)
