--- a/configure	2024-04-13 15:13:24.465294759 +0200
+++ b/configure	2024-04-13 15:14:17.877955548 +0200
@@ -3898,7 +3898,7 @@
 if test "`cd $srcdir && pwd`" != "`pwd`"; then
   # Use -I$(srcdir) only when $(srcdir) != ., so that make's output
   # is not polluted with repeated "-I."
-  am__isrc=' -I$(srcdir)'
+  am__isrc='$(srcdir)'
   # test to see if srcdir already configured
   if test -f $srcdir/config.status; then
     as_fn_error $? "source directory already configured; run \"make distclean\" there first" "$LINENO" 5
--- a/src/Makefile.in	2024-04-13 15:13:34.926424274 +0200
+++ b/src/Makefile.in	2024-04-13 15:13:48.275589477 +0200
@@ -497,7 +497,7 @@
 top_srcdir = @top_srcdir@
 zip_LIBS = @zip_LIBS@
 zstd_LIBS = @zstd_LIBS@
-AM_CPPFLAGS = -I. -I$(srcdir) -I$(top_srcdir)/lib -I.. \
+AM_CPPFLAGS = -I$(srcdir) -I$(top_srcdir)/lib -I.. \
 	-I$(srcdir)/../libelf -I$(srcdir)/../libebl \
 	-I$(srcdir)/../libdw -I$(srcdir)/../libdwelf \
 	-I$(srcdir)/../libdwfl -I$(srcdir)/../libasm -I../debuginfod
