--- a/libelf/elf.h	2016-12-28 10:56:21.585727762 -0800
+++ b/libelf/elf.h	2016-12-28 11:32:13.819801460 -0800
@@ -21,7 +21,9 @@
 
 #include <features.h>
 
-__BEGIN_DECLS
+#ifdef __cplusplus
+extern "C" {
+#endif
 
 /* Standard ELF types.  */
 
@@ -3682,6 +3684,7 @@
 #define R_BPF_NONE		0	/* No reloc */
 #define R_BPF_MAP_FD		1	/* Map fd to pointer */
 
-__END_DECLS
-
+#ifdef __cplusplus
+}
+#endif
 #endif	/* elf.h */
