--- a/lib/fixedsizehash.h	2020-08-05 11:12:01.433358290 +0200
+++ b/lib/fixedsizehash.h	2020-08-05 11:12:06.277399672 +0200
@@ -30,7 +30,6 @@
 #include <errno.h>
 #include <stdlib.h>
 #include <string.h>
-#include <sys/cdefs.h>
 
 #include <system.h>
 
--- a/lib/system.h	2020-08-05 11:26:47.922598610 +0200
+++ b/lib/system.h	2020-08-05 11:26:52.977637856 +0200
@@ -30,7 +30,6 @@
 #define LIB_SYSTEM_H	1
 
 #include <errno.h>
-#include <error.h>
 #include <stddef.h>
 #include <stdint.h>
 #include <sys/param.h>
