--- a/src/utils/platform_unix.cpp	2026-08-29 12:16:58.676749038 +0200
+++ b/src/utils/platform_unix.cpp	2026-08-29 12:17:27.748860822 +0200
@@ -10,6 +10,8 @@
 #endif
 
 #include "utils/platform.h"
+
+#include <cstdlib>
 #include <unistd.h>
 
 //extern char **environ; // unistd.h should have this line
--- src/debuginfo/debugsources.cpp.bak	2026-08-29 12:15:59.676522767 +0200
+++ src/debuginfo/debugsources.cpp	2026-08-29 12:16:25.252620752 +0200
@@ -6,6 +6,8 @@
 #include "debuginfo/debugsources.h"
 #include "utils/hresult.h"
 #include "utils/logger.h"
+
+#include <algorithm>
 #include <list>
 #include <map>
 #include <set>
