--- a/third_party/diagnostics/src/shared/pal/src/include/pal/list.h	2026-08-29 12:19:45.189391539 +0200
+++ b/third_party/diagnostics/src/shared/pal/src/include/pal/list.h	2026-08-29 12:19:57.337438607 +0200
@@ -22,13 +22,13 @@
 #ifndef _LIST_H_INCLUDED
 #define _LIST_H_INCLUDED
 
+#include <cstddef>
+
 #ifdef __cplusplus
 extern "C"
 {
 #endif // __cplusplus
 
-#include <cstddef>
-
 typedef struct _LIST_ENTRY {
    struct _LIST_ENTRY *Flink;
    struct _LIST_ENTRY *Blink;
