--- a/gtk/gtkiconhelper.c
+++ b/gtk/gtkiconhelper.c
@@ -896,7 +896,14 @@ _gtk_icon_helper_draw (GtkIconHelper *self,
 
   if (self->priv->rendered_surface != NULL)
     {
-      gtk_css_style_render_icon_surface (gtk_css_node_get_style (gtk_css_gadget_get_node (GTK_CSS_GADGET (self))),
+      GtkCssNode *node;
+
+      /* Avoid calling gtk_css_node_get_style() because it can
+       * invalidate the style which we don't want while trying
+       * to render the current style.
+       */
+      node = gtk_css_gadget_get_node (GTK_CSS_GADGET (self));
+      gtk_css_style_render_icon_surface (node->style,
                                          cr,
                                          self->priv->rendered_surface,
                                          x, y);
--- a/gtk/gtkrender.c
+++ b/gtk/gtkrender.c
@@ -1153,7 +1153,14 @@
   g_return_if_fail (GTK_IS_STYLE_CONTEXT (context));
   g_return_if_fail (cr != NULL);
 
-  gtk_css_style_render_icon_surface (gtk_style_context_lookup_style (context),
+  GtkCssNode *node;
+
+  /* Avoid calling gtk_style_context_lookup_style() because it can
+   * invalidate the style which we don't want while trying
+   * to render the current style.
+  */
+  node = gtk_style_context_get_node(context);
+  gtk_css_style_render_icon_surface (node->style,
                                      cr,
                                      surface,
                                      x, y); 
