diff --git a/Kvantum/MateriaLight/MateriaLight.kvconfig b/Kvantum/MateriaLight/MateriaLight.kvconfig
index b3588c0..9608b9d 100644
--- a/Kvantum/MateriaLight/MateriaLight.kvconfig
+++ b/Kvantum/MateriaLight/MateriaLight.kvconfig
@@ -67,7 +67,7 @@ reduce_menu_opacity=0
 
 [GeneralColors]
 window.color=#f2f2f2
-base.color=#ffffff
+base.color=#f2f2f2
 alt.base.color=#ffffff
 button.color=#ffffff
 light.color=#ffffff
@@ -131,6 +131,7 @@ text.margin.bottom=3
 text.margin.left=2
 text.margin.right=2
 frame.expansion=0
+focusFrame=true
 
 [PanelButtonTool]
 inherits=PanelButtonCommand
