--- a/opt/visual-studio-code/bin/code	2016-11-21 21:02:07.000000000 +0100
+++ b/opt/visual-studio-code/bin/code	2016-12-01 00:37:51.034074495 +0100
@@ -31,7 +31,7 @@
 	fi
 fi
 
-ELECTRON="$VSCODE_PATH/code"
-CLI="$VSCODE_PATH/resources/app/out/cli.js"
-ELECTRON_RUN_AS_NODE=1 "$ELECTRON" "$CLI" "$@"
-exit $?
\ No newline at end of file
+ELECTRON="/usr/bin/electron"
+CLI="$VSCODE_PATH/app"
+"$ELECTRON" "$CLI" "$@"
+exit $?
