--- a/cmake/Modules/HandleCompilerRT.cmake	2018-11-14 22:56:42.265058281 +0100
+++ b/cmake/Modules/HandleCompilerRT.cmake	2018-11-14 22:57:25.720002785 +0100
@@ -3,7 +3,7 @@
     message(FATAL_ERROR "LIBUNWIND_COMPILE_FLAGS must be defined when using this function")
   endif()
   set(dest "" PARENT_SCOPE)
-  set(CLANG_COMMAND ${CMAKE_CXX_COMPILER} ${LIBUNWIND_COMPILE_FLAGS}
+  set(CLANG_COMMAND ${CMAKE_CXX_COMPILER} ${CMAKE_CXX_COMPILER_ARG1} ${LIBUNWIND_COMPILE_FLAGS}
       "--rtlib=compiler-rt" "--print-libgcc-file-name")
   if (CMAKE_CXX_COMPILER_ID MATCHES Clang AND CMAKE_CXX_COMPILER_TARGET)
     list(APPEND CLANG_COMMAND "--target=${CMAKE_CXX_COMPILER_TARGET}")
@@ -30,7 +30,7 @@
   endif()
   set(dest "" PARENT_SCOPE)
   if (APPLE)
-    set(CLANG_COMMAND ${CMAKE_CXX_COMPILER} ${LIBUNWIND_COMPILE_FLAGS}
+    set(CLANG_COMMAND ${CMAKE_CXX_COMPILER} ${CMAKE_CXX_COMPILER_ARG1} ${LIBUNWIND_COMPILE_FLAGS}
         "-print-file-name=lib")
     execute_process(
         COMMAND ${CLANG_COMMAND}
