diff --git a/cmake/config-ix.cmake b/cmake/config-ix.cmake
index 40a5dd1..c32b43b 100644
--- a/cmake/config-ix.cmake
+++ b/cmake/config-ix.cmake
@@ -24,6 +24,7 @@ if (LIBUNWIND_HAS_NODEFAULTLIBS_FLAG)
   endif ()
   if (LIBUNWIND_USE_COMPILER_RT)
     list(APPEND CMAKE_REQUIRED_FLAGS -rtlib=compiler-rt)
+    list(APPEND CMAKE_REQUIRED_FLAGS unwind)
     find_compiler_rt_library(builtins LIBUNWIND_BUILTINS_LIBRARY)
     list(APPEND CMAKE_REQUIRED_LIBRARIES "${LIBUNWIND_BUILTINS_LIBRARY}")
   elseif (LIBUNWIND_HAS_GCC_S_LIB)
