--- a/ELF/InputFiles.cpp	2016-10-16 21:11:38.801434288 +0200
+++ b/ELF/InputFiles.cpp	2016-10-16 21:14:22.001420860 +0200
@@ -313,7 +313,7 @@
     return &InputSection<ELFT>::Discarded;
   }
 
-  if (Config->Strip != StripPolicy::None && Name.startswith(".debug"))
+  if (Config->Strip != StripPolicy::None && (Name.startswith(".debug") || Name.startswith(".comment")))
     return &InputSection<ELFT>::Discarded;
 
   // The linker merges EH (exception handling) frames and creates a
--- a/ELF/Writer.cpp
+++ b/ELF/Writer.cpp
@@ -279,7 +279,7 @@ template <class ELFT> void Writer<ELFT>::createSyntheticSections() {
     In<ELFT>::Interp = nullptr;
   }
 
-  if (!Config->Relocatable)
+  if (!Config->Relocatable && Config->Strip != StripPolicy::None)
     Symtab<ELFT>::X->Sections.push_back(createCommentSection<ELFT>());
 
   if (Config->Strip != StripPolicy::All) {

--- a/ELF/Options.td	2016-11-07 16:50:58.392900038 +0100
+++ b/ELF/Options.td	2016-11-07 16:52:37.554769701 +0100
@@ -145,6 +145,8 @@
 
 def strip_all: F<"strip-all">, HelpText<"Strip all symbols">;
 
+def nostrip: F<"nostrip">, HelpText<"Do not strip symbols">;
+
 def strip_debug: F<"strip-debug">, HelpText<"Strip debugging information">;
 
 def sysroot: J<"sysroot=">, HelpText<"Set the system root">;
--- a/ELF/Driver.cpp	2016-11-07 16:49:26.581078380 +0100
+++ b/ELF/Driver.cpp	2016-11-07 16:50:39.766389361 +0100
@@ -403,10 +403,7 @@
 }
 
 static StripPolicy getStripOption(opt::InputArgList &Args) {
-  if (auto *Arg = Args.getLastArg(OPT_strip_all, OPT_strip_debug)) {
-    if (Arg->getOption().getID() == OPT_strip_all)
-      return StripPolicy::All;
-    return StripPolicy::Debug;
-  }
+  if (!Args.hasArg(OPT_nostrip))
+        return StripPolicy::All;
   return StripPolicy::None;
 }
 
