--- a/ELF/Driver.cpp  2017-07-19 19:28:24.724889738 +0200
+++ b/ELF/Driver.cpp  2017-07-19 19:33:29.841364526 +0200
@@ -505,11 +505,11 @@
   if (Args.hasArg(OPT_relocatable))
     return StripPolicy::None;
 
-  auto *Arg = Args.getLastArg(OPT_strip_all, OPT_strip_debug);
+  auto *Arg = Args.getLastArg(OPT_strip_none, OPT_strip_debug);
   if (!Arg)
-    return StripPolicy::None;
-  if (Arg->getOption().getID() == OPT_strip_all)
     return StripPolicy::All;
+  if (Arg->getOption().getID() == OPT_strip_none)
+    return StripPolicy::None;
   return StripPolicy::Debug;
 }
 
--- a/ELF/InputFiles.cpp	2016-10-16 21:11:38.801434288 +0200
+++ b/ELF/InputFiles.cpp	2016-10-16 21:14:22.001420860 +0200
@@ -313,7 +313,7 @@
     return &InputSection::Discarded;
   }
 
-  if (Config->Strip != StripPolicy::None && Name.startswith(".debug"))
+  if (Config->Strip != StripPolicy::None && (Name.startswith(".debug") || Name.startswith(".comment")))
     return &InputSection::Discarded;
 
   // If -gdb-index is given, LLD creates .gdb_index section, and that
--- a/ELF/Options.td  2017-07-19 18:01:49.205823604 +0200
+++ b/ELF/Options.td  2017-07-19 19:33:55.657320165 +0200
@@ -230,7 +230,7 @@
 def start_lib: F<"start-lib">,
   HelpText<"Start a grouping of objects that should be treated as if they were together in an archive">;
 
-def strip_all: F<"strip-all">, HelpText<"Strip all symbols">;
+def strip_none: F<"strip-none">, HelpText<"Strip no symbols">;
 
 def strip_debug: F<"strip-debug">, HelpText<"Strip debugging information">;
 
@@ -330,7 +330,7 @@
 def alias_soname_soname: S<"soname">, Alias<soname>;
 def alias_sort_section: J<"sort-section=">, Alias<sort_section>;
 def alias_script: J<"script=">, Alias<script>;
-def alias_strip_all: Flag<["-"], "s">, Alias<strip_all>;
+def alias_strip_all: Flag<["-"], "nostrip">, Alias<strip_none>;
 def alias_strip_debug_S: Flag<["-"], "S">, Alias<strip_debug>;
 def alias_Tbss: J<"Tbss=">, Alias<Tbss>;
 def alias_Tdata: J<"Tdata=">, Alias<Tdata>;
