diff --git a/ELF/LinkerScript.cpp b/ELF/LinkerScript.cpp
index e25175d..f8719f7 100644
--- a/ELF/LinkerScript.cpp
+++ b/ELF/LinkerScript.cpp
@@ -1791,19 +1791,15 @@ void ScriptParser::readVersionDeclaration(StringRef VerStr) {
 }
 
 void ScriptParser::readLocal(StringRef VerStr) {
-  if (consume("*")) {
-    Config->DefaultSymbolVersion = VER_NDX_LOCAL;
-    expect(";");
-    return;
-  }
-
+  std::vector<SymbolVersion> *Locals;
   if (VerStr.empty())
-    setError("locals list for anonymous version is not supported");
+    Locals = &Config->VersionScriptLocals;
+  else
+    Locals = &Config->VersionDefinitions.back().Locals;
 
-  std::vector<SymbolVersion> &Locals = Config->VersionDefinitions.back().Locals;
   while (!Error && peek() != "}") {
     StringRef Tok = next();
-    Locals.push_back({unquote(Tok), false, hasWildcard(Tok)});
+    Locals->push_back({unquote(Tok), false, hasWildcard(Tok)});
     expect(";");
   }
 }
diff --git a/ELF/Config.h b/ELF/Config.h
index a46a8a8..5157bbf 100644
--- a/ELF/Config.h
+++ b/ELF/Config.h
@@ -92,6 +92,7 @@ struct Configuration {
   std::vector<llvm::StringRef> SearchPaths;
   std::vector<llvm::StringRef> Undefined;
   std::vector<SymbolVersion> VersionScriptGlobals;
+  std::vector<SymbolVersion> VersionScriptLocals;
   std::vector<uint8_t> BuildIdVector;
   bool AllowMultipleDefinition;
   bool AsNeeded = false;
diff --git a/ELF/Symbols.h b/ELF/Symbols.h
index 192433e..aec12e9 100644
--- a/ELF/Symbols.h
+++ b/ELF/Symbols.h
@@ -406,6 +406,9 @@ struct Symbol {
   // Version definition index.
   uint16_t VersionId;
 
+  // If a symbol has been processed.
+  bool VersionScriptProcessed = false;
+
   // Symbol visibility. This is the computed minimum visibility of all
   // observed non-DSO symbols.
   unsigned Visibility : 2;
diff --git a/ELF/SymbolTable.cpp b/ELF/SymbolTable.cpp
index efbcba9..dfc5f09 100644
--- a/ELF/SymbolTable.cpp
+++ b/ELF/SymbolTable.cpp
@@ -648,22 +648,30 @@ SymbolTable<ELFT>::findAllDemangled(const StringMatcher &M) {
 // but just specifies symbols visibilities. We assume that the script was
 // in the form of { global: foo; bar; local *; }. So, local is default.
 // In this function, we make specified symbols global.
-template <class ELFT> void SymbolTable<ELFT>::handleAnonymousVersion() {
+template <class ELFT> void SymbolTable<ELFT>::handleAnonymousVersion(
+            std::vector<SymbolVersion> &symbolversions, size_t Version) {
   std::vector<StringRef> Patterns;
-  for (SymbolVersion &Ver : Config->VersionScriptGlobals) {
+  for (SymbolVersion &Ver : symbolversions) {
     if (hasWildcard(Ver.Name)) {
       Patterns.push_back(Ver.Name);
       continue;
     }
-    if (SymbolBody *B = find(Ver.Name))
-      B->symbol()->VersionId = VER_NDX_GLOBAL;
+    if (SymbolBody *B = find(Ver.Name)) {
+      B->symbol()->VersionId = Version;
+      B->symbol()->VersionScriptProcessed = true;
+    }
   }
+
   if (Patterns.empty())
     return;
   StringMatcher M(Patterns);
   std::vector<SymbolBody *> Syms = findAll(M);
-  for (SymbolBody *B : Syms)
-    B->symbol()->VersionId = VER_NDX_GLOBAL;
+  for (SymbolBody *B : Syms) {
+    if (Version == VER_NDX_GLOBAL)
+      B->symbol()->VersionScriptProcessed = true;
+    else if (!B->symbol()->VersionScriptProcessed)
+      B->symbol()->VersionId = Version;
+  }
 }
 
 template <class ELFT>
@@ -687,10 +695,16 @@ void SymbolTable<ELFT>::assignWildcardVersion(SymbolVersion Ver,
 // member of symbols.
 template <class ELFT> void SymbolTable<ELFT>::scanVersionScript() {
   // Handle edge cases first.
-  if (!Config->VersionScriptGlobals.empty()) {
-    handleAnonymousVersion();
+  bool gresult = !Config->VersionScriptGlobals.empty();
+  bool lresult = !Config->VersionScriptLocals.empty();
+
+  if (lresult)
+    handleAnonymousVersion(Config->VersionScriptGlobals, VER_NDX_GLOBAL);
+  if (gresult)
+    handleAnonymousVersion(Config->VersionScriptLocals, VER_NDX_LOCAL);
+
+  if (gresult || lresult)
     return;
-  }
 
   if (Config->VersionDefinitions.empty())
     return;
@@ -727,10 +741,10 @@ template <class ELFT> void SymbolTable<ELFT>::scanVersionScript() {
   // we iterate over the definitions in the reverse order.
   for (size_t I = Config->VersionDefinitions.size() - 1; I != (size_t)-1; --I) {
     VersionDefinition &V = Config->VersionDefinitions[I];
-    for (SymbolVersion &Ver : V.Locals)
-      assignWildcardVersion(Ver, VER_NDX_LOCAL);
     for (SymbolVersion &Ver : V.Globals)
       assignWildcardVersion(Ver, V.Id);
+    for (SymbolVersion &Ver : V.Locals)
+      assignWildcardVersion(Ver, VER_NDX_LOCAL);
   }
 }
 
diff --git a/ELF/SymbolTable.h b/ELF/SymbolTable.h
index d78287c..dc139ba 100644
--- a/ELF/SymbolTable.h
+++ b/ELF/SymbolTable.h
@@ -102,7 +102,7 @@ private:
   std::vector<SymbolBody *> findAllDemangled(const StringMatcher &M);
 
   void initDemangledSyms();
-  void handleAnonymousVersion();
+  void handleAnonymousVersion(std::vector<SymbolVersion> &symbolversions, size_t Version);
   void assignWildcardVersion(SymbolVersion Ver, size_t VersionId);
 
   struct SymIndex {
