--- a/ELF/InputFiles.cpp	2016-11-11 19:48:19.308726806 +0100
+++ b/ELF/InputFiles.cpp	2016-11-11 19:53:59.125750362 +0100
@@ -165,14 +165,17 @@
 }
 
 Optional<MemoryBufferRef> elf::readFile(StringRef Path) {
-  if (Config->Verbose)
-    outs() << Path << "\n";
+  StringRef attempt = "attempt to open ";
 
   auto MBOrErr = MemoryBuffer::getFile(Path);
   if (auto EC = MBOrErr.getError()) {
-    error(EC, "cannot open " + Path);
+    error(EC, attempt + Path + " failed");
     return None;
   }
+  else
+    if (Config->Verbose)
+      outs() << attempt << Path << " succeeded\n";
+
   std::unique_ptr<MemoryBuffer> &MB = *MBOrErr;
   MemoryBufferRef MBRef = MB->getMemBufferRef();
   OwningMBs.push_back(std::move(MB)); // take MB ownership
