--- a/include/lld/ELF/Options.td	2016-10-06 12:37:11.366433675 +0200
+++ b/include/lld/ELF/Options.td	2016-10-06 12:42:05.543448159 +0200
@@ -254,9 +251,11 @@
 def rpath_link: S<"rpath-link">;
 def rpath_link_eq: J<"rpath-link=">;
 def sort_common: F<"sort-common">;
+def stats: Flag<["-"], "stats">;
 def warn_execstack: F<"warn-execstack">;
 def warn_shared_textrel: F<"warn-shared-textrel">;
 def G: Separate<["-"], "G">;
+def Map: Separate<["-"], "Map">;
 
 // Aliases for ignored options
 def alias_define_common_d: Flag<["-"], "d">, Alias<define_common>;
