--- a/ELF/Options.td	2016-10-06 12:37:11.366433675 +0200
+++ b/ELF/Options.td	2016-10-06 12:42:05.543448159 +0200
@@ -295,9 +295,11 @@
   Alias<no_add_needed>;
 def no_dynamic_linker: F<"no-dynamic-linker">;
 def no_fatal_warnings: F<"no-fatal-warnings">;
+def no_keep_memory: F<"no-keep-memory">;
 def no_mmap_output_file: F<"no-mmap-output-file">;
 def no_warn_common: F<"no-warn-common">;
 def no_warn_mismatch: F<"no-warn-mismatch">;
+def retain_symbols_file: Separate<["-"], "retain-symbols-file">;
 def rpath_link: S<"rpath-link">;
 def rpath_link_eq: J<"rpath-link=">;
 def sort_common: F<"sort-common">;
@@ -307,6 +309,7 @@
 def EB : F<"EB">;
 def EL : F<"EL">;
 def G: JoinedOrSeparate<["-"], "G">;
+def Map: Separate<["-"], "Map">;
 def Qy : F<"Qy">;
 
 // Aliases for ignored options
