--- a/ELF/Options.td	2016-10-06 12:37:11.366433675 +0200
+++ b/ELF/Options.td	2016-10-06 12:42:05.543448159 +0200
@@ -332,6 +332,7 @@ def no_copy_dt_needed_entries: F<"no-copy-dt-needed-entries">,
   Alias<no_add_needed>;
 def no_dynamic_linker: F<"no-dynamic-linker">;
 def no_fatal_warnings: F<"no-fatal-warnings">;
+def no_keep_memory: F<"no-keep-memory">;
 def no_mmap_output_file: F<"no-mmap-output-file">;
 def no_warn_common: F<"no-warn-common">;
 def no_warn_mismatch: F<"no-warn-mismatch">;
