diff --git a/ELF/Symbols.cpp b/ELF/Symbols.cpp
index 289bc184..c9643f45 100644
--- a/ELF/Symbols.cpp
+++ b/ELF/Symbols.cpp
@@ -307,9 +307,8 @@ uint8_t Symbol::computeBinding() const {
 }
 
 bool Symbol::includeInDynsym() const {
-  if (computeBinding() == STB_LOCAL)
-    return false;
-  return ExportDynamic || body()->isShared() ||
+  auto cb = computeBinding();
+  return (ExportDynamic && cb != STB_LOCAL) || body()->isShared() ||
          (body()->isUndefined() && Config->Shared);
 }
 
