--- a/include/lld/ELF/Options.td  2016-10-15 20:53:38.996143059 +0200
+++ b/include/lld/ELF/Options.td  2016-10-15 20:54:13.821140193 +0200
@@ -208,7 +208,7 @@
 def alias_o_output: Joined<["--"], "output=">, Alias<o>;
 def alias_pie_pic_executable: F<"pic-executable">, Alias<pie>;
 def alias_relocatable_r: Flag<["-"], "r">, Alias<relocatable>;
-def alias_rpath_R: Joined<["-"], "R">, Alias<rpath>;
+def alias_rpath_R: JoinedOrSeparate<["-"], "R">, Alias<rpath>;
 def alias_rpath_rpath: J<"rpath=">, Alias<rpath>;
 def alias_script_T: JoinedOrSeparate<["-"], "T">, Alias<script>;
 def alias_shared_Bshareable: F<"Bshareable">, Alias<shared>;
