--- a/ELF/InputFiles.cpp	2016-10-16 21:11:38.801434288 +0200
+++ b/ELF/InputFiles.cpp	2016-10-16 21:14:22.001420860 +0200
@@ -313,7 +313,7 @@
     return &InputSection<ELFT>::Discarded;
   }
 
-  if (Config->Strip != StripPolicy::None && Name.startswith(".debug"))
+  if (Config->Strip != StripPolicy::None && (Name.startswith(".debug") || Name.startswith(".comment")))
     return &InputSection<ELFT>::Discarded;
 
   // The linker merges EH (exception handling) frames and creates a
--- a/ELF/Writer.cpp
+++ b/ELF/Writer.cpp
@@ -268,8 +268,8 @@ template <class ELFT> void Writer<ELFT>::createSyntheticSections() {
     Out<ELFT>::VerDef = make<VersionDefinitionSection<ELFT>>();
 
   // Initialize linker generated sections
-  if (!Config->Relocatable)
+  if (!Config->Relocatable && Config->Strip != StripPolicy::None)
     Symtab<ELFT>::X->Sections.push_back(createCommentSection<ELFT>());
 
   if (Config->BuildId == BuildIdKind::Fast)
