reverted:
--- b/cmake/modules/AddLLD.cmake
+++ a/cmake/modules/AddLLD.cmake
@@ -1,5 +1,5 @@
 macro(add_lld_library name)
+  add_llvm_library(${name} ${ARGN})
-  llvm_add_library(${name} ${ARGN})
   set_target_properties(${name} PROPERTIES FOLDER "lld libraries")
 endmacro(add_lld_library)
 
reverted:
--- b/lib/Config/CMakeLists.txt
+++ a/lib/Config/CMakeLists.txt
@@ -4,6 +4,6 @@
   ADDITIONAL_HEADER_DIRS
   ${LLD_INCLUDE_DIR}/lld/Config
 
+  LINK_LIBS
+    LLVMSupport
-  LINK_COMPONENTS
-    Support
   )
reverted:
--- b/lib/Core/CMakeLists.txt
+++ a/lib/Core/CMakeLists.txt
@@ -12,6 +12,6 @@
   ADDITIONAL_HEADER_DIRS
   ${LLD_INCLUDE_DIR}/lld/Core
 
+  LINK_LIBS
+    LLVMSupport
-  LINK_COMPONENTS
-    Support
   )
reverted:
--- b/lib/Driver/CMakeLists.txt
+++ a/lib/Driver/CMakeLists.txt
@@ -8,17 +8,15 @@
   ADDITIONAL_HEADER_DIRS
   ${LLD_INCLUDE_DIR}/lld/Driver
 
-  LINK_COMPONENTS
-    Object
-    Option
-    Support
-
   LINK_LIBS
     lldConfig
     lldMachO
     lldCore
     lldReaderWriter
     lldYAML
+    LLVMObject
+    LLVMOption
+    LLVMSupport
   )
 
 add_dependencies(lldDriver DriverOptionsTableGen)
reverted:
--- b/lib/ReaderWriter/CMakeLists.txt
+++ a/lib/ReaderWriter/CMakeLists.txt
@@ -11,11 +11,9 @@
   ADDITIONAL_HEADER_DIRS
   ${LLD_INCLUDE_DIR}/lld/ReaderWriter
 
-  LINK_COMPONENTS
-    Object
-    Support
-
   LINK_LIBS
     lldCore
     lldYAML
+    LLVMObject
+    LLVMSupport
   )
reverted:
--- b/lib/ReaderWriter/MachO/CMakeLists.txt
+++ a/lib/ReaderWriter/MachO/CMakeLists.txt
@@ -18,16 +18,13 @@
   StubsPass.cpp
   TLVPass.cpp
   WriterMachO.cpp
-
-  LINK_COMPONENTS
-    DebugInfoDWARF
-    Object
-    Support
-    Demangle
-
   LINK_LIBS
     lldCore
     lldYAML
+    LLVMDebugInfoDWARF
+    LLVMObject
+    LLVMSupport
+    LLVMDemangle
     ${PTHREAD_LIB}
   )
 
reverted:
--- b/lib/ReaderWriter/YAML/CMakeLists.txt
+++ a/lib/ReaderWriter/YAML/CMakeLists.txt
@@ -1,9 +1,6 @@
 add_lld_library(lldYAML
   ReaderWriterYAML.cpp
-
-  LINK_COMPONENTS
-    Support
-
   LINK_LIBS
     lldCore
+    LLVMSupport
   )
reverted:
--- b/tools/lld/CMakeLists.txt
+++ a/tools/lld/CMakeLists.txt
@@ -1,7 +1,3 @@
-set(LLVM_LINK_COMPONENTS
-  Support
-  )
-
 add_lld_tool(lld
   lld.cpp
   )
@@ -10,6 +6,7 @@
   lldDriver
   lldCOFF
   lldELF
+  LLVMSupport
   )
 
 install(TARGETS lld
