--- a/mlir/CMakeLists.txt	2023-09-08 15:50:07.726750116 +0200
+++ b/mlir/CMakeLists.txt	2023-09-08 17:56:27.243617964 +0200
@@ -158,8 +158,10 @@
 
 set(CMAKE_INCLUDE_CURRENT_DIR ON)
 
-include_directories( "include")
-include_directories( ${MLIR_INCLUDE_DIR})
+include_directories(BEFORE
+  "include"
+  ${MLIR_INCLUDE_DIR}
+)
 
 # Adding tools/mlir-tblgen here as calling add_tablegen sets some variables like
 # MLIR_TABLEGEN_EXE in PARENT_SCOPE which gets lost if that folder is included
