--- a/ucm2/Qualcomm/x1e80100/Surface12in-HiFi.conf	2026-01-13 08:12:41.000000000 +0100
+++ b/ucm2/Qualcomm/x1e80100/Surface12in-HiFi.conf	2026-04-18 17:49:31.876685799 +0200
@@ -3,8 +3,8 @@
 
 SectionVerb {
 	EnableSequence [
-		cset "name='WSA_CODEC_DMA_RX_0 Audio Mixer MultiMedia2' 1"
-		cset "name='MultiMedia4 Mixer VA_CODEC_DMA_TX_0' 1"
+		cset "name='WSA_CODEC_DMA_RX_0 Audio Mixer MultiMedia1' 1"
+		cset "name='MultiMedia2 Mixer VA_CODEC_DMA_TX_0' 1"
 	]
 
 	Include.wsae.File "/codecs/wsa884x/two-speakers/DefaultEnableSeq.conf"
@@ -25,7 +25,7 @@
 	Value {
 		PlaybackChannels 2
 		PlaybackPriority 100
-		PlaybackPCM "hw:${CardId},1"
+		PlaybackPCM "hw:${CardId},0"
 		PlaybackMixer "default:${CardId}"
 		PlaybackMixerElem "Speakers"
 	}
@@ -41,6 +41,6 @@
 
 	Value {
 		CapturePriority 100
-		CapturePCM "hw:${CardId},3"
+		CapturePCM "hw:${CardId},1"
 	}
 }
