--- a/ucm2/codecs/qcom-lpass/wsa-macro/init.conf	2026-03-26 21:28:36.025397838 +0100
+++ b/ucm2/codecs/qcom-lpass/wsa-macro/init.conf	2026-03-26 21:28:50.673328191 +0100
@@ -1,6 +1,6 @@
 # WSA Digital codec specific volume control settings
 
 BootSequence [
-	cset "name='WSA_RX0 Digital Volume' 84"
-	cset "name='WSA_RX1 Digital Volume' 84"
+	cset "name='WSA WSA_RX0 Digital Volume' 84"
+	cset "name='WSA WSA_RX1 Digital Volume' 84"
 ]
--- a/ucm2/Qualcomm/x1e80100/Surface12in-HiFi.conf	2026-03-26 21:26:52.095315613 +0100
+++ b/ucm2/Qualcomm/x1e80100/Surface12in-HiFi.conf	2026-03-26 21:27:09.442771004 +0100
@@ -3,8 +3,10 @@
 
 SectionVerb {
 	EnableSequence [
-		cset "name='WSA_CODEC_DMA_RX_0 Audio Mixer MultiMedia2' 1"
-		cset "name='MultiMedia4 Mixer VA_CODEC_DMA_TX_0' 1"
+		cset "name='WSA_CODEC_DMA_RX_0 Audio Mixer MultiMedia1' 1"
+		cset "name='DISPLAY_PORT_RX_0 Audio Mixer MultiMedia1' 0"
+		cset "name='DISPLAY_PORT_RX_1 Audio Mixer MultiMedia1' 0"
+		cset "name='MultiMedia2 Mixer VA_CODEC_DMA_TX_0' 1"
 	]
 
 	Include.wsae.File "/codecs/wsa884x/two-speakers/DefaultEnableSeq.conf"
@@ -22,15 +24,80 @@
 	Include.wsmspk1d.File "/codecs/qcom-lpass/wsa-macro/Wsa1SpeakerDisableSeq.conf"
 	Include.wsaspk.File "/codecs/wsa884x/two-speakers/SpeakerSeq.conf"
 
+	EnableSequence [
+		cset "name='DISPLAY_PORT_RX_0 Audio Mixer MultiMedia1' 0"
+		cset "name='DISPLAY_PORT_RX_1 Audio Mixer MultiMedia1' 0"
+		cset "name='WSA_CODEC_DMA_RX_0 Audio Mixer MultiMedia1' 1"
+	]
+
+	DisableSequence [
+		cset "name='WSA_CODEC_DMA_RX_0 Audio Mixer MultiMedia1' 0"
+	]
+
+	ConflictingDevice [
+		"HDMI0"
+		"HDMI1"
+	]
+
 	Value {
 		PlaybackChannels 2
 		PlaybackPriority 100
-		PlaybackPCM "hw:${CardId},1"
+		PlaybackPCM "hw:${CardId},0"
 		PlaybackMixer "default:${CardId}"
 		PlaybackMixerElem "Speakers"
 	}
 }
 
+SectionDevice."HDMI0" {
+	Comment "USB/DisplayPort 0 playback"
+
+	EnableSequence [
+		cset "name='WSA_CODEC_DMA_RX_0 Audio Mixer MultiMedia1' 0"
+		cset "name='DISPLAY_PORT_RX_1 Audio Mixer MultiMedia1' 0"
+		cset "name='DISPLAY_PORT_RX_0 Audio Mixer MultiMedia1' 1"
+	]
+
+	DisableSequence [
+		cset "name='DISPLAY_PORT_RX_0 Audio Mixer MultiMedia1' 0"
+	]
+
+	ConflictingDevice [
+		"Speaker"
+		"HDMI1"
+	]
+
+	Value {
+		PlaybackPriority 200
+		PlaybackPCM "hw:${CardId},0"
+		JackControl "DP0 Jack"
+	}
+}
+
+SectionDevice."HDMI1" {
+	Comment "USB/DisplayPort 1 playback"
+
+	ConflictingDevice [
+		"Speaker"
+		"HDMI0"
+	]
+
+	EnableSequence [
+		cset "name='WSA_CODEC_DMA_RX_0 Audio Mixer MultiMedia1' 0"
+		cset "name='DISPLAY_PORT_RX_0 Audio Mixer MultiMedia1' 0"
+		cset "name='DISPLAY_PORT_RX_1 Audio Mixer MultiMedia1' 1"
+	]
+
+	DisableSequence [
+		cset "name='DISPLAY_PORT_RX_1 Audio Mixer MultiMedia1' 0"
+	]
+
+	Value {
+		PlaybackPriority 200
+		PlaybackPCM "hw:${CardId},0"
+		JackControl "DP1 Jack"
+	}
+}
+
 SectionDevice."Mic" {
 	Comment "Internal microphones"
 
@@ -41,6 +108,6 @@
 
 	Value {
 		CapturePriority 100
-		CapturePCM "hw:${CardId},3"
+		CapturePCM "hw:${CardId},1"
 	}
 }
