--- a/configure
+++ b/configure
@@ -12968,12 +12968,12 @@
 fi
 fi
 if test $text_v = "pango"; then
-	PANGO_LIBS="`pkg-config --libs pangocairo `"
-	PANGO_CFLAGS="`pkg-config --cflags pangocairo `"
+	PANGO_LIBS="`pkg-config --libs pangocairo pangoft2`"
+	PANGO_CFLAGS="`pkg-config --cflags pangocairo pangoft2`"
 
 $as_echo "#define USE_PANGO_RENDERING 1" >>confdefs.h
 
-	GDIPLUS_PKG_REQ="$GDIPLUS_PKG_REQ pangocairo-1.0"
+	GDIPLUS_PKG_REQ="$GDIPLUS_PKG_REQ pangocairo-1.0 pangoft2-1.0"
 fi
 
  if test $text_v = "pango"; then
