diff --git a/Package@swift-4.swift b/Package@swift-4.swift
index 3eb5d1e..452aa1c 100644
--- a/Package@swift-4.swift
+++ b/Package@swift-4.swift
@@ -7,13 +7,8 @@ let package = Package(
         .library(name: "Node", targets: ["Node"]),
         .library(name: "PathIndexable", targets: ["PathIndexable"]),
     ],
-    dependencies: [
-        .package(url: "https://github.com/vapor/core.git", .upToNextMajor(from: "2.0.0")),
-        .package(url: "https://github.com/vapor/bits.git", .upToNextMajor(from: "1.0.0")),
-        .package(url: "https://github.com/vapor/debugging.git", .upToNextMajor(from: "1.0.0")),
-    ],
     targets: [
-        .target(name: "Node", dependencies: ["Bits", "Core", "PathIndexable"]),
+        .target(name: "Node", dependencies: ["PathIndexable"]),
         .testTarget(name: "NodeTests", dependencies: ["Node"]),
         .target(name: "PathIndexable"),
         .testTarget(name: "PathIndexableTests", dependencies: ["PathIndexable"])
