diff --git a/Package.swift b/Package.swift
index a6fbb9c..e242fd6 100644
--- a/Package.swift
+++ b/Package.swift
@@ -7,11 +7,5 @@ let package = Package(
     name: "Node",
     targets: [
         Target(name: "Node")
-    ],
-    dependencies: [
-        .Package(url: "https://github.com/vapor/path-indexable.git", beta2),
-        .Package(url: "https://github.com/vapor/core.git", beta2),
-        .Package(url: "https://github.com/vapor/bits.git", beta1),
-        .Package(url: "https://github.com/vapor/debugging.git", beta1),
     ]
 )
