diff --git a/source/libasync/posix.d b/source/libasync/posix.d
index ce80e7a..86a4311 100644
--- a/source/libasync/posix.d
+++ b/source/libasync/posix.d
@@ -1143,7 +1143,8 @@ package:
 			sigval sigvl;
 			fd_t err;
 			sigvl.sival_ptr = cast(void*) ctxt;
-			try err = pthread_sigqueue(ctxt.pthreadId, fd, sigvl); catch (Throwable) {}
+			import core.thread : getpid;
+			try err = sigqueue(getpid(), fd, sigvl); catch (Throwable) {}
 			if (catchError!"sigqueue"(err)) {
 				return false;
 			}
