--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -8,7 +8,8 @@
 set(CMAKE_C_FLAGS_RELEASE "-O3 ${CMAKE_C_FLAGS_RELEASE}")
 set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS}")
 
-set(libobjc_VERSION 4.6)
+set(libobjc_SOVERSION 4)
+set(libobjc_VERSION 4.6.0)
 
 set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Xclang -fexceptions -Xclang -fobjc-exceptions")
 if (MSVC)
@@ -312,8 +313,9 @@
 
 
 set_target_properties(objc PROPERTIES
-	LINKER_LANGUAGE C
-	SOVERSION ${libobjc_VERSION}
+	LINKER_LANGUAGE C
+	SOVERSION ${libobjc_SOVERSION}
+	VERSION ${libobjc_VERSION}
 	OUTPUT_NAME ${LIBOBJC_NAME}
 	LINK_FLAGS "${objc_LINK_FLAGS}"
 	)
