--- a/.editorconfig	2026-09-08 22:37:16.929377758 +0200
+++ b/.editorconfig	2026-09-08 22:37:23.645410303 +0200
@@ -29,6 +29,7 @@
 # Errors flagged when this file was added. These should be
 # investigated and either fixed or marked as acceptable with
 # a reason why
+dotnet_diagnostic.CA2263.severity = none
 dotnet_diagnostic.IDE0018.severity = none
 dotnet_diagnostic.CA1806.severity = none
 dotnet_diagnostic.IDE0078.severity = none
