--- a/Directory.Build.props	2023-11-17 22:29:23.755861162 +0100
+++ b/Directory.Build.props	2023-11-17 22:29:35.063730645 +0100
@@ -27,6 +27,7 @@
     <FileVersion>$(Version)</FileVersion>
     <InformationalVersion>$(Version)</InformationalVersion>
 
+    <UseRidGraph>true</UseRidGraph>
     <ToolsServiceTargetRuntimes>win-x64;win-x86;win-arm64;osx-x64;osx-arm64;linux-x64;linux-arm64</ToolsServiceTargetRuntimes>
     <EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>
     <EnableNETAnalyzers>true</EnableNETAnalyzers>
--- a/src/Microsoft.SqlTools.SqlCore/Microsoft.SqlTools.SqlCore.csproj	2024-09-30 20:35:04.301974926 +0200
+++ b/src/Microsoft.SqlTools.SqlCore/Microsoft.SqlTools.SqlCore.csproj	2024-09-30 20:35:21.745774646 +0200
@@ -12,7 +12,7 @@
 		<PreserveCompilationContext>true</PreserveCompilationContext>
 		<AssemblyTitle>SqlTools SqlCore Library</AssemblyTitle>
 		<Description>Provides core sql functionality for SQL server editors like Object explorer, Query Execution, and Scripting</Description>
-		<TargetFrameworks>net6.0</TargetFrameworks>
+		<TargetFrameworks>net8.0</TargetFrameworks>
 		<LangVersion>9.0</LangVersion>
 		<AssemblyName>Microsoft.SqlTools.SqlCore</AssemblyName>
 		<Version>1.0.0</Version>
