--- a/src/Microsoft.SqlTools.ManagedBatchParser/BatchParser/PositionStruct.cs	2022-12-17 15:12:01.722821415 +0100
+++ b/src/Microsoft.SqlTools.ManagedBatchParser/BatchParser/PositionStruct.cs	2022-12-17 15:11:30.725423084 +0100
@@ -8,7 +8,7 @@
 namespace Microsoft.SqlTools.ServiceLayer.BatchParser
 {
     [Serializable]
-    public struct PositionStruct
+    public readonly struct PositionStruct
     {
         private readonly int line;
         private readonly int column;
