--- a/Directory.Build.props	2023-11-17 22:29:23.755861162 +0100
+++ b/Directory.Build.props	2023-11-17 22:29:35.063730645 +0100
@@ -27,6 +27,7 @@
     <FileVersion>$(Version)</FileVersion>
     <InformationalVersion>$(Version)</InformationalVersion>
 
+    <UseRidGraph>true</UseRidGraph>
     <ToolsServiceTargetRuntimes>win-x64;win-x86;win-arm64;osx-x64;osx-arm64;linux-x64;linux-arm64</ToolsServiceTargetRuntimes>
     <EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>
     <EnableNETAnalyzers>true</EnableNETAnalyzers>
