--- a/compiler-rt/lib/hwasan/hwasan_exceptions.cpp	2021-04-10 13:08:33.545428168 +0200
+++ b/compile-rt/lib/hwasan/hwasan_exceptions.cpp	2021-04-10 13:09:32.864110986 +0200
@@ -29,6 +29,7 @@
 // is statically linked and the sanitizer runtime and the program are linked
 // against different unwinders. The _Unwind_Context data structure is opaque so
 // it may be incompatible between unwinders.
+typedef uintptr_t _Unwind_Word __attribute__((__mode__(__unwind_word__)));
 typedef _Unwind_Word GetGRFn(_Unwind_Context* context, int index);
 typedef _Unwind_Word GetCFAFn(_Unwind_Context* context);
 
--- a/compiler-rt/lib/fuzzer/FuzzerInterceptors.cpp	2021-04-10 13:07:29.031216523 +0200
+++ b/compiler-rt/lib/fuzzer/FuzzerInterceptors.cpp	2021-04-10 13:08:01.498308259 +0200
@@ -27,6 +27,7 @@
 #include <cassert>
 #include <cstdint>
 #include <dlfcn.h> // for dlsym()
+#include <sys/types.h> // for size_t
 
 static void *getFuncAddr(const char *name, uintptr_t wrapper_addr) {
   void *addr = dlsym(RTLD_NEXT, name);
