--- a/lib/sanitizer_common/sanitizer_platform_limits_posix.h    2021-09-02 20:10:39.137531525 +0200
+++ b/lib/sanitizer_common/sanitizer_platform_limits_posix.h    2021-09-02 20:31:18.127421985 +0200
@@ -367,11 +367,7 @@
   char **gr_mem;
 };
 
-#if defined(__x86_64__) && !defined(_LP64)
 typedef long long __sanitizer_time_t;
-#else
-typedef long __sanitizer_time_t;
-#endif
 
 typedef long __sanitizer_suseconds_t;
 
@@ -475,20 +471,13 @@
   unsigned short d_reclen;
   // more fields that we don't care about
 };
-#elif SANITIZER_ANDROID || defined(__x86_64__)
+#else
 struct __sanitizer_dirent {
   unsigned long long d_ino;
   unsigned long long d_off;
   unsigned short d_reclen;
   // more fields that we don't care about
 };
-#else
-struct __sanitizer_dirent {
-  uptr d_ino;
-  uptr d_off;
-  unsigned short d_reclen;
-  // more fields that we don't care about
-};
 #endif
 
 #if SANITIZER_LINUX && !SANITIZER_ANDROID
