--- a/compiler-rt/lib/sanitizer_common/sanitizer_linux.cpp
+++ b/compiler-rt/lib/sanitizer_common/sanitizer_linux.cpp
@@ -42,6 +42,10 @@
 #undef stat
 #endif
 
+#if defined(__arm)
+#include <asm/unistd.h>
+#endif
+
 #include <dlfcn.h>
 #include <errno.h>
 #include <fcntl.h>
--- a/compiler-rt/lib/sanitizer_common/sanitizer_platform_limits_posix.h
+++ b/compiler-rt/lib/sanitizer_common/sanitizer_platform_limits_posix.h
@@ -367,11 +367,7 @@
   char **gr_mem;
 };
 
-#if defined(__x86_64__) && !defined(_LP64)
 typedef long long __sanitizer_time_t;
-#else
-typedef long __sanitizer_time_t;
-#endif
 
 typedef long __sanitizer_suseconds_t;
 
