--- src/genhomedircon.c.bak	2016-02-23 21:19:01.288243625 +0100
+++ src/genhomedircon.c	2016-02-23 21:21:58.037252327 +0100
@@ -359,7 +359,7 @@
 	if (rbuf == NULL)
 		goto fail;
 	setpwent();
-	while ((retval = getpwent_r(&pwstorage, rbuf, rbuflen, &pwbuf)) == 0) {
+	while ((pwbuf = getpwent()) != NULL) {
 		if (pwbuf->pw_uid < minuid || pwbuf->pw_uid > maxuid)
 			continue;
 		if (!semanage_list_find(shells, pwbuf->pw_shell))
