--- a/build/aclocal.float.m4	2017-01-04 19:44:06.406044538 +0100
+++ b/build/aclocal.float.m4	2017-01-04 19:46:41.220557156 +0100
@@ -31,10 +31,10 @@
 
 ]])], [
 
-if strings -a conftest$ac_exeext | grep noonsees >/dev/null ; then
+if strings conftest$ac_exeext | grep noonsees >/dev/null ; then
   ax_cv_c_float_words_bigendian=yes
 fi
-if strings -a conftest$ac_exeext | grep seesnoon >/dev/null ; then
+if strings conftest$ac_exeext | grep seesnoon >/dev/null ; then
   if test "$ax_cv_c_float_words_bigendian" = unknown; then
     ax_cv_c_float_words_bigendian=no
   else
