--- a/inc/common/UFO/portable_compiler.h
+++ b/inc/common/UFO/portable_compiler.h
@@ -117,7 +117,8 @@
 /* compile-time ASSERT */
 
 #ifndef C_ASSERT
-    #define __UNIQUENAME( a1, a2 )  __CONCAT( a1, a2 )
+    #define __CONCATING( a1, a2 )   a1 ## a2
+    #define __UNIQUENAME( a1, a2 )  __CONCATING( a1, a2 )
     #define UNIQUENAME( __text )    __UNIQUENAME( __text, __COUNTER__ )
 
 
--- a/IGC/AdaptorOCL/OCL/sp/sp_types.h
+++ b/IGC/AdaptorOCL/OCL/sp/sp_types.h
@@ -46,8 +46,9 @@
     typedef unsigned int        ULONG;
 
     #ifndef C_ASSERT
-        #define __UNIQUENAME( a1, a2 )  __CONCAT( a1, a2 )
-            #define UNIQUENAME( __text )    __UNIQUENAME( __text, __COUNTER__ )
+	#define __CONCATING( a1, a2 )   a1 ## a2
+        #define __UNIQUENAME( a1, a2 )  __CONCATING( a1, a2 )
+        #define UNIQUENAME( __text )    __UNIQUENAME( __text, __COUNTER__ )
 
         #define STATIC_ASSERT(e)  typedef char UNIQUENAME(STATIC_ASSERT_)[(e)?1:-1]
 
