--- a/IGC/CMakeLists.txt
+++ b/IGC/CMakeLists.txt
@@ -2766,7 +2766,6 @@
         SET_RAW
           -pipe
           -fmessage-length=0
-          -march=corei7
           -mstackrealign
           -fms-extensions
           -Wno-unused-parameter
@@ -2787,13 +2785,6 @@
           -Wno-write-strings
           -finline
           -fno-strict-aliasing
-          -msse
-          -msse2
-          -msse3
-          -mssse3
-          -msse4
-          -msse4.1
-          -msse4.2
           -Wno-unknown-pragmas
           -fPIC
       )
