diff --git a/Package@swift-4.swift b/Package@swift-4.swift
index 6fbfe8c..d453ec1 100644
--- a/Package@swift-4.swift
+++ b/Package@swift-4.swift
@@ -4,7 +4,7 @@ import PackageDescription
 let package = Package(
     name: "Debugging",
     products: [
-        .library(name: "Debugging", targets: ["Debugging"])
+        .library(name: "Debugging", type: .dynamic, targets: ["Debugging"])
     ],
     targets: [
         .target(name: "Debugging"),
