--- a/tests/test_setup.py	2026-06-26 11:34:56.421918566 +0200
+++ b/tests/test_setup.py	2026-06-26 11:35:17.850002668 +0200
@@ -28,8 +28,6 @@
 import urllib.request
 import zipfile
 
-from invoke import task
-
 from utils.constants import EXTENSIONS_CSPROJ_TEMPLATE, NUGET_CONFIG
 
 ROOT_DIR = pathlib.Path(__file__).parent.parent
@@ -302,7 +300,6 @@
         sys.exit(1)
 
 
-@task
 def extensions(c, clean=False, extensions_dir=None):
     """Build extensions."""
     extensions_dir = extensions_dir or BUILD_DIR / "extensions"
@@ -317,7 +314,6 @@
     print("Extensions installed successfully.")
 
 
-@task
 def vendor_deps(c, target=None):
     """Vendor third-party deps into azure_functions_worker._vendored.
 
@@ -372,7 +368,6 @@
     print("Vendoring complete.")
 
 
-@task
 def build_protos(c, clean=False):
     """Build gRPC bindings."""
 
@@ -389,7 +384,6 @@
     print("gRPC bindings generated successfully.")
 
 
-@task
 def webhost(c, clean=False, webhost_version=None, webhost_dir=None,
             branch_name=None):
     """Builds the webhost"""
@@ -416,7 +410,6 @@
     compile_webhost(webhost_dir)
 
 
-@task
 def clean(c):
     """Clean build directory."""
 
