--- a/project/abstractfilemanagerplugin.cpp	2017-07-11 20:35:42.383387420 +0200
+++ b/project/abstractfilemanagerplugin.cpp	2017-07-11 20:36:17.292386611 +0200
@@ -78,7 +78,7 @@
      * The just returned must be started in one way or another for this method
      * to have any affect. The job will then auto-delete itself upon completion.
      */
-    KIO::Job* eventuallyReadFolder( ProjectFolderItem* item ) Q_REQUIRED_RESULT;
+    Q_REQUIRED_RESULT KIO::Job* eventuallyReadFolder( ProjectFolderItem* item );
     void addJobItems(FileManagerListJob* job,
                      ProjectFolderItem* baseItem,
                      const KIO::UDSEntryList& entries);
