diff --git a/Package.swift b/Package.swift
index 1d5f160f..c27c73c8 100644
--- a/Package.swift
+++ b/Package.swift
@@ -1,16 +1,5 @@
 import PackageDescription
 
-let dependencies: [Package.Dependency] = [
-    // Crypto
-    .Package(url: "https://github.com/vapor/crypto.git", majorVersion: 1),
-
-    // Secure Sockets
-    .Package(url: "https://github.com/vapor/tls.git", versions: Version(1,1,0)..<Version(2,0,0)),
-
-    // CoreComponents
-    .Package(url: "https://github.com/vapor/core.git", majorVersion: 1),
-]
-
 let package = Package(
     name: "Engine",
     targets: [
@@ -59,7 +48,6 @@ let package = Package(
         )
         */
     ],
-    dependencies: dependencies,
     exclude: [
         "Resources",
         "Sources/HTTPExample",
