--- a/limits/util.py	2023-04-17 15:05:49.000000000 +0200
+++ b/limits/util.py	2023-05-18 13:52:32.303942491 +0200
@@ -4,11 +4,11 @@
 import dataclasses
 import re
 import sys
+import importlib
 from collections import UserDict
 from types import ModuleType
 from typing import TYPE_CHECKING, cast
 
-import importlib_resources
 from packaging.version import Version
 
 from limits.typing import Dict, List, NamedTuple, Optional, Tuple, Type, Union
@@ -143,7 +143,7 @@
 
 
 def get_package_data(path: str) -> bytes:
-    return cast(bytes, importlib_resources.files("limits").joinpath(path).read_bytes())
+    return cast(bytes, importlib.resources.files("limits").joinpath(path).read_bytes())
 
 
 def parse_many(limit_string: str) -> List[RateLimitItem]:
