--- a/core/helpers/preamble.h
+++ b/core/helpers/preamble.h
@@ -82,9 +82,9 @@
 
 namespace DebugModeRegisterOffset {
 template <typename GfxFamily>
-constexpr uint32_t registerOffset = 0x20ec;
+static constexpr uint32_t registerOffset = 0x20ec;
 template <typename GfxFamily>
-constexpr uint32_t debugEnabledValue = (1 << 6) | (1 << 22);
+static constexpr uint32_t debugEnabledValue = (1 << 6) | (1 << 22);
 }; // namespace DebugModeRegisterOffset
 
 namespace TdDebugControlRegisterOffset {
--- a/runtime/gen11/reg_configs.h
+++ b/runtime/gen11/reg_configs.h
@@ -25,9 +25,9 @@
 
 namespace DebugModeRegisterOffset {
 template <>
-constexpr uint32_t registerOffset<ICLFamily> = 0x20d8;
+static constexpr uint32_t registerOffset<ICLFamily> = 0x20d8;
 template <>
-constexpr uint32_t debugEnabledValue<ICLFamily> = (1 << 5) | (1 << 21);
+static constexpr uint32_t debugEnabledValue<ICLFamily> = (1 << 5) | (1 << 21);
 }; // namespace DebugModeRegisterOffset
 
 namespace gen11HdcModeRegister {
