--- ./runtime/built_ins/kernels/fill_buffer_stateless.igdrcl_built_in.bak	2020-02-07 22:44:15.299648124 +0100
+++ ./runtime/built_ins/kernels/fill_buffer_stateless.igdrcl_built_in	2020-02-07 22:44:43.974473011 +0100
@@ -21,7 +21,7 @@
     __global uchar* pDst,
     ulong dstOffsetInBytes,
     const __global uchar* pPattern,
-    const size_t patternSizeInEls )
+    const ulong patternSizeInEls )
 {
     size_t gid = get_global_id(0);
     pDst[ gid + dstOffsetInBytes ] = pPattern[ gid & (patternSizeInEls - 1) ];
@@ -46,4 +46,4 @@
     size_t gid = get_global_id(0);
     pDst[ gid + dstOffsetInBytes ] = pPattern[ gid & (patternSizeInEls - 1) ];
 }
-)==="
\ No newline at end of file
+)==="
