diff --git a/app-crypt/mit-krb5/mit-krb5-1.19.2.ebuild b/app-crypt/mit-krb5/mit-krb5-1.19.2.ebuild
index 1ee2e7de82c..79dd5d355c6 100644
--- a/app-crypt/mit-krb5/mit-krb5-1.19.2.ebuild
+++ b/app-crypt/mit-krb5/mit-krb5-1.19.2.ebuild
@@ -3,7 +3,7 @@
 
 EAPI=7
 
-PYTHON_COMPAT=( python3_{7,8,9} )
+PYTHON_COMPAT=( python3_{7,8,9,10} )
 inherit autotools flag-o-matic multilib-minimal python-any-r1 systemd toolchain-funcs
 
 MY_P="${P/mit-}"
diff --git a/app-emulation/libvirt/libvirt-7.8.0.ebuild b/app-emulation/libvirt/libvirt-7.8.0.ebuild
index e20eaa25a6c..5d525b6915e 100644
--- a/app-emulation/libvirt/libvirt-7.8.0.ebuild
+++ b/app-emulation/libvirt/libvirt-7.8.0.ebuild
@@ -110,8 +110,6 @@ RDEPEND="
 	selinux? ( >=sys-libs/libselinux-2.0.85 )
 	virt-network? (
 		net-dns/dnsmasq[dhcp,ipv6,script]
-		net-firewall/ebtables
-		>=net-firewall/iptables-1.4.10[ipv6]
 		net-misc/radvd
 		sys-apps/iproute2[-minimal]
 	)
diff --git a/app-emulation/spice/spice-0.15.0.ebuild b/app-emulation/spice/spice-0.15.0.ebuild
index 30f4642538d..87fb8f6ed9d 100644
--- a/app-emulation/spice/spice-0.15.0.ebuild
+++ b/app-emulation/spice/spice-0.15.0.ebuild
@@ -3,7 +3,7 @@
 
 EAPI=7
 
-PYTHON_COMPAT=( python3_{7,8,9} )
+PYTHON_COMPAT=( python3_{7,8,9,10} )
 inherit autotools python-any-r1 readme.gentoo-r1 xdg-utils
 
 DESCRIPTION="SPICE server"
diff --git a/app-emulation/virt-manager/virt-manager-3.2.0.ebuild b/app-emulation/virt-manager/virt-manager-3.2.0.ebuild
index 8529c5b317c..21e0846d90b 100644
--- a/app-emulation/virt-manager/virt-manager-3.2.0.ebuild
+++ b/app-emulation/virt-manager/virt-manager-3.2.0.ebuild
@@ -3,7 +3,7 @@
 
 EAPI=6
 
-PYTHON_COMPAT=( python3_{8,9} )
+PYTHON_COMPAT=( python3_{8,9,10} )
 DISTUTILS_SINGLE_IMPL=1
 
 DISTUTILS_USE_SETUPTOOLS=no
@@ -40,7 +40,6 @@ RDEPEND="${PYTHON_DEPS}
 		gnome-base/dconf
 		>=net-libs/gtk-vnc-0.3.8[gtk3(+),introspection]
 		net-misc/spice-gtk[usbredir,gtk3,introspection,sasl?]
-		net-misc/x11-ssh-askpass
 		x11-libs/gtk+:3[introspection]
 		x11-libs/gtksourceview:4[introspection]
 		x11-libs/vte:2.91[introspection]
diff --git a/app-misc/solaar/solaar-1.0.5-r2.ebuild b/app-misc/solaar/solaar-1.0.5-r2.ebuild
index de5cdd6f843..4ba6a1c6517 100644
--- a/app-misc/solaar/solaar-1.0.5-r2.ebuild
+++ b/app-misc/solaar/solaar-1.0.5-r2.ebuild
@@ -4,7 +4,7 @@
 EAPI=7
 
 DISTUTILS_SINGLE_IMPL=1
-PYTHON_COMPAT=( python3_{7..9} )
+PYTHON_COMPAT=( python3_{7..10} )
 
 inherit linux-info udev xdg distutils-r1
 
diff --git a/app-text/xmlto/xmlto-0.0.28-r3.ebuild b/app-text/xmlto/xmlto-0.0.28-r3.ebuild
index 6b8268c5f68..e0de1c43e12 100644
--- a/app-text/xmlto/xmlto-0.0.28-r3.ebuild
+++ b/app-text/xmlto/xmlto-0.0.28-r3.ebuild
@@ -20,7 +20,7 @@ RDEPEND=">=app-text/docbook-xsl-stylesheets-1.62.0-r1
 	>=sys-apps/coreutils-6.10-r1
 	|| ( sys-apps/util-linux app-misc/getopt )
 	sys-apps/which
-	text? ( || ( virtual/w3m www-client/lynx www-client/elinks ) )
+	text? ( || ( virtual/w3m www-client/lynx www-client/links ) )
 	latex? ( dev-texlive/texlive-formatsextra )"
 # We only depend on flex when we patch the imput lexer.
 DEPEND="${RDEPEND}"
diff --git a/dev-java/java-config/java-config-2.3.1.ebuild b/dev-java/java-config/java-config-2.3.1.ebuild
index 69ba112dacf..45feed60cc3 100644
--- a/dev-java/java-config/java-config-2.3.1.ebuild
+++ b/dev-java/java-config/java-config-2.3.1.ebuild
@@ -4,7 +4,7 @@
 EAPI=7
 
 # jython depends on java-config, so don't add it or things will break
-PYTHON_COMPAT=( python3_{7..9} )
+PYTHON_COMPAT=( python3_{7..10} )
 DISTUTILS_USE_SETUPTOOLS=no
 
 inherit distutils-r1
diff --git a/dev-java/openjdk/openjdk-17.0.1_p12.ebuild b/dev-java/openjdk/openjdk-17.0.1_p12.ebuild
index b5e1cacf951..2709151130a 100644
--- a/dev-java/openjdk/openjdk-17.0.1_p12.ebuild
+++ b/dev-java/openjdk/openjdk-17.0.1_p12.ebuild
@@ -62,6 +62,8 @@ DEPEND="
 	x11-libs/libXtst
 	javafx? ( dev-java/openjfx:${SLOT}= )
 	|| (
+		dev-java/openjdk-bin:$((${SLOT} - 1))
+		dev-java/openjdk:$((${SLOT} - 1))
 		dev-java/openjdk-bin:${SLOT}
 		dev-java/openjdk:${SLOT}
 	)
@@ -95,7 +97,7 @@ pkg_setup() {
 	openjdk_check_requirements
 	java-vm-2_pkg_setup
 
-	JAVA_PKG_WANT_BUILD_VM="openjdk-${SLOT} openjdk-bin-${SLOT}"
+	JAVA_PKG_WANT_BUILD_VM="openjdk-$((${SLOT} - 1)) openjdk-bin-$((${SLOT} - 1)) openjdk-${SLOT} openjdk-bin-${SLOT}"
 	JAVA_PKG_WANT_SOURCE="${SLOT}"
 	JAVA_PKG_WANT_TARGET="${SLOT}"
 
diff --git a/dev-libs/icu/icu-69.1-r1.ebuild b/dev-libs/icu/icu-69.1-r1.ebuild
index 6f632a02fd2..3cdbad15f5d 100644
--- a/dev-libs/icu/icu-69.1-r1.ebuild
+++ b/dev-libs/icu/icu-69.1-r1.ebuild
@@ -3,7 +3,7 @@
 
 EAPI=7
 
-PYTHON_COMPAT=( python3_{7..9} )
+PYTHON_COMPAT=( python3_{7..10} )
 inherit autotools flag-o-matic multilib-minimal python-any-r1 toolchain-funcs
 
 DESCRIPTION="International Components for Unicode"
diff --git a/dev-libs/jsoncpp/jsoncpp-1.9.4.ebuild b/dev-libs/jsoncpp/jsoncpp-1.9.4.ebuild
index 5c858b99e51..71758758a35 100644
--- a/dev-libs/jsoncpp/jsoncpp-1.9.4.ebuild
+++ b/dev-libs/jsoncpp/jsoncpp-1.9.4.ebuild
@@ -3,7 +3,7 @@
 
 EAPI=7
 
-PYTHON_COMPAT=( python3_{7..9} )
+PYTHON_COMPAT=( python3_{7..10} )
 
 inherit meson python-any-r1
 
diff --git a/dev-libs/libffi/libffi-3.3-r2.ebuild b/dev-libs/libffi/libffi-3.3-r2.ebuild
index 95c00185605..ed29f54b5d2 100644
--- a/dev-libs/libffi/libffi-3.3-r2.ebuild
+++ b/dev-libs/libffi/libffi-3.3-r2.ebuild
@@ -50,19 +50,7 @@ src_prepare() {
 
 multilib_src_configure() {
 	use userland_BSD && export HOST="${CHOST}"
-	# --includedir= path maintains a few properties:
-	# 1. have stable name across libffi versions: some packages like
-	#    dev-lang/ghc or kde-frameworks/networkmanager-qt embed
-	#    ${includedir} at build-time. Don't require those to be
-	#    rebuilt unless SONAME changes. bug #695788
-	#
-	#    We use /usr/.../${PN} (instead of former /usr/.../${P}).
-	#
-	# 2. have ${ABI}-specific location as ffi.h is target-dependent.
-	#
-	#    We use /usr/$(get_libdir)/... to have ABI identifier.
 	econf \
-		--includedir="${EPREFIX}"/usr/$(get_libdir)/${PN}/include \
 		--disable-multi-os-directory \
 		$(use_enable static-libs static) \
 		$(use_enable pax-kernel pax_emutramp) \
diff --git a/dev-libs/zziplib/zziplib-0.13.72.ebuild b/dev-libs/zziplib/zziplib-0.13.72.ebuild
index 7a79e6dbded..326003777fb 100644
--- a/dev-libs/zziplib/zziplib-0.13.72.ebuild
+++ b/dev-libs/zziplib/zziplib-0.13.72.ebuild
@@ -3,7 +3,7 @@
 
 EAPI=7
 
-PYTHON_COMPAT=( python3_{7..9} )
+PYTHON_COMPAT=( python3_{7..10} )
 
 inherit cmake flag-o-matic python-any-r1
 
diff --git a/dev-python/gst-python/gst-python-1.18.4.ebuild b/dev-python/gst-python/gst-python-1.18.4.ebuild
index e58a299ee53..9423cab278f 100644
--- a/dev-python/gst-python/gst-python-1.18.4.ebuild
+++ b/dev-python/gst-python/gst-python-1.18.4.ebuild
@@ -3,7 +3,7 @@
 
 EAPI=7
 
-PYTHON_COMPAT=( python{3_8,3_9} )
+PYTHON_COMPAT=( python{3_8,3_9,3_10} )
 
 inherit meson python-r1 xdg-utils
 
diff --git a/dev-python/pypax/pypax-0.9.5.ebuild b/dev-python/pypax/pypax-0.9.5.ebuild
index 61c08aa64aa..88a7b850a4b 100644
--- a/dev-python/pypax/pypax-0.9.5.ebuild
+++ b/dev-python/pypax/pypax-0.9.5.ebuild
@@ -4,7 +4,7 @@
 EAPI=6
 
 DISTUTILS_USE_SETUPTOOLS=no
-PYTHON_COMPAT=( python3_{8,9} )
+PYTHON_COMPAT=( python3_{8,9,10} )
 
 inherit distutils-r1
 
diff --git a/dev-python/pyudev/pyudev-0.22.0.ebuild b/dev-python/pyudev/pyudev-0.22.0.ebuild
index 7add9fbf588..6c0229bd512 100644
--- a/dev-python/pyudev/pyudev-0.22.0.ebuild
+++ b/dev-python/pyudev/pyudev-0.22.0.ebuild
@@ -2,7 +2,7 @@
 # Distributed under the terms of the GNU General Public License v2
 
 EAPI=7
-PYTHON_COMPAT=( python3_{7..9} )
+PYTHON_COMPAT=( python3_{7..10} )
 
 inherit distutils-r1
 
diff --git a/dev-util/flatpak-builder/flatpak-builder-1.0.11.ebuild b/dev-util/flatpak-builder/flatpak-builder-1.0.11.ebuild
index a8017f003f1..27fdf168897 100644
--- a/dev-util/flatpak-builder/flatpak-builder-1.0.11.ebuild
+++ b/dev-util/flatpak-builder/flatpak-builder-1.0.11.ebuild
@@ -21,6 +21,7 @@ RDEPEND="
 	>=dev-libs/libxml2-2.4:=
 	dev-libs/json-glib:=
 	net-misc/curl:=
+	elibc_musl? ( dev-libs/libdwarf )
 	yaml? ( dev-libs/libyaml:= )
 "
 DEPEND="${RDEPEND}"
diff --git a/eclass/linux-mod.eclass b/eclass/linux-mod.eclass
index 9f4ae64f6b5..09ebe986a4d 100644
--- a/eclass/linux-mod.eclass
+++ b/eclass/linux-mod.eclass
@@ -711,6 +711,7 @@ linux-mod_src_compile() {
 			# fails.
 			eval "emake HOSTCC=\"$(tc-getBUILD_CC)\" \
 						LDFLAGS=\"$(get_abi_LDFLAGS)\" \
+						LLVM_IAS=1 \
 						${BUILD_FIXES} \
 						${BUILD_PARAMS} \
 						${BUILD_TARGETS} " \
diff --git a/gnome-base/gsettings-desktop-schemas/gsettings-desktop-schemas-41.0.ebuild b/gnome-base/gsettings-desktop-schemas/gsettings-desktop-schemas-41.0.ebuild
index a5545f400e7..929c1310ba7 100644
--- a/gnome-base/gsettings-desktop-schemas/gsettings-desktop-schemas-41.0.ebuild
+++ b/gnome-base/gsettings-desktop-schemas/gsettings-desktop-schemas-41.0.ebuild
@@ -3,7 +3,7 @@
 
 EAPI=7
 
-PYTHON_COMPAT=(python3_{6..9})
+PYTHON_COMPAT=(python3_{6..10})
 inherit gnome.org gnome2-utils meson python-any-r1 xdg
 
 DESCRIPTION="Collection of GSettings schemas for GNOME desktop"
diff --git a/kde-plasma/kde-gtk-config/kde-gtk-config-5.23.3.ebuild b/kde-plasma/kde-gtk-config/kde-gtk-config-5.23.3.ebuild
index 8ebcf8b1fc8..52abaf91d76 100644
--- a/kde-plasma/kde-gtk-config/kde-gtk-config-5.23.3.ebuild
+++ b/kde-plasma/kde-gtk-config/kde-gtk-config-5.23.3.ebuild
@@ -18,7 +18,6 @@ KEYWORDS="~amd64 ~arm ~arm64 ~ppc64 ~riscv ~x86"
 IUSE=""
 
 DEPEND="
-	dev-cpp/glibmm:2
 	dev-libs/glib:2
 	>=dev-qt/qtdbus-${QTMIN}:5
 	>=dev-qt/qtgui-${QTMIN}:5
diff --git a/media-libs/graphene/graphene-1.10.6.ebuild b/media-libs/graphene/graphene-1.10.6.ebuild
index ab7fb554651..c7ff2715520 100644
--- a/media-libs/graphene/graphene-1.10.6.ebuild
+++ b/media-libs/graphene/graphene-1.10.6.ebuild
@@ -3,7 +3,7 @@
 
 EAPI=7
 
-PYTHON_COMPAT=( python3_{7..9} )
+PYTHON_COMPAT=( python3_{7..10} )
 inherit xdg-utils meson-multilib python-any-r1
 
 DESCRIPTION="A thin layer of types for graphic libraries"
diff --git a/media-libs/libglvnd/libglvnd-1.3.4.ebuild b/media-libs/libglvnd/libglvnd-1.3.4.ebuild
index 8c83eaea578..4e4423833af 100644
--- a/media-libs/libglvnd/libglvnd-1.3.4.ebuild
+++ b/media-libs/libglvnd/libglvnd-1.3.4.ebuild
@@ -9,7 +9,7 @@ if [[ ${PV} = 9999* ]]; then
 	GIT_ECLASS="git-r3"
 fi
 
-PYTHON_COMPAT=( python3_{7..9} )
+PYTHON_COMPAT=( python3_{7..10} )
 VIRTUALX_REQUIRED=manual
 
 inherit ${GIT_ECLASS} meson-multilib python-any-r1 virtualx
diff --git a/media-video/mkvtoolnix/mkvtoolnix-60.0.0.ebuild b/media-video/mkvtoolnix/mkvtoolnix-60.0.0.ebuild
index cd70327db01..d2d56536d98 100644
--- a/media-video/mkvtoolnix/mkvtoolnix-60.0.0.ebuild
+++ b/media-video/mkvtoolnix/mkvtoolnix-60.0.0.ebuild
@@ -19,7 +19,7 @@ HOMEPAGE="https://mkvtoolnix.download/ https://gitlab.com/mbunkus/mkvtoolnix"
 
 LICENSE="GPL-2"
 SLOT="0"
-IUSE="dbus debug dvd nls pch test"
+IUSE="dbus debug dvd gui nls pch test"
 RESTRICT="!test? ( test )"
 
 # check NEWS.md for build system changes entries for boost/libebml/libmatroska
@@ -43,7 +43,7 @@ RDEPEND="
 	dev-qt/qtwidgets:5
 	dev-qt/qtconcurrent:5
 	dev-qt/qtmultimedia:5
-	app-text/cmark:0=
+	gui? ( app-text/cmark:0= )
 	dbus? ( dev-qt/qtdbus:5 )
 "
 DEPEND="${RDEPEND}
@@ -87,6 +87,7 @@ src_configure() {
 		$(use_enable debug)
 		$(usex pch "" --disable-precompiled-headers)
 		$(use_enable dbus)
+		$(use_enable gui)
 		--disable-qt6
 		--enable-qt5
 		--with-qmake=$(qt5_get_bindir)/qmake
diff --git a/media-video/obs-studio/obs-studio-27.1.3-r1.ebuild b/media-video/obs-studio/obs-studio-27.1.3-r1.ebuild
index 9cd43053e71..693858ef56c 100644
--- a/media-video/obs-studio/obs-studio-27.1.3-r1.ebuild
+++ b/media-video/obs-studio/obs-studio-27.1.3-r1.ebuild
@@ -119,10 +119,6 @@ QA_PREBUILT="
 	usr/lib*/obs-plugins/swiftshader/libGLESv2.so
 "
 
-PATCHES=(
-	"${FILESDIR}/${PN}-26.1.2-python-3.8.patch"
-)
-
 pkg_setup() {
 	use lua && lua-single_pkg_setup
 	use python && python-single-r1_pkg_setup
diff --git a/net-fs/sshfs/Manifest b/net-fs/sshfs/Manifest
index b92f5ff678f..41b848966d7 100644
--- a/net-fs/sshfs/Manifest
+++ b/net-fs/sshfs/Manifest
@@ -1,2 +1,3 @@
 DIST sshfs-3.7.0.tar.xz 52724 BLAKE2B be974475692e8f9a4d0653776b7b7dc62c6beb1a300caf62c55cbc6bb72fedcf08cdd2c65f051415df57e6e2b21e5a3cb00e8525585c40ab92b0bd0b3474ac77 SHA512 bd8bcd45dd9a5e9686c6fb442e877ffdb592ba0d3424d5dab84a955bfafb17e8666abefba6857467833f5b285842bdadd5a9b6c9e8128ac2e564c36df79aa570
 DIST sshfs-3.7.1.tar.xz 52988 BLAKE2B 48fd9b1483a8f4d48e3d9d25dc13c6e265ca3cb623762b2721b955db52fc05a6138cd0fef684e8893c3b380041ac30ecb3df4c5716b067f621790730fa07be08 SHA512 ee91b2eacbad5891006dbac14188ddd591e242b6092c7b7d8234503d79acb52f4b7ea9a15d5eaad83597ff4b4e700580ee2271671edfa6db762d9a8c756d45fd
+DIST sshfs-3.7.2.tar.xz 53160 BLAKE2B 4acf13b5b7c056d1090eadd57bf4df6549d1f712fbfea59db2c568a94cb621ebf31d84324aeba6549c8f8818f3d4685de9d59e09f189c5307b2efb1a992efae2 SHA512 1612d515869a8b38c0bc3d79eaf80e8b4da1a9710c9668e6ec23e6b77313063ae0bee66ab9e9e0f70d98eaa1334425b2087dd93cc4618af3ac6ac191acf7f838
diff --git a/net-libs/gtk-vnc/gtk-vnc-1.2.0.ebuild b/net-libs/gtk-vnc/gtk-vnc-1.2.0.ebuild
index e879d2277a2..2af0190343f 100644
--- a/net-libs/gtk-vnc/gtk-vnc-1.2.0.ebuild
+++ b/net-libs/gtk-vnc/gtk-vnc-1.2.0.ebuild
@@ -2,7 +2,7 @@
 # Distributed under the terms of the GNU General Public License v2
 
 EAPI=7
-PYTHON_COMPAT=( python3_{7..9} )
+PYTHON_COMPAT=( python3_{7..10} )
 
 inherit gnome.org vala meson python-any-r1 xdg
 
@@ -56,7 +56,7 @@ src_configure() {
 		$(meson_feature introspection)
 		$(meson_feature pulseaudio)
 		$(meson_feature sasl)
-		-Dwith-coroutine=auto # gthread on windows, libc ucontext elsewhere; neither has extra deps
+		-Dwith-coroutine=gthread # gthread on windows, libc ucontext elsewhere; neither has extra deps
 		$(meson_feature vala with-vala)
 	)
 	meson_src_configure
diff --git a/net-libs/libpsl/libpsl-0.21.1.ebuild b/net-libs/libpsl/libpsl-0.21.1.ebuild
index 379fbf8773d..7a4d68e7b94 100644
--- a/net-libs/libpsl/libpsl-0.21.1.ebuild
+++ b/net-libs/libpsl/libpsl-0.21.1.ebuild
@@ -3,7 +3,7 @@
 
 EAPI=7
 
-PYTHON_COMPAT=(python3_{7,8,9})
+PYTHON_COMPAT=(python3_{7,8,10})
 inherit multilib-minimal python-any-r1
 
 DESCRIPTION="C library for the Public Suffix List"
diff --git a/net-misc/spice-gtk/spice-gtk-0.39-r1.ebuild b/net-misc/spice-gtk/spice-gtk-0.39-r1.ebuild
index 08fdbf6d9fa..0d1d4fe990c 100644
--- a/net-misc/spice-gtk/spice-gtk-0.39-r1.ebuild
+++ b/net-misc/spice-gtk/spice-gtk-0.39-r1.ebuild
@@ -6,7 +6,7 @@ EAPI=7
 VALA_MIN_API_VERSION="0.14"
 VALA_USE_DEPEND="vapigen"
 
-PYTHON_COMPAT=( python3_{7..9} )
+PYTHON_COMPAT=( python3_{7..10} )
 
 inherit desktop meson python-any-r1 readme.gentoo-r1 vala xdg-utils
 
@@ -95,6 +95,7 @@ src_prepare() {
 
 src_configure() {
 	local emesonargs=(
+		-Dcoroutine=gthread
 		$(meson_feature gtk3 gtk)
 		$(meson_feature introspection)
 		$(meson_use mjpeg builtin-mjpeg)
diff --git a/sys-auth/pambase/pambase-20210201.1.ebuild b/sys-auth/pambase/pambase-20210201.1.ebuild
index b4166689528..6b1d2ac83aa 100644
--- a/sys-auth/pambase/pambase-20210201.1.ebuild
+++ b/sys-auth/pambase/pambase-20210201.1.ebuild
@@ -3,7 +3,7 @@
 
 EAPI=7
 
-PYTHON_COMPAT=( python3_{7..9} )
+PYTHON_COMPAT=( python3_{7..10} )
 
 inherit pam python-any-r1 readme.gentoo-r1
 
diff --git a/sys-devel/clang/clang-13.0.0.ebuild b/sys-devel/clang/clang-13.0.0.ebuild
index 707bb762d95..27a1ed5907a 100644
--- a/sys-devel/clang/clang-13.0.0.ebuild
+++ b/sys-devel/clang/clang-13.0.0.ebuild
@@ -21,6 +21,8 @@ IUSE="debug default-compiler-rt default-libcxx default-lld
 REQUIRED_USE="${PYTHON_REQUIRED_USE}"
 RESTRICT="!test? ( test )"
 
+IUSE="${IUSE} +flang gcc-symlinks"
+
 RDEPEND="
 	~sys-devel/llvm-${PV}:${SLOT}=[debug=,${MULTILIB_USEDEP}]
 	static-analyzer? ( dev-lang/perl:* )
@@ -41,6 +43,10 @@ PDEPEND="
 		llvm-libunwind? ( sys-libs/llvm-libunwind )
 		!llvm-libunwind? ( sys-libs/libunwind )
 	)
+	flang? (
+		dev-libs/libpgmath
+		dev-lang/flang
+		sys-devel/llvm[flang] )
 	default-libcxx? ( >=sys-libs/libcxx-${PV} )
 	default-lld? ( sys-devel/lld )"
 
@@ -219,8 +225,7 @@ get_distribution_components() {
 }
 
 multilib_src_configure() {
-	local llvm_version=$(llvm-config --version) || die
-	local clang_version=$(ver_cut 1-3 "${llvm_version}")
+	local clang_version=$(ver_cut 1-3)
 
 	local mycmakeargs=(
 		-DLLVM_CMAKE_PATH="${EPREFIX}/usr/lib/llvm/${SLOT}/$(get_libdir)/cmake/llvm"
@@ -235,6 +240,7 @@ multilib_src_configure() {
 
 		-DLLVM_TARGETS_TO_BUILD="${LLVM_TARGETS// /;}"
 		-DLLVM_BUILD_TESTS=$(usex test)
+		-DLLVM_ENABLE_CLASSIC_FLANG=$(usex flang)
 
 		# these are not propagated reliably, so redefine them
 		-DLLVM_ENABLE_EH=ON
@@ -248,12 +254,15 @@ multilib_src_configure() {
 		# disable using CUDA to autodetect GPU, just build for all
 		-DCMAKE_DISABLE_FIND_PACKAGE_CUDA=ON
 
-		# override default stdlib and rtlib
+		# override default stdlib and rtlib and libunwind
 		-DCLANG_DEFAULT_CXX_STDLIB=$(usex default-libcxx libc++ "")
 		-DCLANG_DEFAULT_RTLIB=$(usex default-compiler-rt compiler-rt "")
 		-DCLANG_DEFAULT_LINKER=$(usex default-lld lld "")
 		-DCLANG_DEFAULT_UNWINDLIB=$(usex default-compiler-rt libunwind "")
 
+		# create gcc symlinks: gcc g++ cc cpp c++
+		-DCLANG_INSTALL_GCC_SYMLINKS=$(usex gcc-symlinks)
+
 		-DCLANG_ENABLE_ARCMT=$(usex static-analyzer)
 		-DCLANG_ENABLE_STATIC_ANALYZER=$(usex static-analyzer)
 
@@ -345,12 +354,15 @@ src_install() {
 
 	# Apply CHOST and version suffix to clang tools
 	# note: we use two version components here (vs 3 in runtime path)
-	local llvm_version=$(llvm-config --version) || die
-	local clang_version=$(ver_cut 1 "${llvm_version}")
-	local clang_full_version=$(ver_cut 1-3 "${llvm_version}")
-	local clang_tools=( clang clang++ clang-cl clang-cpp )
+	local clang_version=$(ver_cut 1)
+	local clang_full_version=$(ver_cut 1-3)
+	local clang_tools=( clang clang++ clang-cl clang-cpp gcc g++ cpp )
 	local abi i
 
+	if use flang; then
+		clang_tools+=( flang gfortran )
+	fi
+
 	# cmake gives us:
 	# - clang-X
 	# - clang -> clang-X
diff --git a/sys-devel/distcc/distcc-3.4.ebuild b/sys-devel/distcc/distcc-3.4.ebuild
index cb49de0fd3c..f8147fe3366 100644
--- a/sys-devel/distcc/distcc-3.4.ebuild
+++ b/sys-devel/distcc/distcc-3.4.ebuild
@@ -3,7 +3,7 @@
 
 EAPI=7
 
-PYTHON_COMPAT=( python3_{7..9} )
+PYTHON_COMPAT=( python3_{7..10} )
 
 inherit autotools flag-o-matic prefix python-single-r1 systemd \
 	toolchain-funcs xdg-utils
@@ -67,6 +67,7 @@ src_configure() {
 	local myconf=(
 		--disable-Werror
 		--libdir="${EPREFIX}"/usr/lib
+		--without-libiberty
 		$(use_enable ipv6 rfc2553)
 		$(use_with gtk)
 		--without-gnome
diff --git a/sys-devel/distcc/files/distccd.service-1 b/sys-devel/distcc/files/distccd.service-1
index 9bd470297df..fd481026e27 100644
--- a/sys-devel/distcc/files/distccd.service-1
+++ b/sys-devel/distcc/files/distccd.service-1
@@ -4,7 +4,7 @@ After=network.target
 
 [Service]
 User=distcc
-ExecStart=/usr/bin/distccd --no-detach --daemon --port 3632 -N 15 --allow $ALLOWED_SERVERS
+ExecStart=/usr/bin/distccd --no-detach --daemon --port 3632 -N 15 $ALLOWED_SERVERS
 
 [Install]
 WantedBy=multi-user.target
diff --git a/sys-devel/llvm/llvm-13.0.0.ebuild b/sys-devel/llvm/llvm-13.0.0.ebuild
index 4b33fd90585..62755f3c3a3 100644
--- a/sys-devel/llvm/llvm-13.0.0.ebuild
+++ b/sys-devel/llvm/llvm-13.0.0.ebuild
@@ -23,6 +23,8 @@ IUSE="debug doc exegesis +gold libedit +libffi ncurses test xar xml z3
 	kernel_Darwin"
 RESTRICT="!test? ( test )"
 
+IUSE="${IUSE} +flang binutils-symlinks"
+
 RDEPEND="
 	sys-libs/zlib:0=[${MULTILIB_USEDEP}]
 	exegesis? ( dev-libs/libpfm:= )
@@ -304,6 +306,23 @@ get_distribution_components() {
 				docs-llvm-man
 			)
 		fi
+
+		use binutils-symlinks && out+=(
+			addr2line
+			ar
+			c++filt
+			dlltool
+			dwp
+			nm
+			objcopy
+			objdump
+			ranlib
+			readelf
+			size
+			strings
+			strip
+                )
+
 		use doc && out+=(
 			docs-llvm-html
 		)
@@ -335,6 +354,7 @@ multilib_src_configure() {
 		-DLLVM_BUILD_LLVM_DYLIB=ON
 		-DLLVM_LINK_LLVM_DYLIB=ON
 		-DLLVM_DISTRIBUTION_COMPONENTS=$(get_distribution_components)
+		-DLLVM_INSTALL_BINUTILS_SYMLINKS=$(usex binutils-symlinks)
 
 		# cheap hack: LLVM combines both anyway, and the only difference
 		# is that the former list is explicitly verified at cmake time
@@ -351,6 +371,7 @@ multilib_src_configure() {
 		-DLLVM_ENABLE_EH=ON
 		-DLLVM_ENABLE_RTTI=ON
 		-DLLVM_ENABLE_Z3_SOLVER=$(usex z3)
+		-DLLVM_ENABLE_CLASSIC_FLANG=$(usex flang)
 
 		-DLLVM_HOST_TRIPLE="${CHOST}"
 
diff --git a/sys-fs/cryfs/cryfs-0.10.3-r1.ebuild b/sys-fs/cryfs/cryfs-0.10.3-r1.ebuild
index 199c5a6badd..33eb35625eb 100644
--- a/sys-fs/cryfs/cryfs-0.10.3-r1.ebuild
+++ b/sys-fs/cryfs/cryfs-0.10.3-r1.ebuild
@@ -3,7 +3,7 @@
 
 EAPI=7
 
-PYTHON_COMPAT=( python3_{7..9} )
+PYTHON_COMPAT=( python3_{7..10} )
 inherit cmake flag-o-matic linux-info python-any-r1
 
 if [[ ${PV} == 9999 ]] ; then
diff --git a/sys-libs/libblockdev/libblockdev-2.26.ebuild b/sys-libs/libblockdev/libblockdev-2.26.ebuild
index 7a897bf9254..9dbccbc4377 100644
--- a/sys-libs/libblockdev/libblockdev-2.26.ebuild
+++ b/sys-libs/libblockdev/libblockdev-2.26.ebuild
@@ -3,7 +3,7 @@
 
 EAPI=7
 
-PYTHON_COMPAT=( python3_{7..9} )
+PYTHON_COMPAT=( python3_{7..10} )
 inherit autotools python-single-r1 xdg-utils
 
 DESCRIPTION="A library for manipulating block devices"
diff --git a/sys-libs/liburing/liburing-0.7-r1.ebuild b/sys-libs/liburing/liburing-0.7-r1.ebuild
index 715212503eb..3eb405715f9 100644
--- a/sys-libs/liburing/liburing-0.7-r1.ebuild
+++ b/sys-libs/liburing/liburing-0.7-r1.ebuild
@@ -43,7 +43,7 @@ multilib_src_configure() {
 }
 
 multilib_src_compile() {
-	emake V=1 AR="$(tc-getAR)" RANLIB="$(tc-getRANLIB)"
+	emake V=1 AR="$(tc-getAR)" RANLIB="$(tc-getRANLIB)" src
 }
 
 multilib_src_install_all() {
diff --git a/x11-libs/vte/vte-0.64.2.ebuild b/x11-libs/vte/vte-0.64.2.ebuild
index 21873ac96c9..4bb5fc1466d 100644
--- a/x11-libs/vte/vte-0.64.2.ebuild
+++ b/x11-libs/vte/vte-0.64.2.ebuild
@@ -2,7 +2,7 @@
 # Distributed under the terms of the GNU General Public License v2
 
 EAPI=7
-PYTHON_COMPAT=( python3_{7..9} )
+PYTHON_COMPAT=( python3_{7..10} )
 VALA_USE_DEPEND="vapigen"
 VALA_MIN_API_VERSION="0.32"
 
diff --git a/x11-misc/appmenu-gtk-module/appmenu-gtk-module-0.7.3-r2.ebuild b/x11-misc/appmenu-gtk-module/appmenu-gtk-module-0.7.3-r2.ebuild
index a9248abf20d..629a3b10447 100644
--- a/x11-misc/appmenu-gtk-module/appmenu-gtk-module-0.7.3-r2.ebuild
+++ b/x11-misc/appmenu-gtk-module/appmenu-gtk-module-0.7.3-r2.ebuild
@@ -16,7 +16,6 @@ IUSE="wayland"
 
 RDEPEND="
 	dev-libs/glib[dbus]
-	>=x11-libs/gtk+-2.24.0:2
 	>=x11-libs/gtk+-3.22.0:3[wayland=]
 "
 DEPEND="${RDEPEND}
