diff --git a/app-editors/kile/kile-2.9.93.ebuild b/app-editors/kile/kile-2.9.93.ebuild
index e13393dac..89442927b 100644
--- a/app-editors/kile/kile-2.9.93.ebuild
+++ b/app-editors/kile/kile-2.9.93.ebuild
@@ -51,7 +51,6 @@ RDEPEND="${DEPEND}
 	virtual/tex-base
 	pdf? (
 		>=app-text/texlive-core-2014
-		app-text/ghostscript-gpl
 	)
 	png? (
 		app-text/dvipng
diff --git a/app-i18n/fcitx/fcitx-4.2.9.6.ebuild b/app-i18n/fcitx/fcitx-4.2.9.6.ebuild
index 0681078c5..292fb3eb3 100644
--- a/app-i18n/fcitx/fcitx-4.2.9.6.ebuild
+++ b/app-i18n/fcitx/fcitx-4.2.9.6.ebuild
@@ -52,7 +52,7 @@ RDEPEND="dev-libs/glib:2
 		pango? ( x11-libs/pango )
 		!pango? ( media-libs/fontconfig )
 	)
-	enchant? ( app-text/enchant:0= )
+	enchant? ( app-text/enchant:= )
 	gtk2? ( x11-libs/gtk+:2 )
 	gtk3? ( x11-libs/gtk+:3 )
 	introspection? ( dev-libs/gobject-introspection )
diff --git a/app-misc/ca-certificates/ca-certificates-20190110.3.43.ebuild b/app-misc/ca-certificates/ca-certificates-20190110.3.43.ebuild
index 871207e9a..379af5c95 100644
--- a/app-misc/ca-certificates/ca-certificates-20190110.3.43.ebuild
+++ b/app-misc/ca-certificates/ca-certificates-20190110.3.43.ebuild
@@ -26,7 +26,7 @@
 
 EAPI=6
 
-PYTHON_COMPAT=( python3_{6,7} )
+PYTHON_COMPAT=( python3_{6,7,8} )
 
 inherit eutils python-any-r1
 
diff --git a/app-text/iso-codes/iso-codes-4.4.ebuild b/app-text/iso-codes/iso-codes-4.4.ebuild
index 9979cda36..82d2cec0c 100644
--- a/app-text/iso-codes/iso-codes-4.4.ebuild
+++ b/app-text/iso-codes/iso-codes-4.4.ebuild
@@ -2,7 +2,7 @@
 # Distributed under the terms of the GNU General Public License v2
 
 EAPI=7
-PYTHON_COMPAT=( python3_{6,7} )
+PYTHON_COMPAT=( python3_{6,7,8} )
 PLOCALES="ab ace ach af ak am an ar as ast ay az ba bar be bg bi bn bn_IN br bs byn ca ce ch chr ckb crh cs csb cv cy da de dv dz ee el en eo es et eu fa ff fi fo fr frp fur fy ga gez gl gn gu gv ha haw he hi hr ht hu hy ia id io is it iu ja jam ka kab ki kk kl km kn ko kok ku kv kw ky lo lt lv mai mhr mi mk ml mn mo mr ms mt my na nah nb nb_NO ne nl nn nso nv oc or pa pap pi pl ps pt pt_BR ro ru rw sc sd si sk sl so son sq sr sr@latin sv sw ta te tg th ti tig tk tl tr tt tt@iqtelif ug uk ur uz ve vi wa wal wo xh yo zh_CN zh_HK zh_Hant zh_TW zu"
 
 inherit python-any-r1
diff --git a/app-text/libetonyek/libetonyek-0.1.9.ebuild b/app-text/libetonyek/libetonyek-0.1.9.ebuild
index 36303b04a..2d8ce4cde 100644
--- a/app-text/libetonyek/libetonyek-0.1.9.ebuild
+++ b/app-text/libetonyek/libetonyek-0.1.9.ebuild
@@ -22,7 +22,7 @@ RDEPEND="
 	app-text/liblangtag
 	dev-libs/librevenge
 	dev-libs/libxml2
-	=dev-util/mdds-1.4*:1=
+	>=dev-util/mdds-1.4:1=
 	sys-libs/zlib
 "
 DEPEND="${RDEPEND}
@@ -43,7 +43,7 @@ src_prepare() {
 src_configure() {
 	local myeconfargs=(
 		--disable-werror
-		--with-mdds=1.4
+		--with-mdds=1.5
 		$(use_with doc docs)
 		$(use_enable static-libs static)
 		$(use_enable test tests)
diff --git a/app-text/xmlto/xmlto-0.0.28-r1.ebuild b/app-text/xmlto/xmlto-0.0.28-r1.ebuild
index a1209c4c9..908e9a9bb 100644
--- a/app-text/xmlto/xmlto-0.0.28-r1.ebuild
+++ b/app-text/xmlto/xmlto-0.0.28-r1.ebuild
@@ -21,7 +21,7 @@ RDEPEND=">=app-text/docbook-xsl-stylesheets-1.62.0-r1
 	|| ( >=sys-apps/coreutils-6.10-r1 sys-freebsd/freebsd-ubin )
 	|| ( sys-apps/util-linux app-misc/getopt )
 	|| ( sys-apps/which sys-freebsd/freebsd-ubin )
-	text? ( || ( virtual/w3m www-client/lynx www-client/elinks ) )
+	text? ( || ( virtual/w3m www-client/lynx www-client/links ) )
 	latex? ( >=app-text/passivetex-1.25 >=dev-tex/xmltex-1.9-r2 )"
 # We only depend on flex when we patch the imput lexer.
 DEPEND="${RDEPEND}"
diff --git a/dev-java/java-config/java-config-2.2.0-r4.ebuild b/dev-java/java-config/java-config-2.2.0-r4.ebuild
index 8f0cddb82..de3f880c2 100644
--- a/dev-java/java-config/java-config-2.2.0-r4.ebuild
+++ b/dev-java/java-config/java-config-2.2.0-r4.ebuild
@@ -4,7 +4,7 @@
 EAPI=6
 
 # jython depends on java-config, so don't add it or things will break
-PYTHON_COMPAT=( python{3_6,3_7} )
+PYTHON_COMPAT=( python{3_6,3_7,3_8} )
 
 inherit distutils-r1
 
diff --git a/dev-lang/R/R-3.6.2.ebuild b/dev-lang/R/R-3.6.2.ebuild
index d9a391d0f..4829362ee 100644
--- a/dev-lang/R/R-3.6.2.ebuild
+++ b/dev-lang/R/R-3.6.2.ebuild
@@ -85,6 +85,10 @@ pkg_setup() {
 src_prepare() {
 	default
 
+	if use elibc_musl; then
+		append-cxxflags "-D__MUSL__"
+	fi
+
 	# fix packages.html for doc (gentoo bug #205103)
 	sed -e "s:../../../library:../../../../$(get_libdir)/R/library:g" \
 		-i src/library/tools/R/Rd.R || die
diff --git a/dev-lang/mono/mono-6.6.0.161.ebuild b/dev-lang/mono/mono-6.6.0.161.ebuild
index 6b165c573..134827b50 100644
--- a/dev-lang/mono/mono-6.6.0.161.ebuild
+++ b/dev-lang/mono/mono-6.6.0.161.ebuild
@@ -79,6 +79,8 @@ src_prepare() {
 
 multilib_src_configure() {
 	local myeconfargs=(
+		--disable-boehm
+		--without-sigaltstack
 		$(use_with xen xen_opt)
 		--without-ikvm-native
 		--disable-dtrace
diff --git a/dev-libs/glib/glib-2.62.5.ebuild b/dev-libs/glib/glib-2.62.5.ebuild
index 33a36d9a2..11a0f1355 100644
--- a/dev-libs/glib/glib-2.62.5.ebuild
+++ b/dev-libs/glib/glib-2.62.5.ebuild
@@ -2,7 +2,7 @@
 # Distributed under the terms of the GNU General Public License v2
 
 EAPI=7
-PYTHON_COMPAT=( python{3_6,3_7} )
+PYTHON_COMPAT=( python{3_6,3_7,3_8} )
 
 inherit flag-o-matic gnome.org gnome2-utils linux-info meson multilib multilib-minimal python-any-r1 toolchain-funcs xdg
 
diff --git a/dev-libs/icu/icu-65.1-r1.ebuild b/dev-libs/icu/icu-65.1-r1.ebuild
index 3e8ad5669..9fdc147d6 100644
--- a/dev-libs/icu/icu-65.1-r1.ebuild
+++ b/dev-libs/icu/icu-65.1-r1.ebuild
@@ -3,7 +3,7 @@
 
 EAPI=7
 
-PYTHON_COMPAT=( python3_{6,7} )
+PYTHON_COMPAT=( python3_{6,7,8} )
 inherit autotools flag-o-matic multilib-minimal python-any-r1 toolchain-funcs
 
 DESCRIPTION="International Components for Unicode"
diff --git a/dev-libs/libevdev/libevdev-1.8.0.ebuild b/dev-libs/libevdev/libevdev-1.8.0.ebuild
index c3b2cd42d..c2cdeaa1a 100644
--- a/dev-libs/libevdev/libevdev-1.8.0.ebuild
+++ b/dev-libs/libevdev/libevdev-1.8.0.ebuild
@@ -3,7 +3,7 @@
 
 EAPI=7
 
-PYTHON_COMPAT=( python{3_6,3_7} )
+PYTHON_COMPAT=( python{3_6,3_7,3_8} )
 
 inherit multilib-minimal python-any-r1
 
diff --git a/dev-libs/libffi/libffi-3.3-r1.ebuild b/dev-libs/libffi/libffi-3.3-r1.ebuild
index 3c2b1a66d..bcb0e5a2b 100644
--- a/dev-libs/libffi/libffi-3.3-r1.ebuild
+++ b/dev-libs/libffi/libffi-3.3-r1.ebuild
@@ -59,19 +59,7 @@ src_prepare() {
 
 multilib_src_configure() {
 	use userland_BSD && export HOST="${CHOST}"
-	# --includedir= path maintains a few properties:
-	# 1. have stable name across libffi versions: some packages like
-	#    dev-lang/ghc or kde-frameworks/networkmanager-qt embed
-	#    ${includedir} at build-time. Don't require those to be
-	#    rebuilt unless SONAME changes. bug #695788
-	#
-	#    We use /usr/.../${PN} (instead of former /usr/.../${P}).
-	#
-	# 2. have ${ABI}-specific location as ffi.h is target-dependent.
-	#
-	#    We use /usr/$(get_libdir)/... to have ABI identifier.
 	econf \
-		--includedir="${EPREFIX}"/usr/$(get_libdir)/${PN}/include \
 		--disable-multi-os-directory \
 		$(use_enable static-libs static) \
 		$(use_enable pax_kernel pax_emutramp) \
diff --git a/dev-libs/libgit2/libgit2-0.99.0-r1.ebuild b/dev-libs/libgit2/libgit2-0.99.0-r1.ebuild
index c986f95f0..31257dca2 100644
--- a/dev-libs/libgit2/libgit2-0.99.0-r1.ebuild
+++ b/dev-libs/libgit2/libgit2-0.99.0-r1.ebuild
@@ -3,7 +3,7 @@
 
 EAPI=7
 
-PYTHON_COMPAT=( python3_{6,7} )
+PYTHON_COMPAT=( python3_{6,7,8} )
 inherit cmake python-any-r1
 
 DESCRIPTION="A linkable library for Git"
diff --git a/dev-python/h11/h11-0.8.1.ebuild b/dev-python/h11/h11-0.8.1.ebuild
index 5ecef28bc..04b9e4da6 100644
--- a/dev-python/h11/h11-0.8.1.ebuild
+++ b/dev-python/h11/h11-0.8.1.ebuild
@@ -3,7 +3,7 @@
 
 EAPI=6
 
-PYTHON_COMPAT=( python3_{6,7} )
+PYTHON_COMPAT=( python3_{6,7,8} )
 
 inherit distutils-r1
 
diff --git a/dev-python/loguru/loguru-0.4.1.ebuild b/dev-python/loguru/loguru-0.4.1.ebuild
index aed8f0048..dc03e20e9 100644
--- a/dev-python/loguru/loguru-0.4.1.ebuild
+++ b/dev-python/loguru/loguru-0.4.1.ebuild
@@ -3,7 +3,7 @@
 
 EAPI=7
 
-PYTHON_COMPAT=( python3_{6,7} )
+PYTHON_COMPAT=( python3_{6,7,8} )
 
 inherit distutils-r1
 
diff --git a/dev-python/marshmallow/marshmallow-3.0.1.ebuild b/dev-python/marshmallow/marshmallow-3.0.1.ebuild
index 0310d0cfc..b55643801 100644
--- a/dev-python/marshmallow/marshmallow-3.0.1.ebuild
+++ b/dev-python/marshmallow/marshmallow-3.0.1.ebuild
@@ -2,7 +2,7 @@
 # Distributed under the terms of the GNU General Public License v2
 
 EAPI=7
-PYTHON_COMPAT=(  python3_{6,7} )
+PYTHON_COMPAT=(  python3_{6,7,8} )
 
 inherit distutils-r1
 
diff --git a/dev-python/oauthlib/oauthlib-3.0.1.ebuild b/dev-python/oauthlib/oauthlib-3.0.1.ebuild
index b27e06f49..5efc491b0 100644
--- a/dev-python/oauthlib/oauthlib-3.0.1.ebuild
+++ b/dev-python/oauthlib/oauthlib-3.0.1.ebuild
@@ -3,7 +3,7 @@
 
 EAPI=6
 
-PYTHON_COMPAT=( python3_{6,7} )
+PYTHON_COMPAT=( python3_{6,7,8} )
 
 inherit distutils-r1
 
diff --git a/dev-python/rfc3986/rfc3986-1.3.2.ebuild b/dev-python/rfc3986/rfc3986-1.3.2.ebuild
index abbdaf350..b92633c1d 100644
--- a/dev-python/rfc3986/rfc3986-1.3.2.ebuild
+++ b/dev-python/rfc3986/rfc3986-1.3.2.ebuild
@@ -2,7 +2,7 @@
 # Distributed under the terms of the GNU General Public License v2
 
 EAPI=7
-PYTHON_COMPAT=( python2_7 python3_{6,7} pypy3)
+PYTHON_COMPAT=( python2_7 python3_{6,7,8} pypy3)
 
 inherit distutils-r1
 
diff --git a/dev-util/ninja/ninja-1.9.0.ebuild b/dev-util/ninja/ninja-1.9.0.ebuild
index ee77f4aa4..c519d48a3 100644
--- a/dev-util/ninja/ninja-1.9.0.ebuild
+++ b/dev-util/ninja/ninja-1.9.0.ebuild
@@ -3,7 +3,7 @@
 
 EAPI=7
 
-PYTHON_COMPAT=( python{3_6,3_7} )
+PYTHON_COMPAT=( python{3_6,3_7,3_8} )
 
 inherit bash-completion-r1 elisp-common python-any-r1 toolchain-funcs
 
diff --git a/eclass/rust-toolchain.eclass b/eclass/rust-toolchain.eclass
index 66266a5b5..0f912d776 100644
--- a/eclass/rust-toolchain.eclass
+++ b/eclass/rust-toolchain.eclass
@@ -35,7 +35,8 @@ inherit multilib-build
 rust_abi() {
 	local CTARGET=${1:-${CHOST}}
 	case ${CTARGET%%*-} in
-		aarch64*)	  echo aarch64-unknown-linux-gnu;;
+		aarch64*gnu)	  echo aarch64-unknown-linux-gnu;;
+		aarch64*musl)	  echo aarch64-unknown-linux-musl;;
 		mips64*)	  echo mips64-unknown-linux-gnuabi64;;
 		powerpc64le*) echo powerpc64le-unknown-linux-gnu;;
 		powerpc64*)   echo powerpc64-unknown-linux-gnu;;
diff --git a/kde-apps/kwave/kwave-19.12.3.ebuild b/kde-apps/kwave/kwave-19.12.3.ebuild
index 90864a6bd..688149132 100644
--- a/kde-apps/kwave/kwave-19.12.3.ebuild
+++ b/kde-apps/kwave/kwave-19.12.3.ebuild
@@ -61,7 +61,6 @@ RDEPEND="
 	)
 "
 DEPEND="${RDEPEND}
-	>=kde-apps/poxml-${PVCUT}:5
 	>=dev-qt/qtconcurrent-${QTMIN}:5
 "
 
diff --git a/media-libs/libepoxy/libepoxy-1.5.4.ebuild b/media-libs/libepoxy/libepoxy-1.5.4.ebuild
index f68358117..b3e6f5fc4 100644
--- a/media-libs/libepoxy/libepoxy-1.5.4.ebuild
+++ b/media-libs/libepoxy/libepoxy-1.5.4.ebuild
@@ -9,7 +9,7 @@ if [[ ${PV} = 9999* ]]; then
 	GIT_ECLASS="git-r3"
 fi
 
-PYTHON_COMPAT=( python{3_6,3_7} )
+PYTHON_COMPAT=( python{3_6,3_7,3_8} )
 PYTHON_REQ_USE='xml(+)'
 inherit ${GIT_ECLASS} meson multilib-minimal python-any-r1
 
diff --git a/media-libs/libsndfile/libsndfile-1.0.29_pre2_p20191024.ebuild b/media-libs/libsndfile/libsndfile-1.0.29_pre2_p20191024.ebuild
index 1a1ff9ab6..0df25aeb6 100644
--- a/media-libs/libsndfile/libsndfile-1.0.29_pre2_p20191024.ebuild
+++ b/media-libs/libsndfile/libsndfile-1.0.29_pre2_p20191024.ebuild
@@ -3,7 +3,7 @@
 
 EAPI=7
 
-PYTHON_COMPAT=( python{3_6,3_7} pypy3 )
+PYTHON_COMPAT=( python{3_6,3_7,3_8} pypy3 )
 
 inherit autotools python-any-r1 multilib-minimal
 
@@ -32,7 +32,6 @@ DEPEND="${RDEPEND}"
 BDEPEND="
 	virtual/pkgconfig
 	${PYTHON_DEPS}
-	sys-devel/autogen
 	"
 
 S="${WORKDIR}/${PN}-${MY_COMMIT}"
diff --git a/media-libs/mesa/mesa-20.0.1-r1.ebuild b/media-libs/mesa/mesa-20.0.1-r1.ebuild
index 8997f8d06..d62167360 100644
--- a/media-libs/mesa/mesa-20.0.1-r1.ebuild
+++ b/media-libs/mesa/mesa-20.0.1-r1.ebuild
@@ -147,9 +147,10 @@ RDEPEND="${RDEPEND}
 # 1. List all the working slots (with min versions) in ||, newest first.
 # 2. Update the := to specify *max* version, e.g. < 10.
 # 3. Specify LLVM_MAX_SLOT, e.g. 9.
-LLVM_MAX_SLOT="9"
+LLVM_MAX_SLOT="10"
 LLVM_DEPSTR="
 	|| (
+		sys-devel/llvm:10[${MULTILIB_USEDEP}]
 		sys-devel/llvm:9[${MULTILIB_USEDEP}]
 		sys-devel/llvm:8[${MULTILIB_USEDEP}]
 	)
diff --git a/media-libs/vulkan-loader/vulkan-loader-1.2.133.ebuild b/media-libs/vulkan-loader/vulkan-loader-1.2.133.ebuild
index 8d64aee32..d198061f7 100644
--- a/media-libs/vulkan-loader/vulkan-loader-1.2.133.ebuild
+++ b/media-libs/vulkan-loader/vulkan-loader-1.2.133.ebuild
@@ -2,7 +2,7 @@
 # Distributed under the terms of the GNU General Public License v2
 
 EAPI=7
-PYTHON_COMPAT=( python3_{6,7} )
+PYTHON_COMPAT=( python3_{6,7,8} )
 
 if [[ "${PV}" == "9999" ]]; then
 	EGIT_REPO_URI="https://github.com/KhronosGroup/Vulkan-Loader.git"
diff --git a/media-video/vlc/vlc-3.0.8.ebuild b/media-video/vlc/vlc-3.0.8.ebuild
index ff05a37e1..2c5fe067d 100644
--- a/media-video/vlc/vlc-3.0.8.ebuild
+++ b/media-video/vlc/vlc-3.0.8.ebuild
@@ -59,7 +59,7 @@ BDEPEND="
 	x86? ( dev-lang/yasm )
 "
 RDEPEND="
-	net-dns/libidn:=
+	net-dns/libidn2:=
 	sys-libs/zlib[minizip]
 	virtual/libintl
 	virtual/opengl
diff --git a/net-libs/nodejs/nodejs-13.11.0.ebuild b/net-libs/nodejs/nodejs-13.11.0.ebuild
index 00513e897..6e06667d4 100644
--- a/net-libs/nodejs/nodejs-13.11.0.ebuild
+++ b/net-libs/nodejs/nodejs-13.11.0.ebuild
@@ -2,7 +2,7 @@
 # Distributed under the terms of the GNU General Public License v2
 
 EAPI=7
-PYTHON_COMPAT=( python3_{6,7} )
+PYTHON_COMPAT=( python3_{6,7,8} )
 PYTHON_REQ_USE="threads(+)"
 inherit bash-completion-r1 eutils flag-o-matic pax-utils python-any-r1 toolchain-funcs xdg-utils
 
diff --git a/sys-devel/clang/clang-10.0.0_rc3.ebuild b/sys-devel/clang/clang-10.0.0_rc3.ebuild
index 36520c115..1ee17ea23 100644
--- a/sys-devel/clang/clang-10.0.0_rc3.ebuild
+++ b/sys-devel/clang/clang-10.0.0_rc3.ebuild
@@ -32,7 +32,7 @@ LLVM_TARGET_USEDEPS=${ALL_LLVM_TARGETS[@]/%/?}
 LICENSE="Apache-2.0-with-LLVM-exceptions UoI-NCSA MIT"
 SLOT="$(ver_cut 1)"
 KEYWORDS="~amd64 ~arm ~arm64 ~ppc64 ~x86 ~amd64-linux"
-IUSE="debug default-compiler-rt default-libcxx doc +static-analyzer
+IUSE="+gcc-symlinks debug +default-compiler-rt +default-libcxx +default-libunwind doc +static-analyzer
 	test xml kernel_FreeBSD ${ALL_LLVM_TARGETS[*]}"
 REQUIRED_USE="${PYTHON_REQUIRED_USE}
 	|| ( ${ALL_LLVM_TARGETS[*]} )"
@@ -55,7 +55,8 @@ PDEPEND="
 	sys-devel/clang-common
 	~sys-devel/clang-runtime-${PV}
 	default-compiler-rt? ( =sys-libs/compiler-rt-${PV%_*}* )
-	default-libcxx? ( >=sys-libs/libcxx-${PV} )"
+	default-libcxx? ( >=sys-libs/libcxx-${PV} )
+	default-libunwind? ( >=sys-libs/llvm-libunwind-${PV} )"
 
 # least intrusive of all
 CMAKE_BUILD_TYPE=RelWithDebInfo
@@ -231,9 +232,13 @@ multilib_src_configure() {
 		# furthermore, it provides only syntax checking
 		-DCLANG_DEFAULT_OPENMP_RUNTIME=libomp
 
-		# override default stdlib and rtlib
+		# create gcc symlinks: gcc g++ cc cpp c++
+		-DCLANG_INSTALL_GCC_SYMLINKS=$(usex gcc-symlinks)
+
+		# override default stdlib, rtlib, and unwind lib
 		-DCLANG_DEFAULT_CXX_STDLIB=$(usex default-libcxx libc++ "")
 		-DCLANG_DEFAULT_RTLIB=$(usex default-compiler-rt compiler-rt "")
+		-DCLANG_DEFAULT_UNWINDLIB=$(usex default-libunwind libunwind "")
 
 		-DCLANG_ENABLE_ARCMT=$(usex static-analyzer)
 		-DCLANG_ENABLE_STATIC_ANALYZER=$(usex static-analyzer)
diff --git a/sys-devel/distcc/distcc-3.3.3-r1.ebuild b/sys-devel/distcc/distcc-3.3.3-r1.ebuild
index 826019c73..945fada9a 100644
--- a/sys-devel/distcc/distcc-3.3.3-r1.ebuild
+++ b/sys-devel/distcc/distcc-3.3.3-r1.ebuild
@@ -3,7 +3,7 @@
 
 EAPI=6
 
-PYTHON_COMPAT=( python3_6 )
+PYTHON_COMPAT=( python3_{6,8} )
 
 inherit autotools flag-o-matic python-single-r1 systemd \
 	toolchain-funcs user xdg-utils prefix
@@ -78,6 +78,7 @@ src_prepare() {
 src_configure() {
 	local myconf=(
 		--disable-Werror
+		--without-libiberty
 		--libdir=/usr/lib
 		$(use_enable ipv6 rfc2553)
 		$(use_with gtk)
diff --git a/sys-devel/llvm/llvm-10.0.0_rc3.ebuild b/sys-devel/llvm/llvm-10.0.0_rc3.ebuild
index 904db49c7..9119910b0 100644
--- a/sys-devel/llvm/llvm-10.0.0_rc3.ebuild
+++ b/sys-devel/llvm/llvm-10.0.0_rc3.ebuild
@@ -29,7 +29,7 @@ ALL_LLVM_TARGETS=( "${ALL_LLVM_TARGETS[@]/#/llvm_targets_}" )
 LICENSE="Apache-2.0-with-LLVM-exceptions UoI-NCSA BSD public-domain rc"
 SLOT="$(ver_cut 1)"
 KEYWORDS="~amd64 ~arm ~arm64 ~ppc64 ~x86 ~amd64-linux ~ppc-macos ~x64-macos ~x86-macos"
-IUSE="debug doc exegesis gold libedit +libffi ncurses test xar xml z3
+IUSE="+binutils-symlinks debug doc exegesis gold libedit +libffi ncurses test xar xml z3
 	kernel_Darwin ${ALL_LLVM_TARGETS[*]}"
 REQUIRED_USE="|| ( ${ALL_LLVM_TARGETS[*]} )"
 RESTRICT="!test? ( test )"
@@ -266,6 +266,22 @@ get_distribution_components() {
 			opt-viewer
 		)
 
+		use binutils-symlinks && out+=(
+			addr2line
+			ar
+			c++filt
+			dlltool
+			dwp
+			nm
+			objcopy
+			objdump
+			ranlib
+			readelf
+			size
+			strings
+			strip
+		)
+
 		use doc && out+=(
 			docs-dsymutil-man
 			docs-llvm-dwarfdump-man
@@ -300,6 +316,7 @@ multilib_src_configure() {
 		-DLLVM_BUILD_LLVM_DYLIB=ON
 		-DLLVM_LINK_LLVM_DYLIB=ON
 		-DLLVM_DISTRIBUTION_COMPONENTS=$(get_distribution_components)
+		-DLLVM_INSTALL_BINUTILS_SYMLINKS=$(usex binutils-symlinks)
 
 		# cheap hack: LLVM combines both anyway, and the only difference
 		# is that the former list is explicitly verified at cmake time
diff --git a/sys-libs/libblockdev/libblockdev-2.23.ebuild b/sys-libs/libblockdev/libblockdev-2.23.ebuild
index 5b61abdcb..5bcbc93b3 100644
--- a/sys-libs/libblockdev/libblockdev-2.23.ebuild
+++ b/sys-libs/libblockdev/libblockdev-2.23.ebuild
@@ -3,7 +3,7 @@
 
 EAPI=7
 
-PYTHON_COMPAT=( python3_{6,7} )
+PYTHON_COMPAT=( python3_{6,7,8} )
 inherit python-single-r1 xdg-utils
 
 DESCRIPTION="A library for manipulating block devices"
diff --git a/www-client/firefox/firefox-74.0-r1.ebuild b/www-client/firefox/firefox-74.0-r1.ebuild
index 0e6d6565e..fad301fa6 100644
--- a/www-client/firefox/firefox-74.0-r1.ebuild
+++ b/www-client/firefox/firefox-74.0-r1.ebuild
@@ -136,7 +136,7 @@ DEPEND="${CDEPEND}
 			!clang? ( sys-devel/llvm:10 )
 			clang? (
 				=sys-devel/lld-10*
-				sys-devel/llvm:10[gold]
+				sys-devel/llvm:10
 				pgo? ( =sys-libs/compiler-rt-sanitizers-10*[profile] )
 			)
 		)
