diff --git a/app-crypt/libsecret/libsecret-0.20.3.ebuild b/app-crypt/libsecret/libsecret-0.20.3.ebuild
index 07aa303ea6..da8bd74d54 100644
--- a/app-crypt/libsecret/libsecret-0.20.3.ebuild
+++ b/app-crypt/libsecret/libsecret-0.20.3.ebuild
@@ -26,8 +26,8 @@ RDEPEND="
 	crypt? ( >=dev-libs/libgcrypt-1.2.2:0=[${MULTILIB_USEDEP}] )
 	introspection? ( >=dev-libs/gobject-introspection-1.29:= )
 "
-PDEPEND=">=gnome-base/gnome-keyring-3
-"
+#PDEPEND=">=gnome-base/gnome-keyring-3
+#"
 # PDEPEND to avoid circular dep (bug #547456)
 # gnome-keyring needed at runtime as explained at https://bugs.gentoo.org/475182#c2
 # Add ksecrets to PDEPEND when it's added to portage
diff --git a/app-crypt/pinentry/pinentry-1.1.0-r3.ebuild b/app-crypt/pinentry/pinentry-1.1.0-r3.ebuild
index 053f01d8ca..17bca65318 100644
--- a/app-crypt/pinentry/pinentry-1.1.0-r3.ebuild
+++ b/app-crypt/pinentry/pinentry-1.1.0-r3.ebuild
@@ -12,7 +12,7 @@ SRC_URI="mirror://gnupg/${PN}/${P}.tar.bz2"
 LICENSE="GPL-2"
 SLOT="0"
 KEYWORDS="~alpha amd64 arm arm64 hppa ~ia64 ~mips ppc ppc64 ~riscv s390 sparc x86 ~ppc-aix ~x64-cygwin ~amd64-linux ~x86-linux ~ppc-macos ~x64-macos ~x86-macos ~sparc-solaris ~sparc64-solaris ~x64-solaris ~x86-solaris"
-IUSE="caps emacs gnome-keyring fltk gtk ncurses qt5 static"
+IUSE="caps emacs gnome-keyring fltk gtk libsecret ncurses qt5 static"
 
 DEPEND="
 	app-eselect/eselect-pinentry
@@ -21,7 +21,7 @@ DEPEND="
 	>=dev-libs/libgpg-error-1.17
 	caps? ( sys-libs/libcap )
 	fltk? ( x11-libs/fltk )
-	gnome-keyring? ( app-crypt/libsecret )
+	libsecret? ( app-crypt/libsecret )
 	gtk? ( x11-libs/gtk+:2 )
 	ncurses? ( sys-libs/ncurses:0= )
 	qt5? (
@@ -66,7 +66,7 @@ src_configure() {
 	econf \
 		$(use_enable emacs pinentry-emacs) \
 		$(use_enable fltk pinentry-fltk) \
-		$(use_enable gnome-keyring libsecret) \
+		$(use_enable libsecret) \
 		$(use_enable gnome-keyring pinentry-gnome3) \
 		$(use_enable gtk pinentry-gtk2) \
 		$(use_enable ncurses fallback-curses) \
diff --git a/app-editors/kile/kile-2.9.93.ebuild b/app-editors/kile/kile-2.9.93.ebuild
index e13393daca..89442927ba 100644
--- a/app-editors/kile/kile-2.9.93.ebuild
+++ b/app-editors/kile/kile-2.9.93.ebuild
@@ -51,7 +51,6 @@ RDEPEND="${DEPEND}
 	virtual/tex-base
 	pdf? (
 		>=app-text/texlive-core-2014
-		app-text/ghostscript-gpl
 	)
 	png? (
 		app-text/dvipng
diff --git a/app-i18n/fcitx/fcitx-4.2.9.6.ebuild b/app-i18n/fcitx/fcitx-4.2.9.6.ebuild
index abf4e1cbdb..03fca11061 100644
--- a/app-i18n/fcitx/fcitx-4.2.9.6.ebuild
+++ b/app-i18n/fcitx/fcitx-4.2.9.6.ebuild
@@ -57,7 +57,7 @@ DEPEND="dev-libs/glib:2
 		pango? ( x11-libs/pango )
 		!pango? ( media-libs/fontconfig )
 	)
-	enchant? ( app-text/enchant:0= )
+	enchant? ( app-text/enchant:= )
 	gtk2? ( x11-libs/gtk+:2 )
 	gtk3? ( x11-libs/gtk+:3 )
 	lua? ( dev-lang/lua:= )
diff --git a/app-text/xmlto/xmlto-0.0.28-r3.ebuild b/app-text/xmlto/xmlto-0.0.28-r3.ebuild
index 5602e66567..42815c690d 100644
--- a/app-text/xmlto/xmlto-0.0.28-r3.ebuild
+++ b/app-text/xmlto/xmlto-0.0.28-r3.ebuild
@@ -21,7 +21,7 @@ RDEPEND=">=app-text/docbook-xsl-stylesheets-1.62.0-r1
 	>=sys-apps/coreutils-6.10-r1
 	|| ( sys-apps/util-linux app-misc/getopt )
 	sys-apps/which
-	text? ( || ( virtual/w3m www-client/lynx www-client/elinks ) )
+	text? ( || ( virtual/w3m www-client/lynx www-client/links ) )
 	latex? ( dev-texlive/texlive-formatsextra )"
 # We only depend on flex when we patch the imput lexer.
 DEPEND="${RDEPEND}"
diff --git a/dev-lang/R/R-4.0.2.ebuild b/dev-lang/R/R-4.0.2.ebuild
index 83d1ecd938..9725452cda 100644
--- a/dev-lang/R/R-4.0.2.ebuild
+++ b/dev-lang/R/R-4.0.2.ebuild
@@ -92,6 +92,10 @@ pkg_setup() {
 src_prepare() {
 	default
 
+	if use elibc_musl; then
+		append-cxxflags "-D__MUSL__"
+	fi
+
 	# fix packages.html for doc (gentoo bug #205103)
 	sed -e "s:../../../library:../../../../$(get_libdir)/R/library:g" \
 		-i src/library/tools/R/Rd.R || die
diff --git a/dev-lang/mono/mono-6.6.0.161.ebuild b/dev-lang/mono/mono-6.6.0.161.ebuild
index bc6f0cf7f0..001c65de87 100644
--- a/dev-lang/mono/mono-6.6.0.161.ebuild
+++ b/dev-lang/mono/mono-6.6.0.161.ebuild
@@ -79,6 +79,8 @@ src_prepare() {
 
 multilib_src_configure() {
 	local myeconfargs=(
+		--disable-boehm
+		--without-sigaltstack
 		$(use_with xen xen_opt)
 		--without-ikvm-native
 		--disable-dtrace
diff --git a/dev-lang/rust/rust-1.45.2.ebuild b/dev-lang/rust/rust-1.45.2.ebuild
index 60b307bbd5..a3d7ab6e59 100644
--- a/dev-lang/rust/rust-1.45.2.ebuild
+++ b/dev-lang/rust/rust-1.45.2.ebuild
@@ -82,7 +82,6 @@ DEPEND="
 	sys-libs/zlib:=
 	!libressl? ( dev-libs/openssl:0= )
 	libressl? ( dev-libs/libressl:0= )
-	elibc_musl? ( sys-libs/libunwind )
 	system-llvm? (
 		${LLVM_DEPEND}
 	)
@@ -166,6 +165,10 @@ pre_build_checks() {
 	CHECKREQS_DISK_BUILD=${M}M check-reqs_pkg_${EBUILD_PHASE}
 }
 
+clear_vendor_checksums() {
+	sed -i 's/\("files":{\)[^}]*/\1/' vendor/$1/.cargo-checksum.json
+}
+
 pkg_pretend() {
 	pre_build_checks
 }
@@ -200,6 +203,8 @@ src_prepare() {
 			--destdir="${rust_stage0_root}" --prefix=/ || die
 	fi
 
+	clear_vendor_checksums libc
+
 	default
 }
 
@@ -519,6 +524,11 @@ src_install() {
 
 	insinto /etc/env.d/rust
 	doins "${T}/provider-${P}"
+
+	# remove some clutter
+	pushd "${ED}/usr/lib/${P}/rustlib"
+	rm components install.log manifest-* rust-installer-version uninstall.sh || die
+	popd
 }
 
 pkg_postinst() {
@@ -534,15 +544,6 @@ pkg_postinst() {
 	if has_version app-editors/gvim || has_version app-editors/vim; then
 		elog "install app-vim/rust-vim to get vim support for rust."
 	fi
-
-	if use elibc_musl; then
-		ewarn "${PN} on *-musl targets is configured with crt-static"
-		ewarn ""
-		ewarn "you will need to set RUSTFLAGS=\"-C target-feature=-crt-static\" in make.conf"
-		ewarn "to use it with portage, otherwise you may see failures like"
-		ewarn "error: cannot produce proc-macro for serde_derive v1.0.98 as the target "
-		ewarn "x86_64-unknown-linux-musl does not support these crate types"
-	fi
 }
 
 pkg_postrm() {
diff --git a/dev-libs/libffi/libffi-3.3-r1.ebuild b/dev-libs/libffi/libffi-3.3-r1.ebuild
index 996855591a..d9f612fae3 100644
--- a/dev-libs/libffi/libffi-3.3-r1.ebuild
+++ b/dev-libs/libffi/libffi-3.3-r1.ebuild
@@ -52,19 +52,7 @@ pkg_setup() {
 
 multilib_src_configure() {
 	use userland_BSD && export HOST="${CHOST}"
-	# --includedir= path maintains a few properties:
-	# 1. have stable name across libffi versions: some packages like
-	#    dev-lang/ghc or kde-frameworks/networkmanager-qt embed
-	#    ${includedir} at build-time. Don't require those to be
-	#    rebuilt unless SONAME changes. bug #695788
-	#
-	#    We use /usr/.../${PN} (instead of former /usr/.../${P}).
-	#
-	# 2. have ${ABI}-specific location as ffi.h is target-dependent.
-	#
-	#    We use /usr/$(get_libdir)/... to have ABI identifier.
 	econf \
-		--includedir="${EPREFIX}"/usr/$(get_libdir)/${PN}/include \
 		--disable-multi-os-directory \
 		$(use_enable static-libs static) \
 		$(use_enable pax_kernel pax_emutramp) \
diff --git a/dev-util/gn/gn-0.1726.ebuild b/dev-util/gn/gn-0.1726.ebuild
index 5f6a1e3c26..3a8d7bcbe4 100644
--- a/dev-util/gn/gn-0.1726.ebuild
+++ b/dev-util/gn/gn-0.1726.ebuild
@@ -2,7 +2,7 @@
 # Distributed under the terms of the GNU General Public License v2
 
 EAPI=7
-PYTHON_COMPAT=( python3_{6,7} )
+PYTHON_COMPAT=( python3_{6,7,8} )
 
 inherit ninja-utils python-any-r1 toolchain-funcs
 
diff --git a/eclass/rust-toolchain.eclass b/eclass/rust-toolchain.eclass
index 9d7360ed1c..3660a9d468 100644
--- a/eclass/rust-toolchain.eclass
+++ b/eclass/rust-toolchain.eclass
@@ -35,7 +35,8 @@ inherit multilib-build
 rust_abi() {
 	local CTARGET=${1:-${CHOST}}
 	case ${CTARGET%%*-} in
-		aarch64*)	  echo aarch64-unknown-linux-gnu;;
+		aarch64*gnu)	  echo aarch64-unknown-linux-gnu;;
+		aarch64*musl)	  echo aarch64-unknown-linux-musl;;
 		mips64*)	  echo mips64-unknown-linux-gnuabi64;;
 		powerpc64le*) echo powerpc64le-unknown-linux-gnu;;
 		powerpc64*)   echo powerpc64-unknown-linux-gnu;;
diff --git a/media-libs/libsndfile/libsndfile-1.0.29_pre2_p20191024.ebuild b/media-libs/libsndfile/libsndfile-1.0.29_pre2_p20191024.ebuild
index c9272f7fa8..377e48eff2 100644
--- a/media-libs/libsndfile/libsndfile-1.0.29_pre2_p20191024.ebuild
+++ b/media-libs/libsndfile/libsndfile-1.0.29_pre2_p20191024.ebuild
@@ -33,7 +33,6 @@ DEPEND="${RDEPEND}"
 BDEPEND="
 	virtual/pkgconfig
 	${PYTHON_DEPS}
-	sys-devel/autogen
 "
 
 S="${WORKDIR}/${PN}-${MY_COMMIT}"
diff --git a/media-libs/opencv/opencv-4.1.2-r3.ebuild b/media-libs/opencv/opencv-4.1.2-r3.ebuild
index 0f86abb641..9a46ac3241 100644
--- a/media-libs/opencv/opencv-4.1.2-r3.ebuild
+++ b/media-libs/opencv/opencv-4.1.2-r3.ebuild
@@ -3,7 +3,7 @@
 
 EAPI=7
 
-PYTHON_COMPAT=( python3_{6,7} )
+PYTHON_COMPAT=( python3_{6,7,8} )
 
 inherit java-pkg-opt-2 java-ant-2 python-r1 toolchain-funcs cmake-multilib
 
diff --git a/media-video/vlc/vlc-3.0.10-r1.ebuild b/media-video/vlc/vlc-3.0.10-r1.ebuild
index 3e8e9eef5c..8338c2d752 100644
--- a/media-video/vlc/vlc-3.0.10-r1.ebuild
+++ b/media-video/vlc/vlc-3.0.10-r1.ebuild
@@ -59,7 +59,7 @@ BDEPEND="
 	x86? ( dev-lang/yasm )
 "
 RDEPEND="
-	net-dns/libidn:=
+	net-dns/libidn2:=
 	sys-libs/zlib[minizip]
 	virtual/libintl
 	virtual/opengl
diff --git a/sys-devel/clang/clang-10.0.1.ebuild b/sys-devel/clang/clang-10.0.1.ebuild
index e50c60093b..a8cdbb01d8 100644
--- a/sys-devel/clang/clang-10.0.1.ebuild
+++ b/sys-devel/clang/clang-10.0.1.ebuild
@@ -30,7 +30,8 @@ LLVM_TARGET_USEDEPS=${ALL_LLVM_TARGETS[@]/%/?}
 LICENSE="Apache-2.0-with-LLVM-exceptions UoI-NCSA MIT"
 SLOT="$(ver_cut 1)"
 KEYWORDS="~amd64 ~arm ~arm64 ~ppc64 ~x86 ~amd64-linux"
-IUSE="debug default-compiler-rt default-libcxx default-lld doc
+IUSE="debug +default-compiler-rt +default-libcxx +default-lld
+	+default-libunwind doc +gcc-symlinks
 	+static-analyzer test xml kernel_FreeBSD ${ALL_LLVM_TARGETS[*]}"
 REQUIRED_USE="${PYTHON_REQUIRED_USE}
 	|| ( ${ALL_LLVM_TARGETS[*]} )"
@@ -43,7 +44,7 @@ RDEPEND="
 	${PYTHON_DEPS}"
 DEPEND="${RDEPEND}"
 BDEPEND="
-	dev-python/sphinx
+	doc? ( dev-python/sphinx )
 	xml? ( virtual/pkgconfig )
 	${PYTHON_DEPS}"
 RDEPEND="${RDEPEND}
@@ -54,6 +55,7 @@ PDEPEND="
 	~sys-devel/clang-runtime-${PV}
 	default-compiler-rt? ( =sys-libs/compiler-rt-${PV%_*}* )
 	default-libcxx? ( >=sys-libs/libcxx-${PV} )
+	default-libunwind? ( >=sys-libs/llvm-libunwind-${PV} )
 	default-lld? ( sys-devel/lld )"
 
 # Multilib notes:
@@ -196,13 +198,11 @@ get_distribution_components() {
 			find-all-symbols
 			modularize
 			pp-trace
-
-			# manpages
-			docs-clang-man
-			docs-clang-tools-man
 		)
 
 		use doc && out+=(
+			docs-clang-man
+			docs-clang-tools-man
 			docs-clang-html
 			docs-clang-tools-html
 		)
@@ -245,10 +245,14 @@ multilib_src_configure() {
 		# furthermore, it provides only syntax checking
 		-DCLANG_DEFAULT_OPENMP_RUNTIME=libomp
 
-		# override default stdlib and rtlib
+		# create gcc symlinks: gcc g++ cc cpp c++
+		-DCLANG_INSTALL_GCC_SYMLINKS=$(usex gcc-symlinks)
+
+		# override default stdlib and rtlib and unwind lib
 		-DCLANG_DEFAULT_CXX_STDLIB=$(usex default-libcxx libc++ "")
 		-DCLANG_DEFAULT_RTLIB=$(usex default-compiler-rt compiler-rt "")
 		-DCLANG_DEFAULT_LINKER=$(usex default-lld lld "")
+		-DCLANG_DEFAULT_UNWINDLIB=$(usex default-libunwind libunwind "")
 
 		-DCLANG_ENABLE_ARCMT=$(usex static-analyzer)
 		-DCLANG_ENABLE_STATIC_ANALYZER=$(usex static-analyzer)
@@ -262,8 +266,10 @@ multilib_src_configure() {
 		mycmakeargs+=(
 			# normally copied from LLVM_INCLUDE_DOCS but the latter
 			# is lacking value in stand-alone builds
-			-DCLANG_INCLUDE_DOCS=ON
-			-DCLANG_TOOLS_EXTRA_INCLUDE_DOCS=ON
+			-DCLANG_INCLUDE_DOCS=$(usex doc)
+			-DCLANG_TOOLS_EXTRA_INCLUDE_DOCS=$(usex doc)
+		)
+		use doc && mycmakeargs+=(
 			-DLLVM_BUILD_DOCS=ON
 			-DLLVM_ENABLE_SPHINX=ON
 			-DCLANG_INSTALL_SPHINX_HTML_DIR="${EPREFIX}/usr/share/doc/${PF}/html"
@@ -337,7 +343,7 @@ src_install() {
 	local llvm_version=$(llvm-config --version) || die
 	local clang_version=$(ver_cut 1 "${llvm_version}")
 	local clang_full_version=$(ver_cut 1-3 "${llvm_version}")
-	local clang_tools=( clang clang++ clang-cl clang-cpp )
+	local clang_tools=( clang clang++ clang-cl clang-cpp cpp )
 	local abi i
 
 	# cmake gives us:
diff --git a/sys-devel/distcc/distcc-3.3.3-r1.ebuild b/sys-devel/distcc/distcc-3.3.3-r1.ebuild
index f4d8d84f5e..16a1d7cd28 100644
--- a/sys-devel/distcc/distcc-3.3.3-r1.ebuild
+++ b/sys-devel/distcc/distcc-3.3.3-r1.ebuild
@@ -77,6 +77,7 @@ src_configure() {
 	local myconf=(
 		--disable-Werror
 		--libdir=/usr/lib
+		--without-libiberty
 		$(use_enable ipv6 rfc2553)
 		$(use_with gtk)
 		--without-gnome
diff --git a/sys-devel/distcc/files/distccd.service-1 b/sys-devel/distcc/files/distccd.service-1
index 9bd470297d..fd481026e2 100644
--- a/sys-devel/distcc/files/distccd.service-1
+++ b/sys-devel/distcc/files/distccd.service-1
@@ -4,7 +4,7 @@ After=network.target
 
 [Service]
 User=distcc
-ExecStart=/usr/bin/distccd --no-detach --daemon --port 3632 -N 15 --allow $ALLOWED_SERVERS
+ExecStart=/usr/bin/distccd --no-detach --daemon --port 3632 -N 15 $ALLOWED_SERVERS
 
 [Install]
 WantedBy=multi-user.target
diff --git a/sys-devel/llvm/llvm-10.0.1.ebuild b/sys-devel/llvm/llvm-10.0.1.ebuild
index 2bdf59cced..3cef781a38 100644
--- a/sys-devel/llvm/llvm-10.0.1.ebuild
+++ b/sys-devel/llvm/llvm-10.0.1.ebuild
@@ -29,7 +29,7 @@ ALL_LLVM_TARGETS=( "${ALL_LLVM_TARGETS[@]/#/llvm_targets_}" )
 LICENSE="Apache-2.0-with-LLVM-exceptions UoI-NCSA BSD public-domain rc"
 SLOT="$(ver_cut 1)"
 KEYWORDS="~amd64 ~arm ~arm64 ~ppc64 ~x86 ~amd64-linux ~ppc-macos ~x64-macos ~x86-macos"
-IUSE="debug doc exegesis gold libedit +libffi ncurses test xar xml z3
+IUSE="+binutils-symlinks debug doc exegesis gold libedit +libffi ncurses test xar xml z3
 	kernel_Darwin ${ALL_LLVM_TARGETS[*]}"
 REQUIRED_USE="|| ( ${ALL_LLVM_TARGETS[*]} )"
 RESTRICT="!test? ( test )"
@@ -59,10 +59,10 @@ BDEPEND="
 		>=sys-devel/binutils-apple-5.1
 	)
 	libffi? ( virtual/pkgconfig )
-	$(python_gen_any_dep '
+	doc? ( $(python_gen_any_dep '
 		dev-python/sphinx[${PYTHON_USEDEP}]
-		doc? ( dev-python/recommonmark[${PYTHON_USEDEP}] )
-	')"
+		dev-python/recommonmark[${PYTHON_USEDEP}]
+	') )"
 # There are no file collisions between these versions but having :0
 # installed means llvm-config there will take precedence.
 RDEPEND="${RDEPEND}
@@ -77,11 +77,10 @@ PATCHES=(
 )
 
 python_check_deps() {
-	if use doc; then
-		has_version -b "dev-python/recommonmark[${PYTHON_USEDEP}]" ||
-			return 1
-	fi
-	has_version -b "dev-python/sphinx[${PYTHON_USEDEP}]"
+        use doc || return 0
+
+        has_version -b "dev-python/recommonmark[${PYTHON_USEDEP}]" &&
+        has_version -b "dev-python/sphinx[${PYTHON_USEDEP}]"
 }
 
 check_distribution_components() {
@@ -275,15 +274,30 @@ get_distribution_components() {
 
 			# python modules
 			opt-viewer
+		)
 
-			# manpages
+		use binutils-symlinks && out+=(
+			addr2line
+			ar
+			c++filt
+			dlltool
+			dwp
+			nm
+			objcopy
+			objdump
+			ranlib
+			readelf
+			size
+			strings
+			strip
+                )
+
+		use doc && out+=(
+			docs-llvm-html
 			docs-dsymutil-man
 			docs-llvm-dwarfdump-man
 			docs-llvm-man
 		)
-		use doc && out+=(
-			docs-llvm-html
-		)
 
 		use gold && out+=(
 			LLVMgold
@@ -312,6 +326,7 @@ multilib_src_configure() {
 		-DLLVM_BUILD_LLVM_DYLIB=ON
 		-DLLVM_LINK_LLVM_DYLIB=ON
 		-DLLVM_DISTRIBUTION_COMPONENTS=$(get_distribution_components)
+		-DLLVM_INSTALL_BINUTILS_SYMLINKS=$(usex binutils-symlinks)
 
 		# cheap hack: LLVM combines both anyway, and the only difference
 		# is that the former list is explicitly verified at cmake time
@@ -363,11 +378,13 @@ multilib_src_configure() {
 
 	if multilib_is_native_abi; then
 		mycmakeargs+=(
-			-DLLVM_BUILD_DOCS=ON
+			-DLLVM_BUILD_DOCS=$(usex doc)
 			-DLLVM_ENABLE_OCAMLDOC=OFF
-			-DLLVM_ENABLE_SPHINX=ON
+			-DLLVM_ENABLE_SPHINX=$(usex doc)
 			-DLLVM_ENABLE_DOXYGEN=OFF
 			-DLLVM_INSTALL_UTILS=ON
+		)
+		use doc && mycmakeargs+=(
 			-DCMAKE_INSTALL_MANDIR="${EPREFIX}/usr/lib/llvm/${SLOT}/share/man"
 			-DLLVM_INSTALL_SPHINX_HTML_DIR="${EPREFIX}/usr/share/doc/${PF}/html"
 			-DSPHINX_WARNINGS_AS_ERRORS=OFF
diff --git a/www-client/firefox/firefox-79.0-r2.ebuild b/www-client/firefox/firefox-79.0-r2.ebuild
index adf273ae5c..a703dfd04c 100644
--- a/www-client/firefox/firefox-79.0-r2.ebuild
+++ b/www-client/firefox/firefox-79.0-r2.ebuild
@@ -147,7 +147,7 @@ DEPEND="${CDEPEND}
 			!clang? ( sys-devel/llvm:10 )
 			clang? (
 				=sys-devel/lld-10*
-				sys-devel/llvm:10[gold]
+				sys-devel/llvm:10
 				pgo? ( =sys-libs/compiler-rt-sanitizers-10*[profile] )
 			)
 		)
