diff --git a/app-i18n/fcitx/fcitx-4.2.9.6.ebuild b/app-i18n/fcitx/fcitx-4.2.9.6.ebuild
index 0681078c51..292fb3eb3f 100644
--- a/app-i18n/fcitx/fcitx-4.2.9.6.ebuild
+++ b/app-i18n/fcitx/fcitx-4.2.9.6.ebuild
@@ -52,7 +52,7 @@ RDEPEND="dev-libs/glib:2
 		pango? ( x11-libs/pango )
 		!pango? ( media-libs/fontconfig )
 	)
-	enchant? ( app-text/enchant:0= )
+	enchant? ( app-text/enchant:= )
 	gtk2? ( x11-libs/gtk+:2 )
 	gtk3? ( x11-libs/gtk+:3 )
 	introspection? ( dev-libs/gobject-introspection )
diff --git a/app-office/libreoffice-l10n/Manifest b/app-office/libreoffice-l10n/Manifest
index 93f8345d01..66f3cca687 100644
--- a/app-office/libreoffice-l10n/Manifest
+++ b/app-office/libreoffice-l10n/Manifest
@@ -166,3 +166,10 @@ DIST LibreOffice_6.3.4.2_Linux_x86-64_rpm_langpack_xh.tar.gz 288729 BLAKE2B e121
 DIST LibreOffice_6.3.4.2_Linux_x86-64_rpm_langpack_zh-CN.tar.gz 836794 BLAKE2B 561f7dbd4e9df8edd2675d4406ce5896f0b5ec20d79d8ee5cd0f39f429b54e1044aa01690169eb8e772235635c12a35a432a9c7f518a533d7270afb9cf8c4a9d SHA512 5193dfe554eef4b35b6dc6c4de2ac5b46a814a15dfdb07548ca11af20a3a55780819d0229e36a49b289809d0fe66ff40cd1b6dbe40668eb7394a656431ec371c
 DIST LibreOffice_6.3.4.2_Linux_x86-64_rpm_langpack_zh-TW.tar.gz 813840 BLAKE2B 745e1fa9b0656dbb653d698ee89f47bfeca7de4906f2c00fd9244c48294e5f4df53a218f4adcea506878edf85d1d7c75ba09c8411f669f72688eee7f3cb4daa6 SHA512 ac90af321b989a5ea247a0db8b3071c4ebf0aedbfb23f78cd6e141e50931eafb9bbb44f70b26256569ddc447806ee6dcc784e4d70256218dda4cce8812975e99
 DIST LibreOffice_6.3.4.2_Linux_x86-64_rpm_langpack_zu.tar.gz 319374 BLAKE2B 7654828ce36bd19e7aad8a86d58d582cd51a4bf367704744bc835a85df818d67e54ff5e92fa09c5a15b734d7aa8bf0df1d54d7237887f1495a0f33adc8975905 SHA512 b8d408ac7d3556fca53a03fd681b507e62bdcb4035f668474dc1d685390f2fed5758144b62aae526aaff2912c53d83b0e07385af48b0c941826bea052e92540f
+DIST LibreOffice_6.4.0.3_Linux_x86-64_rpm_helppack_de.tar.gz 2584685 BLAKE2B 9c08ee7990d71e0bd84ca432acb4603ff64c03789c90a93908e8a99a52ddfabe9809bf29abe17bf1e02e7d87bd1789dc9030191b9a4dbb734d6bd43da6f39425 SHA512 c3feeef1460852b401b4918423e252cab307b55556a606a95a5b697ea7e934a2883170d413b1568d33b20ab2d74d2c0fc71d33f97c2193a239236eb978335dff
+DIST LibreOffice_6.4.0.3_Linux_x86-64_rpm_helppack_en-GB.tar.gz 2354973 BLAKE2B e7b92e9dc052f33765c1cd9208d32f6142c5ba8149c937299c4167ae004b4ffb324d8db6b5ba0f4845cfa7d712ad336a048fa547139ac807a2aaa0e7bb93f00c SHA512 2b90cf45e63db25f99291e57bcc6b16c9472db966a9126d6bdd9374530d15bab125b58104a38228fe68010a70fb99d955255920dce7f20656a69db14ef4744bf
+DIST LibreOffice_6.4.0.3_Linux_x86-64_rpm_helppack_en-US.tar.gz 2178389 BLAKE2B 0e50c01892b96d0c2763f398a6469ad9e5a244ab45790403c35894f6ff94e72d3f51f8a3140f1b58f81a5784964bed3b52d47cd0b668cc4bb9310dba5952b521 SHA512 081f920d091dff5e73e61233d7b0bba31b9eb10d9f876872d7d78b6182679a8c81e3b0c69fc374c88124a388425a2b03800a53b24ccb0b1289a5c1a5b5836346
+DIST LibreOffice_6.4.0.3_Linux_x86-64_rpm_helppack_ru.tar.gz 2788722 BLAKE2B acb8e848dc6654e5e6366a59e2849c1f052df38da22b7537604e32e1965ccc7e6149fa41196cd7a3a2c56edb83628dc6c1914e11e2f4c9235e89f7fcb70fd836 SHA512 4ba7f4ee2e9816fe3cf21052009824d0f9acb578ebddeeaacbe4bec5fb6ec8891957dccf16a1d693c877ef238fc4932cf959bf07a7faa7d4311ebda7f78bc22e
+DIST LibreOffice_6.4.0.3_Linux_x86-64_rpm_langpack_de.tar.gz 21795400 BLAKE2B dfb29534e220b112e4afbb401757452798f6c625685aa907bd39c0a26a234dd71b5263817ac66163af062618947558792feb43340009bc78503de10666ec5d1a SHA512 8d6cf8fd436dbc88d03edb76e3342796d9c8e0da1896b183ff4b006d50b7d05e4a47b64867a1dfaac2cb2e4cdd2c6b7c04c50c998631e86acfe58deca6975d22
+DIST LibreOffice_6.4.0.3_Linux_x86-64_rpm_langpack_en-GB.tar.gz 7374914 BLAKE2B dbb006a32373d373719106c5019887522014ec423c505ae9d58e78dbd2f23c7abb15bb4436e587aba18a0a07b91a9fa9e8e555a1d3d22e310efffa3013c50cf1 SHA512 16e0fc76e177c0c75305a7b9d81b373238367d36a8e3477f7eeedc2b4ffc4cc62ca19ffb266f97db1042ebb7a664bc9a9bbb3b837f4698fb8b113b4b9a723a74
+DIST LibreOffice_6.4.0.3_Linux_x86-64_rpm_langpack_ru.tar.gz 1941051 BLAKE2B d38af25e4d91ce6cff039a78f79563a607411c2de745b87eb6dcd8d4ccd1676bd0872482c6891ec49d041ab51079166ef7b1544d4329f6c75b2ae0468cfcec1c SHA512 4359dacaa4916d8c4afc1c0c5778956621b3a1fc1e64e082645a783bb0414a575981ab7117e27e187ac47e5b87a00797b0b555468226bff0e34da8fe546bbd63
diff --git a/app-office/libreoffice-l10n/libreoffice-l10n-6.4.0.3.ebuild b/app-office/libreoffice-l10n/libreoffice-l10n-6.4.0.3.ebuild
new file mode 100644
index 0000000000..e69de29bb2
diff --git a/app-text/libetonyek/libetonyek-0.1.9.ebuild b/app-text/libetonyek/libetonyek-0.1.9.ebuild
index 36303b04a6..2d8ce4cdea 100644
--- a/app-text/libetonyek/libetonyek-0.1.9.ebuild
+++ b/app-text/libetonyek/libetonyek-0.1.9.ebuild
@@ -22,7 +22,7 @@ RDEPEND="
 	app-text/liblangtag
 	dev-libs/librevenge
 	dev-libs/libxml2
-	=dev-util/mdds-1.4*:1=
+	>=dev-util/mdds-1.4:1=
 	sys-libs/zlib
 "
 DEPEND="${RDEPEND}
@@ -43,7 +43,7 @@ src_prepare() {
 src_configure() {
 	local myeconfargs=(
 		--disable-werror
-		--with-mdds=1.4
+		--with-mdds=1.5
 		$(use_with doc docs)
 		$(use_enable static-libs static)
 		$(use_enable test tests)
diff --git a/app-text/xmlto/xmlto-0.0.28-r1.ebuild b/app-text/xmlto/xmlto-0.0.28-r1.ebuild
index a1209c4c9d..908e9a9bb8 100644
--- a/app-text/xmlto/xmlto-0.0.28-r1.ebuild
+++ b/app-text/xmlto/xmlto-0.0.28-r1.ebuild
@@ -21,7 +21,7 @@ RDEPEND=">=app-text/docbook-xsl-stylesheets-1.62.0-r1
 	|| ( >=sys-apps/coreutils-6.10-r1 sys-freebsd/freebsd-ubin )
 	|| ( sys-apps/util-linux app-misc/getopt )
 	|| ( sys-apps/which sys-freebsd/freebsd-ubin )
-	text? ( || ( virtual/w3m www-client/lynx www-client/elinks ) )
+	text? ( || ( virtual/w3m www-client/lynx www-client/links ) )
 	latex? ( >=app-text/passivetex-1.25 >=dev-tex/xmltex-1.9-r2 )"
 # We only depend on flex when we patch the imput lexer.
 DEPEND="${RDEPEND}"
diff --git a/dev-java/java-config/java-config-2.2.0-r4.ebuild b/dev-java/java-config/java-config-2.2.0-r4.ebuild
index 8f0cddb82b..de3f880c23 100644
--- a/dev-java/java-config/java-config-2.2.0-r4.ebuild
+++ b/dev-java/java-config/java-config-2.2.0-r4.ebuild
@@ -4,7 +4,7 @@
 EAPI=6
 
 # jython depends on java-config, so don't add it or things will break
-PYTHON_COMPAT=( python{3_6,3_7} )
+PYTHON_COMPAT=( python{3_6,3_7,3_8} )
 
 inherit distutils-r1
 
diff --git a/dev-lang/R/R-3.6.2.ebuild b/dev-lang/R/R-3.6.2.ebuild
index d9a391d0f8..4829362ee4 100644
--- a/dev-lang/R/R-3.6.2.ebuild
+++ b/dev-lang/R/R-3.6.2.ebuild
@@ -85,6 +85,10 @@ pkg_setup() {
 src_prepare() {
 	default
 
+	if use elibc_musl; then
+		append-cxxflags "-D__MUSL__"
+	fi
+
 	# fix packages.html for doc (gentoo bug #205103)
 	sed -e "s:../../../library:../../../../$(get_libdir)/R/library:g" \
 		-i src/library/tools/R/Rd.R || die
diff --git a/dev-lang/mono/mono-6.6.0.161.ebuild b/dev-lang/mono/mono-6.6.0.161.ebuild
index 6b165c5733..134827b507 100644
--- a/dev-lang/mono/mono-6.6.0.161.ebuild
+++ b/dev-lang/mono/mono-6.6.0.161.ebuild
@@ -79,6 +79,8 @@ src_prepare() {
 
 multilib_src_configure() {
 	local myeconfargs=(
+		--disable-boehm
+		--without-sigaltstack
 		$(use_with xen xen_opt)
 		--without-ikvm-native
 		--disable-dtrace
diff --git a/dev-lang/rust/rust-1.41.0.ebuild b/dev-lang/rust/rust-1.41.0.ebuild
index c3b18cdfc5..3ea38a754d 100644
--- a/dev-lang/rust/rust-1.41.0.ebuild
+++ b/dev-lang/rust/rust-1.41.0.ebuild
@@ -50,12 +50,13 @@ IUSE="clippy cpu_flags_x86_sse2 debug doc libressl nightly parallel-compiler rls
 # 3. Specify LLVM_MAX_SLOT, e.g. 9.
 LLVM_DEPEND="
 	|| (
+		sys-devel/llvm:10[llvm_targets_WebAssembly?]
 		sys-devel/llvm:9[llvm_targets_WebAssembly?]
 		wasm? ( =sys-devel/lld-9* )
 	)
-	<sys-devel/llvm-10:=
+	<sys-devel/llvm-11:=
 "
-LLVM_MAX_SLOT=9
+LLVM_MAX_SLOT=10
 
 BOOTSTRAP_DEPEND="|| ( >=dev-lang/rust-1.$(($(ver_cut 2) - 1)).0-r1 >=dev-lang/rust-bin-1.$(($(ver_cut 2) - 1)) )"
 
@@ -66,7 +67,6 @@ COMMON_DEPEND="
 	sys-libs/zlib:=
 	!libressl? ( dev-libs/openssl:0= )
 	libressl? ( dev-libs/libressl:0= )
-	elibc_musl? ( sys-libs/libunwind )
 	system-llvm? (
 		${LLVM_DEPEND}
 	)
@@ -125,6 +125,10 @@ pre_build_checks() {
 	check-reqs_pkg_setup
 }
 
+clear_vendor_checksums() {
+        sed -i 's/\("files":{\)[^}]*/\1/' vendor/$1/.cargo-checksum.json
+}
+
 pkg_pretend() {
 	pre_build_checks
 }
@@ -159,6 +163,9 @@ src_prepare() {
 	fi
 
 	default
+
+	clear_vendor_checksums libc
+	clear_vendor_checksums libgit2-sys
 }
 
 src_configure() {
@@ -280,6 +287,11 @@ src_install() {
 	env DESTDIR="${D}" "${EPYTHON}" ./x.py install -vv --config="${S}"/config.toml \
 	--exclude src/tools/miri || die
 
+	# remove some clutter
+	cd "${D}/usr/lib/rustlib"
+	rm components install.log manifest-* rust-installer-version uninstall.sh || die
+	find "${D}" -name "crt*.o" -delete || die
+
 	# bug #689562, #689160
 	rm "${D}/etc/bash_completion.d/cargo" || die
 	rmdir "${D}"/etc{/bash_completion.d,} || die
@@ -353,15 +365,6 @@ pkg_postinst() {
 	if has_version app-editors/gvim || has_version app-editors/vim; then
 		elog "install app-vim/rust-vim to get vim support for rust."
 	fi
-
-	if use elibc_musl; then
-		ewarn "${PN} on *-musl targets is configured with crt-static"
-		ewarn ""
-		ewarn "you will need to set RUSTFLAGS=\"-C target-feature=-crt-static\" in make.conf"
-		ewarn "to use it with portage, otherwise you may see failures like"
-		ewarn "error: cannot produce proc-macro for serde_derive v1.0.98 as the target "
-		ewarn "x86_64-unknown-linux-musl does not support these crate types"
-	fi
 }
 
 pkg_postrm() {
diff --git a/dev-libs/libffi/libffi-3.3-r1.ebuild b/dev-libs/libffi/libffi-3.3-r1.ebuild
index 3c2b1a66dd..bcb0e5a2bd 100644
--- a/dev-libs/libffi/libffi-3.3-r1.ebuild
+++ b/dev-libs/libffi/libffi-3.3-r1.ebuild
@@ -59,19 +59,7 @@ src_prepare() {
 
 multilib_src_configure() {
 	use userland_BSD && export HOST="${CHOST}"
-	# --includedir= path maintains a few properties:
-	# 1. have stable name across libffi versions: some packages like
-	#    dev-lang/ghc or kde-frameworks/networkmanager-qt embed
-	#    ${includedir} at build-time. Don't require those to be
-	#    rebuilt unless SONAME changes. bug #695788
-	#
-	#    We use /usr/.../${PN} (instead of former /usr/.../${P}).
-	#
-	# 2. have ${ABI}-specific location as ffi.h is target-dependent.
-	#
-	#    We use /usr/$(get_libdir)/... to have ABI identifier.
 	econf \
-		--includedir="${EPREFIX}"/usr/$(get_libdir)/${PN}/include \
 		--disable-multi-os-directory \
 		$(use_enable static-libs static) \
 		$(use_enable pax_kernel pax_emutramp) \
diff --git a/kde-apps/kwave/kwave-19.12.2.ebuild b/kde-apps/kwave/kwave-19.12.2.ebuild
index 90864a6bd2..6881491328 100644
--- a/kde-apps/kwave/kwave-19.12.2.ebuild
+++ b/kde-apps/kwave/kwave-19.12.2.ebuild
@@ -61,7 +61,6 @@ RDEPEND="
 	)
 "
 DEPEND="${RDEPEND}
-	>=kde-apps/poxml-${PVCUT}:5
 	>=dev-qt/qtconcurrent-${QTMIN}:5
 "
 
diff --git a/media-libs/libsndfile/libsndfile-1.0.29_pre2_p20191024.ebuild b/media-libs/libsndfile/libsndfile-1.0.29_pre2_p20191024.ebuild
index 1a1ff9ab65..16422513ef 100644
--- a/media-libs/libsndfile/libsndfile-1.0.29_pre2_p20191024.ebuild
+++ b/media-libs/libsndfile/libsndfile-1.0.29_pre2_p20191024.ebuild
@@ -32,7 +32,6 @@ DEPEND="${RDEPEND}"
 BDEPEND="
 	virtual/pkgconfig
 	${PYTHON_DEPS}
-	sys-devel/autogen
 	"
 
 S="${WORKDIR}/${PN}-${MY_COMMIT}"
diff --git a/media-libs/mesa/mesa-20.0.0.ebuild b/media-libs/mesa/mesa-20.0.0.ebuild
index d02cf2adf0..4710eca2f2 100644
--- a/media-libs/mesa/mesa-20.0.0.ebuild
+++ b/media-libs/mesa/mesa-20.0.0.ebuild
@@ -147,9 +147,10 @@ RDEPEND="${RDEPEND}
 # 1. List all the working slots (with min versions) in ||, newest first.
 # 2. Update the := to specify *max* version, e.g. < 10.
 # 3. Specify LLVM_MAX_SLOT, e.g. 9.
-LLVM_MAX_SLOT="9"
+LLVM_MAX_SLOT="10"
 LLVM_DEPSTR="
 	|| (
+		sys-devel/llvm:10[${MULTILIB_USEDEP}]
 		sys-devel/llvm:9[${MULTILIB_USEDEP}]
 		sys-devel/llvm:8[${MULTILIB_USEDEP}]
 	)
diff --git a/media-video/vlc/vlc-3.0.8.ebuild b/media-video/vlc/vlc-3.0.8.ebuild
index ff05a37e12..2c5fe067db 100644
--- a/media-video/vlc/vlc-3.0.8.ebuild
+++ b/media-video/vlc/vlc-3.0.8.ebuild
@@ -59,7 +59,7 @@ BDEPEND="
 	x86? ( dev-lang/yasm )
 "
 RDEPEND="
-	net-dns/libidn:=
+	net-dns/libidn2:=
 	sys-libs/zlib[minizip]
 	virtual/libintl
 	virtual/opengl
diff --git a/sys-devel/clang/clang-10.0.0.9999.ebuild b/sys-devel/clang/clang-10.0.0.9999.ebuild
index 456493136d..3faa7a2363 100644
--- a/sys-devel/clang/clang-10.0.0.9999.ebuild
+++ b/sys-devel/clang/clang-10.0.0.9999.ebuild
@@ -32,7 +32,7 @@ LLVM_TARGET_USEDEPS=${ALL_LLVM_TARGETS[@]/%/?}
 LICENSE="Apache-2.0-with-LLVM-exceptions UoI-NCSA MIT"
 SLOT="$(ver_cut 1)"
 KEYWORDS=""
-IUSE="debug default-compiler-rt default-libcxx doc +static-analyzer
+IUSE="+gcc-symlinks debug +default-compiler-rt +default-libcxx +default-libunwind doc +static-analyzer
 	test xml kernel_FreeBSD ${ALL_LLVM_TARGETS[*]}"
 REQUIRED_USE="${PYTHON_REQUIRED_USE}
 	|| ( ${ALL_LLVM_TARGETS[*]} )"
@@ -55,7 +55,8 @@ PDEPEND="
 	sys-devel/clang-common
 	~sys-devel/clang-runtime-${PV}
 	default-compiler-rt? ( =sys-libs/compiler-rt-${PV%_*}* )
-	default-libcxx? ( >=sys-libs/libcxx-${PV} )"
+	default-libcxx? ( >=sys-libs/libcxx-${PV} )
+	default-libunwind? ( >=sys-libs/llvm-libunwind-${PV} )"
 
 # least intrusive of all
 CMAKE_BUILD_TYPE=RelWithDebInfo
@@ -231,9 +232,13 @@ multilib_src_configure() {
 		# furthermore, it provides only syntax checking
 		-DCLANG_DEFAULT_OPENMP_RUNTIME=libomp
 
-		# override default stdlib and rtlib
+		# create gcc symlinks: gcc g++ cc cpp c++
+		-DCLANG_INSTALL_GCC_SYMLINKS=$(usex gcc-symlinks)
+
+		# override default stdlib, rtlib, and unwind lib
 		-DCLANG_DEFAULT_CXX_STDLIB=$(usex default-libcxx libc++ "")
 		-DCLANG_DEFAULT_RTLIB=$(usex default-compiler-rt compiler-rt "")
+		-DCLANG_DEFAULT_UNWINDLIB=$(usex default-libunwind libunwind "")
 
 		-DCLANG_ENABLE_ARCMT=$(usex static-analyzer)
 		-DCLANG_ENABLE_STATIC_ANALYZER=$(usex static-analyzer)
diff --git a/sys-devel/distcc/distcc-3.3.3-r1.ebuild b/sys-devel/distcc/distcc-3.3.3-r1.ebuild
index 826019c734..945fada9a8 100644
--- a/sys-devel/distcc/distcc-3.3.3-r1.ebuild
+++ b/sys-devel/distcc/distcc-3.3.3-r1.ebuild
@@ -3,7 +3,7 @@
 
 EAPI=6
 
-PYTHON_COMPAT=( python3_6 )
+PYTHON_COMPAT=( python3_{6,8} )
 
 inherit autotools flag-o-matic python-single-r1 systemd \
 	toolchain-funcs user xdg-utils prefix
@@ -78,6 +78,7 @@ src_prepare() {
 src_configure() {
 	local myconf=(
 		--disable-Werror
+		--without-libiberty
 		--libdir=/usr/lib
 		$(use_enable ipv6 rfc2553)
 		$(use_with gtk)
diff --git a/sys-devel/llvm/llvm-10.0.0.9999.ebuild b/sys-devel/llvm/llvm-10.0.0.9999.ebuild
index 6f27f37321..612b21019e 100644
--- a/sys-devel/llvm/llvm-10.0.0.9999.ebuild
+++ b/sys-devel/llvm/llvm-10.0.0.9999.ebuild
@@ -29,7 +29,7 @@ ALL_LLVM_TARGETS=( "${ALL_LLVM_TARGETS[@]/#/llvm_targets_}" )
 LICENSE="Apache-2.0-with-LLVM-exceptions UoI-NCSA BSD public-domain rc"
 SLOT="$(ver_cut 1)"
 KEYWORDS=""
-IUSE="debug doc exegesis gold libedit +libffi ncurses test xar xml z3
+IUSE="+binutils-symlinks debug doc exegesis gold libedit +libffi ncurses test xar xml z3
 	kernel_Darwin ${ALL_LLVM_TARGETS[*]}"
 REQUIRED_USE="|| ( ${ALL_LLVM_TARGETS[*]} )"
 RESTRICT="!test? ( test )"
@@ -301,6 +301,22 @@ get_distribution_components() {
 			opt-viewer
 		)
 
+		use binutils-symlinks && out+=(
+			addr2line
+			ar
+			c++filt
+			dlltool
+			dwp
+			nm
+			objcopy
+			objdump
+			ranlib
+			readelf
+			size
+			strings
+			strip
+		)
+
 		use doc && out+=(
 			docs-dsymutil-man
 			docs-llvm-dwarfdump-man
@@ -335,6 +351,7 @@ multilib_src_configure() {
 		-DLLVM_BUILD_LLVM_DYLIB=ON
 		-DLLVM_LINK_LLVM_DYLIB=ON
 		-DLLVM_DISTRIBUTION_COMPONENTS=$(get_distribution_components)
+		-DLLVM_INSTALL_BINUTILS_SYMLINKS=$(usex binutils-symlinks)
 
 		# cheap hack: LLVM combines both anyway, and the only difference
 		# is that the former list is explicitly verified at cmake time
diff --git a/sys-libs/libblockdev/libblockdev-2.23.ebuild b/sys-libs/libblockdev/libblockdev-2.23.ebuild
index 5b61abdcb9..5bcbc93b37 100644
--- a/sys-libs/libblockdev/libblockdev-2.23.ebuild
+++ b/sys-libs/libblockdev/libblockdev-2.23.ebuild
@@ -3,7 +3,7 @@
 
 EAPI=7
 
-PYTHON_COMPAT=( python3_{6,7} )
+PYTHON_COMPAT=( python3_{6,7,8} )
 inherit python-single-r1 xdg-utils
 
 DESCRIPTION="A library for manipulating block devices"
diff --git a/www-client/firefox/firefox-73.0.1.ebuild b/www-client/firefox/firefox-73.0.1.ebuild
index 106e5b996e..3fc6e31d56 100644
--- a/www-client/firefox/firefox-73.0.1.ebuild
+++ b/www-client/firefox/firefox-73.0.1.ebuild
@@ -38,7 +38,7 @@ if [[ "${PV}" == *_rc* ]]; then
 	MOZ_SRC_URI="${MOZ_HTTP_URI}/source/${PN}-${MOZ_PV}.source.tar.xz -> $P.tar.xz"
 fi
 
-LLVM_MAX_SLOT=9
+LLVM_MAX_SLOT=10
 
 inherit check-reqs eapi7-ver flag-o-matic toolchain-funcs eutils \
 		gnome2-utils llvm mozcoreconf-v6 pax-utils xdg-utils \
@@ -131,6 +131,15 @@ DEPEND="${CDEPEND}
 	>=sys-devel/binutils-2.30
 	sys-apps/findutils
 	|| (
+		(
+			sys-devel/clang:10
+			!clang? ( sys-devel/llvm:10 )
+			clang? (
+				=sys-devel/lld-10*
+				sys-devel/llvm:10
+				pgo? ( =sys-libs/compiler-rt-sanitizers-10*[profile] )
+			)
+		)
 		(
 			sys-devel/clang:9
 			!clang? ( sys-devel/llvm:9 )
