diff --git a/app-crypt/libsecret/libsecret-0.20.4.ebuild b/app-crypt/libsecret/libsecret-0.20.4.ebuild
index c29109aa42..052651e8fb 100644
--- a/app-crypt/libsecret/libsecret-0.20.4.ebuild
+++ b/app-crypt/libsecret/libsecret-0.20.4.ebuild
@@ -26,7 +26,7 @@ DEPEND="
 "
 RDEPEND="${DEPEND}"
 # gnome-keyring needed at runtime as explained at https://bugs.gentoo.org/475182#c2
-PDEPEND=">=gnome-base/gnome-keyring-3" # to avoid circular dep (bug #547456)
+#PDEPEND=">=gnome-base/gnome-keyring-3" # to avoid circular dep (bug #547456)
 BDEPEND="
 	dev-libs/libxslt
 	dev-util/gdbus-codegen
diff --git a/app-crypt/pinentry/pinentry-1.1.0-r3.ebuild b/app-crypt/pinentry/pinentry-1.1.0-r3.ebuild
index 0a6eb6d4c5..12994f91b3 100644
--- a/app-crypt/pinentry/pinentry-1.1.0-r3.ebuild
+++ b/app-crypt/pinentry/pinentry-1.1.0-r3.ebuild
@@ -12,7 +12,7 @@ SRC_URI="mirror://gnupg/${PN}/${P}.tar.bz2"
 LICENSE="GPL-2"
 SLOT="0"
 KEYWORDS="~alpha amd64 arm arm64 hppa ~ia64 ~mips ppc ppc64 ~riscv s390 sparc x86 ~x64-cygwin ~amd64-linux ~x86-linux ~ppc-macos ~x64-macos ~x86-macos ~sparc-solaris ~sparc64-solaris ~x64-solaris ~x86-solaris"
-IUSE="caps emacs gnome-keyring fltk gtk ncurses qt5"
+IUSE="caps emacs gnome-keyring fltk gtk libsecret ncurses qt5"
 
 DEPEND="
 	app-eselect/eselect-pinentry
@@ -21,7 +21,7 @@ DEPEND="
 	>=dev-libs/libgpg-error-1.17
 	caps? ( sys-libs/libcap )
 	fltk? ( x11-libs/fltk )
-	gnome-keyring? ( app-crypt/libsecret )
+	libsecret? ( app-crypt/libsecret )
 	gtk? ( x11-libs/gtk+:2 )
 	ncurses? ( sys-libs/ncurses:0= )
 	qt5? (
@@ -59,7 +59,7 @@ src_configure() {
 	econf \
 		$(use_enable emacs pinentry-emacs) \
 		$(use_enable fltk pinentry-fltk) \
-		$(use_enable gnome-keyring libsecret) \
+		$(use_enable libsecret) \
 		$(use_enable gnome-keyring pinentry-gnome3) \
 		$(use_enable gtk pinentry-gtk2) \
 		$(use_enable ncurses fallback-curses) \
diff --git a/app-emulation/crun/crun-0.15.ebuild b/app-emulation/crun/crun-0.15.ebuild
index 888df44272..280cabc77a 100644
--- a/app-emulation/crun/crun-0.15.ebuild
+++ b/app-emulation/crun/crun-0.15.ebuild
@@ -21,6 +21,7 @@ DEPEND="
 	>=dev-libs/yajl-2.0.0
 	caps? ( sys-libs/libcap )
 	criu? ( >=sys-process/criu-3.13 )
+	elibc_musl? ( sys-libs/argp-standalone )
 	seccomp? ( sys-libs/libseccomp )
 	systemd? ( sys-apps/systemd:= )
 "
diff --git a/app-emulation/libpod/libpod-2.1.0.ebuild b/app-emulation/libpod/libpod-2.1.0.ebuild
index 8335388f9f..670453c7f1 100644
--- a/app-emulation/libpod/libpod-2.1.0.ebuild
+++ b/app-emulation/libpod/libpod-2.1.0.ebuild
@@ -26,7 +26,6 @@ COMMON_DEPEND="
 	dev-libs/libassuan:=
 	dev-libs/libgpg-error:=
 	>=net-misc/cni-plugins-0.8.6
-	sys-fs/lvm2
 	sys-libs/libseccomp:=
 
 	apparmor? ( sys-libs/libapparmor )
diff --git a/app-emulation/libvirt/libvirt-6.10.0-r2.ebuild b/app-emulation/libvirt/libvirt-6.10.0-r2.ebuild
index d49fae5f1c..939bf632f8 100644
--- a/app-emulation/libvirt/libvirt-6.10.0-r2.ebuild
+++ b/app-emulation/libvirt/libvirt-6.10.0-r2.ebuild
@@ -102,8 +102,6 @@ RDEPEND="
 	selinux? ( >=sys-libs/libselinux-2.0.85 )
 	virt-network? (
 		net-dns/dnsmasq[script]
-		net-firewall/ebtables
-		>=net-firewall/iptables-1.4.10[ipv6]
 		net-misc/radvd
 		sys-apps/iproute2[-minimal]
 	)
diff --git a/app-emulation/spice/spice-0.14.3.ebuild b/app-emulation/spice/spice-0.14.3.ebuild
index 359c95c43d..0b0b021279 100644
--- a/app-emulation/spice/spice-0.14.3.ebuild
+++ b/app-emulation/spice/spice-0.14.3.ebuild
@@ -3,7 +3,7 @@
 
 EAPI=7
 
-PYTHON_COMPAT=( python3_{6,7,8} )
+PYTHON_COMPAT=( python3_{6,7,8,9} )
 inherit autotools python-any-r1 readme.gentoo-r1 xdg-utils
 
 DESCRIPTION="SPICE server"
diff --git a/app-emulation/virt-manager/virt-manager-3.2.0.ebuild b/app-emulation/virt-manager/virt-manager-3.2.0.ebuild
index 91e20874b0..8f7a42e14d 100644
--- a/app-emulation/virt-manager/virt-manager-3.2.0.ebuild
+++ b/app-emulation/virt-manager/virt-manager-3.2.0.ebuild
@@ -41,7 +41,6 @@ RDEPEND="${PYTHON_DEPS}
 		gnome-base/dconf
 		>=net-libs/gtk-vnc-0.3.8[gtk3(+),introspection]
 		net-misc/spice-gtk[usbredir,gtk3,introspection,sasl?]
-		net-misc/x11-ssh-askpass
 		x11-libs/gtk+:3[introspection]
 		x11-libs/gtksourceview:4[introspection]
 		x11-libs/vte:2.91[introspection]
diff --git a/app-misc/solaar/solaar-1.0.3_rc1.ebuild b/app-misc/solaar/solaar-1.0.3_rc1.ebuild
index 37fb735fd5..b0ef183ad1 100644
--- a/app-misc/solaar/solaar-1.0.3_rc1.ebuild
+++ b/app-misc/solaar/solaar-1.0.3_rc1.ebuild
@@ -4,7 +4,7 @@
 EAPI=7
 
 DISTUTILS_USE_SETUPTOOLS=bdepend
-PYTHON_COMPAT=( python3_{6,7,8} )
+PYTHON_COMPAT=( python3_{6,7,8,9} )
 
 inherit linux-info udev xdg distutils-r1
 
diff --git a/app-text/xmlto/xmlto-0.0.28-r3.ebuild b/app-text/xmlto/xmlto-0.0.28-r3.ebuild
index 5602e66567..42815c690d 100644
--- a/app-text/xmlto/xmlto-0.0.28-r3.ebuild
+++ b/app-text/xmlto/xmlto-0.0.28-r3.ebuild
@@ -21,7 +21,7 @@ RDEPEND=">=app-text/docbook-xsl-stylesheets-1.62.0-r1
 	>=sys-apps/coreutils-6.10-r1
 	|| ( sys-apps/util-linux app-misc/getopt )
 	sys-apps/which
-	text? ( || ( virtual/w3m www-client/lynx www-client/elinks ) )
+	text? ( || ( virtual/w3m www-client/lynx www-client/links ) )
 	latex? ( dev-texlive/texlive-formatsextra )"
 # We only depend on flex when we patch the imput lexer.
 DEPEND="${RDEPEND}"
diff --git a/dev-lang/rust/rust-1.48.0.ebuild b/dev-lang/rust/rust-1.48.0.ebuild
index 779372ba4e..5ff181d4d2 100644
--- a/dev-lang/rust/rust-1.48.0.ebuild
+++ b/dev-lang/rust/rust-1.48.0.ebuild
@@ -91,7 +91,6 @@ DEPEND="
 	sys-libs/zlib:=
 	!libressl? ( dev-libs/openssl:0= )
 	libressl? ( dev-libs/libressl:0= )
-	elibc_musl? ( sys-libs/libunwind:= )
 	system-llvm? (
 		${LLVM_DEPEND}
 	)
@@ -208,6 +207,10 @@ pkg_setup() {
 	fi
 }
 
+clear_vendor_checksums() {
+	sed -i 's/\("files":{\)[^}]*/\1/' vendor/$1/.cargo-checksum.json
+}
+
 src_prepare() {
 	if ! use system-bootstrap; then
 		local rust_stage0_root="${WORKDIR}"/rust-stage0
@@ -217,6 +220,8 @@ src_prepare() {
 			--destdir="${rust_stage0_root}" --prefix=/ || die
 	fi
 
+	clear_vendor_checksums libc
+
 	default
 }
 
@@ -225,7 +230,7 @@ src_configure() {
 
 	# Collect rust target names to compile standard libs for all ABIs.
 	for v in $(multilib_get_enabled_abi_pairs); do
-		rust_targets="${rust_targets},\"$(rust_abi $(get_abi_CHOST ${v##*.}))\""
+		rust_targets="${rust_targets},\"$(get_abi_CHOST ${v##*.})\""
 	done
 	if use wasm; then
 		rust_targets="${rust_targets},\"wasm32-unknown-unknown\""
@@ -258,7 +263,7 @@ src_configure() {
 		rust_stage0_root="${WORKDIR}"/rust-stage0
 	fi
 
-	rust_target="$(rust_abi)"
+	rust_target="$(get_abi_CHOST ${v##*.})"
 
 	cat <<- _EOF_ > "${S}"/config.toml
 		[llvm]
@@ -320,7 +325,7 @@ src_configure() {
 	_EOF_
 
 	for v in $(multilib_get_enabled_abi_pairs); do
-		rust_target=$(rust_abi $(get_abi_CHOST ${v##*.}))
+		rust_target=$(get_abi_CHOST ${v##*.})
 		arch_cflags="$(get_abi_CFLAGS ${v##*.})"
 
 		cat <<- _EOF_ >> "${S}"/config.env
@@ -439,7 +444,7 @@ src_compile() {
 	(
 	IFS=$'\n'
 	env $(cat "${S}"/config.env) RUST_BACKTRACE=1\
-		"${EPYTHON}" ./x.py dist -vv --config="${S}"/config.toml -j$(makeopts_jobs) || die
+		"${EPYTHON}" ./x.py build --stage 2 -vv --config="${S}"/config.toml -j$(makeopts_jobs) || die
 	)
 }
 
@@ -591,6 +596,11 @@ src_install() {
 
 	insinto /etc/env.d/rust
 	doins "${T}/provider-${P}"
+
+	# remove some clutter
+	pushd "${ED}/usr/lib/${PN}/${PV}/lib/rustlib"
+	rm components install.log manifest-* rust-installer-version uninstall.sh || die
+	popd
 }
 
 pkg_postinst() {
diff --git a/dev-libs/libffi/libffi-3.3-r2.ebuild b/dev-libs/libffi/libffi-3.3-r2.ebuild
index d6512f716d..a18acf722f 100644
--- a/dev-libs/libffi/libffi-3.3-r2.ebuild
+++ b/dev-libs/libffi/libffi-3.3-r2.ebuild
@@ -50,19 +50,7 @@ src_prepare() {
 
 multilib_src_configure() {
 	use userland_BSD && export HOST="${CHOST}"
-	# --includedir= path maintains a few properties:
-	# 1. have stable name across libffi versions: some packages like
-	#    dev-lang/ghc or kde-frameworks/networkmanager-qt embed
-	#    ${includedir} at build-time. Don't require those to be
-	#    rebuilt unless SONAME changes. bug #695788
-	#
-	#    We use /usr/.../${PN} (instead of former /usr/.../${P}).
-	#
-	# 2. have ${ABI}-specific location as ffi.h is target-dependent.
-	#
-	#    We use /usr/$(get_libdir)/... to have ABI identifier.
 	econf \
-		--includedir="${EPREFIX}"/usr/$(get_libdir)/${PN}/include \
 		--disable-multi-os-directory \
 		$(use_enable static-libs static) \
 		$(use_enable pax_kernel pax_emutramp) \
diff --git a/dev-python/pypax/pypax-0.9.5.ebuild b/dev-python/pypax/pypax-0.9.5.ebuild
index bfa8924e39..57eac2018f 100644
--- a/dev-python/pypax/pypax-0.9.5.ebuild
+++ b/dev-python/pypax/pypax-0.9.5.ebuild
@@ -4,7 +4,7 @@
 EAPI=6
 
 DISTUTILS_USE_SETUPTOOLS=no
-PYTHON_COMPAT=( python{3_6,3_7,3_8} )
+PYTHON_COMPAT=( python{3_6,3_7,3_8,3_9} )
 
 inherit distutils-r1
 
diff --git a/dev-util/flatpak-builder/flatpak-builder-1.0.11.ebuild b/dev-util/flatpak-builder/flatpak-builder-1.0.11.ebuild
index a8017f003f..27fdf16889 100644
--- a/dev-util/flatpak-builder/flatpak-builder-1.0.11.ebuild
+++ b/dev-util/flatpak-builder/flatpak-builder-1.0.11.ebuild
@@ -21,6 +21,7 @@ RDEPEND="
 	>=dev-libs/libxml2-2.4:=
 	dev-libs/json-glib:=
 	net-misc/curl:=
+	elibc_musl? ( dev-libs/libdwarf )
 	yaml? ( dev-libs/libyaml:= )
 "
 DEPEND="${RDEPEND}"
diff --git a/media-libs/vulkan-loader/vulkan-loader-1.2.154.ebuild b/media-libs/vulkan-loader/vulkan-loader-1.2.154.ebuild
index 7a0fa12561..4331608ba0 100644
--- a/media-libs/vulkan-loader/vulkan-loader-1.2.154.ebuild
+++ b/media-libs/vulkan-loader/vulkan-loader-1.2.154.ebuild
@@ -5,7 +5,7 @@ EAPI=7
 
 MY_PN=Vulkan-Loader
 CMAKE_ECLASS="cmake"
-PYTHON_COMPAT=( python3_{6,7,8} )
+PYTHON_COMPAT=( python3_{6,7,8,9} )
 inherit flag-o-matic cmake-multilib python-any-r1 toolchain-funcs
 
 if [[ ${PV} == *9999* ]]; then
diff --git a/media-video/vlc/vlc-3.0.11.1.ebuild b/media-video/vlc/vlc-3.0.11.1.ebuild
index 6de7754010..069bf329e7 100644
--- a/media-video/vlc/vlc-3.0.11.1.ebuild
+++ b/media-video/vlc/vlc-3.0.11.1.ebuild
@@ -59,7 +59,7 @@ BDEPEND="
 	x86? ( dev-lang/yasm )
 "
 RDEPEND="
-	net-dns/libidn:=
+	net-dns/libidn2:=
 	sys-libs/zlib[minizip]
 	virtual/libintl
 	virtual/opengl
diff --git a/net-libs/gtk-vnc/gtk-vnc-1.0.0.ebuild b/net-libs/gtk-vnc/gtk-vnc-1.0.0.ebuild
index 1a09512465..19e6c6f306 100644
--- a/net-libs/gtk-vnc/gtk-vnc-1.0.0.ebuild
+++ b/net-libs/gtk-vnc/gtk-vnc-1.0.0.ebuild
@@ -2,7 +2,7 @@
 # Distributed under the terms of the GNU General Public License v2
 
 EAPI=7
-PYTHON_COMPAT=( python3_{6,7,8} )
+PYTHON_COMPAT=( python3_{6,7,8,9} )
 
 inherit gnome.org vala meson python-any-r1 xdg
 
@@ -60,7 +60,7 @@ src_configure() {
 		$(meson_feature pulseaudio)
 		$(meson_feature sasl)
 		-Denable-werror=false
-		-Dwith-coroutine=auto # gthread on windows, libc ucontext elsewhere; neither has extra deps
+		-Dwith-coroutine=gthread # gthread on windows, libc ucontext elsewhere; neither has extra deps
 		$(meson_use vala with-vala)
 	)
 	meson_src_configure
diff --git a/net-misc/spice-gtk/spice-gtk-0.38.ebuild b/net-misc/spice-gtk/spice-gtk-0.38.ebuild
index f370fa8c0e..b8e0a759a9 100644
--- a/net-misc/spice-gtk/spice-gtk-0.38.ebuild
+++ b/net-misc/spice-gtk/spice-gtk-0.38.ebuild
@@ -93,6 +93,7 @@ src_configure() {
 		$(meson_feature usbredir)
 		$(meson_feature vala vapi)
 		$(meson_feature webdav)
+		-Dcoroutine=gthread
 	)
 
 	if use usbredir; then
diff --git a/sys-devel/clang/clang-11.0.1_rc2.ebuild b/sys-devel/clang/clang-11.0.1_rc2.ebuild
index c5362450b5..42714233a0 100644
--- a/sys-devel/clang/clang-11.0.1_rc2.ebuild
+++ b/sys-devel/clang/clang-11.0.1_rc2.ebuild
@@ -29,6 +29,8 @@ REQUIRED_USE="${PYTHON_REQUIRED_USE}
 	|| ( ${ALL_LLVM_TARGETS[*]} )"
 RESTRICT="!test? ( test )"
 
+IUSE="${IUSE} default-libunwind gcc-symlinks"
+
 RDEPEND="
 	~sys-devel/llvm-${PV}:${SLOT}=[debug=,${MULTILIB_USEDEP}]
 	static-analyzer? ( dev-lang/perl:* )
@@ -54,11 +56,12 @@ PDEPEND="
 	sys-devel/clang-common
 	~sys-devel/clang-runtime-${PV}
 	default-compiler-rt? ( =sys-libs/compiler-rt-${PV%_*}* )
+	default-libunwind? ( >=sys-libs/llvm-libunwind-${PV} )
 	default-libcxx? ( >=sys-libs/libcxx-${PV} )
 	default-lld? ( sys-devel/lld )"
 
 LLVM_COMPONENTS=( clang clang-tools-extra )
-LLVM_MANPAGES=build
+#LLVM_MANPAGES=build
 LLVM_TEST_COMPONENTS=(
 	llvm/lib/Testing/Support
 	llvm/utils/{lit,llvm-lit,unittest}
@@ -261,8 +264,13 @@ multilib_src_configure() {
 		# furthermore, it provides only syntax checking
 		-DCLANG_DEFAULT_OPENMP_RUNTIME=libomp
 
-		# override default stdlib and rtlib
+		# create gcc symlinks: gcc g++ cc cpp c++
+		-DCLANG_INSTALL_GCC_SYMLINKS=$(usex gcc-symlinks)
+
+
+		# override default stdlib and rtlib and libunwind
 		-DCLANG_DEFAULT_CXX_STDLIB=$(usex default-libcxx libc++ "")
+		-DCLANG_DEFAULT_UNWINDLIB=$(usex default-libunwind libunwind "")
 		-DCLANG_DEFAULT_RTLIB=$(usex default-compiler-rt compiler-rt "")
 		-DCLANG_DEFAULT_LINKER=$(usex default-lld lld "")
 
@@ -363,6 +371,7 @@ src_install() {
 	local clang_version=$(ver_cut 1 "${llvm_version}")
 	local clang_full_version=$(ver_cut 1-3 "${llvm_version}")
 	local clang_tools=( clang clang++ clang-cl clang-cpp )
+	local gcc_tools=( gcc g++ cpp )
 	local abi i
 
 	# cmake gives us:
@@ -391,6 +400,18 @@ src_install() {
 		done
 	done
 
+	if use gcc-symlinks; then
+		# and finally create target symlinks for all supported ABIs for compat
+		for abi in $(get_all_abis); do
+			local abi_chost=$(get_abi_CHOST "${abi}")
+			for i in "${gcc_tools[@]}"; do
+				dosym "clang-${clang_version}" \
+			"/usr/lib/llvm/${SLOT}/bin/${abi_chost}-${i}"
+			done
+		done
+	fi
+
+
 	# Remove unnecessary headers on FreeBSD, bug #417171
 	if use kernel_FreeBSD; then
 		rm "${ED}"/usr/lib/clang/${clang_full_version}/include/{std,float,iso,limits,tgmath,varargs}*.h || die
@@ -414,7 +435,7 @@ multilib_src_install_all() {
 	fi
 
 	docompress "/usr/lib/llvm/${SLOT}/share/man"
-	llvm_install_manpages
+	use doc && llvm_install_manpages
 	# match 'html' non-compression
 	use doc && docompress -x "/usr/share/doc/${PF}/tools-extra"
 	# +x for some reason; TODO: investigate
diff --git a/sys-devel/distcc/distcc-3.3.3-r1.ebuild b/sys-devel/distcc/distcc-3.3.3-r1.ebuild
index f4d8d84f5e..16a1d7cd28 100644
--- a/sys-devel/distcc/distcc-3.3.3-r1.ebuild
+++ b/sys-devel/distcc/distcc-3.3.3-r1.ebuild
@@ -77,6 +77,7 @@ src_configure() {
 	local myconf=(
 		--disable-Werror
 		--libdir=/usr/lib
+		--without-libiberty
 		$(use_enable ipv6 rfc2553)
 		$(use_with gtk)
 		--without-gnome
diff --git a/sys-devel/distcc/files/distccd.service-1 b/sys-devel/distcc/files/distccd.service-1
index 9bd470297d..fd481026e2 100644
--- a/sys-devel/distcc/files/distccd.service-1
+++ b/sys-devel/distcc/files/distccd.service-1
@@ -4,7 +4,7 @@ After=network.target
 
 [Service]
 User=distcc
-ExecStart=/usr/bin/distccd --no-detach --daemon --port 3632 -N 15 --allow $ALLOWED_SERVERS
+ExecStart=/usr/bin/distccd --no-detach --daemon --port 3632 -N 15 $ALLOWED_SERVERS
 
 [Install]
 WantedBy=multi-user.target
diff --git a/sys-devel/llvm/llvm-11.0.1_rc2.ebuild b/sys-devel/llvm/llvm-11.0.1_rc2.ebuild
index 9c6edd9239..50c8244faf 100644
--- a/sys-devel/llvm/llvm-11.0.1_rc2.ebuild
+++ b/sys-devel/llvm/llvm-11.0.1_rc2.ebuild
@@ -32,6 +32,8 @@ IUSE="debug doc exegesis gold libedit +libffi ncurses test xar xml z3
 REQUIRED_USE="|| ( ${ALL_LLVM_TARGETS[*]} )"
 RESTRICT="!test? ( test )"
 
+IUSE="${IUSE} binutils-symlinks"
+
 RDEPEND="
 	sys-libs/zlib:0=[${MULTILIB_USEDEP}]
 	exegesis? ( dev-libs/libpfm:= )
@@ -66,7 +68,7 @@ PDEPEND="sys-devel/llvm-common
 	gold? ( >=sys-devel/llvmgold-${SLOT} )"
 
 LLVM_COMPONENTS=( llvm )
-LLVM_MANPAGES=build
+#LLVM_MANPAGES=build
 llvm.org_set_globals
 
 python_check_deps() {
@@ -309,6 +311,23 @@ get_distribution_components() {
 				docs-llvm-man
 			)
 		fi
+
+		use binutils-symlinks && out+=(
+			addr2line
+			ar
+			c++filt
+			dlltool
+			dwp
+			nm
+			objcopy
+			objdump
+			ranlib
+			readelf
+			size
+			strings
+			strip
+                )
+
 		use doc && out+=(
 			docs-llvm-html
 		)
@@ -340,6 +359,7 @@ multilib_src_configure() {
 		-DLLVM_BUILD_LLVM_DYLIB=ON
 		-DLLVM_LINK_LLVM_DYLIB=ON
 		-DLLVM_DISTRIBUTION_COMPONENTS=$(get_distribution_components)
+		-DLLVM_INSTALL_BINUTILS_SYMLINKS=$(usex binutils-symlinks)
 
 		# cheap hack: LLVM combines both anyway, and the only difference
 		# is that the former list is explicitly verified at cmake time
@@ -502,7 +522,9 @@ multilib_src_install_all() {
 	_EOF_
 
 	docompress "/usr/lib/llvm/${SLOT}/share/man"
-	llvm_install_manpages
+	if use doc; then
+		llvm_install_manpages
+	fi
 }
 
 pkg_postinst() {
diff --git a/sys-fs/cryfs/cryfs-0.10.2.ebuild b/sys-fs/cryfs/cryfs-0.10.2.ebuild
index 5ab97c61f3..3f2118cb73 100644
--- a/sys-fs/cryfs/cryfs-0.10.2.ebuild
+++ b/sys-fs/cryfs/cryfs-0.10.2.ebuild
@@ -3,7 +3,7 @@
 
 EAPI=7
 
-PYTHON_COMPAT=( python3_{6,7,8} )
+PYTHON_COMPAT=( python3_{6,7,8,9} )
 inherit cmake flag-o-matic linux-info python-any-r1
 
 if [[ ${PV} == 9999 ]] ; then
diff --git a/sys-libs/liburing/liburing-0.7-r1.ebuild b/sys-libs/liburing/liburing-0.7-r1.ebuild
index dc6b607607..104a31b4f5 100644
--- a/sys-libs/liburing/liburing-0.7-r1.ebuild
+++ b/sys-libs/liburing/liburing-0.7-r1.ebuild
@@ -39,7 +39,7 @@ multilib_src_configure() {
 }
 
 multilib_src_compile() {
-	emake V=1 AR="$(tc-getAR)" RANLIB="$(tc-getRANLIB)"
+	emake V=1 AR="$(tc-getAR)" RANLIB="$(tc-getRANLIB)" src
 }
 
 multilib_src_install_all() {
