diff --git a/app-i18n/fcitx/fcitx-4.2.9.6.ebuild b/app-i18n/fcitx/fcitx-4.2.9.6.ebuild
index 0681078c51..292fb3eb3f 100644
--- a/app-i18n/fcitx/fcitx-4.2.9.6.ebuild
+++ b/app-i18n/fcitx/fcitx-4.2.9.6.ebuild
@@ -52,7 +52,7 @@ RDEPEND="dev-libs/glib:2
 		pango? ( x11-libs/pango )
 		!pango? ( media-libs/fontconfig )
 	)
-	enchant? ( app-text/enchant:0= )
+	enchant? ( app-text/enchant:= )
 	gtk2? ( x11-libs/gtk+:2 )
 	gtk3? ( x11-libs/gtk+:3 )
 	introspection? ( dev-libs/gobject-introspection )
diff --git a/app-text/xmlto/xmlto-0.0.28-r1.ebuild b/app-text/xmlto/xmlto-0.0.28-r1.ebuild
index f75f3368c3..d2970e2d50 100644
--- a/app-text/xmlto/xmlto-0.0.28-r1.ebuild
+++ b/app-text/xmlto/xmlto-0.0.28-r1.ebuild
@@ -21,7 +21,7 @@ RDEPEND=">=app-text/docbook-xsl-stylesheets-1.62.0-r1
 	|| ( >=sys-apps/coreutils-6.10-r1 sys-freebsd/freebsd-ubin )
 	|| ( sys-apps/util-linux app-misc/getopt )
 	|| ( sys-apps/which sys-freebsd/freebsd-ubin )
-	text? ( || ( virtual/w3m www-client/lynx www-client/elinks ) )
+	text? ( || ( virtual/w3m www-client/lynx www-client/links ) )
 	latex? ( >=app-text/passivetex-1.25 >=dev-tex/xmltex-1.9-r2 )"
 # We only depend on flex when we patch the imput lexer.
 DEPEND="${RDEPEND}"
diff --git a/dev-lang/R/R-3.6.2.ebuild b/dev-lang/R/R-3.6.2.ebuild
index b428a11127..c0e0b127c8 100644
--- a/dev-lang/R/R-3.6.2.ebuild
+++ b/dev-lang/R/R-3.6.2.ebuild
@@ -85,6 +85,10 @@ pkg_setup() {
 src_prepare() {
 	default
 
+	if use elibc_musl; then
+		append-cxxflags "-D__MUSL__"
+	fi
+
 	# fix packages.html for doc (gentoo bug #205103)
 	sed -e "s:../../../library:../../../../$(get_libdir)/R/library:g" \
 		-i src/library/tools/R/Rd.R || die
diff --git a/dev-lang/mono/mono-6.6.0.161.ebuild b/dev-lang/mono/mono-6.6.0.161.ebuild
index 93c5f532ff..e50e3dfe59 100644
--- a/dev-lang/mono/mono-6.6.0.161.ebuild
+++ b/dev-lang/mono/mono-6.6.0.161.ebuild
@@ -79,6 +79,8 @@ src_prepare() {
 
 multilib_src_configure() {
 	local myeconfargs=(
+		--disable-boehm
+		--without-sigaltstack
 		--disable-silent-rules
 		$(use_with xen xen_opt)
 		--without-ikvm-native
diff --git a/dev-lang/rust/rust-1.41.0.ebuild b/dev-lang/rust/rust-1.41.0.ebuild
index 0139d99976..e3f00b4fa7 100644
--- a/dev-lang/rust/rust-1.41.0.ebuild
+++ b/dev-lang/rust/rust-1.41.0.ebuild
@@ -50,12 +50,13 @@ IUSE="clippy cpu_flags_x86_sse2 debug doc libressl nightly parallel-compiler rls
 # 3. Specify LLVM_MAX_SLOT, e.g. 9.
 LLVM_DEPEND="
 	|| (
+		sys-devel/llvm:10[llvm_targets_WebAssembly?]
 		sys-devel/llvm:9[llvm_targets_WebAssembly?]
 		wasm? ( =sys-devel/lld-9* )
 	)
-	<sys-devel/llvm-10:=
+	<sys-devel/llvm-11:=
 "
-LLVM_MAX_SLOT=9
+LLVM_MAX_SLOT=10
 
 BOOTSTRAP_DEPEND="|| ( >=dev-lang/rust-1.$(($(ver_cut 2) - 1)).0-r1 >=dev-lang/rust-bin-1.$(($(ver_cut 2) - 1)) )"
 
@@ -66,7 +67,6 @@ COMMON_DEPEND="
 	sys-libs/zlib:=
 	!libressl? ( dev-libs/openssl:0= )
 	libressl? ( dev-libs/libressl:0= )
-	elibc_musl? ( sys-libs/libunwind )
 	system-llvm? (
 		${LLVM_DEPEND}
 	)
@@ -125,6 +125,10 @@ pre_build_checks() {
 	check-reqs_pkg_setup
 }
 
+clear_vendor_checksums() {
+        sed -i 's/\("files":{\)[^}]*/\1/' vendor/$1/.cargo-checksum.json
+}
+
 pkg_pretend() {
 	pre_build_checks
 }
@@ -159,6 +163,9 @@ src_prepare() {
 	fi
 
 	default
+
+	clear_vendor_checksums libc
+	clear_vendor_checksums libgit2-sys
 }
 
 src_configure() {
@@ -280,6 +287,11 @@ src_install() {
 	env DESTDIR="${D}" "${EPYTHON}" ./x.py install -vv --config="${S}"/config.toml \
 	--exclude src/tools/miri || die
 
+	# remove some clutter
+	cd "${D}/usr/lib/rustlib"
+	rm components install.log manifest-* rust-installer-version uninstall.sh || die
+	find "${D}" -name "crt*.o" -delete || die
+
 	# bug #689562, #689160
 	rm "${D}/etc/bash_completion.d/cargo" || die
 	rmdir "${D}"/etc{/bash_completion.d,} || die
@@ -353,15 +365,6 @@ pkg_postinst() {
 	if has_version app-editors/gvim || has_version app-editors/vim; then
 		elog "install app-vim/rust-vim to get vim support for rust."
 	fi
-
-	if use elibc_musl; then
-		ewarn "${PN} on *-musl targets is configured with crt-static"
-		ewarn ""
-		ewarn "you will need to set RUSTFLAGS=\"-C target-feature=-crt-static\" in make.conf"
-		ewarn "to use it with portage, otherwise you may see failures like"
-		ewarn "error: cannot produce proc-macro for serde_derive v1.0.98 as the target "
-		ewarn "x86_64-unknown-linux-musl does not support these crate types"
-	fi
 }
 
 pkg_postrm() {
diff --git a/dev-libs/libffi/libffi-3.3-r1.ebuild b/dev-libs/libffi/libffi-3.3-r1.ebuild
index 2377c79563..d43a929fa2 100644
--- a/dev-libs/libffi/libffi-3.3-r1.ebuild
+++ b/dev-libs/libffi/libffi-3.3-r1.ebuild
@@ -59,19 +59,7 @@ src_prepare() {
 
 multilib_src_configure() {
 	use userland_BSD && export HOST="${CHOST}"
-	# --includedir= path maintains a few properties:
-	# 1. have stable name across libffi versions: some packages like
-	#    dev-lang/ghc or kde-frameworks/networkmanager-qt embed
-	#    ${includedir} at build-time. Don't require those to be
-	#    rebuilt unless SONAME changes. bug #695788
-	#
-	#    We use /usr/.../${PN} (instead of former /usr/.../${P}).
-	#
-	# 2. have ${ABI}-specific location as ffi.h is target-dependent.
-	#
-	#    We use /usr/$(get_libdir)/... to have ABI identifier.
 	econf \
-		--includedir="${EPREFIX}"/usr/$(get_libdir)/${PN}/include \
 		--disable-multi-os-directory \
 		$(use_enable static-libs static) \
 		$(use_enable pax_kernel pax_emutramp) \
diff --git a/kde-apps/kwave/kwave-19.12.2.ebuild b/kde-apps/kwave/kwave-19.12.2.ebuild
index 90864a6bd2..6881491328 100644
--- a/kde-apps/kwave/kwave-19.12.2.ebuild
+++ b/kde-apps/kwave/kwave-19.12.2.ebuild
@@ -61,7 +61,6 @@ RDEPEND="
 	)
 "
 DEPEND="${RDEPEND}
-	>=kde-apps/poxml-${PVCUT}:5
 	>=dev-qt/qtconcurrent-${QTMIN}:5
 "
 
diff --git a/media-libs/libsndfile/libsndfile-1.0.29_pre2_p20191024.ebuild b/media-libs/libsndfile/libsndfile-1.0.29_pre2_p20191024.ebuild
index 435eb04300..551091cbb0 100644
--- a/media-libs/libsndfile/libsndfile-1.0.29_pre2_p20191024.ebuild
+++ b/media-libs/libsndfile/libsndfile-1.0.29_pre2_p20191024.ebuild
@@ -32,7 +32,6 @@ DEPEND="${RDEPEND}"
 BDEPEND="
 	virtual/pkgconfig
 	${PYTHON_DEPS}
-	sys-devel/autogen
 	"
 
 S="${WORKDIR}/${PN}-${MY_COMMIT}"
diff --git a/media-libs/mesa/mesa-20.0.0_rc1.ebuild b/media-libs/mesa/mesa-20.0.0_rc1.ebuild
index 39da329b5d..5bec3f52d1 100644
--- a/media-libs/mesa/mesa-20.0.0_rc1.ebuild
+++ b/media-libs/mesa/mesa-20.0.0_rc1.ebuild
@@ -146,9 +146,10 @@ RDEPEND="${RDEPEND}
 # 1. List all the working slots (with min versions) in ||, newest first.
 # 2. Update the := to specify *max* version, e.g. < 10.
 # 3. Specify LLVM_MAX_SLOT, e.g. 9.
-LLVM_MAX_SLOT="9"
+LLVM_MAX_SLOT="10"
 LLVM_DEPSTR="
 	|| (
+		sys-devel/llvm:10[${MULTILIB_USEDEP}]
 		sys-devel/llvm:9[${MULTILIB_USEDEP}]
 		sys-devel/llvm:8[${MULTILIB_USEDEP}]
 	)
diff --git a/media-video/vlc/vlc-3.0.8.ebuild b/media-video/vlc/vlc-3.0.8.ebuild
index 1365a0718a..e206895dd1 100644
--- a/media-video/vlc/vlc-3.0.8.ebuild
+++ b/media-video/vlc/vlc-3.0.8.ebuild
@@ -59,7 +59,7 @@ BDEPEND="
 	x86? ( dev-lang/yasm )
 "
 RDEPEND="
-	net-dns/libidn:=
+	net-dns/libidn2:=
 	sys-libs/zlib[minizip]
 	virtual/libintl
 	virtual/opengl
diff --git a/sys-devel/clang/clang-10.0.0.9999.ebuild b/sys-devel/clang/clang-10.0.0.9999.ebuild
index 03de91fd9e..8301e80cd0 100644
--- a/sys-devel/clang/clang-10.0.0.9999.ebuild
+++ b/sys-devel/clang/clang-10.0.0.9999.ebuild
@@ -32,7 +32,7 @@ LLVM_TARGET_USEDEPS=${ALL_LLVM_TARGETS[@]/%/?}
 LICENSE="Apache-2.0-with-LLVM-exceptions UoI-NCSA MIT"
 SLOT="$(ver_cut 1)"
 KEYWORDS=""
-IUSE="debug default-compiler-rt default-libcxx doc +static-analyzer
+IUSE="+gcc-symlinks debug +default-compiler-rt +default-libcxx +default-libunwind doc +static-analyzer
 	test xml kernel_FreeBSD ${ALL_LLVM_TARGETS[*]}"
 REQUIRED_USE="${PYTHON_REQUIRED_USE}
 	|| ( ${ALL_LLVM_TARGETS[*]} )"
@@ -55,7 +55,8 @@ PDEPEND="
 	sys-devel/clang-common
 	~sys-devel/clang-runtime-${PV}
 	default-compiler-rt? ( =sys-libs/compiler-rt-${PV%_*}* )
-	default-libcxx? ( >=sys-libs/libcxx-${PV} )"
+	default-libcxx? ( >=sys-libs/libcxx-${PV} )
+	default-libunwind? ( >=sys-libs/llvm-libunwind-${PV} )"
 
 # least intrusive of all
 CMAKE_BUILD_TYPE=RelWithDebInfo
@@ -231,9 +232,13 @@ multilib_src_configure() {
 		# furthermore, it provides only syntax checking
 		-DCLANG_DEFAULT_OPENMP_RUNTIME=libomp
 
-		# override default stdlib and rtlib
+		# create gcc symlinks: gcc g++ cc cpp c++
+		-DCLANG_INSTALL_GCC_SYMLINKS=$(usex gcc-symlinks)
+
+		# override default stdlib, rtlib, and unwind lib
 		-DCLANG_DEFAULT_CXX_STDLIB=$(usex default-libcxx libc++ "")
 		-DCLANG_DEFAULT_RTLIB=$(usex default-compiler-rt compiler-rt "")
+		-DCLANG_DEFAULT_UNWINDLIB=$(usex default-libunwind libunwind "")
 
 		-DCLANG_ENABLE_ARCMT=$(usex static-analyzer)
 		-DCLANG_ENABLE_STATIC_ANALYZER=$(usex static-analyzer)
diff --git a/sys-devel/distcc/distcc-3.3.3.ebuild b/sys-devel/distcc/distcc-3.3.3.ebuild
index 826019c734..64a80a215e 100644
--- a/sys-devel/distcc/distcc-3.3.3.ebuild
+++ b/sys-devel/distcc/distcc-3.3.3.ebuild
@@ -3,7 +3,7 @@
 
 EAPI=6
 
-PYTHON_COMPAT=( python3_6 )
+PYTHON_COMPAT=( python3_{6,7} )
 
 inherit autotools flag-o-matic python-single-r1 systemd \
 	toolchain-funcs user xdg-utils prefix
diff --git a/sys-devel/llvm/llvm-10.0.0.9999.ebuild b/sys-devel/llvm/llvm-10.0.0.9999.ebuild
index 66f38cdf8f..5e7f707d4a 100644
--- a/sys-devel/llvm/llvm-10.0.0.9999.ebuild
+++ b/sys-devel/llvm/llvm-10.0.0.9999.ebuild
@@ -29,7 +29,7 @@ ALL_LLVM_TARGETS=( "${ALL_LLVM_TARGETS[@]/#/llvm_targets_}" )
 LICENSE="Apache-2.0-with-LLVM-exceptions UoI-NCSA BSD public-domain rc"
 SLOT="$(ver_cut 1)"
 KEYWORDS=""
-IUSE="debug doc exegesis gold libedit +libffi ncurses test xar xml z3
+IUSE="+binutils-symlinks debug doc exegesis gold libedit +libffi ncurses test xar xml z3
 	kernel_Darwin ${ALL_LLVM_TARGETS[*]}"
 REQUIRED_USE="|| ( ${ALL_LLVM_TARGETS[*]} )"
 RESTRICT="!test? ( test )"
@@ -304,6 +304,22 @@ get_distribution_components() {
 			opt-viewer
 		)
 
+		use binutils-symlinks && out+=(
+			addr2line
+			ar
+			c++filt
+			dlltool
+			dwp
+			nm
+			objcopy
+			objdump
+			ranlib
+			readelf
+			size
+			strings
+			strip
+		)
+
 		use doc && out+=(
 			docs-dsymutil-man
 			docs-llvm-dwarfdump-man
@@ -338,6 +354,7 @@ multilib_src_configure() {
 		-DLLVM_BUILD_LLVM_DYLIB=ON
 		-DLLVM_LINK_LLVM_DYLIB=ON
 		-DLLVM_DISTRIBUTION_COMPONENTS=$(get_distribution_components)
+		-DLLVM_INSTALL_BINUTILS_SYMLINKS=$(usex binutils-symlinks)
 
 		# cheap hack: LLVM combines both anyway, and the only difference
 		# is that the former list is explicitly verified at cmake time
diff --git a/sys-libs/libapparmor/libapparmor-2.13.3.ebuild b/sys-libs/libapparmor/libapparmor-2.13.3.ebuild
index 52a2615896..56dcb4946a 100644
--- a/sys-libs/libapparmor/libapparmor-2.13.3.ebuild
+++ b/sys-libs/libapparmor/libapparmor-2.13.3.ebuild
@@ -4,7 +4,7 @@
 EAPI=6
 
 DISTUTILS_OPTIONAL=1
-PYTHON_COMPAT=( python{2_7,3_6} )
+PYTHON_COMPAT=( python{2_7,3_6,3_7} )
 GENTOO_DEPEND_ON_PERL="no"
 
 inherit autotools distutils-r1 eapi7-ver perl-functions
diff --git a/www-client/firefox/firefox-72.0.2.ebuild b/www-client/firefox/firefox-72.0.2.ebuild
index 3b3594aa44..9a6f6c806d 100644
--- a/www-client/firefox/firefox-72.0.2.ebuild
+++ b/www-client/firefox/firefox-72.0.2.ebuild
@@ -38,7 +38,7 @@ if [[ "${PV}" == *_rc* ]]; then
 	MOZ_SRC_URI="${MOZ_HTTP_URI}/source/${PN}-${MOZ_PV}.source.tar.xz -> $P.tar.xz"
 fi
 
-LLVM_MAX_SLOT=9
+LLVM_MAX_SLOT=10
 
 inherit check-reqs eapi7-ver flag-o-matic toolchain-funcs eutils \
 		gnome2-utils llvm mozcoreconf-v6 pax-utils xdg-utils \
@@ -133,12 +133,21 @@ DEPEND="${CDEPEND}
 	>=sys-devel/binutils-2.30
 	sys-apps/findutils
 	|| (
+		(
+			sys-devel/clang:10
+			!clang? ( sys-devel/llvm:10 )
+			clang? (
+				=sys-devel/lld-10*
+				sys-devel/llvm:10
+				pgo? ( =sys-libs/compiler-rt-sanitizers-10*[profile] )
+			)
+		)
 		(
 			sys-devel/clang:9
 			!clang? ( sys-devel/llvm:9 )
 			clang? (
 				=sys-devel/lld-9*
-				sys-devel/llvm:9[gold]
+				sys-devel/llvm:9
 				pgo? ( =sys-libs/compiler-rt-sanitizers-9*[profile] )
 			)
 		)
