diff --git a/app-crypt/libsecret/libsecret-0.20.4.ebuild b/app-crypt/libsecret/libsecret-0.20.4.ebuild
index 4b2b1102f1a..d52f5cdcd1e 100644
--- a/app-crypt/libsecret/libsecret-0.20.4.ebuild
+++ b/app-crypt/libsecret/libsecret-0.20.4.ebuild
@@ -29,7 +29,7 @@ DEPEND="
 "
 RDEPEND="${DEPEND}"
 # gnome-keyring needed at runtime as explained at https://bugs.gentoo.org/475182#c2
-PDEPEND=">=gnome-base/gnome-keyring-3" # to avoid circular dep (bug #547456)
+#PDEPEND=">=gnome-base/gnome-keyring-3" # to avoid circular dep (bug #547456)
 BDEPEND="
 	dev-libs/libxslt
 	dev-util/gdbus-codegen
diff --git a/app-emulation/crun/crun-0.17.ebuild b/app-emulation/crun/crun-0.17.ebuild
index bc97b2bddb1..c5115af2510 100644
--- a/app-emulation/crun/crun-0.17.ebuild
+++ b/app-emulation/crun/crun-0.17.ebuild
@@ -21,6 +21,7 @@ DEPEND="
 	>=dev-libs/yajl-2.0.0
 	caps? ( sys-libs/libcap )
 	criu? ( >=sys-process/criu-3.15 )
+	elibc_musl? ( sys-libs/argp-standalone )
 	seccomp? ( sys-libs/libseccomp )
 	systemd? ( sys-apps/systemd:= )
 "
diff --git a/app-emulation/libvirt/libvirt-7.2.0.ebuild b/app-emulation/libvirt/libvirt-7.2.0.ebuild
index 56d210fe894..0660d9cb8bb 100644
--- a/app-emulation/libvirt/libvirt-7.2.0.ebuild
+++ b/app-emulation/libvirt/libvirt-7.2.0.ebuild
@@ -108,8 +108,6 @@ RDEPEND="
 	selinux? ( >=sys-libs/libselinux-2.0.85 )
 	virt-network? (
 		net-dns/dnsmasq[dhcp,ipv6,script]
-		net-firewall/ebtables
-		>=net-firewall/iptables-1.4.10[ipv6]
 		net-misc/radvd
 		sys-apps/iproute2[-minimal]
 	)
diff --git a/app-emulation/virt-manager/virt-manager-3.2.0.ebuild b/app-emulation/virt-manager/virt-manager-3.2.0.ebuild
index 373dc31ea2b..1b9173e029c 100644
--- a/app-emulation/virt-manager/virt-manager-3.2.0.ebuild
+++ b/app-emulation/virt-manager/virt-manager-3.2.0.ebuild
@@ -40,7 +40,6 @@ RDEPEND="${PYTHON_DEPS}
 		gnome-base/dconf
 		>=net-libs/gtk-vnc-0.3.8[gtk3(+),introspection]
 		net-misc/spice-gtk[usbredir,gtk3,introspection,sasl?]
-		net-misc/x11-ssh-askpass
 		x11-libs/gtk+:3[introspection]
 		x11-libs/gtksourceview:4[introspection]
 		x11-libs/vte:2.91[introspection]
diff --git a/app-text/xmlto/xmlto-0.0.28-r3.ebuild b/app-text/xmlto/xmlto-0.0.28-r3.ebuild
index 75fca16f2d6..12963b6ccc8 100644
--- a/app-text/xmlto/xmlto-0.0.28-r3.ebuild
+++ b/app-text/xmlto/xmlto-0.0.28-r3.ebuild
@@ -22,7 +22,7 @@ RDEPEND=">=app-text/docbook-xsl-stylesheets-1.62.0-r1
 	>=sys-apps/coreutils-6.10-r1
 	|| ( sys-apps/util-linux app-misc/getopt )
 	sys-apps/which
-	text? ( || ( virtual/w3m www-client/lynx www-client/elinks ) )
+	text? ( || ( virtual/w3m www-client/lynx www-client/links ) )
 	latex? ( dev-texlive/texlive-formatsextra )"
 # We only depend on flex when we patch the imput lexer.
 DEPEND="${RDEPEND}"
diff --git a/dev-libs/icu/icu-68.2.ebuild b/dev-libs/icu/icu-68.2.ebuild
index 1b834bbdc0c..aba1f56f535 100644
--- a/dev-libs/icu/icu-68.2.ebuild
+++ b/dev-libs/icu/icu-68.2.ebuild
@@ -31,7 +31,6 @@ MULTILIB_CHOST_TOOLS=(
 PATCHES=(
 	"${FILESDIR}/${PN}-65.1-remove-bashisms.patch"
 	"${FILESDIR}/${PN}-64.2-darwin.patch"
-	"${FILESDIR}/${PN}-68.1-nonunicode.patch"
 )
 
 src_prepare() {
diff --git a/dev-libs/libffi/libffi-3.3-r2.ebuild b/dev-libs/libffi/libffi-3.3-r2.ebuild
index 3433babf3ae..553dac48abc 100644
--- a/dev-libs/libffi/libffi-3.3-r2.ebuild
+++ b/dev-libs/libffi/libffi-3.3-r2.ebuild
@@ -50,19 +50,7 @@ src_prepare() {
 
 multilib_src_configure() {
 	use userland_BSD && export HOST="${CHOST}"
-	# --includedir= path maintains a few properties:
-	# 1. have stable name across libffi versions: some packages like
-	#    dev-lang/ghc or kde-frameworks/networkmanager-qt embed
-	#    ${includedir} at build-time. Don't require those to be
-	#    rebuilt unless SONAME changes. bug #695788
-	#
-	#    We use /usr/.../${PN} (instead of former /usr/.../${P}).
-	#
-	# 2. have ${ABI}-specific location as ffi.h is target-dependent.
-	#
-	#    We use /usr/$(get_libdir)/... to have ABI identifier.
 	econf \
-		--includedir="${EPREFIX}"/usr/$(get_libdir)/${PN}/include \
 		--disable-multi-os-directory \
 		$(use_enable static-libs static) \
 		$(use_enable pax_kernel pax_emutramp) \
diff --git a/dev-python/pypax/pypax-0.9.5.ebuild b/dev-python/pypax/pypax-0.9.5.ebuild
index 845cf590fad..38eeea07c72 100644
--- a/dev-python/pypax/pypax-0.9.5.ebuild
+++ b/dev-python/pypax/pypax-0.9.5.ebuild
@@ -4,7 +4,7 @@
 EAPI=6
 
 DISTUTILS_USE_SETUPTOOLS=no
-PYTHON_COMPAT=( python{3_7,3_8} )
+PYTHON_COMPAT=( python{3_7,3_8,3_9} )
 
 inherit distutils-r1
 
diff --git a/dev-util/flatpak-builder/flatpak-builder-1.0.11.ebuild b/dev-util/flatpak-builder/flatpak-builder-1.0.11.ebuild
index a8017f003f1..27fdf168897 100644
--- a/dev-util/flatpak-builder/flatpak-builder-1.0.11.ebuild
+++ b/dev-util/flatpak-builder/flatpak-builder-1.0.11.ebuild
@@ -21,6 +21,7 @@ RDEPEND="
 	>=dev-libs/libxml2-2.4:=
 	dev-libs/json-glib:=
 	net-misc/curl:=
+	elibc_musl? ( dev-libs/libdwarf )
 	yaml? ( dev-libs/libyaml:= )
 "
 DEPEND="${RDEPEND}"
diff --git a/eclass/linux-mod.eclass b/eclass/linux-mod.eclass
index 11b0fd0cfb5..7c73787135c 100644
--- a/eclass/linux-mod.eclass
+++ b/eclass/linux-mod.eclass
@@ -683,6 +683,7 @@ linux-mod_src_compile() {
 			eval "emake HOSTCC=\"$(tc-getBUILD_CC)\" \
 						CROSS_COMPILE=${CHOST}- \
 						LDFLAGS=\"$(get_abi_LDFLAGS)\" \
+						LLVM_IAS=1 \
 						${BUILD_FIXES} \
 						${BUILD_PARAMS} \
 						${BUILD_TARGETS} " \
diff --git a/kde-plasma/kde-gtk-config/kde-gtk-config-5.21.4.ebuild b/kde-plasma/kde-gtk-config/kde-gtk-config-5.21.4.ebuild
index beb08dab9bc..b8aa0039d9f 100644
--- a/kde-plasma/kde-gtk-config/kde-gtk-config-5.21.4.ebuild
+++ b/kde-plasma/kde-gtk-config/kde-gtk-config-5.21.4.ebuild
@@ -21,7 +21,6 @@ BDEPEND="
 	dev-lang/sassc
 "
 DEPEND="
-	dev-cpp/glibmm:2
 	dev-libs/glib:2
 	>=dev-qt/qtdbus-${QTMIN}:5
 	>=dev-qt/qtgui-${QTMIN}:5
diff --git a/media-libs/mesa/mesa-21.0.2.ebuild b/media-libs/mesa/mesa-21.0.2.ebuild
index ebeac9698ee..5e3fe231752 100644
--- a/media-libs/mesa/mesa-21.0.2.ebuild
+++ b/media-libs/mesa/mesa-21.0.2.ebuild
@@ -142,9 +142,10 @@ RDEPEND="${RDEPEND}
 # 1. List all the working slots (with min versions) in ||, newest first.
 # 2. Update the := to specify *max* version, e.g. < 10.
 # 3. Specify LLVM_MAX_SLOT, e.g. 9.
-LLVM_MAX_SLOT="11"
+LLVM_MAX_SLOT="12"
 LLVM_DEPSTR="
 	|| (
+		sys-devel/llvm:12[${MULTILIB_USEDEP}]
 		sys-devel/llvm:11[${MULTILIB_USEDEP}]
 		sys-devel/llvm:10[${MULTILIB_USEDEP}]
 	)
diff --git a/net-libs/gtk-vnc/gtk-vnc-1.0.0.ebuild b/net-libs/gtk-vnc/gtk-vnc-1.0.0.ebuild
index 1229170f286..f1d4da09624 100644
--- a/net-libs/gtk-vnc/gtk-vnc-1.0.0.ebuild
+++ b/net-libs/gtk-vnc/gtk-vnc-1.0.0.ebuild
@@ -2,7 +2,7 @@
 # Distributed under the terms of the GNU General Public License v2
 
 EAPI=7
-PYTHON_COMPAT=( python3_{7,8} )
+PYTHON_COMPAT=( python3_{7,8,9} )
 
 inherit gnome.org vala meson python-any-r1 xdg
 
@@ -60,7 +60,7 @@ src_configure() {
 		$(meson_feature pulseaudio)
 		$(meson_feature sasl)
 		-Denable-werror=false
-		-Dwith-coroutine=auto # gthread on windows, libc ucontext elsewhere; neither has extra deps
+		-Dwith-coroutine=gthread # gthread on windows, libc ucontext elsewhere; neither has extra deps
 		$(meson_use vala with-vala)
 	)
 	meson_src_configure
diff --git a/net-misc/spice-gtk/spice-gtk-0.39.ebuild b/net-misc/spice-gtk/spice-gtk-0.39.ebuild
index a4abdb1a7e2..ab212d382d2 100644
--- a/net-misc/spice-gtk/spice-gtk-0.39.ebuild
+++ b/net-misc/spice-gtk/spice-gtk-0.39.ebuild
@@ -92,6 +92,7 @@ src_prepare() {
 
 src_configure() {
 	local emesonargs=(
+		-Dcoroutine=gthread
 		$(meson_feature gtk3 gtk)
 		$(meson_feature introspection)
 		$(meson_use mjpeg builtin-mjpeg)
diff --git a/sys-devel/clang/clang-12.0.0-r1.ebuild b/sys-devel/clang/clang-12.0.0-r1.ebuild
index 100f9a26b26..4ef1787c1e4 100644
--- a/sys-devel/clang/clang-12.0.0-r1.ebuild
+++ b/sys-devel/clang/clang-12.0.0-r1.ebuild
@@ -29,6 +29,8 @@ REQUIRED_USE="${PYTHON_REQUIRED_USE}
 	|| ( ${ALL_LLVM_TARGETS[*]} )"
 RESTRICT="!test? ( test )"
 
+IUSE="${IUSE} default-libunwind +flang gcc-symlinks"
+
 RDEPEND="
 	~sys-devel/llvm-${PV}:${SLOT}=[debug=,${MULTILIB_USEDEP}]
 	static-analyzer? ( dev-lang/perl:* )
@@ -52,7 +54,12 @@ RDEPEND="${RDEPEND}
 PDEPEND="
 	sys-devel/clang-common
 	~sys-devel/clang-runtime-${PV}
+	flang? (
+		dev-libs/libpgmath
+		dev-lang/flang
+		sys-devel/llvm[flang] )
 	default-compiler-rt? ( =sys-libs/compiler-rt-${PV%_*}* )
+	default-libunwind? ( >=sys-libs/llvm-libunwind-${PV} )
 	default-libcxx? ( >=sys-libs/libcxx-${PV} )
 	default-lld? ( sys-devel/lld )"
 
@@ -246,9 +253,11 @@ multilib_src_configure() {
 		-DBUILD_SHARED_LIBS=OFF
 		-DCLANG_LINK_CLANG_DYLIB=ON
 		-DLLVM_DISTRIBUTION_COMPONENTS=$(get_distribution_components)
+		-DCLANG_INSTALL_GCC_SYMLINKS=$(usex gcc-symlinks)
 
 		-DLLVM_TARGETS_TO_BUILD="${LLVM_TARGETS// /;}"
 		-DLLVM_BUILD_TESTS=$(usex test)
+		-DLLVM_ENABLE_CLASSIC_FLANG=$(usex flang)
 
 		# these are not propagated reliably, so redefine them
 		-DLLVM_ENABLE_EH=ON
@@ -262,11 +271,15 @@ multilib_src_configure() {
 		# disable using CUDA to autodetect GPU, just build for all
 		-DCMAKE_DISABLE_FIND_PACKAGE_CUDA=ON
 
-		# override default stdlib and rtlib
+		# override default stdlib and rtlib and libunwind
 		-DCLANG_DEFAULT_CXX_STDLIB=$(usex default-libcxx libc++ "")
 		-DCLANG_DEFAULT_RTLIB=$(usex default-compiler-rt compiler-rt "")
+		-DCLANG_DEFAULT_UNWINDLIB=$(usex default-libunwind libunwind "")
 		-DCLANG_DEFAULT_LINKER=$(usex default-lld lld "")
 
+		# create gcc symlinks: gcc g++ cc cpp c++
+		-DCLANG_INSTALL_GCC_SYMLINKS=$(usex gcc-symlinks)
+
 		-DCLANG_ENABLE_ARCMT=$(usex static-analyzer)
 		-DCLANG_ENABLE_STATIC_ANALYZER=$(usex static-analyzer)
 
@@ -361,9 +374,13 @@ src_install() {
 	local llvm_version=$(llvm-config --version) || die
 	local clang_version=$(ver_cut 1 "${llvm_version}")
 	local clang_full_version=$(ver_cut 1-3 "${llvm_version}")
-	local clang_tools=( clang clang++ clang-cl clang-cpp )
+	local clang_tools=( clang clang++ clang-cl clang-cpp gcc g++ cpp )
 	local abi i
 
+	if use flang; then
+		clang_tools+=( flang gfortran )
+	fi
+
 	# cmake gives us:
 	# - clang-X
 	# - clang -> clang-X
diff --git a/sys-devel/distcc/distcc-3.3.3-r3.ebuild b/sys-devel/distcc/distcc-3.3.3-r3.ebuild
index c535bb1a9fe..28ca4ccd6a4 100644
--- a/sys-devel/distcc/distcc-3.3.3-r3.ebuild
+++ b/sys-devel/distcc/distcc-3.3.3-r3.ebuild
@@ -75,6 +75,7 @@ src_configure() {
 	local myconf=(
 		--disable-Werror
 		--libdir=/usr/lib
+		--without-libiberty
 		$(use_enable ipv6 rfc2553)
 		$(use_with gtk)
 		--without-gnome
diff --git a/sys-devel/distcc/files/distccd.service-1 b/sys-devel/distcc/files/distccd.service-1
index 9bd470297df..fd481026e27 100644
--- a/sys-devel/distcc/files/distccd.service-1
+++ b/sys-devel/distcc/files/distccd.service-1
@@ -4,7 +4,7 @@ After=network.target
 
 [Service]
 User=distcc
-ExecStart=/usr/bin/distccd --no-detach --daemon --port 3632 -N 15 --allow $ALLOWED_SERVERS
+ExecStart=/usr/bin/distccd --no-detach --daemon --port 3632 -N 15 $ALLOWED_SERVERS
 
 [Install]
 WantedBy=multi-user.target
diff --git a/sys-devel/llvm/llvm-12.0.0.ebuild b/sys-devel/llvm/llvm-12.0.0.ebuild
index e6498e6d32c..e51ebdf5f94 100644
--- a/sys-devel/llvm/llvm-12.0.0.ebuild
+++ b/sys-devel/llvm/llvm-12.0.0.ebuild
@@ -32,6 +32,8 @@ IUSE="debug doc exegesis gold libedit +libffi ncurses test xar xml z3
 REQUIRED_USE="|| ( ${ALL_LLVM_TARGETS[*]} )"
 RESTRICT="!test? ( test )"
 
+IUSE="${IUSE} +flang binutils-symlinks"
+
 RDEPEND="
 	sys-libs/zlib:0=[${MULTILIB_USEDEP}]
 	exegesis? ( dev-libs/libpfm:= )
@@ -318,6 +320,23 @@ get_distribution_components() {
 				docs-llvm-man
 			)
 		fi
+
+		use binutils-symlinks && out+=(
+			addr2line
+			ar
+			c++filt
+			dlltool
+			dwp
+			nm
+			objcopy
+			objdump
+			ranlib
+			readelf
+			size
+			strings
+			strip
+                )
+
 		use doc && out+=(
 			docs-llvm-html
 		)
@@ -349,6 +368,7 @@ multilib_src_configure() {
 		-DLLVM_BUILD_LLVM_DYLIB=ON
 		-DLLVM_LINK_LLVM_DYLIB=ON
 		-DLLVM_DISTRIBUTION_COMPONENTS=$(get_distribution_components)
+		-DLLVM_INSTALL_BINUTILS_SYMLINKS=$(usex binutils-symlinks)
 
 		# cheap hack: LLVM combines both anyway, and the only difference
 		# is that the former list is explicitly verified at cmake time
@@ -365,6 +385,7 @@ multilib_src_configure() {
 		-DLLVM_ENABLE_EH=ON
 		-DLLVM_ENABLE_RTTI=ON
 		-DLLVM_ENABLE_Z3_SOLVER=$(usex z3)
+		-DLLVM_ENABLE_CLASSIC_FLANG=$(usex flang)
 
 		-DLLVM_HOST_TRIPLE="${CHOST}"
 
diff --git a/sys-libs/liburing/liburing-0.7-r1.ebuild b/sys-libs/liburing/liburing-0.7-r1.ebuild
index 715212503eb..3eb405715f9 100644
--- a/sys-libs/liburing/liburing-0.7-r1.ebuild
+++ b/sys-libs/liburing/liburing-0.7-r1.ebuild
@@ -43,7 +43,7 @@ multilib_src_configure() {
 }
 
 multilib_src_compile() {
-	emake V=1 AR="$(tc-getAR)" RANLIB="$(tc-getRANLIB)"
+	emake V=1 AR="$(tc-getAR)" RANLIB="$(tc-getRANLIB)" src
 }
 
 multilib_src_install_all() {
diff --git a/www-client/firefox/firefox-87.0-r1.ebuild b/www-client/firefox/firefox-87.0-r1.ebuild
index eac0a403095..f874bb5a64d 100644
--- a/www-client/firefox/firefox-87.0-r1.ebuild
+++ b/www-client/firefox/firefox-87.0-r1.ebuild
@@ -5,7 +5,7 @@ EAPI="7"
 
 FIREFOX_PATCHSET="firefox-87-patches-04.tar.xz"
 
-LLVM_MAX_SLOT=11
+LLVM_MAX_SLOT=12
 
 PYTHON_COMPAT=( python3_{7..9} )
 PYTHON_REQ_USE="ncurses,sqlite,ssl"
@@ -77,6 +77,14 @@ BDEPEND="${PYTHON_DEPS}
 	virtual/pkgconfig
 	>=virtual/rust-1.47.0
 	|| (
+		(
+			sys-devel/clang:12
+			sys-devel/llvm:12
+			clang? (
+				=sys-devel/lld-12*
+				pgo? ( =sys-libs/compiler-rt-sanitizers-12*[profile] )
+			)
+		)
 		(
 			sys-devel/clang:11
 			sys-devel/llvm:11
