diff --git a/app-crypt/libsecret/libsecret-0.20.3.ebuild b/app-crypt/libsecret/libsecret-0.20.3.ebuild
index e5125875d2..b1b5a96f7d 100644
--- a/app-crypt/libsecret/libsecret-0.20.3.ebuild
+++ b/app-crypt/libsecret/libsecret-0.20.3.ebuild
@@ -26,8 +26,8 @@ RDEPEND="
 	crypt? ( >=dev-libs/libgcrypt-1.2.2:0=[${MULTILIB_USEDEP}] )
 	introspection? ( >=dev-libs/gobject-introspection-1.29:= )
 "
-PDEPEND=">=gnome-base/gnome-keyring-3
-"
+#PDEPEND=">=gnome-base/gnome-keyring-3
+#"
 # PDEPEND to avoid circular dep (bug #547456)
 # gnome-keyring needed at runtime as explained at https://bugs.gentoo.org/475182#c2
 # Add ksecrets to PDEPEND when it's added to portage
diff --git a/app-crypt/pinentry/pinentry-1.1.0-r3.ebuild b/app-crypt/pinentry/pinentry-1.1.0-r3.ebuild
index efc797548a..1bc61719c4 100644
--- a/app-crypt/pinentry/pinentry-1.1.0-r3.ebuild
+++ b/app-crypt/pinentry/pinentry-1.1.0-r3.ebuild
@@ -12,7 +12,7 @@ SRC_URI="mirror://gnupg/${PN}/${P}.tar.bz2"
 LICENSE="GPL-2"
 SLOT="0"
 KEYWORDS="~alpha amd64 arm arm64 hppa ~ia64 ~mips ppc ppc64 ~riscv s390 sparc x86 ~ppc-aix ~x64-cygwin ~amd64-linux ~x86-linux ~ppc-macos ~x64-macos ~x86-macos ~sparc-solaris ~sparc64-solaris ~x64-solaris ~x86-solaris"
-IUSE="caps emacs gnome-keyring fltk gtk ncurses qt5"
+IUSE="caps emacs gnome-keyring fltk gtk libsecret ncurses qt5"
 
 DEPEND="
 	app-eselect/eselect-pinentry
@@ -21,7 +21,7 @@ DEPEND="
 	>=dev-libs/libgpg-error-1.17
 	caps? ( sys-libs/libcap )
 	fltk? ( x11-libs/fltk )
-	gnome-keyring? ( app-crypt/libsecret )
+	libsecret? ( app-crypt/libsecret )
 	gtk? ( x11-libs/gtk+:2 )
 	ncurses? ( sys-libs/ncurses:0= )
 	qt5? (
@@ -59,7 +59,7 @@ src_configure() {
 	econf \
 		$(use_enable emacs pinentry-emacs) \
 		$(use_enable fltk pinentry-fltk) \
-		$(use_enable gnome-keyring libsecret) \
+		$(use_enable libsecret) \
 		$(use_enable gnome-keyring pinentry-gnome3) \
 		$(use_enable gtk pinentry-gtk2) \
 		$(use_enable ncurses fallback-curses) \
diff --git a/app-emulation/docker/docker-19.03.13-r2.ebuild b/app-emulation/docker/docker-19.03.13-r2.ebuild
index 466ac242d0..b06f42310a 100644
--- a/app-emulation/docker/docker-19.03.13-r2.ebuild
+++ b/app-emulation/docker/docker-19.03.13-r2.ebuild
@@ -71,9 +71,7 @@ CONFIG_CHECK="
 	~CGROUPS ~CGROUP_CPUACCT ~CGROUP_DEVICE ~CGROUP_FREEZER ~CGROUP_SCHED ~CPUSETS ~MEMCG
 	~KEYS
 	~VETH ~BRIDGE ~BRIDGE_NETFILTER
-	~IP_NF_FILTER ~IP_NF_TARGET_MASQUERADE
-	~NETFILTER_XT_MATCH_ADDRTYPE ~NETFILTER_XT_MATCH_CONNTRACK ~NETFILTER_XT_MATCH_IPVS
-	~IP_NF_NAT ~NF_NAT
+	~NF_NAT
 	~POSIX_MQUEUE
 
 	~USER_NS
diff --git a/app-i18n/fcitx/fcitx-4.2.9.6.ebuild b/app-i18n/fcitx/fcitx-4.2.9.6.ebuild
index abf4e1cbdb..03fca11061 100644
--- a/app-i18n/fcitx/fcitx-4.2.9.6.ebuild
+++ b/app-i18n/fcitx/fcitx-4.2.9.6.ebuild
@@ -57,7 +57,7 @@ DEPEND="dev-libs/glib:2
 		pango? ( x11-libs/pango )
 		!pango? ( media-libs/fontconfig )
 	)
-	enchant? ( app-text/enchant:0= )
+	enchant? ( app-text/enchant:= )
 	gtk2? ( x11-libs/gtk+:2 )
 	gtk3? ( x11-libs/gtk+:3 )
 	lua? ( dev-lang/lua:= )
diff --git a/app-text/xmlto/xmlto-0.0.28-r3.ebuild b/app-text/xmlto/xmlto-0.0.28-r3.ebuild
index 5602e66567..42815c690d 100644
--- a/app-text/xmlto/xmlto-0.0.28-r3.ebuild
+++ b/app-text/xmlto/xmlto-0.0.28-r3.ebuild
@@ -21,7 +21,7 @@ RDEPEND=">=app-text/docbook-xsl-stylesheets-1.62.0-r1
 	>=sys-apps/coreutils-6.10-r1
 	|| ( sys-apps/util-linux app-misc/getopt )
 	sys-apps/which
-	text? ( || ( virtual/w3m www-client/lynx www-client/elinks ) )
+	text? ( || ( virtual/w3m www-client/lynx www-client/links ) )
 	latex? ( dev-texlive/texlive-formatsextra )"
 # We only depend on flex when we patch the imput lexer.
 DEPEND="${RDEPEND}"
diff --git a/dev-java/java-config/java-config-2.3.1.ebuild b/dev-java/java-config/java-config-2.3.1.ebuild
index 87732bf9ec..8fee8e4942 100644
--- a/dev-java/java-config/java-config-2.3.1.ebuild
+++ b/dev-java/java-config/java-config-2.3.1.ebuild
@@ -4,7 +4,7 @@
 EAPI=7
 
 # jython depends on java-config, so don't add it or things will break
-PYTHON_COMPAT=( python3_{6,7,8} )
+PYTHON_COMPAT=( python3_{6..9} )
 DISTUTILS_USE_SETUPTOOLS=no
 
 inherit distutils-r1
diff --git a/dev-lang/R/R-4.0.2.ebuild b/dev-lang/R/R-4.0.2.ebuild
index 83d1ecd938..9725452cda 100644
--- a/dev-lang/R/R-4.0.2.ebuild
+++ b/dev-lang/R/R-4.0.2.ebuild
@@ -92,6 +92,10 @@ pkg_setup() {
 src_prepare() {
 	default
 
+	if use elibc_musl; then
+		append-cxxflags "-D__MUSL__"
+	fi
+
 	# fix packages.html for doc (gentoo bug #205103)
 	sed -e "s:../../../library:../../../../$(get_libdir)/R/library:g" \
 		-i src/library/tools/R/Rd.R || die
diff --git a/dev-lang/mono/mono-6.6.0.161.ebuild b/dev-lang/mono/mono-6.6.0.161.ebuild
index bc6f0cf7f0..001c65de87 100644
--- a/dev-lang/mono/mono-6.6.0.161.ebuild
+++ b/dev-lang/mono/mono-6.6.0.161.ebuild
@@ -79,6 +79,8 @@ src_prepare() {
 
 multilib_src_configure() {
 	local myeconfargs=(
+		--disable-boehm
+		--without-sigaltstack
 		$(use_with xen xen_opt)
 		--without-ikvm-native
 		--disable-dtrace
diff --git a/dev-lang/rust/rust-1.47.0-r1.ebuild b/dev-lang/rust/rust-1.47.0-r1.ebuild
index cd258a9cb8..5a5f750227 100644
--- a/dev-lang/rust/rust-1.47.0-r1.ebuild
+++ b/dev-lang/rust/rust-1.47.0-r1.ebuild
@@ -82,7 +82,6 @@ DEPEND="
 	sys-libs/zlib:=
 	!libressl? ( dev-libs/openssl:0= )
 	libressl? ( dev-libs/libressl:0= )
-	elibc_musl? ( sys-libs/libunwind:= )
 	system-llvm? (
 		${LLVM_DEPEND}
 	)
@@ -194,6 +193,10 @@ pkg_setup() {
 	fi
 }
 
+clear_vendor_checksums() {
+        sed -i 's/\("files":{\)[^}]*/\1/' vendor/$1/.cargo-checksum.json
+}
+
 src_prepare() {
 	if ! use system-bootstrap; then
 		local rust_stage0_root="${WORKDIR}"/rust-stage0
@@ -203,6 +206,9 @@ src_prepare() {
 			--destdir="${rust_stage0_root}" --prefix=/ || die
 	fi
 
+	clear_vendor_checksums libc
+	clear_vendor_checksums libc-0.2.74
+
 	default
 }
 
@@ -572,6 +578,11 @@ src_install() {
 
 	insinto /etc/env.d/rust
 	doins "${T}/provider-${P}"
+
+	# remove some clutter
+	pushd "${ED}/usr/lib/${PN}/${PV}/lib/rustlib"
+	rm components install.log manifest-* rust-installer-version uninstall.sh || die
+	popd
 }
 
 pkg_postinst() {
diff --git a/dev-libs/gobject-introspection/gobject-introspection-1.64.1-r1.ebuild b/dev-libs/gobject-introspection/gobject-introspection-1.64.1-r1.ebuild
index f8fe7c83e9..2265c468fe 100644
--- a/dev-libs/gobject-introspection/gobject-introspection-1.64.1-r1.ebuild
+++ b/dev-libs/gobject-introspection/gobject-introspection-1.64.1-r1.ebuild
@@ -3,7 +3,7 @@
 
 EAPI=7
 
-PYTHON_COMPAT=( python3_{6,7,8} )
+PYTHON_COMPAT=( python3_{6..9} )
 PYTHON_REQ_USE="xml"
 inherit gnome.org meson python-single-r1 toolchain-funcs xdg
 
diff --git a/dev-libs/libffi/libffi-3.3-r2.ebuild b/dev-libs/libffi/libffi-3.3-r2.ebuild
index 19e01b0832..5f4817a3c2 100644
--- a/dev-libs/libffi/libffi-3.3-r2.ebuild
+++ b/dev-libs/libffi/libffi-3.3-r2.ebuild
@@ -54,19 +54,7 @@ pkg_setup() {
 
 multilib_src_configure() {
 	use userland_BSD && export HOST="${CHOST}"
-	# --includedir= path maintains a few properties:
-	# 1. have stable name across libffi versions: some packages like
-	#    dev-lang/ghc or kde-frameworks/networkmanager-qt embed
-	#    ${includedir} at build-time. Don't require those to be
-	#    rebuilt unless SONAME changes. bug #695788
-	#
-	#    We use /usr/.../${PN} (instead of former /usr/.../${P}).
-	#
-	# 2. have ${ABI}-specific location as ffi.h is target-dependent.
-	#
-	#    We use /usr/$(get_libdir)/... to have ABI identifier.
 	econf \
-		--includedir="${EPREFIX}"/usr/$(get_libdir)/${PN}/include \
 		--disable-multi-os-directory \
 		$(use_enable static-libs static) \
 		$(use_enable pax_kernel pax_emutramp) \
diff --git a/dev-libs/marisa/marisa-0.2.5.ebuild b/dev-libs/marisa/marisa-0.2.5.ebuild
index 4289f375ea..1ecccb7e5c 100644
--- a/dev-libs/marisa/marisa-0.2.5.ebuild
+++ b/dev-libs/marisa/marisa-0.2.5.ebuild
@@ -2,7 +2,7 @@
 # Distributed under the terms of the GNU General Public License v2
 
 EAPI="7"
-PYTHON_COMPAT=(python{3_6,3_7})
+PYTHON_COMPAT=(python3_{6..9})
 DISTUTILS_OPTIONAL="1"
 
 inherit autotools distutils-r1
diff --git a/dev-python/bitarray/bitarray-0.8.3.ebuild b/dev-python/bitarray/bitarray-0.8.3.ebuild
index 43f4112ed5..43bfdb3ee8 100644
--- a/dev-python/bitarray/bitarray-0.8.3.ebuild
+++ b/dev-python/bitarray/bitarray-0.8.3.ebuild
@@ -3,7 +3,7 @@
 
 EAPI=6
 
-PYTHON_COMPAT=( python{3_6,3_7} )
+PYTHON_COMPAT=( python{3_6,3_7,3_8} )
 
 inherit distutils-r1
 
diff --git a/dev-python/pypax/pypax-0.9.5.ebuild b/dev-python/pypax/pypax-0.9.5.ebuild
index b005f93fdc..64efb49fc2 100644
--- a/dev-python/pypax/pypax-0.9.5.ebuild
+++ b/dev-python/pypax/pypax-0.9.5.ebuild
@@ -4,7 +4,7 @@
 EAPI=6
 
 DISTUTILS_USE_SETUPTOOLS=no
-PYTHON_COMPAT=( python{3_6,3_7,3_8} )
+PYTHON_COMPAT=( python3_{6..9} )
 
 inherit distutils-r1
 
diff --git a/dev-util/gdbus-codegen/gdbus-codegen-2.64.5.ebuild b/dev-util/gdbus-codegen/gdbus-codegen-2.64.5.ebuild
index 63279f3aef..57cc6931a2 100644
--- a/dev-util/gdbus-codegen/gdbus-codegen-2.64.5.ebuild
+++ b/dev-util/gdbus-codegen/gdbus-codegen-2.64.5.ebuild
@@ -3,7 +3,7 @@
 
 EAPI=7
 GNOME_ORG_MODULE="glib"
-PYTHON_COMPAT=( python3_{6,7,8} )
+PYTHON_COMPAT=( python3_{6..9} )
 PYTHON_REQ_USE="xml"
 DISTUTILS_SINGLE_IMPL=1
 DISTUTILS_USE_SETUPTOOLS=no
diff --git a/dev-util/itstool/itstool-2.0.6-r1.ebuild b/dev-util/itstool/itstool-2.0.6-r1.ebuild
index a1a296a1d3..2fce6a91a2 100644
--- a/dev-util/itstool/itstool-2.0.6-r1.ebuild
+++ b/dev-util/itstool/itstool-2.0.6-r1.ebuild
@@ -3,7 +3,7 @@
 
 EAPI=7
 
-PYTHON_COMPAT=( python3_{6,7,8} )
+PYTHON_COMPAT=( python3_{6..9} )
 PYTHON_REQ_USE="xml"
 
 inherit python-single-r1
diff --git a/eclass/rust-toolchain.eclass b/eclass/rust-toolchain.eclass
index 9d7360ed1c..4d7b1af38f 100644
--- a/eclass/rust-toolchain.eclass
+++ b/eclass/rust-toolchain.eclass
@@ -35,12 +35,11 @@ inherit multilib-build
 rust_abi() {
 	local CTARGET=${1:-${CHOST}}
 	case ${CTARGET%%*-} in
-		aarch64*)	  echo aarch64-unknown-linux-gnu;;
+		aarch64*gnu)	  echo aarch64-unknown-linux-gnu;;
 		mips64*)	  echo mips64-unknown-linux-gnuabi64;;
 		powerpc64le*) echo powerpc64le-unknown-linux-gnu;;
 		powerpc64*)   echo powerpc64-unknown-linux-gnu;;
 		x86_64*gnu)	  echo x86_64-unknown-linux-gnu;;
-		x86_64*musl)  echo x86_64-unknown-linux-musl;;
 		armv6j*s*)	  echo arm-unknown-linux-gnueabi;;
 		armv6j*h*)	  echo arm-unknown-linux-gnueabihf;;
 		armv7a*h*)	  echo armv7-unknown-linux-gnueabihf;;
diff --git a/media-libs/libsndfile/libsndfile-1.0.29_pre2_p20191024.ebuild b/media-libs/libsndfile/libsndfile-1.0.29_pre2_p20191024.ebuild
index 78c8d372df..0f8305602e 100644
--- a/media-libs/libsndfile/libsndfile-1.0.29_pre2_p20191024.ebuild
+++ b/media-libs/libsndfile/libsndfile-1.0.29_pre2_p20191024.ebuild
@@ -33,7 +33,6 @@ DEPEND="${RDEPEND}"
 BDEPEND="
 	virtual/pkgconfig
 	${PYTHON_DEPS}
-	sys-devel/autogen
 "
 
 S="${WORKDIR}/${PN}-${MY_COMMIT}"
diff --git a/media-libs/mesa/mesa-20.2.0.ebuild b/media-libs/mesa/mesa-20.2.0.ebuild
index 402d1049c1..48d1f0be04 100644
--- a/media-libs/mesa/mesa-20.2.0.ebuild
+++ b/media-libs/mesa/mesa-20.2.0.ebuild
@@ -143,9 +143,10 @@ RDEPEND="${RDEPEND}
 # 1. List all the working slots (with min versions) in ||, newest first.
 # 2. Update the := to specify *max* version, e.g. < 10.
 # 3. Specify LLVM_MAX_SLOT, e.g. 9.
-LLVM_MAX_SLOT="10"
+LLVM_MAX_SLOT="11"
 LLVM_DEPSTR="
 	|| (
+		sys-devel/llvm:11[${MULTILIB_USEDEP}]
 		sys-devel/llvm:10[${MULTILIB_USEDEP}]
 		sys-devel/llvm:9[${MULTILIB_USEDEP}]
 	)
diff --git a/media-libs/opencv/opencv-4.1.2-r3.ebuild b/media-libs/opencv/opencv-4.1.2-r3.ebuild
index 0f86abb641..08b6a65d99 100644
--- a/media-libs/opencv/opencv-4.1.2-r3.ebuild
+++ b/media-libs/opencv/opencv-4.1.2-r3.ebuild
@@ -3,7 +3,7 @@
 
 EAPI=7
 
-PYTHON_COMPAT=( python3_{6,7} )
+PYTHON_COMPAT=( python3_{6..9} )
 
 inherit java-pkg-opt-2 java-ant-2 python-r1 toolchain-funcs cmake-multilib
 
diff --git a/media-video/vlc/vlc-3.0.11.1.ebuild b/media-video/vlc/vlc-3.0.11.1.ebuild
index 6de7754010..069bf329e7 100644
--- a/media-video/vlc/vlc-3.0.11.1.ebuild
+++ b/media-video/vlc/vlc-3.0.11.1.ebuild
@@ -59,7 +59,7 @@ BDEPEND="
 	x86? ( dev-lang/yasm )
 "
 RDEPEND="
-	net-dns/libidn:=
+	net-dns/libidn2:=
 	sys-libs/zlib[minizip]
 	virtual/libintl
 	virtual/opengl
diff --git a/sys-devel/clang/clang-11.0.0.ebuild b/sys-devel/clang/clang-11.0.0.ebuild
index 0dacfd8c5e..565642c4a9 100644
--- a/sys-devel/clang/clang-11.0.0.ebuild
+++ b/sys-devel/clang/clang-11.0.0.ebuild
@@ -31,8 +31,8 @@ LLVM_TARGET_USEDEPS=${ALL_LLVM_TARGETS[@]/%/?}
 LICENSE="Apache-2.0-with-LLVM-exceptions UoI-NCSA MIT"
 SLOT="$(ver_cut 1)"
 KEYWORDS="~amd64 ~arm ~arm64 ~ppc64 ~x86 ~amd64-linux"
-IUSE="debug default-compiler-rt default-libcxx default-lld
-	doc +static-analyzer test xml kernel_FreeBSD ${ALL_LLVM_TARGETS[*]}"
+IUSE="debug +default-compiler-rt +default-libcxx +default-lld +default-libunwind
+	doc +gcc-symlinks +static-analyzer test xml kernel_FreeBSD ${ALL_LLVM_TARGETS[*]}"
 REQUIRED_USE="${PYTHON_REQUIRED_USE}
 	|| ( ${ALL_LLVM_TARGETS[*]} )"
 RESTRICT="!test? ( test )"
@@ -42,9 +42,11 @@ RDEPEND="
 	static-analyzer? ( dev-lang/perl:* )
 	xml? ( dev-libs/libxml2:2=[${MULTILIB_USEDEP}] )
 	${PYTHON_DEPS}"
-DEPEND="${RDEPEND}"
+DEPEND="${RDEPEND}
+	default-compiler-rt? ( =sys-libs/compiler-rt-${PV%_*}* )
+	"
 BDEPEND="
-	dev-python/sphinx
+	doc? ( dev-python/sphinx )
 	xml? ( virtual/pkgconfig )
 	${PYTHON_DEPS}"
 RDEPEND="${RDEPEND}
@@ -53,7 +55,7 @@ RDEPEND="${RDEPEND}
 PDEPEND="
 	sys-devel/clang-common
 	~sys-devel/clang-runtime-${PV}
-	default-compiler-rt? ( =sys-libs/compiler-rt-${PV%_*}* )
+	default-libunwind? ( >=sys-libs/llvm-libunwind-${PV} )
 	default-libcxx? ( >=sys-libs/libcxx-${PV} )
 	default-lld? ( sys-devel/lld )"
 
@@ -190,13 +192,12 @@ get_distribution_components() {
 			find-all-symbols
 			modularize
 			pp-trace
+		)
 
+		use doc && out+=(
 			# manpages
 			docs-clang-man
 			docs-clang-tools-man
-		)
-
-		use doc && out+=(
 			docs-clang-html
 			docs-clang-tools-html
 		)
@@ -239,10 +240,14 @@ multilib_src_configure() {
 		# furthermore, it provides only syntax checking
 		-DCLANG_DEFAULT_OPENMP_RUNTIME=libomp
 
-		# override default stdlib and rtlib
+		# create gcc symlinks: gcc g++ cc cpp c++
+		-DCLANG_INSTALL_GCC_SYMLINKS=$(usex gcc-symlinks)
+
+		# override default stdlib and rtlib and unwind
 		-DCLANG_DEFAULT_CXX_STDLIB=$(usex default-libcxx libc++ "")
 		-DCLANG_DEFAULT_RTLIB=$(usex default-compiler-rt compiler-rt "")
 		-DCLANG_DEFAULT_LINKER=$(usex default-lld lld "")
+		-DCLANG_DEFAULT_UNWINDLIB=$(usex default-libunwind libunwind "")
 
 		-DCLANG_ENABLE_ARCMT=$(usex static-analyzer)
 		-DCLANG_ENABLE_STATIC_ANALYZER=$(usex static-analyzer)
@@ -256,8 +261,10 @@ multilib_src_configure() {
 		mycmakeargs+=(
 			# normally copied from LLVM_INCLUDE_DOCS but the latter
 			# is lacking value in stand-alone builds
-			-DCLANG_INCLUDE_DOCS=ON
-			-DCLANG_TOOLS_EXTRA_INCLUDE_DOCS=ON
+			-DCLANG_INCLUDE_DOCS=$(usex doc)
+			-DCLANG_TOOLS_EXTRA_INCLUDE_DOCS=$(usex doc)
+		)
+		use doc && mycmakeargs+=(
 			-DLLVM_BUILD_DOCS=ON
 			-DLLVM_ENABLE_SPHINX=ON
 			-DCLANG_INSTALL_SPHINX_HTML_DIR="${EPREFIX}/usr/share/doc/${PF}/html"
@@ -332,6 +339,7 @@ src_install() {
 	local clang_version=$(ver_cut 1 "${llvm_version}")
 	local clang_full_version=$(ver_cut 1-3 "${llvm_version}")
 	local clang_tools=( clang clang++ clang-cl clang-cpp )
+	local gcc_tools=( gcc g++ )
 	local abi i
 
 	# cmake gives us:
@@ -360,6 +368,15 @@ src_install() {
 		done
 	done
 
+	# now create target symlinks for all supported ABIs
+	for abi in $(get_all_abis); do
+		local abi_chost=$(get_abi_CHOST "${abi}")
+		for i in "${gcc_tools[@]}"; do
+			dosym "clang-${clang_version}" \
+				"/usr/lib/llvm/${SLOT}/bin/${abi_chost}-${i}"
+		done
+	done
+
 	# Remove unnecessary headers on FreeBSD, bug #417171
 	if use kernel_FreeBSD; then
 		rm "${ED}"/usr/lib/clang/${clang_full_version}/include/{std,float,iso,limits,tgmath,varargs}*.h || die
diff --git a/sys-devel/distcc/distcc-3.3.3-r1.ebuild b/sys-devel/distcc/distcc-3.3.3-r1.ebuild
index f4d8d84f5e..16a1d7cd28 100644
--- a/sys-devel/distcc/distcc-3.3.3-r1.ebuild
+++ b/sys-devel/distcc/distcc-3.3.3-r1.ebuild
@@ -77,6 +77,7 @@ src_configure() {
 	local myconf=(
 		--disable-Werror
 		--libdir=/usr/lib
+		--without-libiberty
 		$(use_enable ipv6 rfc2553)
 		$(use_with gtk)
 		--without-gnome
diff --git a/sys-devel/distcc/files/distccd.service-1 b/sys-devel/distcc/files/distccd.service-1
index 9bd470297d..fd481026e2 100644
--- a/sys-devel/distcc/files/distccd.service-1
+++ b/sys-devel/distcc/files/distccd.service-1
@@ -4,7 +4,7 @@ After=network.target
 
 [Service]
 User=distcc
-ExecStart=/usr/bin/distccd --no-detach --daemon --port 3632 -N 15 --allow $ALLOWED_SERVERS
+ExecStart=/usr/bin/distccd --no-detach --daemon --port 3632 -N 15 $ALLOWED_SERVERS
 
 [Install]
 WantedBy=multi-user.target
diff --git a/sys-devel/llvm/llvm-11.0.0.ebuild b/sys-devel/llvm/llvm-11.0.0.ebuild
index a1a38d4cbd..62e5139b0e 100644
--- a/sys-devel/llvm/llvm-11.0.0.ebuild
+++ b/sys-devel/llvm/llvm-11.0.0.ebuild
@@ -29,7 +29,7 @@ ALL_LLVM_TARGETS=( "${ALL_LLVM_TARGETS[@]/#/llvm_targets_}" )
 LICENSE="Apache-2.0-with-LLVM-exceptions UoI-NCSA BSD public-domain rc"
 SLOT="$(ver_cut 1)"
 KEYWORDS="~amd64 ~arm ~arm64 ~ppc64 ~x86 ~amd64-linux ~ppc-macos ~x64-macos ~x86-macos"
-IUSE="debug doc exegesis gold libedit +libffi ncurses test xar xml z3
+IUSE="+binutils-symlinks debug doc exegesis gold libedit +libffi ncurses test xar xml z3
 	kernel_Darwin ${ALL_LLVM_TARGETS[*]}"
 REQUIRED_USE="|| ( ${ALL_LLVM_TARGETS[*]} )"
 RESTRICT="!test? ( test )"
@@ -59,10 +59,10 @@ BDEPEND="
 		>=sys-devel/binutils-apple-5.1
 	)
 	libffi? ( virtual/pkgconfig )
-	$(python_gen_any_dep '
+	doc? ( $(python_gen_any_dep '
 		dev-python/sphinx[${PYTHON_USEDEP}]
-		doc? ( dev-python/recommonmark[${PYTHON_USEDEP}] )
-	')"
+		dev-python/recommonmark[${PYTHON_USEDEP}]
+	') )"
 # There are no file collisions between these versions but having :0
 # installed means llvm-config there will take precedence.
 RDEPEND="${RDEPEND}
@@ -71,10 +71,9 @@ PDEPEND="sys-devel/llvm-common
 	gold? ( >=sys-devel/llvmgold-${SLOT} )"
 
 python_check_deps() {
-	if use doc; then
-		has_version -b "dev-python/recommonmark[${PYTHON_USEDEP}]" ||
-			return 1
-	fi
+	use doc || return 0
+
+	has_version -b "dev-python/recommonmark[${PYTHON_USEDEP}]" &&
 	has_version -b "dev-python/sphinx[${PYTHON_USEDEP}]"
 }
 
@@ -301,13 +300,29 @@ get_distribution_components() {
 
 			# python modules
 			opt-viewer
+		)
+
+               use binutils-symlinks && out+=(
+			addr2line
+			ar
+			c++filt
+			dlltool
+			dwp
+			nm
+			objcopy
+			objdump
+			ranlib
+			readelf
+			size
+			strings
+			strip
+                )
 
+		use doc && out+=(
 			# manpages
 			docs-dsymutil-man
 			docs-llvm-dwarfdump-man
 			docs-llvm-man
-		)
-		use doc && out+=(
 			docs-llvm-html
 		)
 
@@ -338,6 +353,7 @@ multilib_src_configure() {
 		-DLLVM_BUILD_LLVM_DYLIB=ON
 		-DLLVM_LINK_LLVM_DYLIB=ON
 		-DLLVM_DISTRIBUTION_COMPONENTS=$(get_distribution_components)
+		-DLLVM_INSTALL_BINUTILS_SYMLINKS=$(usex binutils-symlinks)
 
 		# cheap hack: LLVM combines both anyway, and the only difference
 		# is that the former list is explicitly verified at cmake time
@@ -389,11 +405,13 @@ multilib_src_configure() {
 
 	if multilib_is_native_abi; then
 		mycmakeargs+=(
-			-DLLVM_BUILD_DOCS=ON
+			-DLLVM_BUILD_DOCS=$(usex doc)
 			-DLLVM_ENABLE_OCAMLDOC=OFF
-			-DLLVM_ENABLE_SPHINX=ON
+			-DLLVM_ENABLE_SPHINX=$(usex doc)
 			-DLLVM_ENABLE_DOXYGEN=OFF
 			-DLLVM_INSTALL_UTILS=ON
+		)
+		use doc && mycmakeargs+=(
 			-DCMAKE_INSTALL_MANDIR="${EPREFIX}/usr/lib/llvm/${SLOT}/share/man"
 			-DLLVM_INSTALL_SPHINX_HTML_DIR="${EPREFIX}/usr/share/doc/${PF}/html"
 			-DSPHINX_WARNINGS_AS_ERRORS=OFF
diff --git a/sys-libs/libblockdev/libblockdev-2.24.ebuild b/sys-libs/libblockdev/libblockdev-2.24.ebuild
index 04aa60d480..c872526a03 100644
--- a/sys-libs/libblockdev/libblockdev-2.24.ebuild
+++ b/sys-libs/libblockdev/libblockdev-2.24.ebuild
@@ -3,7 +3,7 @@
 
 EAPI=7
 
-PYTHON_COMPAT=( python3_{6,7,8} )
+PYTHON_COMPAT=( python3_{6..9} )
 inherit autotools python-single-r1 xdg-utils
 
 DESCRIPTION="A library for manipulating block devices"
