diff --git a/app-crypt/libsecret/libsecret-0.20.3.ebuild b/app-crypt/libsecret/libsecret-0.20.3.ebuild
index be5c5bb142..daa1b9f246 100644
--- a/app-crypt/libsecret/libsecret-0.20.3.ebuild
+++ b/app-crypt/libsecret/libsecret-0.20.3.ebuild
@@ -26,8 +26,8 @@ RDEPEND="
 	crypt? ( >=dev-libs/libgcrypt-1.2.2:0=[${MULTILIB_USEDEP}] )
 	introspection? ( >=dev-libs/gobject-introspection-1.29:= )
 "
-PDEPEND=">=gnome-base/gnome-keyring-3
-"
+#PDEPEND=">=gnome-base/gnome-keyring-3
+#"
 # PDEPEND to avoid circular dep (bug #547456)
 # gnome-keyring needed at runtime as explained at https://bugs.gentoo.org/475182#c2
 # Add ksecrets to PDEPEND when it's added to portage
diff --git a/app-crypt/pinentry/pinentry-1.1.0-r3.ebuild b/app-crypt/pinentry/pinentry-1.1.0-r3.ebuild
index efc797548a..1bc61719c4 100644
--- a/app-crypt/pinentry/pinentry-1.1.0-r3.ebuild
+++ b/app-crypt/pinentry/pinentry-1.1.0-r3.ebuild
@@ -12,7 +12,7 @@ SRC_URI="mirror://gnupg/${PN}/${P}.tar.bz2"
 LICENSE="GPL-2"
 SLOT="0"
 KEYWORDS="~alpha amd64 arm arm64 hppa ~ia64 ~mips ppc ppc64 ~riscv s390 sparc x86 ~ppc-aix ~x64-cygwin ~amd64-linux ~x86-linux ~ppc-macos ~x64-macos ~x86-macos ~sparc-solaris ~sparc64-solaris ~x64-solaris ~x86-solaris"
-IUSE="caps emacs gnome-keyring fltk gtk ncurses qt5"
+IUSE="caps emacs gnome-keyring fltk gtk libsecret ncurses qt5"
 
 DEPEND="
 	app-eselect/eselect-pinentry
@@ -21,7 +21,7 @@ DEPEND="
 	>=dev-libs/libgpg-error-1.17
 	caps? ( sys-libs/libcap )
 	fltk? ( x11-libs/fltk )
-	gnome-keyring? ( app-crypt/libsecret )
+	libsecret? ( app-crypt/libsecret )
 	gtk? ( x11-libs/gtk+:2 )
 	ncurses? ( sys-libs/ncurses:0= )
 	qt5? (
@@ -59,7 +59,7 @@ src_configure() {
 	econf \
 		$(use_enable emacs pinentry-emacs) \
 		$(use_enable fltk pinentry-fltk) \
-		$(use_enable gnome-keyring libsecret) \
+		$(use_enable libsecret) \
 		$(use_enable gnome-keyring pinentry-gnome3) \
 		$(use_enable gtk pinentry-gtk2) \
 		$(use_enable ncurses fallback-curses) \
diff --git a/app-emulation/crun/crun-0.15.ebuild b/app-emulation/crun/crun-0.15.ebuild
index 888df44272..280cabc77a 100644
--- a/app-emulation/crun/crun-0.15.ebuild
+++ b/app-emulation/crun/crun-0.15.ebuild
@@ -21,6 +21,7 @@ DEPEND="
 	>=dev-libs/yajl-2.0.0
 	caps? ( sys-libs/libcap )
 	criu? ( >=sys-process/criu-3.13 )
+	elibc_musl? ( sys-libs/argp-standalone )
 	seccomp? ( sys-libs/libseccomp )
 	systemd? ( sys-apps/systemd:= )
 "
diff --git a/app-emulation/libpod/libpod-2.1.0.ebuild b/app-emulation/libpod/libpod-2.1.0.ebuild
index 8335388f9f..670453c7f1 100644
--- a/app-emulation/libpod/libpod-2.1.0.ebuild
+++ b/app-emulation/libpod/libpod-2.1.0.ebuild
@@ -26,7 +26,6 @@ COMMON_DEPEND="
 	dev-libs/libassuan:=
 	dev-libs/libgpg-error:=
 	>=net-misc/cni-plugins-0.8.6
-	sys-fs/lvm2
 	sys-libs/libseccomp:=
 
 	apparmor? ( sys-libs/libapparmor )
diff --git a/app-emulation/virt-manager/virt-manager-3.1.0.ebuild b/app-emulation/virt-manager/virt-manager-3.1.0.ebuild
index cb575dc4df..8f524a1272 100644
--- a/app-emulation/virt-manager/virt-manager-3.1.0.ebuild
+++ b/app-emulation/virt-manager/virt-manager-3.1.0.ebuild
@@ -42,7 +42,6 @@ RDEPEND="!app-emulation/virtinst
 		gnome-base/dconf
 		>=net-libs/gtk-vnc-0.3.8[gtk3(+),introspection]
 		net-misc/spice-gtk[usbredir,gtk3,introspection,sasl?]
-		net-misc/x11-ssh-askpass
 		x11-libs/gtk+:3[introspection]
 		x11-libs/gtksourceview:4[introspection]
 		x11-libs/vte:2.91[introspection]
diff --git a/app-i18n/fcitx/fcitx-4.2.9.8.ebuild b/app-i18n/fcitx/fcitx-4.2.9.8.ebuild
index 445151b051..0b9336e8a5 100644
--- a/app-i18n/fcitx/fcitx-4.2.9.8.ebuild
+++ b/app-i18n/fcitx/fcitx-4.2.9.8.ebuild
@@ -59,7 +59,7 @@ DEPEND="dev-libs/glib:2
 		pango? ( x11-libs/pango )
 		!pango? ( media-libs/fontconfig )
 	)
-	enchant? ( app-text/enchant:0= )
+	enchant? ( app-text/enchant:= )
 	gtk2? ( x11-libs/gtk+:2 )
 	gtk3? ( x11-libs/gtk+:3 )
 	lua? ( dev-lang/lua:= )
diff --git a/app-text/xmlto/xmlto-0.0.28-r3.ebuild b/app-text/xmlto/xmlto-0.0.28-r3.ebuild
index 5602e66567..42815c690d 100644
--- a/app-text/xmlto/xmlto-0.0.28-r3.ebuild
+++ b/app-text/xmlto/xmlto-0.0.28-r3.ebuild
@@ -21,7 +21,7 @@ RDEPEND=">=app-text/docbook-xsl-stylesheets-1.62.0-r1
 	>=sys-apps/coreutils-6.10-r1
 	|| ( sys-apps/util-linux app-misc/getopt )
 	sys-apps/which
-	text? ( || ( virtual/w3m www-client/lynx www-client/elinks ) )
+	text? ( || ( virtual/w3m www-client/lynx www-client/links ) )
 	latex? ( dev-texlive/texlive-formatsextra )"
 # We only depend on flex when we patch the imput lexer.
 DEPEND="${RDEPEND}"
diff --git a/dev-lang/rust/rust-1.47.0-r1.ebuild b/dev-lang/rust/rust-1.47.0-r1.ebuild
index eb58cb752e..64815c0564 100644
--- a/dev-lang/rust/rust-1.47.0-r1.ebuild
+++ b/dev-lang/rust/rust-1.47.0-r1.ebuild
@@ -82,7 +82,6 @@ DEPEND="
 	sys-libs/zlib:=
 	!libressl? ( dev-libs/openssl:0= )
 	libressl? ( dev-libs/libressl:0= )
-	elibc_musl? ( sys-libs/libunwind:= )
 	system-llvm? (
 		${LLVM_DEPEND}
 	)
@@ -195,6 +194,10 @@ pkg_setup() {
 	fi
 }
 
+clear_vendor_checksums() {
+        sed -i 's/\("files":{\)[^}]*/\1/' vendor/$1/.cargo-checksum.json
+}
+
 src_prepare() {
 	if ! use system-bootstrap; then
 		local rust_stage0_root="${WORKDIR}"/rust-stage0
@@ -204,6 +207,9 @@ src_prepare() {
 			--destdir="${rust_stage0_root}" --prefix=/ || die
 	fi
 
+	clear_vendor_checksums libc
+	clear_vendor_checksums libc-0.2.74
+
 	default
 }
 
@@ -212,7 +218,7 @@ src_configure() {
 
 	# Collect rust target names to compile standard libs for all ABIs.
 	for v in $(multilib_get_enabled_abi_pairs); do
-		rust_targets="${rust_targets},\"$(rust_abi $(get_abi_CHOST ${v##*.}))\""
+		rust_targets="${rust_targets},\"$(get_abi_CHOST ${v##*.})\""
 	done
 	if use wasm; then
 		rust_targets="${rust_targets},\"wasm32-unknown-unknown\""
@@ -245,7 +251,7 @@ src_configure() {
 		rust_stage0_root="${WORKDIR}"/rust-stage0
 	fi
 
-	rust_target="$(rust_abi)"
+	rust_target="$(get_abi_CHOST ${v##*.})"
 
 	cat <<- _EOF_ > "${S}"/config.toml
 		[llvm]
@@ -307,7 +313,7 @@ src_configure() {
 	_EOF_
 
 	for v in $(multilib_get_enabled_abi_pairs); do
-		rust_target=$(rust_abi $(get_abi_CHOST ${v##*.}))
+		rust_target=$(get_abi_CHOST ${v##*.})
 		arch_cflags="$(get_abi_CFLAGS ${v##*.})"
 
 		cat <<- _EOF_ >> "${S}"/config.env
@@ -425,8 +431,8 @@ src_compile() {
 	# we need \n IFS to have config.env with spaces loaded properly. #734018
 	(
 	IFS=$'\n'
-	env $(cat "${S}"/config.env) RUST_BACKTRACE=1\
-		"${EPYTHON}" ./x.py dist -vv --config="${S}"/config.toml -j$(makeopts_jobs) || die
+	env $(cat "${S}"/config.env) DESTDIR="${D}" \
+		"${EPYTHON}" ./x.py build --stage 2 -vv --config="${S}"/config.toml -j$(makeopts_jobs) || die
 	)
 }
 
@@ -493,7 +499,7 @@ src_install() {
 	(
 	IFS=$'\n'
 	env $(cat "${S}"/config.env) DESTDIR="${D}" \
-		"${EPYTHON}" ./x.py install -vv --config="${S}"/config.toml || die
+		"${EPYTHON}" ./x.py install -vv --config="${S}"/config.toml -j$(makeopts_jobs) || die
 	)
 
 	# bug #689562, #689160
@@ -573,6 +579,11 @@ src_install() {
 
 	insinto /etc/env.d/rust
 	doins "${T}/provider-${P}"
+
+	# remove some clutter
+	pushd "${ED}/usr/lib/${PN}/${PV}/lib/rustlib"
+	rm components install.log manifest-* rust-installer-version uninstall.sh || die
+	popd
 }
 
 pkg_postinst() {
diff --git a/dev-libs/libffi/libffi-3.3-r2.ebuild b/dev-libs/libffi/libffi-3.3-r2.ebuild
index 55138eb082..6a794ea251 100644
--- a/dev-libs/libffi/libffi-3.3-r2.ebuild
+++ b/dev-libs/libffi/libffi-3.3-r2.ebuild
@@ -41,19 +41,7 @@ ECONF_SOURCE=${S}
 
 multilib_src_configure() {
 	use userland_BSD && export HOST="${CHOST}"
-	# --includedir= path maintains a few properties:
-	# 1. have stable name across libffi versions: some packages like
-	#    dev-lang/ghc or kde-frameworks/networkmanager-qt embed
-	#    ${includedir} at build-time. Don't require those to be
-	#    rebuilt unless SONAME changes. bug #695788
-	#
-	#    We use /usr/.../${PN} (instead of former /usr/.../${P}).
-	#
-	# 2. have ${ABI}-specific location as ffi.h is target-dependent.
-	#
-	#    We use /usr/$(get_libdir)/... to have ABI identifier.
 	econf \
-		--includedir="${EPREFIX}"/usr/$(get_libdir)/${PN}/include \
 		--disable-multi-os-directory \
 		$(use_enable static-libs static) \
 		$(use_enable pax_kernel pax_emutramp) \
diff --git a/dev-util/flatpak-builder/flatpak-builder-1.0.11.ebuild b/dev-util/flatpak-builder/flatpak-builder-1.0.11.ebuild
index a8017f003f..27fdf16889 100644
--- a/dev-util/flatpak-builder/flatpak-builder-1.0.11.ebuild
+++ b/dev-util/flatpak-builder/flatpak-builder-1.0.11.ebuild
@@ -21,6 +21,7 @@ RDEPEND="
 	>=dev-libs/libxml2-2.4:=
 	dev-libs/json-glib:=
 	net-misc/curl:=
+	elibc_musl? ( dev-libs/libdwarf )
 	yaml? ( dev-libs/libyaml:= )
 "
 DEPEND="${RDEPEND}"
diff --git a/media-libs/mesa/mesa-20.2.2.ebuild b/media-libs/mesa/mesa-20.2.2.ebuild
index 092aa16286..b128d3746f 100644
--- a/media-libs/mesa/mesa-20.2.2.ebuild
+++ b/media-libs/mesa/mesa-20.2.2.ebuild
@@ -143,9 +143,10 @@ RDEPEND="${RDEPEND}
 # 1. List all the working slots (with min versions) in ||, newest first.
 # 2. Update the := to specify *max* version, e.g. < 10.
 # 3. Specify LLVM_MAX_SLOT, e.g. 9.
-LLVM_MAX_SLOT="10"
+LLVM_MAX_SLOT="11"
 LLVM_DEPSTR="
 	|| (
+		sys-devel/llvm:11[${MULTILIB_USEDEP}]
 		sys-devel/llvm:10[${MULTILIB_USEDEP}]
 		sys-devel/llvm:9[${MULTILIB_USEDEP}]
 	)
diff --git a/media-video/vlc/vlc-3.0.11.1.ebuild b/media-video/vlc/vlc-3.0.11.1.ebuild
index 6de7754010..069bf329e7 100644
--- a/media-video/vlc/vlc-3.0.11.1.ebuild
+++ b/media-video/vlc/vlc-3.0.11.1.ebuild
@@ -59,7 +59,7 @@ BDEPEND="
 	x86? ( dev-lang/yasm )
 "
 RDEPEND="
-	net-dns/libidn:=
+	net-dns/libidn2:=
 	sys-libs/zlib[minizip]
 	virtual/libintl
 	virtual/opengl
diff --git a/net-libs/gtk-vnc/gtk-vnc-1.0.0.ebuild b/net-libs/gtk-vnc/gtk-vnc-1.0.0.ebuild
index 1a09512465..01f1584d81 100644
--- a/net-libs/gtk-vnc/gtk-vnc-1.0.0.ebuild
+++ b/net-libs/gtk-vnc/gtk-vnc-1.0.0.ebuild
@@ -60,7 +60,7 @@ src_configure() {
 		$(meson_feature pulseaudio)
 		$(meson_feature sasl)
 		-Denable-werror=false
-		-Dwith-coroutine=auto # gthread on windows, libc ucontext elsewhere; neither has extra deps
+		-Dwith-coroutine=gthread # gthread on windows, libc ucontext elsewhere; neither has extra deps
 		$(meson_use vala with-vala)
 	)
 	meson_src_configure
diff --git a/net-misc/spice-gtk/spice-gtk-0.38.ebuild b/net-misc/spice-gtk/spice-gtk-0.38.ebuild
index f370fa8c0e..b8e0a759a9 100644
--- a/net-misc/spice-gtk/spice-gtk-0.38.ebuild
+++ b/net-misc/spice-gtk/spice-gtk-0.38.ebuild
@@ -93,6 +93,7 @@ src_configure() {
 		$(meson_feature usbredir)
 		$(meson_feature vala vapi)
 		$(meson_feature webdav)
+		-Dcoroutine=gthread
 	)
 
 	if use usbredir; then
diff --git a/profiles/default/linux/amd64/17.0/musl/parent b/profiles/default/linux/amd64/17.0/musl/parent
index 8b660ec6dc..d8074cd72a 100644
--- a/profiles/default/linux/amd64/17.0/musl/parent
+++ b/profiles/default/linux/amd64/17.0/musl/parent
@@ -1,3 +1,3 @@
-..
+../../17.1
 ../../../../../arch/amd64/no-multilib
 ../../../../../features/musl/
diff --git a/sys-devel/clang/clang-11.0.0.ebuild b/sys-devel/clang/clang-11.0.0.ebuild
index 23ee5eee92..5dd79be307 100644
--- a/sys-devel/clang/clang-11.0.0.ebuild
+++ b/sys-devel/clang/clang-11.0.0.ebuild
@@ -32,8 +32,8 @@ LLVM_TARGET_USEDEPS=${ALL_LLVM_TARGETS[@]/%/?}
 LICENSE="Apache-2.0-with-LLVM-exceptions UoI-NCSA MIT"
 SLOT="$(ver_cut 1)"
 KEYWORDS="amd64 arm ~arm64 ppc64 ~x86 ~amd64-linux"
-IUSE="debug default-compiler-rt default-libcxx default-lld
-	doc +static-analyzer test xml kernel_FreeBSD ${ALL_LLVM_TARGETS[*]}"
+IUSE="debug +default-compiler-rt +default-libcxx +default-lld +default-libunwind
+	doc +gcc-symlinks +static-analyzer test xml kernel_FreeBSD ${ALL_LLVM_TARGETS[*]}"
 REQUIRED_USE="${PYTHON_REQUIRED_USE}
 	|| ( ${ALL_LLVM_TARGETS[*]} )"
 RESTRICT="!test? ( test )"
@@ -43,7 +43,9 @@ RDEPEND="
 	static-analyzer? ( dev-lang/perl:* )
 	xml? ( dev-libs/libxml2:2=[${MULTILIB_USEDEP}] )
 	${PYTHON_DEPS}"
-DEPEND="${RDEPEND}"
+DEPEND="${RDEPEND}
+	default-compiler-rt? ( =sys-libs/compiler-rt-${PV%_*}* )
+	"
 BDEPEND="
 	>=dev-util/cmake-3.16
 	doc? ( dev-python/sphinx )
@@ -55,7 +57,7 @@ RDEPEND="${RDEPEND}
 PDEPEND="
 	sys-devel/clang-common
 	~sys-devel/clang-runtime-${PV}
-	default-compiler-rt? ( =sys-libs/compiler-rt-${PV%_*}* )
+	default-libunwind? ( >=sys-libs/llvm-libunwind-${PV} )
 	default-libcxx? ( >=sys-libs/libcxx-${PV} )
 	default-lld? ( sys-devel/lld )"
 
@@ -245,10 +247,14 @@ multilib_src_configure() {
 		# furthermore, it provides only syntax checking
 		-DCLANG_DEFAULT_OPENMP_RUNTIME=libomp
 
-		# override default stdlib and rtlib
+		# create gcc symlinks: gcc g++ cc cpp c++
+		-DCLANG_INSTALL_GCC_SYMLINKS=$(usex gcc-symlinks)
+
+		# override default stdlib and rtlib and unwind
 		-DCLANG_DEFAULT_CXX_STDLIB=$(usex default-libcxx libc++ "")
 		-DCLANG_DEFAULT_RTLIB=$(usex default-compiler-rt compiler-rt "")
 		-DCLANG_DEFAULT_LINKER=$(usex default-lld lld "")
+		-DCLANG_DEFAULT_UNWINDLIB=$(usex default-libunwind libunwind "")
 
 		-DCLANG_ENABLE_ARCMT=$(usex static-analyzer)
 		-DCLANG_ENABLE_STATIC_ANALYZER=$(usex static-analyzer)
@@ -347,6 +353,7 @@ src_install() {
 	local clang_version=$(ver_cut 1 "${llvm_version}")
 	local clang_full_version=$(ver_cut 1-3 "${llvm_version}")
 	local clang_tools=( clang clang++ clang-cl clang-cpp )
+	local gcc_tools=( gcc g++ cpp )
 	local abi i
 
 	# cmake gives us:
@@ -375,6 +382,15 @@ src_install() {
 		done
 	done
 
+	# and finally create target symlinks for all supported ABIs compat
+	for abi in $(get_all_abis); do
+		local abi_chost=$(get_abi_CHOST "${abi}")
+		for i in "${gcc_tools[@]}"; do
+			dosym "clang-${clang_version}" \
+				"/usr/lib/llvm/${SLOT}/bin/${abi_chost}-${i}"
+		done
+	done
+
 	# Remove unnecessary headers on FreeBSD, bug #417171
 	if use kernel_FreeBSD; then
 		rm "${ED}"/usr/lib/clang/${clang_full_version}/include/{std,float,iso,limits,tgmath,varargs}*.h || die
diff --git a/sys-devel/distcc/distcc-3.3.3-r1.ebuild b/sys-devel/distcc/distcc-3.3.3-r1.ebuild
index f4d8d84f5e..16a1d7cd28 100644
--- a/sys-devel/distcc/distcc-3.3.3-r1.ebuild
+++ b/sys-devel/distcc/distcc-3.3.3-r1.ebuild
@@ -77,6 +77,7 @@ src_configure() {
 	local myconf=(
 		--disable-Werror
 		--libdir=/usr/lib
+		--without-libiberty
 		$(use_enable ipv6 rfc2553)
 		$(use_with gtk)
 		--without-gnome
diff --git a/sys-devel/distcc/files/distccd.service-1 b/sys-devel/distcc/files/distccd.service-1
index 9bd470297d..fd481026e2 100644
--- a/sys-devel/distcc/files/distccd.service-1
+++ b/sys-devel/distcc/files/distccd.service-1
@@ -4,7 +4,7 @@ After=network.target
 
 [Service]
 User=distcc
-ExecStart=/usr/bin/distccd --no-detach --daemon --port 3632 -N 15 --allow $ALLOWED_SERVERS
+ExecStart=/usr/bin/distccd --no-detach --daemon --port 3632 -N 15 $ALLOWED_SERVERS
 
 [Install]
 WantedBy=multi-user.target
diff --git a/sys-devel/llvm/llvm-11.0.0.ebuild b/sys-devel/llvm/llvm-11.0.0.ebuild
index eac3bcb175..dc57b5ac26 100644
--- a/sys-devel/llvm/llvm-11.0.0.ebuild
+++ b/sys-devel/llvm/llvm-11.0.0.ebuild
@@ -30,7 +30,7 @@ ALL_LLVM_TARGETS=( "${ALL_LLVM_TARGETS[@]/#/llvm_targets_}" )
 LICENSE="Apache-2.0-with-LLVM-exceptions UoI-NCSA BSD public-domain rc"
 SLOT="$(ver_cut 1)"
 KEYWORDS="amd64 arm ~arm64 ppc64 ~x86 ~amd64-linux ~ppc-macos ~x64-macos ~x86-macos"
-IUSE="debug doc exegesis gold libedit +libffi ncurses test xar xml z3
+IUSE="+binutils-symlinks debug doc exegesis gold libedit +libffi ncurses test xar xml z3
 	kernel_Darwin ${ALL_LLVM_TARGETS[*]}"
 REQUIRED_USE="|| ( ${ALL_LLVM_TARGETS[*]} )"
 RESTRICT="!test? ( test )"
@@ -319,6 +319,23 @@ get_distribution_components() {
 				docs-llvm-man
 			)
 		fi
+
+		use binutils-symlinks && out+=(
+			addr2line
+			ar
+			c++filt
+			dlltool
+			dwp
+			nm
+			objcopy
+			objdump
+			ranlib
+			readelf
+			size
+			strings
+			strip
+                )
+
 		use doc && out+=(
 			docs-llvm-html
 		)
@@ -350,6 +367,7 @@ multilib_src_configure() {
 		-DLLVM_BUILD_LLVM_DYLIB=ON
 		-DLLVM_LINK_LLVM_DYLIB=ON
 		-DLLVM_DISTRIBUTION_COMPONENTS=$(get_distribution_components)
+		-DLLVM_INSTALL_BINUTILS_SYMLINKS=$(usex binutils-symlinks)
 
 		# cheap hack: LLVM combines both anyway, and the only difference
 		# is that the former list is explicitly verified at cmake time
diff --git a/sys-libs/liburing/liburing-0.7-r1.ebuild b/sys-libs/liburing/liburing-0.7-r1.ebuild
index 284efb838d..fa9a849d62 100644
--- a/sys-libs/liburing/liburing-0.7-r1.ebuild
+++ b/sys-libs/liburing/liburing-0.7-r1.ebuild
@@ -39,7 +39,7 @@ multilib_src_configure() {
 }
 
 multilib_src_compile() {
-	emake V=1 AR="$(tc-getAR)" RANLIB="$(tc-getRANLIB)"
+	emake V=1 AR="$(tc-getAR)" RANLIB="$(tc-getRANLIB)" src
 }
 
 multilib_src_install_all() {
