diff --git a/app-emulation/libvirt/libvirt-8.5.0.ebuild b/app-emulation/libvirt/libvirt-8.5.0.ebuild
index 1ac6f7431bd..f0f495a97c7 100644
--- a/app-emulation/libvirt/libvirt-8.5.0.ebuild
+++ b/app-emulation/libvirt/libvirt-8.5.0.ebuild
@@ -113,8 +113,6 @@ RDEPEND="
 	selinux? ( >=sys-libs/libselinux-2.0.85 )
 	virt-network? (
 		net-dns/dnsmasq[dhcp,ipv6(+),script]
-		net-firewall/ebtables
-		>=net-firewall/iptables-1.4.10[ipv6(+)]
 		net-misc/radvd
 		sys-apps/iproute2[-minimal]
 	)
diff --git a/app-emulation/virt-manager/virt-manager-4.0.0.ebuild b/app-emulation/virt-manager/virt-manager-4.0.0.ebuild
index cdbbb69fd97..8ef4a304349 100644
--- a/app-emulation/virt-manager/virt-manager-4.0.0.ebuild
+++ b/app-emulation/virt-manager/virt-manager-4.0.0.ebuild
@@ -42,7 +42,7 @@ RDEPEND="${PYTHON_DEPS}
 		>=net-libs/gtk-vnc-0.3.8[gtk3(+),introspection]
 		net-misc/spice-gtk[usbredir,gtk3,introspection,sasl?]
 		policykit? ( sys-auth/polkit[introspection] )
-		sys-apps/dbus[X]
+		sys-apps/dbus
 		x11-libs/gtk+:3[introspection]
 		x11-libs/gtksourceview:4[introspection]
 		x11-libs/vte:2.91[introspection]
diff --git a/app-mobilephone/scrcpy/scrcpy-1.24.ebuild b/app-mobilephone/scrcpy/scrcpy-1.24.ebuild
index 7ca29eab260..bcb7a22cbb5 100644
--- a/app-mobilephone/scrcpy/scrcpy-1.24.ebuild
+++ b/app-mobilephone/scrcpy/scrcpy-1.24.ebuild
@@ -15,7 +15,7 @@ LICENSE="Apache-2.0"
 SLOT="0"
 KEYWORDS="~amd64 ~ppc64 ~x86"
 
-RDEPEND="media-libs/libsdl2[X]
+RDEPEND="media-libs/libsdl2
 	media-video/ffmpeg
 	virtual/libusb:1"
 DEPEND="${RDEPEND}"
diff --git a/dev-libs/jansson/jansson-2.14.ebuild b/dev-libs/jansson/jansson-2.14.ebuild
index 8c97fdd3912..f4ca9bf78bc 100644
--- a/dev-libs/jansson/jansson-2.14.ebuild
+++ b/dev-libs/jansson/jansson-2.14.ebuild
@@ -19,8 +19,6 @@ BDEPEND="doc? ( dev-python/sphinx )"
 PATCHES=( "${FILESDIR}/${P}-test-symbols.patch" )
 
 multilib_src_configure() {
-	tc-ld-force-bfd
-
 	ECONF_SOURCE="${S}" econf $(use_enable static-libs static)
 }
 
diff --git a/eclass/ghc-package.eclass b/eclass/ghc-package.eclass
index 8c77ad5bc3c..014c1d1e21e 100644
--- a/eclass/ghc-package.eclass
+++ b/eclass/ghc-package.eclass
@@ -58,11 +58,11 @@ ghc-getghcpkg() {
 # because for some reason the global package file
 # must be specified
 ghc-getghcpkgbin() {
-	local empty_db="${T}/empty.conf.d" ghc_pkg="$(ghc-libdir)/bin/ghc-pkg"
+	local empty_db="${T}/empty.conf.d" ghc_pkg="$(ghc-libdir)/../bin/ghc-pkg"
 	if [[ ! -d ${empty_db} ]]; then
 		"${ghc_pkg}" init "${empty_db}" || die "Failed to initialize empty global db"
 	fi
-	echo "$(ghc-libdir)/bin/ghc-pkg" "--global-package-db=${empty_db}"
+	echo "$(ghc-libdir)/../bin/ghc-pkg" "--global-package-db=${empty_db}"
 }
 
 # @FUNCTION: ghc-version
diff --git a/eclass/linux-mod.eclass b/eclass/linux-mod.eclass
index b7c13cbf7e7..59896741679 100644
--- a/eclass/linux-mod.eclass
+++ b/eclass/linux-mod.eclass
@@ -664,6 +664,7 @@ linux-mod_src_compile() {
 			# fails.
 			eval "emake HOSTCC=\"$(tc-getBUILD_CC)\" \
 						LDFLAGS=\"$(get_abi_LDFLAGS)\" \
+						LLVM_IAS=1 \
 						${BUILD_FIXES} \
 						${BUILD_PARAMS} \
 						${BUILD_TARGETS} " \
diff --git a/eclass/savedconfig.eclass b/eclass/savedconfig.eclass
index 20669c08b33..24c99fcf4eb 100644
--- a/eclass/savedconfig.eclass
+++ b/eclass/savedconfig.eclass
@@ -52,7 +52,7 @@ save_config() {
 	fi
 	[[ $# -eq 0 ]] && die "Usage: save_config <files>"
 
-	local configfile="/etc/portage/savedconfig/${CATEGORY}/${PF}"
+	local configfile="/etc/portage/savedconfig/${CATEGORY}/${PN}"
 
 	if [[ $# -eq 1 && -f $1 ]] ; then
 		# Just one file, so have the ${configfile} be that config file
@@ -65,7 +65,7 @@ save_config() {
 		treecopy "$@" "${ED%/}/${configfile}" || die "failed to save $*"
 	fi
 
-	elog "Your configuration for ${CATEGORY}/${PF} has been saved in "
+	elog "Your configuration for ${CATEGORY}/${PN} has been saved in "
 	elog "\"${configfile}\" for your editing pleasure."
 	elog "You can edit these files by hand and remerge this package with"
 	elog "USE=savedconfig to customise the configuration."
@@ -147,7 +147,7 @@ savedconfig_pkg_postinst() {
 	# are worse :/.
 
 	if use savedconfig ; then
-		find "${EROOT%/}/etc/portage/savedconfig/${CATEGORY}/${PF}" \
+		find "${EROOT%/}/etc/portage/savedconfig/${CATEGORY}/${PN}" \
 			-exec touch {} + 2>/dev/null
 	fi
 }
diff --git a/kde-apps/gwenview/gwenview-22.04.3.ebuild b/kde-apps/gwenview/gwenview-22.04.3.ebuild
index 9a64d0d6970..702ede10def 100644
--- a/kde-apps/gwenview/gwenview-22.04.3.ebuild
+++ b/kde-apps/gwenview/gwenview-22.04.3.ebuild
@@ -87,7 +87,7 @@ src_configure() {
 		$(cmake_use_find_package raw KF5KDcraw)
 		-DGWENVIEW_SEMANTICINFO_BACKEND=$(usex semantic-desktop Baloo None)
 		$(cmake_use_find_package share KF5Purpose)
-		$(cmake_use_find_package X X11)
+		-DWITH_X11=$(usex X)
 	)
 	ecm_src_configure
 }
diff --git a/kde-apps/kdialog/kdialog-22.04.3.ebuild b/kde-apps/kdialog/kdialog-22.04.3.ebuild
index cc270e2c4fd..b39f9590195 100644
--- a/kde-apps/kdialog/kdialog-22.04.3.ebuild
+++ b/kde-apps/kdialog/kdialog-22.04.3.ebuild
@@ -36,7 +36,7 @@ RDEPEND="${DEPEND}"
 
 src_configure() {
 	local mycmakeargs=(
-		$(cmake_use_find_package X X11)
+		-DWITH_X11=$(usex X)
 	)
 	ecm_src_configure
 }
diff --git a/kde-apps/kio-extras/kio-extras-22.04.3.ebuild b/kde-apps/kio-extras/kio-extras-22.04.3.ebuild
index d8f7ab2c6f6..6cd19efae38 100644
--- a/kde-apps/kio-extras/kio-extras-22.04.3.ebuild
+++ b/kde-apps/kio-extras/kio-extras-22.04.3.ebuild
@@ -83,7 +83,7 @@ src_configure() {
 		$(cmake_use_find_package samba Samba)
 		$(cmake_use_find_package sftp libssh)
 		$(cmake_use_find_package taglib Taglib)
-		$(cmake_use_find_package X X11)
+		-DWITH_X11=$(usex X)
 	)
 	use samba && mycmakeargs+=(
 		-DBUILD_KDSoapWSDiscoveryClient=OFF # disable bundled stuff
diff --git a/kde-apps/konsole/konsole-22.04.3.ebuild b/kde-apps/konsole/konsole-22.04.3.ebuild
index 560014d8d4b..f8c17e9a68a 100644
--- a/kde-apps/konsole/konsole-22.04.3.ebuild
+++ b/kde-apps/konsole/konsole-22.04.3.ebuild
@@ -52,7 +52,7 @@ RDEPEND="${DEPEND}"
 
 src_configure() {
 	local mycmakeargs=(
-		$(cmake_use_find_package X X11)
+		-DWITH_X11=$(usex X)
 	)
 
 	ecm_src_configure
diff --git a/kde-frameworks/kinit/kinit-5.96.0.ebuild b/kde-frameworks/kinit/kinit-5.96.0.ebuild
index 98fa7a134a4..8daef1bd943 100644
--- a/kde-frameworks/kinit/kinit-5.96.0.ebuild
+++ b/kde-frameworks/kinit/kinit-5.96.0.ebuild
@@ -43,8 +43,8 @@ src_configure() {
 	local mycmakeargs=(
 		$(cmake_use_find_package caps Libcap)
 		$(cmake_use_find_package man KF5DocTools)
-		$(cmake_use_find_package X X11)
 		$(cmake_use_find_package X XCB)
+		-DWITH_X11=$(usex X)
 	)
 
 	ecm_src_configure
diff --git a/kde-frameworks/kio/kio-5.96.0-r1.ebuild b/kde-frameworks/kio/kio-5.96.0-r1.ebuild
index b0ac9188780..6fe015b71ad 100644
--- a/kde-frameworks/kio/kio-5.96.0-r1.ebuild
+++ b/kde-frameworks/kio/kio-5.96.0-r1.ebuild
@@ -79,7 +79,7 @@ src_configure() {
 		$(cmake_use_find_package handbook KF5DocTools)
 		$(cmake_use_find_package kerberos GSSAPI)
 		$(cmake_use_find_package kwallet KF5Wallet)
-		$(cmake_use_find_package X X11)
+		-DWITH_X11=$(usex X)
 	)
 
 	ecm_src_configure
diff --git a/kde-plasma/kde-gtk-config/kde-gtk-config-5.25.3.ebuild b/kde-plasma/kde-gtk-config/kde-gtk-config-5.25.3.ebuild
index fb20e30ccfa..3ef7c990bee 100644
--- a/kde-plasma/kde-gtk-config/kde-gtk-config-5.25.3.ebuild
+++ b/kde-plasma/kde-gtk-config/kde-gtk-config-5.25.3.ebuild
@@ -18,7 +18,6 @@ KEYWORDS="~amd64 ~arm ~arm64 ~loong ~ppc64 ~riscv ~x86"
 IUSE=""
 
 DEPEND="
-	dev-cpp/glibmm:2
 	dev-libs/glib:2
 	>=dev-qt/qtdbus-${QTMIN}:5
 	>=dev-qt/qtgui-${QTMIN}:5
diff --git a/kde-plasma/libksysguard/libksysguard-5.25.3.ebuild b/kde-plasma/libksysguard/libksysguard-5.25.3.ebuild
index 920ea7b3cac..65cc3c56008 100644
--- a/kde-plasma/libksysguard/libksysguard-5.25.3.ebuild
+++ b/kde-plasma/libksysguard/libksysguard-5.25.3.ebuild
@@ -64,7 +64,7 @@ src_configure() {
 	local mycmakeargs=(
 		$(cmake_use_find_package webengine Qt5WebChannel)
 		$(cmake_use_find_package webengine Qt5WebEngineWidgets)
-		$(cmake_use_find_package X X11)
+		-DWITH_X11=$(usex X)
 	)
 
 	ecm_src_configure
diff --git a/media-libs/mlt/mlt-7.8.0.ebuild b/media-libs/mlt/mlt-7.8.0.ebuild
index 67c1eed409b..6c368511466 100644
--- a/media-libs/mlt/mlt-7.8.0.ebuild
+++ b/media-libs/mlt/mlt-7.8.0.ebuild
@@ -59,7 +59,7 @@ DEPEND="
 	)
 	rubberband? ( media-libs/rubberband )
 	sdl? (
-		media-libs/libsdl2[X,opengl,video]
+		media-libs/libsdl2[opengl,video]
 		media-libs/sdl2-image
 	)
 	vidstab? ( media-libs/vidstab )
diff --git a/media-video/mkvtoolnix/mkvtoolnix-60.0.0.ebuild b/media-video/mkvtoolnix/mkvtoolnix-60.0.0.ebuild
index d249b81b99d..5806e53e49a 100644
--- a/media-video/mkvtoolnix/mkvtoolnix-60.0.0.ebuild
+++ b/media-video/mkvtoolnix/mkvtoolnix-60.0.0.ebuild
@@ -19,7 +19,7 @@ HOMEPAGE="https://mkvtoolnix.download/ https://gitlab.com/mbunkus/mkvtoolnix"
 
 LICENSE="GPL-2"
 SLOT="0"
-IUSE="dbus debug dvd nls pch test"
+IUSE="dbus debug dvd gui nls pch test"
 RESTRICT="!test? ( test )"
 
 # check NEWS.md for build system changes entries for boost/libebml/libmatroska
@@ -44,7 +44,7 @@ RDEPEND="
 	dev-qt/qtwidgets:5
 	dev-qt/qtconcurrent:5
 	dev-qt/qtmultimedia:5
-	app-text/cmark:0=
+	gui? ( app-text/cmark:0= )
 	dbus? ( dev-qt/qtdbus:5 )
 "
 DEPEND="${RDEPEND}
@@ -88,6 +88,7 @@ src_configure() {
 		$(use_enable debug)
 		$(usex pch "" --disable-precompiled-headers)
 		$(use_enable dbus)
+		$(use_enable gui)
 		--disable-qt6
 		--enable-qt5
 		--with-qmake=$(qt5_get_bindir)/qmake
diff --git a/net-misc/spice-gtk/spice-gtk-0.41.ebuild b/net-misc/spice-gtk/spice-gtk-0.41.ebuild
index 4f9503f25b0..bbb46b0fbe7 100644
--- a/net-misc/spice-gtk/spice-gtk-0.41.ebuild
+++ b/net-misc/spice-gtk/spice-gtk-0.41.ebuild
@@ -99,6 +99,7 @@ src_prepare() {
 
 src_configure() {
 	local emesonargs=(
+		-Dcoroutine=gthread
 		$(meson_feature gtk3 gtk)
 		$(meson_feature introspection)
 		$(meson_use mjpeg builtin-mjpeg)
diff --git a/sys-apps/attr/attr-2.5.1.ebuild b/sys-apps/attr/attr-2.5.1.ebuild
index 0d212005d51..343d88c82bb 100644
--- a/sys-apps/attr/attr-2.5.1.ebuild
+++ b/sys-apps/attr/attr-2.5.1.ebuild
@@ -42,7 +42,6 @@ src_configure() {
 	# Remove -flto* from flags as this breaks binaries (bug #644048)
 	filter-flags -flto*
 	append-ldflags "-Wl,--no-gc-sections" #700116
-	tc-ld-force-bfd #644048
 	append-lfs-flags #760857
 	multilib-minimal_src_configure
 }
diff --git a/sys-devel/clang/clang-14.0.6.ebuild b/sys-devel/clang/clang-14.0.6.ebuild
index ed84f3755fa..a2a46062a93 100644
--- a/sys-devel/clang/clang-14.0.6.ebuild
+++ b/sys-devel/clang/clang-14.0.6.ebuild
@@ -23,8 +23,11 @@ IUSE="
 REQUIRED_USE="${PYTHON_REQUIRED_USE}"
 RESTRICT="!test? ( test )"
 
+IUSE="${IUSE} +flang gcc-symlinks"
+
 DEPEND="
 	~sys-devel/llvm-${PV}:${SLOT}=[debug=,${MULTILIB_USEDEP}]
+	default-compiler-rt? ( =sys-libs/compiler-rt-${PV%_*}* )
 	static-analyzer? ( dev-lang/perl:* )
 	xml? ( dev-libs/libxml2:2=[${MULTILIB_USEDEP}] )
 "
@@ -46,10 +49,13 @@ PDEPEND="
 	sys-devel/clang-common
 	~sys-devel/clang-runtime-${PV}
 	default-compiler-rt? (
-		=sys-libs/compiler-rt-${PV%_*}*
 		llvm-libunwind? ( sys-libs/llvm-libunwind )
 		!llvm-libunwind? ( sys-libs/libunwind )
 	)
+	flang? (
+		dev-libs/libpgmath
+		dev-lang/flang
+		sys-devel/llvm[flang] )
 	default-libcxx? ( >=sys-libs/libcxx-${PV} )
 	default-lld? ( sys-devel/lld )
 "
@@ -230,8 +236,7 @@ get_distribution_components() {
 }
 
 multilib_src_configure() {
-	local llvm_version=$(llvm-config --version) || die
-	local clang_version=$(ver_cut 1-3 "${llvm_version}")
+	local clang_version=$(ver_cut 1-3)
 
 	local mycmakeargs=(
 		-DLLVM_CMAKE_PATH="${EPREFIX}/usr/lib/llvm/${SLOT}/$(get_libdir)/cmake/llvm"
@@ -246,6 +251,7 @@ multilib_src_configure() {
 
 		-DLLVM_TARGETS_TO_BUILD="${LLVM_TARGETS// /;}"
 		-DLLVM_BUILD_TESTS=$(usex test)
+		-DLLVM_ENABLE_CLASSIC_FLANG=$(usex flang)
 
 		# these are not propagated reliably, so redefine them
 		-DLLVM_ENABLE_EH=ON
@@ -266,6 +272,10 @@ multilib_src_configure() {
 		-DCLANG_DEFAULT_PIE_ON_LINUX=$(usex pie)
 		-DCLANG_DEFAULT_UNWINDLIB=$(usex default-compiler-rt libunwind "")
 
+		# create gcc symlinks: gcc g++ cc cpp c++
+		-DCLANG_INSTALL_GCC_SYMLINKS=$(usex gcc-symlinks)
+
+
 		-DCLANG_ENABLE_ARCMT=$(usex static-analyzer)
 		-DCLANG_ENABLE_STATIC_ANALYZER=$(usex static-analyzer)
 
@@ -359,12 +369,15 @@ src_install() {
 
 	# Apply CHOST and version suffix to clang tools
 	# note: we use two version components here (vs 3 in runtime path)
-	local llvm_version=$(llvm-config --version) || die
-	local clang_version=$(ver_cut 1 "${llvm_version}")
-	local clang_full_version=$(ver_cut 1-3 "${llvm_version}")
-	local clang_tools=( clang clang++ clang-cl clang-cpp )
+	local clang_version=$(ver_cut 1)
+	local clang_full_version=$(ver_cut 1-3)
+	local clang_tools=( clang clang++ clang-cl clang-cpp gcc g++ cpp )
 	local abi i
 
+	if use flang; then
+		clang_tools+=( flang gfortran )
+	fi
+
 	# cmake gives us:
 	# - clang-X
 	# - clang -> clang-X
diff --git a/sys-devel/distcc/distcc-3.4-r1.ebuild b/sys-devel/distcc/distcc-3.4-r1.ebuild
index 1bee4f2f642..dfee68afeec 100644
--- a/sys-devel/distcc/distcc-3.4-r1.ebuild
+++ b/sys-devel/distcc/distcc-3.4-r1.ebuild
@@ -73,6 +73,7 @@ src_configure() {
 		$(use_enable ipv6 rfc2553)
 		$(use_with gtk)
 		--without-gnome
+		--without-libiberty
 		$(use_with gssapi auth)
 		$(use_with zeroconf avahi)
 	)
diff --git a/sys-devel/distcc/files/distccd.service-1 b/sys-devel/distcc/files/distccd.service-1
index 9bd470297df..fd481026e27 100644
--- a/sys-devel/distcc/files/distccd.service-1
+++ b/sys-devel/distcc/files/distccd.service-1
@@ -4,7 +4,7 @@ After=network.target
 
 [Service]
 User=distcc
-ExecStart=/usr/bin/distccd --no-detach --daemon --port 3632 -N 15 --allow $ALLOWED_SERVERS
+ExecStart=/usr/bin/distccd --no-detach --daemon --port 3632 -N 15 $ALLOWED_SERVERS
 
 [Install]
 WantedBy=multi-user.target
diff --git a/sys-devel/llvm/llvm-14.0.6-r1.ebuild b/sys-devel/llvm/llvm-14.0.6-r1.ebuild
index f21b43b8bd2..00dd14ac38b 100644
--- a/sys-devel/llvm/llvm-14.0.6-r1.ebuild
+++ b/sys-devel/llvm/llvm-14.0.6-r1.ebuild
@@ -22,6 +22,8 @@ KEYWORDS="~amd64 ~arm ~arm64 ~ppc ~ppc64 ~riscv ~sparc ~x86 ~amd64-linux ~ppc-ma
 IUSE="+binutils-plugin debug doc exegesis libedit +libffi ncurses test xar xml z3"
 RESTRICT="!test? ( test )"
 
+IUSE="${IUSE} +flang binutils-symlinks"
+
 RDEPEND="
 	sys-libs/zlib:0=[${MULTILIB_USEDEP}]
 	binutils-plugin? ( >=sys-devel/binutils-2.31.1-r4:*[plugins] )
@@ -320,6 +322,23 @@ get_distribution_components() {
 				docs-llvm-man
 			)
 		fi
+
+		use binutils-symlinks && out+=(
+			addr2line
+			ar
+			c++filt
+			dlltool
+			dwp
+			nm
+			objcopy
+			objdump
+			ranlib
+			readelf
+			size
+			strings
+			strip
+                )
+
 		use doc && out+=(
 			docs-llvm-html
 		)
@@ -351,6 +370,7 @@ multilib_src_configure() {
 		-DLLVM_BUILD_LLVM_DYLIB=ON
 		-DLLVM_LINK_LLVM_DYLIB=ON
 		-DLLVM_DISTRIBUTION_COMPONENTS=$(get_distribution_components)
+		-DLLVM_INSTALL_BINUTILS_SYMLINKS=$(usex binutils-symlinks)
 
 		# cheap hack: LLVM combines both anyway, and the only difference
 		# is that the former list is explicitly verified at cmake time
@@ -367,6 +387,7 @@ multilib_src_configure() {
 		-DLLVM_ENABLE_EH=ON
 		-DLLVM_ENABLE_RTTI=ON
 		-DLLVM_ENABLE_Z3_SOLVER=$(usex z3)
+		-DLLVM_ENABLE_CLASSIC_FLANG=$(usex flang)
 
 		-DLLVM_HOST_TRIPLE="${CHOST}"
 
diff --git a/sys-libs/libcxx/libcxx-14.0.6.ebuild b/sys-libs/libcxx/libcxx-14.0.6.ebuild
index 4d6be5bcc7f..45373883e66 100644
--- a/sys-libs/libcxx/libcxx-14.0.6.ebuild
+++ b/sys-libs/libcxx/libcxx-14.0.6.ebuild
@@ -133,6 +133,7 @@ multilib_src_configure() {
 		-DLLVM_ENABLE_RUNTIMES=libcxx
 		-DLLVM_INCLUDE_TESTS=OFF
 		-DLLVM_LIBDIR_SUFFIX=${libdir#lib}
+		-DCMAKE_POSITION_INDEPENDENT_CODE=ON
 
 		-DLIBCXX_ENABLE_SHARED=ON
 		-DLIBCXX_ENABLE_STATIC=$(usex static-libs)
diff --git a/sys-libs/libcxxabi/libcxxabi-14.0.6.ebuild b/sys-libs/libcxxabi/libcxxabi-14.0.6.ebuild
index 7d864cba362..530c2e82825 100644
--- a/sys-libs/libcxxabi/libcxxabi-14.0.6.ebuild
+++ b/sys-libs/libcxxabi/libcxxabi-14.0.6.ebuild
@@ -79,6 +79,7 @@ multilib_src_configure() {
 		-DLIBCXXABI_INCLUDE_TESTS=$(usex test)
 		-DLIBCXXABI_USE_LLVM_UNWINDER=$(usex libunwind)
 		-DLIBCXXABI_USE_COMPILER_RT=${want_compiler_rt}
+		-DCMAKE_POSITION_INDEPENDENT_CODE=ON
 
 		# upstream is omitting standard search path for this
 		# probably because gcc & clang are bundling their own unwind.h
diff --git a/x11-misc/appmenu-gtk-module/appmenu-gtk-module-0.7.3-r2.ebuild b/x11-misc/appmenu-gtk-module/appmenu-gtk-module-0.7.3-r2.ebuild
index 46a05319bac..24d0dfdc8cf 100644
--- a/x11-misc/appmenu-gtk-module/appmenu-gtk-module-0.7.3-r2.ebuild
+++ b/x11-misc/appmenu-gtk-module/appmenu-gtk-module-0.7.3-r2.ebuild
@@ -16,7 +16,6 @@ IUSE="wayland"
 
 RDEPEND="
 	dev-libs/glib[dbus]
-	>=x11-libs/gtk+-2.24.0:2
 	>=x11-libs/gtk+-3.22.0:3[wayland=]
 "
 DEPEND="${RDEPEND}
