diff --git a/dev-haskell/ghc-exactprint/ghc-exactprint-0.6.3.2.ebuild b/dev-haskell/ghc-exactprint/ghc-exactprint-0.6.3.2.ebuild
index 96cf114f1..1874474e2 100644
--- a/dev-haskell/ghc-exactprint/ghc-exactprint-0.6.3.2.ebuild
+++ b/dev-haskell/ghc-exactprint/ghc-exactprint-0.6.3.2.ebuild
@@ -19,6 +19,7 @@ KEYWORDS="~amd64 ~x86"
 IUSE="roundtrip"
 
 RDEPEND=">=dev-lang/ghc-7.4.1:=
+	dev-haskell/free:=
 	roundtrip? ( dev-haskell/diff:=[profile?]
 			dev-haskell/filemanip:=[profile?]
 			>=dev-haskell/ghc-paths-0.1:=[profile?]
diff --git a/dev-haskell/nats/nats-1.1.2.ebuild b/dev-haskell/nats/nats-1.1.2.ebuild
index 170d4241f..5e0362a31 100644
--- a/dev-haskell/nats/nats-1.1.2.ebuild
+++ b/dev-haskell/nats/nats-1.1.2.ebuild
@@ -24,7 +24,7 @@ RESTRICT=test # avoid circular dep: nats[hashable]->hashable[test]->quickcheck->
 
 RDEPEND=">=dev-lang/ghc-7.4.1:=
 	binary? ( >=dev-haskell/binary-0.2:=[profile?] <dev-haskell/binary-0.9:=[profile?] )
-	hashable? ( >=dev-haskell/hashable-1.1:=[profile?] <dev-haskell/hashable-1.3:=[profile?] )
+	hashable? ( >=dev-haskell/hashable-1.1:=[profile?] <dev-haskell/hashable-1.4:=[profile?] )
 "
 DEPEND="${RDEPEND}"
 
diff --git a/dev-haskell/pointful/pointful-1.1.0.0.ebuild b/dev-haskell/pointful/pointful-1.1.0.0.ebuild
index 01c0fd228..89729620f 100644
--- a/dev-haskell/pointful/pointful-1.1.0.0.ebuild
+++ b/dev-haskell/pointful/pointful-1.1.0.0.ebuild
@@ -30,5 +30,6 @@ src_prepare() {
 	default
 
 	cabal_chdeps \
-		'haskell-src-exts-simple >= 1.18 && < 1.21' 'haskell-src-exts-simple >= 1.18'
+		'haskell-src-exts-simple >= 1.18 && < 1.21' 'haskell-src-exts-simple >= 1.18' \
+		'base                    >= 4.7  && < 4.13 || ^>= 4.13' 'base                    >= 4.7'
 }
diff --git a/dev-lang/idris/idris-1.3.3.ebuild b/dev-lang/idris/idris-1.3.3.ebuild
index cf5b68fa9..f613105ae 100644
--- a/dev-lang/idris/idris-1.3.3.ebuild
+++ b/dev-lang/idris/idris-1.3.3.ebuild
@@ -56,7 +56,7 @@ RDEPEND=">=dev-haskell/aeson-0.6:=[profile?] <dev-haskell/aeson-1.5:=[profile?]
 		dev-libs/gmp )
 "
 DEPEND="${RDEPEND}
-	>=dev-haskell/cabal-2.4 <dev-haskell/cabal-3.1
+	>=dev-haskell/cabal-2.4 <dev-haskell/cabal-3.3
 	test? ( dev-haskell/tagged
 		>=dev-haskell/tasty-0.8
 		>=dev-haskell/tasty-golden-2.0
