diff --git a/dev-haskell/binary/binary-0.8.8.0.ebuild b/dev-haskell/binary/binary-0.8.8.0.ebuild
index 7c2886050..eef12e11a 100644
--- a/dev-haskell/binary/binary-0.8.8.0.ebuild
+++ b/dev-haskell/binary/binary-0.8.8.0.ebuild
@@ -28,4 +28,4 @@ RDEPEND=">=dev-lang/ghc-7.8.2:=
 DEPEND="${RDEPEND}
 "
 
-CABAL_CORE_LIB_GHC_PV="PM:8.10.1 PM:8.10.2 PM:9999"
+CABAL_CORE_LIB_GHC_PV="PM:8.10.1 PM:8.10.2 PM:8.10.3 PM:9999"
diff --git a/dev-haskell/cabal/cabal-3.2.1.0.ebuild b/dev-haskell/cabal/cabal-3.2.1.0.ebuild
index 2e36d7a79..7faebb53d 100644
--- a/dev-haskell/cabal/cabal-3.2.1.0.ebuild
+++ b/dev-haskell/cabal/cabal-3.2.1.0.ebuild
@@ -36,6 +36,8 @@ DEPEND="${RDEPEND}"
 
 S="${WORKDIR}/${MY_P}"
 
+CABAL_CORE_LIB_GHC_PV="PM:8.10.3"
+
 src_configure() {
 	haskell-cabal_src_configure \
 		--flag=-bundled-binary-generic
diff --git a/dev-haskell/exceptions/exceptions-0.10.4.ebuild b/dev-haskell/exceptions/exceptions-0.10.4.ebuild
index dc621f43d..3a9de1cac 100644
--- a/dev-haskell/exceptions/exceptions-0.10.4.ebuild
+++ b/dev-haskell/exceptions/exceptions-0.10.4.ebuild
@@ -31,7 +31,7 @@ DEPEND="${RDEPEND}
 		>=dev-haskell/test-framework-quickcheck2-0.3 <dev-haskell/test-framework-quickcheck2-0.4 )
 "
 
-CABAL_CORE_LIB_GHC_PV="PM:8.10.1 PM:8.10.2"
+CABAL_CORE_LIB_GHC_PV="PM:8.10.1 PM:8.10.2 PM:8.10.3"
 
 src_prepare() {
 	default
diff --git a/dev-haskell/mtl/mtl-2.2.2.ebuild b/dev-haskell/mtl/mtl-2.2.2.ebuild
index 5d80076de..9c8262444 100644
--- a/dev-haskell/mtl/mtl-2.2.2.ebuild
+++ b/dev-haskell/mtl/mtl-2.2.2.ebuild
@@ -24,4 +24,4 @@ RDEPEND=">=dev-lang/ghc-7.10.1:=
 "
 DEPEND="${RDEPEND}"
 
-CABAL_CORE_LIB_GHC_PV="PM:8.4.2_rc1 PM:8.4.2 PM:8.4.3 PM:8.4.4 PM:8.6.1 PM:8.6.2 PM:8.6.3 PM:8.6.4 PM:8.6.5 PM:8.8.1 PM:8.8.2 PM:8.8.3 PM:8.8.4 PM:8.10.1_alpha1 PM:8.10.1_alpha2 PM:8.10.1 PM:8.10.2 PM:9999"
+CABAL_CORE_LIB_GHC_PV="PM:8.4.2_rc1 PM:8.4.2 PM:8.4.3 PM:8.4.4 PM:8.6.1 PM:8.6.2 PM:8.6.3 PM:8.6.4 PM:8.6.5 PM:8.8.1 PM:8.8.2 PM:8.8.3 PM:8.8.4 PM:8.10.1_alpha1 PM:8.10.1_alpha2 PM:8.10.1 PM:8.10.2 PM:8.10.3 PM:9999"
diff --git a/dev-haskell/nats/nats-1.1.2.ebuild b/dev-haskell/nats/nats-1.1.2.ebuild
index 170d4241f..5e0362a31 100644
--- a/dev-haskell/nats/nats-1.1.2.ebuild
+++ b/dev-haskell/nats/nats-1.1.2.ebuild
@@ -24,7 +24,7 @@ RESTRICT=test # avoid circular dep: nats[hashable]->hashable[test]->quickcheck->
 
 RDEPEND=">=dev-lang/ghc-7.4.1:=
 	binary? ( >=dev-haskell/binary-0.2:=[profile?] <dev-haskell/binary-0.9:=[profile?] )
-	hashable? ( >=dev-haskell/hashable-1.1:=[profile?] <dev-haskell/hashable-1.3:=[profile?] )
+	hashable? ( >=dev-haskell/hashable-1.1:=[profile?] <dev-haskell/hashable-1.4:=[profile?] )
 "
 DEPEND="${RDEPEND}"
 
diff --git a/dev-haskell/parsec/parsec-3.1.14.0.ebuild b/dev-haskell/parsec/parsec-3.1.14.0.ebuild
index 7cebba59c..48b28da7c 100644
--- a/dev-haskell/parsec/parsec-3.1.14.0.ebuild
+++ b/dev-haskell/parsec/parsec-3.1.14.0.ebuild
@@ -31,4 +31,4 @@ RDEPEND=">=dev-haskell/fail-4.9:=[profile?] <dev-haskell/fail-4.10:=[profile?]
 "
 DEPEND="${RDEPEND}"
 
-CABAL_CORE_LIB_GHC_PV="PM:8.8.1 PM:8.8.2 PM:8.8.3 PM:8.8.4 PM:8.10.1_alpha1 PM:8.10.1_alpha2 PM:8.10.1 PM:8.10.2 PM:9999"
+CABAL_CORE_LIB_GHC_PV="PM:8.8.1 PM:8.8.2 PM:8.8.3 PM:8.8.4 PM:8.10.1_alpha1 PM:8.10.1_alpha2 PM:8.10.1 PM:8.10.2 PM:8.10.3 PM:9999"
diff --git a/dev-haskell/stm/stm-2.5.0.0.ebuild b/dev-haskell/stm/stm-2.5.0.0.ebuild
index 8cdccf3ae..5e7f1959d 100644
--- a/dev-haskell/stm/stm-2.5.0.0.ebuild
+++ b/dev-haskell/stm/stm-2.5.0.0.ebuild
@@ -25,4 +25,4 @@ DEPEND="${RDEPEND}
 	>=dev-haskell/cabal-1.10
 "
 
-CABAL_CORE_LIB_GHC_PV="PM:8.6.1 PM:8.6.2 PM:8.6.3 PM:8.6.4 PM:8.6.5 PM:8.8.1 PM:8.8.2 PM:8.8.3 PM:8.8.4 PM:8.10.1_alpha1 PM:8.10.1_alpha2 PM:8.10.1 PM:8.10.2 PM:9999"
+CABAL_CORE_LIB_GHC_PV="PM:8.6.1 PM:8.6.2 PM:8.6.3 PM:8.6.4 PM:8.6.5 PM:8.8.1 PM:8.8.2 PM:8.8.3 PM:8.8.4 PM:8.10.1_alpha1 PM:8.10.1_alpha2 PM:8.10.1 PM:8.10.2 PM:8.10.3 PM:9999"
diff --git a/dev-haskell/terminfo/terminfo-0.4.1.4.ebuild b/dev-haskell/terminfo/terminfo-0.4.1.4.ebuild
index 8598129df..2e7a417ea 100644
--- a/dev-haskell/terminfo/terminfo-0.4.1.4.ebuild
+++ b/dev-haskell/terminfo/terminfo-0.4.1.4.ebuild
@@ -24,4 +24,4 @@ DEPEND="${RDEPEND}
 	>=dev-haskell/cabal-1.24.0.0
 "
 
-CABAL_CORE_LIB_GHC_PV="PM:8.8.1 PM:8.8.2 PM:8.8.3 PM:8.8.4 PM:8.10.1_alpha1 PM:8.10.1_alpha2 PM:8.10.1 PM:8.10.2 PM:9999"
+CABAL_CORE_LIB_GHC_PV="PM:8.8.1 PM:8.8.2 PM:8.8.3 PM:8.8.4 PM:8.10.1_alpha1 PM:8.10.1_alpha2 PM:8.10.1 PM:8.10.2 PM:8.10.3 PM:9999"
diff --git a/dev-haskell/text/text-1.2.4.0-r1.ebuild b/dev-haskell/text/text-1.2.4.0-r1.ebuild
index e75fb6f25..020889488 100644
--- a/dev-haskell/text/text-1.2.4.0-r1.ebuild
+++ b/dev-haskell/text/text-1.2.4.0-r1.ebuild
@@ -33,7 +33,7 @@ CABAL_CORE_LIB_GHC_PV="PM:8.8.1 PM:8.8.2 PM:8.8.3 PM:8.8.4 PM:9999"
 # text version downgrade. portage can't figure out what is more
 # important to hold back on an old version: ghc or text.
 # Let's keep a lie until ghc-8.10.3 is ready.
-CABAL_CORE_LIB_GHC_PV+=" PM:8.10.2"
+CABAL_CORE_LIB_GHC_PV+=" PM:8.10.2 PM:8.10.3"
 
 src_configure() {
 	haskell-cabal_src_configure \
diff --git a/dev-haskell/transformers/transformers-0.5.6.2.ebuild b/dev-haskell/transformers/transformers-0.5.6.2.ebuild
index a4d3e18b9..9e3cae2eb 100644
--- a/dev-haskell/transformers/transformers-0.5.6.2.ebuild
+++ b/dev-haskell/transformers/transformers-0.5.6.2.ebuild
@@ -24,4 +24,4 @@ RDEPEND=">=dev-lang/ghc-7.4.1:=
 "
 DEPEND="${RDEPEND}
 "
-CABAL_CORE_LIB_GHC_PV="PM:8.6.4 PM:8.6.5 PM:8.8.1 PM:8.8.2 PM:8.8.3 PM:8.8.4 PM:8.10.1_alpha1 PM:8.10.1_alpha2 PM:8.10.1 PM:8.10.2 PM:9999"
+CABAL_CORE_LIB_GHC_PV="PM:8.6.4 PM:8.6.5 PM:8.8.1 PM:8.8.2 PM:8.8.3 PM:8.8.4 PM:8.10.1_alpha1 PM:8.10.1_alpha2 PM:8.10.1 PM:8.10.2 PM:8.10.3 PM:9999"
diff --git a/dev-haskell/xhtml/xhtml-3000.2.2.1.ebuild b/dev-haskell/xhtml/xhtml-3000.2.2.1.ebuild
index 570b8317b..b1253be56 100644
--- a/dev-haskell/xhtml/xhtml-3000.2.2.1.ebuild
+++ b/dev-haskell/xhtml/xhtml-3000.2.2.1.ebuild
@@ -21,4 +21,4 @@ RDEPEND=">=dev-lang/ghc-8.4.2:="
 DEPEND="${RDEPEND}
 	>=dev-haskell/cabal-1.6"
 
-CABAL_CORE_LIB_GHC_PV="PM:8.4.2_rc1 PM:8.4.2 PM:8.4.3 PM:8.4.4 PM:8.6.1 PM:8.6.2 PM:8.6.3 PM:8.6.4 PM:8.6.5 PM:8.8.1 PM:8.8.2 PM:8.8.3 PM:8.8.4 PM:8.10.1 PM:8.10.2 PM:9999"
+CABAL_CORE_LIB_GHC_PV="PM:8.4.2_rc1 PM:8.4.2 PM:8.4.3 PM:8.4.4 PM:8.6.1 PM:8.6.2 PM:8.6.3 PM:8.6.4 PM:8.6.5 PM:8.8.1 PM:8.8.2 PM:8.8.3 PM:8.8.4 PM:8.10.1 PM:8.10.2 PM:8.10.3 PM:9999"
diff --git a/dev-lang/ghc/ghc-8.10.3.ebuild b/dev-lang/ghc/ghc-8.10.3.ebuild
index 1f1c51fca..9f49f86e3 100644
--- a/dev-lang/ghc/ghc-8.10.3.ebuild
+++ b/dev-lang/ghc/ghc-8.10.3.ebuild
@@ -75,7 +75,7 @@ BUMP_LIBRARIES=(
 
 LICENSE="BSD"
 SLOT="0/${PV}"
-#KEYWORDS="~amd64 ~x86"
+KEYWORDS="~amd64 ~x86"
 IUSE="big-endian doc elfutils ghcbootstrap ghcmakebinary +gmp numa profile test"
 IUSE+=" binary"
 RESTRICT="!test? ( test )"
