diff --git a/dev-haskell/exceptions/exceptions-0.10.4.ebuild b/dev-haskell/exceptions/exceptions-0.10.4.ebuild
index 8a15e77539..0d1f131098 100644
--- a/dev-haskell/exceptions/exceptions-0.10.4.ebuild
+++ b/dev-haskell/exceptions/exceptions-0.10.4.ebuild
@@ -29,7 +29,7 @@ DEPEND="${RDEPEND}
 		>=dev-haskell/test-framework-quickcheck2-0.3 <dev-haskell/test-framework-quickcheck2-0.4 )
 "
 
-CABAL_CORE_LIB_GHC_PV="PM:8.10.5-r1 PM:8.10.6 PM:8.10.7 PM:9.0.1-r1 PM:9.0.2-r1 PM:9999"
+CABAL_CORE_LIB_GHC_PV="PM:8.10.5-r1 PM:8.10.6 PM:8.10.7 PM:9.0.1-r1 PM:9.0.2-r1 PM:9.2.3 PM:9999"
 
 src_prepare() {
 	default
diff --git a/dev-haskell/haskeline/haskeline-0.8.2.ebuild b/dev-haskell/haskeline/haskeline-0.8.2.ebuild
index b3fbea374e..e487626188 100644
--- a/dev-haskell/haskeline/haskeline-0.8.2.ebuild
+++ b/dev-haskell/haskeline/haskeline-0.8.2.ebuild
@@ -36,4 +36,4 @@ src_configure() {
 }
 
 # ghc-9.0.1 actually comes bundled with haskeline-0.8.1.0
-CABAL_CORE_LIB_GHC_PV="PM:8.10.6 PM:8.10.7 PM:9.0.1-r1 PM:9.0.2-r1 PM:9999"
+CABAL_CORE_LIB_GHC_PV="PM:8.10.6 PM:8.10.7 PM:9.0.1-r1 PM:9.0.2-r1 PM:9.2.3 PM:9999"
diff --git a/dev-haskell/mtl/mtl-2.2.2.ebuild b/dev-haskell/mtl/mtl-2.2.2.ebuild
index a5a3cfbf3b..da8044cdbf 100644
--- a/dev-haskell/mtl/mtl-2.2.2.ebuild
+++ b/dev-haskell/mtl/mtl-2.2.2.ebuild
@@ -22,4 +22,4 @@ RDEPEND=">=dev-lang/ghc-7.10.1:=
 "
 DEPEND="${RDEPEND}"
 
-CABAL_CORE_LIB_GHC_PV="PM:8.8.4-r1 PM:8.10.5-r1 PM:8.10.6 PM:8.10.7 PM:9.0.1-r1 PM:9.0.2-r1 PM:9999"
+CABAL_CORE_LIB_GHC_PV="PM:8.8.4-r1 PM:8.10.5-r1 PM:8.10.6 PM:8.10.7 PM:9.0.1-r1 PM:9.0.2-r1 PM:9.2.3 PM:9999"
diff --git a/dev-haskell/terminfo/terminfo-0.4.1.5.ebuild b/dev-haskell/terminfo/terminfo-0.4.1.5.ebuild
index 7fc50e3a6f..2497a5a1db 100644
--- a/dev-haskell/terminfo/terminfo-0.4.1.5.ebuild
+++ b/dev-haskell/terminfo/terminfo-0.4.1.5.ebuild
@@ -22,4 +22,4 @@ DEPEND="${RDEPEND}
 	>=dev-haskell/cabal-2.2.0.1
 "
 
-CABAL_CORE_LIB_GHC_PV="PM:9.0.2-r1 PM:9999"
+CABAL_CORE_LIB_GHC_PV="PM:9.0.2-r1 PM:9.2.3 PM:9999"
diff --git a/dev-haskell/text/text-1.2.5.0.ebuild b/dev-haskell/text/text-1.2.5.0.ebuild
index fbdceb17cc..c99abb4a23 100644
--- a/dev-haskell/text/text-1.2.5.0.ebuild
+++ b/dev-haskell/text/text-1.2.5.0.ebuild
@@ -32,4 +32,4 @@ src_configure() {
 		--flag=-developer
 }
 
-CABAL_CORE_LIB_GHC_PV="PM:9.0.2-r1 PM:9999"
+CABAL_CORE_LIB_GHC_PV="PM:9.0.2-r1 PM:9.2.3 PM:9999"
diff --git a/dev-haskell/transformers/transformers-0.5.6.2.ebuild b/dev-haskell/transformers/transformers-0.5.6.2.ebuild
index db550adb8a..1a91d8533a 100644
--- a/dev-haskell/transformers/transformers-0.5.6.2.ebuild
+++ b/dev-haskell/transformers/transformers-0.5.6.2.ebuild
@@ -22,4 +22,4 @@ RDEPEND=">=dev-lang/ghc-7.4.1:=
 "
 DEPEND="${RDEPEND}
 "
-CABAL_CORE_LIB_GHC_PV="PM:8.8.4-r1 PM:8.10.5-r1 PM:8.10.6 PM:8.10.7 PM:9.0.1-r1 PM:9.0.2-r1 PM:9999"
+CABAL_CORE_LIB_GHC_PV="PM:8.8.4-r1 PM:8.10.5-r1 PM:8.10.6 PM:8.10.7 PM:9.0.1-r1 PM:9.0.2-r1 PM:9.2.3 PM:9999"
diff --git a/dev-haskell/xhtml/xhtml-3000.2.2.1.ebuild b/dev-haskell/xhtml/xhtml-3000.2.2.1.ebuild
index 51debaf7ef..1142ba8118 100644
--- a/dev-haskell/xhtml/xhtml-3000.2.2.1.ebuild
+++ b/dev-haskell/xhtml/xhtml-3000.2.2.1.ebuild
@@ -20,4 +20,4 @@ RDEPEND=">=dev-lang/ghc-8.4.2:="
 DEPEND="${RDEPEND}
 	>=dev-haskell/cabal-1.6"
 
-CABAL_CORE_LIB_GHC_PV="PM:8.8.4-r1 PM:8.10.5-r1 PM:8.10.6 PM:8.10.7 PM:9.0.1-r1 PM:9.0.2-r1 PM:9999"
+CABAL_CORE_LIB_GHC_PV="PM:8.8.4-r1 PM:8.10.5-r1 PM:8.10.6 PM:8.10.7 PM:9.0.1-r1 PM:9.0.2-r1 PM:9.2.3 PM:9999"
diff --git a/eclass/ghc-package.eclass b/eclass/ghc-package.eclass
index 156ee372f3..ee5d208e32 100644
--- a/eclass/ghc-package.eclass
+++ b/eclass/ghc-package.eclass
@@ -58,11 +58,11 @@ ghc-getghcpkg() {
 # because for some reason the global package file
 # must be specified
 ghc-getghcpkgbin() {
-	local empty_db="${T}/empty.conf.d" ghc_pkg="$(ghc-libdir)/bin/ghc-pkg"
+	local empty_db="${T}/empty.conf.d" ghc_pkg="$(ghc-libdir)/../bin/ghc-pkg"
 	if [[ ! -d ${empty_db} ]]; then
 		"${ghc_pkg}" init "${empty_db}" || die "Failed to initialize empty global db"
 	fi
-	echo "$(ghc-libdir)/bin/ghc-pkg" "--global-package-db=${empty_db}"
+	echo "$(ghc-libdir)/../bin/ghc-pkg" "--global-package-db=${empty_db}"
 }
 
 # @FUNCTION: ghc-version
