diff --git a/sys-apps/tcp-wrappers/files/tcp-wrappers-7.6.22-remove-DECLS.patch b/sys-apps/tcp-wrappers/files/tcp-wrappers-7.6.22-remove-DECLS.patch
index 0fbd5ed..0015549 100644
--- a/sys-apps/tcp-wrappers/files/tcp-wrappers-7.6.22-remove-DECLS.patch
+++ b/sys-apps/tcp-wrappers/files/tcp-wrappers-7.6.22-remove-DECLS.patch
@@ -1,22 +1,27 @@
 diff -Naur tcp_wrappers_7.6.orig/tcpd.h tcp_wrappers_7.6/tcpd.h
 --- tcp_wrappers_7.6.orig/tcpd.h	2014-03-24 18:39:52.000000000 +0000
 +++ tcp_wrappers_7.6/tcpd.h	2014-03-24 18:48:21.000000000 +0000
-@@ -11,7 +11,9 @@
+@@ -11,7 +11,10 @@
  #include <netinet/in.h>
  #include <stdio.h>
  
+-__BEGIN_DECLS
++
 +#ifdef __cplusplus
- __BEGIN_DECLS
++extern "C" {
 +#endif
  
  /* Structure to describe one communications endpoint. */
  
-@@ -252,6 +254,8 @@
+@@ -252,6 +255,9 @@
  extern char *my_strtok();
  #endif
  
+-__END_DECLS
++
 +#ifdef __cplusplus
- __END_DECLS
++}
 +#endif
  
  #endif
+
diff --git a/sys-libs/libutempter/libutempter-1.1.6-r2.ebuild b/sys-libs/libutempter/libutempter-1.1.6-r3.ebuild
similarity index 71%
rename from sys-libs/libutempter/libutempter-1.1.6-r2.ebuild
rename to sys-libs/libutempter/libutempter-1.1.6-r3.ebuild
index 2f9f4f1..9275691 100644
--- a/sys-libs/libutempter/libutempter-1.1.6-r2.ebuild
+++ b/sys-libs/libutempter/libutempter-1.1.6-r3.ebuild
@@ -1,26 +1,27 @@
-# Copyright 1999-2017 Gentoo Foundation
+# Copyright 1999-2020 Gentoo Authors
 # Distributed under the terms of the GNU General Public License v2
 
-EAPI="4"
+EAPI=7
 
-inherit user multilib flag-o-matic
+inherit multilib flag-o-matic toolchain-funcs
 
-DESCRIPTION="Library that allows non-privileged apps to write utmp (login) info, which need root access"
-HOMEPAGE="http://altlinux.org/index.php?module=sisyphus&package=libutempter"
+DESCRIPTION="Library that allows non-privileged apps to write utmp (login) info"
+HOMEPAGE="https://altlinux.org/index.php?module=sisyphus&package=libutempter"
 SRC_URI="ftp://ftp.altlinux.org/pub/people/ldv/${PN}/${P}.tar.bz2"
 
 LICENSE="LGPL-2.1"
 SLOT="0"
-KEYWORDS="amd64 arm arm64 ~mips ppc x86"
+KEYWORDS="~alpha amd64 arm arm64 hppa ~ia64 ~m68k ~mips ppc ppc64 s390 sparc x86 ~amd64-linux ~x86-linux"
 IUSE="static-libs elibc_FreeBSD"
 
-RDEPEND="!sys-apps/utempter"
-
-pkg_setup() {
-	enewgroup utmp 406
-}
+RDEPEND="
+	!sys-apps/utempter
+	acct-group/utmp
+"
 
 src_prepare() {
+	default
+
 	local args=(
 		-e "/^libdir /s:/usr/lib:${EPREFIX}/usr/$(get_libdir):"
 		-e '/^libexecdir /s:=.*:= $(libdir)/misc:'
@@ -35,7 +36,7 @@ src_prepare() {
 		)
 	sed -i "${args[@]}" Makefile || die
 
-	epatch "${FILESDIR}"/${PN}-remove-glibc-assumption.patch
+	eapply "${FILESDIR}"/${PN}-remove-glibc-assumption.patch
 }
 
 src_configure() {
@@ -59,12 +60,12 @@ src_install() {
 }
 
 pkg_postinst() {
-	if [ -f "${EROOT}/var/log/wtmp" ] ; then
+	if [[ -f "${EROOT}/var/log/wtmp" ]] ; then
 		chown root:utmp "${EROOT}/var/log/wtmp"
 		chmod 664 "${EROOT}/var/log/wtmp"
 	fi
 
-	if [ -f "${EROOT}/var/run/utmp" ] ; then
+	if [[ -f "${EROOT}/var/run/utmp" ]] ; then
 		chown root:utmp "${EROOT}/var/run/utmp"
 		chmod 664 "${EROOT}/var/run/utmp"
 	fi
