From d02720d8a653bc564b84c27bfa01a6876bf5bec2 Mon Sep 17 00:00:00 2001
From: jbreeden <jared.breeden@gmail.com>
Date: Sat, 23 Jan 2016 17:13:25 -0800
Subject: [PATCH] Export git_stash_apply_init_options

---
 include/git2/stash.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/git2/stash.h b/include/git2/stash.h
index b321dc3..733d75a 100644
--- a/include/git2/stash.h
+++ b/include/git2/stash.h
@@ -150,7 +150,7 @@ typedef struct git_stash_apply_options {
  *        `GIT_STASH_APPLY_OPTIONS_INIT` here.
  * @return Zero on success; -1 on failure.
  */
-int git_stash_apply_init_options(
+GIT_EXTERN(int) git_stash_apply_init_options(
 	git_stash_apply_options *opts, unsigned int version);
 
 /**
