--- a/texk/web2c/luatexdir/image/pdftoepdf.w	2016-09-20 09:49:49.107157491 +0200
+++ b/texk/web2c/luatexdir/image/pdftoepdf.w	2016-09-20 09:50:39.952830027 +0200
@@ -70,7 +70,10 @@
         if (ck == NULL)
             luatex_fail("PDF inclusion: out of memory while processing '%s'",
                         a);
-        snprintf(ck, PDF_CHECKSUM_SIZE, "%" PRIu64 "_%" PRIu64, (uint64_t) size,
+        snprintf(ck, PDF_CHECKSUM_SIZE, "%"
+                 PRIu64
+                 "_%"
+                 PRIu64, (uint64_t) size,
                  (uint64_t) mtime);
    } else {
         switch (fe) {
