--- a/texlive-20160523-source/texk/web2c/luatexdir/image/pdftoepdf.w	2017-02-18 12:18:56.152111484 +0100
+++ b/texlive-20160523-source/texk/web2c/luatexdir/image/pdftoepdf.w	2017-02-18 12:19:54.351991428 +0100
@@ -71,7 +71,10 @@
         ck = (char *) malloc(PDF_CHECKSUM_SIZE);
         if (ck == NULL)
             formatted_error("pdf inclusion","out of memory while processing '%s'", a);
-        snprintf(ck, PDF_CHECKSUM_SIZE, "%" PRIu64 "_%" PRIu64, (uint64_t) size,(uint64_t) mtime);
+        snprintf(ck, PDF_CHECKSUM_SIZE, "%"
+                 PRIu64
+                 "_%"
+                 PRIu64, (uint64_t) size,(uint64_t) mtime);
    } else {
         switch (fe) {
             case FE_FAIL:
