--- a/tools/clang/lib/Driver/Tools.cpp	2016-08-16 15:57:14.404575818 +0200
+++ b/tools/clang/lib/Driver/Tools.cpp	2016-08-16 16:12:24.531443919 +0200
@@ -9358,6 +9358,7 @@
       !Args.hasArg(options::OPT_shared) && !Args.hasArg(options::OPT_static) &&
       !(Args.hasArg(options::OPT_fno_pie) || Args.hasArg(options::OPT_fno_PIE)) &&
       (Args.hasArg(options::OPT_pie) || ToolChain.isPIEDefault());
+  const bool isMusl = (ToolChain.getTriple().getEnvironment() == llvm::Triple::Musl);
   const bool HasCRTBeginEndFiles =
       ToolChain.getTriple().hasEnvironment() ||
       (ToolChain.getTriple().getVendor() != llvm::Triple::MipsTechnologies);
@@ -9443,12 +9446,13 @@
       if (crt1)
         CmdArgs.push_back(Args.MakeArgString(ToolChain.GetFilePath(crt1)));
 
-      CmdArgs.push_back(Args.MakeArgString(ToolChain.GetFilePath("crti.o")));
+      if (!isMusl)
+        CmdArgs.push_back(Args.MakeArgString(ToolChain.GetFilePath("crti.o")));
     }
 
     if (IsIAMCU)
       CmdArgs.push_back(Args.MakeArgString(ToolChain.GetFilePath("crt0.o")));
-    else {
+    else if (!isMusl) {
       const char *crtbegin;
       if (Args.hasArg(options::OPT_static))
         crtbegin = isAndroid ? "crtbegin_static.o" : "crtbeginT.o";
@@ -9566,7 +9570,7 @@
       }
     }
 
-    if (!Args.hasArg(options::OPT_nostartfiles) && !IsIAMCU) {
+    if (!Args.hasArg(options::OPT_nostartfiles) && !IsIAMCU && !isMusl) {
       const char *crtend;
       if (Args.hasArg(options::OPT_shared))
         crtend = isAndroid ? "crtend_so.o" : "crtendS.o";
