--- a/tools/clang/lib/Driver/Tools.cpp	2016-07-28 19:11:02.000000000 +0200
+++ b/tools/clang/lib/Driver/Tools.cpp	2016-08-11 16:02:49.273843730 +0200
@@ -5458,6 +5458,7 @@
               getToolChain().getTriple().getOS() != llvm::Triple::Solaris &&
               getToolChain().getArch() != llvm::Triple::hexagon &&
               getToolChain().getArch() != llvm::Triple::xcore &&
+              getToolChain().getTriple().getEnvironment() != llvm::Triple::Musl &&
               ((getToolChain().getTriple().getVendor() !=
                 llvm::Triple::MipsTechnologies) ||
                getToolChain().getTriple().hasEnvironment())) ||
