--- a/tools/clang/lib/Driver/ToolChains.cpp	2016-07-08 17:15:14.502085957 +0200
+++ b/tools/clang/lib/Driver/ToolChains.cpp	2016-07-08 17:24:08.631042007 +0200
@@ -4241,20 +4241,32 @@
     return Triple.isArch64Bit() ? "/system/bin/linker64" : "/system/bin/linker";
   else if (Triple.isMusl()) {
     std::string ArchName;
+    std::string LibDir;
     switch (Arch) {
     case llvm::Triple::thumb:
       ArchName = "arm";
+      LibDir = "lib";
       break;
     case llvm::Triple::thumbeb:
       ArchName = "armeb";
+      LibDir = "lib";
+      break;
+    case llvm::Triple::x86:
+      LibDir = "lib32";
+      ArchName = "i386";
+      break;
+    case llvm::Triple::x86_64:
+      LibDir = "lib64";
+      ArchName = "x86_64";
       break;
     default:
       ArchName = Triple.getArchName().str();
+      LibDir = "lib";
     }
     if (Triple.getEnvironment() == llvm::Triple::MuslEABIHF)
       ArchName += "hf";
 
-    return "/lib/ld-musl-" + ArchName + ".so.1";
+    return "/" + LibDir + "/ld-musl-" + ArchName + ".so.1";
   }
 
   std::string LibDir;
