--- a/tools/clang/lib/Driver/ToolChains.h	2016-01-07 23:09:47.000000000 +0100
+++ b/tools/clang/lib/Driver/ToolChains.h	2016-02-19 20:03:17.237467235 +0100
@@ -838,6 +838,14 @@
   void AddIAMCUIncludeArgs(const llvm::opt::ArgList &DriverArgs,
                            llvm::opt::ArgStringList &CC1Args) const override;
   bool isPIEDefault() const override;
+
+  unsigned GetDefaultStackProtectorLevel(bool KernelOrKext) const override {
+    StringRef VendorName = Linux::getTriple().getVendorName();
+    if (VendorName.compare("gentoo") == 0)
+      return 2;
+    return 1;
+  }
+
   SanitizerMask getSupportedSanitizers() const override;
   void addProfileRTLibs(const llvm::opt::ArgList &Args,
                         llvm::opt::ArgStringList &CmdArgs) const override;
