--- a/tools/clang/lib/Driver/ToolChains.cpp.bak	2015-08-30 14:35:23.207003151 -0200
+++ b/tools/clang/lib/Driver/ToolChains.cpp	2015-08-30 14:56:29.277967873 -0200
@@ -2870,6 +2870,7 @@
   // in this enum, because some tests are done by integer comparison against
   // the first and last known member in the family, e.g. IsRedHat().
   ArchLinux,
+  GentooLinux,
   DebianLenny,
   DebianSqueeze,
   DebianWheezy,
@@ -2990,6 +2991,9 @@
   if (D.getVFS().exists("/etc/arch-release"))
     return ArchLinux;
 
+  if (D.getVFS().exists("/etc/gentoo-release"))
+    return GentooLinux;
+
   return UnknownDistro;
 }
 
@@ -3158,7 +3162,12 @@
 
   Distro Distro = DetectDistro(Arch);
 
-  if (IsOpenSUSE(Distro) || IsUbuntu(Distro)) {
+  if (Distro == GentooLinux) {
+    ExtraOpts.push_back("-z");
+    ExtraOpts.push_back("now");
+  }
+
+  if (IsOpenSUSE(Distro) || IsUbuntu(Distro) || Distro == GentooLinux) {
     ExtraOpts.push_back("-z");
     ExtraOpts.push_back("relro");
   }
@@ -3178,7 +3187,7 @@
   // ABI requires a mapping between the GOT and the symbol table.
   // Android loader does not support .gnu.hash.
   if (!IsMips && !IsAndroid) {
-    if (IsRedhat(Distro) || IsOpenSUSE(Distro) ||
+    if (IsRedhat(Distro) || IsOpenSUSE(Distro) || Distro == GentooLinux ||
         (IsUbuntu(Distro) && Distro >= UbuntuMaverick))
       ExtraOpts.push_back("--hash-style=gnu");
