--- a/src/github.com/snapcore/snapd/cmd/snap-confine/snap-confine.apparmor.in
+++ b/src/github.com/snapcore/snapd/cmd/snap-confine/snap-confine.apparmor.in
@@ -13,6 +13,7 @@
     # We run privileged, so be fanatical about what we include and don't use
     # any abstractions
     /etc/ld.so.cache r,
+    /etc/ld-musl-x86_64.path r,
     /etc/ld.so.preload r,
     /{,usr/}lib{,32,64,x32}/{,@{multiarch}/}ld-*.so mrix,
     # libc, you are funny
