--- a/src/github.com/snapcore/snapd/cmd/libsnap-confine-private/tool.c
+++ b/src/github.com/snapcore/snapd/cmd/libsnap-confine-private/tool.c
@@ -210,7 +210,11 @@
 		if (apparmor != NULL && aa_profile != NULL) {
 			sc_maybe_aa_change_onexec(apparmor, aa_profile);
 		}
-		fexecve(tool_fd, argv, envp);
+		char path[100];
+		path[0] = '\0';
+		strcat(path, "/usr/lib/snapd/");
+		strcat(path, tool_name);
+		execve(path, argv, envp);
 		die("cannot execute snapd tool %s", tool_name);
 	} else {
 		int status = 0;
