--- a/src/github.com/snapcore/snapd/osutil/buildid.go
+++ b/src/github.com/snapcore/snapd/osutil/buildid.go
@@ -45,8 +45,8 @@
 //
 // See details at http://fedoraproject.org/wiki/Releases/FeatureBuildId
 func ReadBuildID(fname string) (string, error) {
-	const ELF_NOTE_GNU = "GNU\x00"
-	const NT_GNU_BUILD_ID uint32 = 3
+	const ELF_NOTE_GO = "Go\x00\x00"
+	const NT_GO_BUILD_ID uint32 = 4
 
 	// Open the designated ELF file
 	f, err := elf.Open(fname)
@@ -73,7 +73,7 @@
 		}
 
 		// We are looking for a specific type of note
-		if nHdr.Type != NT_GNU_BUILD_ID {
+		if nHdr.Type != NT_GO_BUILD_ID {
 			continue
 		}
 
@@ -83,8 +83,8 @@
 			return "", err
 		}
 
-		// We are only interested in GNU build IDs
-		if string(noteName) != ELF_NOTE_GNU {
+		// We are only interested in Go build IDs
+		if string(noteName) != ELF_NOTE_GO {
 			continue
 		}
 
