--- a/src/github.com/snapcore/snapd/cmd/libsnap-confine-private/apparmor-support.c
+++ b/src/github.com/snapcore/snapd/cmd/libsnap-confine-private/apparmor-support.c
@@ -116,7 +116,7 @@
 	debug("requesting changing of apparmor profile on next exec to %s",
 	      profile);
 	if (aa_change_onexec(profile) < 0) {
-		if (secure_getenv("SNAPPY_LAUNCHER_INSIDE_TESTS") == NULL) {
+		if (getenv("SNAPPY_LAUNCHER_INSIDE_TESTS") == NULL) {
 			die("cannot change profile for the next exec call");
 		}
 	}
