From 7acb26cd54a3b5d96103a0c267752e82e998f17c Mon Sep 17 00:00:00 2001
From: lsh123 <aleksey@aleksey.com>
Date: Thu, 13 Apr 2023 06:48:24 -0700
Subject: [PATCH] Add newline

Issue #626
---
 src/openssl/symkeys.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/openssl/symkeys.c b/src/openssl/symkeys.c
index 2585f19e..629e014f 100644
--- a/src/openssl/symkeys.c
+++ b/src/openssl/symkeys.c
@@ -618,4 +618,4 @@ xmlSecOpenSSLKeyDataPbkdf2Set(xmlSecKeyDataPtr data, const xmlSecByte* buf, xmlS
 
     return(xmlSecBufferSetData(buffer, buf, bufSize));
 }
-#endif /* XMLSEC_NO_PBKDF2 */
\ No newline at end of file
+#endif /* XMLSEC_NO_PBKDF2 */
