--- cargo-0.7.0/src/bin/cargo.rs	2015-12-08 18:00:31.000000000 +0000
+++ cargo-0.7.0/src/bin/cargo.rs	2015-12-27 21:19:47.998824104 +0000
@@ -9,6 +9,7 @@
 use std::env;
 use std::fs;
 use std::path::PathBuf;
+use std::thread::Builder;
 
 use cargo::execute_main_without_stdin;
 use cargo::util::{self, CliResult, lev_distance, Config, human, CargoResult};
@@ -55,7 +56,14 @@
 
 fn main() {
     env_logger::init().unwrap();
-    execute_main_without_stdin(execute, true, USAGE)
+
+    // stack is still to big for usage with musl. So increase stack size.
+    let stack_size = env::var("CARGO_STACK_SIZE").ok()
+                         .and_then(|s| s.parse().ok())
+                         .unwrap_or(8 * 1024 * 1024); // 8MB
+    Builder::new().stack_size(stack_size).spawn(|| {
+        execute_main_without_stdin(execute, true, USAGE)
+    }).unwrap().join().unwrap();
 }
 
 macro_rules! each_subcommand{
