diff --git a/Package.swift b/Package.swift
index bbe10ab..62e37e1 100644
--- a/Package.swift
+++ b/Package.swift
@@ -1,11 +1,5 @@
 import PackageDescription
 
 let package = Package(
-    name: "SQLite",
-    dependencies: [
-        .Package(url: "https://github.com/vapor/csqlite.git", majorVersion: 1),
-        .Package(url: "https://github.com/vapor/core.git", majorVersion: 1),
-        .Package(url: "https://github.com/vapor/node", majorVersion: 1)
-
-    ]
+    name: "SQLite"
 )
