diff --git a/src/cargoexecutionconfig.cpp b/src/cargoexecutionconfig.cpp
index 347ecf9..42331d6 100644
--- a/src/cargoexecutionconfig.cpp
+++ b/src/cargoexecutionconfig.cpp
@@ -30,6 +30,7 @@
 #include <interfaces/iprojectcontroller.h>
 #include <interfaces/iproject.h>
 #include <interfaces/iuicontroller.h>
+#include <interfaces/iruncontroller.h>
 #include <project/projectmodel.h>
 #include <project/builderjob.h>
 #include <serialization/indexedstring.h>
diff --git a/src/cargoplugin.cpp b/src/cargoplugin.cpp
index b93921f..7f58eca 100644
--- a/src/cargoplugin.cpp
+++ b/src/cargoplugin.cpp
@@ -217,7 +217,7 @@ QUrl CargoPlugin::workingDirectory(KDevelop::ILaunchConfiguration* config) const
     return config->project()->path().toUrl();
 }
 
-QString CargoPlugin::environmentGroup(KDevelop::ILaunchConfiguration* /*config*/) const
+QString CargoPlugin::environmentProfileName(KDevelop::ILaunchConfiguration* /*config*/) const
 {
     return QString();
 }
diff --git a/src/cargoplugin.h b/src/cargoplugin.h
index be42673..dbc347c 100644
--- a/src/cargoplugin.h
+++ b/src/cargoplugin.h
@@ -97,7 +97,7 @@ public:
     QUrl executable(KDevelop::ILaunchConfiguration* config, QString& error) const override;
     QStringList arguments(KDevelop::ILaunchConfiguration* config, QString& error) const override;
     KJob* dependencyJob(KDevelop::ILaunchConfiguration* config) const override;
-    QString environmentGroup(KDevelop::ILaunchConfiguration* config) const override;
+    QString environmentProfileName(KDevelop::ILaunchConfiguration* config) const override;
     QString nativeAppConfigTypeId() const override;
     QString terminal(KDevelop::ILaunchConfiguration* config) const override;
     bool useTerminal(KDevelop::ILaunchConfiguration* config) const override;
