--- src/UnliftIO/STM.hs.bak	2018-09-22 17:32:04.564401533 +0200
+++ src/UnliftIO/STM.hs	2018-09-22 17:34:37.051109941 +0200
@@ -81,6 +81,7 @@
 
 import Control.Concurrent.STM (STM, TVar, TMVar, TChan, TQueue, TBQueue)
 import qualified Control.Concurrent.STM as STM
+import Numeric.Natural (Natural)
 import Control.Monad.IO.Unlift
 import System.Mem.Weak (Weak)
 
@@ -165,5 +166,6 @@
 -- | Lifted version of 'STM.newTBQueueIO'
 --
 -- @since 0.2.1.0
-newTBQueueIO :: MonadIO m => Int -> m (TBQueue a)
+newTBQueueIO :: MonadIO m => Natural -> m (TBQueue a)
 newTBQueueIO = liftIO . STM.newTBQueueIO
+
