--- a/configure	2020-03-10 18:30:04.953909904 +0100
+++ b/configure	2020-03-10 18:30:15.300976912 +0100
@@ -26057,7 +26057,7 @@
   echo $ECHO_N "(cached) $ECHO_C" >&6
 else
 
-if test "$cross_compiling" != yes ; then
+if false ; then
 
 # GNU binutils' ld does not involve permissions which may stop ldconfig.
 cf_pathlist=`ld --verbose 2>/dev/null | grep SEARCH_DIR | sed -e 's,SEARCH_DIR[("=][("=]*,,g' -e 's/"[)];//gp' | sort -u`
