--- a/configure	2016-10-05 22:37:56.720212433 +0200
+++ b/configure	2016-10-05 22:39:06.523215869 +0200
@@ -7482,7 +7482,7 @@
       # but apparently some variants of GNU ld only accept -v.
       # Break only if it was the GNU/non-GNU ld that we prefer.
       case `"$lt_cv_path_LD" -v 2>&1 </dev/null` in
-      *GNU* | *'with BFD'*)
+      *GNU* | *'with BFD'* | *'LLD'*)
 	test no != "$with_gnu_ld" && break
 	;;
       *)
@@ -7513,7 +7513,7 @@
 else
   # I'd rather use --version here, but apparently some GNU lds only accept -v.
 case `$LD -v 2>&1 </dev/null` in
-*GNU* | *'with BFD'*)
+*GNU* | *'with BFD'* | *'LLD'*)
   lt_cv_prog_gnu_ld=yes
   ;;
 *)
@@ -12155,7 +12155,7 @@
 	  diet\ *) tmp_diet=yes;;	# linux-dietlibc with static linking (!diet-dyn)
 	esac
       fi
-      if $LD --help 2>&1 | $EGREP ': supported targets:.* elf' > /dev/null \
+      if ($LD --help 2>&1 | $EGREP ': supported targets:.* elf' > /dev/null || $LD --help 2>&1 | $EGREP 'OVERVIEW: lld' > /dev/null) \
 	 && test no = "$tmp_diet"
       then
 	tmp_addflag=' $pic_flag'
