--- a/daemons/lvmpolld/lvmpolld-core.c	2022-10-30 12:53:18.277870834 +0100
+++ b/daemons/lvmpolld/lvmpolld-core.c	2022-10-30 12:53:31.046050905 +0100
@@ -52,14 +52,7 @@
 
 static const char *_strerror_r(int errnum, struct lvmpolld_thread_data *data)
 {
-#ifdef _GNU_SOURCE
-	return strerror_r(errnum, data->buf, sizeof(data->buf)); /* never returns NULL */
-#elif (_POSIX_C_SOURCE >= 200112L || _XOPEN_SOURCE >= 600)
 	return strerror_r(errnum, data->buf, sizeof(data->buf)) ? "" : data->buf;
-#else
-#	warning "Can't decide proper strerror_r implementation. lvmpolld will not issue specific system error messages"
-	return "";
-#endif
 }
 
 static void _usage(const char *prog, FILE *file)
