--- a/Distribution/Gentoo/GHC.hs	2020-02-21 12:18:31.862214140 +0100
+++ b/Distribution/Gentoo/GHC.hs	2020-02-21 12:19:44.011559792 +0100
@@ -67,7 +67,9 @@
 #else
                                                         False   -- is output in binary mode
 #endif /* MIN_VERSION_Cabal(2,1,0) */
-#if MIN_VERSION_Cabal(2,1,0)
+#if MIN_VERSION_Cabal(3,2,0)
+                         return (BL8.unpack out)
+#elif MIN_VERSION_Cabal(2,1,0)
                          case out of
                              ~(DSU.IODataBinary bs) -> return (BL8.unpack bs)
 #else
