--- a/plugins/cpu/loadaverages.cpp	2021-10-20 23:52:08.092762533 +0200
+++ b/plugins/cpu/loadaverages.cpp	2021-10-20 23:52:19.108900409 +0200
@@ -6,6 +6,7 @@
 
 #include "loadaverages.h"
 
+#include <array>
 #include <KLocalizedString>
 
 LoadAverages::LoadAverages(KSysGuard::SensorContainer *parent)
--- a/plugins/lmsensors/lmsensors.cpp	2021-10-20 23:51:21.827183386 +0200
+++ b/plugins/lmsensors/lmsensors.cpp	2021-10-20 23:51:35.865359127 +0200
@@ -6,6 +6,7 @@
 
 #include "lmsensors.h"
 
+#include <array>
 #include <systemstats/SensorContainer.h>
 #include <systemstats/SensorObject.h>
 #include <systemstats/SensorsFeatureSensor.h>
