diff --git a/Package.resolved b/Package.resolved
deleted file mode 100644
index 09a8836..0000000
--- a/Package.resolved
+++ /dev/null
@@ -1,16 +0,0 @@
-{
-  "object": {
-    "pins": [
-      {
-        "package": "Operadics",
-        "repositoryURL": "https://github.com/typelift/Operadics.git",
-        "state": {
-          "branch": "swift-develop",
-          "revision": "743e71210ea319dfd9f0e39268c5e18877debc7f",
-          "version": null
-        }
-      }
-    ]
-  },
-  "version": 1
-}
diff --git a/Package.swift b/Package.swift
index 39d3dbc..8874f68 100644
--- a/Package.swift
+++ b/Package.swift
@@ -9,13 +9,9 @@ let package = Package(
             name: "SwiftCheck",
             targets: ["SwiftCheck"]),
     ],
-    dependencies: [
-		.package(url: "https://github.com/typelift/Operadics.git", .branch("swift-develop"))
-	],
 	targets: [
 		.target(
-            name: "SwiftCheck",
-            dependencies: ["Operadics"]),
+            name: "SwiftCheck"),
         .testTarget(
             name: "SwiftCheckTests",
             dependencies: ["SwiftCheck"]),
