--- source/Core/ArchSpec.cpp	2016-11-17 02:37:42.000000000 +0100
+++ source/Core/ArchSpec.cpp.new	2017-06-11 22:56:13.399037931 +0200
@@ -1125,8 +1125,12 @@
       (rhs == llvm::Triple::Android && lhs == llvm::Triple::EABI) ||
       (lhs == llvm::Triple::GNUEABI && rhs == llvm::Triple::EABI) ||
       (rhs == llvm::Triple::GNUEABI && lhs == llvm::Triple::EABI) ||
+      (lhs == llvm::Triple::MUSLEABI && rhs == llvm::Triple::EABI) ||
+      (rhs == llvm::Triple::MUSLEABI && lhs == llvm::Triple::EABI) ||
       (lhs == llvm::Triple::GNUEABIHF && rhs == llvm::Triple::EABIHF) ||
-      (rhs == llvm::Triple::GNUEABIHF && lhs == llvm::Triple::EABIHF))
+      (rhs == llvm::Triple::GNUEABIHF && lhs == llvm::Triple::EABIHF) ||
+      (lhs == llvm::Triple::MUSLEABIHF && rhs == llvm::Triple::EABIHF) ||
+      (rhs == llvm::Triple::MUSLEABIHF && lhs == llvm::Triple::EABIHF))
     return true;
 
   return false;
