--- a/source/Core/ArchSpec.cpp
+++ b/source/Core/ArchSpec.cpp
@@ -1125,8 +1125,12 @@
       (rhs == llvm::Triple::Android && lhs == llvm::Triple::EABI) ||
       (lhs == llvm::Triple::GNUEABI && rhs == llvm::Triple::EABI) ||
       (rhs == llvm::Triple::GNUEABI && lhs == llvm::Triple::EABI) ||
+      (lhs == llvm::Triple::MuslEABI && rhs == llvm::Triple::EABI) ||
+      (rhs == llvm::Triple::MuslEABI && lhs == llvm::Triple::EABI) ||
       (lhs == llvm::Triple::GNUEABIHF && rhs == llvm::Triple::EABIHF) ||
-      (rhs == llvm::Triple::GNUEABIHF && lhs == llvm::Triple::EABIHF))
+      (rhs == llvm::Triple::GNUEABIHF && lhs == llvm::Triple::EABIHF) ||
+      (lhs == llvm::Triple::MuslEABIHF && rhs == llvm::Triple::EABIHF) ||
+      (rhs == llvm::Triple::MuslEABIHF && lhs == llvm::Triple::EABIHF))
     return true;
 
   return false;
