--- a/source/Symbol/ClangASTContext.cpp	2017-11-26 19:03:08.201590189 +0100
+++ b/source/Symbol/ClangASTContext.cpp	2017-11-26 19:03:56.145494356 +0100
@@ -121,6 +121,7 @@
          // Use Clang for Rust until there is a proper language plugin for it
          language == eLanguageTypeRust ||
          language == eLanguageTypeExtRenderScript ||
+         language == eLanguageTypeCrystal ||
          // Use Clang for D until there is a proper language plugin for it
          language == eLanguageTypeD;
 }
--- a/source/Target/Language.cpp	2017-11-26 19:01:34.593777108 +0100
+++ b/source/Target/Language.cpp	2017-11-26 19:02:57.191612115 +0100
@@ -169,6 +169,7 @@
     {"mipsassem", eLanguageTypeMipsAssembler},
     {"renderscript", eLanguageTypeExtRenderScript},
     // Now synonyms, in arbitrary order
+    {"crystal", eLanguageTypeCrystal},
     {"objc", eLanguageTypeObjC},
     {"objc++", eLanguageTypeObjC_plus_plus},
     {"pascal", eLanguageTypePascal83}};
@@ -293,6 +294,7 @@
   case eLanguageTypeDylan:
   case eLanguageTypeMipsAssembler:
   case eLanguageTypeExtRenderScript:
+  case eLanguageTypeCrystal:
   case eLanguageTypeUnknown:
   default:
     return language;
--- a/include/lldb/lldb-enumerations.h	2017-11-26 18:58:33.521139146 +0100
+++ b/include/lldb/lldb-enumerations.h	2017-11-26 19:01:23.647798900 +0100
@@ -448,6 +448,7 @@
   // assume these can be used as indexes into array g_languages.
   eLanguageTypeMipsAssembler = 0x0024,   ///< Mips_Assembler.
   eLanguageTypeExtRenderScript = 0x0025, ///< RenderScript.
+  eLanguageTypeCrystal = 0x8002, ///< Crystal
   eNumLanguageTypes
 };
 
