diff --git a/source/Symbol/SwiftASTContext.cpp b/source/Symbol/SwiftASTContext.cpp
index 69b0cd84f..65e4a11d1 100644
--- a/source/Symbol/SwiftASTContext.cpp
+++ b/source/Symbol/SwiftASTContext.cpp
@@ -60,12 +60,12 @@
 #include "llvm/Target/TargetOptions.h"
 #include "llvm/Target/TargetSubtargetInfo.h"
 
-#include "swift/../../lib/IRGen/FixedTypeInfo.h"
-#include "swift/../../lib/IRGen/GenEnum.h"
-#include "swift/../../lib/IRGen/GenHeap.h"
-#include "swift/../../lib/IRGen/IRGenModule.h"
-#include "swift/../../lib/IRGen/TypeInfo.h"
-#include "swift/../../lib/IRGen/IRGenMangler.h"
+#include "swift/IRGen/FixedTypeInfo.h"
+#include "swift/IRGen/GenEnum.h"
+#include "swift/IRGen/GenHeap.h"
+#include "swift/IRGen/IRGenModule.h"
+#include "swift/IRGen/TypeInfo.h"
+#include "swift/IRGen/IRGenMangler.h"
 
 #include "swift/Serialization/SerializedModuleLoader.h"
 #include "swift/Strings.h"
