diff --git a/source/Symbol/SwiftASTContext.cpp b/source/Symbol/SwiftASTContext.cpp
index 69b0cd84f..65e4a11d1 100644
--- a/source/Symbol/SwiftASTContext.cpp
+++ b/source/Symbol/SwiftASTContext.cpp
@@ -2389,12 +2389,8 @@ static ConstString GetResourceDir() {
         if (log)
           log->Printf("%s: trying ePathTypeSwiftDir: %s", __FUNCTION__,
                       swift_dir_spec.GetCString());
-        // We can't just check for the Swift directory, because that
-        // always exists.  We have to look for "clang" inside that.
-        FileSpec swift_clang_dir_spec = swift_dir_spec;
-        swift_clang_dir_spec.AppendPathComponent("clang");
 
-        if (swift_clang_dir_spec.IsDirectory()) {
+        if (swift_dir_spec.IsDirectory()) {
           g_cached_resource_dir = ConstString(swift_dir_spec.GetPath());
           if (log)
             log->Printf("%s: found Swift resource dir via "
