diff --git a/tools/repl/swift/CMakeLists.txt b/tools/repl/swift/CMakeLists.txt
index 02c027c5f..549020a63 100644
--- a/tools/repl/swift/CMakeLists.txt
+++ b/tools/repl/swift/CMakeLists.txt
@@ -14,12 +14,13 @@ if(NOT LLDB_SWIFTC)
   if(NOT LLDB_BUILT_STANDALONE)
     set(LLDB_SWIFTC $<TARGET_FILE:swift>c)
   else()
-    set(LLDB_SWIFTC ${LLDB_PATH_TO_SWIFT_BUILD}/bin/swiftc)
+    set(LLDB_SWIFTC /usr/bin/swiftc)
   endif()
 endif()
 
 set(swift_compile_cmd "${LLDB_SWIFTC}"
-                      -resource-dir "${LLDB_PATH_TO_SWIFT_BUILD}/lib/swift"
+                      -Xlinker -nostrip
+                      -resource-dir "/usr/lib/swift"
                       -module-cache-path "${module_cache_dir}")
 
 if(CMAKE_CROSSCOMPILING)
