diff --git a/source/Symbol/SwiftASTContext.cpp b/source/Symbol/SwiftASTContext.cpp
index e8a41dc8b..91d3f7f53 100644
--- a/source/Symbol/SwiftASTContext.cpp
+++ b/source/Symbol/SwiftASTContext.cpp
@@ -2454,7 +2449,6 @@ static ConstString GetResourceDir() {
         // We can't just check for the Swift directory, because that
         // always exists.  We have to look for "clang" inside that.
         FileSpec swift_clang_dir_spec = swift_dir_spec;
-        swift_clang_dir_spec.AppendPathComponent("clang");
 
         if (swift_clang_dir_spec.IsDirectory()) {
           g_cached_resource_dir = ConstString(swift_dir_spec.GetPath());
