diff --git a/tools/repl/swift/CMakeLists.txt b/tools/repl/swift/CMakeLists.txt
index 89d5096..c1b2ad3 100644
--- a/tools/repl/swift/CMakeLists.txt
+++ b/tools/repl/swift/CMakeLists.txt
@@ -11,7 +11,8 @@ if(NOT LLDB_BUILT_STANDALONE)
 endif()
 
 set(swift_compile_cmd "${LLDB_SWIFTC}"
-                      -resource-dir "${LLDB_PATH_TO_SWIFT_BUILD}/lib/swift"
+                      -Xlinker -nostrip
+                      -resource-dir "/usr/lib/swift"
                       -module-cache-path "${module_cache_dir}")
 
 if(CMAKE_CROSSCOMPILING)
