--- a/configure	2022-01-08 21:59:57.000000000 +0100
+++ b/configure	2022-01-09 16:22:24.983344836 +0100
@@ -4358,7 +4358,7 @@
   return 0;
 }
 _ACEOF
-if ac_fn_c_try_compile "$LINENO"; then :
+if ac_fn_c_try_link "$LINENO"; then :
   sc_cv_getprotobynumber_r=1; tmp_bynum_variant=Linux
 else
   cat confdefs.h - <<_ACEOF >conftest.$ac_ext
@@ -4373,7 +4373,7 @@
   return 0;
 }
 _ACEOF
-if ac_fn_c_try_compile "$LINENO"; then :
+if ac_fn_c_try_link "$LINENO"; then :
   sc_cv_getprotobynumber_r=2; tmp_bynum_variant=Solaris
 else
   cat confdefs.h - <<_ACEOF >conftest.$ac_ext
@@ -4388,7 +4388,7 @@
   return 0;
 }
 _ACEOF
-if ac_fn_c_try_compile "$LINENO"; then :
+if ac_fn_c_try_link "$LINENO"; then :
   sc_cv_getprotobynumber_r=3; tmp_bynum_variant=AIX
 else
   sc_cv_getprotobynumber_r=
