diff --git a/3rdparty/spdlog-0.11.0/include/spdlog/details/os.h b/3rdparty/spdlog-0.11.0/include/spdlog/details/os.h
index ed4f45cd..f85b0a71 100644
--- a/3rdparty/spdlog-0.11.0/include/spdlog/details/os.h
+++ b/3rdparty/spdlog-0.11.0/include/spdlog/details/os.h
@@ -342,8 +342,7 @@ inline std::string errno_str(int err_num)
     else
         return "Unkown error";
 
-#elif defined(__FreeBSD__) || defined(__APPLE__) || defined(ANDROID) || \
-      ((_POSIX_C_SOURCE >= 200112L) && ! _GNU_SOURCE) // posix version
+#elif !defined(__GLIBC__)
 
     if (strerror_r(err_num, buf, buf_size) == 0)
         return std::string(buf);
