--- a/libs/editor/settings/wireguardinterfacewidget.cpp
+++ b/libs/editor/settings/wireguardinterfacewidget.cpp
@@ -26,6 +26,9 @@
 #include "simpleiplistvalidator.h"
 #include "wireguardkeyvalidator.h"
 
+#include <QFile>
+#include <QFileInfo>
+#include <QPointer>
 #include <QStandardItemModel>
 
 #include <NetworkManagerQt/Utils>
--- a/libs/editor/settings/wireguardpeerwidget.cpp
+++ b/libs/editor/settings/wireguardpeerwidget.cpp
@@ -26,6 +26,7 @@
 #include "simpleiplistvalidator.h"
 #include "wireguardkeyvalidator.h"
 
+#include <QUrl>
 #include <QStandardItemModel>
 
 #include <NetworkManagerQt/Utils>
