diff --git a/Sources/Fluent/SQLite/SQLiteDriver.swift b/Sources/Fluent/SQLite/SQLiteDriver.swift
index ece5bf1..e72a14b 100644
--- a/Sources/Fluent/SQLite/SQLiteDriver.swift
+++ b/Sources/Fluent/SQLite/SQLiteDriver.swift
@@ -130,7 +130,7 @@
                 case .bool(let bool):
                     try statement.bind(bool)
                 case .bytes(let data):
-                    try statement.bind(data)
+                    try statement.null()
                 case .date(let date):
                     try statement.bind(date.makeNode(in: nil).string ?? "")
                 }
