diff --git a/Package@swift-4.swift b/Package@swift-4.swift
index 1014e49..60d195e 100644
--- a/Package@swift-4.swift
+++ b/Package@swift-4.swift
@@ -4,8 +4,8 @@ import PackageDescription
 let package = Package(
     name: "Fluent",
     products: [
-        .library(name: "Fluent", targets: ["Fluent"]),
-        .library(name: "FluentTester", targets: ["FluentTester"])
+        .library(name: "Fluent", type: .dynamic, targets: ["Fluent"]),
+        .library(name: "FluentTester", type: .dynamic, targets: ["FluentTester"])
     ],
     targets: [
         .target(name: "Fluent"),
