diff --git a/Package.swift b/Package.swift
index 800bcf4..4ac4415 100644
--- a/Package.swift
+++ b/Package.swift
@@ -5,9 +5,5 @@ let package = Package(
     targets: [
         Target(name: "Fluent"),
         Target(name: "FluentTester", dependencies: ["Fluent"])
-    ],
-    dependencies: [
-        // Data structure for converting between multiple representations
-        .Package(url: "https://github.com/vapor/node.git", majorVersion: 1)
     ]
 )
