--- a/src/utils.h
+++ b/src/utils.h
@@ -51,12 +51,12 @@
 
 
 
-bool util_surrounding_get_safe_delta(uint from, uint to, int32_t *delta);
+bool util_surrounding_get_safe_delta(unsigned int from, unsigned int to, int32_t *delta);
 
 bool util_surrounding_get_anchor_pos_from_selection(
     const std::string &surrounding_text,
     const std::string &selected_text,
-    uint cursor_pos,
-    uint *anchor_pos);
+    unsigned int cursor_pos,
+    unsigned int *anchor_pos);
 
 #endif /* __FCITX_ANTHY_UTILS_H__ */
