--- src/utils.h.bak	2019-11-03 14:51:56.354078938 +0100
+++ src/utils.h	2019-11-03 14:52:20.236843206 +0100
@@ -51,12 +51,12 @@
 
 
 
-bool util_surrounding_get_safe_delta(uint from, uint to, int32_t *delta);
+bool util_surrounding_get_safe_delta(unsigned int from, unsigned int to, int32_t *delta);
 
 bool util_surrounding_get_anchor_pos_from_selection(
     const std::string &surrounding_text,
     const std::string &selected_text,
-    uint cursor_pos,
-    uint *anchor_pos);
+    unsigned int cursor_pos,
+    unsigned int *anchor_pos);
 
 #endif /* __FCITX_ANTHY_UTILS_H__ */
