--- a/src/debugger.cpp	2024-10-08 20:35:58.207085159 +0200
+++ b/src/debugger.cpp	2024-10-08 20:39:06.826568794 +0200
@@ -74,6 +74,18 @@
                         u"gdb"_s,
                         u"set width 200\nset backtrace limit 128\nsource %drkonqi_datadir/python/gdb_preamble/preamble.py\npy print_preamble()"_s),
                     .execInputFile = {}}}));
+
+        result.push_back(std::make_shared<Data>( //
+            Data{.displayName = i18nc("@label the debugger called LLDB", "LLDB"),
+                 .codeName = u"lldb"_s,
+                 .tryExec = u"lldb"_s,
+                 .backendData = BackendData{
+                     .command = u"lldb --core %corefile %execpath"_s,
+                     .supportsCommandWithSymbolResolution = false,
+                     .commandWithSymbolResolution = {},
+                     .backtraceBatchCommands = u"settings set term-width 200\nthread info\nbt all"_s,
+                     .preambleCommands = {},
+                     .execInputFile = {}}}));
     }
 
     return result;
