--- a/data/80-iio-sensor-proxy.rules	2026-03-02 20:19:51.000000000 +0100
+++ b/data/80-iio-sensor-proxy.rules	2026-06-02 21:22:33.377318037 +0200
@@ -21,7 +21,7 @@
 # Set the sensor type for all the types we recognise for SDSP/ADSP with libssc.
 # Some devices expose the sensors via a separate SDSP while others re-use ADSP.
 # These DSPs expose itself as /dev/fastrpc-* in the misc subsystem.
-SUBSYSTEM=="misc", KERNEL=="fastrpc-adsp*", ENV{IIO_SENSOR_PROXY_TYPE}+="ssc-light ssc-compass"
+SUBSYSTEM=="misc", KERNEL=="fastrpc-adsp*", ENV{IIO_SENSOR_PROXY_TYPE}+="ssc-light ssc-compass ssc-accel"
 SUBSYSTEM=="misc", KERNEL=="fastrpc-sdsp*", ENV{IIO_SENSOR_PROXY_TYPE}+="ssc-light ssc-compass"
 
 ENV{IIO_SENSOR_PROXY_TYPE}=="", GOTO="iio_sensor_proxy_end"
