--- a/src/compiler/crystal/compiler.cr	2017-11-26 19:34:48.077777091 +0100
+++ b/src/compiler/crystal/compiler.cr	2017-11-26 19:34:15.477842449 +0100
@@ -309,6 +309,7 @@
         link_flags = @link_flags || ""
         link_flags += " -rdynamic"
         link_flags += " -static" if static?
+        link_flags += " -Wl,-nostrip" if debug == Crystal::Debug::All
 
         %(#{CC} #{object_name} -o '#{output_filename}' #{link_flags} #{program.lib_flags})
       end
