diff --git a/exe/Main.hs b/exe/Main.hs
index 9d0adb6..a075bd1 100644
--- a/exe/Main.hs
+++ b/exe/Main.hs
@@ -68,7 +68,6 @@ import Ide.Plugin.Example                 as Example
 import Ide.Plugin.Example2                as Example2
 import Ide.Plugin.GhcIde                  as GhcIde
 import Ide.Plugin.Floskell                as Floskell
-import Ide.Plugin.Fourmolu                as Fourmolu
 import Ide.Plugin.ImportLens              as ImportLens
 import Ide.Plugin.Ormolu                  as Ormolu
 import Ide.Plugin.StylishHaskell          as StylishHaskell
@@ -105,7 +104,6 @@ idePlugins includeExamples = pluginDescToIdePlugins allPlugins
       GhcIde.descriptor  "ghcide"
       , Pragmas.descriptor  "pragmas"
       , Floskell.descriptor "floskell"
-      , Fourmolu.descriptor "fourmolu"
       -- , genericDescriptor     "generic"
       -- , ghcmodDescriptor      "ghcmod"
       , Ormolu.descriptor   "ormolu"
diff --git a/haskell-language-server.cabal b/haskell-language-server.cabal
index 509e4d4..3facb5c 100644
--- a/haskell-language-server.cabal
+++ b/haskell-language-server.cabal
@@ -45,7 +45,6 @@ library
       Ide.Plugin.Eval
       Ide.Plugin.Example
       Ide.Plugin.Example2
-      Ide.Plugin.Fourmolu
       Ide.Plugin.GhcIde
       Ide.Plugin.ImportLens
       Ide.Plugin.Ormolu
@@ -74,7 +73,6 @@ library
     , extra
     , filepath
     , floskell == 0.10.*
-    , fourmolu ^>= 0.1
     , ghc
     , ghc-boot-th
     , ghcide >= 0.1
