diff --git a/exe/Main.hs b/exe/Main.hs
index dfecd73..7f51aa3 100644
--- a/exe/Main.hs
+++ b/exe/Main.hs
@@ -15,7 +15,6 @@ import Ide.Plugin.Example                 as Example
 import Ide.Plugin.Example2                as Example2
 import Ide.Plugin.GhcIde                  as GhcIde
 import Ide.Plugin.Floskell                as Floskell
-import Ide.Plugin.Fourmolu                as Fourmolu
 import Ide.Plugin.ImportLens              as ImportLens
 import Ide.Plugin.Ormolu                  as Ormolu
 import Ide.Plugin.StylishHaskell          as StylishHaskell
@@ -44,7 +43,6 @@ idePlugins includeExamples = pluginDescToIdePlugins allPlugins
       [ GhcIde.descriptor  "ghcide"
       , Pragmas.descriptor  "pragmas"
       , Floskell.descriptor "floskell"
-      , Fourmolu.descriptor "fourmolu"
       , Tactic.descriptor "tactic"
       -- , genericDescriptor     "generic"
       -- , ghcmodDescriptor      "ghcmod"
diff --git a/haskell-language-server.cabal b/haskell-language-server.cabal
index 8ead3a8..6376189 100644
--- a/haskell-language-server.cabal
+++ b/haskell-language-server.cabal
@@ -86,7 +86,6 @@ executable haskell-language-server
     Ide.Plugin.Example
     Ide.Plugin.Example2
     Ide.Plugin.Floskell
-    Ide.Plugin.Fourmolu
     Ide.Plugin.ImportLens
     Ide.Plugin.Ormolu
     Ide.Plugin.Pragmas
@@ -111,7 +110,6 @@ executable haskell-language-server
     , containers
     , deepseq
     , floskell                 ^>=0.10
-    , fourmolu                 ^>=0.1
     , ghc
     , ghc-boot-th
     , ghcide                   >=0.1
