diff --git a/exe/Main.hs b/exe/Main.hs
index a3df59e..f9c1cce 100644
--- a/exe/Main.hs
+++ b/exe/Main.hs
@@ -15,7 +15,6 @@ import           Ide.Plugin.Eval           as Eval
 import           Ide.Plugin.Example        as Example
 import           Ide.Plugin.Example2       as Example2
 import           Ide.Plugin.Floskell       as Floskell
-import           Ide.Plugin.Fourmolu       as Fourmolu
 import           Ide.Plugin.GhcIde         as GhcIde
 import           Ide.Plugin.ImportLens     as ImportLens
 import           Ide.Plugin.Ormolu         as Ormolu
@@ -47,7 +46,6 @@ idePlugins includeExamples = pluginDescToIdePlugins allPlugins
       [ GhcIde.descriptor  "ghcide"
       , Pragmas.descriptor  "pragmas"
       , Floskell.descriptor "floskell"
-      , Fourmolu.descriptor "fourmolu"
       , Tactic.descriptor "tactic"
       -- , genericDescriptor     "generic"
       -- , ghcmodDescriptor      "ghcmod"
diff --git a/haskell-language-server.cabal b/haskell-language-server.cabal
index ea3e80c..ea6b5be 100644
--- a/haskell-language-server.cabal
+++ b/haskell-language-server.cabal
@@ -86,7 +86,6 @@ executable haskell-language-server
     Ide.Plugin.Example
     Ide.Plugin.Example2
     Ide.Plugin.Floskell
-    Ide.Plugin.Fourmolu
     Ide.Plugin.ModuleName
     Ide.Plugin.Ormolu
     Ide.Plugin.Pragmas
@@ -125,7 +124,6 @@ executable haskell-language-server
     , containers
     , deepseq
     , floskell                 ^>=0.10
-    , fourmolu                 ^>=0.3
     , ghc
     , ghc-boot-th
     , ghcide                   >=0.1
