--- a/src/Makefile.am	2016-09-13 19:56:02.796296459 +0200
+++ b/src/Makefile.am	2016-09-13 21:10:49.425528991 +0200
@@ -44,7 +44,7 @@
 	$(X11_LIBS)					\
 	$(UPOWER_LIBS)					\
 	$(GNOME_LIBS)					\
-	-lm
+	-lunwind
 
 gnome_power_statistics_CFLAGS =				\
 	$(WARN_CFLAGS)
@@ -63,7 +63,7 @@
 	$(X11_LIBS)					\
 	$(NOTIFY_LIBS)					\
 	$(GPM_EXTRA_LIBS)				\
-	-lm
+	-lunwind
 
 gnome_power_self_test_CFLAGS =				\
 	$(WARN_CFLAGS)
