--- a/src/adw-fading-label.c	2023-05-06 18:16:49.551025530 +0200
+++ b/src/adw-fading-label.c	2023-05-06 18:17:28.044526009 +0200
@@ -119,9 +119,9 @@
 
   gsk_render_node_get_bounds (node, &bounds);
   bounds.origin.x = 0;
-  bounds.origin.y = floor (bounds.origin.y);
+  bounds.origin.y = ceil (bounds.origin.y);
   bounds.size.width = width;
-  bounds.size.height = ceil (bounds.size.height) + 1;
+  bounds.size.height = ceil (bounds.size.height);
 
   gtk_snapshot_push_mask (snapshot, GSK_MASK_MODE_INVERTED_ALPHA);
 
