--- a/configure.ac.bak	2016-02-03 23:26:12.971051444 +0100
+++ b/configure.ac	2016-02-03 23:26:18.993312939 +0100
@@ -185,7 +185,7 @@
 
 AC_SUBST(WARN_CFLAGS)
 
-AC_CHECK_HEADERS([paths.h shadow.h utmpx.h])
+AC_CHECK_HEADERS([paths.h shadow.h])
 
 AC_CHECK_FUNCS([fgetpwent])
 
--- a/configure.bak	2016-02-03 23:28:06.822968541 +0100
+++ b/configure	2016-02-03 23:28:14.514298845 +0100
@@ -14544,7 +14544,7 @@
 
 
 
-for ac_header in paths.h shadow.h utmpx.h
+for ac_header in paths.h shadow.h
 do :
   as_ac_Header=`$as_echo "ac_cv_header_$ac_header" | $as_tr_sh`
 ac_fn_c_check_header_mongrel "$LINENO" "$ac_header" "$as_ac_Header" "$ac_includes_default"
