diff --color -ur a/src/kwaylandextras.cpp b/src/kwaylandextras.cpp
--- a/src/kwaylandextras.cpp	2026-03-07 21:33:25.000000000 +0100
+++ b/src/kwaylandextras.cpp	2026-03-16 19:13:13.499483052 +0100
@@ -29,7 +29,7 @@
 #if KWINDOWSYSTEM_BUILD_DEPRECATED_SINCE(6, 19)
 void KWaylandExtras::requestXdgActivationToken(QWindow *window, uint32_t serial, const QString &app_id)
 {
-    auto dv2 = dynamic_cast<KWindowSystemPrivateV2 *>(KWindowSystem::d_func());
+    auto dv2 = qobject_cast<KWindowSystemPrivateV2 *>(KWindowSystem::d_func());
     if (!dv2) {
         // Ensure that xdgActivationTokenArrived is always emitted asynchronously
         QTimer::singleShot(0, [serial] {
@@ -44,7 +44,7 @@
 
 quint32 KWaylandExtras::lastInputSerial(QWindow *window)
 {
-    auto dv2 = dynamic_cast<KWindowSystemPrivateV2 *>(KWindowSystem::d_func());
+    auto dv2 = qobject_cast<KWindowSystemPrivateV2 *>(KWindowSystem::d_func());
     if (!dv2) {
         return 0;
     }
@@ -53,21 +53,21 @@
 
 void KWaylandExtras::exportWindow(QWindow *window)
 {
-    if (auto dv2 = dynamic_cast<KWindowSystemPrivateV2 *>(KWindowSystem::d_func())) {
+    if (auto dv2 = qobject_cast<KWindowSystemPrivateV2 *>(KWindowSystem::d_func())) {
         dv2->exportWindow(window);
     }
 }
 
 void KWaylandExtras::unexportWindow(QWindow *window)
 {
-    if (auto dv2 = dynamic_cast<KWindowSystemPrivateV2 *>(KWindowSystem::d_func())) {
+    if (auto dv2 = qobject_cast<KWindowSystemPrivateV2 *>(KWindowSystem::d_func())) {
         dv2->unexportWindow(window);
     }
 }
 
 QFuture<QString> KWaylandExtras::xdgActivationToken(QWindow *window, uint32_t serial, const QString &appId)
 {
-    if (auto dv3 = dynamic_cast<KWindowSystemPrivateV3 *>(KWindowSystem::d_func())) {
+    if (auto dv3 = qobject_cast<KWindowSystemPrivateV3 *>(KWindowSystem::d_func())) {
         return dv3->xdgActivationToken(window, serial, appId);
     } else {
 #if KWINDOWSYSTEM_BUILD_DEPRECATED_SINCE(6, 19)
@@ -99,14 +99,14 @@
 
 void KWaylandExtras::setXdgToplevelTag(QWindow *window, const QString &tag)
 {
-    if (auto dv4 = dynamic_cast<KWindowSystemPrivateV4 *>(KWindowSystem::d_func())) {
+    if (auto dv4 = qobject_cast<KWindowSystemPrivateV4 *>(KWindowSystem::d_func())) {
         dv4->setXdgToplevelTag(window, tag);
     }
 }
 
 void KWaylandExtras::setXdgToplevelDescription(QWindow *window, const QString &description)
 {
-    if (auto dv4 = dynamic_cast<KWindowSystemPrivateV4 *>(KWindowSystem::d_func())) {
+    if (auto dv4 = qobject_cast<KWindowSystemPrivateV4 *>(KWindowSystem::d_func())) {
         dv4->setXdgToplevelDescription(window, description);
     }
 }
diff --color -ur a/src/kwindowsystem.cpp b/src/kwindowsystem.cpp
--- a/src/kwindowsystem.cpp	2026-03-07 21:33:25.000000000 +0100
+++ b/src/kwindowsystem.cpp	2026-03-16 19:13:13.499483052 +0100
@@ -99,7 +99,7 @@
 {
     Q_D(KWindowSystem);
     if (isPlatformWayland()) {
-        if (auto dv2 = dynamic_cast<KWindowSystemPrivateV2 *>(d)) {
+        if (auto dv2 = qobject_cast<KWindowSystemPrivateV2 *>(d)) {
             dv2->setMainWindow(subWindow, mainWindowId);
         }
     } else {
@@ -183,7 +183,7 @@
 void KWindowSystem::setCurrentXdgActivationToken(const QString &token)
 {
     Q_D(KWindowSystem);
-    auto dv2 = dynamic_cast<KWindowSystemPrivateV2 *>(d);
+    auto dv2 = qobject_cast<KWindowSystemPrivateV2 *>(d);
     if (!dv2) {
         return;
     }
diff --color -ur a/src/kwindowsystem_p.h b/src/kwindowsystem_p.h
--- a/src/kwindowsystem_p.h	2026-03-07 21:33:25.000000000 +0100
+++ b/src/kwindowsystem_p.h	2026-03-16 19:13:13.499483052 +0100
@@ -15,8 +15,9 @@
 
 class NETWinInfo;
 
-class KWINDOWSYSTEM_EXPORT KWindowSystemPrivate : public NET
+class KWINDOWSYSTEM_EXPORT KWindowSystemPrivate : public QObject, public NET
 {
+    Q_OBJECT
 public:
     virtual ~KWindowSystemPrivate();
     virtual void activateWindow(QWindow *win, long time = 0) = 0;
@@ -26,6 +27,7 @@
 
 class KWINDOWSYSTEM_EXPORT KWindowSystemPrivateV2 : public KWindowSystemPrivate
 {
+    Q_OBJECT
 public:
 #if KWINDOWSYSTEM_BUILD_DEPRECATED_SINCE(6, 19)
     virtual void requestToken(QWindow *win, uint32_t serial, const QString &app_id) = 0;
@@ -39,12 +41,14 @@
 
 class KWINDOWSYSTEM_EXPORT KWindowSystemPrivateV3 : public KWindowSystemPrivateV2
 {
+    Q_OBJECT
 public:
     virtual QFuture<QString> xdgActivationToken(QWindow *window, uint32_t serial, const QString &appId) = 0;
 };
 
 class KWINDOWSYSTEM_EXPORT KWindowSystemPrivateV4 : public KWindowSystemPrivateV3
 {
+    Q_OBJECT
 public:
     virtual void setXdgToplevelTag(QWindow *window, const QString &tag) = 0;
     virtual void setXdgToplevelDescription(QWindow *window, const QString &description) = 0;
diff --color -ur a/src/platforms/wayland/windowsystem.h b/src/platforms/wayland/windowsystem.h
--- a/src/platforms/wayland/windowsystem.h	2026-03-07 21:33:25.000000000 +0100
+++ b/src/platforms/wayland/windowsystem.h	2026-03-16 19:13:13.503483072 +0100
@@ -12,9 +12,8 @@
 
 class WindowManagement;
 
-class WindowSystem : public QObject, public KWindowSystemPrivateV4
+class WindowSystem : public KWindowSystemPrivateV4
 {
-    Q_OBJECT
 public:
     WindowSystem();
     ~WindowSystem() override;
