--- a/include/libintl.h
+++ b/include/libintl.h
@@ -30,4 +30,11 @@ char *bind_textdomain_codeset(const char *, const char *);
 }
 #endif
 
+#include <stdio.h>
+#define gettext_printf(args...) printf(args)
+
+/* to supply LC_MESSAGES and other stuff GNU expects to be exported when
+   including libintl.h */
+#include <locale.h>
+
 #endif
--- a/src/locale/libintl.c
+++ b/src/locale/libintl.c
@@ -0,0 +1,5 @@
+#include <stdio.h>
+
+/* trick configure tests checking for gnu libintl, as in the copy included in gdb */
+const char *_nl_expand_alias () { return NULL; }
+int _nl_msg_cat_cntr = 0;
