--- a/include/netinet/if_ether.h	2020-02-06 16:27:16.591119887 +0100
+++ b/include/netinet/if_ether.h	2020-02-06 16:27:49.288291028 +0100
@@ -109,11 +109,13 @@
 #define ETH_P_XDSA	0x00F8
 #define ETH_P_MAP	0x00F9
 
+#ifndef __UAPI_DEF_ETHHDR
 struct ethhdr {
 	uint8_t h_dest[ETH_ALEN];
 	uint8_t h_source[ETH_ALEN];
 	uint16_t h_proto;
 };
+#endif
 
 #include <net/ethernet.h>
 #include <net/if_arp.h>
