--- a/Make.defaults	2018-07-19 19:17:25.224964057 +0200
+++ b/Make.defaults	2018-07-19 19:19:49.745916035 +0200
@@ -110,10 +110,10 @@
                             || ( [ $(GCCVERSION) -eq "4" ]      \
                                  && [ $(GCCMINOR) -ge "7" ] ) ) \
                           && echo 1)
-  ifeq ($(GCCNEWENOUGH),1)
-    CPPFLAGS += -DGNU_EFI_USE_MS_ABI -maccumulate-outgoing-args --std=c11
-  else ifeq ($(USING_CLANG),clang)
+  ifeq ($(USING_CLANG),clang)
     CPPFLAGS += -DGNU_EFI_USE_MS_ABI --std=c11
+  else ifeq ($(GCCNEWENOUGH),1)
+    CPPFLAGS += -DGNU_EFI_USE_MS_ABI -maccumulate-outgoing-args --std=c11
   endif
 
   CFLAGS += -mno-red-zone
