--- a/Make.defaults	2016-09-13 13:32:40.965750940 +0200
+++ b/Make.defaults	2016-09-13 13:33:42.765244288 +0200
@@ -108,7 +108,7 @@
                                  && [ $(GCCMINOR) -ge "7" ] ) ) \
                           && echo 1)
   ifeq ($(GCCNEWENOUGH),1)
-    CPPFLAGS += -DGNU_EFI_USE_MS_ABI -maccumulate-outgoing-args --std=c11
+    CPPFLAGS += -DGNU_EFI_USE_MS_ABI --std=c11
   endif
 
   CFLAGS += -mno-red-zone -mno-mmx -mno-sse
@@ -144,9 +144,9 @@
            -fshort-wchar -fno-strict-aliasing \
            -ffreestanding -fno-stack-protector
 else
-CFLAGS  += $(ARCH3264) -g -O2 -fpic -Wall -Wextra -Werror \
+CFLAGS  += $(ARCH3264) -fPIC -Wall -Wextra -Werror \
            -fshort-wchar -fno-strict-aliasing \
-           -fno-merge-constants -ffreestanding -fno-stack-protector \
+           -ffreestanding -fno-stack-protector \
            -fno-stack-check
 endif
 
