--- a/Make.defaults	2017-10-18 16:22:43.135915143 +0200
+++ b/Make.defaults	2017-10-18 16:22:10.094951356 +0200
@@ -109,7 +109,7 @@
                                  && [ $(GCCMINOR) -ge "7" ] ) ) \
                           && echo 1)
   ifeq ($(GCCNEWENOUGH),1)
-    CPPFLAGS += -DGNU_EFI_USE_MS_ABI -maccumulate-outgoing-args --std=c11
+    CPPFLAGS += -DGNU_EFI_USE_MS_ABI --std=c11
   else ifeq ($(USING_CLANG),clang)
     CPPFLAGS += -DGNU_EFI_USE_MS_ABI --std=c11
   endif
@@ -166,7 +166,6 @@
            -fno-stack-check
 endif
 
-ARFLAGS += -U
 ASFLAGS += $(ARCH3264)
 LDFLAGS	+= -nostdlib --warn-common --no-undefined --fatal-warnings \
 	   --build-id=sha1
--- a/lib/Makefile	2017-06-20 16:12:22.000000000 +0200
+++ b/lib/Makefile	2017-10-18 16:22:18.022942395 +0200
@@ -75,7 +75,7 @@
 	for sdir in $(SUBDIRS); do mkdir -p $$sdir; done
 
 libefi.a: $(OBJS)
-	$(AR) rv -U $@ $^
+	$(AR) rv $@ $^
 
 clean:
 	rm -f libefi.a *~ $(OBJS) */*.o
