--- a/shared/n-dhcp4/src/n-dhcp4-c-probe.c
+++ b/shared/n-dhcp4/src/n-dhcp4-c-probe.c
@@ -19,6 +19,43 @@
 #include "n-dhcp4.h"
 #include "n-dhcp4-private.h"
 
+#if defined(__linux__) && !defined(__GLIBC__)
+static uint64_t __rand48_step(unsigned short *xi, unsigned short *lc)
+{
+        uint64_t a, x;
+        x = xi[0] | xi[1]+0U<<16 | xi[2]+0ULL<<32;
+        a = lc[0] | lc[1]+0U<<16 | lc[2]+0ULL<<32;
+        x = a*x + lc[3];
+        xi[0] = x;
+        xi[1] = x>>16;
+        xi[2] = x>>32;
+        return x & 0xffffffffffffull;
+}
+
+
+static int jrand48_r(unsigned short s[3], struct drand48_data *buffer, long *result)
+{
+        *result = (int32_t)(__rand48_step(s, buffer->__x+3) >> 16);
+        return 0;
+}
+
+static int mrand48_r(struct drand48_data *buffer,long *result)
+{
+        *result = jrand48_r(buffer->__x,buffer,result);
+        return 0;
+}
+
+
+static int seed48_r(unsigned short *s, struct drand48_data *buffer)
+{
+        memcpy(buffer->__old_x, buffer->__x, sizeof buffer->__x);
+        memcpy(buffer->__x, s, sizeof buffer->__x);
+        buffer->__c = 0;
+        buffer->__init = 1;
+        buffer->__a = 0;
+        return 0;
+}
+#endif
 
 static int n_dhcp4_client_probe_option_new(NDhcp4ClientProbeOption **optionp,
                                     uint8_t option,
--- a/shared/n-dhcp4/src/n-dhcp4-private.h
+++ b/shared/n-dhcp4/src/n-dhcp4-private.h
@@ -13,6 +13,16 @@
 #include <unistd.h>
 #include "n-dhcp4.h"
 
+#if defined(__linux__) && !defined(__GLIBC__)
+struct drand48_data {
+  unsigned short __x[3];
+  unsigned short __old_x[3];
+  unsigned short __c;
+  unsigned short __init;
+  long long __a;
+};
+#endif
+
 typedef struct NDhcp4CConnection NDhcp4CConnection;
 typedef struct NDhcp4CEventNode NDhcp4CEventNode;
 typedef struct NDhcp4ClientProbeOption NDhcp4ClientProbeOption;
