--- a/src/systemd/src/libsystemd-network/dhcp6-internal.h
+++ b/src/systemd/src/libsystemd-network/dhcp6-internal.h
@@ -5,7 +5,7 @@
   Copyright © 2014-2015 Intel Corporation. All rights reserved.
 ***/
 
-#include <net/ethernet.h>
+//#include <net/ethernet.h>
 #include <netinet/in.h>
 
 #include "sd-event.h"
--- a/src/systemd/src/systemd/sd-dhcp-lease.h
+++ b/src/systemd/src/systemd/sd-dhcp-lease.h
@@ -19,7 +19,7 @@
 ***/
 
 #include <inttypes.h>
-#include <net/ethernet.h>
+//#include <net/ethernet.h>
 #include <netinet/in.h>
 #include <sys/types.h>
 
--- a/src/systemd/src/systemd/sd-dhcp6-client.h
+++ b/src/systemd/src/systemd/sd-dhcp6-client.h
@@ -20,7 +20,7 @@
 ***/
 
 #include <inttypes.h>
-#include <net/ethernet.h>
+//#include <net/ethernet.h>
 #include <sys/types.h>
 
 #include "sd-dhcp6-lease.h"
