--- a/files/71-fastrpc.rules	2026-03-09 21:44:58.416377310 +0100
+++ b/files/71-fastrpc.rules	2026-03-09 21:47:37.540972945 +0100
@@ -6,6 +6,8 @@
 ACTION=="add", SUBSYSTEM=="misc", KERNEL=="fastrpc-adsp-secure", OWNER="fastrpc", GROUP="fastrpc", MODE="0640", TAG+="systemd", ENV{SYSTEMD_WANTS}+="adsprpcd.service"
 ACTION=="add", SUBSYSTEM=="misc", KERNEL=="fastrpc-adsp", OWNER="fastrpc", GROUP="fastrpc", MODE="0640", TAG+="systemd", ENV{SYSTEMD_WANTS}+="adsprpcd_audiopd.service"
 ACTION=="add", SUBSYSTEM=="misc", KERNEL=="fastrpc-adsp-secure", OWNER="fastrpc", GROUP="fastrpc", MODE="0640", TAG+="systemd", ENV{SYSTEMD_WANTS}+="adsprpcd_audiopd.service"
+ACTION=="add", SUBSYSTEM=="misc", KERNEL=="fastrpc-adsp", OWNER="fastrpc", GROUP="fastrpc", MODE="0640", TAG+="systemd", ENV{SYSTEMD_WANTS}+="adsprpcd_sensorspd.service"
+ACTION=="add", SUBSYSTEM=="misc", KERNEL=="fastrpc-adsp-secure", OWNER="fastrpc", GROUP="fastrpc", MODE="0640", TAG+="systemd", ENV{SYSTEMD_WANTS}+="adsprpcd_sensorspd.service"
 
 # cDSP devices
 ACTION=="add", SUBSYSTEM=="misc", KERNEL=="fastrpc-cdsp", OWNER="fastrpc", GROUP="fastrpc", MODE="0640", TAG+="systemd", ENV{SYSTEMD_WANTS}+="cdsprpcd.service"
--- /dev/null	2026-03-09 20:19:38.421380375 +0100
+++ b/files/adsprpcd_sensorspd.service	2026-03-09 21:46:33.672736013 +0100
@@ -0,0 +1,11 @@
+[Unit]
+Description=sensorspd aDSP RPC daemon
+After=dev-fastrpc-adsp.device dev-fastrpc-adsp-secure.device
+ConditionPathExists=|/dev/fastrpc-adsp
+ConditionPathExists=|/dev/fastrpc-adsp-secure
+
+[Service]
+Type=exec
+ExecStart=/usr/bin/adsprpcd sensorspd
+Restart=on-failure
+RestartSec=5
--- a/files/Makefile.am	2026-03-09 21:50:36.249623633 +0100
+++ b/files/Makefile.am	2026-03-09 21:51:00.781711771 +0100
@@ -7,6 +7,7 @@
 dist_systemdsystemunit_DATA = \
 	adsprpcd.service \
 	adsprpcd_audiopd.service \
+	adsprpcd_sensorspd.service \
 	cdsprpcd.service \
 	cdsp1rpcd.service \
 	gdsp0rpcd.service \
