--- src/Data/Text/Prettyprint/Doc/Render/Util/SimpleDocTree.hs.bak	2019-09-17 10:22:45.358096893 +0200
+++ src/Data/Text/Prettyprint/Doc/Render/Util/SimpleDocTree.hs	2019-09-17 10:23:10.894536312 +0200
@@ -126,8 +126,9 @@
         (a', s') <- p s
         (a'', s'') <- runParser (f a') s'
         pure (a'', s'') )
-
+#if !MONAD_FAIL
     fail _err = empty
+#endif
 
 #if MONAD_FAIL
 instance MonadFail (UniqueParser s) where
