--- a/build/make/Makefile	2016-10-16 13:33:57.881693859 +0200
+++ b/build/make/Makefile	2016-10-16 13:34:15.405692417 +0200
@@ -296,7 +296,7 @@
 	$(if $(quiet),@echo "    [LD] $$@")
 	$(qexec)$$(LD) -shared $$(LDFLAGS) \
             -Wl,--no-undefined -Wl,-soname,$$(SONAME) \
-            -Wl,--version-script,$$(EXPORTS_FILE) -o $$@ \
+            -Wl,--sandbox-version-script,$$(EXPORTS_FILE) -o $$@ \
             $$(filter %.o,$$^) $$(extralibs)
 endef
 
