--- a/backend/src/llvm/llvm_to_gen.cpp	2018-01-14 11:14:22.758133907 +0100
+++ b/backend/src/llvm/llvm_to_gen.cpp	2018-01-14 11:15:16.394005641 +0100
@@ -322,7 +322,11 @@
     DataLayout DL(&mod);
     
     gbeDiagnosticContext dc;
+#if LLVM_VERSION_MAJOR * 10 + LLVM_VERSION_MINOR >= 60
+    mod.getContext().setDiagnosticHandlerCallBack(&gbeDiagnosticHandler,&dc);
+#else
     mod.getContext().setDiagnosticHandler(&gbeDiagnosticHandler,&dc);
+#endif
 
 #if LLVM_VERSION_MAJOR * 10 + LLVM_VERSION_MINOR >= 37
     mod.setDataLayout(DL);
--- a/backend/src/llvm/llvm_unroll.cpp	2018-01-14 11:15:50.373923743 +0100
+++ b/backend/src/llvm/llvm_unroll.cpp	2018-01-14 11:18:24.148555622 +0100
@@ -205,7 +205,9 @@
           if (parentTripCount != 0 && currTripCount * parentTripCount > 32) {
             //Don't change the unrollID if doesn't force unroll.
             //setUnrollID(parentL, false);
-#if LLVM_VERSION_MAJOR * 10 + LLVM_VERSION_MINOR >= 38
+#if LLVM_VERSION_MAJOR * 10 + LLVM_VERSION_MINOR >= 60
+            loopInfo.erase(parentL);
+#elif LLVM_VERSION_MAJOR * 10 + LLVM_VERSION_MINOR >= 38
             loopInfo.markAsRemoved(parentL);
 #else
             LPM.deleteLoopFromQueue(parentL);
