commit f7bac47051c3e593444fe5c6183649a6e11e4a30
Author: Mias van Klei <miasvanklei@gmail.com>
Date:   Tue Oct 4 16:51:27 2016 +0200

    musl fixes

diff --git a/backend/src/backend/gen/gen_mesa_disasm.c b/backend/src/backend/gen/gen_mesa_disasm.c
index 5653275..d228ffe 100644
--- a/backend/src/backend/gen/gen_mesa_disasm.c
+++ b/backend/src/backend/gen/gen_mesa_disasm.c
@@ -1115,7 +1115,7 @@ static uint32_t __conv_half_to_float(uint16_t h)
       // Adjust mantissa so it's normalized (and keep
       // track of exponent adjustment)
       int e = -1;
-      uint m = i.mantissa;
+      unsigned int m = i.mantissa;
       do {
         e++;
         m <<= 1;
diff --git a/utests/utest_helper.cpp b/utests/utest_helper.cpp
index d12dccf..bcfe8a0 100644
--- a/utests/utest_helper.cpp
+++ b/utests/utest_helper.cpp
@@ -978,7 +978,7 @@ uint32_t __half_to_float(uint16_t h, bool* isInf, bool* infSign)
       // Adjust mantissa so it's normalized (and keep
       // track of exponent adjustment)
       int e = -1;
-      uint m = i.mantissa;
+      unsigned int m = i.mantissa;
       do {
         e++;
         m <<= 1;
diff --git a/utests/utest_helper.hpp b/utests/utest_helper.hpp
index 034a411..0e2d5dc 100644
--- a/utests/utest_helper.hpp
+++ b/utests/utest_helper.hpp
@@ -34,6 +34,7 @@
 #include <cassert>
 #include <cstdio>
 #include <cstdlib>
+#include <sys/time.h>
 
 #if defined(__ANDROID__)
 #define __thread
diff --git a/src/cl_api_kernel.c b/src/cl_api_kernel.c
index ef494e6..28240f5 100644
--- a/src/cl_api_kernel.c
+++ b/src/cl_api_kernel.c
@@ -117,7 +117,7 @@ clEnqueueNDRangeKernel(cl_command_queue command_queue,
         fixed_local_sz[0] = 16;
         fixed_local_sz[1] = 1;
       } else {
-        uint j, maxDimSize = 64 /* from 64? */, maxGroupSize = 256; //MAX_WORK_GROUP_SIZE may too large
+        unsigned int j, maxDimSize = 64 /* from 64? */, maxGroupSize = 256; //MAX_WORK_GROUP_SIZE may too large
         size_t realGroupSize = 1;
         for (i = 0; i < work_dim; i++) {
           for (j = maxDimSize; j > 1; j--) {
diff --git a/src/cl_alloc.c b/src/cl_alloc.c
index e532569..2ac5ee4 100644
--- a/src/cl_alloc.c
+++ b/src/cl_alloc.c
@@ -23,6 +23,7 @@
 #include <stdlib.h>
 #include <assert.h>
 #include <malloc.h>
+#include <sys/types.h>
 
 static volatile int32_t cl_alloc_n = 0;
 
diff --git a/src/cl_enqueue.c b/src/cl_enqueue.c
index fbcd7b6..f8d28d7 100644
--- a/src/cl_enqueue.c
+++ b/src/cl_enqueue.c
@@ -493,7 +493,7 @@ error:
 cl_int cl_enqueue_svm_free(enqueue_data *data, cl_int status) {
   int i;
   void **pointers = data->pointers;
-  uint num_svm_ptrs = data->size;
+  unsigned int num_svm_ptrs = data->size;
   cl_int err = CL_SUCCESS;
 
   if (status != CL_COMPLETE)
