--- vendor/brightray/vendor/libchromiumcontent/vendor/chromium/src/components/test_runner/test_runner.gyp.bak	2015-12-01 15:51:54.943261142 +0100
+++ vendor/brightray/vendor/libchromiumcontent/vendor/chromium/src/components/test_runner/test_runner.gyp	2015-12-01 15:52:38.543148202 +0100
@@ -158,8 +158,6 @@
             'files': [
               'resources/fonts/AHEM____.TTF',
               'resources/fonts/fonts.conf',
-              '../../third_party/gardiner_mod/GardinerModBug.ttf',
-              '../../third_party/gardiner_mod/GardinerModCat.ttf',
             ]
           }],
         }],
--- chromium_src/chrome/renderer/spellchecker/spellcheck_worditerator.h.bak	2015-12-01 16:37:07.552353086 +0100
+++ chromium_src/chrome/renderer/spellchecker/spellcheck_worditerator.h	2015-12-01 16:36:55.676384346 +0100
@@ -14,7 +14,7 @@
 #include "base/basictypes.h"
 #include "base/memory/scoped_ptr.h"
 #include "base/strings/string16.h"
-#include "third_party/icu/source/common/unicode/uscript.h"
+#include <unicode/uscript.h>
 
 namespace base {
 namespace i18n {
--- atom/renderer/api/atom_api_spell_check_client.cc.bak	2015-12-01 16:38:16.027172825 +0100
+++ atom/renderer/api/atom_api_spell_check_client.cc	2015-12-01 16:38:41.902104697 +0100
@@ -11,7 +11,7 @@
 #include "base/logging.h"
 #include "native_mate/converter.h"
 #include "native_mate/dictionary.h"
-#include "third_party/icu/source/common/unicode/uscript.h"
+#include <unicode/uscript.h>
 #include "third_party/WebKit/public/web/WebTextCheckingCompletion.h"
 #include "third_party/WebKit/public/web/WebTextCheckingResult.h"
 
--- chromium_src/library_loaders/libspeechd.h.bak	2015-12-01 16:39:40.001951709 +0100
+++ chromium_src/library_loaders/libspeechd.h	2015-12-01 16:40:21.452842562 +0100
@@ -4,7 +4,7 @@
 #ifndef LIBRARY_LOADER_OUT_RELEASE_GEN_LIBRARY_LOADERS_LIBSPEECHD_H
 #define LIBRARY_LOADER_OUT_RELEASE_GEN_LIBRARY_LOADERS_LIBSPEECHD_H
 
-#include "third_party/speech-dispatcher/libspeechd.h"
+#include <speech-dispatcher/libspeechd.h>
 #define LIBRARY_LOADER_OUT_RELEASE_GEN_LIBRARY_LOADERS_LIBSPEECHD_H_DLOPEN
 
 
--- chromium_src/chrome/renderer/spellchecker/spellcheck_worditerator.cc.bak	2015-12-01 16:41:07.052722469 +0100
+++ chromium_src/chrome/renderer/spellchecker/spellcheck_worditerator.cc	2015-12-01 16:41:51.802604621 +0100
@@ -14,10 +14,10 @@
 #include "base/logging.h"
 #include "base/strings/stringprintf.h"
 #include "base/strings/utf_string_conversions.h"
-#include "third_party/icu/source/common/unicode/normlzr.h"
-#include "third_party/icu/source/common/unicode/schriter.h"
-#include "third_party/icu/source/common/unicode/uscript.h"
-#include "third_party/icu/source/i18n/unicode/ulocdata.h"
+#include <unicode/normlzr.h>
+#include <unicode/schriter.h>
+#include <unicode/uscript.h>
+#include <unicode/ulocdata.h>
 
 // SpellcheckCharAttribute implementation:
 
