--- vendor/brightray/vendor/libchromiumcontent/script/update    2015-11-30 20:05:53.730852491 +0100
+++ vendor/brightray/vendor/libchromiumcontent/script/update    2015-11-30 20:06:07.555856360 +0100
@@ -27,17 +27,9 @@
 def main():
   args = parse_args()
 
-  version = chromium_version()
-  if not is_source_tarball_updated(version):
-    download_source_tarball(version)
-
-  if sys.platform == 'linux2':
-    install_sysroot()
-
   target_arch = args.target_arch
   return (apply_patches() or
           copy_chromiumcontent_files() or
-          update_clang() or
           run_gyp(target_arch))
 
 
@@ -119,6 +111,8 @@
   gyp_defines = 'disable_nacl=1 component={0}'.format(component)
 
   env['GYP_GENERATORS'] = 'ninja'
+  env['CC'] = 'clang'
+  env['CXX'] = 'clang++'
 
   if sys.platform in ['win32', 'cygwin']:
     # Don't use the auto-installed VS2013 Express toolchain from depot_tools.
@@ -137,6 +131,43 @@
 
   # Always build on 64-bit machine.
   gyp_defines += ' host_arch=x64'
+  gyp_defines += ' disable_sse2=1'
+  gyp_defines += ' disable_pnacl=1'
+  gyp_defines += ' disable_nacl=1'
+  gyp_defines += ' disable_glibc=1'
+  gyp_defines += ' disable_fatal_linker_warnings=1'
+  gyp_defines += ' use_allocator=none'
+  gyp_defines += ' clang=1'
+  gyp_defines += ' is_clang=true'
+  gyp_defines += ' host_clang=0'
+  gyp_defines += ' clang_use_chrome_plugins=0'
+  gyp_defines += ' werror='
+  gyp_defines += ' fastbuild=2'
+  gyp_defines += ' use_system_bzip2=1'
+  gyp_defines += ' use_system_flac=1'
+  gyp_defines += ' use_system_harfbuzz=1'
+  gyp_defines += ' use_system_icu=1'
+  gyp_defines += ' use_system_jsoncpp=1'
+  gyp_defines += ' use_system_libevent=1'
+  gyp_defines += ' use_system_libjpeg=1'
+  gyp_defines += ' use_system_libpng=1'
+  gyp_defines += ' use_system_libwebp=1'
+  gyp_defines += ' use_system_libxml=1'
+  gyp_defines += ' use_system_libxslt=1'
+  gyp_defines += ' use_system_minizip=1'
+  gyp_defines += ' use_system_nspr=1'
+  gyp_defines += ' use_system_re2=1'
+  gyp_defines += ' use_system_snappy=1'
+  gyp_defines += ' use_system_speex=1'
+  gyp_defines += ' use_system_xdg_utils=1'
+  gyp_defines += ' use_system_zlib=1'
+  gyp_defines += ' icu_use_data_file_flag=0'
+  gyp_defines += ' use_system_yasm=1'
+  gyp_defines += ' logging_like_official_build=1'
+  gyp_defines += ' linux_link_gsettings=1'
+  gyp_defines += ' linux_link_libpci=1'
+  gyp_defines += ' linux_link_libspeechd=1'
+  gyp_defines += ' libspeechd_h_prefix=speech-dispatcher/'

   # Build 32-bit or 64-bit.
   gyp_defines += ' target_arch={0}'.format(target_arch)
--- vendor/brightray/vendor/libchromiumcontent/chromiumcontent/chromiumcontent.gypi   2015-11-30 23:40:14.144818880 +0100
+++ vendor/brightray/vendor/libchromiumcontent/chromiumcontent/chromiumcontent.gypi   2015-11-30 23:40:29.943753461 +0100
@@ -26,7 +26,7 @@
         'use_dbus': 1,
         # Make Linux build contain debug symbols, this flag will add '-g' to
         # cflags.
-        'linux_dump_symbols': 1,
+        'linux_dump_symbols': 0,
         # The Linux build of libchromiumcontent.so depends on, but doesn't
         # provide, tcmalloc by default.  Disabling tcmalloc here also prevents
         # any conflicts when linking to binaries or libraries that don't use
