--- script/bootstrap.py	2015-11-29 17:51:56.740061544 +0100
+++ script/bootstrap.py	2015-11-29 17:48:57.325561651 +0100
@@ -50,10 +50,6 @@
   if sys.platform == 'cygwin':
     update_win32_python()
 
-  if PLATFORM != 'win32':
-    update_clang()
-
-  update_submodules()
   setup_python_libs()
   update_node_modules('.')
   bootstrap_brightray(args.dev, args.url, args.target_arch,
@@ -61,9 +57,6 @@
                       args.libcc_shared_library_path,
                       args.libcc_static_library_path)
 
-  if args.target_arch in ['arm', 'ia32'] and PLATFORM == 'linux':
-    download_sysroot(args.target_arch)
-
   create_chrome_version_h()
   touch_config_gypi()
   run_update()
@@ -144,8 +137,8 @@
     # Use prebuilt clang for building native modules.
     llvm_dir = os.path.join(SOURCE_ROOT, 'vendor', 'llvm-build',
                             'Release+Asserts', 'bin')
-    env['CC']  = os.path.join(llvm_dir, 'clang')
-    env['CXX'] = os.path.join(llvm_dir, 'clang++')
+    env['CC']  = 'clang'
+    env['CXX'] = 'clang++'
     env['npm_config_clang'] = '1'
   with scoped_cwd(dirname):
     args = [NPM, 'install']
