--- common.gypi	2015-11-26 07:12:28.000000000 +0100
+++ common.gypi	2015-11-29 15:43:20.749584401 +0100
@@ -18,11 +18,11 @@
     'node_install_npm': 'false',
     'node_prefix': '',
     'node_shared_cares': 'false',
-    'node_shared_http_parser': 'false',
-    'node_shared_libuv': 'false',
-    'node_shared_openssl': 'false',
+    'node_shared_http_parser': 'true',
+    'node_shared_libuv': 'true',
+    'node_shared_openssl': 'true',
     'node_shared_v8': 'true',
-    'node_shared_zlib': 'false',
+    'node_shared_zlib': 'true',
     'node_tag': '',
     'node_use_dtrace': 'false',
     'node_use_etw': 'false',
@@ -172,6 +172,12 @@
               '-Wl,--whole-archive',
               '<@(libchromiumcontent_v8_libraries)',
               '-Wl,--no-whole-archive',
+	     ],
+	    'libraries': [
+	      '-lssl',
+	      '-luv',
+	      '-lhttp_parser',
+	      '-lz',
             ],
           }, {
             'libraries': [ '<@(libchromiumcontent_v8_libraries)' ],
--- atom.gyp.bak	2015-12-01 15:38:01.666781894 +0100
+++ atom.gyp	2015-12-01 15:38:06.767783033 +0100
@@ -157,7 +157,6 @@
                 '<@(copied_libraries)',
                 '<(libchromiumcontent_dir)/libEGL.dll',
                 '<(libchromiumcontent_dir)/libGLESv2.dll',
-                '<(libchromiumcontent_dir)/icudtl.dat',
                 '<(libchromiumcontent_dir)/content_resources_200_percent.pak',
                 '<(libchromiumcontent_dir)/content_shell.pak',
                 '<(libchromiumcontent_dir)/ui_resources_200_percent.pak',
@@ -203,7 +202,6 @@
               'destination': '<(PRODUCT_DIR)',
               'files': [
                 '<@(copied_libraries)',
-                '<(libchromiumcontent_dir)/icudtl.dat',
                 '<(libchromiumcontent_dir)/content_shell.pak',
                 '<(libchromiumcontent_dir)/natives_blob.bin',
                 '<(libchromiumcontent_dir)/snapshot_blob.bin',
@@ -328,6 +326,30 @@
               # Make native module dynamic loading work.
               '-rdynamic',
             ],
+	     'libraries': [
+             '-luv',
+	      '-lz',
+             '-licuuc',
+	      '-lpng',
+	      '-lharfbuzz',
+	      '-lxml2',
+	      '-lxslt',
+	      '-lpciaccess',
+	      '-licui18n',
+             '-lFLAC',
+             '-lFLAC++',
+             '-lspeex',
+             '-ljpeg',
+             '-lminizip',
+             '-levent',
+             '-lspeechd',
+             '-lre2',
+             '-lpci',
+             '-lsnappy',
+             '-lleveldb',
+             '-lwebp',
+             '-lwebpdemux',
+            ],
           },
           # Required settings of using breakpad.
           'cflags_cc': [
@@ -426,7 +448,6 @@
           'mac_bundle_resources': [
             'atom/common/resources/mac/MainMenu.xib',
             '<(libchromiumcontent_dir)/content_shell.pak',
-            '<(libchromiumcontent_dir)/icudtl.dat',
             '<(libchromiumcontent_dir)/natives_blob.bin',
             '<(libchromiumcontent_dir)/snapshot_blob.bin',
           ],
