Description: Fix type ambiguity
Index: ufraw-0.20/dcraw.cc
===================================================================
--- ufraw-0.20.orig/dcraw.cc
+++ ufraw-0.20/dcraw.cc
@@ -113,6 +113,8 @@ extern "C" {
 
 /* All definitions of global variables are defined inside a class in dcraw.h */
 #include "dcraw.h"
+#define abs(i) abs((long)i)
+#define sqrt(r) sqrt((double)r)
 
 #ifdef DCRAW_NOMAIN
 extern

