diff --git a/CMakeLists.txt b/CMakeLists.txt
index 9ff4583..16f5852 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -149,10 +149,6 @@ install(TARGETS
         DESTINATION
           ${CMAKE_INSTALL_FULL_LIBDIR})
 install(FILES
-          ${PTHREAD_WORKQUEUE_HEADERS}
-        DESTINATION
-          ${CMAKE_INSTAL_FULL_INCLUDEDIR})
-install(FILES
           ${PROJECT_SOURCE_DIR}/pthread_workqueue.3
         DESTINATION
           ${CMAKE_INSTALL_FULL_MANDIR})
