--- a/clutter/clutter/clutter-main.c	2017-05-24 15:06:49.345332075 +0200
+++ b/clutter/clutter/clutter-main.c	2017-05-24 15:07:33.007261778 +0200
@@ -3688,8 +3688,6 @@
   fmt = g_strconcat (stamp, ":", format, NULL);
   g_free (stamp);
 
-  g_logv (G_LOG_DOMAIN, G_LOG_LEVEL_MESSAGE, fmt, var_args);
-
   g_free (fmt);
 }
 
@@ -3728,7 +3726,6 @@
   fmt = g_strconcat ("[DIAGNOSTIC]: ", format, NULL);
 
   va_start (args, format);
-  g_logv (G_LOG_DOMAIN, G_LOG_LEVEL_MESSAGE, fmt, args);
   va_end (args);
 
   g_free (fmt);
