--- src/main/main.cpp.bak	2017-07-08 12:03:47.315205913 +0200
+++ src/main/main.cpp	2017-07-08 12:04:47.059208854 +0200
@@ -22,7 +22,7 @@
 #include <QMessageBox> // For QT_REQUIRE_VERSION
 #include <iostream>
 
-#if defined(Q_OS_LINUX) || defined(__GLIBC__) || defined(__FreeBSD__) || defined(__HAIKU__)
+#if defined(Q_OS_LINUX) && defined(__GLIBC__) || defined(__FreeBSD__) || defined(__HAIKU__)
 #include <signal.h>
 #include <execinfo.h>
 
@@ -129,7 +129,7 @@
     qInstallMessageHandler(&msgHandler);
 #endif
 
-#if defined(Q_OS_LINUX) || defined(__GLIBC__) || defined(__FreeBSD__)
+#if defined(Q_OS_LINUX) && defined(__GLIBC__) || defined(__FreeBSD__)
     signal(SIGSEGV, qupzilla_signal_handler);
 #endif
 
