diff --git a/pyproject.toml b/pyproject.toml
index 9517a06..05809c6 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -25,7 +25,7 @@ audio-metadata = "^0.6"
 google-music = "^3.2"
 google-music-proto = "^2.5.1"
 google-music-utils = "^2.1"
-loguru = "^0.3"
+loguru = "0.4.*"
 pendulum = "^2.0"
 pprintpp = "0.*"
 natsort = ">=5.0,<7.0"
diff --git a/src/google_music_scripts/commands.py b/src/google_music_scripts/commands.py
index b124a41..3233d35 100644
--- a/src/google_music_scripts/commands.py
+++ b/src/google_music_scripts/commands.py
@@ -78,7 +78,7 @@ def do_delete(args):
 				)
 		else:
 			logger.info("No songs deleted")
-	elif logger._min_level <= 15:
+	elif logger._core.min_level <= 15:
 		for song in to_delete:
 			title = song.get('title', "<empty>")
 			artist = song.get('artist', "<empty>")
@@ -170,7 +170,7 @@ def do_download(args):
 
 			logger.info("Found {} songs already exist by audio hash", len(existing_songs))
 
-			if logger._min_level <= 5:
+			if logger._core.min_level <= 5:
 				for song in existing_songs:
 					title = song.get('title', "<title>")
 					artist = song.get('artist', "<artist>")
@@ -224,7 +224,7 @@ def do_download(args):
 				len(existing_songs)
 			)
 
-			if logger._min_level <= 5:
+			if logger._core.min_level <= 5:
 				for song in existing_songs:
 					title = song.get('title', "<title>")
 					artist = song.get('artist', "<artist>")
@@ -257,7 +257,7 @@ def do_download(args):
 
 	if not args.dry_run:
 		download_songs(mm, to_download, template=args.output)
-	elif logger._min_level <= 15:
+	elif logger._core.min_level <= 15:
 		for song in to_download:
 			title = song.get('title', "<title>")
 			artist = song.get('artist', "<artist>")
@@ -428,7 +428,7 @@ def do_upload(args):
 
 		logger.info("Found {} songs already exist by audio hash", len(existing_songs))
 
-		if logger._min_level <= 5:
+		if logger._core.min_level <= 5:
 			for song in natsorted(existing_songs):
 				logger.trace(song)
 
@@ -461,7 +461,7 @@ def do_upload(args):
 
 			logger.info("Found {} songs already exist by metadata", len(existing_songs))
 
-			if logger._min_level <= 5:
+			if logger._core.min_level <= 5:
 				for song in existing_songs:
 					logger.trace(song)
 
@@ -482,7 +482,7 @@ def do_upload(args):
 			no_sample=args.no_sample,
 			delete_on_success=args.delete_on_success
 		)
-	elif logger._min_level <= 15:
+	elif logger._core.min_level <= 15:
 		for song in to_upload:
 			logger.log(
 				'ACTION_SUCCESS',
diff --git a/src/google_music_scripts/core.py b/src/google_music_scripts/core.py
index a46cd9c..def00b1 100644
--- a/src/google_music_scripts/core.py
+++ b/src/google_music_scripts/core.py
@@ -251,7 +251,7 @@ def upload_songs(
 				no_sample=no_sample
 			)
 
-			if logger._min_level <= 15:
+			if logger._core.min_level <= 15:
 				if result['reason'] == 'Uploaded':
 					logger.log(
 						'ACTION_SUCCESS',
--- a/setup.py	2020-01-26 14:03:21.699404208 +0100
+++ b/setup.py	2020-01-26 14:05:10.171206600 +0100
@@ -17,12 +17,11 @@
  'google-music-proto>=2.5.1,<3.0.0',
  'google-music-utils>=2.1,<3.0',
  'google-music>=3.2,<4.0',
- 'loguru>=0.3,<0.4',
+ 'loguru>=0.4,<0.5',
  'natsort>=5.0,<7.0',
  'pendulum>=2.0,<3.0',
  'pprintpp<1.0.0',
- 'tbm-utils>=1.0.0,<2.0.0',
- 'tomlkit>=0.5,<0.6']
+ 'tbm-utils>=1.0.0,<2.0.0']
 
 extras_require = \
 {'dev': ['flake8>=3.5,<4.0',
