diff --git a/libclock/clock.cpp b/libclock/clock.cpp
index 1d424be050aab51555602ebcf129cf1a597f11db..d4cf9e36146e0e3a18486e5782ecd9dd6d11e79b 100644
--- a/libclock/clock.cpp
+++ b/libclock/clock.cpp
@@ -136,6 +136,7 @@ void Clock::componentComplete()
         setupTimeZone(QTimeZone(QTimeZone::systemTimeZoneId()));
     }
     setupTickConnections();
+    Q_EMIT timeChanged();
 }
 
 void Clock::update()
