--- a/support/utils.cpp	2016-04-02 12:03:07.425761971 +0200
+++ b/support/utils.cpp	2016-04-02 12:05:03.649758732 +0200
@@ -773,7 +773,7 @@
     #elif defined Q_OS_MAC
     return fixPath(QCoreApplication::applicationDirPath())+app;
     #else
-    return QString(INSTALL_PREFIX "/"LINUX_LIB_DIR"/")+QCoreApplication::applicationName()+constDirSep+app;
+    return QString(INSTALL_PREFIX "/" LINUX_LIB_DIR "/")+QCoreApplication::applicationName()+constDirSep+app;
     #endif
 }
 
