--- tirpc/rpcsvc/nis.h	2015-08-21 15:23:12.061629002 +0000
+++ tiprc/rpcsvc/nis.h	2015-08-21 15:23:15.000000000 +0000
@@ -1,49 +1,50 @@
 /*
- * Copyright (c) 2010, Oracle America, Inc.
+ * Sun RPC is a product of Sun Microsystems, Inc. and is provided for
+ * unrestricted use provided that this legend is included on all tape
+ * media and as a part of the software program in whole or part.  Users
+ * may copy or modify Sun RPC without charge, but are not authorized
+ * to license or distribute it to anyone else except as part of a product or
+ * program developed by the user or with the express written consent of
+ * Sun Microsystems, Inc.
  *
- * Redistribution and use in source and binary forms, with or without
- * modification, are permitted provided that the following conditions are
- * met:
+ * SUN RPC IS PROVIDED AS IS WITH NO WARRANTIES OF ANY KIND INCLUDING THE
+ * WARRANTIES OF DESIGN, MERCHANTIBILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE, OR ARISING FROM A COURSE OF DEALING, USAGE OR TRADE PRACTICE.
  *
- *     * Redistributions of source code must retain the above copyright
- *       notice, this list of conditions and the following disclaimer.
- *     * Redistributions in binary form must reproduce the above
- *       copyright notice, this list of conditions and the following
- *       disclaimer in the documentation and/or other materials
- *       provided with the distribution.
- *     * Neither the name of the "Oracle America, Inc." nor the names of its
- *       contributors may be used to endorse or promote products derived
- *       from this software without specific prior written permission.
+ * Sun RPC is provided with no support and without any obligation on the
+ * part of Sun Microsystems, Inc. to assist in its use, correction,
+ * modification or enhancement.
  *
- *   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
- *   "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
- *   LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
- *   FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
- *   COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
- *   INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
- *   DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
- *   GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
- *   INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
- *   WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
- *   NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
- *   OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ * SUN MICROSYSTEMS, INC. SHALL HAVE NO LIABILITY WITH RESPECT TO THE
+ * INFRINGEMENT OF COPYRIGHTS, TRADE SECRETS OR ANY PATENTS BY SUN RPC
+ * OR ANY PART THEREOF.
+ *
+ * In no event will Sun Microsystems, Inc. be liable for any lost revenue
+ * or profits or other special, indirect and consequential damages, even if
+ * Sun has been advised of the possibility of such damages.
+ *
+ * Sun Microsystems, Inc.
+ * 2550 Garcia Avenue
+ * Mountain View, California  94043
  */
 
 #ifndef _RPCSVC_NIS_H
 #define _RPCSVC_NIS_H 1
 
 #include <features.h>
-#include <rpc/rpc.h>
-#include <rpcsvc/nis_tags.h>
+//#include <rpc/rpc.h>
+//#include <rpcsvc/nis_tags.h>
 
-__BEGIN_DECLS
+#ifdef __cplusplus
+extern "C" {
+#endif
 
 /*
  *	nis.h
  *
  *	This file is the main include file for NIS clients. It contains
  *	both the client library function defines and the various data
- *	structures used by the NIS service. It includes the file nis_tags.h
+ * 	structures used by the NIS service. It includes the file nis_tags.h
  *	which defines the tag values. This allows the tags to change without
  *	having to change the nis.x file.
  *
@@ -120,8 +121,8 @@
 typedef enum nstype nstype;
 
 struct oar_mask {
-	uint32_t oa_rights;
-	zotypes oa_otype;
+        uint32_t oa_rights;
+        zotypes oa_otype;
 };
 typedef struct oar_mask oar_mask;
 
@@ -517,7 +518,7 @@
 
 
 /* Prototypes, and extern declarations for the NIS library functions. */
-#include <rpcsvc/nislib.h>
+//#include <rpcsvc/nislib.h>
 #endif
 
 /*
@@ -537,80 +538,8 @@
 #define NIS_PROG 100300
 #define NIS_VERSION 3
 
-#define NIS_LOOKUP 1
-extern  nis_result * nis_lookup_3 (ns_request *, CLIENT *) __THROW;
-extern  nis_result * nis_lookup_3_svc (ns_request *, struct svc_req *) __THROW;
-#define NIS_ADD 2
-extern  nis_result * nis_add_3 (ns_request *, CLIENT *) __THROW;
-extern  nis_result * nis_add_3_svc (ns_request *, struct svc_req *) __THROW;
-#define NIS_MODIFY 3
-extern  nis_result * nis_modify_3 (ns_request *, CLIENT *) __THROW;
-extern  nis_result * nis_modify_3_svc (ns_request *, struct svc_req *) __THROW;
-#define NIS_REMOVE 4
-extern  nis_result * nis_remove_3 (ns_request *, CLIENT *) __THROW;
-extern  nis_result * nis_remove_3_svc (ns_request *, struct svc_req *) __THROW;
-#define NIS_IBLIST 5
-extern  nis_result * nis_iblist_3 (ib_request *, CLIENT *) __THROW;
-extern  nis_result * nis_iblist_3_svc (ib_request *, struct svc_req *) __THROW;
-#define NIS_IBADD 6
-extern  nis_result * nis_ibadd_3 (ib_request *, CLIENT *) __THROW;
-extern  nis_result * nis_ibadd_3_svc (ib_request *, struct svc_req *) __THROW;
-#define NIS_IBMODIFY 7
-extern  nis_result * nis_ibmodify_3 (ib_request *, CLIENT *) __THROW;
-extern  nis_result * nis_ibmodify_3_svc (ib_request *, struct svc_req *)
-     __THROW;
-#define NIS_IBREMOVE 8
-extern  nis_result * nis_ibremove_3 (ib_request *, CLIENT *) __THROW;
-extern  nis_result * nis_ibremove_3_svc (ib_request *, struct svc_req *)
-     __THROW;
-#define NIS_IBFIRST 9
-extern  nis_result * nis_ibfirst_3 (ib_request *, CLIENT *) __THROW;
-extern  nis_result * nis_ibfirst_3_svc (ib_request *, struct svc_req *)
-     __THROW;
-#define NIS_IBNEXT 10
-extern  nis_result * nis_ibnext_3 (ib_request *, CLIENT *) __THROW;
-extern  nis_result * nis_ibnext_3_svc (ib_request *, struct svc_req *) __THROW;
-#define NIS_FINDDIRECTORY 12
-extern  fd_result * nis_finddirectory_3 (fd_args *, CLIENT *) __THROW;
-extern  fd_result * nis_finddirectory_3_svc (fd_args *,
-					     struct svc_req *) __THROW;
-#define NIS_STATUS 14
-extern  nis_taglist * nis_status_3 (nis_taglist *, CLIENT *) __THROW;
-extern  nis_taglist * nis_status_3_svc (nis_taglist *, struct svc_req *)
-     __THROW;
-#define NIS_DUMPLOG 15
-extern  log_result * nis_dumplog_3 (dump_args *, CLIENT *) __THROW;
-extern  log_result * nis_dumplog_3_svc (dump_args *, struct svc_req *) __THROW;
-#define NIS_DUMP 16
-extern  log_result * nis_dump_3 (dump_args *, CLIENT *) __THROW;
-extern  log_result * nis_dump_3_svc (dump_args *, struct svc_req *) __THROW;
-#define NIS_CALLBACK 17
-extern  bool_t * nis_callback_3 (netobj *, CLIENT *) __THROW;
-extern  bool_t * nis_callback_3_svc (netobj *, struct svc_req *) __THROW;
-#define NIS_CPTIME 18
-extern  uint32_t * nis_cptime_3 (nis_name *, CLIENT *) __THROW;
-extern  uint32_t * nis_cptime_3_svc (nis_name *, struct svc_req *) __THROW;
-#define NIS_CHECKPOINT 19
-extern  cp_result * nis_checkpoint_3 (nis_name *, CLIENT *) __THROW;
-extern  cp_result * nis_checkpoint_3_svc (nis_name *, struct svc_req *)
-     __THROW;
-#define NIS_PING 20
-extern  void * nis_ping_3 (ping_args *, CLIENT *) __THROW;
-extern  void * nis_ping_3_svc (ping_args *, struct svc_req *) __THROW;
-#define NIS_SERVSTATE 21
-extern  nis_taglist * nis_servstate_3 (nis_taglist *, CLIENT *) __THROW;
-extern  nis_taglist * nis_servstate_3_svc (nis_taglist *,
-					   struct svc_req *) __THROW;
-#define NIS_MKDIR 22
-extern  nis_error * nis_mkdir_3 (nis_name *, CLIENT *) __THROW;
-extern  nis_error * nis_mkdir_3_svc (nis_name *, struct svc_req *) __THROW;
-#define NIS_RMDIR 23
-extern  nis_error * nis_rmdir_3 (nis_name *, CLIENT *) __THROW;
-extern  nis_error * nis_rmdir_3_svc (nis_name *, struct svc_req *) __THROW;
-#define NIS_UPDKEYS 24
-extern  nis_error * nis_updkeys_3 (nis_name *, CLIENT *) __THROW;
-extern  nis_error * nis_updkeys_3_svc (nis_name *, struct svc_req *) __THROW;
-
-__END_DECLS
+#ifdef __cplusplus
+}
+#endif
 
 #endif /* ! _RPCSVC_NIS_H */
