--- source/creator/creator.c.bak	2016-04-04 16:35:46.633388933 +0200
+++ source/creator/creator.c	2016-04-04 16:36:06.580389915 +0200
@@ -618,7 +618,7 @@
 	 * set breakpoints on fpe_handler */
 	signal(SIGFPE, fpe_handler);
 
-# if defined(__linux__) && defined(__GNUC__)
+# if defined(__linux__) && defined(__GLIBC__)
 	feenableexcept(FE_DIVBYZERO | FE_INVALID | FE_OVERFLOW);
 # endif /* defined(__linux__) && defined(__GNUC__) */
 # if defined(OSX_SSE_FPE)
