--- a/src/bootstrap/compile.rs	2016-10-08 14:32:50.900901529 +0200
+++ b/src/bootstrap/compile.rs	2016-10-08 14:33:04.700902208 +0200
@@ -114,10 +114,6 @@
     let libdir = build.sysroot_libdir(target_compiler, target);
     add_to_sysroot(&libdir, &libstd_stamp(build, compiler, target));
 
-    if target.contains("musl") && !target.contains("mips") {
-        copy_musl_third_party_objects(build, target, &libdir);
-    }
-
     if build.config.sanitizers && compiler.stage != 0 && target == "x86_64-apple-darwin" {
         // The sanitizers are only built in stage1 or above, so the dylibs will
         // be missing in stage0 and causes panic. See the `std()` function above
@@ -126,15 +122,6 @@
     }
 }
 
-/// Copies the crt(1,i,n).o startup objects
-///
-/// Only required for musl targets that statically link to libc
-fn copy_musl_third_party_objects(build: &Build, target: &str, into: &Path) {
-    for &obj in &["crt1.o", "crti.o", "crtn.o"] {
-        copy(&build.musl_root(target).unwrap().join("lib").join(obj), &into.join(obj));
-    }
-}
-
 fn copy_apple_sanitizer_dylibs(native_dir: &Path, platform: &str, into: &Path) {
     for &sanitizer in &["asan", "tsan"] {
         let filename = format!("libclang_rt.{}_{}_dynamic.dylib", sanitizer, platform);
