--- a/src/librustc_back/target/x86_64_unknown_linux_musl.rs	2016-10-08 14:19:51.995863180 +0200
+++ b/src/librustc_back/target/x86_64_unknown_linux_musl.rs	2016-10-08 14:20:24.596864785 +0200
@@ -15,6 +15,7 @@
     base.max_atomic_width = Some(64);
     base.pre_link_args.get_mut(&LinkerFlavor::Gcc).unwrap().push("-m64".to_string());
+    base.no_default_libraries = false;
 
     Ok(Target {
         llvm_target: "x86_64-unknown-linux-musl".to_string(),
