--- a/src/libunwind/build.rs	2016-11-10 16:33:04.611193849 +0100
+++ b/src/libunwind/build.rs	2016-11-10 16:33:23.011113762 +0100
@@ -17,7 +17,7 @@
 
     if target.contains("linux") {
         if target.contains("musl") && !target.contains("mips") {
-            println!("cargo:rustc-link-lib=static=unwind");
+            println!("cargo:rustc-link-lib=unwind");
         } else if !target.contains("android") {
             println!("cargo:rustc-link-lib=gcc_s");
         
