--- a/src/librustc_llvm/build.rs	2017-03-18 15:00:25.871282618 +0100
+++ b/src/librustc_llvm/build.rs	2017-03-18 15:00:52.012218593 +0100
@@ -25,13 +25,7 @@
         .filter_map(|s| s.parse::<u32>().ok());
     if let (Some(major), Some(minor)) = (parts.next(), parts.next()) {
         if major > 3 || (major == 3 && minor >= 9) {
-            // Force the link mode we want, preferring static by default, but
-            // possibly overridden by `configure --enable-llvm-link-shared`.
-            if env::var_os("LLVM_LINK_SHARED").is_some() {
-                return ("dylib", Some("--link-shared"));
-            } else {
-                return ("static", Some("--link-static"));
-            }
+            return ("dylib", Some("--link-shared"));
         } else if major == 3 && minor == 8 {
             // Find out LLVM's default linking mode.
             let mut mode_cmd = Command::new(llvm_config);
