--- a/src/librustc_llvm/build.rs	2016-10-21 15:35:12.331742403 +0200
+++ b/src/librustc_llvm/build.rs	2016-10-21 15:35:28.181741099 +0200
@@ -203,7 +203,7 @@
         } else if cxxflags.contains("stdlib=libc++") {
             println!("cargo:rustc-link-lib=c++");
         } else {
-            println!("cargo:rustc-link-lib=stdc++");
+            println!("cargo:rustc-link-lib=c++");
         }
     }
 }
