--- a/src/libstd/lib.rs	2016-11-10 16:38:41.460851478 +0100
+++ b/src/libstd/lib.rs	2016-11-10 16:38:59.136786859 +0100
@@ -245,7 +245,6 @@
 #![feature(collections)]
 #![feature(collections_bound)]
 #![feature(collections_range)]
-#![feature(compiler_builtins_lib)]
 #![feature(const_fn)]
 #![feature(core_float)]
 #![feature(core_intrinsics)]
@@ -333,9 +332,6 @@
 #[cfg(stage0)]
 extern crate alloc_system;
 
-// compiler-rt intrinsics
-extern crate compiler_builtins;
-
 // During testing, this crate is not actually the "real" std library, but rather
 // it links to the real std library, which was compiled from this same source
 // code. So any lang items std defines are conditionally excluded (or else they
--- a/src/libstd/Cargo.toml	2016-11-10 16:36:43.186296681 +0100
+++ b/src/libstd/Cargo.toml	2016-11-10 16:36:58.511237638 +0100
@@ -19,7 +19,6 @@
 core = { path = "../libcore" }
 libc = { path = "../rustc/libc_shim" }
 rand = { path = "../librand" }
-compiler_builtins = { path = "../libcompiler_builtins" }
 std_unicode = { path = "../libstd_unicode" }
 unwind = { path = "../libunwind" }
