--- a/src/bootstrap/tool.rs
+++ b/src/bootstrap/tool.rs
@@ -313,6 +313,10 @@ impl Step for Rustdoc {
         cargo.env("RUSTC_DEBUGINFO", builder.config.rust_debuginfo.to_string())
              .env("RUSTC_DEBUGINFO_LINES", builder.config.rust_debuginfo_lines.to_string());
 
+        if build.config.llvm_link_shared {
+            cargo.env("LLVM_LINK_SHARED", "1");
+        }
+
         build.run(&mut cargo);
         // Cargo adds a number of paths to the dylib search path on windows, which results in
         // the wrong rustdoc being executed. To avoid the conflicting rustdocs, we name the "tool"
--- a/src/tools/rustdoc/Cargo.toml
+++ b/src/tools/rustdoc/Cargo.toml
@@ -12,3 +12,4 @@ path = "main.rs"
 
 [dependencies]
 rustdoc = { path = "../../librustdoc" }
+rustc_llvm = { path = "../../librustc_llvm" }
--- a/src/tools/rustdoc/main.rs
+++ b/src/tools/rustdoc/main.rs
@@ -9,5 +9,6 @@
 // except according to those terms.
 
 extern crate rustdoc;
+extern crate rustc_llvm;
 
 fn main() { rustdoc::main() }
--- a/src/tools/rls/Cargo.toml
+++ b/src/tools/rls/Cargo.toml
@@ -22,6 +22,7 @@
 rls-rustc = "0.1"
 rls-span = { version = "0.4", features = ["serialize-serde"] }
 rls-vfs = { version = "0.4", features = ["racer-impls"] }
+rustc_llvm = { path = "../../librustc_llvm" }
 rustfmt-nightly = "0.2.16"
 serde = "1.0"
 serde_json = "1.0"
--- a/src/tools/rls/src/main.rs
+++ b/src/tools/rls/src/main.rs
@@ -14,6 +14,8 @@
 #![feature(integer_atomics)]
 #![feature(fnbox)]
 
+extern crate rustc_llvm;
+
 extern crate cargo;
 extern crate env_logger;
 extern crate languageserver_types as ls_types;
--- a/src/tools/clippy/Cargo.toml	2018-01-01 22:52:59.000000000 +0100
+++ b/src/tools/clippy/Cargo.toml.new	2018-01-06 17:51:36.848306374 +0100
@@ -41,6 +41,7 @@
 # end automatic update
 cargo_metadata = "0.2"
 regex = "0.2"
+rustc_llvm = { path = "../../librustc_llvm" }
 
 [dev-dependencies]
 compiletest_rs = "0.3"
--- a/src/tools/clippy/src/main.rs	2018-01-01 22:52:59.000000000 +0100
+++ b/src/tools/clippy/src/main.rs.new	2018-01-06 17:52:05.185239927 +0100
@@ -7,6 +7,7 @@
 use std::process;
 use std::io::{self, Write};
 
+extern crate rustc_llvm;
 extern crate cargo_metadata;
 
 use std::path::Path;
