diff --git a/include/nss.h b/include/nss.h
index f02addb..b5a046c 100644
--- a/include/nss.h
+++ b/include/nss.h
@@ -1,6 +1,8 @@
 #ifndef NSS__H
 #define NSS__H
 
+#include <stdint.h>
+
 enum nss_status
 {
 	NSS_STATUS_TRYAGAIN = -2,
@@ -10,4 +12,13 @@ enum nss_status
 	NSS_STATUS_RETURN = 2
 };
 
+struct gaih_addrtuple
+{
+        struct gaih_addrtuple *next;
+        char *name;
+        int family;
+        uint32_t addr[4];
+        uint32_t scopeid;
+};
+
 #endif
