--- a/lib/Makefile	2016-10-16 13:23:55.802743400 +0200
+++ b/lib/Makefile	2016-10-16 13:24:11.202742133 +0200
@@ -59,7 +59,7 @@
  ifdef PCI_HAVE_PM_DARWIN_DEVICE
 	$(CC) -shared $(LDFLAGS) $(SONAME) -Wl,-install_name,$(LIBDIR)/$(PCILIB) -o $@ $^ $(LIB_LDLIBS)
  else
-	$(CC) -shared $(LDFLAGS) $(SONAME) -Wl,--version-script=libpci.ver -o $@ $^ $(LIB_LDLIBS)
+	$(CC) -shared $(LDFLAGS) $(SONAME) -Wl,--sandbox-version-script=libpci.ver -o $@ $^ $(LIB_LDLIBS)
  endif
 endif
 
