--- a/src/Data/Compression/Inflate.hs	2018-09-22 22:45:14.525262647 +0200
+++ b/src/Data/Compression/Inflate.hs	2018-09-22 22:53:39.780887970 +0200
@@ -40,6 +40,7 @@
 import qualified Data.Char
 import Control.Applicative
 import Control.Monad
+import Control.Monad.Fail
 
 import Data.Bits
 import Data.Word
@@ -126,6 +127,9 @@
  -- return :: a -> InfM a
     return x = InfM $ \s -> (x, s)
 
+instance MonadFail InfM where
+    fail = error "something failed"
+
 instance Applicative InfM where
     pure = return
     (<*>) = ap
