--- a/include/qemu/host-utils.h	2017-11-12 10:48:11.332923204 +0100
+++ b/include/qemu/host-utils.h	2017-11-12 10:48:40.763866951 +0100
@@ -207,7 +207,7 @@
  */
 static inline int clrsb32(uint32_t val)
 {
-#if QEMU_GNUC_PREREQ(4, 7)
+#if QEMU_GNUC_PREREQ(4, 7) && !defined(__clang__)
     return __builtin_clrsb(val);
 #else
     return clz32(val ^ ((int32_t)val >> 1)) - 1;
@@ -223,7 +223,7 @@
  */
 static inline int clrsb64(uint64_t val)
 {
-#if QEMU_GNUC_PREREQ(4, 7)
+#if QEMU_GNUC_PREREQ(4, 7) && !defined(__clang__)
     return __builtin_clrsbll(val);
 #else
     return clz64(val ^ ((int64_t)val >> 1)) - 1;
