--- a/fts.h	2016-09-13 18:01:10.680722494 +0200
+++ b/fts.h	2016-09-13 18:02:11.277761552 +0200
@@ -35,6 +35,17 @@
 #ifndef	_FTS_H_
 #define	_FTS_H_
 
+#include <sys/stat.h>
+#include <sys/param.h>
+
+#include <dirent.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <fts.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
 typedef struct {
 	struct _ftsent *fts_cur;	/* current node */
 	struct _ftsent *fts_child;	/* linked list of children */
--- a/fts.c	2016-09-13 18:01:28.040248783 +0200
+++ b/fts.c	2016-09-13 18:01:57.774752461 +0200
@@ -29,16 +29,7 @@
  * SUCH DAMAGE.
  */
 
-#include <sys/param.h>
-#include <sys/stat.h>
-
-#include <dirent.h>
-#include <errno.h>
-#include <fcntl.h>
 #include <fts.h>
-#include <stdlib.h>
-#include <string.h>
-#include <unistd.h>
 
 /* Largest alignment size needed, minus one.
  *    Usually long double is the worst case.  */

