--- a/src/k3bsystemproblemdialog.cpp	2017-07-08 22:27:09.715566034 +0200
+++ b/src/k3bsystemproblemdialog.cpp	2017-07-08 22:31:47.011579686 +0200
@@ -57,9 +57,6 @@
 #include <langinfo.h>
 #endif
 
-#ifndef Q_OS_WIN32
-#include <fstab.h>
-#endif
 #include <unistd.h>
 
 
@@ -389,23 +386,6 @@
             dvd_r_dl = true;
     }
 
-#ifndef Q_OS_WIN32
-    // check automounted devices
-    QList<K3b::Device::Device*> automountedDevices = checkForAutomounting();
-    for( QList<K3b::Device::Device *>::const_iterator it = automountedDevices.constBegin();
-         it != automountedDevices.constEnd(); ++it ) {
-        problems.append( K3b::SystemProblem( K3b::SystemProblem::NON_CRITICAL,
-                                           i18n("Device %1 - %2 is automounted.",
-                                                (*it)->vendor(),(*it)->description()),
-                                           i18n("K3b is not able to unmount automounted devices. Thus, especially "
-                                                "DVD+RW rewriting might fail. There is no need to report this as "
-                                                "a bug or feature wish; it is not possible to solve this problem "
-                                                "from within K3b."),
-                                           i18n("Replace the automounting entries in /etc/fstab with old-fashioned "
-                                                "ones or use a user-space mounting solution like pmount or ivman.") ) );
-    }
-
-#endif
     if( atapiWriter ) {
         if( !K3b::plainAtapiSupport() &&
             !K3b::hackedAtapiSupport() ) {
@@ -676,40 +656,6 @@
 }
 
 
-#ifndef Q_OS_WIN32
-QList<K3b::Device::Device*> K3b::SystemProblemDialog::checkForAutomounting()
-{
-    QList<K3b::Device::Device *> l;
-    ::setfsent();
-
-    struct fstab * mountInfo = 0;
-    while( (mountInfo = ::getfsent()) )
-    {
-        // check if the entry corresponds to a device
-        QString md = K3b::resolveLink( QFile::decodeName( mountInfo->fs_spec ) );
-        QString type = QFile::decodeName( mountInfo->fs_vfstype );
-
-        if( type == "supermount" || type == "subfs" ) {
-            // parse the device
-            QStringList opts = QString::fromLocal8Bit(mountInfo->fs_mntops).split( ',' );
-            for( QStringList::const_iterator it = opts.constBegin(); it != opts.constEnd(); ++it ) {
-                if( (*it).startsWith("dev=") ) {
-                    md = (*it).mid( 4 );
-                    break;
-                }
-            }
-
-            if( K3b::Device::Device* dev = k3bcore->deviceManager()->findDevice( md ) )
-                l.append( dev );
-        }
-    } // while mountInfo
-
-    ::endfsent();
-    return l;
-}
-#endif
-
-
 bool K3b::SystemProblemDialog::readCheckSystemConfig()
 {
     KConfigGroup cfgGrp( KSharedConfig::openConfig(), "General Options" );
--- a/src/k3bsystemproblemdialog.h	2017-07-08 22:31:04.851577611 +0200
+++ b/src/k3bsystemproblemdialog.h	2017-07-08 22:31:10.051577867 +0200
@@ -93,9 +93,6 @@
                              bool showBinSettingsButton,
                              QWidget* parent = 0);
         static int dmaActivated( Device::Device* );
-#ifndef Q_OS_WIN32
-        static QList<Device::Device*> checkForAutomounting();
-#endif
         QCheckBox* m_checkDontShowAgain;
     };
 }
