--- a/src/kcatalog.cpp.bak	2016-02-17 19:21:12.829135921 +0100
+++ b/src/kcatalog.cpp	2016-02-17 19:21:31.253134823 +0100
@@ -169,7 +169,7 @@
 
         // Magic to make sure GNU Gettext doesn't use stale cached translation
         // from previous language.
-#if defined(__USE_GNU_GETTEXT)
+#if defined(__GLIBC__)
 #if defined(Q_OS_WIN)
         extern "C" int __declspec(dllimport) _nl_msg_cat_cntr;
 #else
