--- a/Package@swift-4.swift	2017-09-22 11:04:31.651910853 +0200
+++ b/Package@swift-4.swift	2017-09-22 11:04:43.384910581 +0200
@@ -4,8 +4,8 @@
 let package = Package(
     name: "Routing",
     products: [
-        .library(name: "Branches", targets: ["Branches"]),
-        .library(name: "Routing", targets: ["Routing"]),
+        .library(name: "Branches", type: .dynamic, targets: ["Branches"]),
+        .library(name: "Routing", type: .dynamic, targets: ["Routing"]),
     ],
     targets: [
         .target(name: "Branches"),
