diff --git a/Package.swift b/Package.swift
index ef2a09c8..a06aa0bf 100644
--- a/Package.swift
+++ b/Package.swift
@@ -20,3 +20,10 @@ let package = Package(
         "Sources/TypeSafeGenerator"
     ]
 )
+
+let rlib = Product(name: "Routing", type: .Library(.Dynamic), modules: "Routing")
+products.append(rlib)
+let hlib = Product(name: "HTTPRouting", type: .Library(.Dynamic), modules: "HTTPRouting")
+products.append(hlib)
+let tlib = Product(name: "TypeSafeRouting", type: .Library(.Dynamic), modules: "TypeSafeRouting")
+products.append(tlib)
