diff --git a/Package.swift b/Package.swift
index ef2a09c8..12fe2350 100644
--- a/Package.swift
+++ b/Package.swift
@@ -11,12 +11,7 @@ let package = Package(
         Target(name: "Branches"),
         Target(name: "Routing", dependencies: ["Branches"]),
     ],
-    dependencies: [
-        // Core vapor transport layer
-        .Package(url: "https://github.com/vapor/engine.git", Version(2,0,0, prereleaseIdentifiers: ["beta"])),
-        .Package(url: "https://github.com/vapor/node.git", Version(2,0,0, prereleaseIdentifiers: ["beta"]))
-    ],
     exclude: [
         "Sources/TypeSafeGenerator"
     ]
