diff --git a/Package.swift b/Package.swift
index ef2a09c8..12fe2350 100644
--- a/Package.swift
+++ b/Package.swift
@@ -11,12 +11,7 @@ let package = Package(
         Target(name: "TypeSafeRouting", dependencies: ["Routing", "HTTPRouting"]),
         // Target(name: "TypeSafeGenerator"),
     ],
-    dependencies: [
-        // Core vapor transport layer
-        .Package(url: "https://github.com/vapor/engine.git", majorVersion: 1),
-        .Package(url: "https://github.com/vapor/node.git", majorVersion: 1)
-    ],
     exclude: [
         "Sources/TypeSafeGenerator"
     ]
