diff --git a/Package@swift-4.swift b/Package@swift-4.swift
index 128f35e1..99133b8c 100644
--- a/Package@swift-4.swift
+++ b/Package@swift-4.swift
@@ -7,16 +7,10 @@ let package = Package(
         .library(name: "Branches", targets: ["Branches"]),
         .library(name: "Routing", targets: ["Routing"]),
     ],
-    dependencies: [
-        .package(url: "https://github.com/vapor/engine.git", .upToNextMajor(from: "2.2.0")),
-        .package(url: "https://github.com/vapor/core.git", .upToNextMajor(from: "2.1.2")),
-        .package(url: "https://github.com/vapor/node.git", .upToNextMajor(from: "2.1.1")),
-        .package(url: "https://github.com/vapor/debugging.git", .upToNextMajor(from: "1.1.0")),
-    ],
     targets: [
-        .target(name: "Branches", dependencies: ["Core", "Node"]),
-        .testTarget(name: "BranchesTests", dependencies: ["Branches", "HTTP"]),
-        .target(name: "Routing", dependencies: ["Branches", "HTTP", "WebSockets"]),
+        .target(name: "Branches"),
+        .testTarget(name: "BranchesTests", dependencies: ["Branches"]),
+        .target(name: "Routing", dependencies: ["Branches"]),
         .testTarget(name: "RoutingTests", dependencies: ["Routing"]),
     ]
-)
\ No newline at end of file
+)
