diff --git a/Package.swift b/Package.swift
index f8290dc..4af9fcf 100644
--- a/Package.swift
+++ b/Package.swift
@@ -3,3 +3,9 @@ import PackageDescription
 let package = Package(
   name: "Socks"
 )
+
+let slib = Product(name: "Socks", type: .Library(.Dynamic), modules: "Socks")
+products.append(slib)
+
+let sclib = Product(name: "SocksCore", type: .Library(.Dynamic), modules: "SocksCore")
+products.append(sclib)
