--- a/lib/Target/X86/X86TargetObjectFile.cpp	2017-07-19 10:59:09.357644885 +0200
+++ b/lib/Target/X86/X86TargetObjectFile.cpp	2017-07-19 10:59:23.276619732 +0200
@@ -83,7 +83,7 @@
 X86LinuxNaClTargetObjectFile::Initialize(MCContext &Ctx,
                                          const TargetMachine &TM) {
   TargetLoweringObjectFileELF::Initialize(Ctx, TM);
-  InitializeELF(TM.Options.UseInitArray);
+  InitializeELF(true);
 }
 
 const MCExpr *X86WindowsTargetObjectFile::lowerRelativeReference(
