--- a/tools/llvm-objcopy/llvm-objcopy.cpp
+++ b/tools/llvm-objcopy/llvm-objcopy.cpp
@@ -188,9 +188,8 @@
 /// format-agnostic modifications, i.e. preserving dates.
 static void executeObjcopy(const CopyConfig &Config) {
   sys::fs::file_status Stat;
-  if (Config.PreserveDates)
-    if (auto EC = sys::fs::status(Config.InputFilename, Stat))
-      reportError(Config.InputFilename, EC);
+  if (auto EC = sys::fs::status(Config.InputFilename, Stat))
+    reportError(Config.InputFilename, EC);
 
   if (Config.InputFormat == "binary") {
     auto BufOrErr = MemoryBuffer::getFile(Config.InputFilename);
@@ -212,6 +211,8 @@
     }
   }
 
+  sys::fs::setPermissions(Config.OutputFilename, Stat.permissions());
+
   if (Config.PreserveDates) {
     restoreDateOnFile(Config.OutputFilename, Stat);
     if (!Config.SplitDWO.empty())
--- a/tools/llvm-objcopy/StripOpts.td
+++ b/tools/llvm-objcopy/StripOpts.td
@@ -43,7 +43,7 @@
                   HelpText<"Remove debugging symbols only">;
 def d : Flag<["-"], "d">, Alias<strip_debug>;
 def g : Flag<["-"], "g">, Alias<strip_debug>;
-def S : Flag<["-"], "S">, Alias<strip_debug>;
+def S : Flag<["-"], "S">, Alias<strip_all_gnu>;
 def strip_unneeded : Flag<["-", "--"], "strip-unneeded">,
                      HelpText<"Remove all symbols not needed by relocations">;
 
