diff --git a/lib/Target/X86/X86.td b/lib/Target/X86/X86.td
index 3112f00c91f..31dd1ea4d79 100644
--- a/lib/Target/X86/X86.td
+++ b/lib/Target/X86/X86.td
@@ -886,6 +886,7 @@ def ProcessorFeatures {
                                        FeatureCLZERO,
                                        FeatureCMOV,
                                        Feature64Bit,
+                                       FeatureCMPXCHG8B,
                                        FeatureCMPXCHG16B,
                                        FeatureF16C,
                                        FeatureFMA,
