--- a/tools/lld/ELF/DriverUtils.cpp	2016-07-16 21:16:15.000000000 +0200
+++ b/tools/lld/ELF/DriverUtils.cpp	2016-10-05 23:14:28.135320326 +0200
@@ -95,14 +95,22 @@
 void elf::printHelp(const char *Argv0) {
   ELFOptTable Table;
   Table.PrintHelp(outs(), Argv0, "lld", false);
+  # keep track of supported targets in Driver.cpp
+  std::string Targets = std::string("\nld: supported targets: elf32-i386");
+  Targets += std::string(" elf-iamcu elf_amd64 elf_x86_64 elf32-x86-64");
+  Targets += std::string(" elf64ppc elf64ltsmip elf64btsmip elf32ppc");
+  Targets += std::string(" elf32ltsmip elf32btsmip aarch64elf");
+  Targets += std::string(" aarch64linux armelf_linux_eabi\n");
+  outs() << Targets;
 }
 
 std::string elf::getVersionString() {
   std::string Version = getLLDVersion();
   std::string Repo = getLLDRepositoryVersion();
   if (Repo.empty())
-    return "LLD " + Version + "\n";
-  return "LLD " + Version + " " + Repo + "\n";
+  # kind of a hack, but it works with libtool
+    return "GNU LD " + Version + "\n";
+  return "GNU LD " + Version + " " + Repo + "\n";
 }
 
 // Makes a given pathname an absolute path first, and then remove
