--- a/tools/lld/ELF/Driver.cpp	2016-10-05 19:02:18.635575434 +0200
+++ b/tools/lld/ELF/Driver.cpp	2016-10-05 19:02:41.635576566 +0200
@@ -442,10 +442,6 @@
 
   for (auto *Arg : Args.filtered(OPT_export_dynamic_symbol))
     Config->DynamicList.push_back(Arg->getValue());
-
-  if (auto *Arg = Args.getLastArg(OPT_version_script))
-    if (Optional<MemoryBufferRef> Buffer = readFile(Arg->getValue()))
-      parseVersionScript(*Buffer);
 }
 
 void LinkerDriver::createFiles(opt::InputArgList &Args) {
@@ -555,7 +551,6 @@
   Symtab.scanUndefinedFlags();
   Symtab.scanShlibUndefined();
   Symtab.scanDynamicList();
-  Symtab.scanVersionScript();
   Symtab.scanSymbolVersions();
 
   Symtab.addCombinedLtoObject();
