--- a/tools/lld/ELF/SymbolListFile.cpp	2016-10-11 17:58:38.580414162 +0200
+++ b/tools/lld/ELF/SymbolListFile.cpp	2016-10-11 19:12:32.560632466 +0200
@@ -72,7 +72,7 @@
   void parseExtern(std::vector<SymbolVersion> *Globals);
   void parseVersion(StringRef VerStr);
   void parseGlobal(StringRef VerStr);
-  void parseLocal();
+  void parseLocal(StringRef VerStr);
 };
 
 size_t elf::defineSymbolVersion(StringRef VerStr) {
@@ -89,7 +89,7 @@
   if (skip("global:") || peek() != "local:")
     parseGlobal(VerStr);
   if (skip("local:"))
-    parseLocal();
+    parseLocal(VerStr);
   expect("}");
 
   // Each version may have a parent version. For example, "Ver2" defined as
@@ -101,10 +101,16 @@
   expect(";");
 }
 
-void VersionScriptParser::parseLocal() {
+void VersionScriptParser::parseLocal(StringRef VerStr) {
   Config->DefaultSymbolVersion = VER_NDX_LOCAL;
-  expect("*");
-  expect(";");
+  if(peek() == "*") {
+    expect("*");
+    expect(";");
+  }
+  else {
+    while(!VerStr.empty() && peek() != "}")
+      next();
+  }
 }
 
 void VersionScriptParser::parseExtern(std::vector<SymbolVersion> *Globals) {
