--- a/tools/lld/ELF/Options.td	2016-10-06 12:37:11.366433675 +0200
+++ a/tools/lld/ELF/Options.td	2016-10-06 12:42:05.543448159 +0200
@@ -165,9 +165,6 @@
 
 def version: F<"version">, HelpText<"Display the version number">;
 
-def version_script: S<"version-script">,
-  HelpText<"Read a version script">;
-
 def warn_common: F<"warn-common">,
   HelpText<"Warn about duplicate common symbols">;
 
@@ -254,9 +251,12 @@
 def rpath_link: S<"rpath-link">;
 def rpath_link_eq: J<"rpath-link=">;
 def sort_common: F<"sort-common">;
+def stats: Flag<["-"], "stats">;
+def version_script: S<"version-script">;
 def warn_execstack: F<"warn-execstack">;
 def warn_shared_textrel: F<"warn-shared-textrel">;
 def G: Separate<["-"], "G">;
+def Map: Separate<["-"], "Map">;
 
 // Aliases for ignored options
 def alias_define_common_d: Flag<["-"], "d">, Alias<define_common>;
