--- a/tools/llvm-readobj/llvm-readobj.cpp	2016-07-14 04:24:01.000000000 +0200
+++ b/tools/llvm-readobj/llvm-readobj.cpp	2016-10-06 14:32:20.934773864 +0200
@@ -57,10 +57,10 @@
     cl::desc("Alias for --file-headers"),
     cl::aliasopt(FileHeaders));
 
-  // -sections, -s
+  // -sections, -S
   cl::opt<bool> Sections("sections",
     cl::desc("Display all sections."));
-  cl::alias SectionsShort("s",
+  cl::alias SectionsShort("S",
     cl::desc("Alias for --sections"),
     cl::aliasopt(Sections));
 
@@ -96,10 +96,13 @@
   cl::opt<bool> DynRelocs("dyn-relocations",
     cl::desc("Display the dynamic relocation entries in the file"));
 
-  // -symbols, -t
+  // -symbols, -s
   cl::opt<bool> Symbols("symbols",
     cl::desc("Display the symbol table"));
-  cl::alias SymbolsShort("t",
+  cl::alias SymbolsShort("s",
+    cl::desc("Alias for --symbols"),
+    cl::aliasopt(Symbols));
+  cl::alias SymbolsShorter("syms",
     cl::desc("Alias for --symbols"),
     cl::aliasopt(Symbols));
 
@@ -117,9 +117,12 @@
     cl::desc("Alias for --unwind"),
     cl::aliasopt(UnwindInfo));
 
-  // -dynamic-table
+  // -dynamic-table, -d
   cl::opt<bool> DynamicTable("dynamic-table",
     cl::desc("Display the ELF .dynamic section table"));
+  cl::alias DynamicTableShort("d",
+    cl::desc("Alias for -dynamic-table"),
+    cl::aliasopt(DynamicTable));
 
   // -needed-libs
   cl::opt<bool> NeededLibraries("needed-libs",
