--- a/tools/clang/lib/Driver/Tools.cpp	2016-08-11 16:56:11.686754497 +0200
+++ b/tools/clang/lib/Driver/Tools.cpp	2016-08-11 16:55:24.320755817 +0200
@@ -9362,6 +9362,7 @@
   const bool HasCRTBeginEndFiles =
       ToolChain.getTriple().hasEnvironment() ||
       (ToolChain.getTriple().getVendor() != llvm::Triple::MipsTechnologies);
+  const bool isMusl = ToolChain.getTriple().getEnvironment() == llvm::Triple::Musl;
 
   ArgStringList CmdArgs;
 
@@ -9461,7 +9462,7 @@
       else
         crtbegin = isAndroid ? "crtbegin_dynamic.o" : "crtbegin.o";
 
-      if (HasCRTBeginEndFiles)
+      if (HasCRTBeginEndFiles && !isMusl)
         CmdArgs.push_back(Args.MakeArgString(ToolChain.GetFilePath(crtbegin)));
     }
 
@@ -9577,7 +9578,7 @@
       else
         crtend = isAndroid ? "crtend_android.o" : "crtend.o";
 
-      if (HasCRTBeginEndFiles)
+      if (HasCRTBeginEndFiles && !isMusl)
         CmdArgs.push_back(Args.MakeArgString(ToolChain.GetFilePath(crtend)));
       if (!isAndroid)
         CmdArgs.push_back(Args.MakeArgString(ToolChain.GetFilePath("crtn.o")));
