--- tools/clang/lib/Basic/Targets.cpp	2016-05-24 21:36:54.000000000 +0200
+++ tools/clang/lib/Basic/Targets.cpp.new	2016-07-16 12:04:33.175430669 +0200
@@ -2625,7 +2625,6 @@
     setFeatureEnabledImpl(Features, "lzcnt", true);
     setFeatureEnabledImpl(Features, "bmi", true);
     setFeatureEnabledImpl(Features, "bmi2", true);
-    setFeatureEnabledImpl(Features, "rtm", true);
     setFeatureEnabledImpl(Features, "fma", true);
     // FALLTHROUGH
   case CK_IvyBridge:
