--- a/tools/clang/lib/Driver/Types.cpp        2014-12-21 08:00:00.000000000 +0100
+++ b/tools/clang/lib/Driver/Types.cpp.new    2015-02-28 16:06:09.089090828 +0100
@@ -156,8 +156,6 @@
            .Case("hh", TY_CXXHeader)
            .Case("ll", TY_LLVM_IR)
            .Case("hpp", TY_CXXHeader)
-           .Case("ads", TY_Ada)
-           .Case("adb", TY_Ada)
            .Case("ast", TY_AST)
            .Case("c++", TY_CXX)
            .Case("C++", TY_CXX)

