--- cmake/modules/AddLLVM.cmake.bak	2016-01-29 01:22:31.684720741 +0100
+++ cmake/modules/AddLLVM.cmake	2016-01-29 01:22:59.559667505 +0100
@@ -468,6 +468,11 @@
     endif()
   endif()
 
+  set_target_properties(${name}
+      PROPERTIES
+      SOVERSION ${LLVM_VERSION_MAJOR}.${LLVM_VERSION_MINOR}
+      VERSION ${LLVM_VERSION_MAJOR}.${LLVM_VERSION_MINOR}.${LLVM_VERSION_PATCH}${LLVM_VERSION_SUFFIX})
+
   # Add the explicit dependency information for this library.
   #
   # It would be nice to verify that we have the dependencies for this library
