--- a/tools/clang/lib/Frontend/CompilerInvocation.cpp	2016-08-05 19:53:52.354985818 +0200
+++ b/tools/clang/lib/Frontend/CompilerInvocation.cpp	2016-08-05 19:54:19.203983608 +0200
@@ -1513,7 +1513,7 @@
     case IK_PreprocessedCXX:
     case IK_ObjCXX:
     case IK_PreprocessedObjCXX:
-      LangStd = LangStandard::lang_gnucxx98;
+      LangStd = LangStandard::lang_gnucxx14;
       break;
     case IK_RenderScript:
       LangStd = LangStandard::lang_c99;
