--- a/src/main/printutils.c	2021-07-14 21:28:25.591997860 +0200
+++ b/src/main/printutils.c	2021-07-14 21:28:53.658276401 +0200
@@ -752,7 +752,7 @@
 			for(int j = 0; j < res; j++) *q++ = *p++;
 		    } else {
 # if !defined (__STDC_ISO_10646__) && !defined (Win32)
-			if(!use_ucs) Unicode_warning = TRUE;
+			Unicode_warning = TRUE;
 # endif
 			if(k > 0xffff)
 			    snprintf(buf, 13, "\\U{%06x}", k);
