--- a/src/framework/mlt_property.h	2018-03-06 19:23:06.250095105 +0100
+++ b/src/framework/mlt_property.h	2018-03-06 19:23:59.839982519 +0100
@@ -25,18 +25,7 @@
 
 #include "mlt_types.h"
 
-#if defined(__FreeBSD__)
-/* This header has existed since 1994 and defines __FreeBSD_version below. */
-#include <sys/param.h>
-#endif
-
-#if defined(__GLIBC__) && !defined(__APPLE__)
-#  include <locale.h>
-#elif defined(__APPLE__) || (__FreeBSD_version >= 900506)
-#  include <xlocale.h>
-#else
-typedef char* locale_t;
-#endif
+#include <locale.h>
 
 extern mlt_property mlt_property_init( );
 extern int mlt_property_set_int( mlt_property self, int value );
