--- a/Source/JavaScriptCore/runtime/VMEntryScope.cpp	2017-01-27 15:39:31.656848341 +0100
+++ b/Source/JavaScriptCore/runtime/VMEntryScope.cpp	2017-01-27 15:39:45.464808121 +0100
@@ -84,8 +84,8 @@
     //
     // These sizes were derived from the stack usage of a number of sites when
     // layout occurs when we've already consumed most of the C stack.
-    const size_t requiredStack = 128 * KB;
-    const size_t errorModeRequiredStack = 64 * KB;
+    const size_t requiredStack = 32 * KB;
+    const size_t errorModeRequiredStack = 16 * KB;
 
     size_t requiredCapacity = interpreter->isInErrorHandlingMode() ? errorModeRequiredStack : requiredStack;
     RELEASE_ASSERT(m_stack.size() >= requiredCapacity);
