--- a/configure	2016-10-09 19:27:23.860483448 +0200
+++ b/configure	2016-10-09 19:27:47.587484616 +0200
@@ -9100,7 +9100,7 @@
   fp_cv_prog_ar_args="q"
 else
   touch conftest.dummy
-  for fp_var in clqsZ clqs cqs clq cq ; do
+  for fp_var in clqs cqs clq cq ; do
      rm -f conftest.a
      if "$fp_prog_ar" $fp_var conftest.a conftest.dummy > /dev/null 2> /dev/null; then
         fp_cv_prog_ar_args=$fp_var
--- a/aclocal.m4        2016-10-09 19:58:17.697574721 +0200
+++ b/aclocal.m4        2016-10-09 19:58:26.972575177 +0200
@@ -1123,7 +1123,7 @@
   fp_cv_prog_ar_args="q"
 else
   touch conftest.dummy
-  for fp_var in clqsZ clqs cqs clq cq ; do
+  for fp_var in clqs cqs clq cq ; do
      rm -f conftest.a
      if "$fp_prog_ar" $fp_var conftest.a conftest.dummy > /dev/null 2> /dev/null; then
         fp_cv_prog_ar_args=$fp_var
