--- a/compiler/main/SysTools.hs	2017-10-03 00:06:30.000000000 +0200
+++ b/compiler/main/SysTools.hs	2017-11-25 13:32:17.627055626 +0100
@@ -796,6 +796,10 @@
           -- ELF specific flag, see Note [ELF needed shared libs]
           return (GnuGold [Option "-Wl,--no-as-needed"])
 
+        | any ("LLD" `isPrefixOf`) stdo =
+          -- lld is just like GNU gold but better.
+          return (GnuGold [Option "-Wl,--no-as-needed"])
+
          -- Unknown linker.
         | otherwise = fail "invalid --version output, or linker is unsupported"
 
