--- a/src/backends/julia/juliaserver/juliaserver.cpp	2017-07-15 14:32:28.362341348 +0200
+++ b/src/backends/julia/juliaserver/juliaserver.cpp	2017-07-15 14:32:47.910339739 +0200
@@ -37,10 +37,9 @@
     jl_atexit_hook(0);
 }
 
-void JuliaServer::login(const QString &path) const
+void JuliaServer::login() const
 {
-    QString dir_path = QFileInfo(path).dir().absolutePath();
-    jl_init(dir_path.toLatin1().constData());
+    jl_init();
 }
 
 void JuliaServer::runJuliaCommand(const QString &command)
--- a/src/backends/julia/juliaserver/juliaserver.h	2017-07-15 14:32:58.511338867 +0200
+++ b/src/backends/julia/juliaserver/juliaserver.h	2017-07-15 14:33:35.607335815 +0200
@@ -40,10 +40,8 @@
 public Q_SLOTS:
     /**
      * Initializer for JuliaServer. Call this first before using it
-     *
-     * @param path path to julia executable
      */
-    Q_SCRIPTABLE void login(const QString &path) const;
+    Q_SCRIPTABLE void login() const;
 
     /**
      * Runs a piece of julia code. After this returns use getOutput, getError,
