--- a/Cargo.toml	2022-02-11 11:20:59.000000000 +0100
+++ b/Cargo.toml	2022-02-22 17:51:35.562908429 +0100
@@ -10,7 +10,7 @@
 argon2 = {version = "0.3", features = ["parallel"]}
 napi = {version = "2", default-features = false, features = ["napi3"]}
 napi-derive = {version = "2", default-features = false, features = ["type-def"]}
-rand = {git = "https://github.com/rust-random/rand", version = "0.8", rev = "a407bdf", features = ["nightly", "simd_support"]}
+rand = {version = "0.8.5", features = ["nightly", "simd_support"]}
 
 [build-dependencies]
 napi-build = "1"
