--- a/bin/estrip	2018-04-17 11:28:43.923258400 +0200
+++ b/bin/estrip	2018-04-17 11:29:37.017160823 +0200
@@ -137,6 +137,13 @@
 	DEF_STRIP_FLAGS="-R .comment -R .GCC.command.line -R .note.gnu.gold-version"
 	SPLIT_STRIP_FLAGS=
 	;;
+*) # sys-devel/llvm
+	# We'll leave out -R .note for now until we can check out the relevance
+	# of the section when it has the ALLOC flag set on it ...
+	SAFE_STRIP_FLAGS=""
+	DEF_STRIP_FLAGS="-S -R .comment -R .GCC.command.line -R .note.gnu.gold-version"
+	SPLIT_STRIP_FLAGS=
+	;;
 esac
 : ${PORTAGE_STRIP_FLAGS=${SAFE_STRIP_FLAGS} ${DEF_STRIP_FLAGS}}
 
@@ -416,18 +423,7 @@
 	# elf types though cause that stuff is good.
 
 	buildid=
-	if [[ ${f} == *"current ar archive"* ]] ; then
-		__vecho "   ${x:${#ED}}"
-		if ${strip_this} ; then
-			# If we have split debug enabled, then do not strip this.
-			# There is no concept of splitdebug for objects not yet
-			# linked in (only for finally linked ELFs), so we have to
-			# retain the debug info in the archive itself.
-			if ! ${FEATURES_splitdebug} || ${RESTRICT_splitdebug} ; then
-				${STRIP} -g "${x}"
-			fi
-		fi
-	elif [[ ${f} == *"SB executable"* || ${f} == *"SB pie executable"* ||
+	if [[ ${f} == *"SB executable"* || ${f} == *"SB pie executable"* ||
 		${f} == *"SB shared object"* ]] ; then
 		process_elf "${x}" "${inode_link}" ${PORTAGE_STRIP_FLAGS}
 	elif [[ ${f} == *"SB relocatable"* ]] ; then
