--- a/runtime/druntime/src/ldc/eh/libunwind.d	2016-09-05 21:14:32.124681185 +0200
+++ b/runtime/druntime/src/ldc/eh/libunwind.d	2016-09-05 21:15:06.173694166 +0200
@@ -66,7 +66,7 @@
 
     alias void function(_Unwind_Reason_Code, _Unwind_Exception*) _Unwind_Exception_Cleanup_Fn;
 
-    struct _Unwind_Exception
+    align(16) struct _Unwind_Exception
     {
         ulong exception_class;
         _Unwind_Exception_Cleanup_Fn exception_cleanup;

