--- a/gen/nested.cpp	2022-09-16 15:59:41.521259112 +0200
+++ b/gen/nested.cpp	2022-09-16 16:00:48.969230588 +0200
@@ -228,7 +228,7 @@
     unsigned idx = getVthisIdx(decl);
     llvm::StructType *st = getIrAggr(decl, true)->getLLStructType();
     LLValue *gep = DtoGEP(st, value, 0, idx, ".vthis");
-    DtoStore(DtoBitCast(nest, gep->getType()->getContainedType(0)), gep);
+    DtoStore(DtoBitCast(nest, gep->getType()->getScalarType()), gep);
   }
 }
 
