--- a/driver/linker-gcc.cpp
+++ b/driver/linker-gcc.cpp
@@ -250,6 +250,7 @@
       break;
     }
     args.push_back("-lrt");
+    args.push_back("-lunwind");
   // fallthrough
   case llvm::Triple::Darwin:
   case llvm::Triple::MacOSX:
--- a/runtime/CMakeLists.txt
+++ b/runtime/CMakeLists.txt
@@ -84,7 +84,7 @@
 # Auto-detect C system libraries
 if("${C_SYSTEM_LIBS}" STREQUAL "AUTO")
     if("${TARGET_SYSTEM}" MATCHES "Linux")
-        set(C_SYSTEM_LIBS m pthread rt dl)
+        set(C_SYSTEM_LIBS m pthread rt dl unwind)
     else()
         set(C_SYSTEM_LIBS m pthread)
     endif()
