--- a/driver/linker.cpp	2016-08-11 20:00:04.168447085 +0200
+++ b/driver/linker.cpp	2016-08-11 20:00:47.892445867 +0200
@@ -250,6 +250,7 @@
       break;
     }
     args.push_back("-lrt");
+    args.push_back("-lunwind");
   // fallthrough
   case llvm::Triple::Darwin:
   case llvm::Triple::MacOSX:
--- a/runtime/CMakeLists.txt	2016-08-11 20:01:08.768445285 +0200
+++ b/runtime/CMakeLists.txt	2016-08-11 20:03:21.143441596 +0200
@@ -425,7 +425,7 @@
 
     if(BUILD_SHARED_LIBS)
         if(${CMAKE_SYSTEM} MATCHES "Linux")
-            set(dso_system_libs "m" "pthread" "rt" "dl")
+            set(dso_system_libs "unwind")
         else()
             set(dso_system_libs "m" "pthread")
         endif()
