--- a/driver/ldmd.cpp
+++ b/driver/ldmd.cpp
@@ -108,8 +108,8 @@ char *concat(const char *a, int b) {
 int execute(const std::string &exePath, const char **args) {
 #if LDC_LLVM_VER >= 700
   std::vector<llvm::StringRef> argv;
-  for (auto arg = args; arg != nullptr; ++arg) {
-    argv.push_back(*arg);
+  for (auto pArg = args; *pArg; ++pArg) {
+    argv.push_back(*pArg);
   }
   auto envVars = llvm::None;
 #else
--- a/gen/uda.cpp
+++ b/gen/uda.cpp
@@ -20,7 +20,9 @@
 // Auto-generate:
 // Attribute::AttrKind getAttrKindFromName(StringRef AttrName) { ... }
 #define GET_ATTR_KIND_FROM_NAME
-#if LDC_LLVM_VER >= 400
+#if LDC_LLVM_VER >= 700
+#include "llvm/IR/Attributes.inc"
+#elif LDC_LLVM_VER >= 400
 #include "llvm/IR/Attributes.gen"
 #else
 #include "llvm/IR/Attributes.inc"

