--- a/CMakeLists.txt	2022-11-04 21:43:08.399206121 +0100
+++ b/CMakeLists.txt	2022-11-04 21:43:27.312431746 +0100
@@ -606,8 +606,6 @@
     if(NOT (LDC_LLVM_VER LESS 1000))
         if(MSVC)
             list(APPEND LDC_LINKERFLAG_LIST LLVMSymbolize.lib)
-        else()
-            set(LDC_LINKERFLAG_LIST -lLLVMSymbolize ${LDC_LINKERFLAG_LIST})
         endif()
     endif()
     if (LDC_LLVM_VER LESS 1200 OR NOT LDC_LLVM_VER LESS 1400)
