diff --git a/runtime/jit-rt/cpp-so/disassembler.cpp b/runtime/jit-rt/cpp-so/disassembler.cpp
index b0943da4..3a71d595 100644
--- a/runtime/jit-rt/cpp-so/disassembler.cpp
+++ b/runtime/jit-rt/cpp-so/disassembler.cpp
@@ -266,8 +266,8 @@ void disassemble(const llvm::TargetMachine &tm,
 
   llvm::MCTargetOptions opts;
   auto mab = unique(target.createMCAsmBackend(
-#if LDC_LLVM_VER >= 700
-      *sti,*mri,opts)
+#if LDC_LLVM_VER >= 600
+      *sti, *mri, opts)
 #else
       *mri, tm.getTargetTriple().getTriple(), tm.getTargetCPU(), opts)
 #endif
