--- a/runtime/druntime/src/core/sys/posix/net/if_.d	2024-01-14 20:04:20.161689866 +0100
+++ b/runtime/druntime/src/core/sys/posix/net/if_.d	2024-01-14 20:05:54.856825644 +0100
@@ -158,3 +158,18 @@
     if_nameindex_t* if_nameindex();
     void            if_freenameindex(if_nameindex_t*);
 }
+else version (CRuntime_Musl)
+{
+    struct if_nameindex_t
+    {
+        uint    if_index;
+        char*   if_name;
+    }
+
+    enum IF_NAMESIZE = 16;
+
+    uint            if_nametoindex(const scope char*);
+    char*           if_indextoname(uint, char*);
+    if_nameindex_t* if_nameindex();
+    void            if_freenameindex(if_nameindex_t*);
+}
