diff --git a/driver/linker.cpp b/driver/linker.cpp
index 496fd877..faabfba5 100644
--- a/driver/linker.cpp
+++ b/driver/linker.cpp
@@ -335,6 +335,12 @@ static int linkObjToBinaryGcc(bool sharedLib, bool fullyStatic) {
   // create path to exe
   CreateDirectoryOnDisk(gExePath);
 
+  // we strip by default, we do not want this when debugging
+  // so pass -nostrip when debugging
+  if (global.params.symdebug) {
+    args.push_back("-Wl,-nostrip");
+  }
+
   // Pass sanitizer arguments to linker. Requires clang.
   if (opts::sanitize == opts::AddressSanitizer) {
     args.push_back("-fsanitize=address");
diff --git a/runtime/CMakeLists.txt b/runtime/CMakeLists.txt
index d4c6838e..304448e1 100644
--- a/runtime/CMakeLists.txt
+++ b/runtime/CMakeLists.txt
@@ -499,7 +499,7 @@ macro(build_runtime_variants d_flags c_flags ld_flags path_suffix outlist_target
     build_runtime(
         "${d_flags};${D_FLAGS};${D_FLAGS_DEBUG}"
         "${c_flags}"
-        "${ld_flags}"
+        "${ld_flags} -Wl,-nostrip"
         "-debug"
         "${path_suffix}"
         ${outlist_targets}
