diff --git a/driver/linker.cpp b/driver/linker.cpp
index 496fd877..faabfba5 100644
--- a/driver/linker-gcc.cpp
+++ b/driver/linker-gcc.cpp
diff --git a/driver/linker-gcc.cpp b/driver/linker-gcc.cpp
index 0d7a8e33..2445f772 100644
--- a/driver/linker-gcc.cpp
+++ b/driver/linker-gcc.cpp
@@ -209,6 +209,13 @@ void ArgsBuilder::build(llvm::StringRef outputPath,
 
   addSanitizers();
 
+  // we strip by default, we do not want this when debugging
+  // so pass -nostrip when debugging
+  if (global.params.symdebug) {
+    args.push_back("-Wl,-nostrip");
+  }
+
+
 #if LDC_LLVM_VER >= 309
   // Add LTO link flags before adding the user link switches, such that the user
   // can pass additional options to the LTO plugin.
diff --git a/runtime/CMakeLists.txt b/runtime/CMakeLists.txt
index d4c6838e..304448e1 100644
--- a/runtime/CMakeLists.txt
+++ b/runtime/CMakeLists.txt
@@ -499,7 +499,7 @@ macro(build_runtime_variants d_flags c_flags ld_flags path_suffix outlist_target
     build_runtime(
         "${d_flags};${D_FLAGS};${D_FLAGS_DEBUG}"
         "${c_flags}"
-        "${ld_flags}"
+        "${ld_flags} -Wl,-nostrip"
         "-debug"
         "${path_suffix}"
         ${outlist_targets}
