--- a/runtime/druntime/src/rt/qsort.d	2016-10-23 22:57:29.099892120 +0200
+++ b/runtime/druntime/src/rt/qsort.d	2016-10-23 22:57:36.474891914 +0200
@@ -18,21 +18,6 @@
 
 private import core.stdc.stdlib;
 
-version (CRuntime_Glibc)
-{
-    alias extern (C) int function(const void *, const void *, void *) Cmp;
-    extern (C) void qsort_r(void *base, size_t nmemb, size_t size, Cmp cmp, void *arg);
-
-    extern (C) void[] _adSort(void[] a, TypeInfo ti)
-    {
-        extern (C) int cmp(in void* p1, in void* p2, void* ti)
-        {
-            return (cast(TypeInfo)ti).compare(p1, p2);
-        }
-        qsort_r(a.ptr, a.length, ti.tsize, &cmp, cast(void*)ti);
-        return a;
-    }
-}
-else version (FreeBSD)
+version (FreeBSD)
 {
     alias extern (C) int function(void *, const void *, const void *) Cmp;
     extern (C) void qsort_r(void *base, size_t nmemb, size_t size, void *thunk, Cmp cmp);
