--- a/runtime/druntime/src/rt/backtrace/dwarf.d
+++ b/runtime/druntime/src/rt/backtrace/dwarf.d
@@ -12,8 +12,7 @@
 
 module rt.backtrace.dwarf;
 
-version(CRuntime_Glibc) version = linux_or_freebsd;
-else version(FreeBSD) version = linux_or_freebsd;
+version(FreeBSD) version = linux_or_freebsd;
 
 version(linux_or_freebsd):
 
--- a/runtime/druntime/src/core/runtime.d
+++ b/runtime/druntime/src/core/runtime.d
@@ -401,9 +401,7 @@
 extern (C) bool runModuleUnitTests()
 {
     // backtrace
-    version( CRuntime_Glibc )
-        import core.sys.linux.execinfo;
-    else version( Darwin )
+    version( Darwin )
         import core.sys.darwin.execinfo;
     else version( FreeBSD )
         import core.sys.freebsd.execinfo;
@@ -481,9 +481,7 @@
 Throwable.TraceInfo defaultTraceHandler( void* ptr = null )
 {
     // backtrace
-    version( CRuntime_Glibc )
-        import core.sys.linux.execinfo;
-    else version( Darwin )
+    version( Darwin )
         import core.sys.darwin.execinfo : backtrace_symbols;
     else version( FreeBSD )
         import core.sys.freebsd.execinfo;
