--- a/src/plugins/windowsystem/windowsystem.h	2026-03-03 10:32:10.000000000 +0100
+++ b/src/plugins/windowsystem/windowsystem.h	2026-03-19 20:34:12.743745426 +0100
@@ -12,9 +12,8 @@
 namespace KWin
 {
 
-class WindowSystem : public QObject, public KWindowSystemPrivateV3
+class WindowSystem : public KWindowSystemPrivateV3
 {
-    Q_OBJECT
 public:
     WindowSystem();
     void activateWindow(QWindow *win, long time) override;
