--- a/src/inputmethod.cpp	2025-12-09 10:50:20.000000000 +0100
+++ b/src/inputmethod.cpp	2025-12-26 14:49:09.751545164 +0100
@@ -313,7 +313,7 @@
 {
     Q_ASSERT(panel->isInputMethod());
     if (m_panel) {
-        qCWarning(KWIN_VIRTUALKEYBOARD) << "Replacing input panel" << m_panel << "with" << panel;
+        qCDebug(KWIN_VIRTUALKEYBOARD) << "Replacing input panel" << m_panel << "with" << panel;
         m_panel->destroyWindow();
     }
 
