--- a/configure.ac	2019-11-03 14:31:15.655207071 +0100
+++ b/configure.ac	2019-11-03 14:33:59.536786696 +0100
@@ -830,7 +830,7 @@
 AS_IF([test "x$HAVE_ALSA" = "x1"],
     [
         save_CPPFLAGS="$CPPFLAGS"; CPPFLAGS="$CPPFLAGS $ASOUNDLIB_CFLAGS"
-        AC_CHECK_HEADERS([use-case.h], HAVE_ALSA_UCM=1, HAVE_ALSA_UCM=0)
+        AC_CHECK_HEADERS([alsa/use-case.h], HAVE_ALSA_UCM=1, HAVE_ALSA_UCM=0)
         CPPFLAGS="$save_CPPFLAGS"
     ],
     HAVE_ALSA_UCM=0)
--- a/src/modules/alsa/alsa-ucm.h	2019-11-03 14:32:33.122481271 +0100
+++ b/src/modules/alsa/alsa-ucm.h	2019-11-03 14:32:45.562525269 +0100
@@ -23,7 +23,7 @@
 ***/
 
 #ifdef HAVE_ALSA_UCM
-#include <use-case.h>
+#include <alsa/use-case.h>
 #else
 typedef void snd_use_case_mgr_t;
 #endif
