--- a/lib/freebl/Makefile	2016-10-06 12:13:08.382362631 +0200
+++ b/lib/freebl/Makefile	2016-10-06 12:13:53.407364848 +0200
@@ -221,10 +221,12 @@
     DEFINES += -DMP_IS_LITTLE_ENDIAN
 #   DEFINES += -DMPI_AMD64_ADD
     # comment the next four lines to turn off Intel HW acceleration.
+ifndef CC_IS_CLANG
     DEFINES += -DUSE_HW_AES -DINTEL_GCM
     ASFILES += intel-aes.s intel-gcm.s
     EXTRA_SRCS += intel-gcm-wrap.c
     INTEL_GCM = 1
+endif
     MPI_SRCS += mpi_amd64.c mp_comba.c
 endif
 ifeq ($(CPU_ARCH),x86)
@@ -462,10 +462,12 @@
 	    ASFLAGS += -march=opteron -m64 -fPIC
 	    MPI_SRCS += mp_comba.c
 	    # comment the next four lines to turn off Intel HW acceleration
+ifndef CC_IS_CLANG
 	    ASFILES += intel-gcm.s
 	    EXTRA_SRCS += intel-gcm-wrap.c
 	    INTEL_GCM = 1
 	    DEFINES += -DINTEL_GCM
+endif
 	else
 	    ASFILES  = arcfour-amd64-sun.s mpi_amd64_sun.s sha-fast-amd64-sun.s
  	    ASFILES += mp_comba_amd64_sun.s mpcpucache_amd64.s
