--- a/nss/lib/freebl/Makefile	2019-09-13 12:14:22.194383238 +0200
+++ b/nss/lib/freebl/Makefile	2019-09-13 12:03:32.035968011 +0200
@@ -749,9 +749,9 @@
 # symbolic names to registers, for example,
 #     .set  Htbl, %rdi
 # So we can't use Clang's integrated assembler with intel-gcm.s.
-ifdef CC_IS_CLANG
-$(OBJDIR)/$(PROG_PREFIX)intel-gcm$(OBJ_SUFFIX): CFLAGS += -no-integrated-as
-endif
+#ifdef CC_IS_CLANG
+#$(OBJDIR)/$(PROG_PREFIX)intel-gcm$(OBJ_SUFFIX): CFLAGS += -no-integrated-as
+#endif
 endif
 
 ifdef INTEL_GCM_CLANG_CL
