--- a/scripts/Kbuild.include	2016-12-27 16:13:38.304005103 +0100
+++ b/scripts/Kbuild.include	2016-12-27 16:13:49.265972467 +0100
@@ -51,26 +51,26 @@
 # as-option
 # Usage: cflags-y += $(call as-option, -Wa$(comma)-isa=foo,)
 
-as-option = $(shell if $(CC) $(CFLAGS) $(1) -Wa,-Z -c -o /dev/null \
-	     -xassembler /dev/null > /dev/null 2>&1; then echo "$(1)"; \
+as-option = $(shell if $(CC) $(CFLAGS) $(1) -Wa,-Z -c -o .test \
+	     -xassembler .test > .test 2>&1; then echo "$(1)"; \
 	     else echo "$(2)"; fi ;)
 
 # cc-option
 # Usage: cflags-y += $(call cc-option, -march=winchip-c6, -march=i586)
 
-cc-option = $(shell if $(CC) $(CFLAGS) $(1) -S -o /dev/null -xc /dev/null \
-             > /dev/null 2>&1; then echo "$(1)"; else echo "$(2)"; fi ;)
+cc-option = $(shell if $(CC) $(CFLAGS) $(1) -S -o .test -xc .test \
+             > .test 2>&1; then echo "$(1)"; else echo "$(2)"; fi ;)
 
 # hostcc-option
 # Usage: hostcflags-y += $(call hostcc-option, -march=winchip-c6, -march=i586)
 
-hostcc-option = $(shell if $(HOSTCC) $(HOSTCFLAGS) $(1) -S -o /dev/null -xc /dev/null \
-             > /dev/null 2>&1; then echo "$(1)"; else echo "$(2)"; fi ;)
+hostcc-option = $(shell if $(HOSTCC) $(HOSTCFLAGS) $(1) -S -o .test -xc .test \
+             > .test 2>&1; then echo "$(1)"; else echo "$(2)"; fi ;)
 
 # cc-option-yn
 # Usage: flag := $(call cc-option-yn, -march=winchip-c6)
-cc-option-yn = $(shell if $(CC) $(CFLAGS) $(1) -S -o /dev/null -xc /dev/null \
-                > /dev/null 2>&1; then echo "y"; else echo "n"; fi;)
+cc-option-yn = $(shell if $(CC) $(CFLAGS) $(1) -S -o .test -xc .test \
+                > .test 2>&1; then echo "y"; else echo "n"; fi;)
 
 # cc-option-align
 # Prefix align with either -falign or -malign
--- a/Makefile.flags	2016-12-27 16:17:30.377331740 +0100
+++ b/Makefile.flags	2016-12-27 16:17:59.526249380 +0100
@@ -127,7 +127,7 @@
 # gcc-4.2.1 fails if we try to feed C source on stdin:
 #  echo 'int main(void){return 0;}' | $(CC) $(CFLAGS) -lcrypt -o /dev/null -xc -
 # fall back to using a temp file:
-CRYPT_AVAILABLE := $(shell echo 'int main(void){return 0;}' >crypttest.c; $(CC) $(CFLAGS) -lcrypt -o /dev/null crypttest.c >/dev/null 2>&1 && echo "y"; rm crypttest.c)
+CRYPT_AVAILABLE := $(shell echo 'int main(void){return 0;}' >crypttest.c; $(CC) $(CFLAGS) -lcrypt -o cryptotest.o crypttest.c >/dev/null 2>&1 && echo "y"; rm crypttest.c)
 ifeq ($(CRYPT_AVAILABLE),y)
 LDLIBS += m crypt
 else
