--- third_party/WebKit/Source/platform/heap/GCInfo.cpp.old	2015-10-16 14:01:38.151359760 +0200
+++ third_party/WebKit/Source/platform/heap/GCInfo.cpp	2015-10-16 14:02:02.725270920 +0200
@@ -5,6 +5,7 @@
 #include "config.h"
 #include "platform/heap/GCInfo.h"
 
+#include <malloc.h>
 #include "platform/heap/Handle.h"
 #include "platform/heap/Heap.h"
 
--- third_party/ots/include/opentype-sanitiser.hbak	2015-10-16 13:58:40.026003858 +0200
+++ third_party/ots/include/opentype-sanitiser.h	2015-10-16 13:59:10.049895272 +0200
@@ -22,6 +22,7 @@
 #else
 #include <arpa/inet.h>
 #include <stdint.h>
+#include <sys/types.h>
 #endif
 
 #include <algorithm>
