--- a/libkwave/MultiTrackSink.h
+++ b/libkwave/MultiTrackSink.h
@@ -32,7 +32,7 @@
 
     template <class SINK, const bool INITIALIZE>
     class MultiTrackSink: public Kwave::SampleSink,
-                          private QList<SINK *>
+                          public QList<SINK *>
     {
     public:
 	/**
--- a/libkwave/MultiTrackSource.h
+++ b/libkwave/MultiTrackSource.h
@@ -38,7 +38,7 @@
      */
     template <class SOURCE, const bool INITIALIZE>
     class Q_DECL_EXPORT MultiTrackSource: public Kwave::SampleSource,
-	                                  private QList<SOURCE *>
+	                                  public QList<SOURCE *>
     {
     public:
 	/**
