diff --git a/CMakeLists.txt b/CMakeLists.txt
index ee28c68230..98c2a447b1 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -28,7 +28,7 @@ include(ECMSetupVersion)
 include(ECMAddTests)
 include(ECMAddAppIcon)
 
-find_package( Qt5 REQUIRED COMPONENTS Core DBus Widgets Test Gui Svg Xml Script Sql Quick Qml ScriptTools WebKitWidgets)
+find_package( Qt5 REQUIRED COMPONENTS Core DBus Widgets Test Gui Svg Xml Script Sql Quick Qml ScriptTools)
 find_package( KF5 REQUIRED COMPONENTS CoreAddons Solid KDELibs4Support Codecs I18n ThreadWeaver Plasma PlasmaQuick KCMUtils NewStuff TextEditor GlobalAccel KIO NotifyConfig Archive DNSSD)
 
 ###############
