--- a/Makefile	2017-11-12 10:18:43.440370161 +0100
+++ b/Makefile	2017-11-12 10:18:57.655341657 +0100
@@ -206,8 +206,8 @@
 AUTOMAKE_OPTIONS = 1.6
 
 #Fools KDevelop into including all files
-SUBDIRS = libcelt tests tools
-DIST_SUBDIRS = libcelt tests tools
+SUBDIRS = libcelt tools
+DIST_SUBDIRS = libcelt tools
 pkgconfigdir = $(libdir)/pkgconfig
 pkgconfig_DATA = celt051.pc 
 EXTRA_DIST = celt051.pc.in Doxyfile Doxyfile.devel
