--- a/CMakeLists.txt	2022-02-03 21:09:38.045461416 +0100
+++ b/CMakeLists.txt	2022-02-03 21:10:49.716415200 +0100
@@ -131,8 +131,9 @@
 set_package_properties(KF5KDcraw PROPERTIES URL "https://phabricator.kde.org/source/libkdcraw/" DESCRIPTION "C++ interface around LibRaw library used to decode RAW picture files" TYPE OPTIONAL)
 
 
-find_package(X11)
-if(X11_FOUND)
+option(WITH_X11 "Build with X11 integration" ON)
+if (WITH_X11)
+    find_package(X11 REQUIRED)
     if (QT_MAJOR_VERSION STREQUAL "5")
         find_package(Qt5 ${QT_MIN_VERSION} CONFIG REQUIRED X11Extras)
     endif()
