--- a/src/GoZipTool/GoZipTool.csproj	2025-09-13 20:47:35.757551502 +0200
+++ b/src/GoZipTool/GoZipTool.csproj	2025-09-13 20:48:41.388892517 +0200
@@ -41,6 +41,9 @@
       <GoOS Condition="'$(ResolvedRuntimeIdentifier)' == 'linux-x64'">linux</GoOS>
       <GoArch Condition="'$(ResolvedRuntimeIdentifier)' == 'linux-x64'">amd64</GoArch>
 
+      <GoOS Condition="'$(ResolvedRuntimeIdentifier)' == 'linux-musl-arm64'">linux</GoOS>
+      <GoArch Condition="'$(ResolvedRuntimeIdentifier)' == 'linux-musl-arm64'">arm64</GoArch>
+
       <GoOS Condition="'$(ResolvedRuntimeIdentifier)' == 'osx-x64'">darwin</GoOS>
       <GoArch Condition="'$(ResolvedRuntimeIdentifier)' == 'osx-x64'">amd64</GoArch>
 
