--- a/src/Azure.Functions.Cli/Helpers/PythonHelpers.cs	2025-02-18 12:25:24.346036451 +0100
+++ b/src/Azure.Functions.Cli/Helpers/PythonHelpers.cs	2025-02-18 12:27:57.492343683 +0100
@@ -167,7 +167,7 @@
         {
             if (pythonVersion?.Version == null)
             {
-                var message = "Could not find a Python version. 3.7.x, 3.8.x, 3.9.x, 3.10.x, 3.11.x or 3.12.x is recommended, and used in Azure Functions.";
+                var message = "Could not find a Python version. 3.7.x, 3.8.x, 3.9.x, 3.10.x, 3.11.x, 3.12.x or 3.13.x is recommended, and used in Azure Functions.";
                 if (errorIfNoVersion) throw new CliException(message);
                 ColoredConsole.WriteLine(WarningColor(message));
                 return;
@@ -175,23 +175,23 @@
 
             ColoredConsole.WriteLine(AdditionalInfoColor($"Found Python version {pythonVersion.Version} ({pythonVersion.ExecutablePath})."));
 
-            // Python 3.[7|8|9|10|11|12] (supported)
+            // Python 3.[7|8|9|10|11|12|13] (supported)
             if (IsVersionSupported(pythonVersion))
             {
                 return;
             }
 
-            // Python 3.x (but not 3.[7|8|9|10|11|12]), not recommended, may fail. E.g.: 3.4, 3.5.
+            // Python 3.x (but not 3.[7|8|9|10|11|12|13]), not recommended, may fail. E.g.: 3.4, 3.5.
             if (pythonVersion.Major == 3)
             {
                 if (errorIfNotSupported)
-                    throw new CliException($"Python 3.7.x to 3.12.x is required for this operation. " +
-                        $"Please install Python 3.7, 3.8, 3.9, 3.10, 3.11 or 3.12 and use a virtual environment to switch to Python 3.7, 3.8, 3.9, 3.10, 3.11 or 3.12.");
-                ColoredConsole.WriteLine(WarningColor("Python  3.7.x, 3.8.x, 3.9.x, 3.10.x, 3.11.x or 3.12.x is recommended, and used in Azure Functions."));
+                    throw new CliException($"Python 3.7.x to 3.13.x is required for this operation. " +
+                        $"Please install Python 3.7, 3.8, 3.9, 3.10, 3.11, 3.12 or 3.13 and use a virtual environment to switch to Python 3.7, 3.8, 3.9, 3.10, 3.11, 3.12 or 3.13.");
+                ColoredConsole.WriteLine(WarningColor("Python  3.7.x, 3.8.x, 3.9.x, 3.10.x, 3.11.x, 3.12.x or 3.13.x is recommended, and used in Azure Functions."));
             }
 
             // No Python 3
-            var error = "Python 3.x (recommended version 3.[7|8|9|10|11|12]) is required.";
+            var error = "Python 3.x (recommended version 3.[7|8|9|10|11|12|13]) is required.";
             if (errorIfNoVersion) throw new CliException(error);
             ColoredConsole.WriteLine(WarningColor(error));
         }
@@ -226,6 +226,7 @@
             var python310GetVersionTask = GetVersion("python3.10");
             var python311GetVersionTask = GetVersion("python3.11");
             var python312GetVersionTask = GetVersion("python3.12");
+            var python313GetVersionTask = GetVersion("python3.13");
 
             var versions = new List<WorkerLanguageVersionInfo>
             {
@@ -239,6 +240,7 @@
                 await python39GetVersionTask,
                 await python310GetVersionTask,
                 await python311GetVersionTask,
-                await python312GetVersionTask
+                await python312GetVersionTask,
+                await python313GetVersionTask
             };
 
             // Highest preference -- Go through the list, if we find the first python 3.6 or python 3.7 worker, we prioritize that.
@@ -599,6 +601,7 @@
             {
                 switch (info?.Minor)
                 {
+                    case 13:
                     case 12:
                     case 11:
                     case 10:
