--- a/libtransmission/bitfield.c	2016-10-16 14:25:14.398440714 +0200
+++ b/libtransmission/bitfield.c	2016-10-16 14:25:43.448438323 +0200
@@ -13,6 +13,7 @@
 #include <assert.h>
 #include <stdlib.h> /* realloc () */
 #include <string.h> /* memset */
+#include <sys/types.h> /* ssize_t */
 
 #include "transmission.h"
 #include "bitfield.h"
--- a/libtransmission/fdlimit.h	2016-10-16 14:26:14.324435783 +0200
+++ b/libtransmission/fdlimit.h	2016-10-16 14:26:31.523434368 +0200
@@ -14,6 +14,8 @@
  #error only libtransmission should #include this header.
 #endif
 
+#include <sys/types.h>
+
 #include "transmission.h"
 #include "net.h"
 
