--- a/makefile	2016-08-27 12:19:10.040688148 +0200
+++ b/makefile	2016-08-27 12:21:37.289518208 +0200
@@ -149,7 +149,7 @@
 	${LDC} ${CLIENT_SRC} ${LDC_CLIENT_FLAGS}
 
 $(OBJ_DIR)/%.o: $(SERVER_SRC)
-	$(LDC) $*.d $(LDC_SERVER_FLAGS) -od=$(OBJ_DIR) -op -c
+	$(LDC) $*.d $(LDC_SERVER_FLAGS) -od=$(OBJ_DIR) -of=$*.o -op -c
 
 ldcserver: githash $(SERVER_OBJS)
 	${LDC} ${SERVER_OBJS} ${LDC_SERVER_FLAGS} -of=bin/dcd-server
