--- include/__config
+++ include/__config
@@ -314,14 +314,10 @@ typedef __char32_t char32_t;
 #elif defined(__linux__)
 #if !defined(_LIBCPP_HAS_MUSL_LIBC)
 # include <features.h>
-#if __GLIBC_PREREQ(2, 15)
 #define _LIBCPP_HAS_QUICK_EXIT
-#endif
-#if __GLIBC_PREREQ(2, 17)
 #define _LIBCPP_HAS_C11_FEATURES
-#endif
 #else // defined(_LIBCPP_HAS_MUSL_LIBC)
 #define _LIBCPP_HAS_QUICK_EXIT
 #define _LIBCPP_HAS_C11_FEATURES
