--- CMakeLists.txt.bal	2016-01-23 17:28:28.655496499 +0100
+++ CMakeLists.txt	2016-01-23 17:28:47.015253956 +0100
@@ -263,9 +263,9 @@
 remove_flags(-Wno-pedantic -pedantic-errors -pedantic)
 
 # Required flags ==============================================================
-add_compile_flags_if_supported(-std=c++11)
-if (NOT MSVC AND NOT LIBCXX_SUPPORTS_STD_EQ_CXX11_FLAG)
-  message(FATAL_ERROR "C++11 is required but the compiler does not support -std=c++11")
+add_compile_flags_if_supported(-std=c++14)
+if (NOT MSVC AND NOT LIBCXX_SUPPORTS_STD_EQ_CXX14_FLAG)
+  message(FATAL_ERROR "C++14 is required but the compiler does not support -std=c++14")
 endif()
 
 # On all systems the system c++ standard library headers need to be excluded.
