--- cmake/Modules/HandleOutOfTreeLLVM.cmake.bak	2016-01-23 17:21:29.246168520 +0100
+++ cmake/Modules/HandleOutOfTreeLLVM.cmake	2016-01-23 17:21:44.884815747 +0100
@@ -11,8 +11,7 @@
     message(STATUS "Found LLVM_CONFIG as ${LLVM_CONFIG}")
     set(CONFIG_COMMAND ${LLVM_CONFIG}
       "--includedir"
-      "--prefix"
-      "--src-root")
+      "--prefix")
     execute_process(
       COMMAND ${CONFIG_COMMAND}
       RESULT_VARIABLE HAD_ERROR
