--- a/src/CMakeLists.txt	2025-03-15 17:52:44.642190079 +0100
+++ b/src/CMakeLists.txt	2025-03-15 18:14:13.097499396 +0100
@@ -33,7 +33,6 @@
 kcoreaddons_add_plugin(admin SOURCES  kioadmin_debug.cpp worker.cpp dbustypes.cpp ${admin_SRCS} INSTALL_NAMESPACE "kf${QT_MAJOR_VERSION}/kio")
 target_link_libraries(admin
     PUBLIC KF${QT_MAJOR_VERSION}::KIOCore
-    PRIVATE atomic
     PolkitQt${QT_MAJOR_VERSION}-1::Core
     Qt::Core
     Qt::DBus)
