--- a/arch/arm64/Makefile	2019-03-14 17:41:01.096478563 +0100
+++ b/arch/arm64/Makefile	2019-03-14 17:41:16.725366330 +0100
@@ -50,7 +50,7 @@
 else
 KBUILD_CPPFLAGS	+= -mlittle-endian
 AS		+= -EL
-LD		+= -EL
+LD		+= -EL -maarch64elf
 endif
 
 CHECKFLAGS	+= -D__aarch64__
--- a/arch/arm64/kernel/vdso/Makefile	2019-03-14 17:40:22.537751210 +0100
+++ b/arch/arm64/kernel/vdso/Makefile	2019-03-14 17:40:34.133669869 +0100
@@ -54,7 +54,7 @@
 
 # Actual build commands
 quiet_cmd_vdsold = VDSOL   $@
-      cmd_vdsold = $(CC) $(c_flags) -Wl,-n -Wl,-T $^ -o $@
+      cmd_vdsold = $(CC) $(c_flags) -Wl,-T $^ -o $@
 quiet_cmd_vdsoas = VDSOA   $@
       cmd_vdsoas = $(CC) $(a_flags) -c -o $@ $<
 
--- a/drivers/usb/dwc_otg_310/Kconfig	2019-03-08 23:28:36.000000000 +0100
+++ b/drivers/usb/dwc_otg_310/Kconfig	2019-03-14 17:40:02.389891153 +0100
@@ -18,5 +18,5 @@
 
 config DWC_OTG_310
 	tristate
-	default y if USB20_HOST || USB20_OTG
+	default n if USB20_HOST || USB20_OTG
 endmenu
