--- a/idris-1.3.2/Setup.hs
+++ b/idris-1.3.2/Setup.hs
@@ -215,6 +215,7 @@
       autogenComponentModulesDir lbi _ = autogenModulesDir lbi
 #endif
 
+#if !(MIN_VERSION_Cabal(3,0,0))
 idrisPreSDist args flags = do
   let dir = S.fromFlag (S.sDistDirectory flags)
   let verb = S.fromFlag (S.sDistVerbosity flags)
@@ -244,6 +245,7 @@
                               removeFile targetFile)
              (\e -> let e' = (e :: SomeException) in return ())
   postSDist simpleUserHooks args flags desc lbi
+#endif
 
 #if !(MIN_VERSION_Cabal(2,0,0))
 rewriteFileEx :: Verbosity -> FilePath -> String -> IO ()
@@ -353,8 +355,10 @@
    , postInst = \_ flags pkg local ->
                   idrisInstall (S.fromFlag $ S.installVerbosity flags)
                                NoCopyDest pkg local
+#if !(MIN_VERSION_Cabal(3,0,0))
    , preSDist = idrisPreSDist
    , sDistHook = idrisSDist (sDistHook simpleUserHooks)
    , postSDist = idrisPostSDist
+#endif
    , testHook = idrisTestHook
    }
