--- a/libcxx/src/CMakeLists.txt	2023-06-11 00:58:16.000000000 +0200
+++ b/libcxx/src/CMakeLists.txt	2023-06-14 21:13:36.956991676 +0200
@@ -301,6 +301,7 @@
       COMPILE_FLAGS "${LIBCXX_COMPILE_FLAGS}"
       LINK_FLAGS    "${LIBCXX_LINK_FLAGS}"
       OUTPUT_NAME   "${LIBCXX_STATIC_OUTPUT_NAME}"
+      POSITION_INDEPENDENT_CODE ON
   )
   cxx_add_common_build_flags(cxx_static)
   cxx_set_common_defines(cxx_static)
