--- a/flang/CMakeLists.txt	2022-09-15 23:22:53.138769870 +0200
+++ b/flang/CMakeLists.txt	2022-09-15 23:23:21.174144242 +0200
@@ -92,8 +92,9 @@
   get_filename_component(MLIR_DIR_ABSOLUTE ${MLIR_DIR} REALPATH)
   list(APPEND CMAKE_MODULE_PATH ${MLIR_DIR_ABSOLUTE})
   include(AddMLIR)
-  find_program(MLIR_TABLEGEN_EXE "mlir-tblgen" ${LLVM_TOOLS_BINARY_DIR}
-    NO_DEFAULT_PATH)
+  unset(MLIR_TABLEGEN_EXE)
+  find_program(MLIR_TABLEGEN_EXE "mlir-tblgen" HINTS ${LLVM_TOOLS_BINARY_DIR}
+    NO_DEFAULT_PATH REQUIRED)
 
   option(LLVM_INSTALL_TOOLCHAIN_ONLY
     "Only include toolchain files in the 'install' target." OFF)
