--- a/src/librustc_target/spec/linux_musl_base.rs
+++ b/src/librustc_target/spec/linux_musl_base.rs
@@ -14,6 +14,8 @@
     base.crt_static_default = false;
     // These targets allow the user to choose between static and dynamic linking.
     base.crt_static_respected = true;
+    // Static position-independent executables are supported.
+    base.static_position_independent_executables = true;
 
     base
 }
