--- a/compiler/rustc_target/src/spec/crt_objects.rs   2022-11-06 21:28:03.937194468 +0100
+++ b/compiler/rustc_target/src/spec/crt_objects.rs   2022-11-06 21:28:22.668483163 +0100
@@ -63,28 +63,6 @@
     ])
 }
 
-pub(super) fn pre_musl_self_contained() -> CrtObjects {
-    new(&[
-        (LinkOutputKind::DynamicNoPicExe, &["crt1.o", "crti.o", "crtbegin.o"]),
-        (LinkOutputKind::DynamicPicExe, &["Scrt1.o", "crti.o", "crtbeginS.o"]),
-        (LinkOutputKind::StaticNoPicExe, &["crt1.o", "crti.o", "crtbegin.o"]),
-        (LinkOutputKind::StaticPicExe, &["rcrt1.o", "crti.o", "crtbeginS.o"]),
-        (LinkOutputKind::DynamicDylib, &["crti.o", "crtbeginS.o"]),
-        (LinkOutputKind::StaticDylib, &["crti.o", "crtbeginS.o"]),
-    ])
-}
-
-pub(super) fn post_musl_self_contained() -> CrtObjects {
-    new(&[
-        (LinkOutputKind::DynamicNoPicExe, &["crtend.o", "crtn.o"]),
-        (LinkOutputKind::DynamicPicExe, &["crtendS.o", "crtn.o"]),
-        (LinkOutputKind::StaticNoPicExe, &["crtend.o", "crtn.o"]),
-        (LinkOutputKind::StaticPicExe, &["crtendS.o", "crtn.o"]),
-        (LinkOutputKind::DynamicDylib, &["crtendS.o", "crtn.o"]),
-        (LinkOutputKind::StaticDylib, &["crtendS.o", "crtn.o"]),
-    ])
-}
-
 pub(super) fn pre_mingw_self_contained() -> CrtObjects {
     new(&[
         (LinkOutputKind::DynamicNoPicExe, &["crt2.o", "rsbegin.o"]),
--- a/compiler/rustc_target/src/spec/base/linux_musl.rs	2020-12-31 19:07:19.216270236 +0100
+++ b/compiler/rustc_target/src/spec/base/linux_musl.rs	2020-12-31 19:07:53.639438587 +0100
@@ -1,11 +1,8 @@
-use crate::spec::{Env, LinkSelfContainedDefault, TargetOptions, base, crt_objects};
+use crate::spec::{Env, TargetOptions, base};

 pub(crate) fn opts() -> TargetOptions {
     TargetOptions {
         env: Env::Musl,
-        pre_link_objects_self_contained: crt_objects::pre_musl_self_contained(),
-        post_link_objects_self_contained: crt_objects::post_musl_self_contained(),
-        link_self_contained: LinkSelfContainedDefault::InferredForMusl,
         ..base::linux::opts()
     }
 }
--- a/src/bootstrap/src/core/config/target_selection.rs	2025-08-09 10:52:57.423193421 +0200
+++ b/src/bootstrap/src/core/config/target_selection.rs	2025-08-09 10:53:40.186580248 +0200
@@ -93,7 +93,7 @@
     }
 
     pub fn needs_crt_begin_end(&self) -> bool {
-        self.contains("musl") && !self.contains("unikraft")
+        false
     }
 
     /// Path to the file defining the custom target, if any.
