--- a/gnome-session/main.c	2016-09-13 21:06:30.721027175 +0200
+++ b/gnome-session/main.c	2016-09-13 21:07:06.870057648 +0200
@@ -313,7 +313,7 @@
 
         debug_string = g_getenv ("GNOME_SESSION_DEBUG");
         if (debug_string != NULL) {
-                debug = rpmatch (debug_string) == TRUE || atoi (debug_string) == 1;
+                debug = atoi (debug_string) == 1;
         }
 
         error = NULL;
