--- pommed/conffile.c.bak	2011-06-02 11:24:05.000000000 +0200
+++ pommed/conffile.c	2015-10-06 14:56:26.254326201 +0200
@@ -22,6 +22,7 @@
 #include <string.h>
 
 #include <syslog.h>
+#include <pthread.h>
 
 #include <confuse.h>
 
--- pommed/mactel/ambient.c.bak	2016-02-03 22:53:52.417115713 +0100
+++ pommed/mactel/ambient.c	2016-02-03 22:54:04.408625684 +0100
@@ -26,6 +26,7 @@
 #include <fcntl.h>
 #include <string.h>
 #include <errno.h>
+#include <limits.h>
 
 #include <syslog.h>
 
