diff --color -ur a/ell/ell.sym b/ell/ell.sym
--- a/ell/ell.sym	2021-05-02 13:06:43.000000000 +0200
+++ b/ell/ell.sym	2021-05-05 15:36:43.504110239 +0200
@@ -4,6 +4,7 @@
 	l_malloc;
 	l_realloc;
 	l_memdup;
+	l_free_cleanup;
 	l_free;
 	l_strdup;
 	l_strndup;
@@ -436,6 +437,7 @@
 	l_ringbuf_append;
 	/* settings */
 	l_settings_new;
+	l_settings_free_cleanup;
 	l_settings_free;
 	l_settings_load_from_data;
 	l_settings_to_data;
