--- a/ell/cert.c	2021-03-30 12:56:49.180305112 +0200
+++ b/ell/cert.c	2021-03-30 12:59:35.437505269 +0200
@@ -435,6 +435,11 @@
 	return NULL;
 }
 
+static void cert_keyring_cleanup(struct l_keyring **p)
+{
+	l_keyring_free(*p);
+}
+
 #define RETURN_ERROR(msg, args...)	\
 	do {	\
 		if (error) {	\
@@ -449,7 +454,8 @@
 					const char **error)
 {
 	struct l_keyring *ca_ring = NULL;
-	_auto_(l_keyring_free) struct l_keyring *verify_ring = NULL;
+	L_AUTO_CLEANUP_VAR(struct l_keyring *, verify_ring,
+				cert_keyring_cleanup) = NULL;
 	struct l_cert *cert;
 	struct l_key *prev_key = NULL;
 	int verified = 0;
--- a/ell/util.h	2021-03-30 12:56:49.180305112 +0200
+++ b/ell/util.h	2021-03-30 12:57:30.957599969 +0200
@@ -226,6 +226,9 @@
 	L_PUT_UNALIGNED(L_CPU_TO_BE64(val), (uint64_t *) ptr);
 }
 
+#define L_AUTO_CLEANUP_VAR(vartype,varname,destroy) \
+	vartype varname __attribute__((cleanup(destroy)))
+
 #define L_AUTO_FREE_VAR(vartype,varname) \
 	vartype varname __attribute__((cleanup(auto_free)))
 
