--- b/gpsoauth/google.py
+++ a/gpsoauth/google.py
@@ -1,8 +1,8 @@
 import base64
 import hashlib
 
+from Crypto.PublicKey import RSA
+from Crypto.Cipher import PKCS1_OAEP
-from Cryptodome.PublicKey import RSA
-from Cryptodome.Cipher import PKCS1_OAEP
 
 from .util import bytes_to_long, long_to_bytes
 
--- b/setup.py
+++ a/setup.py
@@ -34,7 +34,7 @@
     packages=['gpsoauth'],
     include_package_data=True,
     install_requires=[
+        'pycryptodome >= 3.0',
-        'pycryptodomex >= 3.0',
         'requests',
     ],
     license='MIT',
