--- a/runtime/druntime/src/rt/sections_elf_shared.d	2016-01-03 21:18:32.000000000 +0100
+++ b/runtime/druntime/src/rt/sections_elf_shared.d	2016-01-26 13:03:31.281469463 +0100
@@ -646,8 +646,6 @@
             if (dyn.d_tag == DT_STRTAB)
             {
                 version (linux)
-                    strtab = cast(const(char)*)dyn.d_un.d_ptr;
-                else version (FreeBSD)
                     strtab = cast(const(char)*)(info.dlpi_addr + dyn.d_un.d_ptr); // relocate
                 else
                     static assert(0, "unimplemented");
