--- a/src/mi/backtrace.c	2016-12-18 01:17:03.464617464 +0100
+++ b/src/mi/backtrace.c	2016-12-18 01:17:10.863585935 +0100
@@ -74,8 +74,4 @@
 
   return n;
 }
-
-extern int backtrace (void **buffer, int size)
-  WEAK ALIAS(unw_backtrace);
-
 #endif /* !UNW_REMOTE_ONLY */
--- a/configure	2016-12-18 01:20:51.366704575 +0100
+++ b/configure	2016-12-18 01:20:40.389746125 +0100
@@ -17263,7 +17263,7 @@
 fi
 
 if test x$GCC = xyes -a x$intel_compiler != xyes -a x$qcc_compiler != xyes; then
-  LIBCRTS="-lgcc_s"
+  LIBCRTS="-lunwind"
 fi
 
 { $as_echo "$as_me:${as_lineno-$LINENO}: checking for __builtin___clear_cache" >&5
