--- a/src/test/test-sizeof.c	2017-09-25 18:52:41.028550848 +0200
+++ b/src/test/test-sizeof.c	2017-09-25 18:52:48.315538673 +0200
@@ -48,8 +48,6 @@
         info(unsigned);
         info(long unsigned);
         info(long long unsigned);
-        info(__syscall_ulong_t);
-        info(__syscall_slong_t);
 
         info(float);
         info(double);
@@ -59,7 +57,6 @@
         info(ssize_t);
         info(time_t);
         info(usec_t);
-        info(__time_t);
 
         info(enum Enum);
         info(enum BigEnum);
