From 9b146a0a80bf6cdce8ad20256b60777beb426070 Mon Sep 17 00:00:00 2001
From: maxice8 <thinkabit.ukim@gmail.com>
Date: Fri, 27 Apr 2018 10:05:25 -0300
Subject: [PATCH] musl_missing.h: add FTW_* macros missing from musl libc.

---
 src/shared/musl_missing.h | 21 +++++++++++++++++++++
 1 file changed, 21 insertions(+)

diff --git a/src/shared/musl_missing.h b/src/shared/musl_missing.h
index d42f03c34..15f5f8473 100644
--- a/src/shared/musl_missing.h
+++ b/src/shared/musl_missing.h
@@ -107,6 +107,27 @@ typedef __compar_fn_t comparison_fn_t;
  */
 #define __register_atfork(prepare,parent,child,dso) pthread_atfork(prepare,parent,child)
 
+/* 
+ * Missing FTW macros in musl, define them if not defined
+ * taken from
+ * https://git.yoctoproject.org/cgit.cgi/poky/plain/meta/recipes-core/systemd/systemd/0028-add-missing-FTW_-macros-for-musl.patch
+ */
+#ifndef FTW_ACTIONRETVAL
+#define FTW_ACTIONRETVAL 16
+#endif
+
+#ifndef FTW_CONTINUE
+#define FTW_CONTINUE 0
+#endif
+
+#ifndef FTW_STOP
+#define FTW_STOP 1
+#endif
+
+#ifndef FTW_SKIP_SUBTREE
+#define FTW_SKIP_SUBTREE 2
+#endif
+
 #endif // !defined(__GLIBC__)
 
 #endif // ELOGIND_BASIC_MUSL_MISSING_H_INCLUDED
