--- a/arch/arm/Makefile	2018-10-27 21:13:57.390791660 +0200
+++ b/arch/arm/Makefile	2018-10-27 21:17:20.525583851 +0200
@@ -10,7 +10,7 @@
 #
 # Copyright (C) 1995-2001 by Russell King
 
-LDFLAGS_vmlinux	:= --no-undefined -X --pic-veneer
+LDFLAGS_vmlinux	:= --no-undefined -X
 ifeq ($(CONFIG_CPU_ENDIAN_BE8),y)
 LDFLAGS_vmlinux	+= --be8
 KBUILD_LDFLAGS_MODULE	+= --be8
@@ -51,6 +51,9 @@
 KBUILD_LDFLAGS	+= -EL
 endif
 
+# Set cpu arch
+KBUILD_LDFLAGS	+= -m armelf_linux_eabi
+
 #
 # The Scalar Replacement of Aggregates (SRA) optimization pass in GCC 4.9 and
 # later may result in code being generated that handles signed short and signed 
