--- a/Makefile	2019-02-01 23:33:19.307522218 +0100
+++ b/Makefile	2019-02-01 23:34:37.531390992 +0100
@@ -512,6 +512,8 @@
 
 KBUILD_CFLAGS	+= $(call cc-option,-fno-PIE)
 KBUILD_AFLAGS	+= $(call cc-option,-fno-PIE)
+KBUILD_CFLAGS_MODULE	+= $(call cc-option,-fno-PIE)
+KBUILD_AFLAGS_MODULE	+= $(call cc-option,-fno-PIE)
 
 # The expansion should be delayed until arch/$(SRCARCH)/Makefile is included.
 # Some architectures define CROSS_COMPILE in arch/$(SRCARCH)/Makefile.
