--- a/arch/arm64/boot/dts/rockchip/rk3328.dtsi	2019-09-22 14:13:25.821093280 +0200
+++ b/arch/arm64/boot/dts/rockchip/rk3328.dtsi	2019-09-22 14:28:12.427006256 +0200
@@ -105,22 +105,22 @@
 		};
 		opp-816000000 {
 			opp-hz = /bits/ 64 <816000000>;
-			opp-microvolt = <1000000>;
+			opp-microvolt = <1050000>;
 			clock-latency-ns = <40000>;
 		};
 		opp-1008000000 {
 			opp-hz = /bits/ 64 <1008000000>;
-			opp-microvolt = <1100000>;
+			opp-microvolt = <1150000>;
 			clock-latency-ns = <40000>;
 		};
 		opp-1200000000 {
 			opp-hz = /bits/ 64 <1200000000>;
-			opp-microvolt = <1225000>;
+			opp-microvolt = <1275000>;
 			clock-latency-ns = <40000>;
 		};
 		opp-1296000000 {
 			opp-hz = /bits/ 64 <1296000000>;
-			opp-microvolt = <1300000>;
+			opp-microvolt = <1350000>;
 			clock-latency-ns = <40000>;
 		};
 	};
@@ -596,6 +597,17 @@
 		status = "disabled";
 	};
 
+	vpu: video-codec@ff350000 {
+		compatible = "rockchip,rk3328-vpu";
+		reg = <0x0 0xff350000 0x0 0x800>;
+		interrupts = <GIC_SPI 9 IRQ_TYPE_LEVEL_HIGH>;
+		interrupt-names = "vdpu";
+		clocks = <&cru ACLK_VPU>, <&cru HCLK_VPU>;
+		clock-names = "aclk", "hclk";
+		iommus = <&vpu_mmu>;
+		power-domains = <&power RK3328_PD_VPU>;
+	};
+
 	vpu_mmu: iommu@ff350800 {
 		compatible = "rockchip,iommu";
 		reg = <0x0 0xff350800 0x0 0x40>;
--- a/arch/arm64/boot/dts/rockchip/rk3328-rock64.dts	2019-09-22 14:07:54.243379846 +0200
+++ b/arch/arm64/boot/dts/rockchip/rk3328-rock64.dts	2019-09-22 14:18:23.028816748 +0200
@@ -11,7 +11,7 @@
 	compatible = "pine64,rock64", "rockchip,rk3328";
 
 	chosen {
-		stdout-path = "serial2:1500000n8";
+		stdout-path = "serial2:115200n8";
 	};
 
 	gmac_clkin: external-gmac-clock {
