--- a/arch/arm64/boot/dts/qcom/purwa.dtsi	2026-01-30 16:32:40.988562775 +0100
+++ b/arch/arm64/boot/dts/qcom/purwa.dtsi	2026-01-30 16:34:35.152881281 +0100
@@ -166,6 +166,22 @@
 	};
 };
 
+&csiphy0 {
+	compatible = "qcom,x1p42100-mipi-csi2-combo-phy";
+};
+
+&csiphy1 {
+	compatible = "qcom,x1p42100-mipi-csi2-combo-phy";
+};
+
+&csiphy2 {
+	compatible = "qcom,x1p42100-mipi-csi2-combo-phy";
+};
+
+&csiphy4 {
+	compatible = "qcom,x1p42100-mipi-csi2-combo-phy";
+};
+
 &soc {
 	/* The PCIe3 PHY on X1P42100 uses a different IP block */
 	pcie3_phy: phy@1bd4000 {
