--- a/drivers/media/i2c/ov13858.c	2026-02-08 22:03:27.000000000 +0100
+++ b/drivers/media/i2c/ov13858.c	2026-03-06 22:56:54.098976876 +0100
@@ -3,6 +3,7 @@
 
 #include <linux/acpi.h>
 #include <linux/clk.h>
+#include <linux/delay.h>
 #include <linux/i2c.h>
 #include <linux/module.h>
 #include <linux/pm_runtime.h>
@@ -40,6 +41,10 @@
 #define OV13858_REG_MIPI_SC_CTRL0	0x3016
 #define OV13858_REG_MIPI_SC_CTRL1	0x3022
 
+#define OV13858_MCLK                    19200000
+#define OV13858_DATA_LANES		4
+#define OV13858_RGB_DEPTH		10
+
 /* Chip ID */
 #define OV13858_REG_CHIP_ID		0x300a
 #define OV13858_CHIP_ID			0x00d855
@@ -949,8 +954,8 @@
  */
 static u64 link_freq_to_pixel_rate(u64 f)
 {
-	f *= 2 * 4;
-	do_div(f, 10);
+	f *= 2 * OV13858_DATA_LANES;
+	do_div(f, OV13858_RGB_DEPTH);
 
 	return f;
 }
@@ -1028,9 +1033,15 @@
 	}
 };
 
+static const char * const ov13858_supply_names[] = {
+	"afvdd",	/* Autofocus power */
+	"dovdd",	/* Digital I/O power */
+	"avdd",		/* Analog power */
+	"dvdd",		/* Digital core power */
+};
+
 struct ov13858 {
 	struct device *dev;
-	struct clk *clk;
 
 	struct v4l2_subdev sd;
 	struct media_pad pad;
@@ -1043,9 +1054,17 @@
 	struct v4l2_ctrl *hblank;
 	struct v4l2_ctrl *exposure;
 
+	struct clk *img_clk;
+	struct regulator_bulk_data supplies[ARRAY_SIZE(ov13858_supply_names)];
+	struct gpio_desc *reset;
+
 	/* Current mode */
 	const struct ov13858_mode *cur_mode;
 
+	/* MIPI lanes info */
+	u32 link_freq_index;
+	u8 mipi_lanes;
+
 	/* Mutex for serialized access */
 	struct mutex mutex;
 };
@@ -1655,6 +1674,128 @@
 	mutex_destroy(&ov13858->mutex);
 }
 
+static int ov13858_get_pm_resources(struct device *dev)
+{
+	struct v4l2_subdev *sd = dev_get_drvdata(dev);
+	struct ov13858 *ov13858 = to_ov13858(sd);
+	int i;
+
+	ov13858->reset = devm_gpiod_get_optional(dev, "reset", GPIOD_OUT_HIGH);
+	if (IS_ERR(ov13858->reset))
+		return dev_err_probe(dev, PTR_ERR(ov13858->reset),
+				     "failed to get reset gpio\n");
+
+	for (i = 0; i < ARRAY_SIZE(ov13858_supply_names); i++)
+		ov13858->supplies[i].supply = ov13858_supply_names[i];
+
+	return devm_regulator_bulk_get(dev, ARRAY_SIZE(ov13858_supply_names),
+				       ov13858->supplies);
+}
+
+static int ov13858_power_off(struct device *dev)
+{
+	struct v4l2_subdev *sd = dev_get_drvdata(dev);
+	struct ov13858 *ov13858 = to_ov13858(sd);
+
+	if (ov13858->reset)
+		gpiod_set_value_cansleep(ov13858->reset, 1);
+
+	regulator_bulk_disable(ARRAY_SIZE(ov13858_supply_names),
+			       ov13858->supplies);
+
+	clk_disable_unprepare(ov13858->img_clk);
+
+	return 0;
+}
+
+static int ov13858_power_on(struct device *dev)
+{
+	struct v4l2_subdev *sd = dev_get_drvdata(dev);
+	struct ov13858 *ov13858 = to_ov13858(sd);
+	int ret;
+
+	/* Assert reset */
+	if (ov13858->reset)
+		gpiod_set_value_cansleep(ov13858->reset, 1);
+
+	/* Enable clock first */
+	ret = clk_prepare_enable(ov13858->img_clk);
+	if (ret < 0) {
+		dev_err(dev, "failed to enable imaging clock: %d\n", ret);
+		return ret;
+	}
+
+	ret = regulator_bulk_enable(ARRAY_SIZE(ov13858_supply_names),
+				    ov13858->supplies);
+	if (ret < 0) {
+		dev_err(dev, "failed to enable regulators\n");
+		goto disable_clk;
+	}
+
+	usleep_range(5000, 10000);
+	gpiod_set_value_cansleep(ov13858->reset, 1);
+	msleep(20);
+
+	return 0;
+
+disable_clk:
+	clk_disable_unprepare(ov13858->img_clk);
+
+	return ret;
+}
+
+static int ov13858_check_hwcfg(struct ov13858 *ov13858)
+{
+	struct v4l2_fwnode_endpoint bus_cfg = {
+		.bus_type = V4L2_MBUS_CSI2_DPHY
+	};
+	struct device *dev = ov13858->dev;
+	struct fwnode_handle *ep;
+	struct fwnode_handle *fwnode = dev_fwnode(dev);
+	unsigned long link_freq_bitmap;
+	int ret;
+
+	ep = fwnode_graph_get_next_endpoint(fwnode, NULL);
+	if (!ep)
+		return dev_err_probe(dev, -EPROBE_DEFER,
+				     "waiting for fwnode graph endpoint\n");
+
+	ret = v4l2_fwnode_endpoint_alloc_parse(ep, &bus_cfg);
+	fwnode_handle_put(ep);
+	if (ret)
+		return dev_err_probe(dev, ret, "parsing endpoint failed\n");
+
+	if (bus_cfg.bus.mipi_csi2.num_data_lanes != OV13858_DATA_LANES) {
+		dev_err(dev, "number of CSI2 data lanes %d is not supported\n",
+			bus_cfg.bus.mipi_csi2.num_data_lanes);
+		ret = -EINVAL;
+		goto out_err;
+	}
+
+	if (!bus_cfg.nr_of_link_frequencies) {
+		dev_err(dev, "no link frequencies defined\n");
+		ret = -EINVAL;
+		goto out_err;
+	}
+
+	ret = v4l2_link_freq_to_bitmap(dev, bus_cfg.link_frequencies,
+				       bus_cfg.nr_of_link_frequencies,
+				       link_freq_menu_items,
+				       ARRAY_SIZE(link_freq_menu_items),
+				       &link_freq_bitmap);
+	if (ret)
+		goto out_err;
+
+	/* v4l2_link_freq_to_bitmap() guarantees at least 1 bit is set */
+	ov13858->link_freq_index = ffs(link_freq_bitmap) - 1;
+	ov13858->mipi_lanes = bus_cfg.bus.mipi_csi2.num_data_lanes;
+
+out_err:
+	v4l2_fwnode_endpoint_free(&bus_cfg);
+
+	return ret;
+}
+
 static int ov13858_probe(struct i2c_client *client)
 {
 	struct ov13858 *ov13858;
@@ -1667,13 +1808,13 @@
 
 	ov13858->dev = &client->dev;
 
-	ov13858->clk = devm_v4l2_sensor_clk_get(ov13858->dev, NULL);
-	if (IS_ERR(ov13858->clk))
-		return dev_err_probe(ov13858->dev, PTR_ERR(ov13858->clk),
-				     "failed to get clock\n");
+	ov13858->img_clk = devm_v4l2_sensor_clk_get(ov13858->dev, NULL);
+	if (IS_ERR(ov13858->img_clk))
+		return dev_err_probe(ov13858->dev, PTR_ERR(ov13858->img_clk),
+				     "failed to get imaging clock\n");
 
-	freq = clk_get_rate(ov13858->clk);
-	if (freq != 19200000)
+	freq = clk_get_rate(ov13858->img_clk);
+	if (freq != OV13858_MCLK)
 		return dev_err_probe(ov13858->dev, -EINVAL,
 				     "external clock %lu is not supported\n",
 				     freq);
@@ -1681,11 +1822,26 @@
 	/* Initialize subdev */
 	v4l2_i2c_subdev_init(&ov13858->sd, client, &ov13858_subdev_ops);
 
+	/* Check HW config */
+	ret = ov13858_check_hwcfg(ov13858);
+	if (ret)
+		return ret;
+
+	ret = ov13858_get_pm_resources(ov13858->dev);
+	if (ret)
+		return ret;
+
+	ret = ov13858_power_on(ov13858->dev);
+	if (ret) {
+		dev_err_probe(ov13858->dev, ret, "failed to power on\n");
+		return ret;
+	}
+
 	/* Check module identity */
 	ret = ov13858_identify_module(ov13858);
 	if (ret) {
 		dev_err(ov13858->dev, "failed to find sensor: %d\n", ret);
-		return ret;
+		goto probe_error_power_off;
 	}
 
 	/* Set default mode to max resolution */
@@ -1693,7 +1849,8 @@
 
 	ret = ov13858_init_controls(ov13858);
 	if (ret)
-		return ret;
+		dev_err(ov13858->dev, "failed to init controls: %d\n", ret);
+		goto error_handler_free;
 
 	/* Initialize subdev */
 	ov13858->sd.internal_ops = &ov13858_internal_ops;
@@ -1710,16 +1867,13 @@
 		goto error_handler_free;
 	}
 
+	pm_runtime_set_active(ov13858->dev);
+	pm_runtime_enable(ov13858->dev);
+
 	ret = v4l2_async_register_subdev_sensor(&ov13858->sd);
 	if (ret < 0)
 		goto error_media_entity;
 
-	/*
-	 * Device is already turned on by i2c-core with ACPI domain PM.
-	 * Enable runtime PM and turn off the device.
-	 */
-	pm_runtime_set_active(ov13858->dev);
-	pm_runtime_enable(ov13858->dev);
 	pm_runtime_idle(ov13858->dev);
 
 	return 0;
@@ -1729,7 +1883,9 @@
 
 error_handler_free:
 	ov13858_free_controls(ov13858);
-	dev_err(ov13858->dev, "%s failed:%d\n", __func__, ret);
+
+probe_error_power_off:
+	ov13858_power_off(ov13858->dev);
 
 	return ret;
 }
@@ -1744,8 +1900,16 @@
 	ov13858_free_controls(ov13858);
 
 	pm_runtime_disable(ov13858->dev);
+
+	if (!pm_runtime_status_suspended(ov13858->dev)) {
+		ov13858_power_off(ov13858->dev);
+		pm_runtime_set_suspended(ov13858->dev);
+	}
 }
 
+static DEFINE_RUNTIME_DEV_PM_OPS(ov13858_pm_ops, ov13858_power_off,
+				 ov13858_power_on, NULL);
+
 static const struct i2c_device_id ov13858_id_table[] = {
 	{ "ov13858" },
 	{}
@@ -1762,10 +1926,18 @@
 MODULE_DEVICE_TABLE(acpi, ov13858_acpi_ids);
 #endif
 
+static const struct of_device_id ov13858_of_match[] = {
+	{ .compatible = "ovti,ov13858" },
+	{ /* sentinel */ }
+};
+MODULE_DEVICE_TABLE(of, ov13858_of_match);
+
 static struct i2c_driver ov13858_i2c_driver = {
 	.driver = {
 		.name = "ov13858",
+		.pm = pm_sleep_ptr(&ov13858_pm_ops),
 		.acpi_match_table = ACPI_PTR(ov13858_acpi_ids),
+		.of_match_table = ov13858_of_match,
 	},
 	.probe = ov13858_probe,
 	.remove = ov13858_remove,
