--- a/f_check	2023-07-30 16:29:15.392815946 +0200
+++ b/f_check	2023-07-30 16:29:30.359022776 +0200
@@ -365,7 +365,7 @@
         case "$flag" in -l*)
             case "$flag" in
                 *ibrary*|*gfortranbegin*|*flangmain*|*frtbegin*|*pathfstart*|\
-                    *crt[0-9]*|*gcc*|*user32*|*kernel32*|*advapi32*|*shell32*|\
+                    *crt[0-9]*|*gcc*|*wp|*user32*|*kernel32*|*advapi32*|*shell32*|\
                     -l) ;;
                 *omp*)
                     case "$vendor" in
