--- a/Makefile.system	2023-07-30 17:54:49.238902111 +0200
+++ b/Makefile.system	2023-07-30 18:01:21.231531337 +0200
@@ -820,9 +820,12 @@
 ifeq ($(F_COMPILER), GFORTRAN)
 FCOMMON_OPT +=  -fdefault-integer-8
 endif
-ifeq ($(F_COMPILER), FLANG)
+ifeq ($(F_COMPILER), CLASSICFLANG)
 FCOMMON_OPT += -i8
 endif
+ifeq ($(F_COMPILER), FLANG)
+FCOMMON_OPT += -fdefault-integer-8
+endif
 endif
 endif
 endif
@@ -835,9 +838,12 @@
 ifeq ($(F_COMPILER), GFORTRAN)
 FCOMMON_OPT +=  -fdefault-integer-8
 endif
-ifeq ($(F_COMPILER), FLANG)
+ifeq ($(F_COMPILER), CLASSICFLANG)
 FCOMMON_OPT += -i8
 endif
+ifeq ($(F_COMPILER), FLANG)
+FCOMMON_OPT += -fdefault-integer-8
+endif
 endif
 endif
 endif
@@ -850,9 +856,12 @@
 ifeq ($(F_COMPILER), GFORTRAN)
 FCOMMON_OPT +=  -fdefault-integer-8
 endif
-ifeq ($(F_COMPILER), FLANG)
+ifeq ($(F_COMPILER), CLASSICFLANG)
 FCOMMON_OPT += -i8
 endif
+ifeq ($(F_COMPILER), FLANG)
+FCOMMON_OPT += -fdefault-integer-8
+endif
 endif
 endif
 endif
@@ -1025,7 +1034,7 @@
 endif
 endif
 
-ifeq ($(F_COMPILER), FLANG)
+ifeq ($(F_COMPILER), CLASSICFLANG)
 CCOMMON_OPT += -DF_INTERFACE_FLANG
 FCOMMON_OPT += -Mrecursive -Kieee
 ifeq ($(OSNAME), Linux)
@@ -1055,6 +1055,42 @@
 endif
 endif
 
+ifeq ($(F_COMPILER), FLANG)
+CCOMMON_OPT += -DF_INTERFACE_GFORT
+FCOMMON_OPT += -Wall
+ifdef NO_BINARY_MODE
+ifeq ($(ARCH), $(filter $(ARCH),mips64))
+ifdef BINARY64
+FCOMMON_OPT += -mabi=64
+else
+FCOMMON_OPT += -mabi=n32
+endif
+else ifeq ($(ARCH), $(filter $(ARCH),mips))
+FCOMMON_OPT += -mabi=32
+endif
+else
+ifdef BINARY64
+ifneq ($(OSNAME), AIX)
+ifneq ($(ARCH), riscv64)
+FCOMMON_OPT += -m64
+endif
+endif
+ifdef INTERFACE64
+ifneq ($(INTERFACE64), 0)
+FCOMMON_OPT +=  -fdefault-integer-8
+endif
+endif
+else
+ifneq ($(OSNAME), AIX)
+FCOMMON_OPT += -m32
+endif
+endif
+endif
+ifeq ($(USE_OPENMP), 1)
+FCOMMON_OPT += -fopenmp
+endif
+endif
+
 ifeq ($(F_COMPILER), G77)
 CCOMMON_OPT += -DF_INTERFACE_G77
 FCOMMON_OPT += -Wall
--- a/exports/Makefile	2023-04-01 22:18:01.000000000 +0200
+++ b/exports/Makefile	2023-07-30 18:03:28.968366267 +0200
@@ -185,6 +185,11 @@
 	-Wl,--whole-archive $< -Wl,--no-whole-archive \
 	-Wl,-soname,$(INTERNALNAME) $(EXTRALIB)
 	$(CC) $(CFLAGS) $(LDFLAGS) -w -o linktest linktest.c ../$(LIBSONAME) $(FEXTRALIB) && echo OK.
+else ifeq ($(F_COMPILER), CLASSICFLANG)
+	$(FC) $(FFLAGS) $(LDFLAGS) -shared -o ../$(LIBSONAME) \
+	-Wl,--whole-archive $< -Wl,--no-whole-archive \
+	-Wl,-soname,$(INTERNALNAME) $(EXTRALIB)
+	$(CC) $(CFLAGS) $(LDFLAGS) -w -o linktest linktest.c ../$(LIBSONAME) $(FEXTRALIB) && echo OK.
 else ifeq ($(F_COMPILER), FLANG)
 	$(FC) $(FFLAGS) $(LDFLAGS) -shared -o ../$(LIBSONAME) \
 	-Wl,--whole-archive $< -Wl,--no-whole-archive \
--- a/test/Makefile	2023-04-01 22:18:01.000000000 +0200
+++ b/test/Makefile	2023-07-30 18:02:31.300537848 +0200
@@ -263,8 +263,11 @@
 
 
 ifeq ($(USE_OPENMP), 1)
-ifeq ($(F_COMPILER), GFORTRAN)
 ifeq ($(C_COMPILER), CLANG)
+ifeq ($(F_COMPILER), GFORTRAN)
+CEXTRALIB = -lomp
+endif
+ifeq ($(F_COMPILER), FLANG)
 CEXTRALIB = -lomp
 endif
 endif
--- a/f_check	2023-04-01 22:18:01.000000000 +0200
+++ b/f_check	2023-07-30 18:07:36.347730563 +0200
@@ -81,6 +81,10 @@
                 vendor=FUJITSU
                 openmp='-Kopenmp'
                 ;;
+            *FIRModule*)
+                vendor=FLANG
+                openmp='-fopenmp'
+                ;;
 	    *Hewlett*)
 		vendor=CRAY
                openmp='-fopenmp'
@@ -99,7 +102,7 @@
                 else
                     case "$compiler" in
                         *flang*)
-                            vendor=FLANG
+                            vendor=CLASSICFLANG
                             openmp='-fopenmp'
                             ;;
                         *ifort*|*ifx*)
@@ -401,7 +401,7 @@
     done
 fi
 
-if [ "$vendor" = "FLANG" ]; then
+if [ "$vendor" = "CLASSICFLANG" ]; then
     linker_a="$linker_a -lflang"
 fi
 
