--- a/src/Language/Haskell/Exts/Simple/Fixity.hs
+++ b/src/Language/Haskell/Exts/Simple/Fixity.hs
@@ -19,7 +19,7 @@
 
 -- * Functions
 
-applyFixities :: (AppFixity ast, Functor ast, Monad m) => [Fixity] -> ast () -> m (ast ())
+applyFixities :: (AppFixity ast, Functor ast, MonadFail m, Monad m) => [Fixity] -> ast () -> m (ast ())
 applyFixities fixs = liftM (fmap (const ())) . H.applyFixities fixs . fmap (const noSrcSpan)
 
 -- omitted fixity functionality, new in haskell-source-exts-1.20:
