--- a/Breeze-gtk/gtk-2.0/gtkrc.bak	2016-07-02 21:17:27.587830869 +0200
+++ b/Breeze-gtk/gtk-2.0/gtkrc	2016-07-02 21:18:20.411830595 +0200
@@ -5,11 +5,11 @@
 gtk-color-scheme = "base_color:#fcfcfc"
 gtk-color-scheme = "insensitive_base_color:#e5e5e5"
 gtk-color-scheme = "fg_color:#31363b"
-gtk-color-scheme = "bg_color:#eff0f1"
+gtk-color-scheme = "bg_color:#ffffff"
 gtk-color-scheme = "selected_fg_color:#eff0f1"
 gtk-color-scheme = "selected_bg_color:#3daee9"
 gtk-color-scheme = "button_fg_color:#31363b"
-gtk-color-scheme = "tooltip_fg_color:#eff0f1"
+gtk-color-scheme = "tooltip_fg_color:#000000"
 gtk-color-scheme = "tooltip_bg_color:#31363b"
 gtk-color-scheme = "insensitive_fg_color:#aaadaf"
 gtk-color-scheme = "insensitive_text_color:#b3b4b6"
