--- a/Breeze-gtk/gtk-2.0/gtkrc.bak	2016-03-27 12:41:46.015949921 +0200
+++ b/Breeze-gtk/gtk-2.0/gtkrc	2016-03-27 12:42:17.991949180 +0200
@@ -2,9 +2,9 @@
 # Theme:        Breezy
 # Description:  Breeze theme for GTK+2.0
 
-gtk-color-scheme = "bg_color:#EFF0F1\nselected_bg_color:#3DAEE9\nbase_color:#FCFCFC" # Background, base.
+gtk-color-scheme = "bg_color:#FFFFFF\nselected_bg_color:#3DAEE9\nbase_color:#FCFCFC" # Background, base.
 gtk-color-scheme = "fg_color:#31363B\nselected_fg_color:#FFFFFF\ninsensitive_fg_color:#BDC3C7\ntext_color:#31363B" # Foreground, text.
-gtk-color-scheme = "tooltip_bg_color:#31363B\ntooltip_fg_color:#FCFCFC" # Tooltips.
+gtk-color-scheme = "tooltip_bg_color:#31363B\ntooltip_fg_color:#000000" # Tooltips.
 
 gtk-auto-mnemonics	= 1
 
