--- a/src/scopes/audioscopes/spectrogram.cpp	2016-03-06 12:38:14.156205118 +0100
+++ b/src/scopes/audioscopes/spectrogram.cpp	2016-03-06 12:38:54.430207101 +0100
@@ -245,7 +245,7 @@
                 x = leftDist + (m_innerScopeRect.width()-1) * ((float)hz)/m_freqMax;
 
                 // Hide text if it would overlap with the text drawn at the mouse position
-                hideText = m_aTrackMouse->isChecked() && m_mouseWithinWidget && abs(x-(leftDist + mouseX + 20)) < (int) minDistX + 16
+                hideText = m_aTrackMouse->isChecked() && m_mouseWithinWidget && (x-(leftDist + mouseX + 20)) < (int) minDistX + 16
                         && mouseX < m_innerScopeRect.width() && mouseX >= 0;
 
                 if (x <= rightBorder) {
@@ -273,7 +273,7 @@
             }
             // Draw the line at the very right (maximum frequency)
             x = leftDist + m_innerScopeRect.width()-1;
-            hideText = m_aTrackMouse->isChecked() && m_mouseWithinWidget && abs(x-(leftDist + mouseX + 30)) < (int) minDistX
+            hideText = m_aTrackMouse->isChecked() && m_mouseWithinWidget && (x-(leftDist + mouseX + 30)) < (int) minDistX
                     && mouseX < m_innerScopeRect.width() && mouseX >= 0;
             davinci.drawLine(x, topDist, x, topDist + m_innerScopeRect.height()+6);
             if (!hideText) {
--- a/src/dvdwizard/dvdwizardmenu.h	2016-03-06 12:39:21.931208455 +0100
+++ b/src/dvdwizard/dvdwizardmenu.h	2016-03-06 12:39:32.959208997 +0100
@@ -31,6 +31,7 @@
 #include <QGraphicsRectItem>
 #include <QDomElement>
 
+#include <cmath>
 #include <QDebug>
 #include <KMessageWidget>
 
