--- Make.defaults.bak	2016-10-02 16:09:51.583998429 +0200
+++ Make.defaults	2016-10-02 16:10:05.383997293 +0200
@@ -26,7 +26,7 @@
 gcc_cflags =
 cflags	= $(CFLAGS) $(SUBDIR_CFLAGS) \
 	 -Wall -Wextra -Wsign-compare -Wstrict-aliasing \
-	-std=gnu11 -fshort-wchar -fPIC \
+	-std=gnu11 -fPIC \
 	-D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -DLOCALEDIR=\"$(localedir)\" \
 	-DEFIBOOTMGR_VERSION="\"$(VERSION)\"" \
 	$(if $(findstring clang,$(CC)),$(clang_cflags),) \
