--- a/Make.defaults	2016-09-13 17:57:25.847813836 +0200
+++ b/Make.defaults	2016-09-13 17:57:36.472704876 +0200
@@ -26,7 +26,7 @@
 gcc_cflags =
 cflags	= $(CFLAGS) $(SUBDIR_CFLAGS) \
 	-Werror -Wall -Wextra -Wsign-compare -Wstrict-aliasing \
-	-std=gnu11 -fshort-wchar -fPIC -flto \
+	-std=gnu11 -fPIC -flto \
 	-D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -DLOCALEDIR=\"$(localedir)\" \
 	$(if $(filter $(CC),clang),$(clang_cflags),) \
 	$(if $(filter $(CC),gcc),$(gcc_cflags),) \
