--- src/pl-dict.c	2015-12-21 13:51:16.000000000 +0100
+++ src/pl-dict.c	2016-01-22 10:52:09.166944481 +0100
@@ -111,6 +111,8 @@
   return (ss->compar)(aa, bb, ss->arg);
 }
 
+#define __nonnull(params) __attribute__ ((__nonnull__ params))
+
 #if (defined _GNU_SOURCE || defined __GNU__ || defined __linux__)
 
 typedef int(* __compar_d_fn_t)(const void *, const void *, void *);
--- src/pl-thread.c	2015-12-21 13:51:16.000000000 +0100
+++ src/pl-thread.c	2016-01-22 11:52:04.378200909 +0100
@@ -5640,7 +5640,7 @@
 void
 __assert_fail(const char *assertion,
 	      const char *file,
-	      unsigned int line,
+	      int line,
 	      const char *function)
 { int tid = PL_thread_self();
   atom_t alias;
