--- a/Distribution/Simple/Program/Ar.hs	2018-02-06 21:16:52.444914714 +0100
+++ b/Distribution/Simple/Program/Ar.hs	2017-07-24 02:19:03.000000000 +0200
@@ -68,20 +68,20 @@
   -- call on OSX we use "ar qs" so that it'll make the index.
 
   let simpleArgs  = case hostOS of
-             OSX -> ["-r", "-s"]
-             _   -> ["-r"]
+             OSX -> "rs"
+             _   -> "r"
 
-      initialArgs = ["-q"]
+      initialArgs = "q"
       finalArgs   = case hostOS of
-             OSX -> ["-q", "-s"]
-             _   -> ["-q"]
+             OSX -> "qs"
+             _   -> "q"
 
-      extraArgs   = verbosityOpts verbosity ++ [tmpPath]
+      extraArgs   = verbosityOpts verbosity
 
-      simple  = programInvocation ar (simpleArgs  ++ extraArgs)
-      initial = programInvocation ar (initialArgs ++ extraArgs)
+      simple  = programInvocation ar ((simpleArgs  ++ extraArgs) : [tmpPath])
+      initial = programInvocation ar ((initialArgs ++ extraArgs) : [tmpPath])
       middle  = initial
-      final   = programInvocation ar (finalArgs   ++ extraArgs)
+      final   = programInvocation ar ((finalArgs   ++ extraArgs) : [tmpPath])
 
   sequence_
         [ runProgramInvocation verbosity inv
@@ -95,8 +95,8 @@
 
       invokeWithResponesFile :: FilePath -> ProgramInvocation
       invokeWithResponesFile atFile =
-        programInvocation ar $
-        simpleArgs ++ extraArgs ++ ['@' : atFile]
+        programInvocation ar
+        ((simpleArgs ++ extraArgs) : ['@' : atFile])
 
   if oldVersionManualOverride || responseArgumentsNotSupported
     then
@@ -118,9 +118,9 @@
     progDb = withPrograms lbi
     Platform hostArch hostOS = hostPlatform lbi
     verbosityOpts v
-      | v >= deafening = ["-v"]
-      | v >= verbose   = []
-      | otherwise      = ["-c"] -- Do not warn if library had to be created.
+      | v >= deafening = "v"
+      | v >= verbose   = ""
+      | otherwise      = "c" -- Do not warn if library had to be created.
 
 -- | @ar@ by default includes various metadata for each object file in their
 -- respective headers, so the output can differ for the same inputs, making
