--- a/src/translate.h	2016-09-19 17:48:15.386251913 +0200
+++ b/src/translate.h	2016-09-19 17:51:04.985885357 +0200
@@ -545,7 +545,7 @@
 	int decimal_sep;
 	int max_digits;    // max number of digits which can be spoken as an integer number (rather than individual digits)
 	const char *ordinal_indicator;   // UTF-8 string
-	const char *roman_suffix;    // add this (ordinal) suffix to Roman numbers (LANG=an)
+	const unsigned char *roman_suffix;    // add this (ordinal) suffix to Roman numbers (LANG=an)
 
 	// bit 0, accent name before the letter name, bit 1 "capital" after letter name
 	int accents;
--- a/src/tr_languages.cpp	2016-09-19 17:47:05.811472633 +0200
+++ b/src/tr_languages.cpp	2016-09-19 17:51:13.436871673 +0200
@@ -198,7 +198,7 @@
 	0x200d,  1, // zero width joiner
 	0, 0 };
 
-const char string_ordinal[] = {0xc2,0xba,0};  // masculine ordinal character, UTF-8
+const unsigned char string_ordinal[] = {0xc2,0xba,0};  // masculine ordinal character, UTF-8
 
 
 static Translator* NewTranslator(void)
@@ -296,7 +296,7 @@
 	tr->langopts.replace_chars = NULL;
 	tr->langopts.ascii_language[0] = 0;    // Non-Latin alphabet languages, use this language to speak Latin words, default is English
 	tr->langopts.alt_alphabet_lang = L('e','n');
-	tr->langopts.roman_suffix = "";
+	tr->langopts.roman_suffix = 0;
 
 	SetLengthMods(tr,201);
 //	tr->langopts.length_mods = length_mods_en;
