diff --git a/dispatch/dispatch.h b/dispatch/dispatch.h
index 2d45b83..56da4b5 100644
--- a/dispatch/dispatch.h
+++ b/dispatch/dispatch.h
@@ -48,6 +48,16 @@
 #endif
 #endif
 
+#ifndef __BEGIN_DECLS
+#if defined(__cplusplus)
+#define __BEGIN_DECLS extern "C" {
+#define __END_DECLS }
+#else
+#define __BEGIN_DECLS
+#define __END_DECLS
+#endif
+#endif
+
 #define DISPATCH_API_VERSION 20170124
 
 #ifndef __DISPATCH_BUILDING_DISPATCH__
diff --git a/os/linux_base.h b/os/linux_base.h
index c8b9cad..2838ad0 100644
--- a/os/linux_base.h
+++ b/os/linux_base.h
@@ -15,10 +15,6 @@
 
 #include <sys/param.h>
 
-#if HAVE_SYS_CDEFS_H
-#include <sys/cdefs.h>
-#endif
-
 #ifndef API_AVAILABLE
 #define API_AVAILABLE(...)
 #endif
diff --git a/src/internal.h b/src/internal.h
index 0536db1..4c2d9fe 100644
--- a/src/internal.h
+++ b/src/internal.h
@@ -248,11 +248,7 @@ DISPATCH_EXPORT DISPATCH_NOTHROW void dispatch_atfork_child(void);
 #if !TARGET_OS_WIN32
 #include <sys/mount.h>
 #include <sys/queue.h>
-#ifdef __ANDROID__
 #include <linux/sysctl.h>
-#else
-#include <sys/sysctl.h>
-#endif /* __ANDROID__ */
 #include <sys/socket.h>
 #include <sys/time.h>
 #include <sys/mman.h>
diff --git a/dispatch/source.h b/dispatch/source.h
index 6992d42..dd79bc8 100644
--- a/dispatch/source.h
+++ b/dispatch/source.h
@@ -32,7 +32,7 @@
 #endif
 
 #if !TARGET_OS_WIN32
-#include <sys/signal.h>
+#include <signal.h>
 #endif
 
 DISPATCH_ASSUME_NONNULL_BEGIN
