diff --git a/Makefile.am b/Makefile.am
index 63c8b17..6eb594d 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -18,8 +18,6 @@ endif
 
 SUBDIRS=						\
 	dispatch					\
-	$(MAYBE_PTHREAD_WORKQUEUES)	\
-	$(MAYBE_KQUEUES)			\
 	man							\
 	os							\
 	private						\
diff --git a/src/internal.h b/src/internal.h
index 8934b2c..a25756e 100644
--- a/src/internal.h
+++ b/src/internal.h
@@ -246,11 +246,7 @@ DISPATCH_EXPORT DISPATCH_NOTHROW void dispatch_atfork_child(void);
 #include <sys/event.h>
 #include <sys/mount.h>
 #include <sys/queue.h>
-#ifdef __ANDROID__
 #include <linux/sysctl.h>
-#else
-#include <sys/sysctl.h>
-#endif /* __ANDROID__ */
 #include <sys/socket.h>
 #include <sys/time.h>
 #include <sys/mman.h>
