--- a/chromium/chrome/browser/ui/libgtkui/BUILD.gn	2017-07-31 14:46:27.896012390 +0200
+++ b/chromium/chrome/browser/ui/libgtkui/BUILD.gn	2017-07-31 14:46:44.309012010 +0200
@@ -135,17 +135,3 @@
     "//chrome/browser:theme_properties",
   ]
 }
-
-component("libgtk3ui") {
-  sources = common_sources
-  configs += common_configs
-  defines = [ "LIBGTKUI_IMPLEMENTATION" ]
-
-  deps = common_deps + [
-           "//build/config/linux/gtk3",
-           "//build/config/linux/gtk3:gtkprint3",
-         ]
-  public_deps = [
-    "//chrome/browser:theme_properties",
-  ]
-}
--- a/chromium/BUILD.gn	2017-07-31 14:59:00.827994950 +0200
+++ b/chromium/BUILD.gn	2017-07-31 14:59:37.981994089 +0200
@@ -360,9 +360,12 @@
   }
 
   if (is_linux && !is_chromeos && !is_chromecast && !use_ozone) {
-    # TODO(thomasanderson): Remove this once we build using
-    # GTK3 by default. (crbug.com/132847, crbug.com/79722)
-    deps += [ "//chrome/browser/ui/libgtkui:libgtk3ui" ]
+    # These are the only components that can interact with gtk.
+    if (use_gtk3) {
+      deps += [ "//chrome/browser/ui/libgtkui:libgtk3ui" ]
+    } else {
+      deps += [ "//chrome/browser/ui/libgtkui:libgtk2ui" ]
+    }
   }
 
   if (use_ozone) {
