--- a/include/uapi/linux/if_ether.h	2015-06-23 15:04:38.000000000 +0000
+++ b/include/uapi/linux/if_ether.h	2015-08-20 11:28:37.875454420 +0000
@@ -134,11 +134,14 @@
  *	This is an Ethernet frame header.
  */
 
+#if !defined(HAVE_IF_ETHER)
+#define HAVE_IF_ETHER
 struct ethhdr {
 	unsigned char	h_dest[ETH_ALEN];	/* destination eth addr	*/
 	unsigned char	h_source[ETH_ALEN];	/* source ether addr	*/
 	__be16		h_proto;		/* packet type ID field	*/
 } __attribute__((packed));
+#endif
 

 #endif /* _UAPI_LINUX_IF_ETHER_H */
--- a/include/uapi/linux/if_arp.h	2015-06-23 15:04:38.000000000 +0000
+++ b/include/uapi/linux/if_arp.h	2015-08-20 11:37:56.962823172 +0000
@@ -110,6 +110,8 @@
 
 
 /* ARP ioctl request. */
+#if !defined(HAVE_ARP_H)
+#define HAVE_ARP_H
 struct arpreq {
   struct sockaddr	arp_pa;		/* protocol address		*/
   struct sockaddr	arp_ha;		/* hardware address		*/
@@ -154,8 +156,9 @@
 	unsigned char		ar_tha[ETH_ALEN];	/* target hardware address	*/
 	unsigned char		ar_tip[4];		/* target IP address		*/
 #endif
+#endif
 
 };
 
 
 #endif /* _UAPI_LINUX_IF_ARP_H */
