--- a/arch/x86/Makefile	2018-09-19 13:42:30.858708405 +0200
+++ b/arch/x86/Makefile	2018-09-19 13:43:08.648688299 +0200
@@ -180,10 +180,6 @@
   endif
 endif
 
-ifndef CC_HAVE_ASM_GOTO
-  $(error Compiler lacks asm-goto support.)
-endif
-
 #
 # Jump labels need '-maccumulate-outgoing-args' for gcc < 4.5.2 to prevent a
 # GCC bug (https://gcc.gnu.org/bugzilla/show_bug.cgi?id=46226).  There's no way
