--- a/libs/lensfun/cpuid.cpp	2017-01-04 20:01:01.051916410 +0100
+++ b/libs/lensfun/cpuid.cpp	2017-01-04 20:01:12.242882417 +0100
@@ -92,9 +92,9 @@
 {
 #define cpuid(cmd) \
     __asm volatile ( \
-        "push %%"R_BX"\n" \
+        "push %%" R_BX "\n" \
         "cpuid\n" \
-        "pop %%"R_BX"\n" \
+        "pop %%" R_BX "\n" \
        : "=a" (ax), "=c" (cx),  "=d" (dx) \
        : "0" (cmd))
 
