--- a/security/sandbox/linux/Sandbox.cpp	2025-12-12 16:40:29.829694287 +0100
+++ b/security/sandbox/linux/Sandbox.cpp	2025-12-12 16:41:53.005083641 +0100
@@ -775,6 +775,8 @@
   files->Add("/proc/net/unix", SandboxOpenedFile::Error{});
   files->Add("/proc/self/maps", SandboxOpenedFile::Error{});
   (void)dlopen("libgcc_s.so.1", RTLD_GLOBAL|RTLD_LAZY);
+  // add widevine_compat.so.0 for Widevine
+  files->Add("/lib/libwidevine_compat.so.1");
   // Finally, start the sandbox.
   SetCurrentProcessSandbox(GetMediaSandboxPolicy(files));
 }
