--- a/security/sandbox/linux/SandboxOpenedFiles.cpp
+++ b/security/sandbox/linux/SandboxOpenedFiles.cpp
@@ -11,6 +11,7 @@
 
 #include <fcntl.h>
 #include <unistd.h>
+#include <errno.h>
 
 namespace mozilla {
 
--- a/toolkit/components/remote/RemoteUtils.h
+++ b/toolkit/components/remote/RemoteUtils.h
@@ -6,6 +6,8 @@
 #ifndef RemoteUtils_h__
 #define RemoteUtils_h__
 
+#include <sys/types.h>
+
 char* ConstructCommandLine(int32_t argc, char** argv,
                            const char* aDesktopStartupID,
                            int* aCommandLineLength);
--- a/media/mtransport/third_party/nICEr/src/stun/addrs-netlink.c
+++ b/media/mtransport/third_party/nICEr/src/stun/addrs-netlink.c
@@ -31,6 +31,7 @@
 */
 
 #if defined(LINUX)
+#include <net/if.h>
 #include "addrs-netlink.h"
 #include <csi_platform.h>
 #include <assert.h>
@@ -46,7 +47,6 @@
 #include <sys/utsname.h>
 #include <sys/ioctl.h>
 #include <netinet/in.h>
-#include <net/if.h>
 #include <unistd.h>
 #include <errno.h>
 #include <linux/netlink.h>
