--- a/security/sandbox/linux/SandboxOpenedFiles.cpp	2018-05-18 15:53:29.737816486 +0200
+++ b/security/sandbox/linux/SandboxOpenedFiles.cpp	2018-05-18 15:53:40.855793940 +0200
@@ -11,6 +11,7 @@
 
 #include <fcntl.h>
 #include <unistd.h>
+#include <errno.h>
 
 namespace mozilla {
 
--- a/widget/xremoteclient/RemoteUtils.h	2018-05-18 15:54:36.097679927 +0200
+++ b/widget/xremoteclient/RemoteUtils.h	2018-05-18 15:54:45.943659952 +0200
@@ -6,9 +6,11 @@
 #ifndef RemoteUtils_h__
 #define RemoteUtils_h__
 
+#include <sys/types.h>
+
 char*
 ConstructCommandLine(int32_t argc, char **argv,
                      const char* aDesktopStartupID,
                      int *aCommandLineLength);
 
-#endif // RemoteUtils_h__
\ No newline at end of file
+#endif // RemoteUtils_h__
