--- a/security/sandbox/linux/SandboxOpenedFiles.cpp
+++ b/security/sandbox/linux/SandboxOpenedFiles.cpp
@@ -11,6 +11,7 @@
 
 #include <fcntl.h>
 #include <unistd.h>
+#include <errno.h>
 
 namespace mozilla {
 
--- a/widget/xremoteclient/RemoteUtils.h
+++ b/widget/xremoteclient/RemoteUtils.h
@@ -6,6 +6,8 @@
 #ifndef RemoteUtils_h__
 #define RemoteUtils_h__
 
+#include <sys/types.h>
+
 char* ConstructCommandLine(int32_t argc, char** argv,
                            const char* aDesktopStartupID,
                            int* aCommandLineLength);
--- a/toolkit/mozapps/update/common/updatedefines.h
+++ b/toolkit/mozapps/update/common/updatedefines.h
@@ -102,7 +102,7 @@
 
 #ifdef SOLARIS
 #include <sys/stat.h>
-#else
+#elif !defined(__linux__) || defined(__GLIBC__)
 #include <fts.h>
 #endif
 #include <dirent.h>
