--- a/python/mozbuild/mozbuild/action/check_binary.py	2021-04-02 13:00:37.710480674 +0200
+++ b/python/mozbuild/mozbuild/action/check_binary.py	2021-04-02 13:00:50.538625272 +0200
@@ -36,7 +36,7 @@
 TARGET = {
     "MOZ_LIBSTDCXX_VERSION": buildconfig.substs.get("MOZ_LIBSTDCXX_TARGET_VERSION"),
     "platform": buildconfig.substs["OS_TARGET"],
-    "readelf": "{}readelf".format(buildconfig.substs.get("TOOLCHAIN_PREFIX", "")),
+    "readelf": "readelf"
 }
 
 ADDR_RE = re.compile(r"[0-9a-f]{8,16}")
--- a/toolkit/library/build/dependentlibs.py	2021-04-02 16:46:27.677201295 +0200
+++ b/toolkit/library/build/dependentlibs.py	2021-04-02 16:47:13.512655697 +0200
@@ -44,7 +44,7 @@
 def dependentlibs_readelf(lib):
     """Returns the list of dependencies declared in the given ELF .so"""
     proc = subprocess.Popen(
-        [substs.get("TOOLCHAIN_PREFIX", "") + "readelf", "-d", lib],
+        ["readelf", "-d", lib],
         stdout=subprocess.PIPE,
         universal_newlines=True,
     )
