--- a/vendor/adb/sysdeps.h	2022-11-07 10:11:09.388101610 +0100
+++ b/vendor/adb/sysdeps.h	2022-11-07 09:55:47.171356711 +0100
@@ -522,7 +522,7 @@
 }
 
 static inline int adb_pread(borrowed_fd fd, void* buf, size_t len, off64_t offset) {
-#if defined(__APPLE__)
+#if defined(__APPLE__) || !defined(__GLIBC__) && defined(__linux__)
     return TEMP_FAILURE_RETRY(pread(fd.get(), buf, len, offset));
 #else
     return TEMP_FAILURE_RETRY(pread64(fd.get(), buf, len, offset));
@@ -544,7 +544,7 @@
 }
 
 static inline int adb_pwrite(int fd, const void* buf, size_t len, off64_t offset) {
-#if defined(__APPLE__)
+#if defined(__APPLE__) || !defined(__GLIBC__) && defined(__linux__)
     return TEMP_FAILURE_RETRY(pwrite(fd, buf, len, offset));
 #else
     return TEMP_FAILURE_RETRY(pwrite64(fd, buf, len, offset));
@@ -557,7 +557,7 @@
 #define pwrite ___xxx_pwrite
 
 static inline int64_t adb_lseek(borrowed_fd fd, int64_t pos, int where) {
-#if defined(__APPLE__)
+#if defined(__APPLE__) || !defined(__GLIBC__) && defined(__linux__)
     return lseek(fd.get(), pos, where);
 #else
     return lseek64(fd.get(), pos, where);
--- a/vendor/core/libutils/include/utils/Compat.h	2022-11-07 10:12:32.828132463 +0100
+++ b/vendor/core/libutils/include/utils/Compat.h	2022-11-07 09:53:25.863437916 +0100
@@ -23,7 +23,7 @@
 #include <sys/mman.h>
 #endif
 
-#if defined(__APPLE__)
+#if defined(__APPLE__) || !defined(__GLIBC__) && defined(__linux__)
 
 /* Mac OS has always had a 64-bit off_t, so it doesn't have off64_t. */
 static_assert(sizeof(off_t) >= 8, "This code requires that Mac OS have at least a 64-bit off_t.");
--- a/vendor/libbase/include/android-base/off64_t.h	2022-11-07 09:46:22.888228861 +0100
+++ b/vendor/libbase/include/android-base/off64_t.h	2022-11-07 09:46:54.439725726 +0100
@@ -16,7 +16,7 @@
 
 #pragma once
 
-#if defined(__APPLE__)
+#if defined(__APPLE__) || !defined(__GLIBC__) && defined(__linux__)
 /** Mac OS has always had a 64-bit off_t, so it doesn't have off64_t. */
 typedef off_t off64_t;
 #endif
--- a/vendor/libziparchive/zip_archive.cc	2022-11-07 10:15:06.062019700 +0100
+++ b/vendor/libziparchive/zip_archive.cc	2022-11-07 10:15:15.166131627 +0100
@@ -36,7 +36,7 @@
 #include <span>
 #include <vector>
 
-#if defined(__APPLE__)
+#if defined(__APPLE__) || !defined(__GLIBC__) && defined(__linux__)
 #define lseek64 lseek
 #endif
 
