diff --git a/source/app.d b/source/app.d
index 01a1625..c70ae76 100644
--- a/source/app.d
+++ b/source/app.d
@@ -19,8 +19,8 @@ import std.conv;
 
 import source.info;
 
-__gshared File stdin, stdout;
-shared static this() {
+File stdin, stdout;
+static this() {
 	stdin = std.stdio.stdin;
 	stdout = std.stdio.stdout;
 	std.stdio.stdin = File("/dev/null", "r");
@@ -340,9 +340,6 @@ int main(string[] args)
 	}
 	else
 	{
-		static if (is(typeof(registerMemoryErrorHandler)))
-			registerMemoryErrorHandler();
-
 		writeMutex = new Mutex;
 		commandMutex = new Mutex;
 
