diff --git a/src/qwaylandinputpanelsurface.cpp b/src/qwaylandinputpanelsurface.cpp
index 200b4b84a02f2f3ce9f25a7de6d2285f9b0b4ca3..e08d0b1d4ec9ab5b6ab6e61ea8d04f09ae69dcc8 100644
--- a/src/qwaylandinputpanelsurface.cpp
+++ b/src/qwaylandinputpanelsurface.cpp
@@ -47,11 +47,10 @@ void QWaylandInputPanelSurface::applyConfigure()
         }
 
         set_toplevel(screen->output(), position_center_bottom);
+        window()->display()->handleWindowActivated(window());
         break;
     }
     }
-
-    window()->display()->handleWindowActivated(window());
 }
 
 QT_END_NAMESPACE
