--- a/document-portal/document-portal.c	2021-01-19 23:19:52.612470980 +0100
+++ b/document-portal/document-portal.c	2021-01-19 23:19:58.853555514 +0100
@@ -1639,7 +1639,6 @@
 
   if (set_one_signal_handler (SIGHUP, exit_handler, 0) == -1 ||
       set_one_signal_handler (SIGINT, exit_handler, 0) == -1 ||
-      set_one_signal_handler (SIGTERM, exit_handler, 0) == -1 ||
       set_one_signal_handler (SIGPIPE, SIG_IGN, 0) == -1)
     exit (5);
 
