diff --git a/backend/src/backend/gen_insn_selection_optimize.cpp b/backend/src/backend/gen_insn_selection_optimize.cpp
index 2ab2a7f9..8775beb6 100644
--- a/backend/src/backend/gen_insn_selection_optimize.cpp
+++ b/backend/src/backend/gen_insn_selection_optimize.cpp
@@ -457,19 +457,19 @@ namespace gbe
             }
             else if(src0.type == GEN_TYPE_UD || src1.type == GEN_TYPE_UD)
             {
-              unsigned int s0 = src0.value.ud;
+              int s0 = src0.value.ud;
               if (src0.absolute)
                 s0 = abs(s0);
               if (src0.negation)
                 s0 = -s0;
 
-              unsigned int s1 = src1.value.ud;
+              int s1 = src1.value.ud;
               if (src1.absolute)
                 s1 = abs(s1);
               if (src1.negation)
                 s1 = -s1;
 
-              unsigned int result;
+              int result;
               RESULT();
               insn->src(0) = GenRegister::immud(result);
             }
