--- a/eng/common/dotnet-install.sh	2020-05-20 17:14:59.545961704 +0200
+++ b/eng/common/dotnet-install.sh	2020-05-20 17:15:13.698070860 +0200
@@ -49,10 +49,10 @@
 
 . "$scriptroot/tools.sh"
 dotnetRoot="$repo_root/.dotnet"
-InstallDotNet $dotnetRoot $version "$architecture" $runtime true $runtimeSourceFeed $runtimeSourceFeedKey || {
-  local exit_code=$?
-  echo "dotnet-install.sh failed (exit code '$exit_code')." >&2
-  ExitWithExitCode $exit_code
-}
+#InstallDotNet $dotnetRoot $version "$architecture" $runtime true $runtimeSourceFeed $runtimeSourceFeedKey || {
+#  local exit_code=$?
+#  echo "dotnet-install.sh failed (exit code '$exit_code')." >&2
+#  ExitWithExitCode $exit_code
+#}
 
 ExitWithExitCode 0
