--- a/src/shared/user-record-nss.c	2021-07-09 23:05:57.795285798 +0200
+++ b/src/shared/user-record-nss.c	2021-07-09 22:34:29.547591543 +0200
@@ -317,6 +317,7 @@
         return 0;
 }
 
+#if ENABLE_GSHADOW
 int nss_group_to_group_record(
                 const struct group *grp,
                 const struct sgrp *sgrp,
@@ -529,3 +530,4 @@
         (*ret)->incomplete = incomplete;
         return 0;
 }
+#endif
--- a/src/shared/user-record-nss.h	2021-07-09 23:06:02.173249153 +0200
+++ b/src/shared/user-record-nss.h	2021-07-09 23:06:40.536928109 +0200
@@ -2,7 +2,9 @@
 #pragma once
 
 #include <grp.h>
+#if ENABLE_GSHADOW
 #include <gshadow.h>
+#endif
 #include <pwd.h>
 #include <shadow.h>
 
@@ -17,8 +19,10 @@
 int nss_user_record_by_name(const char *name, bool with_shadow, UserRecord **ret);
 int nss_user_record_by_uid(uid_t uid, bool with_shadow, UserRecord **ret);
 
+#if ENABLE_GSHADOW
 int nss_group_to_group_record(const struct group *grp, const struct sgrp *sgrp, GroupRecord **ret);
 int nss_sgrp_for_group(const struct group *grp, struct sgrp *ret_sgrp, char **ret_buffer);
 
 int nss_group_record_by_name(const char *name, bool with_shadow, GroupRecord **ret);
 int nss_group_record_by_gid(gid_t gid, bool with_shadow, GroupRecord **ret);
+#endif
--- a/src/shared/userdb.c	2021-07-09 23:05:11.132676446 +0200
+++ b/src/shared/userdb.c	2021-07-09 22:34:06.746779085 +0200
@@ -922,6 +922,7 @@
         }
 
 
+#if ENABLE_GSHADOW
         if (!FLAGS_SET(flags, USERDB_EXCLUDE_NSS) && !(iterator && iterator->nss_covered)) {
                 r = userdb_iterator_block_nss_systemd(iterator);
                 if (r >= 0) {
@@ -930,6 +931,7 @@
                                 return r;
                 }
         }
+#endif
 
         if (!FLAGS_SET(flags, USERDB_DONT_SYNTHESIZE)) {
                 if (streq(name, "root"))
@@ -972,6 +974,7 @@
                         return r;
         }
 
+#if ENABLE_GSHADOW
         if (!FLAGS_SET(flags, USERDB_EXCLUDE_NSS) && !(iterator && iterator->nss_covered)) {
                 r = userdb_iterator_block_nss_systemd(iterator);
                 if (r >= 0) {
@@ -980,6 +983,7 @@
                                 return r;
                 }
         }
+#endif
 
         if (!FLAGS_SET(flags, USERDB_DONT_SYNTHESIZE)) {
                 if (gid == 0)
@@ -1033,6 +1037,7 @@
         return 0;
 }
 
+#if ENABLE_GSHADOW
 int groupdb_iterator_get(UserDBIterator *iterator, GroupRecord **ret) {
         int r;
 
@@ -1134,6 +1139,7 @@
 
         return r;
 }
+#endif
 
 static void discover_membership_dropins(UserDBIterator *i, UserDBFlags flags) {
         int r;
@@ -1219,6 +1225,7 @@
 
         qr = userdb_start_query(iterator, "io.systemd.UserDatabase.GetMemberships", true, query, flags);
 
+#if ENABLE_GSHADOW
         if (!FLAGS_SET(flags, USERDB_EXCLUDE_NSS) && (qr < 0 || !iterator->nss_covered)) {
                 _cleanup_(group_record_unrefp) GroupRecord *gr = NULL;
 
@@ -1240,6 +1247,7 @@
                                 return -ENOMEM;
                 }
         }
+#endif
 
         if (!FLAGS_SET(flags, USERDB_EXCLUDE_DROPIN) && (qr < 0 || !iterator->dropin_covered))
                 discover_membership_dropins(iterator, flags);
