--- a/src/portable/portable.c	2022-05-21 15:23:41.000000000 +0200
+++ b/src/portable/portable.c	2022-05-22 20:45:04.363500648 +0200
@@ -1,5 +1,6 @@
 /* SPDX-License-Identifier: LGPL-2.1-or-later */
 
+#include <sys/file.h>
 #include <linux/loop.h>
 
 #include "bus-common-errors.h" 
--- a/src/test/test-fs-util.c	2023-04-16 15:43:18.875169117 +0200
+++ b/src/test/test-fs-util.c	2023-04-16 15:42:11.884275419 +0200
@@ -1,6 +1,7 @@
 /* SPDX-License-Identifier: LGPL-2.1-or-later */
 
 #include <unistd.h>
+#include <sys/file.h>
 
 #include "alloc-util.h"
 #include "copy.h"
--- a/src/basic/missing_fs.h	2023-05-03 23:13:50.104931727 +0200
+++ b/src/basic/missing_fs.h	2023-05-03 23:15:43.374638653 +0200
@@ -1,6 +1,8 @@
 /* SPDX-License-Identifier: LGPL-2.1-or-later */
 #pragma once
 
+#include <linux/types.h>
+
 /* linux/fs.h */
 #ifndef RENAME_NOREPLACE /* 0a7c3937a1f23f8cb5fc77ae01661e9968a51d0c (3.15) */
 #define RENAME_NOREPLACE (1 << 0)
--- src/basic/lock-util.h	2023-06-22 22:55:20.154391168 +0200
+++ src/basic/lock-util.h	2023-06-22 22:55:49.865811861 +0200
@@ -2,6 +2,7 @@
 #pragma once
 
 #include <fcntl.h>
+#include <sys/file.h>
 
 typedef struct LockFile {
         int dir_fd;
