--- a/src/portable/portabled-operation.c
+++ b/src/portable/portabled-operation.c
@@ -1,5 +1,6 @@
 /* SPDX-License-Identifier: LGPL-2.1+ */
 
+#include <sys/wait.h>
 #include "alloc-util.h"
 #include "fd-util.h"
 #include "portabled-operation.h"
--- a/src/udev/udev-event.c
+++ b/src/udev/udev-event.c
@@ -7,6 +7,7 @@
 #include <stddef.h>
 #include <stdio.h>
 #include <stdlib.h>
+#include <sys/wait.h>
 #include <unistd.h>
 
 #include "sd-event.h"
--- a/src/cgroups-agent/cgroups-agent.c
+++ b/src/cgroups-agent/cgroups-agent.c
@@ -1,6 +1,7 @@
 /* SPDX-License-Identifier: LGPL-2.1+ */
 
 #include <stdlib.h>
+#include <string.h>
 #include <sys/socket.h>
 
 #include "fd-util.h"
