--- a/src/libsystemd-network/sd-dhcp6-client.c	2020-02-06 16:17:25.080003621 +0100
+++ b/src/libsystemd-network/sd-dhcp6-client.c	2020-02-06 16:18:31.708357365 +0100
@@ -5,7 +5,6 @@
 
 #include <errno.h>
 #include <sys/ioctl.h>
-#include <linux/if_arp.h>
 #include <linux/if_infiniband.h>
 
 #include "sd-dhcp6-client.h"
--- a/src/network/networkd-dhcp4.c	2020-02-06 16:17:35.153057162 +0100
+++ b/src/network/networkd-dhcp4.c	2020-02-06 16:18:11.515250248 +0100
@@ -2,7 +2,6 @@
 
 #include <netinet/in.h>
 #include <linux/if.h>
-#include <linux/if_arp.h>
 
 #include "alloc-util.h"
 #include "dhcp-client-internal.h"
--- a/src/network/networkd-dhcp6.c	2020-02-06 16:17:44.522106940 +0100
+++ b/src/network/networkd-dhcp6.c	2020-02-06 16:18:19.435292269 +0100
@@ -5,7 +5,6 @@
 
 #include <netinet/in.h>
 #include <linux/if.h>
-#include <linux/if_arp.h>
 
 #include "sd-dhcp6-client.h"
 
--- a/src/network/networkd-link.c	2020-02-06 16:17:52.978151851 +0100
+++ b/src/network/networkd-link.c	2020-02-06 16:17:57.922178101 +0100
@@ -2,7 +2,6 @@
 
 #include <netinet/in.h>
 #include <linux/if.h>
-#include <linux/if_arp.h>
 #include <linux/if_link.h>
 #include <unistd.h>
 
