--- a/src/libsystemd-network/sd-dhcp6-client.c	2020-02-06 16:17:25.080003621 +0100
+++ b/src/libsystemd-network/sd-dhcp6-client.c	2020-02-06 16:18:31.708357365 +0100
@@ -5,7 +5,6 @@
 
 #include <errno.h>
 #include <sys/ioctl.h>
-#include <linux/if_arp.h>
 #include <linux/if_infiniband.h>
 
 #include "sd-dhcp6-client.h"
--- a/src/network/networkd-dhcp4.c	2020-02-06 16:17:35.153057162 +0100
+++ b/src/network/networkd-dhcp4.c	2020-02-06 16:18:11.515250248 +0100
@@ -2,7 +2,6 @@
 
 #include <netinet/in.h>
 #include <linux/if.h>
-#include <linux/if_arp.h>
 
 #include "alloc-util.h"
 #include "dhcp-client-internal.h"
--- a/src/network/networkd-dhcp6.c	2020-02-06 16:17:44.522106940 +0100
+++ b/src/network/networkd-dhcp6.c	2020-02-06 16:18:19.435292269 +0100
@@ -5,7 +5,6 @@
 
 #include <netinet/in.h>
 #include <linux/if.h>
-#include <linux/if_arp.h>
 
 #include "sd-dhcp6-client.h"
 
--- a/src/network/networkd-link.c	2020-02-06 16:17:52.978151851 +0100
+++ b/src/network/networkd-link.c	2020-02-06 16:17:57.922178101 +0100
@@ -2,7 +2,6 @@
 
 #include <netinet/in.h>
 #include <linux/if.h>
-#include <linux/if_arp.h>
 #include <linux/if_link.h>
 #include <unistd.h>
 
--- a/src/network/networkd-dhcp-common.c	2020-11-26 22:38:21.065742239 +0100
+++ b/src/network/networkd-dhcp-common.c	2020-11-26 22:40:36.656842867 +0100
@@ -1,7 +1,7 @@
 /* SPDX-License-Identifier: LGPL-2.1-or-later */
 
 #include <netinet/in.h>
-#include <linux/if_arp.h>
+#include <net/if.h>
 
 #include "dhcp-internal.h"
 #include "dhcp6-internal.h"
--- a/src/network/networkd-dhcp-server.c	2020-11-26 22:39:00.812064588 +0100
+++ b/src/network/networkd-dhcp-server.c	2020-11-26 22:39:05.615103556 +0100
@@ -1,7 +1,6 @@
 /* SPDX-License-Identifier: LGPL-2.1-or-later */
 
 #include <netinet/in.h>
-#include <linux/if_arp.h>
 #include <linux/if.h>
 
 #include "sd-dhcp-server.h"
--- a/src/network/networkd-route.c	2020-11-26 22:35:53.784550176 +0100
+++ b/src/network/networkd-route.c	2020-11-26 22:37:52.589511442 +0100
@@ -1,5 +1,6 @@
 /* SPDX-License-Identifier: LGPL-2.1-or-later */
 
+#include <netinet/in.h>
 #include <linux/icmpv6.h>
 #include <linux/ipv6_route.h>
 
--- a/src/home/homed-home.c	2021-03-10 16:56:00.043792752 +0100
+++ b/src/home/homed-home.c	2021-03-10 16:56:14.243930237 +0100
@@ -5,6 +5,7 @@
 #endif
 
 #include <sys/mman.h>
+#include <linux/quota.h>
 #include <sys/quota.h>
 #include <sys/vfs.h>
 
--- a/src/home/homework-quota.c	2021-03-10 16:55:35.409553655 +0100
+++ b/src/home/homework-quota.c	2021-03-10 16:56:32.753109080 +0100
@@ -1,5 +1,6 @@
 /* SPDX-License-Identifier: LGPL-2.1-or-later */
 #include <sys/quota.h>
+#include <linux/quota.h>
 
 #include "blockdev-util.h"
 #include "btrfs-util.h"
