--- a/src/test/test-recurse-dir.c	2021-12-24 13:37:35.162657348 +0100
+++ b/src/test/test-recurse-dir.c	2021-12-24 13:37:48.396808391 +0100
@@ -16,7 +16,7 @@
                 struct FTW *ftwbuf) {
 
         if (ftwbuf->level == 0) /* skip top-level */
-                return FTW_CONTINUE;
+                return 0;
 
         switch (typeflag) {
 
@@ -49,7 +49,7 @@
                 assert_not_reached();
         }
 
-        return FTW_CONTINUE;
+        return 0;
 }
 
 static int recurse_dir_callback(
