--- a/src/systemctl/systemctl-edit.c	2024-06-30 11:45:32.468278548 +0200
+++ b/src/systemctl/systemctl-edit.c	2024-06-30 11:45:48.838287381 +0200
@@ -316,7 +316,7 @@
                 .marker_end = DROPIN_MARKER_END,
                 .remove_parent = !arg_full,
                 .overwrite_with_origin = true,
-                .stdin = arg_stdin,
+                .from_stdin = arg_stdin,
         };
         _cleanup_strv_free_ char **names = NULL;
         sd_bus *bus;
--- a/src/shared/edit-util.c	2024-06-30 11:43:52.478223664 +0200
+++ b/src/shared/edit-util.c	2024-06-30 11:44:52.364923405 +0200
@@ -212,7 +212,7 @@
         if (fchmod(fileno(f), 0644) < 0)
                 return log_error_errno(errno, "Failed to change mode of temporary file '%s': %m", temp);
 
-        if (e->context->stdin) {
+        if (e->context->from_stdin) {
                 if (fwrite(contents, 1, contents_size, f) != contents_size)
                         return log_error_errno(SYNTHETIC_ERRNO(EIO),
                                                "Failed to copy input to temporary file '%s'.", temp);
@@ -326,7 +326,7 @@
         if (!tmp)
                 return log_oom();
 
-        if (e->context->marker_start && !e->context->stdin) {
+        if (e->context->marker_start && !e->context->from_stdin) {
                 /* Trim out the lines between the two markers */
                 char *contents_start, *contents_end;
 
@@ -374,7 +374,7 @@
         if (context->n_files == 0)
                 return log_debug_errno(SYNTHETIC_ERRNO(ENOENT), "Got no files to edit.");
 
-        if (context->stdin) {
+        if (context->from_stdin) {
                 r = read_full_stream(stdin, &data, &data_size);
                 if (r < 0)
                         return log_error_errno(r, "Failed to read stdin: %m");
@@ -386,7 +386,7 @@
                         return r;
         }
 
-        if (!context->stdin) {
+        if (!context->from_stdin) {
                 r = run_editor(context);
                 if (r < 0)
                         return r;
--- a/src/shared/edit-util.h	2024-06-30 11:42:57.811526217 +0200
+++ b/src/shared/edit-util.h	2024-06-30 11:42:52.594856581 +0200
@@ -15,7 +15,7 @@
         const char *marker_end;
         bool remove_parent;
         bool overwrite_with_origin; /* Always overwrite target with original file. */
-        bool stdin;                 /* Read contents from stdin instead of launching an editor. */
+        bool from_stdin;                 /* Read contents from stdin instead of launching an editor. */
 } EditFileContext;
 
 void edit_file_context_done(EditFileContext *context);
