--- a/src/basic/copy.c
+++ b/src/basic/copy.c
@@ -3,6 +3,7 @@
 #include <dirent.h>
 #include <errno.h>
 #include <fcntl.h>
+#include <signal.h>
 #include <stddef.h>
 #include <stdio.h>
 #include <stdlib.h>
--- a/src/portable/portabled-operation.c
+++ b/src/portable/portabled-operation.c
@@ -1,5 +1,7 @@
 /* SPDX-License-Identifier: LGPL-2.1+ */
 
+#include <sys/wait.h>
+
 #include "alloc-util.h"
 #include "fd-util.h"
 #include "portabled-operation.h"
--- a/src/coredump/coredump-vacuum.c
+++ b/src/coredump/coredump-vacuum.c
@@ -16,6 +16,7 @@
 #include "string-util.h"
 #include "time-util.h"
 #include "user-util.h"
+#include "missing.h"
 
 #define DEFAULT_MAX_USE_LOWER (uint64_t) (1ULL*1024ULL*1024ULL)           /* 1 MiB */
 #define DEFAULT_MAX_USE_UPPER (uint64_t) (4ULL*1024ULL*1024ULL*1024ULL)   /* 4 GiB */
--- a/src/resolve/resolvectl.c
+++ b/src/resolve/resolvectl.c
@@ -30,6 +30,7 @@
 #include "strv.h"
 #include "terminal-util.h"
 #include "verbs.h"
+#include "missing.h"
 
 static int arg_family = AF_UNSPEC;
 static int arg_ifindex = 0;
