diff --git a/src/shared/utmp-wtmp.h b/src/shared/utmp-wtmp.h
index 36e4203b4f7d..188d011fdd71 100644
--- a/src/shared/utmp-wtmp.h
+++ b/src/shared/utmp-wtmp.h
@@ -59,7 +59,7 @@ static inline int utmp_wall(
                 const char *message,
                 const char *username,
                 const char *origin_tty,
-                bool (*match_tty)(const char *tty, void *userdata),
+                bool (*match_tty)(const char *tty, bool is_local, void *userdata),
                 void *userdata) {
         return 0;
 }
