--- a/meson.build	2020-12-03 23:05:06.929350504 +0100
+++ b/meson.build	2020-12-03 23:05:21.029463431 +0100
@@ -516,7 +516,7 @@
                                  #include <sys/stat.h>
                                  #include <unistd.h>'''],
         ['explicit_bzero' ,   '''#include <string.h>'''],
-        ['reallocarray',      '''#include <malloc.h>'''],
+        ['reallocarray',      '''#include <stdlib.h>'''],
         ['set_mempolicy',     '''#include <stdlib.h>
                                  #include <unistd.h>'''],
         ['get_mempolicy',     '''#include <stdlib.h>
