--- a/src/basic/copy.c
+++ b/src/basic/copy.c
@@ -3,6 +3,7 @@
 #include <errno.h>
 #include <fcntl.h>
 #include <stddef.h>
+#include <signal.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <sys/sendfile.h>
--- a/src/portable/portabled-operation.c
+++ b/src/portable/portabled-operation.c
@@ -1,5 +1,7 @@
 /* SPDX-License-Identifier: LGPL-2.1+ */
 
+#include <sys/wait.h>
+
 #include "alloc-util.h"
 #include "fd-util.h"
 #include "portabled-operation.h"
--- a/src/login/logind-brightness.c
+++ b/src/login/logind-brightness.c
@@ -1,5 +1,7 @@
 /* SPDX-License-Identifier: LGPL-2.1+ */
 
+#include <sys/wait.h>
+
 #include "bus-util.h"
 #include "device-util.h"
 #include "hash-funcs.h"
--- a/src/partition/makefs.c
+++ b/src/partition/makefs.c
@@ -3,6 +3,7 @@
 #include <fcntl.h>
 #include <sys/prctl.h>
 #include <sys/stat.h>
+#include <sys/file.h>
 #include <sys/types.h>
 #include <unistd.h>
 
