--- a/src/basic/build-path.c	2024-06-30 15:10:44.714414216 +0200
+++ b/src/basic/build-path.c	2024-06-30 15:12:03.259993617 +0200
@@ -33,11 +33,11 @@
                         break;
 
                 case DT_STRTAB:
-                        /* On MIPS and RISC-V DT_STRTAB records an offset, not a valid address, so it has to be adjusted
+                        /* On MIPS, RISC-V, and musl DT_STRTAB records an offset, not a valid address, so it has to be adjusted
                          * using the bias calculated earlier. */
                         if (d->d_un.d_val != 0)
                                 strtab = (const char *) ((uintptr_t) d->d_un.d_val
-#if defined(__mips__) || defined(__riscv)
+#if defined(__mips__) || defined(__riscv) || !defined(__GLIBC__)
                                          + bias
 #endif
                                 );
