--- a/src/boot/meson.build	2023-04-16 15:43:18.647166096 +0200
+++ b/src/boot/meson.build	2023-04-16 15:42:11.762273780 +0200
@@ -130,6 +130,7 @@
 efi_c_args = [
         '-DSD_BOOT=1',
         '-ffreestanding',
+        '-nostdlibinc',
         '-fno-strict-aliasing',
         '-fshort-wchar',
         '-include', 'efi_config.h',
@@ -150,7 +151,6 @@
 endif
 
 efi_c_ld_args = [
-        '-lgcc',
         '-nostdlib',
         '-static-pie',
         '-Wl,--entry=efi_main',
