--- a/src/fundamental/string-util-fundamental.h	2023-11-29 20:39:06.800401041 +0100
+++ b/src/fundamental/string-util-fundamental.h	2023-11-29 20:39:31.892048867 +0100
@@ -115,3 +115,7 @@
         /* A pure ASCII, locale independent version of isalpha() */
         return (a >= 'a' && a <= 'z') || (a >= 'A' && a <= 'Z');
 }
+
+#if !defined(__GLIBC__)
+#define basename(src) (strrchr(src,'/') ? strrchr(src,'/')+1 : src)
+#endif
