--- a/src/basic/fileio.c	2020-02-06 17:32:07.088701329 +0100
+++ b/src/basic/fileio.c	2020-02-06 17:33:11.246085640 +0100
@@ -205,6 +205,11 @@
                 goto fail;
         }
 
+#ifndef __GLIBC__ /// systemd must not disable buffers on musl-libc based systems when going this route
+        if (flags & WRITE_STRING_FILE_DISABLE_BUFFER)
+            flags ^= WRITE_STRING_FILE_DISABLE_BUFFER;
+#endif // __GLIBC__
+
         r = fdopen_unlocked(fd, "w", &f);
         if (r < 0) {
                 safe_close(fd);
