--- a/src/basic/fileio.c	2020-02-06 15:38:48.821384735 +0100
+++ b/src/basic/fileio.c	2020-02-06 15:41:52.533276606 +0100
@@ -210,6 +210,10 @@
                         safe_close(fd);
                         goto fail;
                 }
+#ifndef __GLIBC__ /// systemd must not disable buffers on musl-libc based systems when going this route
+                if (flags & WRITE_STRING_FILE_DISABLE_BUFFER)
+                        flags ^= WRITE_STRING_FILE_DISABLE_BUFFER;
+#endif // __GLIBC__
         }
 
         if (flags & WRITE_STRING_FILE_DISABLE_BUFFER)
