--- a/src/test/test-json.c
+++ b/src/test/test-json.c
@@ -11,6 +11,10 @@
 #include "tests.h"
 #include "util.h"
 
+#ifndef M_PIl
+#define M_PIl 3.141592653589793238462643383279502884L
+#endif
+
 static void test_tokenizer(const char *data, ...) {
         unsigned line = 0, column = 0;
         void *state = NULL;
--- a/src/test/test-cpu-set-util.c
+++ b/src/test/test-cpu-set-util.c
@@ -5,6 +5,10 @@
 #include "string-util.h"
 #include "macro.h"
 
+#ifndef __cpu_mask
+#define __cpu_mask unsigned long
+#endif
+
 static void test_parse_cpu_set(void) {
         CPUSet c = {};
         _cleanup_free_ char *str = NULL;
--- a/src/test/test-sizeof.c
+++ b/src/test/test-sizeof.c
@@ -9,6 +9,10 @@
 
 #include "time-util.h"
 
+#ifndef __cpu_mask
+#define __cpu_mask unsigned long
+#endif
+
 /* Print information about various types. Useful when diagnosing
  * gcc diagnostics on an unfamiliar architecture. */
 
