diff --git a/meson.build b/meson.build
index c8adc85089006..58ced9d4cbac3 100644
--- a/meson.build
+++ b/meson.build
@@ -72,7 +72,10 @@ conf.set10('SD_BOOT', false)
 
 # Create a title-less summary section early, so it ends up first in the output.
 # More items are added later after they have been detected.
-summary({'build mode' : get_option('mode')})
+summary({
+        'libc'       : get_option('libc'),
+        'build mode' : get_option('mode'),
+})
 
 #####################################################################
 
@@ -2045,7 +2048,14 @@ dbus_programs = []
 boot_stubs = []
 
 # This is similar to system_includes below, but for passing custom_target().
-system_include_args = [
+system_include_args = []
+if get_option('libc') == 'musl'
+        system_include_args += [
+                '-isystem', meson.project_build_root()  / 'src/include/musl',
+                '-isystem', meson.project_source_root() / 'src/include/musl',
+        ]
+endif
+system_include_args += [
         '-isystem', meson.project_build_root()  / 'src/include/override',
         '-isystem', meson.project_source_root() / 'src/include/override',
         '-isystem', meson.project_build_root()  / 'src/include/uapi',
@@ -2063,6 +2073,9 @@ system_includes = [
                 is_system : true,
         ),
 ]
+if get_option('libc') == 'musl'
+        system_includes += include_directories('src/include/musl', is_system : true)
+endif
 
 basic_includes = [
         include_directories(
diff --git a/meson_options.txt b/meson_options.txt
index d8dec33ec42e7..ad203ba301234 100644
--- a/meson_options.txt
+++ b/meson_options.txt
@@ -397,6 +397,8 @@ option('ima', type : 'boolean',
 option('ipe', type : 'boolean',
        description : 'IPE support')
 
+option('libc', type : 'combo', choices : ['glibc', 'musl'],
+       description : 'libc implementation to be used')
 option('acl', type : 'feature', deprecated : { 'true' : 'enabled', 'false' : 'disabled' },
        description : 'libacl support')
 option('audit', type : 'feature', deprecated : { 'true' : 'enabled', 'false' : 'disabled' },
diff --git a/src/libc/meson.build b/src/libc/meson.build
index eeee98c9d6c6d..306512ffd7002 100644
--- a/src/libc/meson.build
+++ b/src/libc/meson.build
@@ -16,6 +16,8 @@ libc_wrapper_sources = files(
         'xattr.c',
 )
 
+subdir('musl')
+
 sources += libc_wrapper_sources
 
 libc_wrapper_static = static_library(
diff --git a/src/libc/musl/meson.build b/src/libc/musl/meson.build
new file mode 100644
index 0000000000000..a876230c67f87
--- /dev/null
+++ b/src/libc/musl/meson.build
@@ -0,0 +1,5 @@
+# SPDX-License-Identifier: LGPL-2.1-or-later
+
+if get_option('libc') != 'musl'
+        subdir_done()
+endif
