diff --git a/src/boot/meson.build b/src/boot/meson.build
index ba0b309acff65..bea96bb2e3499 100644
--- a/src/boot/meson.build
+++ b/src/boot/meson.build
@@ -179,6 +179,11 @@ if get_option('mode') == 'developer' and get_option('debug')
         efi_c_args += '-DEFI_DEBUG'
 endif
 
+if get_option('libc') == 'musl'
+        # To make musl not define wchar_t as int, rather than short.
+        efi_c_args += '-D__DEFINED_wchar_t'
+endif
+
 efi_c_ld_args = [
         '-nostdlib',
         '-static-pie',
