diff --git a/meson.build b/meson.build
index f49dcc3bfb3a7..33c6b108e68cd 100644
--- a/meson.build
+++ b/meson.build
@@ -1640,7 +1640,6 @@ foreach term : ['analyze',
                 'hibernate',
                 'hostnamed',
                 'hwdb',
-                'idn',
                 'ima',
                 'ipe',
                 'initrd',
@@ -1679,6 +1678,13 @@ endforeach
 
 enable_sysusers = conf.get('ENABLE_SYSUSERS') == 1
 
+have = get_option('idn')
+if have and get_option('libc') == 'musl'
+        warning('IDN support is requested but it is not supported when building with musl, disabling it.')
+        have = false
+endif
+conf.set10('ENABLE_IDN', have)
+
 foreach tuple : [['nss-mymachines', 'machined'],
                  ['nss-resolve',    'resolve']]
         want = get_option(tuple[0])
