diff --git a/src/include/musl/string.h b/src/include/musl/string.h
new file mode 100644
index 0000000000000..cc3da63012d08
--- /dev/null
+++ b/src/include/musl/string.h
@@ -0,0 +1,7 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+#pragma once
+
+#include_next <string.h>
+
+char* strerror_r_gnu(int errnum, char *buf, size_t buflen);
+#define strerror_r strerror_r_gnu
diff --git a/src/libc/musl/meson.build b/src/libc/musl/meson.build
index 3e205e1eb1012..a64f292081b77 100644
--- a/src/libc/musl/meson.build
+++ b/src/libc/musl/meson.build
@@ -7,4 +7,5 @@ endif
 libc_wrapper_sources += files(
         'printf.c',
         'time.c',
+        'string.c',
 )
diff --git a/src/libc/musl/string.c b/src/libc/musl/string.c
new file mode 100644
index 0000000000000..c2a9f4a1690ea
--- /dev/null
+++ b/src/libc/musl/string.c
@@ -0,0 +1,26 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <errno.h>
+#include <stdio.h>
+#include <string.h>
+
+/* The header errno.h overrides strerror_r with strerror_r_gnu, hence we need to undef it here. */
+#undef strerror_r
+
+char* strerror_r_gnu(int errnum, char *buf, size_t buflen) {
+        int saved_errno = errno;
+        const char *msg = strerror(errnum);
+        size_t l = strlen(msg);
+        if (l >= buflen) {
+                if (buflen > 0) {
+                        if (buflen > 1)
+                                memcpy(buf, msg, buflen - 1);
+                        buf[buflen - 1] = '\0';
+                }
+                errno = ERANGE;
+        } else {
+                memcpy(buf, msg, l + 1);
+                errno = saved_errno;
+        }
+        return buf;
+}
