diff --git a/src/include/musl/utmpx.h b/src/include/musl/utmpx.h
new file mode 100644
index 0000000000000..0c687a83f39f5
--- /dev/null
+++ b/src/include/musl/utmpx.h
@@ -0,0 +1,14 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+#pragma once
+
+#include_next <utmpx.h>
+
+#ifndef ACCOUNTING
+#define ACCOUNTING 9
+#endif
+#ifndef UTMPX_FILE
+#define UTMPX_FILE _PATH_UTMP
+#endif
+#ifndef WTMPX_FILE
+#define WTMPX_FILE _PATH_WTMP
+#endif
