diff --git a/meson.build b/meson.build
index 441334860ac54..3ddc114243079 100644
--- a/meson.build
+++ b/meson.build
@@ -571,6 +571,13 @@ conf.set('SIZEOF_RLIM_T', cc.sizeof('rlim_t', prefix : '#include <sys/resource.h
 conf.set('SIZEOF_TIME_T', cc.sizeof('time_t', prefix : '#include <sys/time.h>'))
 conf.set('SIZEOF_TIMEX_MEMBER', cc.sizeof('typeof(((struct timex *)0)->freq)', prefix : '#include <sys/timex.h>'))
 
+if get_option('libc') == 'musl'
+        if conf.get('SIZEOF_RLIM_T') != 8
+                error('Unexpected size of rlim_t: @0@'.format(conf.get('SIZEOF_RLIM_T')))
+        endif
+        conf.set_quoted('RLIM_FMT', '%llu')
+endif
+
 long_max = cc.compute_int(
         'LONG_MAX',
         prefix : '#include <limits.h>',
diff --git a/src/basic/format-util.h b/src/basic/format-util.h
index e42f788ce6f26..d40ca5818a10a 100644
--- a/src/basic/format-util.h
+++ b/src/basic/format-util.h
@@ -39,12 +39,14 @@ assert_cc(sizeof(gid_t) == sizeof(uint32_t));
 #  error Unknown timex member size
 #endif
 
-#if SIZEOF_RLIM_T == 8
-#  define RLIM_FMT "%" PRIu64
-#elif SIZEOF_RLIM_T == 4
-#  define RLIM_FMT "%" PRIu32
-#else
-#  error Unknown rlim_t size
+#ifndef RLIM_FMT
+#  if SIZEOF_RLIM_T == 8
+#    define RLIM_FMT "%" PRIu64
+#  elif SIZEOF_RLIM_T == 4
+#    define RLIM_FMT "%" PRIu32
+#  else
+#    error Unknown rlim_t size
+#  endif
 #endif
 
 #if SIZEOF_DEV_T == 8
