--- a/bin/estrip
+++ b/bin/estrip
@@ -434,9 +434,7 @@
 	# elf types though cause that stuff is good.
 
 	buildid=
-	if [[ ${f} == *"current ar archive"* ]] ; then
-		process_ar "${x}"
-	elif [[ ${f} == *"SB executable"* || ${f} == *"SB pie executable"* ||
+	if [[ ${f} == *"SB executable"* || ${f} == *"SB pie executable"* ||
 		${f} == *"SB shared object"* ]] ; then
 		process_elf "${x}" "${inode_link}" ${PORTAGE_STRIP_FLAGS}
 	elif [[ ${f} == *"SB relocatable"* ]] ; then
