--- a/bin/estrip
+++ b/bin/estrip
@@ -133,7 +133,7 @@
 *GNU*) # sys-devel/binutils
 	# We'll leave out -R .note for now until we can check out the relevance
 	# of the section when it has the ALLOC flag set on it ...
-	SAFE_STRIP_FLAGS="--strip-unneeded -N __gentoo_check_ldflags__"
+	SAFE_STRIP_FLAGS="-S"
 	DEF_STRIP_FLAGS="-R .comment -R .GCC.command.line -R .note.gnu.gold-version"
 	SPLIT_STRIP_FLAGS=
 	;;
@@ -434,9 +434,7 @@
 	# elf types though cause that stuff is good.
 
 	buildid=
-	if [[ ${f} == *"current ar archive"* ]] ; then
-		process_ar "${x}"
-	elif [[ ${f} == *"SB executable"* || ${f} == *"SB pie executable"* ||
+	if [[ ${f} == *"SB executable"* || ${f} == *"SB pie executable"* ||
 		${f} == *"SB shared object"* ]] ; then
 		process_elf "${x}" "${inode_link}" ${PORTAGE_STRIP_FLAGS}
 	elif [[ ${f} == *"SB relocatable"* ]] ; then
