diff --git a/tools/lld/CMakeLists.txt b/tools/lld/CMakeLists.txt
index a15e296e31d..4b3e3f0835b 100644
--- a/tools/lld/CMakeLists.txt
+++ b/tools/lld/CMakeLists.txt
@@ -20,7 +20,7 @@ install(TARGETS lld
   RUNTIME DESTINATION bin)
 
 if(NOT LLD_SYMLINKS_TO_CREATE)
-  set(LLD_SYMLINKS_TO_CREATE lld-link ld.lld ld64.lld wasm-ld)
+  set(LLD_SYMLINKS_TO_CREATE lld-link ld.lld ld64.lld wasm-ld ld)
 endif()
 
 foreach(link ${LLD_SYMLINKS_TO_CREATE})
