--- a/lld/tools/lld/CMakeLists.txt.bak	2021-04-01 23:35:22.467460176 +0200
+++ b/lld/tools/lld/CMakeLists.txt	2021-04-01 23:35:32.679579629 +0200
@@ -25,7 +25,7 @@
 
 if(NOT LLD_SYMLINKS_TO_CREATE)
   set(LLD_SYMLINKS_TO_CREATE
-      lld-link ld.lld ld64.lld ld64.lld.darwinnew ld64.lld.darwinold wasm-ld)
+      lld-link ld.lld ld64.lld ld64.lld.darwinnew ld64.lld.darwinold wasm-ld ld)
 endif()
 
 foreach(link ${LLD_SYMLINKS_TO_CREATE})
