--- a/src/auditctl.c.bak	2015-08-17 19:17:23.403170528 +0000
+++ b/src/auditctl.c	2015-08-17 19:19:15.034830655 +0000
@@ -33,6 +33,7 @@
 #include <ctype.h>
 #include <unistd.h>
 #include <sys/utsname.h>
+#include <sys/select.h>
 #include <fcntl.h>
 #include <errno.h>
 #include <libgen.h>	/* For basename */
--- a/audisp/audispd.c.bak	2015-08-17 19:21:36.242401255 +0000
+++ b/audisp/audispd.c	2015-08-17 19:22:29.595231947 +0000
@@ -29,6 +29,7 @@
 #include <errno.h>
 #include <string.h>
 #include <sys/wait.h>
+#include <limits.h>
 #include <pthread.h>
 #include <dirent.h>
 #include <fcntl.h>
