diff --git a/haddock-api/src/Documentation/Haddock.hs b/haddock-api/src/Documentation/Haddock.hs
index 10d6849a..22a1c922 100644
--- a/haddock-api/src/Documentation/Haddock.hs
+++ b/haddock-api/src/Documentation/Haddock.hs
@@ -67,7 +67,11 @@ module Documentation.Haddock (
   haddock,
   haddockWithGhc,
   getGhcDirs,
-  withGhc
+  withGhc,
+
+  -- * Types
+  Wrap(..),
+  showWrapped
 ) where
 
 import Documentation.Haddock.Markup (markup)
