diff --git a/src/llvm-api.cpp b/src/llvm-api.cpp
index 847180029..d589ba6db 100644
--- a/src/llvm-api.cpp
+++ b/src/llvm-api.cpp
@@ -241,12 +241,6 @@ extern "C" JL_DLLEXPORT LLVMContextRef LLVMExtraGetValueContext(LLVMValueRef V)
     return wrap(&unwrap(V)->getContext());
 }
 
-extern ModulePass *createNVVMReflectPass();
-extern "C" JL_DLLEXPORT void LLVMExtraAddMVVMReflectPass(LLVMPassManagerRef PM)
-{
-    createNVVMReflectPass();
-}
-
 extern "C" JL_DLLEXPORT void
 LLVMExtraAddTargetLibraryInfoByTiple(const char *T, LLVMPassManagerRef PM)
 {
