--- a/src/debuginfo.cpp
+++ b/src/debuginfo.cpp
@@ -728,15 +728,6 @@ static void get_function_name_and_base(const object::ObjectFile *object, bool in
     bool needs_name = name && (!*name || untrusted_dladdr);
     // Try platform specific methods first since they are usually faster
     if (needs_saddr) {
-#if (defined(_OS_LINUX_) || defined(_OS_FREEBSD_)) && !defined(JL_DISABLE_LIBUNWIND)
-        unw_proc_info_t pip;
-        // Seems that libunwind may return NULL IP depending on what info it finds...
-        if (unw_get_proc_info_by_ip(unw_local_addr_space, pointer,
-                                    &pip, NULL) == 0 && pip.start_ip) {
-            *saddr = (void*)pip.start_ip;
-            needs_saddr = false;
-        }
-#endif
 #if defined(_OS_WINDOWS_) && defined(_CPU_X86_64_)
         DWORD64 ImageBase;
         PRUNTIME_FUNCTION fn = RtlLookupFunctionEntry(pointer, &ImageBase, NULL);
diff --git a/Make.inc b/Make.inc
index 79a38b0ca9..460f692a1a 100644
--- a/Make.inc
+++ b/Make.inc
@@ -863,7 +863,7 @@ LIBUNWIND:=
 else
 ifeq ($(USE_SYSTEM_LIBUNWIND), 1)
 ifneq ($(OS),Darwin)
-LIBUNWIND:=-lunwind-generic -lunwind
+LIBUNWIND:=-lunwind
 # Only for linux since we want to use not yet released libunwind features
 JCFLAGS+=-DSYSTEM_LIBUNWIND
 JCPPFLAGS+=-DSYSTEM_LIBUNWIND
