--- a/src/jitlayers.cpp 2019-07-14 21:13:54.631876662 +0000
+++ b/src/jitlayers.cpp 2019-07-14 21:13:39.822875586 +0000
@@ -504,8 +504,14 @@
     // Force LLVM to emit the module so that we can register the symbols
     // in our lookup table.
     auto Err = CompileLayer.emitAndFinalize(key);
-    // Check for errors to prevent LLVM from crashing the program.
-    assert(!Err);
+    if(Err) {
+        std::string Buf;
+        raw_string_ostream OS(Buf);
+        logAllUnhandledErrors(std::move(Err), OS, "");
+        OS.flush();
+        llvm::report_fatal_error("FATAL: Unable to compile LLVM Module: '" + Buf + "'\n"
+                             "The module's content was printed above. Please file a bug report");
+    }
 }
 
 void JuliaOJIT::removeModule(ModuleHandleT H)
