--- a/Make.inc	2018-06-27 17:49:47.004109137 +0200
+++ b/Make.inc	2018-06-27 18:49:22.475117160 +0200
@@ -1207,12 +1207,6 @@
 
 JULIA_BUILD_MODE := release
 JULIA_LIBSUFFIX:=
-ifeq (,$(findstring release,$(MAKECMDGOALS)))
-ifneq (,$(findstring debug,$(MAKECMDGOALS)))
-JULIA_BUILD_MODE := debug
-JULIA_LIBSUFFIX:=-debug
-endif
-endif
 
 JULIA_EXECUTABLE_debug := $(build_bindir)/julia-debug$(EXE)
 JULIA_EXECUTABLE_release := $(build_bindir)/julia$(EXE)
