--- a/src/Makefile	2023-07-28 16:49:24.548321010 +0200
+++ b/src/Makefile	2023-07-28 16:48:37.591493052 +0200
@@ -119,7 +119,7 @@
 ifneq ($(USE_SYSTEM_LLVM),0)
 # USE_SYSTEM_LLVM != 0
 CG_LLVMLINK += $(LLVM_LDFLAGS) $(shell $(LLVM_CONFIG_HOST) --libs --system-libs)
-LLVM_SHLIB_SYMBOL_VERSION := $(shell nm -D --with-symbol-versions $(shell $(LLVM_CONFIG_HOST) --libfiles --link-shared | awk '{print $1; exit}') | \
+LLVM_SHLIB_SYMBOL_VERSION := $(shell nm -D $(shell $(LLVM_CONFIG_HOST) --libfiles --link-shared | awk '{print $1; exit}') | \
                                grep _ZN4llvm3Any6TypeId | head -n 1 | sed -e 's/.*@//')
 
 # HACK: llvm-config doesn't correctly point to shared libs on all platforms
