--- a/Makefile	2025-01-06 16:29:26.353695702 +0100
+++ b/Makefile	2025-01-06 16:32:25.823810788 +0100
@@ -289,7 +289,6 @@
 
 
 install: $(build_depsbindir)/stringreplace
-	@$(MAKE) $(QUIET_MAKE) $(JULIA_BUILD_MODE)
 	@for subdir in $(bindir) $(datarootdir)/julia/stdlib/$(VERSDIR) $(docdir) $(man1dir) $(includedir)/julia $(libdir) $(private_libdir) $(sysconfdir) $(private_libexecdir); do \
 		mkdir -p $(DESTDIR)$$subdir; \
 	done
