--- a/stdlib/SuiteSparse_jll/src/SuiteSparse_jll.jl	2023-07-30 21:06:34.666408054 +0200
+++ b/stdlib/SuiteSparse_jll/src/SuiteSparse_jll.jl	2023-07-30 21:13:49.868059562 +0200
@@ -75,8 +75,8 @@
     const libcolamd = "libcolamd.so.3"
     const libklu = "libklu.so.2"
     const libldl = "libldl.so.3"
-    const librbio = "librbio.so.4"
-    const libspqr = "libspqr.so.4"
+    const librbio = "librbio.so.3"
+    const libspqr = "libspqr.so.3"
     const libsuitesparseconfig = "libsuitesparseconfig.so.7"
     const libumfpack = "libumfpack.so.6"
 end
