--- a/doc/Makefile	2017-07-23 15:34:14.372994303 +0200
+++ b/doc/Makefile	2017-07-24 01:00:23.780456834 +0200
@@ -30,15 +30,7 @@
 cleanall: clean
 
 html: deps
-	@echo "Building HTML documentation."
-ifneq ($(OS),WINNT)
-	$(JULIA_EXECUTABLE) $(call cygpath_w,$(SRCDIR)/make.jl)
-else
-# work around issue #11727, windows output redirection breaking on buildbot
-	$(JULIA_EXECUTABLE) $(call cygpath_w,$(SRCDIR)/make.jl) > docbuild.log 2>&1
-	@cat docbuild.log
-endif
-	@echo "Build finished. The HTML pages are in _build/html."
+	@echo "NOOP"
 
 pdf: deps
 	@echo "Building PDF documentation."
