--- a/stdlib/LibGit2_jll/src/LibGit2_jll.jl
+++ b/stdlib/LibGit2_jll/src/LibGit2_jll.jl
@@ -21,9 +21,9 @@ libgit2_path = ""
 if Sys.iswindows()
     const libgit2 = "libgit2.dll"
 elseif Sys.isapple()
-    const libgit2 = "@rpath/libgit2.1.6.dylib"
+    const libgit2 = "@rpath/libgit2.1.6.dylib"
 else
-    const libgit2 = "libgit2.so.1.6"
+    const libgit2 = "libgit2.so"
 end
 
 function __init__() 
