--- a/src/support/Makefile	2017-05-18 21:32:19.472432407 +0200
+++ b/src/support/Makefile	2017-05-18 21:32:51.111393795 +0200
@@ -64,11 +64,11 @@
 
 $(BUILDDIR)/libsupport.a: $(OBJS)
 	rm -rf $@
-	@$(call PRINT_LINK, $(AR) -rcs $@ $^)
+	@$(call PRINT_LINK, $(AR) rcs $@ $^)
 
 $(BUILDDIR)/libsupport-debug.a: $(DOBJS)
 	rm -rf $@
-	@$(call PRINT_LINK, $(AR) -rcs $@ $^)
+	@$(call PRINT_LINK, $(AR) rcs $@ $^)
 
 clean:
 	rm -f $(BUILDDIR)/*.o
--- a/src/flisp/Makefile	2017-05-18 21:32:27.776421194 +0200
+++ b/src/flisp/Makefile	2017-05-18 21:32:38.711406659 +0200
@@ -64,11 +64,11 @@
 
 $(BUILDDIR)/$(LIBTARGET)-debug.a: $(DOBJS) | $(BUILDDIR)
 	rm -rf $@
-	@$(call PRINT_LINK, $(AR) -rcs $@ $(DOBJS))
+	@$(call PRINT_LINK, $(AR) rcs $@ $(DOBJS))
 
 $(BUILDDIR)/$(LIBTARGET).a: $(OBJS) | $(BUILDDIR)
 	rm -rf $@
-	@$(call PRINT_LINK, $(AR) -rcs $@ $(OBJS))
+	@$(call PRINT_LINK, $(AR) rcs $@ $(OBJS))
 
 ifneq ($(USEMSVC), 1)
 CCLD := $(CC)
