--- a/stdlib/LibGit2_jll/src/LibGit2_jll.jl
+++ b/stdlib/LibGit2_jll/src/LibGit2_jll.jl
@@ -21,9 +21,9 @@ libgit2_path = ""
 if Sys.iswindows()
     const libgit2 = "libgit2.dll"
 elseif Sys.isapple()
-    const libgit2 = "@rpath/libgit2.1.6.dylib"
+    const libgit2 = "@rpath/libgit2.1.6.dylib"
 else
-    const libgit2 = "libgit2.so.1.6"
+    const libgit2 = "libgit2.so"
 end
 
 function __init__() 
--- a/stdlib/OpenBLAS_jll/src/OpenBLAS_jll.jl	2023-11-08 13:18:25.935654245 +0100
+++ b/stdlib/OpenBLAS_jll/src/OpenBLAS_jll.jl	2023-11-08 13:16:12.837367050 +0100
@@ -28,7 +28,7 @@
 libopenblas_path::String = ""
 
 if Base.USE_BLAS64
-    const libsuffix = "64_"
+    const libsuffix = "64"
 else
     const libsuffix = ""
 end
