--- a/src/client/web-process/util-composer.vala	2016-10-25 13:55:16.385774576 +0200
+++ b/src/client/web-process/util-composer.vala	2016-10-25 13:55:31.534774154 +0200
@@ -398,8 +398,8 @@
             return;
         }
 
-        left = node.clone_node(false);
-        right = node.clone_node(false);
+        left = node.clone_node_with_error(false);
+        right = node.clone_node_with_error(false);
 
         // Move the first $offset children to the left node.
         for (long i = 0; i < offset; i++)
--- a/src/client/web-process/util-webkit.vala	2016-10-25 13:54:39.660775600 +0200
+++ b/src/client/web-process/util-webkit.vala	2016-10-25 13:55:01.809774983 +0200
@@ -28,7 +28,7 @@
     public WebKit.DOM.HTMLElement? clone_node(WebKit.DOM.Node node, bool deep = true) {
         WebKit.DOM.HTMLElement? clone = null;
         try {
-            clone = node.clone_node(deep) as WebKit.DOM.HTMLElement;
+            clone = node.clone_node_with_error(deep) as WebKit.DOM.HTMLElement;
         } catch (Error err) {
             debug("Error selecting cloning node: %s", err.message);
         }
@@ -178,7 +178,7 @@
                             new_nodes.add(anchor);
                         } else {
                             // Text part.
-                            WebKit.DOM.Node new_node = node.clone_node(false);
+                            WebKit.DOM.Node new_node = node.clone_node_with_error(false);
                             new_node.set_node_value(pieces[i]);
                             new_nodes.add(new_node);
                         }
