diff --git a/CMakeLists.txt b/CMakeLists.txt
index aacb5259..18b4f71b 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -149,14 +149,6 @@ endif ()
 include(GlibCompileResourcesSupport)
 
 #
-# Unit tests.
-#
-# We don't use CMake's enable_testing/add_test built-ins because they
-# use ctest. It's not called "test" because that is cmake reserved.
-add_custom_target(tests ALL geary-test)
-add_dependencies(tests geary-test)
-
-#
 # Uninstall target
 #
 configure_file(
@@ -232,4 +224,3 @@ add_subdirectory(po)
 add_subdirectory(sql)
 add_subdirectory(ui)
 add_subdirectory(src)
-add_subdirectory(test)
