diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 51396b4..49a7416 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -421,12 +421,7 @@ ensure_vala_version("0.26.0" MINIMUM)
 include(ValaPrecompile)
 
 # Vapigen
-find_program(VAPIGEN vapigen)
-if (VAPIGEN STREQUAL "VAPIGEN-NOTFOUND")
-    message(FATAL_ERROR "vapigen must be installed to build Geary.")
-else ()
-    message(STATUS "Found vapigen: " ${VAPIGEN})
-endif ()
+set(VAPIGEN $ENV{VAPIGEN})
 
 # Packages
 if (LIBMESSAGINGMENU_FOUND)
