--- a/Makefile.am	2016-09-28 14:08:05.670338956 +0200
+++ b/Makefile.am	2016-09-28 14:08:31.420336838 +0200
@@ -79,10 +79,6 @@
 no_undefined = -no-undefined
 endif
 
-# REQUIRED_CFLAGS absolutely get appended to CFLAGS, whatever the
-# the value of CFLAGS in the environment.
-REQUIRED_CXXFLAGS=-Wl,-lstdc++
-
 # Because -fPIC can interfere with compilation on 32-bit platforms but
 # is absolutely necessary for AMD64, we check what the target machine's
 # word size is, and set our required flags based upon that.
--- a/Makefile.in	2016-09-28 14:10:14.321328371 +0200
+++ b/Makefile.in	2016-09-28 14:10:37.246326484 +0200
@@ -474,10 +474,6 @@
 # compilation setup. #
 @PLATFORM_WIN32_TRUE@no_undefined = -no-undefined
 
-# REQUIRED_CFLAGS absolutely get appended to CFLAGS, whatever the
-# the value of CFLAGS in the environment.
-REQUIRED_CXXFLAGS = -Wl,-lstdc++ $(am__append_1)
-AM_CXXFLAGS = $(REQUIRED_CXXFLAGS)
 lib@PACKAGE_NAME@_la_SOURCES = $(GEXIV2_sources) $(GEXIV2_public_HEADERS)
 lib@PACKAGE_NAME@_la_LIBADD = $(EXIV2_LIBS) $(GLIB_LIBS)
 lib@PACKAGE_NAME@_la_CPPFLAGS = $(EXIV2_CFLAGS) $(GLIB_CFLAGS)
