--- a/lib/Driver/Distro.cpp	2017-07-19 17:05:48.464242836 +0200
+++ b/lib/Driver/Distro.cpp	2017-07-19 17:05:14.896297856 +0200
@@ -129,6 +129,9 @@
   if (VFS.exists("/etc/arch-release"))
     return Distro::ArchLinux;
 
+  if (VFS.exists("/etc/gentoo-release"))
+    return Distro::GentooLinux;
+
   return Distro::UnknownDistro;
 }
 
--- a/include/clang/Driver/Distro.h	2017-07-19 17:01:58.903617193 +0200
+++ b/include/clang/Driver/Distro.h	2017-07-19 17:03:38.760455154 +0200
@@ -33,6 +33,7 @@
     DebianJessie,
     DebianStretch,
     Exherbo,
+    GentooLinux,
     RHEL5,
     RHEL6,
     RHEL7,
@@ -114,6 +115,10 @@
     return DistroVal >= UbuntuHardy && DistroVal <= UbuntuArtful;
   }
 
+  bool IsGentoo() const {
+    return DistroVal == GentooLinux;
+  }
+
   /// @}
 };
 
