--- a/lib/Headers/unwind.h	2017-06-24 11:09:47.899882590 +0200
+++ b/lib/Headers/unwind.h	2017-06-24 11:10:01.234881493 +0200
@@ -26,7 +26,7 @@
 #ifndef __CLANG_UNWIND_H
 #define __CLANG_UNWIND_H
 
-#if defined(__APPLE__) && __has_include_next(<unwind.h>)
+#if __has_include_next(<unwind.h>)
 /* Darwin (from 11.x on) provide an unwind.h. If that's available,
  * use it. libunwind wraps some of its definitions in #ifdef _GNU_SOURCE,
  * so define that around the include.*/
