--- a/lib/Driver/Tools.cpp	2016-08-19 13:59:21.260816904 +0200
+++ b/lib/Driver/Tools.cpp	2016-08-19 14:08:56.557903229 +0200
@@ -3001,7 +3001,10 @@
 // and available, check for possible errors, etc.
 static void addClangRT(const ToolChain &TC, const ArgList &Args,
                        ArgStringList &CmdArgs) {
+  CmdArgs.push_back("--as-needed");
   CmdArgs.push_back(TC.getCompilerRTArgString(Args, "builtins"));
+  CmdArgs.push_back("-lunwind");
+  CmdArgs.push_back("--no-as-needed");
 }
 
 namespace {
