--- a/lib/Basic/Targets/OSTargets.h
+++ b/lib/Basic/Targets/OSTargets.h
@@ -354,9 +354,12 @@
       this->PlatformMinVersion = VersionTuple(Maj, Min, Rev);
       if (Maj)
         Builder.defineMacro("__ANDROID_API__", Twine(Maj));
-    } else {
+    } else if (!Triple.isMusl()) {
         Builder.defineMacro("__gnu_linux__");
     }
+    else {
+        Builder.defineMacro("__STDC_ISO_10646__", "201206L");
+    }
     if (Opts.POSIXThreads)
       Builder.defineMacro("_REENTRANT");
     if (Opts.CPlusPlus)
