--- a/lib/Frontend/CompilerInvocation.cpp 2017-07-19 17:43:16.481656709 +0200
+++ b/lib/Frontend/CompilerInvocation.cpp 2017-07-19 17:44:40.533524918 +0200
@@ -1667,7 +1667,7 @@
       if (T.isPS4())
         LangStd = LangStandard::lang_gnucxx11;
       else
-        LangStd = LangStandard::lang_gnucxx98;
+        LangStd = LangStandard::lang_gnucxx14;
       break;
     case InputKind::RenderScript:
       LangStd = LangStandard::lang_c99;
