--- a/lib/Driver/ToolChains.cpp	2016-10-02 12:03:54.025212727 +0200
+++ b/lib/Driver/ToolChains.cpp	2016-10-02 12:06:06.350201839 +0200
@@ -4233,7 +4233,7 @@
     if (Triple.getEnvironment() == llvm::Triple::MuslEABIHF)
       ArchName += "hf";
 
-    return "/lib/ld-musl-" + ArchName + ".so.1";
+    return "/" + getOSLibDir(Triple, Args).str() + "/ld-musl-" + ArchName + ".so.1";
   }
 
   std::string LibDir;
