diff --git a/tools/driver/CMakeLists.txt b/tools/driver/CMakeLists.txt
index 22a498422ae..fd9d5b78b47 100644
--- a/tools/driver/CMakeLists.txt
+++ b/tools/driver/CMakeLists.txt
@@ -62,7 +62,7 @@
 add_dependencies(clang clang-headers)
 
 if(NOT CLANG_LINKS_TO_CREATE)
-  set(CLANG_LINKS_TO_CREATE clang++ clang-cl clang-cpp flang)
+  set(CLANG_LINKS_TO_CREATE clang++ clang-cl clang-cpp flang gcc g++ cc c++ cpp gfortran)
 endif()
 
 foreach(link ${CLANG_LINKS_TO_CREATE})
