--- a/lib/Frontend/InitPreprocessor.cpp	2016-10-16 12:08:18.516116742 +0200
+++ b/lib/Frontend/InitPreprocessor.cpp	2016-10-16 12:10:20.727106686 +0200
@@ -446,6 +446,9 @@
     Builder.defineMacro("__ASSEMBLER__");
   if (LangOpts.CUDA)
     Builder.defineMacro("__CUDA__");
+
+  // define standard iso, used by libedit for example
+  Builder.defineMacro("__STDC_ISO_10646__", "201206L");
 }
 
 /// Initialize the predefined C++ language feature test macros defined in
