--- a/lib/Driver/ToolChains.h	2016-01-07 23:09:47.000000000 +0100
+++ b/lib/Driver/ToolChains.h	2016-02-19 20:03:17.237467235 +0100
@@ -838,6 +838,9 @@
   void AddIAMCUIncludeArgs(const llvm::opt::ArgList &DriverArgs,
                            llvm::opt::ArgStringList &CC1Args) const override;
   bool isPIEDefault() const override;
+
+  unsigned GetDefaultStackProtectorLevel(bool KernelOrKext) const override;
+
   SanitizerMask getSupportedSanitizers() const override;
   void addProfileRTLibs(const llvm::opt::ArgList &Args,
                         llvm::opt::ArgStringList &CmdArgs) const override;
--- a/lib/Driver/ToolChains.cpp	2016-08-18 19:56:48.000000000 +0200
+++ b/lib/Driver/ToolChains.cpp	2016-09-30 16:00:53.952224677 +0200
@@ -4329,6 +4371,13 @@
   return "/" + LibDir + "/" + Loader;
 }
 
+unsigned Linux::GetDefaultStackProtectorLevel(bool KernelOrKext) const {
+  StringRef VendorName = Linux::getTriple().getVendorName();
+  if (VendorName.compare("gentoo") == 0)
+    return 2;
+  return 1;
+}
+
 void Linux::AddClangSystemIncludeArgs(const ArgList &DriverArgs,
                                       ArgStringList &CC1Args) const {
   const Driver &D = getDriver();
