--- a/lib/Driver/ToolChains.cpp	2016-02-16 20:56:48.000000000 +0100
+++ b/lib/Driver/ToolChains.cpp	2016-07-01 11:14:02.041762774 +0200
@@ -2239,12 +2239,12 @@
   // in the second half of each row is 1 + the number of path separators
   // in the first half.
   const std::string LibAndInstallSuffixes[][2] = {
-      {"/gcc/" + CandidateTriple.str(), "/../../.."},
+      {"/clang/" + CandidateTriple.str(), "/../../.."},
 
       // Debian puts cross-compilers in gcc-cross
       {"/gcc-cross/" + CandidateTriple.str(), "/../../.."},
 
-      {"/" + CandidateTriple.str() + "/gcc/" + CandidateTriple.str(),
+      {"/" + CandidateTriple.str() + "/clang/" + CandidateTriple.str(),
        "/../../../.."},
 
       // The Freescale PPC SDK has the gcc libraries in
