--- a/lib/Driver/Tools.cpp	2016-08-19 13:59:21.260816904 +0200
+++ b/lib/Driver/Tools.cpp	2016-08-19 14:08:56.557903229 +0200
@@ -3000,8 +3000,8 @@
 // FIXME: Make sure we can also emit shared objects if they're requested
 // and available, check for possible errors, etc.
 static void addClangRT(const ToolChain &TC, const ArgList &Args,
-                       ArgStringList &CmdArgs) {
-  CmdArgs.push_back(TC.getCompilerRTArgString(Args, "builtins"));
+                       ArgStringList &CmdArgs, bool Shared) {
+  CmdArgs.push_back(TC.getCompilerRTArgString(Args, "builtins", Shared));
 }
 
 namespace {
@@ -9270,7 +9270,10 @@
       llvm_unreachable("unsupported OS");
     case llvm::Triple::Win32:
     case llvm::Triple::Linux:
-      addClangRT(TC, Args, CmdArgs);
+      if (Args.hasArg(options::OPT_static))
+        addClangRT(TC, Args, CmdArgs, false);
+      else
+        addClangRT(TC, Args, CmdArgs, true);
       break;
     }
     break;
