--- a/lib/Driver/Tools.cpp	2016-10-16 21:22:07.780382534 +0200
+++ b/lib/Driver/Tools.cpp	2016-10-16 21:23:22.306376402 +0200
@@ -9398,7 +9398,7 @@
   if (Args.hasArg(options::OPT_rdynamic))
     CmdArgs.push_back("-export-dynamic");
 
-  if (Args.hasArg(options::OPT_s))
+  if (!Args.hasArg(options::OPT_nostrip))
     CmdArgs.push_back("-s");
 
   if (Arch == llvm::Triple::armeb || Arch == llvm::Triple::thumbeb)
--- a/include/clang/Driver/Options.td	2016-10-18 14:40:43.252321932 +0200
+++ b/include/clang/Driver/Options.td	2016-10-18 14:43:36.200307701 +0200
@@ -1880,6 +1880,7 @@
            "system header.">;
 def : Separate<["--"], "no-system-header-prefix">, Alias<no_system_header_prefix>;
 def s : Flag<["-"], "s">;
+def nostrip : Flag<["-"], "nostrip">;
 def target : Joined<["--"], "target=">, Flags<[DriverOption, CoreOption]>,
   HelpText<"Generate code for the given target">;
 def gcc_toolchain : Joined<["--"], "gcc-toolchain=">, Flags<[DriverOption]>,
