--- a/lib/Driver/Tools.cpp	2016-10-16 21:22:07.780382534 +0200
+++ b/lib/Driver/Tools.cpp	2016-10-16 21:23:22.306376402 +0200
@@ -9398,7 +9398,7 @@
   if (Args.hasArg(options::OPT_rdynamic))
     CmdArgs.push_back("-export-dynamic");
 
-  if (Args.hasArg(options::OPT_s))
+  if (!Args.hasArg(options::OPT_gTune_Group) && !Args.hasArg(options::OPT_ggdbN_Group) && !Args.hasArg(options::OPT_g_Group) && !Args.hasArg(options::OPT_gN_Group))
     CmdArgs.push_back("-s");
 
   if (Arch == llvm::Triple::armeb || Arch == llvm::Triple::thumbeb)
