--- a/lib/Driver/ToolChain.cpp	2016-10-05 19:11:54.189603771 +0200
+++ b/lib/Driver/ToolChain.cpp	2016-10-05 19:12:26.688605371 +0200
@@ -354,6 +354,9 @@
       if (UseLinker.empty() || UseLinker == "ld")
         return GetProgramPath("ld");
 
+      if (UseLinker == "gold")
+        return GetProgramPath("ld.lld");
+
       llvm::SmallString<8> LinkerName("ld.");
       LinkerName.append(UseLinker);
 
@@ -366,7 +369,7 @@
     return "";
   }
 
-  return GetProgramPath(DefaultLinker);
+  return GetProgramPath("ld.lld");
 }
 
 types::ID ToolChain::LookupTypeForExtension(const char *Ext) const {
