diff --git a/Package@swift-4.swift b/Package@swift-4.swift
index 81b5f1b..0cc35fd 100644
--- a/Package@swift-4.swift
+++ b/Package@swift-4.swift
@@ -6,12 +6,8 @@ let package = Package(
     products: [
     	.library(name: "JWT", targets: ["JWT"]),
     ],
-    dependencies: [
-        .package(url: "https://github.com/vapor/crypto.git", .upToNextMajor(from: "2.0.0")),
-        .package(url: "https://github.com/vapor/json.git", .upToNextMajor(from: "2.0.0")),
-    ],
     targets: [
-    	.target(name: "JWT", dependencies: ["Crypto", "JSON"]),
+        .target(name: "JWT"),
     	.testTarget(name: "JWTTests", dependencies: ["JWT"]),
     ]
 )
