--- wscript.bak	2016-09-02 12:02:09.749732302 +0200
+++ wscript	2016-09-02 12:03:02.048363036 +0200
@@ -125,9 +125,9 @@
 	ctx.load('compiler_c vala')
 	ctx.add_option('--with-unity', action='store_true', default=False, dest='unity', help="Build functionality dependent on libunity")
 	ctx.add_option('--noopt', action='store_true', default=False, dest='noopt', help="Turn off compiler optimizations")
-	ctx.add_option('--debug', action='store_true', default=True, dest='debug', help="Turn on debugging symbols")
-	ctx.add_option('--no-debug', action='store_false', dest='debug', help="Turn off debugging symbols")
-	ctx.add_option('--no-system-hooks', action='store_false', default=True, dest='system_hooks', help="Don't run system hooks after installation (ldconfig, icon cache update, ...")
+	ctx.add_option('--debug', action='store_true', default=False, dest='debug', help="Turn on debugging symbols")
+	ctx.add_option('--no-debug', action='store_false', default=True, dest='debug', help="Turn off debugging symbols")
+	ctx.add_option('--no-system-hooks', action='store_false', default=False, dest='system_hooks', help="Don't run system hooks after installation (ldconfig, icon cache update, ...")
 	ctx.add_option('--platform', default=_PLATFORM, help="Target platform")
 	ctx.add_option('--with-apps-alpha', action='store_true', default=False, dest='apps_alpha', help="Include Nuvola Apps Alpha launcher.")
 
