--- src/mapi/entry_x86-64_tls.h.bak	2015-03-28 19:20:39.000000000 +0100
+++ src/mapi/entry_x86-64_tls.h	2015-03-31 15:40:44.514358152 +0200
@@ -62,8 +62,8 @@
 {
 }
 
-static char
-x86_64_entry_start[];
+extern char
+x86_64_entry_start[] __attribute__((visibility("hidden")));
 
 mapi_func
 entry_get_public(int slot)
--- src/mapi/entry_x86_tls.h.bak	2015-05-03 15:47:23.193729158 +0200
+++ src/mapi/entry_x86_tls.h	2015-05-03 15:48:07.417532336 +0200
@@ -72,8 +72,10 @@
 extern unsigned long
 x86_current_tls();
 
-static char x86_entry_start[];
-static char x86_entry_end[];
+extern
+char x86_entry_start[] __attribute__((visibility("hidden")));
+extern
+char x86_entry_end[] __attribute__((visibility("hidden")));
 
 void
 entry_patch_public(void)
