--- a/src/gallium/targets/vdpau/Makefile.in	2016-10-06 19:47:01.181703421 +0200
+++ b/src/gallium/targets/vdpau/Makefile.in	2016-10-06 19:48:25.381707567 +0200
@@ -226,6 +226,7 @@
 	$(top_builddir)/src/gallium/auxiliary/libgalliumvlwinsys.la \
 	$(top_builddir)/src/gallium/auxiliary/libgalliumvl.la \
 	$(top_builddir)/src/gallium/auxiliary/libgallium.la \
+        $(top_builddir)/src/compiler/nir/libnir.la \
 	$(top_builddir)/src/util/libmesautil.la $(am__DEPENDENCIES_1) \
 	$(am__DEPENDENCIES_1) $(am__DEPENDENCIES_2) \
 	$(am__DEPENDENCIES_10) $(am__append_19) $(am__DEPENDENCIES_11)
@@ -697,6 +697,7 @@
 	$(top_builddir)/src/gallium/auxiliary/libgalliumvlwinsys.la \
 	$(top_builddir)/src/gallium/auxiliary/libgalliumvl.la \
 	$(top_builddir)/src/gallium/auxiliary/libgallium.la \
+        $(top_builddir)/src/compiler/nir/libnir.la \
 	$(top_builddir)/src/util/libmesautil.la $(VL_LIBS) \
 	$(LIBDRM_LIBS) $(GALLIUM_COMMON_LIB_DEPS) $(am__append_14) \
 	$(am__append_15) $(am__append_16)
--- a/src/gallium/targets/va/Makefile.in	2016-10-06 19:46:48.556702800 +0200
+++ b/src/gallium/targets/va/Makefile.in	2016-10-06 19:48:09.631706791 +0200
@@ -216,6 +216,7 @@
 	$(top_builddir)/src/gallium/auxiliary/libgalliumvlwinsys.la \
 	$(top_builddir)/src/gallium/auxiliary/libgalliumvl.la \
 	$(top_builddir)/src/gallium/auxiliary/libgallium.la \
+        $(top_builddir)/src/compiler/nir/libnir.la \
 	$(top_builddir)/src/util/libmesautil.la $(am__DEPENDENCIES_1) \
 	$(am__DEPENDENCIES_1) $(am__DEPENDENCIES_2) \
 	$(am__DEPENDENCIES_9) $(am__append_15) $(am__DEPENDENCIES_10)
@@ -708,6 +708,7 @@
 	$(top_builddir)/src/gallium/auxiliary/libgalliumvlwinsys.la \
 	$(top_builddir)/src/gallium/auxiliary/libgalliumvl.la \
 	$(top_builddir)/src/gallium/auxiliary/libgallium.la \
+        $(top_builddir)/src/compiler/nir/libnir.la \
 	$(top_builddir)/src/util/libmesautil.la $(VL_LIBS) \
 	$(LIBDRM_LIBS) $(GALLIUM_COMMON_LIB_DEPS) $(am__append_18) \
 	$(am__append_19) $(am__append_20)
