diff --git a/compiler/nimfix/nimfix.nim b/compiler/nimfix/nimfix.nim
index b4007cd..865ec94 100644
--- a/compiler/nimfix/nimfix.nim
+++ b/compiler/nimfix/nimfix.nim
@@ -12,8 +12,8 @@
 import strutils, os, parseopt
 import compiler/options, compiler/commands, compiler/modules, compiler/sem,
   compiler/passes, compiler/passaux, compiler/nimfix/pretty,
-  compiler/msgs, compiler/nimconf,
-  compiler/extccomp, compiler/condsyms, compiler/lists
+  compiler/msgs, compiler/nimconf, compiler/modulegraphs,
+  compiler/extccomp, compiler/condsyms, compiler/lists, compiler/idents
 
 const Usage = """
 Nimfix - Tool to patch Nim code
@@ -34,7 +34,7 @@ Options:
 In addition, all command line options of Nim are supported.
 """
 
-proc mainCommand =
+proc mainCommand*(graph: ModuleGraph; cache: IdentCache) =
   registerPass verbosePass
   registerPass semPass
   gCmd = cmdPretty
@@ -43,7 +43,7 @@ proc mainCommand =
     # current path is always looked first for modules
     prependStr(searchPaths, gProjectPath)
 
-  compileProject()
+  compileProject(graph, cache)
   pretty.overwriteFiles()
 
 proc processCmdLine*(pass: TCmdLinePass, cmd: string) =
@@ -80,6 +80,8 @@ proc processCmdLine*(pass: TCmdLinePass, cmd: string) =
       options.gProjectName = unixToNativePath(p.key)
       # if processArgument(pass, p, argsCount): break
 
+proc mainCommand*() = mainCommand(newModuleGraph(), newIdentCache())
+
 proc handleCmdLine() =
   if paramCount() == 0:
     stdout.writeLine(Usage)
