--- a/make/autoconf/toolchain.m4	2018-11-09 12:46:01.413685934 +0100
+++ b/make/autoconf/toolchain.m4	2018-11-09 12:45:26.950182944 +0100
@@ -37,7 +37,7 @@
 VALID_TOOLCHAINS_all="gcc clang xlc microsoft"
 
 # These toolchains are valid on different platforms
-VALID_TOOLCHAINS_linux="gcc clang"
+VALID_TOOLCHAINS_linux="clang gcc"
 VALID_TOOLCHAINS_macosx="gcc clang"
 VALID_TOOLCHAINS_aix="xlc"
 VALID_TOOLCHAINS_windows="microsoft"
