--- a/make/autoconf/jdk-options.m4	2018-10-06 00:05:45.000000000 +0200
+++ b/make/autoconf/jdk-options.m4	2018-11-09 16:42:58.992525891 +0100
@@ -231,7 +231,7 @@
   AC_SUBST(UNLIMITED_CRYPTO)
 
   # Should we build the serviceability agent (SA)?
-  INCLUDE_SA=true
+  AC_CHECK_HEADER(thread_db.h,INCLUDE_SA=true,INCLUDE_SA=false)
   if HOTSPOT_CHECK_JVM_VARIANT(zero); then
     INCLUDE_SA=false
   fi
