--- a/devices/musicbrainz.cpp	2016-04-02 12:02:50.525762442 +0200
+++ b/devices/musicbrainz.cpp	2016-04-02 12:04:49.874759116 +0200
@@ -248,7 +248,7 @@
     }
     DBUG << "Should lookup " << discId;
 
-    MusicBrainz5::CQuery Query("cantata-"PACKAGE_VERSION_STRING);
+    MusicBrainz5::CQuery Query("cantata-" PACKAGE_VERSION_STRING);
     QList<CdAlbum> m;
     QList<QNetworkProxy> proxies=NetworkProxyFactory::self()->queryProxy(QNetworkProxyQuery(QUrl("http://musicbrainz.org")));
     foreach (const QNetworkProxy &p, proxies) {
--- a/support/utils.cpp	2016-04-02 12:03:07.425761971 +0200
+++ b/support/utils.cpp	2016-04-02 12:05:03.649758732 +0200
@@ -773,7 +773,7 @@
     #elif defined Q_OS_MAC
     return fixPath(QCoreApplication::applicationDirPath())+app;
     #else
-    return QString(INSTALL_PREFIX "/"LINUX_LIB_DIR"/")+QCoreApplication::applicationName()+constDirSep+app;
+    return QString(INSTALL_PREFIX "/" LINUX_LIB_DIR "/")+QCoreApplication::applicationName()+constDirSep+app;
     #endif
 }
 
--- a/online/podcastwidget.cpp	2016-04-22 09:26:53.495119811 +0200
+++ b/online/podcastwidget.cpp	2016-04-22 09:27:10.244120636 +0200
@@ -295,13 +295,4 @@
 
 void PodcastWidget::controlActions()
 {
-    SinglePageWidget::controlActions();
-
-    QModelIndexList selected=view->selectedIndexes(false); // Dont need sorted selection here...
-    unSubscribeAction->setEnabled(1==selected.count() && static_cast<PodcastService::Item *>(proxy.mapToSource(selected.first()).internalPointer())->isPodcast());
-    downloadAction->setEnabled(!selected.isEmpty());
-    deleteAction->setEnabled(!selected.isEmpty());
-    cancelDownloadAction->setEnabled(!selected.isEmpty());
-    markAsNewAction->setEnabled(!selected.isEmpty());
-    markAsListenedAction->setEnabled(!selected.isEmpty());
 }
