--- a/CMakeLists.txt	2017-07-30 21:13:01.000000000 +0200
+++ b/CMakeLists.txt	2017-08-03 14:24:32.848926437 +0200
@@ -302,7 +302,7 @@
     endif ()
 endif ()
 
-include_directories(${CMAKE_SOURCE_DIR}/3rdparty ${CMAKE_SOURCE_DIR} ${CMAKE_BINARY_DIR} ${CMAKE_CURRENT_SOURCE_DIR} ${CMAKE_CURRENT_BINARY_DIR} ${QTINCLUDES} ${ZLIB_INCLUDE_DIRS})
+include_directories(${CMAKE_SOURCE_DIR}/3rdparty /usr/include/KF5/Solid ${CMAKE_SOURCE_DIR} ${CMAKE_BINARY_DIR} ${CMAKE_CURRENT_SOURCE_DIR} ${CMAKE_CURRENT_BINARY_DIR} ${QTINCLUDES} ${ZLIB_INCLUDE_DIRS})
 
 set(CANTATA_SRCS ${CANTATA_SRCS}
     gui/settings.cpp gui/application.cpp gui/initialsettingswizard.cpp gui/mainwindow.cpp gui/preferencesdialog.cpp gui/customactionssettings.cpp
@@ -439,7 +439,7 @@
         include_directories(${TAGLIB_INCLUDES}/..)
     endif ()
     if (ENABLE_DEVICES_SUPPORT)
-        set(CANTATA_LIBS ${CANTATA_LIBS} solidlite)
+        set(CANTATA_LIBS ${CANTATA_LIBS} KF5Solid)
     endif ()
     if (ENABLE_DEVICES_SUPPORT AND ENABLE_REMOTE_DEVICES AND NOT WIN32 AND NOT APPLE)
         set(CANTATA_LIBS ${CANTATA_LIBS} avahi)
@@ -456,7 +456,6 @@
     add_subdirectory(tags)
 
     if (ENABLE_DEVICES_SUPPORT)
-        add_subdirectory(3rdparty/solid-lite)
         if (MTP_FOUND)
             set(CANTATA_SRCS ${CANTATA_SRCS} devices/mtpdevice.cpp)
             set(CANTATA_MOC_HDRS ${CANTATA_MOC_HDRS} devices/mtpdevice.h)
--- a/INSTALL	2017-06-24 19:56:56.000000000 +0200
+++ b/INSTALL	2017-08-03 14:24:32.848926437 +0200
@@ -58,7 +58,7 @@
         Default: OFF
 
     -DENABLE_UDISKS2=ON
-        Build UDisks2 backend for solid-lite.
+        Build UDisks2 backend for solid.
         Default: ON
 
     -DINSTALL_UBUNTU_ICONS=<ON if Ubuntu, OFF otherwise>
--- a/devices/audiocddevice.cpp	2017-07-23 12:22:18.000000000 +0200
+++ b/devices/audiocddevice.cpp	2017-08-03 14:24:32.835926438 +0200
@@ -39,7 +39,7 @@
 #include <QDir>
 #include <QUrl>
 #include <QUrlQuery>
-#include "solid-lite/block.h"
+#include "solid/block.h"
 
 const QLatin1String AudioCdDevice::constAnyDev("-");
 
--- a/devices/audiocddevice.h	2017-04-07 20:04:18.000000000 +0200
+++ b/devices/audiocddevice.h	2017-08-03 14:24:32.834926438 +0200
@@ -27,7 +27,7 @@
 #include "device.h"
 #include "gui/covers.h"
 #include "http/httpserver.h"
-#include "solid-lite/opticaldrive.h"
+#include "solid/opticaldrive.h"
 #include <QImage>
 
 class CddbInterface;
--- a/devices/device.cpp	2017-04-07 20:54:06.000000000 +0200
+++ b/devices/device.cpp	2017-08-03 14:24:32.835926438 +0200
@@ -47,11 +47,11 @@
 #include "models/musiclibraryitemalbum.h"
 #include "models/musiclibraryitemsong.h"
 #include "models/musiclibrarymodel.h"
-#include "solid-lite/portablemediaplayer.h"
-#include "solid-lite/storageaccess.h"
-#include "solid-lite/storagedrive.h"
-#include "solid-lite/opticaldisc.h"
-#include "solid-lite/genericinterface.h"
+#include "solid/portablemediaplayer.h"
+#include "solid/storageaccess.h"
+#include "solid/storagedrive.h"
+#include "solid/opticaldisc.h"
+#include "solid/genericinterface.h"
 #endif // ENABLE_DEVICES_SUPPORT
 
 void Device::moveDir(const QString &from, const QString &to, const QString &base, const QString &coverFile)
--- a/devices/device.h	2017-04-07 20:04:17.000000000 +0200
+++ b/devices/device.h	2017-08-03 14:24:32.833926438 +0200
@@ -30,7 +30,7 @@
 #include "config.h"
 #ifdef ENABLE_DEVICES_SUPPORT
 #include "deviceoptions.h"
-#include "solid-lite/device.h"
+#include "solid/device.h"
 #endif
 
 class QWidget;
--- a/devices/mtpdevice.h	2017-04-11 18:31:14.000000000 +0200
+++ b/devices/mtpdevice.h	2017-08-03 14:24:32.834926438 +0200
@@ -27,7 +27,7 @@
 #include "fsdevice.h"
 #include "mpd-interface/song.h"
 #include "config.h"
-#include "solid-lite/portablemediaplayer.h"
+#include "solid/portablemediaplayer.h"
 #include <libmtp.h>
 
 class MusicLibraryItemRoot;
--- a/devices/umsdevice.cpp	2017-04-07 20:54:06.000000000 +0200
+++ b/devices/umsdevice.cpp	2017-08-03 14:24:32.834926438 +0200
@@ -30,7 +30,7 @@
 #include <QFile>
 #include <QFileInfo>
 #include <QTextStream>
-#include "solid-lite/storagedrive.h"
+#include "solid/storagedrive.h"
 
 static const QLatin1String constSettingsFile("/.is_audio_player");
 static const QLatin1String constMusicFolderKey("audio_folder");
--- a/devices/umsdevice.h	2017-04-07 20:04:17.000000000 +0200
+++ b/devices/umsdevice.h	2017-08-03 14:24:32.833926438 +0200
@@ -25,7 +25,7 @@
 #define UMSDEVICE_H
 
 #include "fsdevice.h"
-#include "solid-lite/storageaccess.h"
+#include "solid/storageaccess.h"
 
 class UmsDevice : public FsDevice
 {
--- a/models/devicesmodel.cpp	2017-04-08 11:21:34.000000000 +0200
+++ b/models/devicesmodel.cpp	2017-08-03 14:24:32.847926437 +0200
@@ -46,14 +46,14 @@
 #include <QStringList>
 #include <QMimeData>
 #include <QTimer>
-#include "solid-lite/device.h"
-#include "solid-lite/deviceinterface.h"
-#include "solid-lite/devicenotifier.h"
-#include "solid-lite/portablemediaplayer.h"
-#include "solid-lite/storageaccess.h"
-#include "solid-lite/storagedrive.h"
-#include "solid-lite/storagevolume.h"
-#include "solid-lite/opticaldisc.h"
+#include "solid/device.h"
+#include "solid/deviceinterface.h"
+#include "solid/devicenotifier.h"
+#include "solid/portablemediaplayer.h"
+#include "solid/storageaccess.h"
+#include "solid/storagedrive.h"
+#include "solid/storagevolume.h"
+#include "solid/opticaldisc.h"
 
 
 #include <QDebug>
--- a/replaygain/CMakeLists.txt	2017-04-08 11:46:29.000000000 +0200
+++ b/replaygain/CMakeLists.txt	2017-08-03 14:24:32.823926438 +0200
@@ -15,11 +15,11 @@
         QT5_WRAP_CPP(CANTATA_RG_MOC_SRCS ${CANTATA_RG_MOC_HDRS})
 
     if (WIN32)
-        set(CMAKE_BUILD_TYPE "Release")
+#IGNORE         set(CMAKE_BUILD_TYPE "Release")
         ADD_EXECUTABLE(cantata-replaygain WIN32 ${CANTATA_RG_SRCS} ${CANTATA_RG_MOC_SRCS} ${CANTATA_PO})
         install(TARGETS cantata-replaygain DESTINATION ${CMAKE_INSTALL_PREFIX})
     elseif (APPLE)
-        set(CMAKE_BUILD_TYPE "Release")
+#IGNORE         set(CMAKE_BUILD_TYPE "Release")
         ADD_EXECUTABLE(cantata-replaygain ${CANTATA_RG_SRCS} ${CANTATA_RG_MOC_SRCS} ${CANTATA_PO})
 #        install(TARGETS cantata-replaygain DESTINATION ${MACOSX_BUNDLE_APP_DIR})
     else ()
--- a/tags/CMakeLists.txt	2017-04-08 11:45:29.000000000 +0200
+++ b/tags/CMakeLists.txt	2017-08-03 14:24:32.824926438 +0200
@@ -5,11 +5,11 @@
 QT5_WRAP_CPP(CANTATA_TAGS_MOC_SRCS ${CANTATA_TAGS_MOC_HDRS})
 
 if (WIN32)
-    set(CMAKE_BUILD_TYPE "Release")
+#IGNORE     set(CMAKE_BUILD_TYPE "Release")
     ADD_EXECUTABLE(cantata-tags WIN32 ${CANTATA_TAGS_SRCS} ${CANTATA_TAGS_MOC_SRCS})
     install(TARGETS cantata-tags DESTINATION ${CMAKE_INSTALL_PREFIX})
 elseif (APPLE)
-    set(CMAKE_BUILD_TYPE "Release")
+#IGNORE     set(CMAKE_BUILD_TYPE "Release")
     ADD_EXECUTABLE(cantata-tags ${CANTATA_TAGS_SRCS} ${CANTATA_TAGS_MOC_SRCS})
     #install(TARGETS cantata-tags DESTINATION ${MACOSX_BUNDLE_APP_DIR})
 else ()
