diff --git a/src/GhcProject/Types.hs b/src/GhcProject/Types.hs
index ab01f6d..4d794bc 100644
--- a/src/GhcProject/Types.hs
+++ b/src/GhcProject/Types.hs
@@ -136,6 +136,9 @@ instance Binary ChModuleName where
 instance Binary ChComponentName where
     put = ggput . from
     get = to `fmap` ggget
+instance Binary ChLibraryName where
+    put = ggput . from
+    get = to `fmap` ggget
 instance Binary ChEntrypoint where
     put = ggput . from
     get = to `fmap` ggget
