diff --git a/src/Development/IDE/Core/Preprocessor.hs b/src/Development/IDE/Core/Preprocessor.hs
index 440edb8..01dba29 100644
--- a/src/Development/IDE/Core/Preprocessor.hs
+++ b/src/Development/IDE/Core/Preprocessor.hs
@@ -24,6 +24,7 @@ import SysTools (Option (..), runUnlit, runPp)
 import Control.Monad.Trans.Except
 import qualified GHC.LanguageExtensions as LangExt
 import Data.Maybe
+import ToolSettings
 import Control.Exception.Safe (catch, throw)
 import Data.IORef (IORef, modifyIORef, newIORef, readIORef)
 import Data.Text (Text)
@@ -166,12 +167,11 @@ runLhs dflags filename contents = withTempDir $ \dir -> do
     escape (c:cs)    = c : escape cs
     escape []        = []
 
-
 modifyOptP :: ([String] -> [String]) -> DynFlags -> DynFlags
-modifyOptP op = onSettings (onOptP op)
+modifyOptP op = onToolSettings (onOptP op)
   where
-    onSettings f x = x{settings = f $ settings x}
-    onOptP f x = x{sOpt_P = f $ sOpt_P x}
+    onToolSettings f x = x{toolSettings = f $ toolSettings x}
+    onOptP f x = x{toolSettings_opt_P = f $ toolSettings_opt_P x}
 
 -- | Run CPP on a file
 runCpp :: DynFlags -> FilePath -> Maybe SB.StringBuffer -> IO SB.StringBuffer
