--- a/support/errno.c	2017-05-11 16:07:24.991627923 +0200
+++ b/support/errno.c	2017-05-11 16:08:03.362643120 +0200
@@ -86,7 +86,7 @@
 	mph_return_if_size_t_overflow (n);
 
 	/* first, check for valid errnum */
-#if HOST_ANDROID
+#ifndef __GLIBC__
 	/* Android NDK defines _GNU_SOURCE but strerror_r follows the XSI semantics
 	 * not the GNU one. XSI version returns an integer, as opposed to the GNU one
 	 * which returns pointer to the buffer.
