--- CMakeLists.txt	2016-01-23 16:28:26.330122671 +0100
+++ CMakeLists.txt	2016-01-23 16:28:34.215429400 +0100
@@ -21,7 +21,7 @@
     set(LLVM_CMAKE_PATH "${LLVM_PATH}/cmake/modules")
   elseif (LLVM_CONFIG)
     message(STATUS "Found LLVM_CONFIG as ${LLVM_CONFIG}")
-    set(CONFIG_COMMAND ${LLVM_CONFIG} "--includedir" "--prefix" "--src-root")
+    set(CONFIG_COMMAND ${LLVM_CONFIG} "--includedir" "--prefix")
     execute_process(COMMAND ${CONFIG_COMMAND}
                     RESULT_VARIABLE HAD_ERROR
                     OUTPUT_VARIABLE CONFIG_OUTPUT)
