--- a/Package@swift-4.swift	2017-09-22 10:34:09.850953051 +0200
+++ b/Package@swift-4.swift	2017-09-22 10:34:21.047952792 +0200
@@ -4,8 +4,8 @@
 let package = Package(
     name: "Sockets",
     products: [
-        .library(name: "Transport", targets: ["Transport"]),
-        .library(name: "Sockets", targets: ["Sockets"]),
+        .library(name: "Transport", type: .dynamic, targets: ["Transport"]),
+        .library(name: "Sockets", type: .dynamic, targets: ["Sockets"]),
     ],
     targets: [
         .target(name: "Transport"),
