diff --git a/Package.swift b/Package.swift
index a9443d0..b6ec3dc 100644
--- a/Package.swift
+++ b/Package.swift
@@ -5,9 +5,5 @@ let package = Package(
     targets: [
         Target(name: "Transport"),
         Target(name: "Sockets", dependencies: ["Transport"])
-    ],
-    dependencies: [
-        // Core extensions, type-aliases, and functions that facilitate common tasks.
-        .Package(url: "https://github.com/vapor/core.git", Version(2,0,0, prereleaseIdentifiers: ["beta"]))
     ]
 )
