--- a/src/lldb/src/sb/sbcommandinterpreter.rs 2025-02-17 10:32:14.000000000 +0100
+++ b/src/lldb/src/sb/sbcommandinterpreter.rs 2025-02-22 21:45:58.927098949 +0100
@@ -1,12 +1,5 @@
 use super::*;

-cpp! {{
-namespace std {
-    // Prevent cpp from auto-deriving Default, which would use the default SBCommandInterpreter constructor
-    // introduced in v18, causing codelldb to be incompatible with earlier versions.
-    template<> struct is_default_constructible<lldb::SBCommandInterpreter> : std::false_type {};
-}
-}}
 cpp_class!(pub unsafe struct SBCommandInterpreter as "SBCommandInterpreter");

 unsafe impl Send for SBCommandInterpreter {}
