--- a/weaklink-a97ecc2e6467ae6316e14f2d8855275b7983b5a7/weaklink_build/src/stub_gen/mod.rs	2025-10-15 21:19:52.865689419 +0200
+++ b/weaklink-a97ecc2e6467ae6316e14f2d8855275b7983b5a7/weaklink_build/src/stub_gen/mod.rs	2025-10-15 21:20:16.117455424 +0200
@@ -14,7 +14,7 @@
         write_lines!(text,
             "global_asm!{{\""
             ".data"
-            ".p2align 2, 0x0"
+            ".p2align 3, 0x0"
             "{pfx}{symbol_table}:"
             "{entries}"
             "\"}}",
@@ -31,7 +31,7 @@
                 write_lines!(text,
                     "global_asm!{{\""
                     ".text"
-                    ".p2align 2, 0x0"
+                    ".p2align 3, 0x0"
                     ".global \\\"{symbol}\\\"" // Will be unescaped the 2nd time when compiling the generated module.
                     //".type   \\\"{symbol}\\\", function"
                     "\\\"{symbol}\\\":",
