diff --color -ur a/System.Native/CMakeLists.txt b/System.Native/CMakeLists.txt
--- a/System.Native/CMakeLists.txt	2024-07-16 19:44:51.000000000 +0200
+++ b/System.Native/CMakeLists.txt	2024-09-24 22:00:14.139212022 +0200
@@ -4,12 +4,6 @@
     add_definitions(-DHAS_CONSOLE_SIGNALS)
 endif ()
 
-if (CLR_CMAKE_TARGET_LINUX_MUSL)
-    # Fix up the thread stack size for MUSL to more reasonable size.
-    # TODO: https://github.com/dotnet/runtimelab/issues/791
-    add_definitions(-DENSURE_PRIMARY_STACK_SIZE)
-endif ()
-
 if (CLR_CMAKE_TARGET_OSX)
     add_definitions(-D_DARWIN_C_SOURCE)
 endif ()
diff --color -ur a/System.Native/pal_threading.c b/System.Native/pal_threading.c
--- a/System.Native/pal_threading.c	2024-07-16 19:44:51.000000000 +0200
+++ b/System.Native/pal_threading.c	2024-09-24 22:00:14.139212022 +0200
@@ -233,14 +233,6 @@
     error = pthread_attr_setdetachstate(&attrs, PTHREAD_CREATE_DETACHED);
     assert(error == 0);
 
-#ifdef ENSURE_PRIMARY_STACK_SIZE
-    // TODO: https://github.com/dotnet/runtimelab/issues/791
-    if (stackSize == 0)
-    {
-        stackSize = 1536 * 1024;
-    }
-#endif
-
     if (stackSize > 0)
     {
         if (stackSize < (uintptr_t)PTHREAD_STACK_MIN)
