--- a/src/native/libs/CMakeLists.txt	2025-02-15 17:38:23.031666000 +0100
+++ b/src/native/libs/CMakeLists.txt	2025-02-15 17:38:45.119443937 +0100
@@ -68,32 +68,6 @@
 
     string(TOUPPER ${CMAKE_BUILD_TYPE} UPPERCASE_CMAKE_BUILD_TYPE)
 
-    if (UPPERCASE_CMAKE_BUILD_TYPE STREQUAL DEBUG OR UPPERCASE_CMAKE_BUILD_TYPE STREQUAL CHECKED)
-        if (UPPERCASE_CMAKE_BUILD_TYPE STREQUAL DEBUG)
-            add_compile_options(-O0)
-        elseif (UPPERCASE_CMAKE_BUILD_TYPE STREQUAL CHECKED)
-            add_compile_options(-O2)
-        endif ()
-
-        add_definitions(-DDEBUG)
-    elseif (UPPERCASE_CMAKE_BUILD_TYPE STREQUAL RELEASE)
-        # Use O1 option when the clang version is smaller than 3.9
-        # Otherwise use O3 option in release build
-        if (CLR_CMAKE_TARGET_ARCH_ARMV7L AND DEFINED ENV{CROSSCOMPILE} AND CMAKE_C_COMPILER_VERSION VERSION_LESS 3.9)
-            add_compile_options (-O1)
-        else ()
-            if(CLR_CMAKE_TARGET_ANDROID)
-                # -O2 optimization generates faster/smaller code on Android
-                # TODO: This duplicates the settings in eng/native/configureoptimization.cmake, we should unify it
-                add_compile_options (-O2)
-            else()
-                add_compile_options (-O3)
-            endif ()
-        endif ()
-    else ()
-        message(FATAL_ERROR "Unknown build type. Set CMAKE_BUILD_TYPE to DEBUG or RELEASE.")
-    endif ()
-
     if (CLR_CMAKE_TARGET_APPLE)
         add_definitions(-D__APPLE_USE_RFC_3542)
     endif ()
