--- a/eng/native/configurecompiler.cmake  2025-08-21 17:38:47.000000000 +0200
+++ b/eng/native/configurecompiler.cmake  2025-09-12 17:54:13.122725763 +0200
@@ -563,11 +563,6 @@
     endif()
   endif(CLR_CMAKE_HOST_OSX OR CLR_CMAKE_HOST_MACCATALYST)
 
-  # Suppress warnings-as-errors in release branches to reduce servicing churn
-  if (PRERELEASE)
-    add_compile_options(-Werror)
-  endif(PRERELEASE)
-
   # Disabled common warnings
   add_compile_options(-Wno-unused-variable)
   add_compile_options(-Wno-unused-value)
