--- a/clang/lib/Driver/Types.cpp        2014-12-21 08:00:00.000000000 +0100
+++ b/clang/lib/Driver/Types.cpp        2015-02-28 16:06:09.089090828 +0100
@@ -156,8 +156,6 @@
            .Case("mi", TY_PP_ObjC)
            .Case("mm", TY_ObjCXX)
            .Case("rs", TY_RenderScript)
-           .Case("adb", TY_Ada)
-           .Case("ads", TY_Ada)
            .Case("asm", TY_PP_Asm)
            .Case("ast", TY_AST)
            .Case("ccm", TY_CXXModule)
