--- a/clang/lib/Driver/ToolChains/Gnu.cpp	2022-09-15 20:28:56.374355236 +0200
+++ b/clang/lib/Driver/ToolChains/Gnu.cpp	2022-09-16 10:59:00.879819290 +0200
@@ -599,6 +599,7 @@
   if (D.IsFlangMode()) {
     addFortranRuntimeLibraryPath(ToolChain, Args, CmdArgs);
     addFortranRuntimeLibs(ToolChain, CmdArgs);
+    ToolChain.AddCXXStdlibLibArgs(Args, CmdArgs);
     CmdArgs.push_back("-lm");
   }
 
