--- a/clang/lib/Driver/ToolChains/CommonArgs.cpp	2022-09-16 11:43:14.568959939 +0200
+++ b/clang/lib/Driver/ToolChains/CommonArgs.cpp	2022-09-16 11:43:30.178174739 +0200
@@ -758,12 +758,6 @@
 void tools::addFortranRuntimeLibraryPath(const ToolChain &TC,
                                          const llvm::opt::ArgList &Args,
                                          ArgStringList &CmdArgs) {
-  // NOTE: Generating executables by Flang is considered an "experimental"
-  // feature and hence this is guarded with a command line option.
-  // TODO: Make this work unconditionally once Flang is mature enough.
-  if (!Args.hasArg(options::OPT_flang_experimental_exec))
-    return;
-
   // Default to the <driver-path>/../lib directory. This works fine on the
   // platforms that we have tested so far. We will probably have to re-fine
   // this in the future. In particular, on some platforms, we may need to use
