--- a/src/cmd/link/internal/ld/lib.go	2025-04-11 14:52:08.325736104 +0200
+++ b/src/cmd/link/internal/ld/lib.go	2025-04-11 14:53:16.201020628 +0200
@@ -1670,21 +1670,6 @@
 			// Use lld to avoid errors from default linker (issue #38838)
 			altLinker = "lld"
 		}
-
-		if ctxt.Arch.InFamily(sys.ARM64) && buildcfg.GOOS == "linux" {
-			// Use ld.bfd as the default linker
-			altLinker = "bfd"
-
-			// Provide a useful error if ld.bfd is missing
-			name, args := flagExtld[0], flagExtld[1:]
-			args = append(args, "-fuse-ld=bfd", "-Wl,--version")
-			cmd := exec.Command(name, args...)
-			if out, err := cmd.CombinedOutput(); err == nil {
-				if !bytes.Contains(out, []byte("GNU ld")) {
-					log.Fatalf("ARM64 external linker must be ld.bfd, but is not: %s", out)
-				}
-			}
-		}
 	}
 	if ctxt.Arch.Family == sys.ARM64 && buildcfg.GOOS == "freebsd" {
 		// Switch to ld.bfd on freebsd/arm64.
