diff --git a/src/Language/Haskell/Brittany/Internal/Layouters/Expr.hs b/src/Language/Haskell/Brittany/Internal/Layouters/Expr.hs
index dac10c9..9288a60 100644
--- a/src/Language/Haskell/Brittany/Internal/Layouters/Expr.hs
+++ b/src/Language/Haskell/Brittany/Internal/Layouters/Expr.hs
@@ -1,4 +1,5 @@
 {-# LANGUAGE DataKinds #-}
+{-# LANGUAGE ViewPatterns #-}
 
 module Language.Haskell.Brittany.Internal.Layouters.Expr
   ( layoutExpr
@@ -17,7 +18,7 @@ import           Language.Haskell.Brittany.Internal.Config.Types
 
 import qualified Language.Haskell.GHC.ExactPrint.Types as ExactPrint.Types
 
-import           GHC ( runGhc, GenLocated(L), SrcSpan, moduleNameString, AnnKeywordId(..), RdrName(..) )
+import           GHC ( runGhc, GenLocated(L), SrcSpan, moduleNameString, AnnKeywordId(..), RdrName(..), dL )
 import           HsSyn
 import           Name
 import qualified FastString
@@ -112,8 +113,8 @@ layoutExpr lexpr@(L _ expr) = do
           -- by wrapping it in docSeq below. We _could_ add alignments for
           -- stuff like lists-of-lambdas. Nothing terribly important..)
           let shouldPrefixSeparator = case p of
-                (L _ LazyPat{}) -> isFirst
-                (L _ BangPat{}) -> isFirst
+                (dL->L _ LazyPat{}) -> isFirst
+                (dL->L _ BangPat{}) -> isFirst
                 _               -> False
           patDocSeq <- layoutPat p
           fixed <- case Seq.viewl patDocSeq of
