diff --git a/src/Language/Haskell/Brittany/Internal/Transformations/Alt.hs b/src/Language/Haskell/Brittany/Internal/Transformations/Alt.hs
index f4094ff..d1fc658 100644
--- a/src/Language/Haskell/Brittany/Internal/Transformations/Alt.hs
+++ b/src/Language/Haskell/Brittany/Internal/Transformations/Alt.hs
@@ -22,7 +22,7 @@ import           Language.Haskell.Brittany.Internal.Config.Types
 import           Language.Haskell.Brittany.Internal.Types
 
 import qualified Control.Monad.Memo as Memo
-
+import qualified Control.Monad.Fail as Fail
 
 
 data AltCurPos = AltCurPos
@@ -337,7 +337,7 @@ transformAlts =
           acp :: AltCurPos <- mGet
           tellDebugMess $ "transformAlts: BDFDEBUG " ++ s ++ " (node-id=" ++ show brDcId ++ "): acp=" ++ show acp
           reWrap . BDFDebug s <$> rec bd
-    processSpacingSimple :: (MonadMultiReader
+    processSpacingSimple :: (Fail.MonadFail m, MonadMultiReader
                                                      Config m,
                                                    MonadMultiState AltCurPos m, MonadMultiWriter (Seq String) m) => BriDocNumbered -> m ()
     processSpacingSimple bd = getSpacing bd >>= \case
@@ -362,7 +362,7 @@ transformAlts =
 
 getSpacing
   :: forall m
-   . (MonadMultiReader Config m, MonadMultiWriter (Seq String) m)
+   . (Fail.MonadFail m, MonadMultiReader Config m, MonadMultiWriter (Seq String) m)
   => BriDocNumbered
   -> m (LineModeValidity VerticalSpacing)
 getSpacing !bridoc = rec bridoc
