--- a/Cheapskate/ParserCombinators.hs
+++ b/Cheapskate/ParserCombinators.hs
@@ -109,7 +109,6 @@
 
 instance Monad Parser where
   return x = Parser $ \st -> Right (st, x)
-  fail e = Parser $ \st -> Left $ ParseError (position st) e
   p >>= g = Parser $ \st ->
     case evalParser p st of
          Left e        -> Left e
@@ -117,6 +116,9 @@
   {-# INLINE return #-}
   {-# INLINE (>>=) #-}
 
+instance MonadFail Parser where
+  fail e = Parser $ \st -> Left $ ParseError (position st) e
+
 instance MonadPlus Parser where
   mzero = Parser $ \st -> Left $ ParseError (position st) "(mzero)"
   mplus p1 p2 = Parser $ \st ->
