--- a/Source/JavaScriptCore/API/tests/GlobalContextWithFinalizerTest.cpp	2017-06-20 11:45:56.617113181 +0200
+++ b/Source/JavaScriptCore/API/tests/GlobalContextWithFinalizerTest.cpp	2017-06-20 11:41:02.425098697 +0200
@@ -23,6 +23,7 @@
  * THE POSSIBILITY OF SUCH DAMAGE.
  */
 
+#include <stdio.h>
 #include "config.h"
 #include "GlobalContextWithFinalizerTest.h"
 
--- a/Source/JavaScriptCore/API/JSRemoteInspector.h	2017-06-20 11:46:42.929115461 +0200
+++ b/Source/JavaScriptCore/API/JSRemoteInspector.h	2017-06-20 11:38:56.529092498 +0200
@@ -26,6 +26,7 @@
 #ifndef JSRemoteInspector_h
 #define JSRemoteInspector_h
 
+#include <sys/types.h>
 #include <JavaScriptCore/JSBase.h>
 #include <JavaScriptCore/WebKitAvailability.h>
 
--- a/Source/WebKit2/UIProcess/API/C/WKContextPrivate.h	2017-06-20 12:16:55.124204684 +0200
+++ b/Source/WebKit2/UIProcess/API/C/WKContextPrivate.h	2017-06-20 12:17:06.268205232 +0200
@@ -26,6 +26,7 @@
 #ifndef WKContextPrivate_h
 #define WKContextPrivate_h
 
+#include <sys/types.h>
 #include <WebKit/WKBase.h>
 #include <WebKit/WKContext.h>
 
--- a/Source/WebKit2/UIProcess/API/C/WKPagePrivate.h	2017-06-20 12:19:50.244213305 +0200
+++ b/Source/WebKit2/UIProcess/API/C/WKPagePrivate.h	2017-06-20 12:20:02.620213915 +0200
@@ -26,6 +26,7 @@
 #ifndef WKPagePrivate_h
 #define WKPagePrivate_h
 
+#include <sys/types.h>
 #include <WebKit/WKBase.h>
 #include <WebKit/WKPage.h>
 
