diff --git a/Package.swift b/Package.swift
index d5902c2..2de4ebc 100644
--- a/Package.swift
+++ b/Package.swift
@@ -10,12 +10,5 @@ let package = Package(
 
 		// RFC 2388
         Target(name: "FormData", dependencies: ["Multipart"])
-    ],
-    dependencies: [
-        // Core extensions, type-aliases, and functions that facilitate common tasks
-        .Package(url: "https://github.com/vapor/core.git", beta),
-
-        // HTTP package for HeaderKey type
-        .Package(url: "https://github.com/vapor/engine.git", beta),
     ]
 )
