diff --git a/cabal-plan.cabal b/cabal-plan.cabal
index 6ef84c8..b0666b5 100644
--- a/cabal-plan.cabal
+++ b/cabal-plan.cabal
@@ -69,64 +69,6 @@ library
 
   ghc-options: -Wall
 
-library topograph
-  default-language:    Haskell2010
-  other-extensions:    RankNTypes ScopedTypeVariables RecordWildCards
-  exposed-modules:     Topograph
-
-  build-depends:       base              (>= 4.6 && <4.10) || ^>= 4.10 || ^>= 4.11
-                     , base-compat       ^>= 0.9.3 || ^>=0.10.1
-                     , base-orphans      ^>= 0.6 || ^>=0.7 || ^>=0.8
-                     , containers        ^>= 0.5.0
-                     , vector            ^>= 0.12.0.1
-
-  hs-source-dirs:      src-topograph
-
-executable cabal-plan
-  default-language:    Haskell2010
-  other-extensions:    RecordWildCards
-
-  hs-source-dirs: src-exe
-  main-is: cabal-plan.hs
-  other-modules: Paths_cabal_plan, LicenseReport
-  autogen-modules: Paths_cabal_plan
-
-  if flag(exe)
-    -- dependencies w/ inherited version ranges via 'cabal-plan' library
-    build-depends: cabal-plan
-                 , topograph
-                 , base
-                 , text
-                 , containers
-                 , bytestring
-                 , directory
-
-    -- dependencies which require version bounds
-    build-depends: mtl            ^>= 2.2.1
-                 , ansi-terminal  ^>= 0.6.2 || ^>= 0.8.0.2
-                 , base-compat    ^>= 0.9.3 || ^>=0.10.1
-                 , optparse-applicative ^>= 0.13.0 || ^>= 0.14.0
-                 , parsec         ^>= 3.1.11
-                 , vector         ^>= 0.12.0.1
-
-
-    if flag(license-report)
-      build-depends: Cabal    ^>= 2.2.0.1
-                   , tar      ^>= 0.5.1.0
-                   , zlib     ^>= 0.6.2
-                   , filepath ^>= 1.4.1.2
-
-    if !impl(ghc >= 8.0)
-      build-depends:
-                   semigroups     ^>= 0.18.3
-
-    if flag(_)
-      cpp-options: -DUNDERLINE_SUPPORT
-  else
-    buildable: False
-
-  ghc-options: -Wall
-
 source-repository head
   type: git
   location: https://github.com/hvr/cabal-plan
