--- a/analitza/expression.h	2017-07-15 13:36:39.253616923 +0200
+++ b/analitza/expression.h	2017-07-15 13:37:00.891615143 +0200
@@ -116,7 +116,7 @@
         /**
          *    Returns the expression of the lambda body (without resolving the dependencies)
          */
-        Expression lambdaBody() const Q_REQUIRED_RESULT;
+        Q_REQUIRED_RESULT Expression lambdaBody() const;
         
         /**
          *    Returns whether it is a vector expression.
