diff --git a/lib/Distribution/Helper.hs b/lib/Distribution/Helper.hs
index 7dbfb2f..4c81900 100644
--- a/lib/Distribution/Helper.hs
+++ b/lib/Distribution/Helper.hs
@@ -120,7 +120,7 @@ import Text.Printf
 import GHC.Generics
 import Prelude
 
-import Paths_cabal_helper (getLibexecDir)
+import Paths_cabal_helper (getBinDir)
 import CabalHelper.Shared.InterfaceTypes
 import CabalHelper.Shared.Sandbox
 
@@ -462,7 +462,7 @@ instance Show LibexecNotFoundError where
 
 findLibexecExe :: IO FilePath
 findLibexecExe = do
-    libexecdir <- getLibexecDir
+    libexecdir <- getBinDir
     let exeName = "cabal-helper-wrapper"
         exe = libexecdir </> exeName FP.<.> exeExtension'
 
