--- a/C/hashes/hash.h	2023-02-06 11:37:46.000000000 +0100
+++ b/C/hashes/hash.h	2023-03-02 19:43:36.776061112 +0100
@@ -42,20 +42,4 @@
 
 #include "../7zTypes.h"
 
-#ifndef _UINT32_T_DECLARED
-typedef UInt32 uint32_t;
-#define _UINT32_T_DECLARED
-#endif
-
-#ifndef _UINT64_T_DECLARED
-typedef UInt64 uint64_t;
-#define _UINT64_T_DECLARED
-#endif
-
-//#include <stdint.h>
-
-#ifndef min
-#define min(a,b) (((a)>(b))?(b):(a))
-#endif
-
 #endif /* __hash_h__ */
