--- a/Source/WebCore/platform/UserAgentQuirks.cpp	2023-04-29 20:00:00.153501797 +0200
+++ b/Source/WebCore/platform/UserAgentQuirks.cpp	2023-04-29 20:00:22.657825179 +0200
@@ -119,6 +119,7 @@
     // agent to use the light version. Earlier versions even block users from
     // accessing the calendar.
     if (domain == "outlook.office.com"_s
+        || domain == "outlook.live.com"_s
         || domain == "mail.ntu.edu.tw"_s
         || domain == "exchange.tu-berlin.de"_s)
         return true;
