--- a/shared/nm-glib-aux/nm-macros-internal.h
+++ b/shared/nm-glib-aux/nm-macros-internal.h
@@ -1526,12 +1526,10 @@
 static inline char *
 _nm_strndup_a_step (char *s, const char *str, gsize len)
 {
-	NM_PRAGMA_WARNING_DISABLE ("-Wstringop-truncation");
 	if (len > 0)
 		strncpy (s, str, len);
 	s[len] = '\0';
 	return s;
-	NM_PRAGMA_WARNING_REENABLE;
 }
 
 /* Similar to g_strndup(), however, if the string (including the terminating
