--- a/shared/nm-utils/nm-shared-utils.h	2019-03-16 09:36:18.432316779 +0100
+++ b/shared/nm-utils/nm-shared-utils.h	2019-03-16 09:36:42.239231502 +0100
@@ -280,12 +280,10 @@
 static inline char *
 _nm_strndup_a_step (char *s, const char *str, gsize len)
 {
-	NM_PRAGMA_WARNING_DISABLE ("-Wstringop-truncation");
 	if (len > 0)
 		strncpy (s, str, len);
 	s[len] = '\0';
 	return s;
-	NM_PRAGMA_WARNING_REENABLE;
 }
 
 /* Similar to g_strndup(), however, if the string (including the terminating
