--- a/shared/nm-glib-aux/nm-json-aux.c	2021-02-04 15:22:31.550129499 +0100
+++ b/shared/nm-glib-aux/nm-json-aux.c	2021-02-04 15:22:52.674850158 +0100
@@ -172,7 +172,7 @@
 
     handle = dlopen(soname,
                     RTLD_LAZY | RTLD_LOCAL | RTLD_NODELETE
-#if !defined(ASAN_BUILD)
+#if !defined(ASAN_BUILD) && defined(__GLIBC__)
                         | RTLD_DEEPBIND
 #endif
                         | 0);
