--- a/src/versit/vobject.c.bak	2016-02-17 19:16:11.177153901 +0100
+++ b/src/versit/vobject.c	2016-02-17 19:16:44.777151899 +0100
@@ -46,6 +46,7 @@
 
 #include "vobject.h"
 #include <string.h>
+#include <strings.h>
 #include <stdio.h>
 
 #define NAME_OF(o)          o->id
--- a/src/versit/vcc.y.bak	2016-02-17 19:16:33.403152577 +0100
+++ b/src/versit/vcc.y	2016-02-17 19:16:59.752151006 +0100
@@ -99,6 +99,7 @@
 #endif
 
 #include <string.h>
+#include <strings.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <ctype.h>
--- a/src/versit/vcc.c.bak	2016-02-17 19:19:00.129143831 +0100
+++ b/src/versit/vcc.c	2016-02-17 19:19:12.977143065 +0100
@@ -164,6 +164,7 @@
 #endif
 
 #include <string.h>
+#include <strings.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <ctype.h>
