--- a/Makefile.am	2018-11-25 19:51:29.756923422 +0100
+++ b/Makefile.am	2018-11-25 19:51:47.188896516 +0100
@@ -14,7 +14,7 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
-SUBDIRS = marisa-glib libkkc tools data tests po
+SUBDIRS = marisa-glib libkkc tools data po
 DISTCHECK_CONFIGURE_FLAGS = --enable-docs
 ACLOCAL_AMFLAGS = -I m4
 
@@ -22,7 +22,7 @@
 SUBDIRS += docs
 endif
 
-DIST_SUBDIRS = marisa-glib libkkc tools data tests po docs
+DIST_SUBDIRS = marisa-glib libkkc tools data po docs
 
 GITIGNOREFILES =							\
 	ABOUT-NLS							\
