--- a/rdma/rdma.c	2023-12-04 21:59:14.711211789 +0100
+++ b/rdma/rdma.c	2023-12-04 21:59:28.543049030 +0100
@@ -4,6 +4,8 @@
  * Authors:     Leon Romanovsky <leonro@mellanox.com>
  */
 
+#include <libgen.h>
+
 #include "rdma.h"
 #include "version.h"
 #include "color.h"
--- a/lib/bpf_legacy.c	2023-12-04 21:58:14.535920287 +0100
+++ b/lib/bpf_legacy.c	2023-12-04 21:58:35.827669519 +0100
@@ -18,6 +18,7 @@
 #include <stdarg.h>
 #include <limits.h>
 #include <assert.h>
+#include <libgen.h>
 
 #ifdef HAVE_ELF
 #include <libelf.h>
