--- a/tc/f_flower.c
+++ b/tc/f_flower.c
@@ -17,6 +17,7 @@
 #include <linux/if_arp.h>
 #include <linux/if_ether.h>
 #include <linux/ip.h>
+#include <linux/limits.h>
 #include <linux/tc_act/tc_vlan.h>
 #include <linux/mpls.h>
 
--- a/devlink/devlink.c
+++ b/devlink/devlink.c
@@ -13,6 +13,7 @@
 #include <stdlib.h>
 #include <stdarg.h>
 #include <string.h>
+#include <signal.h>
 #include <stdbool.h>
 #include <unistd.h>
 #include <getopt.h>
