diff --git a/include/libnetlink.h b/include/libnetlink.h
index 30f0c2d2..7074e913 100644
--- a/include/libnetlink.h
+++ b/include/libnetlink.h
@@ -4,6 +4,7 @@

 #include <stdio.h>
 #include <string.h>
+#include <endian.h>
 #include <asm/types.h>
 #include <linux/netlink.h>
 #include <linux/rtnetlink.h>
diff --git a/rdma/rdma.c b/rdma/rdma.c
index 131c6b2a..f58c8c3b 100644
--- a/rdma/rdma.c
+++ b/rdma/rdma.c
@@ -8,6 +8,8 @@
 #include "version.h"
 #include "color.h"

+#include <libgen.h>
+
 /* Global utils flags */
 int json;
