--- a/src/libsystemd/sd-event/sd-event.c	2016-07-17 16:52:32.111542591 +0200
+++ b/src/libsystemd/sd-event/sd-event.c	2016-07-17 16:52:38.235542087 +0200
@@ -2286,7 +2286,7 @@
                          s->child.siginfo.si_code == CLD_KILLED ||
                          s->child.siginfo.si_code == CLD_DUMPED;
 
-                r = s->child.callback(s, &s->child.siginfo, s->userdata);
+//                r = s->child.callback(s, &s->child.siginfo, s->userdata);
 
                 /* Now, reap the PID for good. */
                 if (zombie)
