--- a/src/basic/missing.h
+++ b/src/basic/missing.h
@@ -1162,6 +1162,11 @@ struct input_mask {
 #define RENAME_NOREPLACE (1 << 0)
 #endif
 
+#if !defined(__GLIBC__)
+typedef int (*__compar_fn_t)(const void *, const void *);
+typedef __compar_fn_t comparison_fn_t;
+#endif
+
 #ifndef KCMP_FILE
 #define KCMP_FILE 0
 #endif
--- a/src/basic/strbuf.c
+++ b/src/basic/strbuf.c
@@ -24,6 +24,7 @@ 
 #include "alloc-util.h"
 #include "strbuf.h"
 #include "util.h"
+#include "missing.h"
 
 /*
  * Strbuf stores given strings in a single continuous allocated memory
