--- a/src/basic/parse-util.c
+++ b/src/basic/parse-util.c
@@ -23,7 +23,7 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
-#include <xlocale.h>
+#include "missing.h"
 
 #include "alloc-util.h"
 #include "extract-word.h"
--- a/src/shared/machine-image.c
+++ b/src/shared/machine-image.c
@@ -25,6 +25,7 @@
 #include <string.h>
 #include <sys/file.h>
 #include <sys/stat.h>
+#include <sys/file.h>
 #include <unistd.h>
 #include <linux/fs.h>
 #include "alloc-util.h"
--- a/src/shared/machine-pool.c
+++ b/src/shared/machine-pool.c
@@ -31,6 +31,7 @@
 #include <sys/stat.h>
 #include <sys/statfs.h>
 #include <sys/statvfs.h>
+#include <sys/file.h>
 #include <unistd.h>
 
 #include "sd-bus-protocol.h"
--- a/src/nspawn/nspawn-network.c	2016-07-17 16:46:57.983570084 +0200
+++ b/src/nspawn/nspawn-network.c	2016-07-17 16:47:41.357566516 +0200
@@ -19,6 +19,7 @@
 
 #include <linux/veth.h>
 #include <net/if.h>
+#include <sys/file.h>
 
 #include "libudev.h"
 #include "sd-id128.h"
--- a/src/nspawn/nspawn-stub-pid1.c	2016-07-17 16:46:22.333573018 +0200
+++ b/src/nspawn/nspawn-stub-pid1.c	2016-07-17 16:46:29.407572436 +0200
@@ -18,7 +18,7 @@
 ***/
 
 #include <sys/reboot.h>
-#include <sys/unistd.h>
+#include <unistd.h>
 #include <sys/wait.h>
 
 #include "fd-util.h"
--- a/src/network/netdev/tunnel.c	2017-03-02 16:09:08.142401823 +0100
+++ b/src/network/netdev/tunnel.c	2017-03-02 16:09:16.139671091 +0100
@@ -20,7 +20,6 @@
 #include <arpa/inet.h>
 #include <net/if.h>
 #include <linux/ip.h>
-#include <linux/if_tunnel.h>
 #include <linux/ip6_tunnel.h>
 
 #include "sd-netlink.h"
