From b2dbc5b510233c4993ebaff160a63e7cadec694c Mon Sep 17 00:00:00 2001
From: Emil Renner Berthing <systemd@esmil.dk>
Date: Sat, 12 Sep 2015 19:56:52 +0000
Subject: [PATCH] socket-util: don't fail if libc doesn't support IDN

Signed-off-by: Emil Renner Berthing <systemd@esmil.dk>
Signed-off-by: Khem Raj <raj.khem@gmail.com>
---
 src/basic/socket-util.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/src/basic/socket-util.c b/src/basic/socket-util.c
index 0f38f9a..951c30d 100644
--- a/src/basic/socket-util.c
+++ b/src/basic/socket-util.c
@@ -44,6 +44,15 @@
 #include "string-util.h"
 #include "user-util.h"
 #include "util.h"
+/* Don't fail if the standard library
+ * doesn't support IDN */
+#ifndef NI_IDN
+#define NI_IDN 0
+#endif
+
+#ifndef NI_IDN_USE_STD3_ASCII_RULES
+#define NI_IDN_USE_STD3_ASCII_RULES 0
+#endif
 
 int socket_address_parse(SocketAddress *a, const char *s) {
         char *e, *n;
