--- a/Makefile.am	2016-07-25 21:49:47.000000000 +0200
+++ b/Makefile.am	2016-11-22 14:11:59.662456683 +0100
@@ -2867,9 +2867,8 @@
 	$(AM_V_CC)$(EFI_CC) $(efi_cppflags) $(efi_cflags) -c $< -o $@
 
 $(systemd_boot_solib): $(systemd_boot_objects)
-	$(AM_V_CCLD)$(LD) $(efi_ldflags) $(systemd_boot_objects) \
-		-o $@ -lefi -lgnuefi $(shell $(CC) -print-libgcc-file-name); \
-	nm -D -u $@ | grep ' U ' && exit 1 || :
+	$(AM_V_CCLD)ld.bfd -nostrip $(efi_ldflags) $(systemd_boot_objects) \
+		-o $@ -lefi -lgnuefi $(shell $(CC) -print-libgcc-file-name);
 
 $(systemd_boot): $(systemd_boot_solib)
 	$(AM_V_GEN)$(OBJCOPY) -j .text -j .sdata -j .data -j .dynamic \
@@ -2917,9 +2916,8 @@
 	$(AM_V_CC)$(EFI_CC) $(efi_cppflags) $(efi_cflags) -c $< -o $@
 
 $(stub_solib): $(stub_objects)
-	$(AM_V_CCLD)$(LD) $(efi_ldflags) $(stub_objects) \
-		-o $@ -lefi -lgnuefi $(shell $(CC) -print-libgcc-file-name); \
-	nm -D -u $@ | grep ' U ' && exit 1 || :
+	$(AM_V_CCLD)ld.bfd -nostrip $(efi_ldflags) $(stub_objects) \
+		-o $@ -lefi -lgnuefi $(shell $(CC) -print-libgcc-file-name);
 
 $(stub): $(stub_solib)
 	$(AM_V_GEN)$(OBJCOPY) -j .text -j .sdata -j .data -j .dynamic \
