--- a/configure.ac	2017-03-02 15:24:02.627249466 +0100
+++ b/configure.ac	2017-03-02 15:36:50.844504796 +0100
@@ -167,8 +167,6 @@
         -Werror=implicit-function-declaration \
         -Werror=missing-declarations \
         -Werror=return-type \
-        -Werror=incompatible-pointer-types \
-        -Werror=format=2 \
         -Wstrict-prototypes \
         -Wredundant-decls \
         -Wmissing-noreturn \
