--- src/shared/pager.c	2015-10-30 22:15:32.736483956 +0100
+++ src/shared/pager.c	2015-10-30 22:19:21.488237983 +0100
@@ -152,11 +152,9 @@
                 return;
 
         /* Inform pager that we are done */
-        fclose(stdout);
-        stdout = NULL;
+        fclose(stdout);
 
-        fclose(stderr);
-        stderr = NULL;
+        fclose(stderr);
 
         kill(pager_pid, SIGCONT);
         (void) wait_for_terminate(pager_pid, NULL);
