--- a/Modules/selectmodule.c	2017-01-04 17:23:18.517130959 +0100
+++ b/Modules/selectmodule.c	2017-01-04 17:22:14.485216076 +0100
@@ -1932,8 +1932,10 @@
     PyModule_AddIntConstant(m, "KQ_EV_EOF", EV_EOF);
     PyModule_AddIntConstant(m, "KQ_EV_ERROR", EV_ERROR);
 
+#ifndef __linux__
     /* READ WRITE filter flag */
     PyModule_AddIntConstant(m, "KQ_NOTE_LOWAT", NOTE_LOWAT);
+#endif
 
     /* VNODE filter flags  */
     PyModule_AddIntConstant(m, "KQ_NOTE_DELETE", NOTE_DELETE);
@@ -1942,7 +1944,9 @@
     PyModule_AddIntConstant(m, "KQ_NOTE_ATTRIB", NOTE_ATTRIB);
     PyModule_AddIntConstant(m, "KQ_NOTE_LINK", NOTE_LINK);
     PyModule_AddIntConstant(m, "KQ_NOTE_RENAME", NOTE_RENAME);
+#ifndef __linux__
     PyModule_AddIntConstant(m, "KQ_NOTE_REVOKE", NOTE_REVOKE);
+#endif
 
     /* PROC filter flags  */
     PyModule_AddIntConstant(m, "KQ_NOTE_EXIT", NOTE_EXIT);
--- a/setup.py	2017-01-04 17:34:17.435041248 +0100
+++ b/setup.py	2017-01-04 17:38:13.957598005 +0100
@@ -693,7 +693,10 @@
             missing.append('spwd')
 
         # select(2); not on ancient System V
-        exts.append( Extension('select', ['selectmodule.c']) )
+        libs = []
+        if (config_h_vars.get('HAVE_KQUEUE', True)):
+            libs = ['kqueue']
+        exts.append( Extension('select', ['selectmodule.c'], libraries=libs) )
 
         # Fred Drake's interface to the Python parser
         exts.append( Extension('parser', ['parsermodule.c']) )
