--- a/configure.ac	2016-09-09 11:16:56.638129779 +0200
+++ b/configure.ac	2016-09-09 11:25:02.614607062 +0200
@@ -16,7 +16,6 @@
 AC_CONFIG_AUX_DIR([build-aux])
 AC_CONFIG_SRCDIR(src/core/display.c)
 AC_CONFIG_HEADERS(config.h)
-AC_CONFIG_SUBDIRS([cogl clutter])
 
 AM_INIT_AUTOMAKE([1.11 foreign no-dist-gzip dist-xz tar-ustar subdir-objects])
 m4_ifdef([AM_SILENT_RULES],[AM_SILENT_RULES([yes])],)
@@ -63,12 +62,16 @@
 
 LIBWACOM_VERSION=0.13
 
+CLUTTER_PACKAGE=mutter-clutter-1.0
+
 MUTTER_PC_MODULES="
    gtk+-3.0 >= 3.19.8
    gio-unix-2.0 >= 2.35.1
    pango >= 1.2.0
    cairo >= 1.10.0
    gsettings-desktop-schemas >= 3.21.4
+   $CLUTTER_PACKAGE
+   mutter-cogl-1.0
    json-glib-1.0
    upower-glib >= 0.99.0
    gnome-desktop-3.0
@@ -239,7 +242,7 @@
 
 PKG_CHECK_MODULES(MUTTER, $MUTTER_PC_MODULES)
 
-MUTTER_NATIVE_BACKEND_MODULES="libdrm libsystemd libinput >= 1.4 gudev-1.0 gbm >= 10.3"
+MUTTER_NATIVE_BACKEND_MODULES="mutter-clutter-egl-1.0 libdrm libsystemd libinput >= 1.4 gudev-1.0 gbm >= 10.3"
 
 AC_ARG_ENABLE(native-backend,
   AS_HELP_STRING([--disable-native-backend], [disable mutter native (KMS) backend]),,
@@ -255,7 +258,7 @@
 ])
 AM_CONDITIONAL([HAVE_NATIVE_BACKEND],[test "$have_native_backend" = "yes"])
 
-MUTTER_WAYLAND_MODULES="wayland-server >= 1.6.90"
+MUTTER_WAYLAND_MODULES="mutter-clutter-wayland-1.0 clutter-wayland-compositor-1.0 wayland-server >= 1.6.90"
 
 AC_ARG_ENABLE(wayland,
   AS_HELP_STRING([--disable-wayland], [disable mutter on wayland support]),,
@@ -281,6 +284,10 @@
 PKG_CHECK_EXISTS([xi >= 1.6.99.1],
                  AC_DEFINE([HAVE_XI23],[1],[Define if you have support for XInput 2.3 or greater]))
 
+# This is used for plugins
+AC_SUBST(CLUTTER_PACKAGE)
+PKG_CHECK_MODULES(CLUTTER, $CLUTTER_PACKAGE)
+
 AC_PATH_XTRA
 
 ALL_X_LIBS="$X_LIBS $X_PRE_LIBS -lX11 $X_EXTRA_LIBS"
--- a/src/Makefile.am	2016-09-09 11:20:46.641985434 +0200
+++ b/src/Makefile.am	2016-09-09 11:37:45.888604069 +0200
@@ -24,15 +24,9 @@
 	-I$(srcdir)/core						\
 	-I$(srcdir)/ui							\
 	-I$(srcdir)/compositor						\
-	-I$(top_srcdir)/cogl						\
 	-I$(top_srcdir)/cogl/cogl					\
 	-I$(top_srcdir)/cogl/cogl/winsys				\
-	-I$(top_builddir)/cogl/cogl					\
-	-I$(top_builddir)/cogl						\
-	-I$(top_srcdir)/clutter						\
 	-I$(top_srcdir)/clutter/clutter					\
-	-I$(top_builddir)/clutter					\
-	-I$(top_builddir)/clutter/clutter				\
 	-DMUTTER_LIBEXECDIR=\"$(libexecdir)\"				\
 	-DMUTTER_LOCALEDIR=\"$(localedir)\"				\
 	-DMUTTER_PKGDATADIR=\"$(pkgdatadir)\"				\
@@ -383,11 +377,7 @@
 nodist_libmutter_la_SOURCES = $(mutter_built_sources)
 
 libmutter_la_LDFLAGS = -no-undefined -export-symbols-regex "^(meta|ag)_.*"
-libmutter_la_LIBADD  = \
-	$(MUTTER_LIBS) \
-	$(MUTTER_NATIVE_BACKEND_LIBS)	\
-	$(top_builddir)/clutter/clutter/libmutter-clutter-1.0.la \
-	$(NULL)
+libmutter_la_LIBADD  = $(MUTTER_LIBS) $(MUTTER_NATIVE_BACKEND_LIBS)
 
 # Headers installed for plugins; introspected information will
 # be extracted into Mutter-<version>.gir
@@ -472,8 +462,6 @@
 typelib_DATA = Meta-$(api_version).typelib
 
 INTROSPECTION_GIRS = Meta-$(api_version).gir
-INTROSPECTION_SCANNER_ARGS = --add-include-path=$(top_builddir)/clutter/clutter --add-include-path=$(top_builddir)/cogl/cogl --add-include-path=$(top_builddir)/cogl/cogl-pango
-INTROSPECTION_COMPILER_ARGS = --includedir=$(top_builddir)/clutter/clutter --includedir=$(top_builddir)/cogl/cogl --includedir=$(top_builddir)/cogl/cogl-pango
 
 Meta-$(api_version).gir: libmutter.la
 @META_GIR@_INCLUDES = GObject-2.0 GDesktopEnums-3.0 Gdk-3.0 Gtk-3.0 Clutter-1.0 xlib-2.0 xfixes-4.0 Cogl-1.0
--- a/src/compositor/plugins/Makefile.am	2016-09-09 11:22:21.340737959 +0200
+++ b/src/compositor/plugins/Makefile.am	2016-09-09 11:40:43.095872356 +0200
@@ -5,10 +5,6 @@
 	$(MUTTER_CFLAGS)						\
 	-I$(top_builddir)/src						\
 	-I$(top_srcdir)/src						\
-	-I$(top_srcdir)/cogl						\
-	-I$(top_builddir)/cogl						\
-	-I$(top_srcdir)/clutter						\
-	-I$(top_builddir)/clutter					\
 	-I$(top_builddir)/clutter/clutter				\
 	-DMUTTER_LIBEXECDIR=\"$(libexecdir)\"				\
 	-DMUTTER_LOCALEDIR=\"$(localedir)\"				\
