--- a/src/coreclr/src/pal/src/CMakeLists.txt	2020-11-14 00:56:09.442947775 +0100
+++ b/src/coreclr/src/pal/src/CMakeLists.txt	2020-11-14 00:56:19.858033703 +0100
@@ -82,9 +82,6 @@
 if(CLR_CMAKE_HOST_ALPINE_LINUX)
   # Setting RLIMIT_NOFILE breaks debugging of coreclr on Alpine Linux for some reason
   add_definitions(-DDONT_SET_RLIMIT_NOFILE)
-  # On Alpine Linux, we need to ensure that the reported stack range for the primary thread is
-  # larger than the initial committed stack size.
-  add_definitions(-DENSURE_PRIMARY_STACK_SIZE)
 endif(CLR_CMAKE_HOST_ALPINE_LINUX)
 
 # turn off capability to remove unused functions (which was enabled in debug build with sanitizers)
