--- a/src/coreclr/src/vm/CMakeLists.txt	2020-05-20 15:13:48.834035410 +0200
+++ b/src/coreclr/src/vm/CMakeLists.txt	2020-05-20 15:17:18.821263054 +0200
@@ -847,8 +847,6 @@
 elseif(CLR_CMAKE_TARGET_ARCH_ARM64)
     set(VM_SOURCES_DAC_AND_WKS_ARCH
         ${ARCH_SOURCES_DIR}/stubs.cpp
-        exceptionhandling.cpp
-        gcinfodecoder.cpp
     )
 
     set(VM_HEADERS_DAC_AND_WKS_ARCH
@@ -858,6 +856,8 @@
 
     set(VM_SOURCES_WKS_ARCH
         ${ARCH_SOURCES_DIR}/profiler.cpp
+        exceptionhandling.cpp
+        gcinfodecoder.cpp
     )
 
     if(CLR_CMAKE_HOST_UNIX)
