--- a/CMakeLists.txt	2021-11-08 21:00:28.949045256 +0100
+++ b/CMakeLists.txt	2021-11-08 21:03:28.528178648 +0100
@@ -203,10 +203,12 @@
 
     find_program(DOTNETCLI dotnet PATHS "${DOTNET_DIR}" ENV PATH NO_DEFAULT_PATH)
 
+    file(GLOB MANAGED_SRC "${CMAKE_CURRENT_SOURCE_DIR}/managed/*.cs")
+
     add_custom_command(OUTPUT ${DOTNET_BUILD_RESULT}
       COMMAND ${DOTNETCLI} publish ${MANAGEDPART_PROJECT} -o ${CMAKE_CURRENT_BINARY_DIR} /p:BaseIntermediateOutputPath=${CMAKE_CURRENT_BINARY_DIR}/obj/ /p:BaseOutputPath=${CMAKE_CURRENT_BINARY_DIR}/bin/
       WORKING_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}"
-      DEPENDS "${CMAKE_CURRENT_SOURCE_DIR}/managed/*.cs" "${MANAGEDPART_PROJECT}"
+      DEPENDS "${MANAGED_SRC}" "${MANAGEDPART_PROJECT}"
       COMMENT "Compiling ${MANAGEDPART_DLL_NAME}"
       VERBATIM
     )
--- a/compileoptions.cmake	2021-11-08 21:09:01.878111330 +0100
+++ b/compileoptions.cmake	2021-11-08 21:07:14.963852969 +0100
@@ -1,5 +1,5 @@
 if (CLR_CMAKE_PLATFORM_UNIX)
-  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++14")
 
   # this allows pretty-printers in gdb to work with clang
   set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -D_GLIBCXX_DEBUG")
