--- a/configure.ac	2016-10-16 13:17:44.158773980 +0200
+++ b/configure.ac	2016-10-16 13:18:08.125772008 +0200
@@ -118,36 +118,6 @@
 ALSA_PKGCONF_DIR="$pkgconfdir"
 AC_SUBST(ALSA_PKGCONF_DIR)
 
-dnl Check for versioned symbols
-AC_MSG_CHECKING(for versioned symbols)
-AC_ARG_WITH(versioned,
-  AS_HELP_STRING([--with-versioned],
-    [shared library will be compiled with versioned symbols (default = yes)]),
-  versioned="$withval", versioned="yes")
-if test "$versioned" = "yes"; then
-  # it seems that GNU ld versions since 2.10 are not broken
-  xres=`grep '^VERSION=' ${srcdir}/ltmain.sh | cut -d = -f 2 | cut -d \" -f 2`
-  major=`echo $xres | cut -d . -f 1`
-  minor=`echo $xres | cut -d . -f 2`
-  pass=0
-  if test $major -eq 1 && test $minor -gt 3; then
-    pass=1
-  else
-    if test $major -gt 1; then
-      pass=1
-    fi
-  fi
-  if test $pass -eq 1; then
-    AC_DEFINE(VERSIONED_SYMBOLS,,[compiled with versioned symbols])
-    AC_MSG_RESULT(yes)
-  else
-    AC_MSG_RESULT(broken libtool - use libtool v1.4+; no versions)
-  fi
-else
-  AC_MSG_RESULT(no)
-fi
-AM_CONDITIONAL([VERSIONED_SYMBOLS], [test x$versioned = xyes])
-
 dnl Check for symbolic-functions
 AC_MSG_CHECKING(for symbolic-functions)
 AC_ARG_ENABLE(symbolic-functions,
@@ -159,10 +129,10 @@
     AC_MSG_RESULT(yes)
   else
     AC_MSG_RESULT(not supported by ld)
-    symfuncs="no"
+    symfuncs="yes"
   fi
 else
-  AC_MSG_RESULT(no)
+  AC_MSG_RESULT(yes)
 fi
 AM_CONDITIONAL([SYMBOLIC_FUNCTIONS], [test x"$symfuncs" = xyes])
 
--- a/src/Makefile.am	2016-10-16 13:19:21.679765956 +0200
+++ b/src/Makefile.am	2016-10-16 13:19:50.075763619 +0200
@@ -1,12 +1,6 @@
 EXTRA_DIST=Versions
 COMPATNUM=@LIBTOOL_VERSION_INFO@
 
-if VERSIONED_SYMBOLS
-VSYMS = -Wl,--version-script=Versions
-else
-VSYMS =
-endif
-
 if SYMBOLIC_FUNCTIONS
 SYMFUNCS = -Wl,-Bsymbolic-functions
 else
@@ -53,7 +47,7 @@
 SUBDIRS += conf
 libasound_la_LIBADD += @ALSA_DEPLIBS@
 
-libasound_la_LDFLAGS = -version-info $(COMPATNUM) $(VSYMS) $(SYMFUNCS) $(LDFLAGS_NOUNDEFINED)
+libasound_la_LDFLAGS = -version-info $(COMPATNUM) $(SYMFUNCS) $(LDFLAGS_NOUNDEFINED)
 
 control/libcontrol.la:
 	$(MAKE) -C control libcontrol.la
