--- a/src/libtracker-extract/tracker-module-manager.h	2016-10-05 13:52:18.000000000 +0200
+++ b/src/libtracker-extract/tracker-module-manager.h	2016-11-25 16:27:02.527712924 +0100
@@ -68,7 +68,7 @@
 typedef gboolean (* TrackerExtractMetadataFunc) (TrackerExtractInfo *info);
 
 
-gboolean  tracker_extract_module_manager_init                (void) G_GNUC_CONST;
+gboolean  tracker_extract_module_manager_init                (void) G_GNUC_CONST __attribute__ ((optnone));
 GModule * tracker_extract_module_manager_get_for_mimetype    (const gchar                  *mimetype,
                                                               TrackerExtractInitFunc       *init_func,
                                                               TrackerExtractShutdownFunc   *shutdown_func,
