--- a/configure.ac	2017-02-23 11:41:22.000000000 +0100
+++ b/configure.ac	2017-05-13 12:16:45.013098016 +0200
@@ -2592,21 +2592,6 @@
 AM_CONDITIONAL(HAVE_ARTWORK, test "x$have_artwork" = "xyes")
 
 ##################################################################
-# Check for libseccomp
-##################################################################
-
-PKG_CHECK_MODULES(LIBSECCOMP,
-                  [libseccomp >= 2.0],
-                  [have_libseccomp=yes],
-                  [have_libseccomp=no])
-
-if test "$have_libseccomp" = "yes"; then
-   LIBTRACKER_COMMON_LIBS="$LIBTRACKER_COMMON_LIBS $LIBSECCOMP_LIBS"
-   LIBTRACKER_COMMON_CFLAGS="$LIBTRACKER_COMMON_CFLAGS $LIBSECCOMP_CFLAGS"
-   AC_DEFINE(HAVE_LIBSECCOMP, [], [Define if we have libseccomp])
-fi
-
-##################################################################
 # Write generated files
 ##################################################################
 
