--- a/configure.ac	2017-05-24 17:19:53.804350087 +0200
+++ b/configure.ac	2017-05-24 17:20:13.932349526 +0200
@@ -2606,24 +2606,6 @@
 AM_CONDITIONAL(HAVE_ARTWORK, test "x$have_artwork" = "xyes")
 
 ##################################################################
-# Check for libseccomp
-##################################################################
-
-PKG_CHECK_MODULES(LIBSECCOMP,
-                  [libseccomp >= 2.0],
-                  [have_libseccomp=yes],
-                  [have_libseccomp=no])
-
-if test "$have_libseccomp" = "yes"; then
-   LIBTRACKER_COMMON_LIBS="$LIBTRACKER_COMMON_LIBS $LIBSECCOMP_LIBS"
-   LIBTRACKER_COMMON_CFLAGS="$LIBTRACKER_COMMON_CFLAGS $LIBSECCOMP_CFLAGS"
-   AC_DEFINE(HAVE_LIBSECCOMP, [], [Define if we have libseccomp])
-# seccomp is a linux thing
-elif test "$tracker_os_linux" = "yes"; then
-   AC_MSG_ERROR([Libseccomp is mandatory for sandboxed metadata extraction])
-fi
-
-##################################################################
 # Write generated files
 ##################################################################
 
