--- a/src/extra/intl/Makefile.in	2017-07-08 22:51:52.366639031 +0200
+++ b/src/extra/intl/Makefile.in	2017-07-08 22:53:34.981644083 +0200
@@ -57,7 +57,7 @@
 
 libintl.a: $(libintl_a_OBJECTS)
 	@rm -f $@
-	$(AR) -cr $@ $(libintl_a_OBJECTS)
+	$(AR) cr $@ $(libintl_a_OBJECTS)
 	$(RANLIB) $@
 
 headers: libgnuintl.h
--- a/src/extra/tre/Makefile.in	2017-07-08 22:52:03.301639570 +0200
+++ b/src/extra/tre/Makefile.in	2017-07-08 22:53:50.684644856 +0200
@@ -50,7 +50,7 @@
 
 libtre.a: $(libtre_a_OBJECTS)
 	@rm -f $@
-	$(AR) -cr $@ $(libtre_a_OBJECTS)
+	$(AR) cr $@ $(libtre_a_OBJECTS)
 	$(RANLIB) $@
 
 mostlyclean: clean
--- a/src/extra/tzone/Makefile.in	2017-07-08 22:52:11.301639963 +0200
+++ b/src/extra/tzone/Makefile.in	2017-07-08 22:53:59.221645277 +0200
@@ -24,7 +24,7 @@
 
 libtz.a: $(OBJECTS)
 	@rm -f $@
-	$(AR) -cr $@ $(OBJECTS)
+	$(AR) cr $@ $(OBJECTS)
 	@$(RANLIB) $@
 
 
--- a/src/extra/xdr/Makefile.in	2017-07-08 22:52:21.118640447 +0200
+++ b/src/extra/xdr/Makefile.in	2017-07-08 22:54:08.293645723 +0200
@@ -45,7 +45,7 @@
 
 libxdr.a: $(libxdr_a_OBJECTS)
 	@rm -f $@
-	$(AR) -cr $@ $(libxdr_a_OBJECTS)
+	$(AR) cr $@ $(libxdr_a_OBJECTS)
 	$(RANLIB) $@
 
 
--- a/src/main/Makefile.in	2017-07-08 22:52:32.293640997 +0200
+++ b/src/main/Makefile.in	2017-07-08 22:54:19.325646267 +0200
@@ -148,7 +148,7 @@
 	-@mkdir libs
 	@(cd libs; for l in $(STATIC_LIBS); do $(AR) -x ../$$l; done)
 	@rm -rf $@
-	$(AR) -cr $@ $(OBJECTS) libs/*o
+	$(AR) cr $@ $(OBJECTS) libs/*o
 	$(RANLIB) $@
 	@rm -rf libs
 
--- a/src/nmath/standalone/Makefile.in	2017-07-08 22:53:08.694642789 +0200
+++ b/src/nmath/standalone/Makefile.in	2017-07-08 22:54:35.157647046 +0200
@@ -123,7 +123,7 @@
 
 libRmath.a: $(libRmath_a_OBJECTS)
 	@rm -rf $@
-	$(AR) -cr $@ $(libRmath_a_OBJECTS)
+	$(AR) cr $@ $(libRmath_a_OBJECTS)
 	$(RANLIB) $@
 
 ## under peculiar circumstances, $(LIBM) here helps.
--- a/src/nmath/Makefile.in	2017-07-08 22:52:48.542641797 +0200
+++ b/src/nmath/Makefile.in	2017-07-08 22:54:28.077646697 +0200
@@ -73,7 +73,7 @@
 
 libnmath.a: $(OBJECTS)
 	@rm -rf $@
-	$(AR) -cr $@ $(OBJECTS)
+	$(AR) cr $@ $(OBJECTS)
 	$(RANLIB) $@
 
 mostlyclean: clean
--- a/src/unix/Makefile.in	2017-07-08 22:49:28.093631928 +0200
+++ b/src/unix/Makefile.in	2017-07-08 22:49:39.732632501 +0200
@@ -54,7 +54,7 @@
 
 libunix.a: $(OBJECTS)
 	@rm -rf $@
-	$(AR) -cr $@ $(OBJECTS)
+	$(AR) cr $@ $(OBJECTS)
 	$(RANLIB) $@
 
 mostlyclean: clean
--- a/src/appl/Makefile.in	2017-07-08 23:40:44.472783392 +0200
+++ b/src/appl/Makefile.in	2017-07-08 23:40:53.711783846 +0200
@@ -48,7 +48,7 @@
 
 libappl.a: $(OBJECTS)
 	@rm -f $@
-	$(AR) -cr $@ $(OBJECTS)
+	$(AR) cr $@ $(OBJECTS)
 	$(RANLIB) $@
 
 blas.o: $(top_srcdir)/src/extra/blas/blas.f
