--- libgc/os_dep.c.bak	2015-10-25 19:00:55.308241610 +0100
+++ libgc/os_dep.c	2015-10-25 19:01:10.982170990 +0100
@@ -41,7 +41,7 @@
 #     else /* not 2 <= __GLIBC__ */
         /* libc5 doesn't have <sigcontext.h>: go directly with the kernel   */
         /* one.  Check LINUX_VERSION_CODE to see which we should reference. */
-#       include <asm/sigcontext.h>
+//#       include <asm/sigcontext.h>
 #     endif /* 2 <= __GLIBC__ */
 #   endif
 # endif
