diff --git a/GhcMod/Target.hs b/GhcMod/Target.hs
index 02f447e..b2d944b 100644
--- a/GhcMod/Target.hs
+++ b/GhcMod/Target.hs
@@ -193,7 +193,7 @@ runGmlTWith' efnmns' mdf mUpdateHooks wrapper action = do
     cfns <- mapM getCanonicalFileNameSafe ccfns
     let serfnmn = Set.fromList $ map Right mns ++ map Left cfns
     (opts, mappedStrs) <- targetGhcOptions crdl serfnmn
-    let opts' = opts ++ ["-O0"] ++ optGhcUserOptions
+    let opts' = opts ++ ["-O0", "-fno-it"] ++ optGhcUserOptions
 
     gmVomit
       "session-ghc-options"
