--- a/drivers/platform/surface/surface_aggregator_registry.c	2026-01-12 19:56:39.361435704 +0100
+++ b/drivers/platform/surface/surface_aggregator_registry.c	2026-01-12 20:00:02.693765127 +0100
@@ -51,6 +51,12 @@
 	.parent = &ssam_node_root,
 };
 
+/* RTC */
+static const struct software_node ssam_node_rtc = {
+        .name = "ssam:01:01:01:00:00",
+        .parent = &ssam_node_root,
+};
+
 /* AC adapter. */
 static const struct software_node ssam_node_bat_ac = {
 	.name = "ssam:01:02:01:01:01",
@@ -408,6 +414,7 @@
 
 static const struct software_node *ssam_node_group_sp12[] = {
 	&ssam_node_root,
+	&ssam_node_rtc,
 	&ssam_node_hub_kip,
 	&ssam_node_tmp_sensors,
 	&ssam_node_hid_kip_keyboard,
