--- a/vendor/spdlog/spdlog/details/os.h	2022-12-17 14:20:20.256262930 +0100
+++ b/vendor/spdlog/spdlog/details/os.h	2022-12-17 14:25:39.266964147 +0100
@@ -247,7 +247,7 @@
 #else // unix
     int fd = fileno(f);
     //64 bits(but not in osx or cygwin, where fstat64 is deprecated)
-#if !defined(__FreeBSD__) && !defined(__APPLE__) && (defined(__x86_64__) || defined(__ppc64__)) && !defined(__CYGWIN__)
+#if defined(__GLIBC__) && (defined(__x86_64__) || defined(__ppc64__))
     struct stat64 st;
     if (fstat64(fd, &st) == 0)
         return static_cast<size_t>(st.st_size);
