--- a/src/Yi/Rope.hs	2018-04-16 00:40:55.812454382 +0200
+++ b/src/Yi/Rope.hs	2018-04-16 00:44:34.714057315 +0200
@@ -2,6 +2,7 @@
 {-# LANGUAGE DeriveDataTypeable #-}
 {-# LANGUAGE MultiParamTypeClasses #-}
 {-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE CPP #-}
 {-# OPTIONS_HADDOCK show-extensions #-}
 
 -- |
@@ -73,7 +74,9 @@
 import           Data.FingerTree hiding (null, empty, reverse, split)
 import qualified Data.List as L (foldl')
 import           Data.Maybe
-import           Data.Monoid
+#if __GLASGOW_HASKELL__ < 804
+import Data.Semigroup
+#endif
 import           Data.String (IsString(..))
 import qualified Data.Text as TX
 import qualified Data.Text.Encoding as TXE
@@ -116,9 +119,12 @@
 countNl :: TX.Text -> Int
 countNl = TX.count "\n"
 
+instance Semigroup Size where
+  Indices c l <> Indices c' l' = Indices (c + c') (l + l')
+
 instance Monoid Size where
   mempty = Indices 0 0
-  Indices c l `mappend` Indices c' l' = Indices (c + c') (l + l')
+  mappend = (<>)
 
 instance Measured Size YiChunk where
   measure (Chunk l t) = Indices l (countNl t)
@@ -154,9 +160,12 @@
 instance IsString YiString where
   fromString = Yi.Rope.fromString
 
+instance Semigroup YiString where
+  (<>) = Yi.Rope.append
+
 instance Monoid YiString where
   mempty = Yi.Rope.empty
-  mappend = Yi.Rope.append
+  mappend = (<>)
   mconcat = Yi.Rope.concat
 
 instance Ord YiString where
