--- a/lib/utils.c	2022-12-17 12:09:48.047449867 +0100
+++ b/lib/utils.c	2022-12-17 12:10:03.412568872 +0100
@@ -112,9 +112,9 @@
 	char tmp[BUFSIZ];
 	size_t next_read;
 	ssize_t bytes_r;
-	off64_t r;
+	off_t r;
 
-	r = lseek64(fd, bytes, SEEK_CUR);
+	r = lseek(fd, bytes, SEEK_CUR);
 	if (r > 0)
 		return 0;
 	if (r < 0 && errno != ESPIPE)
--- a/lib/utils_wipe.c	2022-12-17 12:10:57.394982953 +0100
+++ b/lib/utils_wipe.c	2022-12-17 12:11:16.154125459 +0100
@@ -180,7 +180,7 @@
 	if (r)
 		goto out;
 
-	if (lseek64(devfd, offset, SEEK_SET) < 0) {
+	if (lseek(devfd, offset, SEEK_SET) < 0) {
 		log_err(cd, _("Cannot seek to device offset."));
 		r = -EINVAL;
 		goto out;
