--- a/src/backends/meta-monitor-manager.c	2016-09-24 19:55:50.538824951 +0200
+++ b/src/backends/meta-monitor-manager.c	2016-09-24 19:57:36.038796219 +0200
@@ -590,7 +590,7 @@
 
   g_variant_builder_init (&crtc_builder, G_VARIANT_TYPE ("a(uxiiiiiuaua{sv})"));
   g_variant_builder_init (&output_builder, G_VARIANT_TYPE ("a(uxiausauaua{sv})"));
-  g_variant_builder_init (&mode_builder, G_VARIANT_TYPE ("a(uxuud)"));
+  g_variant_builder_init (&mode_builder, G_VARIANT_TYPE ("a(uxuudu)"));
 
   for (i = 0; i < manager->n_crtcs; i++)
     {
@@ -714,12 +714,13 @@
     {
       MetaMonitorMode *mode = &manager->modes[i];
 
-      g_variant_builder_add (&mode_builder, "(uxuud)",
+      g_variant_builder_add (&mode_builder, "(uxuudu)",
                              i, /* ID */
                              (gint64)mode->mode_id,
                              (guint32)mode->width,
                              (guint32)mode->height,
-                             (double)mode->refresh_rate);
+                             (double)mode->refresh_rate,
+                             (guint32) 0);
     }
 
   meta_dbus_display_config_complete_get_resources (skeleton,
--- a/src/org.gnome.Mutter.DisplayConfig.xml	2016-09-24 19:56:07.563821173 +0200
+++ b/src/org.gnome.Mutter.DisplayConfig.xml	2016-09-24 19:56:57.787807922 +0200
@@ -133,7 +133,7 @@
       <arg name="serial" direction="out" type="u" />
       <arg name="crtcs" direction="out" type="a(uxiiiiiuaua{sv})" />
       <arg name="outputs" direction="out" type="a(uxiausauaua{sv})" />
-      <arg name="modes" direction="out" type="a(uxuud)" />
+      <arg name="modes" direction="out" type="a(uxuudu)" />
       <arg name="max_screen_width" direction="out" type="i" />
       <arg name="max_screen_height" direction="out" type="i" />
     </method>
