--- a/solenv/gbuild/platform/unxgcc.mk	2016-10-23 14:12:08.231770426 +0200
+++ b/solenv/gbuild/platform/unxgcc.mk	2016-10-23 14:12:15.855770214 +0200
@@ -158,7 +158,7 @@
 define gb_LinkTarget__command_staticlink
 $(call gb_Helper_abbreviate_dirs,\
 	rm -f $(1) && \
-	$(gb_AR) $(gb_LTOPLUGINFLAGS) -rsu $(1) \
+	$(gb_AR) $(gb_LTOPLUGINFLAGS) rsu $(1) \
 		$(if $(LD_PLUGIN),--plugin $(LD_PLUGIN)) \
 		$(foreach object,$(COBJECTS),$(call gb_CObject_get_target,$(object))) \
 		$(foreach object,$(CXXOBJECTS),$(call gb_CxxObject_get_target,$(object))) \
