From: Dominik Kriegner <dominik.kriegner@gmail.com>
Date: Tue, 30 Nov 2016 20:30:00 +0000
Subject: [PATCH] use systems libconfig libary

diff -urN a/nvagent/src/CMakeLists.txt b/nvagent/src/CMakeLists.txt
--- a/nvagent/src/CMakeLists.txt	2016-11-30 19:11:24.949262776 +0100
+++ b/nvagent/src/CMakeLists.txt	2016-11-30 20:20:17.627577305 +0100
@@ -20,7 +20,6 @@
 
 message(STATUS "CMAKE_SOURCE_DIR" ${CMAKE_SOURCE_DIR})
 
-link_directories("${CMAKE_SOURCE_DIR}/libconfig/lib/.libs")
 link_directories("${CMAKE_SOURCE_DIR}/tapcfg/build")
 
 if (WITH_GUI)
@@ -103,12 +102,12 @@
 	target_link_libraries(netvirt-agent
 		-Wl,-Bstatic
 		tapcfg
-		config
 		nvcore
 		-Wl,-Bdynamic
 		ssl
 		crypto
 		pthread
+		config
 	)
 endif()
 
diff -urN a/nvctrler/src/CMakeLists.txt b/nvctrler/src/CMakeLists.txt
--- a/nvctrler/src/CMakeLists.txt	2016-11-30 19:11:24.951262484 +0100
+++ b/nvctrler/src/CMakeLists.txt	2016-11-30 20:21:53.859030787 +0100
@@ -21,9 +21,7 @@
 find_library(EVENT_SSL_LIBRARY event_openssl HINTS /usr/local/lib)
 
 include_directories("/usr/local/include")
-include_directories("${CMAKE_SOURCE_DIR}/libconfig/lib/")
 include_directories("${CMAKE_SOURCE_DIR}/libnvcore/src/protocol/")
-link_directories("${CMAKE_SOURCE_DIR}/libconfig/lib/.libs")
 target_link_libraries(netvirt-ctrler
 	nvcore
 	${EVENT_CORE_LIBRARY}
@@ -34,7 +32,7 @@
 	crypto
 	jansson
 	pthread
-	${CMAKE_SOURCE_DIR}/libconfig/lib/.libs/libconfig.a
+	config
 )
 
 install(TARGETS netvirt-ctrler
diff -urN a/nvswitch/src/CMakeLists.txt b/nvswitch/src/CMakeLists.txt
--- a/nvswitch/src/CMakeLists.txt	2016-11-30 19:11:24.952262338 +0100
+++ b/nvswitch/src/CMakeLists.txt	2016-11-30 20:21:30.212612668 +0100
@@ -22,7 +22,6 @@
 include_directories("${CMAKE_SOURCE_DIR}/libconfig/lib/")
 include_directories("${CMAKE_SOURCE_DIR}/libnvcore/src/")
 include_directories("${CMAKE_SOURCE_DIR}/libnvcore/src/protocol/")
-link_directories("${CMAKE_SOURCE_DIR}/libconfig/lib/.libs")
 target_link_libraries(netvirt-switch
 	nvcore
 	${EVENT_CORE_LIBRARY}
@@ -31,7 +30,7 @@
 	crypto
 	pthread
 	jansson
-	${CMAKE_SOURCE_DIR}/libconfig/lib/.libs/libconfig.a
+	config
 )
 
 install(TARGETS netvirt-switch
