--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1188,16 +1188,6 @@ set(MAX_EXTENDED_ALIGNMENT "${ALIGNOF_DOUBLE16}")
 ##########################################################
 
 
-string(TOUPPER "${CMAKE_BUILD_TYPE}" BTYPE)
-if("${CMAKE_C_FLAGS_${BTYPE}}" MATCHES "DNDEBUG")
-  set(POCL_ASSERTS_BUILD 0)
-else()
-  set(POCL_ASSERTS_BUILD 1)
-endif()
-
-##########################################################
-
-
 # cmake docs:
 # SOVERSION: What version number is this target.
 
