--- a/src/PantheonTerminalWindow.vala	2016-09-09 16:51:10.250620961 +0200
+++ b/src/PantheonTerminalWindow.vala	2016-09-09 16:51:23.024758620 +0200
@@ -467,14 +467,6 @@
                     PantheonTerminal.saved_state.window_state = PantheonTerminalWindowState.NORMAL;
                 }
         
-                /* Save window size */
-                if (PantheonTerminal.saved_state.window_state == PantheonTerminalWindowState.NORMAL) {
-                    int width, height;
-                    get_size (out width, out height);
-                    PantheonTerminal.saved_state.window_width = width;
-                    PantheonTerminal.saved_state.window_height = height;
-                }
-        
                 /* Save window position */
                 int root_x, root_y;
                 get_position (out root_x, out root_y);
