--- a/src/PantheonTerminalWindow.vala	2016-09-02 00:09:48.781702317 +0200
+++ b/src/PantheonTerminalWindow.vala	2016-09-02 00:09:57.629770172 +0200
@@ -107,7 +107,7 @@
             set_application (app);
 
 
-            Gtk.Settings.get_default ().gtk_application_prefer_dark_theme = true;
+            Gtk.Settings.get_default ().gtk_application_prefer_dark_theme = false;
 
             set_visual (Gdk.Screen.get_default ().get_rgba_visual ());
 
