--- Makefile.am	2016-02-05 18:43:17.237162560 +0100
+++ Makefile.am	2016-02-05 18:44:21.372632844 +0100
@@ -1,6 +1,6 @@
 ## Makefile.am - procress this file with automake to produce Makefile.in
 lib_LTLIBRARIES = libobstack.la
-libobstack_la_SOURCES = obstack.c
+libobstack_la_SOURCES = obstack.c obstack_printf.c
 libobstack_la_HEADERS = obstack.h
 libobstack_ladir = $(includedir)
 ACLOCAL_AMFLAGS = -Im4
--- obstack.h	2015-09-10 09:00:20.000000000 +0200
+++ obstack.h	2016-02-05 18:46:45.283160929 +0100
@@ -538,6 +538,8 @@
 
 #endif /* not __GNUC__ or not __STDC__ */
 
+int obstack_printf(struct obstack *obst, const char *fmt, ...);
+
 #ifdef __cplusplus
 }	/* C++ */
 #endif
--- obstack_printf.c	2016-02-05 18:45:15.874728684 +0100
+++ obstack_printf.c	2016-02-05 18:44:33.939116341 +0100
@@ -0,0 +1,36 @@
+/*
+ * This file is part of cparser.
+ * Copyright (C) 2007-2008 Matthias Braun <matze@braunis.de>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
+ * 02111-1307, USA.
+ */
+#include <stdarg.h>
+#include <stdio.h>
+#include "obstack.h"
+
+int obstack_printf(struct obstack *obst, const char *fmt, ...)
+{
+  char buf[1024];
+  va_list ap;
+  int len;
+
+  va_start(ap, fmt);
+  len = vsnprintf(buf, sizeof(buf), fmt, ap);
+  obstack_grow(obst, buf, len);
+  va_end(ap);
+
+  return len;
+}
