--- a/Source/WebCore/platform/gtk/UserAgentGtk.cpp	2017-01-12 17:11:23.857914681 +0100
+++ b/Source/WebCore/platform/gtk/UserAgentGtk.cpp	2017-01-12 17:12:32.545706482 +0100
@@ -201,6 +201,10 @@
     if (baseDomain == "yahoo.com")
         return true;
 
+    // Ooutlook displays a ugly text version
+    if (baseDomain == "outlook.com")
+        return true;
+
     // taobao.com displays a mobile version with our standard user agent.
     if (baseDomain == "taobao.com")
         return true;
