diff --git a/Source/WebCore/platform/UserAgentQuirks.cpp b/Source/WebCore/platform/UserAgentQuirks.cpp
index bdd82dab461449e2fe3da757e7813bbaf7cc8e35..a0506b1ee307e1f466c52fd4415a042c2faa2212 100644
--- a/Source/WebCore/platform/UserAgentQuirks.cpp
+++ b/Source/WebCore/platform/UserAgentQuirks.cpp
@@ -41,8 +41,9 @@ static bool isGoogle(const URL& url)
     // Our Google UA is *very* complicated to get right. Read
     // https://webkit.org/b/142074 carefully before changing. Test that Earth
     // view is available in Google Maps. Test Google Calendar. Test downloading
-    // the Hangouts browser plugin. Change platformVersionForUAString() to
-    // return "FreeBSD amd64" and test Maps and Calendar again.
+    // the Hangouts browser plugin. Test logging out and logging in to a Google
+    // account. Change platformVersionForUAString() to return "FreeBSD amd64"
+    // and test everything again.
     if (baseDomain.startsWith("google."))
         return true;
     if (baseDomain == "gstatic.com")
@@ -75,7 +76,7 @@ static bool urlRequiresChromeBrowser(const URL& url)
 
 static bool urlRequiresFirefoxBrowser(const URL& url)
 {
-    return isGoogle(url);
+    return isGoogle(url) && url.host() != "accounts.google.com";
 }
 
 static bool urlRequiresMacintoshPlatform(const URL& url)
