--- a/Source/WebCore/platform/linux/MemoryPressureHandlerLinux.cpp	2016-09-02 15:03:41.097828820 +0200
+++ b/Source/WebCore/platform/linux/MemoryPressureHandlerLinux.cpp	2016-09-02 15:03:54.270827122 +0200
@@ -33,6 +33,7 @@
 
 #include <fcntl.h>
 #include <malloc.h>
+#include <stdio.h>
 #include <sys/eventfd.h>
 #include <sys/stat.h>
 #include <sys/types.h>
--- a/Source/WebKit2/UIProcess/linux/MemoryPressureMonitor.cpp	2016-09-14 14:26:00.253663597 +0200
+++ b/Source/WebKit2/UIProcess/linux/MemoryPressureMonitor.cpp	2016-09-14 14:26:11.126642591 +0200
@@ -33,6 +33,7 @@
 #include <fcntl.h>
 #include <mutex>
 #include <stdlib.h>
+#include <stdio.h>
 #include <string.h>
 #include <sys/eventfd.h>
 #include <sys/stat.h>
