--- a/Source/JavaScriptCore/runtime/Options.h	2016-08-20 22:40:30.839528477 +0200
+++ b/Source/JavaScriptCore/runtime/Options.h	2016-08-20 22:40:49.913502382 +0200
@@ -113,8 +113,8 @@
     v(bool, reportMustSucceedExecutableAllocations, false, nullptr) \
     \
     v(unsigned, maxPerThreadStackUsage, 4 * MB, nullptr) \
-    v(unsigned, reservedZoneSize, 128 * KB, nullptr) \
-    v(unsigned, errorModeReservedZoneSize, 64 * KB, nullptr) \
+    v(unsigned, reservedZoneSize, 32 * KB, nullptr) \
+    v(unsigned, errorModeReservedZoneSize, 16 * KB, nullptr) \
     \
     v(bool, crashIfCantAllocateJITMemory, false, nullptr) \
     v(unsigned, jitMemoryReservationSize, 0, "Set this number to change the executable allocation size in ExecutableAllocatorFixedVMPool. (In bytes.)") \
