--- udisks/udisksclient.c.bak	2016-02-04 23:01:34.797392924 +0100
+++ udisks/udisksclient.c	2016-02-04 23:03:12.771143607 +0100
@@ -2001,10 +2001,8 @@
             {
               /* we know better than the compiler here */
 #if defined( __GNUC_PREREQ) || defined(__clang__)
-# if __GNUC_PREREQ(4,6) || __clang__
 #  pragma GCC diagnostic push
 #  pragma GCC diagnostic ignored "-Wformat-nonliteral"
-# endif
 #endif
               if (long_string)
                 ret = g_strdup_printf (g_dpgettext2 (GETTEXT_PACKAGE, "fs-type", id_type[n].long_name), version);
@@ -2012,9 +2010,7 @@
                 ret = g_strdup_printf (g_dpgettext2 (GETTEXT_PACKAGE, "fs-type", id_type[n].short_name), version);
               goto out;
 #if defined( __GNUC_PREREQ) || defined(__clang__)
-# if __GNUC_PREREQ(4,6) || __clang__
 #  pragma GCC diagnostic pop
-# endif
 #endif
             }
         }
