--- a/GNUmakefile	2023-10-01 18:38:21.000000000 +0200
+++ b/GNUmakefile	2025-01-31 21:30:06.664372772 +0100
@@ -991,14 +991,6 @@
   endif # X86/X32/X64
 endif # SunOS
 
-ifneq ($(IS_LINUX)$(IS_HURD),00)
-  ifeq ($(findstring -fopenmp,$(CXXFLAGS)),-fopenmp)
-    ifeq ($(findstring -lgomp,$(LDLIBS)),)
-      LDLIBS += -lgomp
-    endif # LDLIBS
-  endif # OpenMP
-endif # IS_LINUX or IS_HURD
-
 # Add -errtags=yes to get the name for a warning suppression
 ifneq ($(SUN_COMPILER),0)	# override flags for CC Sun C++ compiler
 # Add to all Solaris
