--- a/GNUmakefile	2018-05-02 10:46:52.511157210 +0200
+++ b/GNUmakefile	2018-05-02 10:47:49.581052880 +0200
@@ -24,7 +24,6 @@
 MKDIR ?= mkdir -p
 
 LN ?= ln -sf
-LDCONF ?= /sbin/ldconfig -n
 
 INSTALL = install
 INSTALL_PROGRAM = $(INSTALL)
@@ -872,8 +871,8 @@
 	@-$(MKDIR) $(DESTDIR)$(LIBDIR)
 	$(INSTALL_PROGRAM) libcryptopp.so$(SOLIB_VERSION_SUFFIX) $(DESTDIR)$(LIBDIR)
 ifeq ($(HAS_SOLIB_VERSION),1)
-	-$(LN) libcryptopp.so$(SOLIB_VERSION_SUFFIX) $(DESTDIR)$(LIBDIR)/libcryptopp.so
-	$(LDCONF) $(DESTDIR)$(LIBDIR)
+	-$(LN) libcryptopp.so$(SOLIB_VERSION_SUFFIX) $(DESTDIR)$(LIBDIR)/libcryptopp.so$(SOLIB_COMPAT_SUFFIX)
+	-$(LN) libcryptopp.so$(SOLIB_COMPAT_SUFFIX) $(DESTDIR)$(LIBDIR)/libcryptopp.so
 endif
 endif
 ifneq ($(wildcard libcryptopp.pc),)
