diff --git a/Sources/Configs/Config+Directory.swift b/Sources/Configs/Config+Directory.swift
index 52e0aefa..ae33ed0f 100644
--- a/Sources/Configs/Config+Directory.swift
+++ b/Sources/Configs/Config+Directory.swift
@@ -47,11 +47,7 @@ extension FileManager {
     fileprivate func isDirectory(path: String) -> Bool {
         var isDirectory: ObjCBool = false
         _ = fileExists(atPath: path, isDirectory: &isDirectory)
-        #if os(Linux)
-            return isDirectory
-        #else
-            return isDirectory.boolValue
-        #endif
+        return isDirectory.boolValue
     }
 
     fileprivate func files(path: String) throws -> [String] {
