diff --git a/Sources/Vapor/File/FileManager.swift b/Sources/Vapor/File/FileManager.swift
index b1ea7074..a36338b1 100644
--- a/Sources/Vapor/File/FileManager.swift
+++ b/Sources/Vapor/File/FileManager.swift
@@ -41,7 +41,7 @@ class FileManager {
         let path = try self.expandPath(path).finished(with: "/")
         let pattern = strdup(path + "{*,.*}")
 
-        switch glob(pattern, GLOB_MARK | GLOB_NOSORT | GLOB_BRACE, nil, &gt) {
+        switch glob(pattern, GLOB_MARK | GLOB_NOSORT, nil, &gt) {
         case GLOB_NOMATCH:
             return [ ]
         case GLOB_ABORTED:
