--- a/Package@swift-4.swift	2017-09-22 11:09:34.475903839 +0200
+++ b/Package@swift-4.swift	2017-09-22 11:09:58.720903277 +0200
@@ -4,11 +4,11 @@
 let package = Package(
     name: "Vapor",
     products: [
-        .library(name: "Cache", targets: ["Cache"]),
-        .library(name: "Configs", targets: ["Configs"]),
-        .library(name: "Sessions", targets: ["Sessions"]),
-        .library(name: "Testing", targets: ["Testing"]),
-        .library(name: "Vapor", targets: ["Vapor"]),
+        .library(name: "Cache", type: .dynamic, targets: ["Cache"]),
+        .library(name: "Configs", type: .dynamic, targets: ["Configs"]),
+        .library(name: "Sessions", type: .dynamic, targets: ["Sessions"]),
+        .library(name: "Testing", type: .dynamic, targets: ["Testing"]),
+        .library(name: "Vapor", type: .dynamic, targets: ["Vapor"]),
     ],
     targets: [
         .target(name: "Cache"),
