diff --git a/Package.swift b/Package.swift
index 71c29fd..b0d5a96 100644
--- a/Package.swift
+++ b/Package.swift
@@ -12,4 +12,11 @@ var package = Package(
             name: "TurnstileWeb",
             dependencies: [.Target(name: "Turnstile")]),
         ]
-)
\ No newline at end of file
+)
+
+let tlib = Product(name: "Turnstile", type: .Library(.Dynamic), modules: "Turnstile")
+products.append(tlib)
+let tclib = Product(name: "TurnstileCrypto", type: .Library(.Dynamic), modules: "TurnstileCrypto")
+products.append(tclib)
+let twlib = Product(name: "TurnstileWeb", type: .Library(.Dynamic), modules: "TurnstileWeb")
+products.append(twlib)
