--- a/include/string.h
+++ b/include/string.h
@@ -93,6 +93,7 @@
 char *strcasestr(const char *, const char *);
 void *memmem(const void *, size_t, const void *, size_t);
 void *memrchr(const void *, int, size_t);
+void *rawmemchr(const void *, int);
 void *mempcpy(void *, const void *, size_t);
 #ifndef __cplusplus
 char *basename();
--- /dev/null
+++ b/src/string/rawmemchr.c
@@ -0,0 +1,11 @@
+#define _GNU_SOURCE
+#include <string.h>
+#include "libc.h"
+
+void *__rawmemchr(const void *s, int c)
+{
+        register const unsigned char *r = s;
+        while (*r != ((unsigned char)c)) ++r;
+        return (void *) r;
+}
+weak_alias(__rawmemchr, rawmemchr);
