--- /dev/null
+++ b/src/network/__cmsg_nxthdr.c
@@ -0,0 +1,8 @@
+#include <sys/socket.h>
+
+struct cmsghdr* __cmsg_nxthdr(struct msghdr* mhdr, struct cmsghdr* cmsg)
+{
+	struct cmsghdr * __ptr = (struct cmsghdr *)((unsigned char *)cmsg + CMSG_ALIGN(cmsg->cmsg_len));
+
+	return ((size_t)cmsg->cmsg_len < sizeof (struct cmsghdr) || (unsigned char *)(__ptr + 1) > (unsigned char *)mhdr->msg_control + mhdr->msg_controllen) ? 0 : __ptr;
+}
--- a/include/sys/socket.h
+++ b/include/sys/socket.h
@@ -284,9 +284,7 @@
 #define __MHDR_END(mhdr) ((unsigned char *)(mhdr)->msg_control + (mhdr)->msg_controllen)
 
 #define CMSG_DATA(cmsg) ((unsigned char *) (((struct cmsghdr *)(cmsg)) + 1))
-#define CMSG_NXTHDR(mhdr, cmsg) ((cmsg)->cmsg_len < sizeof (struct cmsghdr) || \
-	__CMSG_LEN(cmsg) + sizeof(struct cmsghdr) >= __MHDR_END(mhdr) - (unsigned char *)(cmsg) \
-	? 0 : (struct cmsghdr *)__CMSG_NEXT(cmsg))
+#define CMSG_NXTHDR(mhdr, cmsg) __cmsg_nxthdr(mhdr, cmsg)
 #define CMSG_FIRSTHDR(mhdr) ((size_t) (mhdr)->msg_controllen >= sizeof (struct cmsghdr) ? (struct cmsghdr *) (mhdr)->msg_control : (struct cmsghdr *) 0)
 
 #define CMSG_ALIGN(len) (((len) + sizeof (size_t) - 1) & (size_t) ~(sizeof (size_t) - 1))
@@ -307,6 +305,9 @@
 	unsigned long __ss_align;
 };
 
+struct cmsghdr *__cmsg_nxthdr (struct msghdr *, struct cmsghdr *);
+
+
 int socket (int, int, int);
 int socketpair (int, int, int, int [2]);
 
