--- /dev/null
+++ b/src/math/isinf.c
@@ -0,0 +1,19 @@
+#include <math.h>
+#include "libc.h"
+
+int __isinff(float arg)
+{
+    return isinf(arg);
+}
+
+int __isinf(double x)
+{
+	return isinf(x);
+}
+weak_alias(__isinf, isinf);
+
+int __isinfl(long double x)
+{
+	return isinf(x);
+}
+weak_alias(__isinfl, isinfl);
--- /dev/null
+++ b/src/math/isnan.c
@@ -0,0 +1,20 @@
+#include <math.h>
+#include "libc.h"
+
+int __isnanf(float arg)
+{
+    return isnan(arg);
+}
+weak_alias(__isnanf, isnanf);
+
+int __isnan(double x)
+{
+	return isnan(x);
+}
+weak_alias(__isnan, isnan);
+
+int __isnanl(long double x)
+{
+	return isnan(x);
+}
+weak_alias(__isnanl, isnanl);
--- b/src/math/finite.c	2015-06-04 22:08:24.000000000 -0200
+++ b/src/math/finite.c	2015-08-23 12:40:08.000014438 -0200
@@ -1,7 +1,15 @@
 #define _GNU_SOURCE
 #include <math.h>
+#include "libc.h"
 
-int finite(double x)
+int __finite(double x)
 {
 	return isfinite(x);
 }
+weak_alias(__finite, finite);
+
+int __finitel(long double x)
+{
+	return isfinite(x);
+}
+weak_alias(__finitel, finitel);
--- a/src/math/finitef.c	2016-02-19 20:07:43.992445286 +0100
+++ b/src/math/finitef.c	2016-02-19 20:08:27.888441674 +0100
@@ -1,7 +1,9 @@
 #define _GNU_SOURCE
 #include <math.h>
+#include "libc.h"
 
-int finitef(float x)
+int __finitef(float x)
 {
 	return isfinite(x);
 }
+weak_alias(__finitef, finitef);
