--- src/math/finitef.c	2016-01-26 11:14:43.990349718 +0100
+++ src/math/finitef.c	2016-01-26 11:14:25.302503400 +0100
@@ -1,7 +1,9 @@
 #define _GNU_SOURCE
 #include <math.h>
+#include "libc.h"
 
-int finitef(float x)
+int __finitef(float x)
 {
 	return isfinite(x);
 }
+weak_alias(__finitef, finitef);
--- src/math/nan.c	2016-01-26 11:14:44.001350214 +0100
+++ src/math/nan.c	2016-01-26 11:14:35.063946748 +0100
@@ -4,3 +4,8 @@
 {
 	return NAN;
 }
+
+int __isnanl(long double arg)
+{
+	return isnan(arg);
+}
--- src/string/strtok_r.c	2016-01-31 19:24:57.742649710 +0100
+++ src/string/strtok_r.c.new	2016-01-31 19:25:52.428469125 +0100
@@ -1,6 +1,7 @@
 #include <string.h>
+#include "libc.h"
 
-char *strtok_r(char *restrict s, const char *restrict sep, char **restrict p)
+char *__strtok_r(char *restrict s, const char *restrict sep, char **restrict p)
 {
 	if (!s && !(s = *p)) return NULL;
 	s += strspn(s, sep);
@@ -10,3 +11,4 @@
 	else *p = 0;
 	return s;
 }
+weak_alias(__strtok_r, strtok_r);
