--- a/src/misc/getopt_long.c	2016-10-14 14:02:10.383503009 +0200
+++ b/src/misc/getopt_long.c	2016-10-14 14:04:47.908490047 +0200
@@ -53,7 +53,7 @@
 {
 	optarg = 0;
 	if (longopts && argv[optind][0] == '-' &&
-		((longonly && argv[optind][1]) ||
+		((longonly && argv[optind][1] && argv[optind][1] != '-') ||
 		 (argv[optind][1] == '-' && argv[optind][2])))
 	{
 		int colon = optstring[optstring[0]=='+'||optstring[0]=='-']==':';
