--- a/src/locale/dcngettext.c	2017-03-04 08:05:45.755115338 +0000
+++ b/src/locale/dcngettext.c	2017-03-04 08:05:45.755115338 +0000
@@ -121,6 +121,7 @@
 char *dcngettext(const char *domainname, const char *msgid1, const char *msgid2, unsigned long int n, int category)
 {
+	if (!msgid1) goto notrans;
 	static struct msgcat *volatile cats;
 	struct msgcat *p;
 	struct __locale_struct *loc = CURRENT_LOCALE;
 	const struct __locale_map *lm;
