--- a/include/net/if_arp.h	2015-08-21 09:52:48.740508005 +0000
+++ b/include/net/if_arp.h	2015-08-20 12:01:10.000000000 +0000
@@ -19,6 +19,8 @@
 #define	ARPOP_InREPLY	9
 #define	ARPOP_NAK	10
 
+#if !defined(HAVE_ARP_H)
+#define HAVE_ARP_H
 struct arphdr {
 	uint16_t ar_hrd;
 	uint16_t ar_pro;
@@ -109,6 +111,7 @@
 	int arp_flags;
 	struct sockaddr arp_netmask;
 };
+#endif
 
 #define ATF_COM		0x02
 #define	ATF_PERM	0x04
--- a/include/netinet/if_ether.h	2015-08-21 09:52:48.744507992 +0000
+++ b/include/netinet/if_ether.h	2015-08-20 12:01:11.000000000 +0000
@@ -93,11 +93,14 @@
 #define ETH_P_IEEE802154 0x00F6
 #define ETH_P_CAIF	0x00F7
 
+#if !defined(HAVE_IF_ETHER)
+#define HAVE_IF_ETHER
 struct ethhdr {
 	uint8_t h_dest[ETH_ALEN];
 	uint8_t h_source[ETH_ALEN];
 	uint16_t h_proto;
 };
+#endif
 
 #include <net/ethernet.h>
 #include <net/if_arp.h>
