--- a/include/net/if_arp.h	2015-08-21 09:52:48.740508005 +0000
+++ b/include/net/if_arp.h	2015-08-20 12:01:10.000000000 +0000
@@ -19,6 +19,8 @@
 #define	ARPOP_InREPLY	9
 #define	ARPOP_NAK	10
 
+#if !defined(HAVE_ARP_H)
+#define HAVE_ARP_H
 struct arphdr {
 	uint16_t ar_hrd;
 	uint16_t ar_pro;
@@ -109,6 +111,7 @@
 	int arp_flags;
 	struct sockaddr arp_netmask;
 };
+#endif
 
 #define ATF_COM		0x02
 #define	ATF_PERM	0x04
