--- include/malloc.h.bak	2015-09-07 19:58:26.289156324 +0200
+++ include/malloc.h	2015-09-07 19:59:05.631158261 +0200
@@ -18,6 +18,21 @@
 
 size_t malloc_usable_size(void *);
 
+struct mallinfo mallinfo (void);
+
+struct mallinfo {
+    int arena;     /* Non-mmapped space allocated (bytes) */
+    int ordblks;   /* Number of free chunks */
+    int smblks;    /* Number of free fastbin blocks */
+    int hblks;     /* Number of mmapped regions */
+    int hblkhd;    /* Space allocated in mmapped regions (bytes) */
+    int usmblks;   /* Maximum total allocated space (bytes) */
+    int fsmblks;   /* Space in freed fastbin blocks (bytes) */
+    int uordblks;  /* Total allocated space (bytes) */
+    int fordblks;  /* Total free space (bytes) */
+    int keepcost;  /* Top-most, releasable space (bytes) */
+};
+
 #ifdef __cplusplus
 }
 #endif
--- /dev/null	2015-09-07 09:56:29.856378154 +0200
+++ src/malloc/mallinfo.c	2015-09-07 20:00:06.399161253 +0200
@@ -0,0 +1,7 @@
+#include <malloc.h>
+
+struct mallinfo mallinfo(void)
+{
+    struct mallinfo st;
+    return st;
+}
