--- a/include/ixion/model_context.hpp	2020-02-19 21:40:12.692890309 +0100
+++ b/include/ixion/model_context.hpp	2020-02-19 21:39:39.185730245 +0100
@@ -61,7 +61,7 @@
         } value;
 
         /** Initializes the cell to be empty. */
-        input_cell(nullptr_t);
+        input_cell(std::nullptr_t);
         /** Boolean cell value. */
         input_cell(bool b);
         /** The char array must be null-terminated. */
