--- a/src/widgets/widgets/qmenu.cpp.bak	2016-04-03 19:19:45.721551358 +0200
+++ b/src/widgets/widgets/qmenu.cpp	2016-04-03 19:19:56.921550437 +0200
@@ -3234,7 +3234,7 @@
     item->setShortcut(action->shortcut());
     item->setCheckable(action->isCheckable());
     item->setChecked(action->isChecked());
-    item->setHasExclusiveGroup(action->actionGroup() && action->actionGroup()->isExclusive());
+//    item->setHasExclusiveGroup(action->actionGroup() && action->actionGroup()->isExclusive());
     item->setFont(action->font());
     item->setRole((QPlatformMenuItem::MenuRole) action->menuRole());
     item->setEnabled(action->isEnabled());
