--- a/libpolyml/xwindows.cpp	2018-07-08 17:38:27.887474958 +0200
+++ b/libpolyml/xwindows.cpp	2018-07-08 17:39:04.110373694 +0200
@@ -7303,7 +7303,7 @@
                              : printf("Old Display <%p>",P.AsAddress()));
             return;
             
-        default:         printf("Pointer "ZERO_X"%p",P.AsAddress());
+        default:         printf("Pointer " ZERO_X "%p",P.AsAddress());
             return;
         }
     }
