--- a/Makefile.am	2018-07-08 17:38:26.151479905 +0200
+++ b/Makefile.am	2018-07-08 17:39:44.173265792 +0200
@@ -75,9 +75,7 @@
 OBJECTSUFFIX=o
 endif
 
-if EXPMACHO
-poly_LDFLAGS += -Wl,-no_pie
-endif
+poly_LDFLAGS += -nopie
 
 poly_SOURCES = 
 poly_LDADD = $(POLYOBJECTFILE) $(POLYRESOURCES) libpolymain/libpolymain.la libpolyml/libpolyml.la 
--- a/polyc.in	2018-07-21 10:56:49.578928162 +0200
+++ b/polyc.in	2018-07-21 10:57:43.504993672 +0200
@@ -29,7 +29,7 @@
 @NATIVE_WINDOWS_FALSE@TEMPORARYDIR=${TMPDIR:-/tmp}
 
 # Extra options for Mac OS X
-@EXPMACHO_TRUE@EXTRALDFLAGS="-Wl,-no_pie"
+EXTRALDFLAGS="-nopie"
 
 TMPOBJFILE=${TEMPORARYDIR}/polyobj.$$.$SUFFIX
 trap 'rm -f "$TMPOBJFILE"' 0
