diff --git a/lib/target.py b/lib/target.py
index bdeecdf..8bcf72b 100644
--- a/lib/target.py
+++ b/lib/target.py
@@ -59,7 +59,7 @@ class ArchType:
     @staticmethod
     def to_string(value):
         if value == ArchType.armv7:
-            return "armv7"
+            return "arm"
         if value == ArchType.armv6:
             return "armv6"
         if value == ArchType.armeb:
@@ -150,6 +150,8 @@ class ArchType:
             return ArchType.armv7
         if string == "armv7":
             return ArchType.armv7
+        if string == "armv7a":
+            return ArchType.armv7
         if string == "armv7l":
             return ArchType.armv7
         if string == "armv6":
