diff --git a/lib/target.py b/lib/target.py
index bdeecdf..606f9f9 100644
--- a/lib/target.py
+++ b/lib/target.py
@@ -53,6 +53,7 @@ class ArchType:
     shave       = 39
     armv6       = 40
     s390x       = 41
+    thumbv7     = 42
 # Do not assume that these are 1:1 mapping.  This should follow
 # canonical naming conventions for arm, etc. architectures.
 # See apple/swift PR #608    
@@ -106,6 +107,8 @@ class ArchType:
             return "tce"
         if value == ArchType.thumb:
             return "armv7"
+        if value == ArchType.thumbv7:
+            return "thumbv7"
         if value == ArchType.thumbeb:
             return "thumbeb"
         if value == ArchType.x86:
@@ -198,6 +201,8 @@ class ArchType:
             return ArchType.tce
         if string == "thumb":
             return ArchType.thumb
+        if string == "thumbv7":
+            return ArchType.thumbv7
         if string == "thumbeb":
             return ArchType.thumbeb
         if string == "x86":
