From abd7ee6dce8fe692897ac9201301c7c2038cecf6 Mon Sep 17 00:00:00 2001
From: Mike Frysinger <vapier@gentoo.org>
Date: Mon, 14 Nov 2016 09:57:01 -0500
Subject: [PATCH] paxldso: fix building when ldso support is disabled

---
 paxldso.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/paxldso.c b/paxldso.c
index 2227387..93eaf97 100644
--- a/paxldso.c
+++ b/paxldso.c
@@ -362,8 +362,10 @@ static void ldso_config_cleanup(void)
 # define ldso_config_cleanup()
 #endif /* PAX_LDSO_CONFIG */
 
+#ifndef paxldso_cleanup
 void paxldso_cleanup(void)
 {
 	ldso_cache_cleanup();
 	ldso_config_cleanup();
 }
+#endif
