From 061efe12b5e1f15341b30088588513e1855eab77 Mon Sep 17 00:00:00 2001
From: Mike Frysinger <vapier@gentoo.org>
Date: Mon, 14 Nov 2016 23:02:34 -0500
Subject: [PATCH] paxldso: fix unused arg warnings when support is disabled

---
 paxldso.h | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/paxldso.h b/paxldso.h
index 1517e5e..16cbbac 100644
--- a/paxldso.h
+++ b/paxldso.h
@@ -26,7 +26,7 @@
 #if PAX_LDSO_CACHE
 extern char *ldso_cache_lookup_lib(elfobj *elf, const char *fname);
 #else
-static inline char *ldso_cache_lookup_lib(elfobj *elf, const char *fname)
+static inline char *ldso_cache_lookup_lib(__unused__ elfobj *elf, __unused__ const char *fname)
 {
 	return NULL;
 }
@@ -54,7 +54,7 @@ extern array_t *ldpaths;
 #if PAX_LDSO_CONFIG
 extern int ldso_config_load(const char *fname);
 #else
-static inline int ldso_config_load(const char *fname)
+static inline int ldso_config_load(__unused__ const char *fname)
 {
 	return 0;
 }
