--- a/Makefile.in	2025-04-09 18:31:28.470566122 +0200
+++ b/Makefile.in	2025-04-09 18:32:42.709789587 +0200
@@ -7600,7 +7600,7 @@
 src_kill_LDADD = $(LDADD) $(LIBTHREAD)
 src_link_LDADD = $(LDADD)
 src_ln_LDADD = $(LDADD) $(CLOCK_TIME_LIB)
-src_logname_LDADD = $(LDADD)
+src_logname_LDADD = $(LDADD) $(READUTMP_LIB)
 
 # for cap_get_file
 src_ls_LDADD = $(LDADD) $(LIB_SELINUX) $(LIB_SMACK) $(CLOCK_TIME_LIB) \
