diff --git a/Package@swift-4.swift b/Package@swift-4.swift
index 4c16a6e..55e5c66 100644
--- a/Package@swift-4.swift
+++ b/Package@swift-4.swift
@@ -6,13 +6,8 @@ let package = Package(
     products: [
         .library(name: "TLS", targets: ["TLS"]),
     ],
-    dependencies: [
-        .package(url: "https://github.com/vapor/core.git", .upToNextMajor(from: "2.1.1")),
-      	.package(url: "https://github.com/vapor/sockets.git", .upToNextMajor(from: "2.1.0")),
-        .package(url: "https://github.com/vapor/ctls.git", .upToNextMajor(from: "1.1.0")),
-    ],
     targets: [
-        .target(name: "TLS", dependencies: ["Core", "CTLS", "Sockets"]),
+        .target(name: "TLS"),
         .testTarget(name: "TLSTests", dependencies: ["TLS"]),
     ]
 )
