--- a/shell/shell.cpp
+++ b/shell/shell.cpp
@@ -211,7 +211,8 @@
     KParts::ReadWritePart* const part = m_tabs[0].part;
 
     // Return false if we can't open new tabs and the only part is occupied
-    if ( !dynamic_cast<Okular::ViewerInterface*>(part)->openNewFilesInTabs()
+    if ( dynamic_cast<Okular::ViewerInterface*>(part)
+         && !dynamic_cast<Okular::ViewerInterface*>(part)->openNewFilesInTabs()
          && !part->url().isEmpty()
          && !ShellUtils::unique(serializedOptions))
     {
