--- a/src/framework/mlt_property.h
+++ b/src/framework/mlt_property.h
@@ -30,13 +30,7 @@
 #include <sys/param.h>
 #endif
 
-#if (defined(__GLIBC__) && !defined(__APPLE__)) || defined(HAVE_LOCALE_H)
-#  include <locale.h>
-#elif defined(__APPLE__) || (defined(__FreeBSD_version) && __FreeBSD_version >= 900506)
-#  include <xlocale.h>
-#else
-typedef char* locale_t;
-#endif
+#include <locale.h>
 
 extern mlt_property mlt_property_init( );
 extern void mlt_property_clear( mlt_property self );
