--- a/Source/Core/Core/HW/EXI/BBA/TAPServerConnection.cpp	2026-03-12 00:36:49.000000000 +0100
+++ b/Source/Core/Core/HW/EXI/BBA/TAPServerConnection.cpp	2026-05-16 14:14:37.659343199 +0200
@@ -13,6 +13,7 @@
 #include <sys/socket.h>
 #include <sys/un.h>
 #include <unistd.h>
+#include <sys/time.h>
 #endif
 
 #include "Common/Logging/Log.h"
--- a/Source/Core/VideoCommon/Assets/CustomTextureData.h	2026-06-25 00:36:03.000000000 +0200
+++ b/Source/Core/VideoCommon/Assets/CustomTextureData.h	2026-08-09 18:11:11.933972588 +0200
@@ -5,6 +5,7 @@
 
 #include <span>
 #include <string>
+#include <vector>
 
 #include "Common/Buffer.h"
 #include "Common/CommonTypes.h"
