--- src/mtpFilesystemErrors.h.bak	2017-10-03 13:33:51.649570827 +0200
+++ src/mtpFilesystemErrors.h	2017-10-03 13:34:04.679570525 +0200
@@ -22,6 +22,8 @@
 #define MTPFILESYSTEMERRORS_H_
 
 #include <stdexcept>
+#include <cerrno>
+#include <string>
 
 // There appears to be a bug in Android 4.0.4 on the Galaxy Nexus (and maybe
 // other Android versions as well) where if you specify a filename greater
