diff --git a/src/client/qwaylandshellsurface_p.h b/src/client/qwaylandshellsurface_p.h
index 6499a2bb..0cb55442 100644
--- a/src/client/qwaylandshellsurface_p.h
+++ b/src/client/qwaylandshellsurface_p.h
@@ -82,7 +82,7 @@ public:
     QPlatformWindow *platformWindow();
     struct wl_surface *wlSurface();
 
-    virtual std::any surfaceRole() const { return std::any(); };
+    virtual void *surfaceRole() const { return nullptr; };
 
     virtual void attachPopup(QWaylandShellSurface *popup) { Q_UNUSED(popup); }
     virtual void detachPopup(QWaylandShellSurface *popup) { Q_UNUSED(popup); }
diff --git a/src/plugins/shellintegration/fullscreen-shell-v1/qwaylandfullscreenshellv1surface.h b/src/plugins/shellintegration/fullscreen-shell-v1/qwaylandfullscreenshellv1surface.h
index 0a82e5ee..4c55c077 100644
--- a/src/plugins/shellintegration/fullscreen-shell-v1/qwaylandfullscreenshellv1surface.h
+++ b/src/plugins/shellintegration/fullscreen-shell-v1/qwaylandfullscreenshellv1surface.h
@@ -18,7 +18,7 @@ class Q_WAYLANDCLIENT_EXPORT QWaylandFullScreenShellV1Surface : public QWaylandS
 {
 public:
     QWaylandFullScreenShellV1Surface(QtWayland::zwp_fullscreen_shell_v1 *shell, QWaylandWindow *window);
-    std::any surfaceRole() const override { return m_shell->object(); }
+    void *surfaceRole() const override { return m_shell->object(); }
 
 private:
     QtWayland::zwp_fullscreen_shell_v1 *m_shell = nullptr;
diff --git a/src/plugins/shellintegration/ivi-shell/qwaylandivisurface_p.h b/src/plugins/shellintegration/ivi-shell/qwaylandivisurface_p.h
index fc97a835..f4bf94d7 100644
--- a/src/plugins/shellintegration/ivi-shell/qwaylandivisurface_p.h
+++ b/src/plugins/shellintegration/ivi-shell/qwaylandivisurface_p.h
@@ -27,7 +27,7 @@ public:
 
     void applyConfigure() override;
 
-    std::any surfaceRole() const override { return ivi_surface::object(); };
+    void *surfaceRole() const override { return (void*)ivi_surface::object(); };
 
 private:
     void createExtendedSurface(QWaylandWindow *window);
diff --git a/src/plugins/shellintegration/qt-shell/qwaylandqtsurface_p.h b/src/plugins/shellintegration/qt-shell/qwaylandqtsurface_p.h
index 6431bd55..bed255d7 100644
--- a/src/plugins/shellintegration/qt-shell/qwaylandqtsurface_p.h
+++ b/src/plugins/shellintegration/qt-shell/qwaylandqtsurface_p.h
@@ -41,7 +41,7 @@ public:
     void raise() override;
     void lower() override;
 
-    std::any surfaceRole() const override { return object(); };
+    void *surfaceRole() const override { return (void*)object(); };
 
 private:
     void resetConfiguration();
diff --git a/src/plugins/shellintegration/wl-shell/qwaylandwlshellsurface_p.h b/src/plugins/shellintegration/wl-shell/qwaylandwlshellsurface_p.h
index 24600302..52823c15 100644
--- a/src/plugins/shellintegration/wl-shell/qwaylandwlshellsurface_p.h
+++ b/src/plugins/shellintegration/wl-shell/qwaylandwlshellsurface_p.h
@@ -57,7 +57,7 @@ public:
     void applyConfigure() override;
     bool wantsDecorations() const override;
 
-    std::any surfaceRole() const override { return object(); };
+    void *surfaceRole() const override { return (void*)object(); };
 
 protected:
     void requestWindowStates(Qt::WindowStates states) override;
diff --git a/src/plugins/shellintegration/xdg-shell/qwaylandxdgshell.cpp b/src/plugins/shellintegration/xdg-shell/qwaylandxdgshell.cpp
index 2ff84e63..7ab0f5ed 100644
--- a/src/plugins/shellintegration/xdg-shell/qwaylandxdgshell.cpp
+++ b/src/plugins/shellintegration/xdg-shell/qwaylandxdgshell.cpp
@@ -442,13 +442,13 @@ void *QWaylandXdgSurface::nativeResource(const QByteArray &resource)
     return nullptr;
 }
 
-std::any QWaylandXdgSurface::surfaceRole() const
+void *QWaylandXdgSurface::surfaceRole() const
 {
     if (m_toplevel)
         return m_toplevel->object();
     if (m_popup)
         return m_popup->object();
-    return {};
+    return nullptr;
 }
 
 void QWaylandXdgSurface::requestWindowStates(Qt::WindowStates states)
diff --git a/src/plugins/shellintegration/xdg-shell/qwaylandxdgshell_p.h b/src/plugins/shellintegration/xdg-shell/qwaylandxdgshell_p.h
index e2dc12dd..46e83b3e 100644
--- a/src/plugins/shellintegration/xdg-shell/qwaylandxdgshell_p.h
+++ b/src/plugins/shellintegration/xdg-shell/qwaylandxdgshell_p.h
@@ -72,7 +72,7 @@ public:
 
     void *nativeResource(const QByteArray &resource);
 
-    std::any surfaceRole() const override;
+    void *surfaceRole() const override;
 
 protected:
     void requestWindowStates(Qt::WindowStates states) override;
