--- a/src/Distribution/Client/PackageHash.hs	2022-06-25 16:19:56.619145497 +0200
+++ b/src/Distribution/Client/PackageHash.hs	2022-06-25 16:22:32.855329932 +0200
@@ -272,9 +272,9 @@
       , entry "platform" prettyShow pkgHashPlatform
       , opt   "flags" mempty showFlagAssignment pkgHashFlagAssignment
       , opt   "configure-script" [] unwords pkgHashConfigureScriptArgs
-      , opt   "vanilla-lib" True  prettyShow pkgHashVanillaLib
-      , opt   "shared-lib"  False prettyShow pkgHashSharedLib
-      , opt   "dynamic-exe" False prettyShow pkgHashDynExe
+      , opt   "vanilla-lib" False  prettyShow pkgHashVanillaLib
+      , opt   "shared-lib"  True prettyShow pkgHashSharedLib
+      , opt   "dynamic-exe" True prettyShow pkgHashDynExe
       , opt   "fully-static-exe" False prettyShow pkgHashFullyStaticExe
       , opt   "ghci-lib"    False prettyShow pkgHashGHCiLib
       , opt   "prof-lib"    False prettyShow pkgHashProfLib
--- a/src/Distribution/Client/ProjectPlanning.hs	2022-06-25 16:39:10.306411882 +0200
+++ b/src/Distribution/Client/ProjectPlanning.hs	2022-06-25 16:39:42.369739301 +0200
@@ -1808,10 +1808,10 @@
 
         elabPkgDescriptionOverride = descOverride
 
-        elabVanillaLib    = perPkgOptionFlag pkgid True packageConfigVanillaLib --TODO: [required feature]: also needs to be handled recursively
+        elabVanillaLib    = perPkgOptionFlag pkgid False packageConfigVanillaLib --TODO: [required feature]: also needs to be handled recursively
         elabSharedLib     = pkgid `Set.member` pkgsUseSharedLibrary
         elabStaticLib     = perPkgOptionFlag pkgid False packageConfigStaticLib
-        elabDynExe        = perPkgOptionFlag pkgid False packageConfigDynExe
+        elabDynExe        = perPkgOptionFlag pkgid True packageConfigDynExe
         elabFullyStaticExe = perPkgOptionFlag pkgid False packageConfigFullyStaticExe
         elabGHCiLib       = perPkgOptionFlag pkgid False packageConfigGHCiLib --TODO: [required feature] needs to default to enabled on windows still
 
