diff --git a/haskell-ide-engine.cabal b/haskell-ide-engine.cabal
index 405e7b3f..dbee03c6 100644
--- a/haskell-ide-engine.cabal
+++ b/haskell-ide-engine.cabal
@@ -145,32 +142,6 @@ executable hie-wrapper
      ghc-options:      -Werror
   default-language:    Haskell2010
 
-library hie-test-utils
-  hs-source-dirs:      test/utils
-  exposed-modules:     TestUtils
-  build-depends:       base
-                     , haskell-ide-engine
-                     , haskell-lsp
-                     , hie-plugin-api
-                     , aeson
-                     , blaze-markup
-                     , containers
-                     , data-default
-                     , directory
-                     , filepath
-                     , ghc-mod-core
-                     , hslogger
-                     , hspec
-                     , hspec-core
-                     , stm
-                     , text
-                     , unordered-containers
-                     , yaml
-  ghc-options:         -Wall -Wredundant-constraints
-  if flag(pedantic)
-     ghc-options:      -Werror
-  default-language:    Haskell2010
-
 test-suite unit-test
   type:                exitcode-stdio-1.0
   hs-source-dirs:      test/unit
