diff --git a/src/Haskell/Ide/Engine/Plugin/Haddock.hs b/src/Haskell/Ide/Engine/Plugin/Haddock.hs
index 7cbefc3c..c290c800 100644
--- a/src/Haskell/Ide/Engine/Plugin/Haddock.hs
+++ b/src/Haskell/Ide/Engine/Plugin/Haddock.hs
@@ -109,7 +109,7 @@ getDocsForName df name = do
   case mf of
     Nothing -> return Nothing
     Just f -> do
-      ehi <- readInterfaceFile nameCacheFromIdeM f
+      ehi <- readInterfaceFile nameCacheFromIdeM f False
       case ehi of
         Left message -> do
           debugm $ "Haddock docs couldn't be loaded as readInterfaceFile failed with: " ++ message
@@ -157,14 +157,14 @@ prettyprintType n t = T.unlines
 renderDocs :: MDoc Name -> T.Text
 renderDocs = markup renderMarkDown . _doc
 
-renderMarkDown :: DocMarkup Name T.Text
+renderMarkDown :: DocMarkup (Wrap Name) T.Text
 renderMarkDown =
   Markup { markupEmpty = ""
          , markupString = T.pack
          , markupParagraph = (<> "\n\n")
          , markupAppend = mappend
-         , markupIdentifier = surround "`" . T.pack . getOccString
-         , markupIdentifierUnchecked = T.pack . occNameString . snd
+         , markupIdentifier = surround "`" . T.pack . showWrapped getOccString
+         , markupIdentifierUnchecked = T.pack . showWrapped (occNameString . snd)
          , markupModule = surround "**" . T.pack
          , markupWarning = surround "*"
          , markupEmphasis = surround "*"
@@ -179,7 +179,7 @@ renderMarkDown =
              T.pack $ maybe
                (hyperlinkUrl h)
                (\l -> "["<>l<>"]("<>hyperlinkUrl h<>")")
-               (hyperlinkLabel h)
+               (T.unpack <$> hyperlinkLabel h)
          , markupAName = T.pack
          , markupPic = const ""
          , markupMathInline = T.pack
