--- a/lib/autoconf/fortran.m4	2019-10-18 19:09:13.011447170 +0200
+++ b/lib/autoconf/fortran.m4	2019-10-18 19:09:27.338424467 +0200
@@ -661,7 +661,7 @@
 	     [_AC_LINKER_OPTION([$ac_arg], ac_cv_[]_AC_LANG_ABBREV[]_libs)])
 	  ;;
 	  # Ignore these flags.
-	-lang* | -lcrt*.o | -lc | -lgcc* | -lSystem | -libmil | -little \
+	-lang* | -lcrt*.o | -lc | -lwp | -lgcc* | -lSystem | -libmil | -little \
 	  |-LANG:=* | -LIST:* | -LNO:* | -link)
 	  ;;
 	-lkernel32)
