--- a/src/core/config/desktop_linux.pri	2016-02-15 16:18:46.000000000 +0100
+++ b/src/core/config/desktop_linux.pri	2016-03-06 15:15:10.880668744 +0100
@@ -6,5 +6,4 @@
     desktop_linux=1 \
     enable_plugins=1 \
 
-linux-clang: GYP_CONFIG += clang=1 host_clang=1 clang_use_chrome_plugins=0 make_clang_dir=/usr
-else: GYP_CONFIG += clang=0 host_clang=0
+GYP_CONFIG += clang=1 host_clang=1 clang_use_chrome_plugins=0 make_clang_dir=/usr
