--- a/src/3rdparty/chromium/sandbox/linux/suid/process_util.h.orig
+++ b/src/3rdparty/chromium/sandbox/linux/suid/process_util.h
@@ -11,6 +11,14 @@
 #include <stdbool.h>
 #include <sys/types.h>
 
+// Some additional functions
+# define TEMP_FAILURE_RETRY(expression) \
+	(__extension__			\
+	 ({ long int __result;		\
+	  do __result = (long int) (expression); \
+	  while (__result == -1L && errno == EINTR); \
+	  __result; }))
+
 // This adjusts /proc/process/oom_score_adj so the Linux OOM killer
 // will prefer certain process types over others. The range for the
 // adjustment is [-1000, 1000], with [0, 1000] being user accessible.
--- a/src/3rdparty/chromium/sandbox/linux/seccomp-bpf/trap.cc.orig
+++ b/src/3rdparty/chromium/sandbox/linux/seccomp-bpf/trap.cc
@@ -156,7 +156,7 @@
   // most versions of glibc don't include this information in siginfo_t. So,
   // we need to explicitly copy it into a arch_sigsys structure.
   struct arch_sigsys sigsys;
-  memcpy(&sigsys, &info->_sifields, sizeof(sigsys));
+  memcpy(&sigsys, &info->__si_fields, sizeof(sigsys));
 
 #if defined(__mips__)
   // When indirect syscall (syscall(__NR_foo, ...)) is made on Mips, the
--- a/src/3rdparty/chromium/net/base/dns_reloader.cc.orig
+++ b/src/3rdparty/chromium/net/base/dns_reloader.cc
@@ -5,7 +5,7 @@
 #include "net/base/dns_reloader.h"
 
 #if defined(OS_POSIX) && !defined(OS_MACOSX) && !defined(OS_OPENBSD) && \
-    !defined(OS_ANDROID)
+    !defined(OS_ANDROID) && !defined(_GNU_SOURCE)
 
 #include <resolv.h>
 
--- a/src/3rdparty/chromium/net/dns/dns_config_service_posix.cc.orig
+++ b/src/3rdparty/chromium/net/dns/dns_config_service_posix.cc
@@ -116,7 +116,7 @@
 ConfigParsePosixResult ReadDnsConfig(DnsConfig* config) {
   ConfigParsePosixResult result;
   config->unhandled_options = false;
-#if defined(OS_OPENBSD)
+#if defined(OS_OPENBSD) || defined(_GNU_SOURCE)
   // Note: res_ninit in glibc always returns 0 and sets RES_INIT.
   // res_init behaves the same way.
   memset(&_res, 0, sizeof(_res));
--- a/src/3rdparty/chromium/third_party/ffmpeg/libavutil/cpu.c.orig
+++ b/src/3rdparty/chromium/third_party/ffmpeg/libavutil/cpu.c
@@ -38,7 +38,6 @@
 #include <sys/param.h>
 #endif
 #include <sys/types.h>
-#include <sys/sysctl.h>
 #endif
 #if HAVE_UNISTD_H
 #include <unistd.h>
--- a/src/3rdparty/chromium/third_party/WebKit/Source/platform/heap/StackFrameDepth.cpp.orig
+++ b/src/3rdparty/chromium/third_party/WebKit/Source/platform/heap/StackFrameDepth.cpp
@@ -58,7 +58,7 @@
     // FIXME: On Mac OSX and Linux, this method cannot estimate stack size
     // correctly for the main thread.
 
-#if defined(__GLIBC__) || OS(ANDROID) || OS(FREEBSD)
+#if OS(LINUX) || OS(ANDROID) || OS(FREEBSD)
     // pthread_getattr_np() can fail if the thread is not invoked by
     // pthread_create() (e.g., the main thread of webkit_unit_tests).
     // In this case, this method returns 0 and the caller must handle it.
@@ -105,7 +105,7 @@
 
 void* StackFrameDepth::getStackStart()
 {
-#if defined(__GLIBC__) || OS(ANDROID) || OS(FREEBSD)
+#if OS(LINUX) || OS(ANDROID) || OS(FREEBSD)
     pthread_attr_t attr;
     int error;
 #if OS(FREEBSD)
--- a/src/3rdparty/chromium/third_party/ffmpeg/libavutil/autorename_libavutil_cpu.c.orig
+++ b/src/3rdparty/chromium/third_party/ffmpeg/libavutil/autorename_libavutil_cpu.c
@@ -38,7 +38,6 @@
 #include <sys/param.h>
 #endif
 #include <sys/types.h>
-#include <sys/sysctl.h>
 #endif
 #if HAVE_UNISTD_H
 #include <unistd.h>
--- a/src/3rdparty/chromium/device/serial/serial_io_handler_posix.cc.orig
+++ b/src/3rdparty/chromium/device/serial/serial_io_handler_posix.cc
@@ -6,6 +6,7 @@
 
 #include <sys/ioctl.h>
 #include <termios.h>
+#include <asm-generic/ioctls.h>
 
 #include "base/posix/eintr_wrapper.h"
 
--- a/src/3rdparty/chromium/net/dns/host_resolver_impl.cc.orig
+++ b/src/3rdparty/chromium/net/dns/host_resolver_impl.cc
@@ -1881,8 +1881,7 @@
 #endif
   NetworkChangeNotifier::AddIPAddressObserver(this);
   NetworkChangeNotifier::AddDNSObserver(this);
-#if defined(OS_POSIX) && !defined(OS_MACOSX) && !defined(OS_OPENBSD) && \
-    !defined(OS_ANDROID)
+#if defined(__GLIBC__)  
   EnsureDnsReloaderInit();
 #endif
 
--- a/src/3rdparty/chromium/net/dns/host_resolver_proc.cc.orig
+++ b/src/3rdparty/chromium/net/dns/host_resolver_proc.cc
@@ -193,8 +193,7 @@
   // Restrict result set to only this socket type to avoid duplicates.
   hints.ai_socktype = SOCK_STREAM;
 
-#if defined(OS_POSIX) && !defined(OS_MACOSX) && !defined(OS_OPENBSD) && \
-    !defined(OS_ANDROID)
+#if defined(__GLIBC__)  
   DnsReloaderMaybeReload();
 #endif
   int err = getaddrinfo(host.c_str(), NULL, &hints, &ai);
--- a/src/3rdparty/chromium/third_party/WebKit/Source/platform/heap/GCInfo.cpp.bak	2015-11-29 20:57:42.365387784 +0100
+++ b/src/3rdparty/chromium/third_party/WebKit/Source/platform/heap/GCInfo.cpp	2015-11-29 20:58:08.738551520 +0100
@@ -5,6 +5,7 @@
 #include "config.h"
 #include "platform/heap/GCInfo.h"
 
+#include <malloc.h>
 #include "platform/heap/Handle.h"
 #include "platform/heap/Heap.h"
 
--- a/src/3rdparty/chromium/third_party/ots/include/opentype-sanitiser.h.bak	2015-11-29 20:57:44.090326971 +0100
+++ b/src/3rdparty/chromium/third_party/ots/include/opentype-sanitiser.h	2015-11-29 20:58:20.963225505 +0100
@@ -24,6 +24,7 @@
 #include <stdint.h>
 #endif
 
+#include <sys/types.h>
 #include <algorithm>
 #include <cassert>
 #include <cstddef>
--- a/src/3rdparty/chromium/third_party/libxml/src/threads.c.bak	2016-03-06 15:20:33.508684629 +0100
+++ b/src/3rdparty/chromium/third_party/libxml/src/threads.c	2016-03-06 15:20:45.957685242 +0100
@@ -77,8 +77,6 @@
 	   __attribute((weak));
 extern int pthread_cond_wait ()
 	   __attribute((weak));
-extern int pthread_equal ()
-	   __attribute((weak));
 extern pthread_t pthread_self ()
 	   __attribute((weak));
 extern int pthread_key_create ()
