--- a/src/3rdparty/chromium/third_party/libxml/src/threads.c	2016-04-22 09:54:12.479200506 +0200
+++ b/src/3rdparty/chromium/third_party/libxml/src/threads.c	2016-04-22 09:54:22.978201022 +0200
@@ -77,8 +77,6 @@
 	   __attribute((weak));
 extern int pthread_cond_wait ()
 	   __attribute((weak));
-extern int pthread_equal ()
-	   __attribute((weak));
 extern pthread_t pthread_self ()
 	   __attribute((weak));
 extern int pthread_key_create ()
