--- ./src/3rdparty/chromium/base/trace_event/malloc_dump_provider.cc.orig
+++ ./src/3rdparty/chromium/base/trace_event/malloc_dump_provider.cc
@@ -29,6 +29,7 @@
 // Called at trace dump point time. Creates a snapshot the memory counters for
 // the current process.
 bool MallocDumpProvider::OnMemoryDump(ProcessMemoryDump* pmd) {
+#if defined(__GLIBC__)
   struct mallinfo info = mallinfo();
   DCHECK_GE(info.arena + info.hblkhd, info.uordblks);
 
@@ -46,7 +47,7 @@
   MemoryAllocatorDump* inner_dump = pmd->CreateAllocatorDump(kAllocatedObjects);
   inner_dump->AddScalar(MemoryAllocatorDump::kNameSize,
                         MemoryAllocatorDump::kUnitsBytes, info.uordblks);
-
+#endif
   return true;
 }
 
--- ./src/3rdparty/chromium/content/child/content_child_helpers.cc
+++ ./src/3rdparty/chromium/content/child/content_child_helpers.cc
@@ -18,7 +18,7 @@
 
 namespace content {
 
-#if defined(OS_LINUX) || defined(OS_ANDROID)
+#if defined(_GLIBC_) || defined(OS_ANDROID)
 size_t GetMemoryUsageKB() {
   struct mallinfo minfo = mallinfo();
   uint64_t mem_usage =
