--- a/tools/qmake/mkspecs/features/functions.prf	2016-02-15 16:18:46.000000000 +0100
+++ b/tools/qmake/mkspecs/features/functions.prf.new	2016-03-06 15:11:57.954659246 +0100
@@ -4,7 +4,7 @@
       skipBuild("Qt WebEngine on Linux requires clang or GCC.")
       return(false)
     }
-    gcc:!clang:!isGCCVersionSupported(): return(false)
+    gcc:!clang: return(false)
   } else:win32 {
     winrt {
       skipBuild("WinRT is not supported.")
