--- a/src/3rdparty/chromium/tools/gn/bootstrap/bootstrap.py	2017-10-26 21:47:29.119621617 +0200
+++ b/src/3rdparty/chromium/tools/gn/bootstrap/bootstrap.py	2017-10-26 21:47:40.287600830 +0200
@@ -726,9 +726,6 @@
   cmd.append('gn')
   check_call(cmd)
 
-  if not options.debug and not is_win:
-    check_call(['strip', os.path.join(build_dir, 'gn')])
-
 
 if __name__ == '__main__':
   sys.exit(main(sys.argv[1:]))
