diff --git a/shim.h b/shim.h
index 3c98d85..912847c 100644
--- a/shim.h
+++ b/shim.h
@@ -2,7 +2,6 @@
 #define __COPENSSL_SHIM_H__
 
 #include <openssl/conf.h>
-#include <openssl/evp.h>
 #include <openssl/err.h>
 #include <openssl/bio.h>
 #include <openssl/ssl.h>
@@ -14,5 +13,6 @@
 #include <openssl/ripemd.h>
 #include <openssl/pkcs12.h>
 #include <openssl/x509v3.h>
+#include <openssl/evp.h>
 
 #endif
