--- a/Source/JavaScriptCore/runtime/VMStackBounds.h.bak	2015-10-18 11:26:09.578266128 +0200
+++ b/Source/JavaScriptCore/runtime/VMStackBounds.h	2015-10-18 11:26:24.653177666 +0200
@@ -53,8 +53,8 @@
         //
         // These sizes were derived from the stack usage of a number of sites when
         // layout occurs when we've already consumed most of the C stack.
-        const size_t requiredStack = 128 * KB;
-        const size_t errorModeRequiredStack = 64 * KB;
+        const size_t requiredStack = 64 * KB;
+        const size_t errorModeRequiredStack = 16 * KB;
 
         size_t requiredCapacity = interpreter->isInErrorHandlingMode() ? errorModeRequiredStack : requiredStack;
         RELEASE_ASSERT(m_bounds.size() >= requiredCapacity);
