diff --git a/Sources/Build/BuildPlan.swift b/Sources/Build/BuildPlan.swift
index be8e2bf7..f8e9d177 100644
--- a/Sources/Build/BuildPlan.swift
+++ b/Sources/Build/BuildPlan.swift
@@ -260,7 +260,6 @@ public final class SwiftTargetDescription {
     /// The arguments needed to compile this target.
     public func compileArguments() -> [String] {
         var args = [String]()
-        args += ["-swift-version", String(swiftVersion)]
         args += buildParameters.toolchain.extraSwiftCFlags
         args += buildParameters.swiftCompilerFlags
         args += optimizationArguments
diff --git a/Sources/PackageLoading/ManifestLoader.swift b/Sources/PackageLoading/ManifestLoader.swift
index 86f8107d..4906b9c1 100644
--- a/Sources/PackageLoading/ManifestLoader.swift
+++ b/Sources/PackageLoading/ManifestLoader.swift
@@ -304,7 +304,6 @@ public final class ManifestLoader: ManifestLoaderProtocol {
     ) -> [String] {
         var cmd = [String]()
         let runtimePath = self.runtimePath(for: manifestVersion)
-        cmd += ["-swift-version", String(manifestVersion.rawValue)]
         cmd += ["-I", runtimePath.asString]
       #if os(macOS)
         cmd += ["-target", "x86_64-apple-macosx10.10"]
diff --git a/Utilities/bootstrap b/Utilities/bootstrap
index 0e3710f4..fa04dc81 100755
--- a/Utilities/bootstrap
+++ b/Utilities/bootstrap
@@ -216,9 +216,6 @@ class Target(object):
         if args.libdispatch_source_dir:
             other_args.extend(["-Xcc", "-fblocks"])
 
-	# FIXME: Default to swift version 3 for now.
-        other_args.extend(["-swift-version", "3"])
-        
         print("  %s:" % json.dumps(compile_swift_node), file=output)
         print("    tool: swift-compiler", file=output)
         print("    executable: %s" % json.dumps(args.swiftc_path), file=output)
