diff --git a/Sources/Basic/OutputByteStream.swift b/Sources/Basic/OutputByteStream.swift
index aa0bb0af..42cf9793 100644
--- a/Sources/Basic/OutputByteStream.swift
+++ b/Sources/Basic/OutputByteStream.swift
@@ -566,7 +566,7 @@ public class FileOutputByteStream: OutputByteStream {
 public final class LocalFileOutputByteStream: FileOutputByteStream {
 
     /// The pointer to the file.
-    let fp: UnsafeMutablePointer<FILE>
+    let fp: OpaquePointer
 
     /// True if there were any IO error during writing.
     private var error: Bool = false
@@ -575,7 +575,7 @@ public final class LocalFileOutputByteStream: FileOutputByteStream {
     private var closeOnDeinit: Bool
 
     /// Instantiate using the file pointer.
-    init(filePointer: UnsafeMutablePointer<FILE>, closeOnDeinit: Bool = true) throws {
+    init(filePointer: OpaquePointer, closeOnDeinit: Bool = true) throws {
         self.fp = filePointer
         self.closeOnDeinit = closeOnDeinit
         super.init()
diff --git a/Sources/Utility/InterruptHandler.swift b/Sources/Utility/InterruptHandler.swift
index 51738b41..22fb61d4 100644
--- a/Sources/Utility/InterruptHandler.swift
+++ b/Sources/Utility/InterruptHandler.swift
@@ -41,7 +41,7 @@ public final class InterruptHandler {
       #if os(macOS)
         action.__sigaction_u.__sa_handler = signalHandler
       #else
-        action.__sigaction_handler = unsafeBitCast(signalHandler, to: sigaction.__Unnamed_union___sigaction_handler.self)
+        action.__sa_handler = unsafeBitCast(signalHandler, to: sigaction.__Unnamed_union___sa_handler.self)
       #endif
         // Install the new handler.
         sigaction(SIGINT, &action, &oldAction)
diff --git a/Sources/Commands/SwiftTool.swift b/Sources/Commands/SwiftTool.swift
index 79bf5d1f..60c20e2a 100644
--- a/Sources/Commands/SwiftTool.swift
+++ b/Sources/Commands/SwiftTool.swift
@@ -173,7 +173,7 @@ public class SwiftTool<Options: ToolOptions> {
               #if os(macOS)
                 action.__sigaction_u.__sa_handler = SIG_DFL
               #else
-                action.__sigaction_handler = unsafeBitCast(SIG_DFL, to: sigaction.__Unnamed_union___sigaction_handler.self)
+                action.__sa_handler = unsafeBitCast(SIG_DFL, to: sigaction.__Unnamed_union___sa_handler.self)
               #endif
                 sigaction(SIGINT, &action, nil)
 
