diff --git a/Sources/Commands/Destination.swift b/Sources/Commands/Destination.swift
index f5dc5093..77af5ddb 100644
--- a/Sources/Commands/Destination.swift
+++ b/Sources/Commands/Destination.swift
@@ -123,7 +123,7 @@ public struct Destination {
             dynamicLibraryExtension: "so",
             extraCCFlags: ["-fPIC"],
             extraSwiftCFlags: [],
-            extraCPPFlags: ["-lstdc++"]
+            extraCPPFlags: ["-lc++"]
         )
       #endif
     }
diff --git a/Sources/Commands/UserToolchain.swift b/Sources/Commands/UserToolchain.swift
index 38d0fae5..be5dd404 100644
--- a/Sources/Commands/UserToolchain.swift
+++ b/Sources/Commands/UserToolchain.swift
@@ -111,12 +111,10 @@ public struct UserToolchain: Toolchain {
         }
 
         self.extraSwiftCFlags = [
-            "-target", destination.target,
             "-sdk", destination.sdk.asString
         ] + destination.extraSwiftCFlags
 
         self.extraCCFlags = [
-            "-target", destination.target,
             "--sysroot", destination.sdk.asString
         ] + destination.extraCCFlags
 
