--- a/hw/scsi/mptsas.c	2017-11-12 10:50:04.109707173 +0100
+++ b/hw/scsi/mptsas.c	2017-11-12 10:50:17.924681363 +0100
@@ -1238,9 +1238,6 @@
      * and fail migration instead of asserting here.
      * When we do, we might be able to re-enable NDEBUG below.
      */
-#ifdef NDEBUG
-#error building with NDEBUG is not supported
-#endif
     assert(n >= 0);
 
     pci_dma_sglist_init(&req->qsg, pci, n);
--- a/hw/virtio/virtio.c	2017-11-12 10:51:24.405554057 +0100
+++ b/hw/virtio/virtio.c	2017-11-12 10:51:41.324521131 +0100
@@ -1027,9 +1027,6 @@
      * of asserting here.
      * When we do, we might be able to re-enable NDEBUG below.
      */
-#ifdef NDEBUG
-#error building with NDEBUG is not supported
-#endif
     assert(ARRAY_SIZE(data.in_addr) >= data.in_num);
     assert(ARRAY_SIZE(data.out_addr) >= data.out_num);
 
