--- a/include/qemu/osdep.h	2018-02-14 11:46:13.816383800 +0100
+++ b/include/qemu/osdep.h	2018-02-14 11:46:48.646295225 +0100
@@ -120,9 +120,9 @@
  * submit patches to remove any side-effects inside an assertion, or
  * fixing error handling that should use Error instead of assert.
  */
-#ifdef NDEBUG
-#error building with NDEBUG is not supported
-#endif
+//#ifdef NDEBUG
+//#error building with NDEBUG is not supported
+//#endif
 #ifdef G_DISABLE_ASSERT
 #error building with G_DISABLE_ASSERT is not supported
 #endif
