--- a/mkspecs/linux-clang/qplatformdefs.h	2017-10-02 08:43:38.000000000 +0000
+++ b/mkspecs/linux-clang/qplatformdefs.h	2017-11-08 19:58:19.901889505 +0000
@@ -84,11 +84,7 @@
 
 #undef QT_SOCKLEN_T
 
-#if defined(__GLIBC__) && (__GLIBC__ >= 2)
 #define QT_SOCKLEN_T            socklen_t
-#else
-#define QT_SOCKLEN_T            int
-#endif
 
 #if defined(_XOPEN_SOURCE) && (_XOPEN_SOURCE >= 500)
 #define QT_SNPRINTF             ::snprintf
