--- a/src/network/kernel/qnetworkinterface_linux.cpp	2023-03-23 20:06:47.000000000 +0100
+++ b/src/network/kernel/qnetworkinterface_linux.cpp	2023-05-14 17:50:59.198703949 +0200
@@ -75,9 +75,6 @@
         return QNetworkInterface::Loopback;
 
     case ARPHRD_ETHER:
-        // check if it's a WiFi interface
-        if (qt_safe_ioctl(socket, SIOCGIWMODE, req) >= 0)
-            return QNetworkInterface::Wifi;
         return QNetworkInterface::Ethernet;
 
     case ARPHRD_SLIP:
