--- a/Package@swift-4.swift	2017-09-22 10:56:55.147921427 +0200
+++ b/Package@swift-4.swift	2017-09-22 10:57:19.191920870 +0200
@@ -4,12 +4,12 @@
 let package = Package(
     name: "Engine",
     products: [
-        .library(name: "CHTTP", targets: ["CHTTP"]),
-        .library(name: "URI", targets: ["URI"]),
-        .library(name: "Cookies", targets: ["Cookies"]),
-        .library(name: "HTTP", targets: ["HTTP"]),
-        .library(name: "WebSockets", targets: ["WebSockets"]),
-        .library(name: "SMTP", targets: ["SMTP"]),
+        .library(name: "CHTTP", type: .dynamic, targets: ["CHTTP"]),
+        .library(name: "URI", type: .dynamic, targets: ["URI"]),
+        .library(name: "Cookies", type: .dynamic, targets: ["Cookies"]),
+        .library(name: "HTTP", type: .dynamic, targets: ["HTTP"]),
+        .library(name: "WebSockets", type: .dynamic, targets: ["WebSockets"]),
+        .library(name: "SMTP", type: .dynamic, targets: ["SMTP"]),
     ],
     targets: [
         .target(name: "CHTTP"),
