diff --git a/Package.swift b/Package.swift
index 1d5f160f..c27c73c8 100644
--- a/Package.swift
+++ b/Package.swift
@@ -67,3 +55,14 @@ let package = Package(
         "Sources/SMTPExample",
     ]
 )
+
+let ulib = Product(name: "URI", type: .Library(.Dynamic), modules: "URI")
+let tlib = Product(name: "Transport", type: .Library(.Dynamic), modules: "Transport")
+let hlib = Product(name: "HTTP", type: .Library(.Dynamic), modules: "HTTP")
+let wlib = Product(name: "WebSockets", type: .Library(.Dynamic), modules: "WebSockets")
+let slib = Product(name: "SMTP", type: .Library(.Dynamic), modules: "SMTP")
+products.append(ulib)
+products.append(tlib)
+products.append(hlib)
+products.append(wlib)
+products.append(slib)
