diff --git a/Package.swift b/Package.swift
index 7f4c75d1..48387a18 100644
--- a/Package.swift
+++ b/Package.swift
@@ -1,15 +1,5 @@
 import PackageDescription
 
-let beta = Version(2,0,0, prereleaseIdentifiers: ["beta"])
-
-let dependencies: [Package.Dependency] = [
-    // Crypto
-    .Package(url: "https://github.com/vapor/crypto.git", beta),
-
-    // Secure Sockets
-    .Package(url: "https://github.com/vapor/tls.git", beta)
-]
-
 let package = Package(
     name: "Engine",
     targets: [
@@ -21,7 +11,6 @@ let package = Package(
         Target(name: "SMTP")
         // Target(name: "Performance", dependencies: ["HTTP"])
     ],
-    dependencies: dependencies,
     exclude: [
         "Sources/Performance"
     ]
