--- a/src/config.c	2018-09-19 16:52:59.000000000 +0200
+++ b/src/config.c	2018-09-23 19:24:20.833880608 +0200
@@ -462,7 +462,7 @@
 
       p = ssh_config_get_str_tok(&s, NULL);
       if (p && *parsing) {
-#ifdef HAVE_GLOB
+#if defined(HAVE_GLOB) && defined(HAVE_GLOB_GL_FLAGS_MEMBER)
         local_parse_glob(session, p, parsing, seen);
 #else
         local_parse_file(session, p, parsing, seen);
