--- a/ConfigureChecks.cmake	2018-08-29 21:52:42.000000000 +0200
+++ b/ConfigureChecks.cmake	2018-09-21 15:32:59.771215805 +0200
@@ -63,7 +63,6 @@
 check_include_file(sys/param.h HAVE_SYS_PARAM_H)
 check_include_file(arpa/inet.h HAVE_ARPA_INET_H)
 check_include_file(byteswap.h HAVE_BYTESWAP_H)
-check_include_file(glob.h HAVE_GLOB_H)
 
 if (WIN32)
   check_include_file(io.h HAVE_IO_H)
@@ -155,10 +154,6 @@
 check_function_exists(explicit_bzero HAVE_EXPLICIT_BZERO)
 check_function_exists(memset_s HAVE_MEMSET_S)
 
-if (HAVE_GLOB_H)
-  check_function_exists(glob HAVE_GLOB)
-endif (HAVE_GLOB_H)
-
 if (NOT WIN32)
   check_function_exists(vsnprintf HAVE_VSNPRINTF)
   check_function_exists(snprintf HAVE_SNPRINTF)
