--- a/arch/arm64/boot/dts/rockchip/rk3399-pinebook-pro.dts
+++ b/arch/arm64/boot/dts/rockchip/rk3399-pinebook-pro.dts
@@ -102,12 +109,11 @@
 
 		red_led: led-1 {
 			color = <LED_COLOR_ID_RED>;
-			default-state = "off";
+			default-state = "on";
 			function = LED_FUNCTION_STANDBY;
-			gpios = <&gpio0 RK_PA2 GPIO_ACTIVE_HIGH>;
-			label = "red:standby";
+			gpios = <&gpio0 RK_PA2 GPIO_ACTIVE_LOW>;
+			label = "red:power";
 			panic-indicator;
-			retain-state-suspended;
 		};
 	};
 
