--- a/arch/arm64/boot/dts/rockchip/rk3399-pinebook-pro.dts
+++ b/arch/arm64/boot/dts/rockchip/rk3399-pinebook-pro.dts
@@ -102,12 +109,11 @@
 
 		red_led: led-1 {
 			color = <LED_COLOR_ID_RED>;
-			default-state = "off";
+			default-state = "on";
 			function = LED_FUNCTION_STANDBY;
-			gpios = <&gpio0 RK_PA2 GPIO_ACTIVE_HIGH>;
-			label = "red:standby";
+			gpios = <&gpio0 RK_PA2 GPIO_ACTIVE_LOW>;
+			label = "red:power";
 			panic-indicator;
-			retain-state-suspended;
 		};
 	};
 
@@ -372,8 +378,19 @@
 	};
 };
 
+ /* Don't use vopl for dp, save it for edp */
+ /delete-node/ &dp_in_vopl;
+ /delete-node/ &vopl_out_dp;
+ /delete-node/ &vopb_out_edp;
+ /delete-node/ &edp_in_vopb;
+
 &cdn_dp {
 	status = "okay";
+	extcon = <&fusb0>;
+};
+
+&cdn_dp_sound {
+	status = "okay";
 };
 
 &cpu_b0 {
@@ -429,10 +446,6 @@
 	status = "okay";
 };
 
-&hdmi_sound {
-	status = "okay";
-};
-
 &i2c0 {
 	clock-frequency = <400000>;
 	i2c-scl-falling-time-ns = <4>;
@@ -698,7 +711,7 @@
 
 		connector {
 			compatible = "usb-c-connector";
-			data-role = "host";
+			data-role = "dual";
 			label = "USB-C";
 			op-sink-microwatt = <1000000>;
 			power-role = "dual";
@@ -708,6 +721,9 @@
 				<PDO_FIXED(5000, 1400, PDO_FIXED_USB_COMM)>;
 			try-power-role = "sink";
 
+			extcon-cables = <1 2 5 6 9 10 12 44>;
+			typec-altmodes = <0xff01 1 0x001c0000 1>;
+
 			ports {
 				#address-cells = <1>;
 				#size-cells = <0>;
@@ -751,6 +785,10 @@
 	status = "okay";
 };
 
+&i2s2 {
+	status = "okay";
+};
+
 &io_domains {
 	audio-supply = <&vcc_3v0>;
 	gpio1830-supply = <&vcc_3v0>;
@@ -957,6 +997,7 @@
 };
 
 &tcphy0 {
+	extcon = <&fusb0>;
 	status = "okay";
 };
 
