--- a/arch/arm64/boot/dts/rockchip/rk3399-pinebook-pro.dts	2020-06-01 01:49:15.000000000 +0200
+++ b/arch/arm64/boot/dts/rockchip/rk3399-pinebook-pro.dts	2020-06-01 16:05:48.292178199 +0200
@@ -28,6 +28,13 @@
 		pwms = <&pwm0 0 740740 0>;
 	};
 
+	bat: battery {
+		compatible = "simple-battery";
+		charge-full-design-microamp-hours = <9800000>;
+		voltage-max-design-microvolt = <4350000>;
+		voltage-min-design-microvolt = <3000000>;
+	};
+
 	edp_panel: edp-panel {
 		compatible = "boe,nv140fhmn49";
 		backlight = <&backlight>;
@@ -102,12 +109,11 @@
 
 		red-led {
 			color = <LED_COLOR_ID_RED>;
-			default-state = "off";
+			default-state = "on";
 			function = LED_FUNCTION_STANDBY;
-			gpios = <&gpio0 RK_PA2 GPIO_ACTIVE_HIGH>;
-			label = "red:standby";
+			gpios = <&gpio0 RK_PA2 GPIO_ACTIVE_LOW>;
+			label = "red:power";
 			panic-indicator;
-			retain-state-suspended;
 		};
 	};
 
@@ -372,8 +378,19 @@
 	};
 };
 
+ /* Don't use vopl for dp, save it for edp */
+ /delete-node/ &dp_in_vopl;
+ /delete-node/ &vopl_out_dp;
+ /delete-node/ &vopb_out_edp;
+ /delete-node/ &edp_in_vopb;
+
 &cdn_dp {
 	status = "okay";
+	extcon = <&fusb0>;
+};
+
+&cdn_dp_sound {
+	status = "okay";
 };
 
 &cpu_b0 {
@@ -429,10 +446,6 @@
 	status = "okay";
 };
 
-&hdmi_sound {
-	status = "okay";
-};
-
 &i2c0 {
 	clock-frequency = <400000>;
 	i2c-scl-falling-time-ns = <4>;
@@ -708,6 +720,9 @@
 				<PDO_FIXED(5000, 1400, PDO_FIXED_USB_COMM)>;
 			try-power-role = "sink";
 
+			extcon-cables = <1 2 5 6 9 10 12 44>;
+			typec-altmodes = <0xff01 1 0x001c0000 1>;
+
 			ports {
 				#address-cells = <1>;
 				#size-cells = <0>;
@@ -741,17 +756,39 @@
 			};
 		};
 	};
+
+	cw2015@62 {
+		compatible = "cellwise,cw2015";
+		reg = <0x62>;
+		cellwise,battery-profile = /bits/ 8 <
+			0x17 0x67 0x80 0x73 0x6E 0x6C 0x6B 0x63
+			0x77 0x51 0x5C 0x58 0x50 0x4C 0x48 0x36
+			0x15 0x0C 0x0C 0x19 0x5B 0x7D 0x6F 0x69
+			0x69 0x5B 0x0C 0x29 0x20 0x40 0x52 0x59
+			0x57 0x56 0x54 0x4F 0x3B 0x1F 0x7F 0x17
+			0x06 0x1A 0x30 0x5A 0x85 0x93 0x96 0x2D
+			0x48 0x77 0x9C 0xB3 0x80 0x52 0x94 0xCB
+			0x2F 0x00 0x64 0xA5 0xB5 0x11 0xF0 0x11
+		>;
+		cellwise,monitor-interval-ms = <5000>;
+		monitored-battery = <&bat>;
+		power-supplies = <&mains_charger>, <&fusb0>;
+	};
 };
 
 &i2s1 {
-	#sound-dai-cells = <0>;
-	pinctrl-names = "default";
 	pinctrl-0 = <&i2s_8ch_mclk_gpio>, <&i2s1_2ch_bus>;
 	rockchip,capture-channels = <8>;
 	rockchip,playback-channels = <8>;
 	status = "okay";
 };
 
+&i2s2 {
+	rockchip,capture-channels = <8>;
+	rockchip,playback-channels = <8>;
+	status = "okay";
+};
+
 &io_domains {
 	audio-supply = <&vcc_3v0>;
 	gpio1830-supply = <&vcc_3v0>;
@@ -958,6 +995,7 @@
 };
 
 &tcphy0 {
+	extcon = <&fusb0>;
 	status = "okay";
 };
 
