--- a/coreclr-1.1.1/compileoptions.cmake	2017-09-19 09:51:48.030879983 +0200
+++ b/coreclr-1.1.1/compileoptions.cmake	2017-09-19 09:52:49.922775025 +0200
@@ -23,9 +23,6 @@
   # after hitting just about 20 errors.
   add_compile_options(-ferror-limit=4096)
 
-  # All warnings that are not explicitly disabled are reported as errors
-  add_compile_options(-Werror)
-
   # Disabled warnings
   add_compile_options(-Wno-unused-private-field)
   add_compile_options(-Wno-unused-variable)
