--- a/coreclr-1.1.1/src/ToolBox/superpmi/superpmi-shared/methodcontext.cpp	2017-09-19 09:51:10.272943604 +0200
+++ b/coreclr-1.1.1/src/ToolBox/superpmi/superpmi-shared/methodcontext.cpp	2017-09-19 09:51:24.041919971 +0200
@@ -285,7 +285,7 @@
 
 void MethodContext::MethodInitHelper(unsigned char *buff2, unsigned int totalLen)
 {
-    MethodContext::MethodContext();
+    MethodContext();
 
     unsigned int buffIndex = 0;
     unsigned int localsize = 0;
