--- a/src/ToolBox/SOS/Strike/strike.cpp
+++ b/src/ToolBox/SOS/Strike/strike.cpp
@@ -82,6 +82,7 @@
 #include <dbghelp.h>
 
 #include <stdio.h>
+#include <sys/types.h>
 #include <stdlib.h>
 #include <string.h>
 #include <stddef.h>
--- a/src/debug/createdump/crashinfo.h
+++ b/src/debug/createdump/crashinfo.h
@@ -3,6 +3,10 @@
 // See the LICENSE file in the project root for more information.
 
 // typedef for our parsing of the auxv variables in /proc/pid/auxv.
+
+#include <array>
+#include <string>
+
 #if defined(__i386) || defined(__ARM_EABI__) 
 typedef Elf32_auxv_t elf_aux_entry;
 #elif defined(__x86_64) || defined(__aarch64__)
--- a/src/debug/createdump/memoryregion.h
+++ b/src/debug/createdump/memoryregion.h
@@ -2,6 +2,8 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+#define PAGE_SHIFT 12
+
 enum MEMORY_REGION_FLAGS : uint32_t
 {
     // PF_X        = 0x01,      // Execute
--- a/src/debug/createdump/threadinfo.cpp
+++ b/src/debug/createdump/threadinfo.cpp
@@ -4,6 +4,10 @@
 
 #include "createdump.h"
 
+#ifndef __GLIBC__
+typedef int __ptrace_request;
+#endif
+
 #define FPREG_ErrorOffset(fpregs) *(DWORD*)&((fpregs).rip)
 #define FPREG_ErrorSelector(fpregs) *(((WORD*)&((fpregs).rip)) + 2)
 #define FPREG_DataOffset(fpregs) *(DWORD*)&((fpregs).rdp)
--- a/src/debug/daccess/nidump.cpp
+++ b/src/debug/daccess/nidump.cpp
@@ -9239,7 +9239,7 @@
 #define assert(a)
 #define NOTHROW
 #define GC_NOTRIGGER
-#include <gcdecoder.cpp>
+//#include <gcdecoder.cpp>
 #undef NOTHROW
 #undef GC_NOTRIGGER
 
@@ -9253,16 +9253,16 @@
 #undef _ASSERTE
 #define _ASSERTE(a) do {} while (0)
 #ifdef _TARGET_X86_
-#include <gcdump.cpp>
+//#include <gcdump.cpp>
 #endif
 
 #undef LIMITED_METHOD_CONTRACT
 #undef WRAPPER_NO_CONTRACT
 #ifdef _TARGET_X86_
-#include <i386/gcdumpx86.cpp>
+//#include <i386/gcdumpx86.cpp>
 #else // !_TARGET_X86_
 #undef PREGDISPLAY
-#include <gcdumpnonx86.cpp>
+//#include <gcdumpnonx86.cpp>
 #endif // !_TARGET_X86_
 
 #ifdef __MSC_VER
--- a/src/pal/tools/clang-compiler-override.txt
+++ b/src/pal/tools/clang-compiler-override.txt
@@ -1,16 +1,5 @@
 SET (CMAKE_C_FLAGS_INIT                "-Wall -std=c11")
-SET (CMAKE_C_FLAGS_DEBUG_INIT          "-g -O0")
-SET (CLR_C_FLAGS_CHECKED_INIT          "-g -O2")
-# Refer to the below instruction to support __thread with -O2/-O3 on Linux/ARM
-# https://github.com/dotnet/coreclr/blob/master/Documentation/building/linux-instructions.md
-SET (CMAKE_C_FLAGS_RELEASE_INIT        "-g -O3")
-SET (CMAKE_C_FLAGS_RELWITHDEBINFO_INIT "-g -O2")
-
 SET (CMAKE_CXX_FLAGS_INIT                "-Wall -Wno-null-conversion -std=c++11")
-SET (CMAKE_CXX_FLAGS_DEBUG_INIT          "-g -O0")
-SET (CLR_CXX_FLAGS_CHECKED_INIT          "-g -O2")
-SET (CMAKE_CXX_FLAGS_RELEASE_INIT        "-g -O3")
-SET (CMAKE_CXX_FLAGS_RELWITHDEBINFO_INIT "-g -O2")
 
 SET (CLR_DEFINES_DEBUG_INIT              DEBUG _DEBUG _DBG URTBLDENV_FRIENDLY=Checked BUILDENV_CHECKED=1)
 SET (CLR_DEFINES_CHECKED_INIT            DEBUG _DEBUG _DBG URTBLDENV_FRIENDLY=Checked BUILDENV_CHECKED=1)
--- a/build.sh	2018-06-04 14:41:09.000000000 +0200
+++ b/build.sh.	2018-06-27 22:01:34.248069855 +0200
@@ -386,7 +386,7 @@
 
     if [ "$__HostArch" == "x64" ]; then
         if [ "$__HostOS" == "Linux" ]; then
-            __isMSBuildOnNETCoreSupported=1
+            __isMSBuildOnNETCoreSupported=0
             # note: the RIDs below can use globbing patterns
             UNSUPPORTED_RIDS=("debian.9-x64" "ubuntu.17.04-x64")
             for UNSUPPORTED_RID in "${UNSUPPORTED_RIDS[@]}"
--- a/CMakeLists.txt	2018-05-04 17:36:32.867992320 +0200
+++ b/CMakeLists.txt	2018-05-04 17:36:44.561970508 +0200
@@ -177,6 +177,8 @@
             set(CLR_CMAKE_TARGET_TIZEN_LINUX 1)
         elseif(CLR_CMAKE_LINUX_ID STREQUAL alpine)
             set(CLR_CMAKE_PLATFORM_ALPINE_LINUX 1)
+        elseif(CLR_CMAKE_LINUX_ID STREQUAL gentoo)
+            set(CLR_CMAKE_PLATFORM_ALPINE_LINUX 1)
         endif()
         if(CLR_CMAKE_LINUX_ID STREQUAL ubuntu)
             set(CLR_CMAKE_PLATFORM_UBUNTU_LINUX 1)
--- a/src/pal/src/CMakeLists.txt	2018-10-13 03:07:42.000000000 +0200
+++ b/src/pal/src/CMakeLists.txt	2018-11-11 11:55:22.595649021 +0100
@@ -292,15 +292,6 @@
     find_library(INTL intl)
   endif()
 
-  # On Android, we don't need to link with gcc_s, pthread and rt
-  if(NOT CLR_CMAKE_PLATFORM_ANDROID)
-    target_link_libraries(coreclrpal
-      gcc_s
-      pthread
-      rt
-    )
-  endif()
-
   if(CLR_CMAKE_PLATFORM_ANDROID)
     find_library(LZMA NAMES lzma)
 
