--- a/CMakeLists.txt	2018-05-04 17:36:32.867992320 +0200
+++ b/CMakeLists.txt	2018-05-04 17:36:44.561970508 +0200
@@ -177,6 +177,8 @@
             set(CLR_CMAKE_TARGET_TIZEN_LINUX 1)
         elseif(CLR_CMAKE_LINUX_ID STREQUAL alpine)
             set(CLR_CMAKE_PLATFORM_ALPINE_LINUX 1)
+        elseif(CLR_CMAKE_LINUX_ID STREQUAL gentoo)
+            set(CLR_CMAKE_PLATFORM_ALPINE_LINUX 1)
         endif()
         if(CLR_CMAKE_LINUX_ID STREQUAL ubuntu)
             set(CLR_CMAKE_PLATFORM_UBUNTU_LINUX 1)
