--- a/src/Native/Unix/CMakeLists.txt	2018-12-06 16:47:25.807729229 +0100
+++ b/src/Native/Unix/CMakeLists.txt	2018-12-06 16:47:51.622698883 +0100
@@ -21,9 +21,7 @@
 add_compile_options(-fPIC)
 add_compile_options(-I${CMAKE_CURRENT_SOURCE_DIR}/Common)
 add_compile_options(-I${CMAKE_CURRENT_BINARY_DIR}/Common)
-add_compile_options(-g)
 add_compile_options(-Wthread-safety)
-add_compile_options(-Werror)
 
 if(CMAKE_SYSTEM_NAME STREQUAL Emscripten)
     set(CLR_CMAKE_PLATFORM_WASM 1)
