diff --git a/Package.swift b/Package.swift
index a42bc88..b95f87d 100644
--- a/Package.swift
+++ b/Package.swift
@@ -5,7 +5,7 @@ import PackageDescription
 let package = Package(
     name: "Operadics",
     products: [
-        .library(name: "Operadics", targets: ["Operadics"]),
+        .library(name: "Operadics", type: .dynamic, targets: ["Operadics"]),
     ],
     targets: [
         .target(name: "Operadics")
