--- a/sql-common/client_authentication.cc	2017-06-02 19:42:10.000000000 +0200
+++ b/sql-common/client_authentication.cc	2017-08-16 14:21:25.523584069 +0200
@@ -84,7 +84,7 @@
 
   if (mysql->options.extension != NULL &&
       mysql->options.extension->server_public_key_path != NULL &&
-      mysql->options.extension->server_public_key_path != '\0')
+      *mysql->options.extension->server_public_key_path != '\0')
   {
     pub_key_file= fopen(mysql->options.extension->server_public_key_path,
                         "r");
