--- a/include/my_stacktrace.h	2017-06-02 19:42:10.000000000 +0200
+++ b/include/my_stacktrace.h	2017-08-16 14:21:13.211603297 +0200
@@ -19,8 +19,8 @@
 #include <my_global.h>
 
 #ifdef TARGET_OS_LINUX
-#if defined (__x86_64__) || defined (__i386__) || \
-    (defined(__alpha__) && defined(__GNUC__))
+#if (defined (__x86_64__) || defined (__i386__) || \
+    (defined(__alpha__) && defined(__GNUC__))) && defined(__GLIBC__)
 #define HAVE_STACKTRACE 1
 #endif
 #elif defined(__WIN__) || defined(HAVE_PRINTSTACK)
