--- a/src/Documentation/Haddock.hs
+++ b/src/Documentation/Haddock.hs
@@ -34,6 +34,7 @@ module Documentation.Haddock (
 
   -- * Documentation comments
   Doc,
+  MDoc,
   DocH(..),
   Example(..),
   Hyperlink(..),
--- a/src/Documentation/Haddock.hs
+++ b/src/Documentation/Haddock.hs
@@ -16,6 +16,7 @@ module Documentation.Haddock (
   -- * Interface
   Interface(..),
   InstalledInterface(..),
+  toInstalledIface,
   createInterfaces,
   processModules,
 
