From 0b61aa6a72de7b8710857c0d078aa202e952f678 Mon Sep 17 00:00:00 2001
From: Bodigrim <andrew.lelechenko@gmail.com>
Date: Sat, 27 Apr 2019 18:02:02 +0100
Subject: [PATCH] Remove fail from Monad instance for newer GHCs

---
 Test/QuickCheck/Monadic.hs | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/Test/QuickCheck/Monadic.hs b/Test/QuickCheck/Monadic.hs
index a8f3043..1470828 100644
--- a/Test/QuickCheck/Monadic.hs
+++ b/Test/QuickCheck/Monadic.hs
@@ -123,7 +123,9 @@ instance Applicative (PropertyM m) where
 instance Monad m => Monad (PropertyM m) where
   return = pure
   (>>=) = bind
+#ifdef NO_MONADFAIL
   fail = fail_
+#endif
 
 #ifndef NO_MONADFAIL
 instance Monad m => Fail.MonadFail (PropertyM m) where
