diff --git a/Package@swift-4.swift b/Package@swift-4.swift
index 16ac0a3..7260c8a 100644
--- a/Package@swift-4.swift
+++ b/Package@swift-4.swift
@@ -6,12 +6,8 @@ let package = Package(
     products: [
         .library(name: "JSON", targets: ["JSON"]),
     ],
-    dependencies: [
-        .package(url: "https://github.com/vapor/core.git", .upToNextMajor(from: "2.1.0")),
-        .package(url: "https://github.com/vapor/node.git", .upToNextMajor(from: "2.1.0")),
-    ],
     targets: [
-        .target(name: "JSON", dependencies: ["Core", "Node"]),
+        .target(name: "JSON"),
         .testTarget(name: "JSONTests", dependencies: ["JSON"]),
     ]
-)
\ No newline at end of file
+)
