--- a/bridges/source/cpp_uno/gcc3_linux_x86-64/share.hxx	2018-09-10 22:08:33.000000000 +0200
+++ b/bridges/source/cpp_uno/gcc3_linux_x86-64/share.hxx	2018-09-19 20:00:15.937637794 +0200
@@ -113,7 +113,7 @@
 #endif
     std::type_info * exceptionType;
     void (* exceptionDestructor)(void *);
-    std::unexpected_handler unexpectedHandler;
+    void (*unexpectedHandler)();
     std::terminate_handler terminateHandler;
     __cxa_exception * nextException;
     int handlerCount;
