--- a/bridges/source/cpp_uno/gcc3_linux_aarch64/cpp2uno.cxx	2025-10-24 17:29:01.000000000 +0200
+++ b/bridges/source/cpp_uno/gcc3_linux_aarch64/cpp2uno.cxx	2026-01-01 16:32:49.329116941 +0100
@@ -583,24 +583,9 @@
 void bridges::cpp_uno::shared::VtableFactory::flushCode(
     unsigned char const * begin, unsigned char const * end)
 {
-#if !defined ANDROID && !defined MACOSX
-   static void (*clear_cache)(unsigned char const *, unsigned char const *)
-       = (void (*)(unsigned char const *, unsigned char const *)) dlsym(
-           RTLD_DEFAULT, "__clear_cache");
-   (*clear_cache)(begin, end);
-#else
-    // GCC clarified with
-    // <http://gcc.gnu.org/git/?p=gcc.git;a=commit;h=a90b0cdd444f6dde1084a439862cf507f6d3b2ae>
-    // "extend.texi (__clear_cache): Correct signature" that __builtin___clear_cache takes void*
-    // parameters, while Clang uses char* ever since
-    // <https://github.com/llvm/llvm-project/commit/c491a8d4577052bc6b3b4c72a7db6a7cfcbc2ed0> "Add
-    // support for __builtin___clear_cache in Clang" (TODO: see
-    // <https://bugs.llvm.org/show_bug.cgi?id=48489> "__builtin___clear_cache() has a different
-    // prototype than GCC"; once fixed for our Clang baseline, we can drop the reinterpret_casts):
     __builtin___clear_cache(
-        reinterpret_cast<char *>(const_cast<unsigned char *>(begin)),
-        reinterpret_cast<char *>(const_cast<unsigned char *>(end)));
-#endif
+        const_cast<unsigned char *>(begin),
+        const_cast<unsigned char *>(end));
 }
 
 /* vim:set shiftwidth=4 softtabstop=4 expandtab: */
