--- a/drivers/firmware/qcom/qcom_scm.c	2026-04-06 12:59:47.627004221 +0200
+++ b/drivers/firmware/qcom/qcom_scm.c	2026-04-06 13:00:51.569756167 +0200
@@ -2309,6 +2309,7 @@
 	{ .compatible = "microsoft,denali", },
 	{ .compatible = "microsoft,romulus13", },
 	{ .compatible = "microsoft,romulus15", },
+	{ .compatible = "microsoft,sp12" },
 	{ .compatible = "qcom,hamoa-iot-evk" },
 	{ .compatible = "qcom,sc8180x-primus" },
 	{ .compatible = "qcom,x1e001de-devkit" },
