--- a/drivers/bluetooth/btusb.c	2025-05-23 10:08:42.081878019 +0200
+++ b/drivers/bluetooth/btusb.c	2025-05-23 10:10:51.052548072 +0200
@@ -2952,6 +2952,7 @@
 	return 0;
 }
 
+#define QCA_DEBUG_ACL_HANDLE 0x2EDC
 #define QCA_MEMDUMP_ACL_HANDLE 0x2EDD
 #define QCA_MEMDUMP_SIZE_MAX  0x100000
 #define QCA_MEMDUMP_VSE_CLASS 0x01
@@ -3181,6 +3182,15 @@
 {
 	if (acl_pkt_is_dump_qca(hdev, skb))
 		return handle_dump_pkt_qca(hdev, skb);
+
+	/* We receive debug logs from chip as an ACL packets.
+	 * Instead of sending the data to ACL to decode the
+	 * received data, we are pushing them to the above layers
+	 * as a diagnostic packet.
+	 */
+	if (get_unaligned_le16(skb->data) == QCA_DEBUG_ACL_HANDLE)
+		return hci_recv_diag(hdev, skb);
+
 	return hci_recv_frame(hdev, skb);
 }
 
