--- a/drivers/net/wireless/ath/ath12k/mac.c	2025-12-12 23:24:55.872432365 +0100
+++ b/drivers/net/wireless/ath/ath12k/mac.c	2025-12-12 23:27:02.829198134 +0100
@@ -13735,11 +13735,13 @@
 	int ret, i, j;
 	u32 ht_cap = U32_MAX, antennas_rx = 0, antennas_tx = 0;
 	bool is_6ghz = false, is_raw_mode = false, is_monitor_disable = false;
-	u8 *mac_addr = NULL;
+	u8 mac_addr[ETH_ALEN] = {};
 	u8 mbssid_max_interfaces = 0;
 
 	wiphy->max_ap_assoc_sta = 0;
 
+	device_get_mac_address(ab->dev, mac_addr);
+
 	for_each_ar(ah, ar, i) {
 		u32 ht_cap_info = 0;
 
