--- a/drivers/firmware/qcom/qcom_scm.c	2025-11-30 23:42:10.000000000 +0100
+++ b/drivers/firmware/qcom/qcom_scm.c	2025-12-01 14:10:58.923205679 +0100
@@ -2009,6 +2009,8 @@
 	{ .compatible = "microsoft,blackrock" },
 	{ .compatible = "microsoft,romulus13", },
 	{ .compatible = "microsoft,romulus15", },
+	{ .compatible = "ntmer,tw220" },
+	{ .compatible = "ntmer,tw220" },
 	{ .compatible = "qcom,hamoa-iot-evk" },
 	{ .compatible = "qcom,sc8180x-primus" },
 	{ .compatible = "qcom,x1e001de-devkit" },
