diff --git a/Source/SwiftyTextTable/TextTable.swift b/Source/SwiftyTextTable/TextTable.swift
index 263dfce..b0a5659 100644
--- a/Source/SwiftyTextTable/TextTable.swift
+++ b/Source/SwiftyTextTable/TextTable.swift
@@ -8,17 +8,11 @@
 
 import Foundation
 
-#if os(macOS)
-    typealias Regex = NSRegularExpression
-#else
-    typealias Regex = RegularExpression
-#endif
-
 private let strippingPattern = "(?:\u{001B}\\[(?:[0-9]|;)+m)*(.*?)(?:\u{001B}\\[0m)+"
 
 // We can safely force try this regex because the pattern has be tested to work.
 // swiftlint:disable:next force_try
-private let strippingRegex = try! Regex(pattern: strippingPattern, options: [])
+private let strippingRegex = try! NSRegularExpression(pattern: strippingPattern, options: [])
 
 private extension String {
     func stripped() -> String {
