--- a/setup.py
+++ b/setup.py
@@ -4,30 +4,6 @@
 packages = \
 ['pytzdata', 'pytzdata.commands']
 
-package_data = \
-{'': ['*'],
- 'pytzdata': ['zoneinfo/*',
-              'zoneinfo/Africa/*',
-              'zoneinfo/America/*',
-              'zoneinfo/America/Argentina/*',
-              'zoneinfo/America/Indiana/*',
-              'zoneinfo/America/Kentucky/*',
-              'zoneinfo/America/North_Dakota/*',
-              'zoneinfo/Antarctica/*',
-              'zoneinfo/Arctic/*',
-              'zoneinfo/Asia/*',
-              'zoneinfo/Atlantic/*',
-              'zoneinfo/Australia/*',
-              'zoneinfo/Brazil/*',
-              'zoneinfo/Canada/*',
-              'zoneinfo/Chile/*',
-              'zoneinfo/Etc/*',
-              'zoneinfo/Europe/*',
-              'zoneinfo/Indian/*',
-              'zoneinfo/Mexico/*',
-              'zoneinfo/Pacific/*',
-              'zoneinfo/US/*']}
-
 setup_kwargs = {
     'name': 'pytzdata',
     'version': '2018.7',
@@ -13,7 +13,6 @@
     'author_email': 'sebastien@eustace.io',
     'url': 'https://github.com/sdispater/pytzdata',
     'packages': packages,
-    'package_data': package_data,
     'python_requires': '>=2.7, !=3.0.*, !=3.1.*, !=3.2.*, !=3.3.*',
 }
 
