diff --git a/src/qwaylandlayersurface.cpp b/src/qwaylandlayersurface.cpp
index 89f47ef..023f192 100644
--- a/src/qwaylandlayersurface.cpp
+++ b/src/qwaylandlayersurface.cpp
@@ -102,10 +102,10 @@ void QWaylandLayerSurface::zwlr_layer_surface_v1_configure(uint32_t serial, uint
 
 void QWaylandLayerSurface::attachPopup(QtWaylandClient::QWaylandShellSurface *popup)
 {
-    std::any anyRole = popup->surfaceRole();
+    void *anyRole = popup->surfaceRole();
 
-    if (auto role = std::any_cast<::xdg_popup *>(&anyRole)) {
-        get_popup(*role);
+    if (auto role = (::xdg_popup *)anyRole) {
+        get_popup(role);
     } else {
         qCWarning(LAYERSHELLQT) << "Cannot attach popup of unknown type";
     }
