--- a/drivers/remoteproc/qcom_q6v5_pas.c	2026-09-07 00:07:20.000000000 +0200
+++ b/drivers/remoteproc/qcom_q6v5_pas.c	2026-09-11 17:05:48.858145807 +0200
@@ -1249,6 +1249,7 @@
 	.lite_dtb_pas_id = 0x29,
 	.minidump_id = 5,
 	.auto_boot = true,
+	.early_boot = true,
 	.proxy_pd_names = (char*[]){
 		"lcx",
 		"lmx",
@@ -1268,6 +1269,7 @@
 	.dtb_pas_id = 0x25,
 	.minidump_id = 7,
 	.auto_boot = true,
+	.early_boot = true,
 	.proxy_pd_names = (char*[]){
 		"cx",
 		"mxc",
