--- a/Network/Socks5/Parse.hs	2019-09-15 17:24:51.161838220 +0200
+++ b/Network/Socks5/Parse.hs	2019-09-15 17:25:23.495593967 +0200
@@ -71,10 +71,11 @@
     { runParser :: forall r . ByteString -> Failure r -> Success a r -> Result r }
 
 instance Monad Parser where
-    fail errorMsg = Parser $ \buf err _ -> err buf ("failed: " ++ errorMsg)
     return v = Parser $ \buf _ ok -> ok buf v
     m >>= k = Parser $ \buf err ok ->
          runParser m buf err (\buf' a -> runParser (k a) buf' err ok)
+instance MonadFail Parser where
+    fail errorMsg = Parser $ \buf err _ -> err buf ("failed: " ++ errorMsg)
 instance MonadPlus Parser where
     mzero = fail "Parser.MonadPlus.mzero"
     mplus f g = Parser $ \buf err ok ->
