diff --git a/products/swift-bindings/CMakeLists.txt b/products/swift-bindings/CMakeLists.txt
index 14da42f..e033dab 100644
--- a/products/swift-bindings/CMakeLists.txt
+++ b/products/swift-bindings/CMakeLists.txt
@@ -4,6 +4,7 @@ set(SOURCES llbuild.swift)
 # Link C API.
 list(APPEND additional_args -import-underlying-module -lllbuild)
 list(APPEND additional_args -I ${CMAKE_CURRENT_SOURCE_DIR}/../libllbuild/public-api)
+list(APPEND additional_args -o ${CMAKE_CURRENT_BINARY_DIR}/CMakeFiles/llbuild.swift.o)
 
 # Add swift bindings target if swift compiler is present.
 if (SWIFTC_FOUND)
