--- a/ksysguardd/CMakeLists.txt	2017-07-07 15:59:48.736052099 +0200
+++ b/ksysguardd/CMakeLists.txt	2017-07-07 16:00:21.760053725 +0200
@@ -6,6 +6,7 @@
 add_definitions( -DKSYSGUARDDRCFILE="${KDE_INSTALL_FULL_SYSCONFDIR}/ksysguarddrc" )
 
 add_definitions(-DOSTYPE_${CMAKE_SYSTEM_NAME})
+add_definitions(-D_GNU_SOURCE)
 
 set(HAVE_LMSENSORS ${SENSORS_FOUND})
 
--- a/ksysguardd/Linux/netstat.c	2017-07-07 15:58:26.184048034 +0200
+++ b/ksysguardd/Linux/netstat.c	2017-07-07 15:59:01.840049790 +0200
@@ -264,10 +264,10 @@
 {
 	FILE *netstat;
 	char buffer[1024];
-	uint local_addr, local_port;
-	uint remote_addr, remote_port;
+	unsigned int local_addr, local_port;
+	unsigned int remote_addr, remote_port;
 	int uid;
-	uint state;
+	unsigned int state;
 	SocketInfo *socket_info;
 
 	if (strstr(cmd, "tcp")) {
--- a/ksysguardd/Linux/diskstat.c	2017-07-07 15:56:31.088042368 +0200
+++ b/ksysguardd/Linux/diskstat.c	2017-07-07 15:56:41.871042899 +0200
@@ -24,6 +24,7 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
+#include <strings.h>
 #include <sys/statvfs.h>
 #include <time.h>
 #include <unistd.h>
