--- a/ksysguardd/Linux/netstat.c.bak	2016-02-17 19:28:16.881110646 +0100
+++ b/ksysguardd/Linux/netstat.c	2016-02-17 19:28:42.855109098 +0100
@@ -264,10 +264,10 @@
 {
 	FILE *netstat;
 	char buffer[1024];
-	uint local_addr, local_port;
-	uint remote_addr, remote_port;
+	unsigned int local_addr, local_port;
+	unsigned int remote_addr, remote_port;
 	int uid;
-	uint state;
+	unsigned int state;
 	SocketInfo *socket_info;
 
 	if (strstr(cmd, "tcp")) {
