--- a/ghcide.cabal	2021-07-10 16:14:56.609189157 +0200
+++ b/ghcide.cabal	2021-07-10 16:24:42.506201485 +0200
@@ -32,6 +32,10 @@
     default:     False
     manual:      True
 
+flag ghcide-bench
+    description: benchmark ghcide
+    default:     False
+
 library
     default-language:   Haskell2010
     build-depends:
@@ -214,7 +218,6 @@
     hs-source-dirs: bench/hist bench/lib
     other-modules: Experiments.Types
     build-tool-depends:
-        ghcide:ghcide-bench,
         hp2pretty:hp2pretty,
         implicit-hie:gen-hie
     default-extensions:
@@ -382,7 +385,8 @@
         TypeApplications
         ViewPatterns
 
-executable ghcide-bench
+if flag(ghcide-bench)
+  executable ghcide-bench
     default-language: Haskell2010
     build-tool-depends:
         ghcide:ghcide
