--- js/src/jsinterp.cpp.bak	2015-10-09 20:01:09.138381404 +0200
+++ js/src/jsinterp.cpp	2015-10-09 20:02:01.088291159 +0200
@@ -3664,6 +3664,7 @@
 BEGIN_CASE(JSOP_LEAVEFORLETIN)
 BEGIN_CASE(JSOP_LEAVEBLOCKEXPR)
 {
+#if DEBUG
     DebugOnly<uint32_t> blockDepth = regs.fp()->blockChain().stackDepth();
 
     regs.fp()->popBlock(cx);
@@ -3683,6 +3684,7 @@
         len = JSOP_LEAVEFORLETIN_LENGTH;
         DO_NEXT_OP(len);
     }
+#endif
 }
 END_CASE(JSOP_LEAVEBLOCK)
 
