--- a/Sources/Commandant/LinuxSupport.swift	2017-01-08 12:32:31.690856006 +0100
+++ /dev/null	2017-01-05 12:04:57.287313775 +0100
@@ -1,14 +0,0 @@
-//
-//  LinuxSupport.swift
-//  Commandant
-//
-//  Created by Norio Nomura on 3/26/16.
-//  Copyright © 2016 Carthage. All rights reserved.
-//
-
-import Foundation
-
-// swift-corelibs-foundation is still written in Swift 2 API.
-#if os(Linux)
-	typealias Process = Task
-#endif
