diff --git a/Package.swift b/Package.swift
index 6cdbcbd..0d3e11c 100644
--- a/Package.swift
+++ b/Package.swift
@@ -5,7 +5,7 @@ import PackageDescription
 let package = Package(
     name: "Commandant",
     products: [
-        .library(name: "Commandant", targets: ["Commandant"]),
+        .library(name: "Commandant", type: .dynamic, targets: ["Commandant"]),
     ],
     targets: [
         .target(name: "Commandant"),
