diff --git a/hls-explicit-imports-plugin.cabal b/hls-explicit-imports-plugin.cabal
index 276c8b567d..f4b8fd0641 100644
--- a/hls-explicit-imports-plugin.cabal
+++ b/hls-explicit-imports-plugin.cabal
@@ -28,22 +28,6 @@ library
     , text
     , unordered-containers
 
-  if impl(ghc < 8.10.5)
-    build-depends:
-      ghc-api-compat ==8.6
-  elif impl(ghc == 8.10.5)
-    build-depends:
-      ghc-api-compat ==8.10.5
-  elif impl(ghc == 8.10.6)
-    build-depends:
-      ghc-api-compat ==8.10.6
-  elif impl(ghc == 8.10.7)
-    build-depends:
-      ghc-api-compat ==8.10.7
-  elif impl(ghc == 9.0.1)
-    build-depends:
-      ghc-api-compat ==9.0.1
-
   default-language:   Haskell2010
   default-extensions:
     DataKinds
diff --git a/src/Ide/Plugin/ExplicitImports.hs b/src/Ide/Plugin/ExplicitImports.hs
index 4194a79e27..d094c197d0 100644
--- a/src/Ide/Plugin/ExplicitImports.hs
+++ b/src/Ide/Plugin/ExplicitImports.hs
@@ -35,16 +35,6 @@ import           Ide.PluginUtils                      (mkLspCommand)
 import           Ide.Types
 import           Language.LSP.Server
 import           Language.LSP.Types
-#if MIN_VERSION_ghc(9,0,0)
-import           GHC.Builtin.Names                    (pRELUDE)
-#else
-import           PrelNames                            (pRELUDE)
-#endif
-import           RnNames                              (findImportUsage,
-                                                       getMinimalImports)
-import qualified SrcLoc
-import           TcRnMonad                            (initTcWithGbl)
-import           TcRnTypes                            (TcGblEnv (tcg_used_gres))
 
 importCommandId :: CommandId
 importCommandId = "ImportLensCommand"
@@ -197,13 +187,13 @@ minimalImportsRule = define $ \MinimalImports nfp -> do
   (imports, mbMinImports) <- liftIO $ extractMinimalImports hsc tmr
   let importsMap =
         Map.fromList
-          [ (SrcLoc.realSrcSpanStart l, T.pack (prettyPrint i))
-            | L (OldRealSrcSpan l) i <- fromMaybe [] mbMinImports
+          [ (realSrcSpanStart l, T.pack (prettyPrint i))
+            | L (RealSrcSpan l _) i <- fromMaybe [] mbMinImports
           ]
       res =
-        [ (i, Map.lookup (SrcLoc.realSrcSpanStart l) importsMap)
+        [ (i, Map.lookup (realSrcSpanStart l) importsMap)
           | i <- imports
-          , OldRealSrcSpan l <- [getLoc i]
+          , RealSrcSpan l _ <- [getLoc i]
         ]
   return ([], MinimalImportsResult res <$ mbMinImports)
 
@@ -240,7 +230,7 @@ mkExplicitEdit pred posMapping (L src imp) explicit
   | ImportDecl {ideclHiding = Just (False, _)} <- imp =
     Nothing
   | not (isQualifiedImport imp),
-    OldRealSrcSpan l <- src,
+    RealSrcSpan l _ <- src,
     L _ mn <- ideclName imp,
     -- (almost) no one wants to see an explicit import list for Prelude
     pred mn,
