--- buildlib.bak	2016-02-04 20:08:15.655639965 +0100
+++ buildlib	2016-02-04 20:08:23.880966051 +0100
@@ -31,7 +31,7 @@
 LIB=libBlocksRuntime.a
 SRC=BlocksRuntime
 if test -f $LIB; then rm $LIB; fi
-"$CC" -c $CFLAGS -o $SRC/data.o $SRC/data.c \
-&& "$CC" -c $CFLAGS -o $SRC/runtime.o -I . $SRC/runtime.c \
+$CC -c $CFLAGS -o $SRC/data.o $SRC/data.c \
+&& $CC -c $CFLAGS -o $SRC/runtime.o -I . $SRC/runtime.c \
 && "$AR" cr $LIB $SRC/data.o $SRC/runtime.o \
 && "$RANLIB" $LIB
