--- a/source3/Makefile.in	2016-10-16 13:53:20.578598189 +0200
+++ b/source3/Makefile.in	2016-10-16 13:53:34.913597009 +0200
@@ -2050,7 +2050,7 @@
 
 $(LIBTALLOC_STATIC_TARGET): $(BINARY_PREREQS) $(LIBTALLOC_OBJ0)
 	@echo Linking non-shared library $@
-	@-$(AR) -rc $@ $(LIBTALLOC_OBJ0)
+	@-$(AR) rc $@ $(LIBTALLOC_OBJ0)
 libtalloc: $(LIBTALLOC)
 
 cleanlibtalloc::
@@ -2141,7 +2141,7 @@
 
 $(LIBTEVENT_STATIC_TARGET): $(BINARY_PREREQS) $(LIBTEVENT_OBJ0)
 	@echo Linking non-shared library $@
-	@-$(AR) -rc $@ $(LIBTEVENT_OBJ0)
+	@-$(AR) rc $@ $(LIBTEVENT_OBJ0)
 libtevent: $(LIBTEVENT)
 
 cleanlibtevent::
@@ -2234,7 +2234,7 @@
 
 $(LIBTDB_STATIC_TARGET): $(BINARY_PREREQS) $(LIBTDB_OBJ0)
 	@echo Linking non-shared library $@
-	@-$(AR) -rc $@ $(LIBTDB_OBJ0)
+	@-$(AR) rc $@ $(LIBTDB_OBJ0)
 
 libtdb: $(LIBTDB)
 
@@ -2333,7 +2333,7 @@
 
 $(LIBWBCLIENT_STATIC_TARGET): $(BINARY_PREREQS) $(LIBWBCLIENT_OBJ0) $(WBCOMMON_OBJ)
 	@echo Linking non-shared library $@
-	@-$(AR) -rc $@ $(LIBWBCLIENT_OBJ0) $(WBCOMMON_OBJ)
+	@-$(AR) rc $@ $(LIBWBCLIENT_OBJ0) $(WBCOMMON_OBJ)
 
 libwbclient: $(LIBWBCLIENT)
 
@@ -2412,7 +2412,7 @@
 
 $(LIBADDNS_STATIC_TARGET): $(BINARY_PREREQS) $(LIBADDNS_OBJ0)
 	@echo Linking non-shared library $@
-	@-$(AR) -rc $@ $(LIBADDNS_OBJ0)
+	@-$(AR) rc $@ $(LIBADDNS_OBJ0)
 
 libaddns: $(LIBADDNS)
 
@@ -2524,7 +2524,7 @@
 
 $(LIBNETAPI_STATIC_TARGET): $(BINARY_PREREQS) $(LIBNETAPI_OBJ0)
 	@echo Linking non-shared library $@
-	@-$(AR) -rc $@ $(LIBNETAPI_OBJ0)
+	@-$(AR) rc $@ $(LIBNETAPI_OBJ0)
 
 libnetapi: $(LIBNETAPI)
 
@@ -2605,7 +2605,7 @@
 
 $(LIBSMBCLIENT_STATIC_TARGET): $(BINARY_PREREQS) $(LIBSMBCLIENT_OBJ1)
 	@echo Linking non-shared library $@
-	@-$(AR) -rc $@ $(LIBSMBCLIENT_OBJ1)
+	@-$(AR) rc $@ $(LIBSMBCLIENT_OBJ1)
 
 libsmbclient: $(LIBSMBCLIENT)
 
@@ -2692,7 +2692,7 @@
 
 $(LIBSMBSHAREMODES_STATIC_TARGET): $(BINARY_PREREQS) $(LIBSMBSHAREMODES_OBJ0)
 	@echo Linking non-shared library $@
-	@-$(AR) -rc $@ $(LIBSMBSHAREMODES_OBJ0)
+	@-$(AR) rc $@ $(LIBSMBSHAREMODES_OBJ0)
 
 libsmbsharemodes: $(LIBSMBSHAREMODES)
 
