diff --git a/Source/SourceKittenFramework/LinuxCompatibility.swift b/Source/SourceKittenFramework/LinuxCompatibility.swift
index cc71b53..ed111e5 100644
--- a/Source/SourceKittenFramework/LinuxCompatibility.swift
+++ b/Source/SourceKittenFramework/LinuxCompatibility.swift
@@ -9,21 +9,19 @@
 import Foundation
 
 #if os(Linux)
-public typealias Process = Task
-public typealias NSRegularExpression = RegularExpression
 
 extension CharacterSet {
     public func bridge() -> NSCharacterSet {
         return _bridgeToObjectiveC()
     }
 }
-extension TextCheckingResult {
+extension NSTextCheckingResult {
     public func rangeAt(_ index: Int) -> NSRange {
         return range(at: index)
     }
 }
 extension NSString {
-    public var isAbsolutePath: Bool { return absolutePath }
+    public var isAbsolutePath: Bool { return hasPrefix("~") || hasPrefix("/") }
 }
 extension Dictionary {
     public func bridge() -> NSDictionary {
