diff --git a/Package@swift-4.swift b/Package@swift-4.swift
index 4f92135..9b03e87 100644
--- a/Package@swift-4.swift
+++ b/Package@swift-4.swift
@@ -5,29 +5,17 @@ let package = Package(
     name: "SourceKitten",
     products: [
         .executable(name: "sourcekitten", targets: ["sourcekitten"]),
-        .library(name: "SourceKittenFramework", targets: ["SourceKittenFramework"])
-    ],
-    dependencies: [
-        .package(url: "https://github.com/Carthage/Commandant.git", from: "0.12.0"),
-        .package(url: "https://github.com/drmohundro/SWXMLHash.git", from: "4.1.0"),
-        .package(url: "https://github.com/jpsim/Yams.git", from: "0.3.3"),
-        .package(url: "https://github.com/norio-nomura/Clang_C.git", from: "1.0.0"),
-        .package(url: "https://github.com/norio-nomura/SourceKit.git", from: "1.0.0"),
+        .library(name: "SourceKittenFramework", type: .dynamic, targets: ["SourceKittenFramework"])
     ],
     targets: [
         .target(
             name: "sourcekitten",
             dependencies: [
-                "Commandant",
                 "SourceKittenFramework",
             ]
         ),
         .target(
             name: "SourceKittenFramework",
-            dependencies: [
-                "SWXMLHash",
-                "Yams",
-            ],
             exclude: [
                 "clang-c",
                 "sourcekitd.h",
