diff --git a/Source/SourceKittenFramework/LinuxCompatibility.swift b/Source/SourceKittenFramework/LinuxCompatibility.swift
index cc71b53..c2a7a73 100644
--- a/Source/SourceKittenFramework/LinuxCompatibility.swift
+++ b/Source/SourceKittenFramework/LinuxCompatibility.swift
@@ -10,7 +10,6 @@ import Foundation
 
 #if os(Linux)
 public typealias Process = Task
-public typealias NSRegularExpression = RegularExpression
 
 extension CharacterSet {
     public func bridge() -> NSCharacterSet {
