diff --git a/Source/sourcekitten/IndexCommand.swift b/Source/sourcekitten/IndexCommand.swift
index c2a50d7..3fc389f 100644
--- a/Source/sourcekitten/IndexCommand.swift
+++ b/Source/sourcekitten/IndexCommand.swift
@@ -36,6 +36,9 @@ struct IndexCommand: CommandProtocol {
         guard !options.file.isEmpty else {
             return .failure(.invalidArgument(description: "file must be set when calling index"))
         }
+        guard !options.compilerargs.isEmpty else {
+            return .failure(.invalidArgument(description: "compilerargs must be set when calling index"))
+        }
         let absoluteFile = options.file.bridge().absolutePathRepresentation()
         let request = Request.index(file: absoluteFile, arguments: options.compilerargs.components(separatedBy: " "))
         print(toJSON(toNSDictionary(request.send())))
