diff --git a/Source/sourcekitten/CompleteCommand.swift b/Source/sourcekitten/CompleteCommand.swift
index 4aa1358..6c7bd40 100644
--- a/Source/sourcekitten/CompleteCommand.swift
+++ b/Source/sourcekitten/CompleteCommand.swift
@@ -55,9 +55,6 @@ struct CompleteCommand: CommandProtocol {
         var args: [String]
         if options.spmModule.isEmpty {
             args = ["-c", path] + options.compilerargs
-            if args.index(of: "-sdk") == nil {
-                args.append(contentsOf: ["-sdk", sdkPath()])
-            }
         } else {
             guard let module = Module(spmName: options.spmModule) else {
                 return .failure(.invalidArgument(description: "Bad module name"))
