diff --git a/Package.swift b/Package.swift
index 2f7e0a5..9b03e87 100644
--- a/Package.swift
+++ b/Package.swift
@@ -5,7 +5,7 @@ let package = Package(
     name: "SourceKitten",
     products: [
         .executable(name: "sourcekitten", targets: ["sourcekitten"]),
-        .library(name: "SourceKittenFramework", targets: ["SourceKittenFramework"])
+        .library(name: "SourceKittenFramework", type: .dynamic, targets: ["SourceKittenFramework"])
     ],
     targets: [
         .target(
