--- src/polkitbackendlocalauthorizationstore.c	2013-05-07 18:23:47.000000000 +0000
+++ src/polkitbackendlocalauthorizationstore.c.new	2015-08-19 17:17:06.558002729 +0000
@@ -776,6 +776,7 @@
 		  && g_pattern_match_string ((GPatternSpec *) ll->data, identity_string))
 		break;
 	    }
+#if defined HAVE_GETNETGRENT
 
 	  /* if no identity specs matched and identity is a user, match against netgroups */
 	  if (ll == NULL && POLKIT_IS_UNIX_USER (identity))
@@ -791,6 +792,7 @@
                 break;
 		}
 	    }
+#endif
 	}
 
       if (ll == NULL)
