--- a/CMakeLists.txt	2022-02-08 15:07:48.035754728 +0100
+++ b/CMakeLists.txt	2022-02-08 15:09:58.784410706 +0100
@@ -65,24 +65,27 @@
                   SOVERSION 9
 )
 
-find_package(X11)
-set_package_properties(X11 PROPERTIES DESCRIPTION "X11 libraries"
-                        URL "https://www.x.org"
-                        TYPE OPTIONAL
-                        PURPOSE "Required for building the X11 based workspace"
-                    )
-if(X11_FOUND)
+
+option(WITH_X11 "Build with X11 integration" ON)
+if(WITH_X11)
+    find_package(X11 REQUIRED)
+    set_package_properties(X11 PROPERTIES DESCRIPTION "X11 libraries"
+                            URL "https://www.x.org"
+                            TYPE OPTIONAL
+                            PURPOSE "Required for building the X11 based workspace"
+                        )
+
     if (QT_MAJOR_VERSION STREQUAL "5")
         find_package(Qt5X11Extras REQUIRED)
     elseif (QT_MAJOR_VERSION STREQUAL "6")
         find_package(Qt${QT_MAJOR_VERSION} ${REQUIRED_QT_VERSION} CONFIG REQUIRED Gui) # qtx11extras_p.h
     endif()
-    set(HAVE_X11 1)
+    set(HAVE_X11 TRUE)
     find_library(X11_XRes_LIB XRes ${X11_LIB_SEARCH_PATH})
     find_path(X11_XRes_INCLUDE_PATH X11/extensions/XRes.h ${X11_INC_SEARCH_PATH})
 
     if(X11_XRes_LIB AND X11_XRes_INCLUDE_PATH)
-        set(X11_XRes_FOUND TRUE)
+        set(HAVE_XRES TRUE)
     endif()
 endif()
 
@@ -116,8 +119,6 @@
         PURPOSE "Used for reading hardware sensors")
 endif()
 
-set(HAVE_X11 ${X11_FOUND})
-set(HAVE_XRES ${X11_XRes_FOUND})
 configure_file(config-ksysguard.h.cmake ${CMAKE_CURRENT_BINARY_DIR}/config-ksysguard.h )
 
 option(BUILD_DESIGNERPLUGIN "Build plugin for Qt Designer" ON)
@@ -148,7 +149,7 @@
 add_subdirectory( systemstats )
 
 if(BUILD_TESTING)
-    add_subdirectory( autotests )
+#DONOTCOMPILE     add_subdirectory( autotests )
 endif()
 
 # add clang-format target for all our real source files
--- a/processui/CMakeLists.txt	2022-02-08 15:12:03.639130880 +0100
+++ b/processui/CMakeLists.txt	2022-02-08 15:12:47.829766937 +0100
@@ -26,7 +26,7 @@
 add_library(processui ${processui_LIB_SRCS})
 add_library(KSysGuard::ProcessUi ALIAS processui)
 
-if(X11_XRes_FOUND)
+if(HAVE_XRES)
     target_link_libraries(processui PRIVATE ${X11_XRes_LIB} ${X11_LIBRARIES})
     include_directories(${X11_XRes_INCLUDE_PATH})
 endif()
@@ -59,7 +59,7 @@
         "$<BUILD_INTERFACE:${libksysguard_SOURCE_DIR}>"
         "$<INSTALL_INTERFACE:${KDE_INSTALL_INCLUDEDIR}/ksysguard>"
 )
-if(X11_FOUND)
+if(HAVE_X11)
     if (QT_MAJOR_VERSION STREQUAL "5")
        target_link_libraries(processui PRIVATE Qt::X11Extras KF5::WindowSystem)
     else() 
