--- a/CMakeLists.txt	2019-12-07 21:21:51.413816716 +0100
+++ b/CMakeLists.txt	2019-12-07 21:22:15.575905321 +0100
@@ -25,7 +25,7 @@
 include(CheckLibraryExists)
 include(FeatureSummary)
 
-find_package(Qt5 ${QT_MIN_VERSION} REQUIRED CONFIG COMPONENTS DBus Network Widgets WebChannel)
+find_package(Qt5 ${QT_MIN_VERSION} REQUIRED CONFIG COMPONENTS DBus Network Widgets)
 find_package(Qt5WebEngineWidgets ${QT_MIN_VERSION} CONFIG)
 set_package_properties(Qt5WebEngineWidgets PROPERTIES
                        URL "git://code.qt.org/qt/qtwebenginewidgets.git"
--- a/processui/CMakeLists.txt	2019-12-07 21:22:26.595945713 +0100
+++ b/processui/CMakeLists.txt	2019-12-07 21:22:35.584978648 +0100
@@ -37,7 +37,6 @@
         KF5::ConfigCore
     PRIVATE
         Qt5::DBus
-        Qt5::WebChannel
         KF5::I18n
         KF5::WindowSystem
         KF5::AuthCore
