--- a/Source/cmTarget.cxx	2022-08-17 18:54:57.000000000 +0200
+++ b/Source/cmTarget.cxx	2022-08-26 11:03:50.860378428 +0200
@@ -709,6 +709,7 @@
     initProp("ENABLE_EXPORTS");
   }
   if (this->impl->TargetType == cmStateEnums::SHARED_LIBRARY ||
+      this->impl->TargetType == cmStateEnums::STATIC_LIBRARY ||
       this->impl->TargetType == cmStateEnums::MODULE_LIBRARY) {
     this->SetProperty("POSITION_INDEPENDENT_CODE", "True");
   } else if (this->CanCompileSources()) {
