--- src/Makefile.in.bak	2015-08-17 20:31:07.737722579 +0000
+++ src/Makefile.in	2015-08-17 20:31:18.409690799 +0000
@@ -257,7 +257,7 @@
 #
 LIBVERSION = 1:0:1
 AM_CPPFLAGS = ${regular_CPPFLAGS} -I${top_srcdir}/include
-AM_CFLAGS = ${regular_CFLAGS} ${GCC_FVISIBILITY_HIDDEN}
+AM_CFLAGS = ${regular_CFLAGS}
 lib_LTLIBRARIES = libmnl.la
 libmnl_la_LDFLAGS = -Wl,--version-script=$(srcdir)/libmnl.map -version-info $(LIBVERSION)
 libmnl_la_SOURCES = socket.c callback.c nlmsg.c attr.c internal.h libmnl.map
