diff --git a/Sources/Cloud/Commands/Playground.swift b/Sources/Cloud/Commands/Playground.swift
index 2284012..0c3de69 100644
--- a/Sources/Cloud/Commands/Playground.swift
+++ b/Sources/Cloud/Commands/Playground.swift
@@ -588,11 +588,7 @@ extension FileManager {
     func createVaporConfigFolderIfNecessary() throws {
         var isDirectory: ObjCBool = false
         _ = fileExists(atPath: vaporConfigDir, isDirectory: &isDirectory)
-        #if os(Linux)
-            guard !isDirectory else { return }
-        #else
-            guard !isDirectory.boolValue else { return }
-        #endif
+        guard !isDirectory.boolValue else { return }
         try createDirectory(atPath: vaporConfigDir, withIntermediateDirectories: true)
     }
 }
diff --git a/Sources/Cloud/Redis.swift b/Sources/Cloud/Redis.swift
index f737385..fcaf36b 100644
--- a/Sources/Cloud/Redis.swift
+++ b/Sources/Cloud/Redis.swift
@@ -287,11 +287,7 @@ public final class CloudRedis {
                 else { return }
             
             guard log != "EXIT!" else {
-                do {
-                    exit(0)
-                } catch {
-                    exit(1)
-                }
+                exit(0)
             }
             
             console.print(log)
diff --git a/Sources/VaporToolbox/Config.swift b/Sources/VaporToolbox/Config.swift
index bc453a4..76e40ae 100644
--- a/Sources/VaporToolbox/Config.swift
+++ b/Sources/VaporToolbox/Config.swift
@@ -87,11 +87,7 @@ extension FileManager {
     private func isDirectory(path: String) -> Bool {
         var isDirectory: ObjCBool = false
         _ = fileExists(atPath: path, isDirectory: &isDirectory)
-        #if os(Linux)
-            return isDirectory
-        #else
-            return isDirectory.boolValue
-        #endif
+        return isDirectory.boolValue
     }
 }
 
