diff --git a/Package.swift b/Package.swift
index b5e6caa..76c2b0b 100644
--- a/Package.swift
+++ b/Package.swift
@@ -5,10 +5,5 @@ let package = Package(
     targets: [
         Target(name: "VaporToolbox"),
         Target(name: "Executable", dependencies: ["VaporToolbox"])
-    ],
-    dependencies: [
-        // Console protocols, terminal, and commands
-        .Package(url: "https://github.com/vapor/console.git", majorVersion: 2),
-        .Package(url: "https://github.com/vapor/json.git", majorVersion: 2)
     ]
 )
