--- a/SuiteSparse_config/cmake_modules/SuiteSparseBLAS.cmake	2025-11-05 16:33:17.000000000 +0100
+++ b/SuiteSparse_config/cmake_modules/SuiteSparseBLAS.cmake	2025-11-29 20:43:18.091746598 +0100
@@ -109,10 +109,9 @@
 # setting is ignored).
 
 if ( SUITESPARSE_USE_64BIT_BLAS )
-
-    # Look for Intel MKL BLAS with 64-bit integers
-    message ( STATUS "Looking for Intel 64-bit BLAS" )
-    set ( BLA_VENDOR Intel10_64ilp )
+    # Look for FlexiBLAS with 64-bit integers
+    message ( STATUS "Looking for FlexiBLAS" )
+    set ( BLA_VENDOR FlexiBLAS )
     set ( BLA_SIZEOF_INTEGER 8 )
     find_package ( BLAS )
     if ( BLAS_FOUND )
@@ -121,46 +120,6 @@
         return ( )
     endif ( )
 
-    # Look for ARM BLAS with 64-bit integers
-    message ( STATUS "Looking for ARM 64-bit BLAS" )
-    set ( BLA_VENDOR Arm_ilp64_mp )
-    set ( BLA_SIZEOF_INTEGER 8 )
-    find_package ( BLAS )
-    if ( BLAS_FOUND )
-        include ( SuiteSparseBLAS64 )
-        return ( )
-    endif ( )
-
-    # Look for IBM BLAS with 64-bit integers
-    message ( STATUS "Looking for IBM ESSL 64-bit BLAS" )
-    set ( BLA_VENDOR IBMESSL_SMP )
-    set ( BLA_SIZEOF_INTEGER 8 )
-    find_package ( BLAS )
-    if ( BLAS_FOUND )
-        include ( SuiteSparseBLAS64 )
-        return ( )
-    endif ( )
-
-    # Look for OpenBLAS with 64-bit integers
-    message ( STATUS "Looking for 64-bit OpenBLAS" )
-    set ( BLA_VENDOR OpenBLAS )
-    set ( BLA_SIZEOF_INTEGER 8 )
-    find_package ( BLAS )
-    if ( BLAS_FOUND )
-        include ( SuiteSparseBLAS64 )
-        return ( )
-    endif ( )
-
-    # Look for any 64-bit BLAS
-    unset ( BLA_VENDOR )
-    message ( STATUS "Looking for any 64-bit BLAS" )
-    set ( BLA_SIZEOF_INTEGER 8 )
-    find_package ( BLAS )
-    if ( BLAS_FOUND )
-        include ( SuiteSparseBLAS64 )
-        return ( )
-    endif ( )
-
     # report an error if strict
     if ( SUITESPARSE_USE_STRICT )
         message ( FATAL_ERROR "64-bit BLAS required, but not found" )
