--- a/Source/WebCore/platform/PlatformPasteboard.h	2016-10-01 20:32:00.235819478 +0200
+++ b/Source/WebCore/platform/PlatformPasteboard.h	2016-10-01 20:32:37.084816446 +0200
@@ -30,6 +30,7 @@
 #include <wtf/RefCounted.h>
 #include <wtf/RetainPtr.h>
 #include <wtf/Vector.h>
+#include <functional>
 
 #if PLATFORM(MAC)
 OBJC_CLASS NSPasteboard;
--- a/Source/WebCore/platform/gtk/PasteboardHelper.h	2016-10-01 20:31:10.160823598 +0200
+++ b/Source/WebCore/platform/gtk/PasteboardHelper.h	2016-10-01 20:31:25.284822354 +0200
@@ -29,6 +29,7 @@
 #include <wtf/Noncopyable.h>
 #include <wtf/Vector.h>
 #include <wtf/glib/GRefPtr.h>
+#include <functional>
 
 namespace WebCore {
 
