diff --git a/src/polkitbackend/polkitbackendinteractiveauthority.c b/src/polkitbackend/polkitbackendinteractiveauthority.c
index 3bd2f0b..9d2ec61 100644
--- a/src/polkitbackend/polkitbackendinteractiveauthority.c
+++ b/src/polkitbackend/polkitbackendinteractiveauthority.c
@@ -2103,6 +2103,7 @@ get_users_in_group (PolkitIdentity                    *group,
   return ret;
 }
 
+#if defined HAVE_GETNETGRENT
 static GList *
 get_users_in_net_group (PolkitIdentity                    *group,
                         gboolean                           include_root)
@@ -2154,6 +2155,7 @@ get_users_in_net_group (PolkitIdentity                    *group,
   endnetgrent ();
   return ret;
 }
+#endif
 
 /* ---------------------------------------------------------------------------------------------------- */
 
@@ -2243,10 +2245,12 @@ authentication_agent_initiate_challenge (AuthenticationAgent         *agent,
         {
           user_identities = g_list_concat (user_identities, get_users_in_group (identity, FALSE));
         }
+#if defined HAVE_GETNETGRENT
       else if (POLKIT_IS_UNIX_NETGROUP (identity))
         {
           user_identities =  g_list_concat (user_identities, get_users_in_net_group (identity, FALSE));
         }
+#endif
       else
         {
           g_warning ("Unsupported identity");
diff --git a/src/polkitbackend/polkitbackendjsauthority.cpp b/src/polkitbackend/polkitbackendjsauthority.cpp
index bc2fe22..b84c110 100644
--- a/src/polkitbackend/polkitbackendjsauthority.cpp
+++ b/src/polkitbackend/polkitbackendjsauthority.cpp
@@ -33,6 +33,7 @@
 #include <glib/gstdio.h>
 #include <locale.h>
 #include <glib/gi18n-lib.h>
+#include <sys/wait.h>
 
 #include <polkit/polkit.h>
 #include "polkitbackendjsauthority.h"
@@ -1493,12 +1494,16 @@
 {
   /* PolkitBackendJsAuthority *authority = POLKIT_BACKEND_JS_AUTHORITY (JS_GetContextPrivate (cx)); */
   bool ret = false;
-  char *user;
-  char *netgroup;
   bool is_in_netgroup = false;
 
   JS::CallArgs args = JS::CallArgsFromVp (argc, vp);
 
+
+#if defined(HAVE_INNETGR)
+  {
+  char *netgroup;
+  char *user;
+
   user = JS_EncodeString (cx, args[0].toString());
   netgroup = JS_EncodeString (cx, args[1].toString());
 
@@ -1512,6 +1517,8 @@
 
   JS_free (cx, netgroup);
   JS_free (cx, user);
+  }
+#endif
 
   ret = true;
 
