diff --git a/Package@swift-4.swift b/Package@swift-4.swift
index fe30082..d32b55e 100644
--- a/Package@swift-4.swift
+++ b/Package@swift-4.swift
@@ -6,12 +6,8 @@ let package = Package(
     products: [
         .library(name: "Random", targets: ["Random"]),
     ],
-    dependencies: [
-        // Core extensions, type-aliases, and functions that facilitate common tasks.
-        .package(url: "https://github.com/vapor/core.git", .upToNextMajor(from: "2.1.1")),
-    ],
     targets: [
-        .target(name: "Random", dependencies: ["Core"]),
+        .target(name: "Random"),
         .testTarget(name: "RandomTests", dependencies: ["Random"]),
     ]
 )
