diff --git a/Package.swift b/Package.swift
index 9aa9ca1..4aff7c3 100644
--- a/Package.swift
+++ b/Package.swift
@@ -3,9 +3,5 @@
 import PackageDescription
 
 let package = Package(
-    name: "Random",
-    dependencies: [
-        // Core extensions, type-aliases, and functions that facilitate common tasks.
-        .Package(url: "https://github.com/vapor/core.git", Version(2,0,0, prereleaseIdentifiers: ["beta"]))
-    ]
+    name: "Random"
 )
