--- a/include/string.h	2016-08-20 09:55:57.497735569 +0200
+++ b/include/string.h	2016-08-20 09:56:26.021655839 +0200
@@ -85,7 +85,19 @@
 #endif
 
 #ifdef _GNU_SOURCE
-#define	strdupa(x)	strcpy(alloca(strlen(x)+1),x)
+#ifdef __GNUC__
+#define strdupa(s) \
+       (__extension__ ({const char *__in = (s); \
+                        size_t __len = strlen (__in) + 1; \
+                        char * __out = (char *) __builtin_alloca (__len); \
+                        (char *) memcpy (__out, __in, __len);}))
+#define strndupa(s, n) \
+       (__extension__ ({const char *__in = (s); \
+                        size_t __len = strnlen (__in, (n)) + 1; \
+                        char *__out = (char *) __builtin_alloca (__len); \
+                        __out[__len-1] = '\0'; \
+                        (char *) memcpy (__out, __in, __len-1);}))
+#endif
 int strverscmp (const char *, const char *);
 int strcasecmp_l (const char *, const char *, locale_t);
 int strncasecmp_l (const char *, const char *, size_t, locale_t);
