--- a/include/netdb.h	2018-06-22 22:53:17.823318130 +0200
+++ b/include/netdb.h	2018-06-22 22:53:47.749266653 +0200
@@ -133,6 +133,8 @@
 int gethostbyaddr_r(const void *, socklen_t, int, struct hostent *, char *, size_t, struct hostent **, int *);
 int getservbyport_r(int, const char *, struct servent *, char *, size_t, struct servent **);
 int getservbyname_r(const char *, const char *, struct servent *, char *, size_t, struct servent **);
+#define NETDB_INTERNAL  -1
+#define NETDB_SUCCESS  0
 #define EAI_NODATA     -5
 #define EAI_ADDRFAMILY -9
 #define EAI_INPROGRESS -100
