--- a/include/locale.h	2016-11-09 17:41:21.660739755 +0100
+++ b/include/locale.h	2016-11-09 17:43:39.125081341 +0100
@@ -7,7 +7,9 @@
 
 #include <features.h>
 
-#ifdef __cplusplus
+#if __cplusplus >= 201103L
+#define NULL __null
+#elif __cplusplus
 #define NULL 0L
 #else
 #define NULL ((void*)0)
--- a/include/string.h	2016-11-09 17:41:21.660739755 +0100
+++ b/include/string.h	2016-11-09 17:43:39.125081341 +0100
@@ -7,7 +7,9 @@
 
 #include <features.h>
 
-#ifdef __cplusplus
+#if __cplusplus >= 201103L
+#define NULL __null
+#elif __cplusplus
 #define NULL 0L
 #else
 #define NULL ((void*)0)
--- a/include/stddef.h	2016-11-09 17:41:21.660739755 +0100
+++ b/include/stddef.h	2016-11-09 17:43:39.125081341 +0100
@@ -7,7 +7,9 @@
 
 #include <features.h>
 
-#ifdef __cplusplus
+#if __cplusplus >= 201103L
+#define NULL __null
+#elif __cplusplus
 #define NULL 0L
 #else
 #define NULL ((void*)0)
--- a/include/stdio.h	2016-11-09 17:41:21.660739755 +0100
+++ b/include/stdio.h	2016-11-09 17:43:39.125081341 +0100
@@ -7,7 +7,9 @@
 
 #include <features.h>
 
-#ifdef __cplusplus
+#if __cplusplus >= 201103L
+#define NULL __null
+#elif __cplusplus
 #define NULL 0L
 #else
 #define NULL ((void*)0)
--- a/include/stdlib.h	2016-11-09 17:41:21.660739755 +0100
+++ b/include/stdlib.h	2016-11-09 17:43:39.125081341 +0100
@@ -7,7 +7,9 @@
 
 #include <features.h>
 
-#ifdef __cplusplus
+#if __cplusplus >= 201103L
+#define NULL __null
+#elif __cplusplus
 #define NULL 0L
 #else
 #define NULL ((void*)0)
--- a/include/time.h	2016-11-09 17:41:21.660739755 +0100
+++ b/include/time.h	2016-11-09 17:43:39.125081341 +0100
@@ -7,7 +7,9 @@
 
 #include <features.h>
 
-#ifdef __cplusplus
+#if __cplusplus >= 201103L
+#define NULL __null
+#elif __cplusplus
 #define NULL 0L
 #else
 #define NULL ((void*)0)
--- a/include/unistd.h	2016-11-09 17:41:21.660739755 +0100
+++ b/include/unistd.h	2016-11-09 17:43:39.125081341 +0100
@@ -7,7 +7,9 @@
 
 #include <features.h>
 
-#ifdef __cplusplus
+#if __cplusplus >= 201103L
+#define NULL __null
+#elif __cplusplus
 #define NULL 0L
 #else
 #define NULL ((void*)0)
--- a/include/wchar.h	2016-11-09 17:41:21.660739755 +0100
+++ b/include/wchar.h	2016-11-09 17:43:39.125081341 +0100
@@ -7,7 +7,9 @@
 
 #include <features.h>
 
-#ifdef __cplusplus
+#if __cplusplus >= 201103L
+#define NULL __null
+#elif __cplusplus
 #define NULL 0L
 #else
 #define NULL ((void*)0)
