--- a/src/systemd/src/basic/parse-util.h	2016-08-08 10:28:12.377905000 +0200
+++ b/src/systemd/src/basic/parse-util.h	2016-08-08 10:28:31.327904472 +0200
@@ -107,3 +107,12 @@
 int parse_fractional_part_u(const char **s, size_t digits, unsigned *res);
 
 int parse_percent(const char *p);
+
+#define strndupa(s, n) \
+  ({ \
+    const char *__old = (s); \
+    size_t __len = strnlen(__old, (n)); \
+    char *__new = (char *)alloca(__len + 1); \
+    __new[__len] = '\0'; \
+    (char *)memcpy(__new, __old, __len); \
+  })
