--- a/configure	2016-10-16 15:14:23.641198041 +0200
+++ b/configure	2016-10-16 15:14:52.094195699 +0200
@@ -6592,7 +6592,7 @@
 
 # Prefer gcc-* variants; the ones libtool would choose don't work with LTO
 if test -n "$ac_tool_prefix"; then
-  for ac_prog in gcc-ar ar
+  for ac_prog in  ar
   do
     # Extract the first word of "$ac_tool_prefix$ac_prog", so it can be a program name with args.
 set dummy $ac_tool_prefix$ac_prog; ac_word=$2
@@ -6636,7 +6636,7 @@
 fi
 if test -z "$AR"; then
   ac_ct_AR=$AR
-  for ac_prog in gcc-ar ar
+  for ac_prog in  ar
 do
   # Extract the first word of "$ac_prog", so it can be a program name with args.
 set dummy $ac_prog; ac_word=$2
@@ -6692,7 +6692,7 @@
 fi
 
 if test -n "$ac_tool_prefix"; then
-  for ac_prog in gcc-ranlib ranlib
+  for ac_prog in  ranlib
   do
     # Extract the first word of "$ac_tool_prefix$ac_prog", so it can be a program name with args.
 set dummy $ac_tool_prefix$ac_prog; ac_word=$2
@@ -6736,7 +6736,7 @@
 fi
 if test -z "$RANLIB"; then
   ac_ct_RANLIB=$RANLIB
-  for ac_prog in gcc-ranlib ranlib
+  for ac_prog in  ranlib
 do
   # Extract the first word of "$ac_prog", so it can be a program name with args.
 set dummy $ac_prog; ac_word=$2
