diff --git a/hls-ormolu-plugin.cabal b/hls-ormolu-plugin.cabal
index 93c27a704..496493203 100644
--- a/hls-ormolu-plugin.cabal
+++ b/hls-ormolu-plugin.cabal
@@ -23,12 +23,13 @@ library
     , base            >=4.12  && <5
     , filepath
     , ghc
+    , ghc-api-compat
     , ghc-boot-th
     , ghcide          >=1.2   && <1.5
     , hls-plugin-api  >=1.1   && <1.3
     , lens
     , lsp
-    , ormolu          ^>=0.1.2
+    , ormolu          ^>=0.1.2 || ^>= 0.2
     , text
 
   default-language: Haskell2010
