diff --git a/hls-ormolu-plugin.cabal b/hls-ormolu-plugin.cabal
index 89cd715fab..c1f03546c1 100644
--- a/hls-ormolu-plugin.cabal
+++ b/hls-ormolu-plugin.cabal
@@ -33,22 +33,6 @@ library
 
   default-language: Haskell2010
 
-  if impl(ghc < 8.10.5)
-    build-depends:
-      ghc-api-compat ==8.6
-  elif impl(ghc == 8.10.5)
-    build-depends:
-      ghc-api-compat ==8.10.5
-  elif impl(ghc == 8.10.6)
-    build-depends:
-      ghc-api-compat ==8.10.6
-  elif impl(ghc == 8.10.7)
-    build-depends:
-      ghc-api-compat ==8.10.7
-  elif impl(ghc == 9.0.1)
-    build-depends:
-      ghc-api-compat ==9.0.1
-
 test-suite tests
   type:             exitcode-stdio-1.0
   default-language: Haskell2010
diff --git a/src/Ide/Plugin/Ormolu.hs b/src/Ide/Plugin/Ormolu.hs
index 4a19566f87..d1a465eb64 100644
--- a/src/Ide/Plugin/Ormolu.hs
+++ b/src/Ide/Plugin/Ormolu.hs
@@ -7,21 +7,20 @@ module Ide.Plugin.Ormolu
   )
 where
 
-import           Control.Exception           (try)
-import           Control.Monad.IO.Class      (liftIO)
-import qualified Data.Text                   as T
-import           Development.IDE             hiding (pluginHandlers)
-import           Development.IDE.GHC.Compat  (moduleNameString)
-import qualified DynFlags                    as D
-import qualified EnumSet                     as S
+import           Control.Exception               (try)
+import           Control.Monad.IO.Class          (liftIO)
+import qualified Data.Text                       as T
+import           Development.IDE                 hiding (pluginHandlers)
+import           Development.IDE.GHC.Compat      (moduleNameString, hsc_dflags)
+import qualified Development.IDE.GHC.Compat      as D
+import qualified Development.IDE.GHC.Compat.Util as S
 import           GHC.LanguageExtensions.Type
-import           GhcPlugins                  (HscEnv (hsc_dflags))
 import           Ide.PluginUtils
 import           Ide.Types
-import           Language.LSP.Server         hiding (defaultConfig)
+import           Language.LSP.Server             hiding (defaultConfig)
 import           Language.LSP.Types
 import           Ormolu
-import           System.FilePath             (takeFileName)
+import           System.FilePath                 (takeFileName)
 
 -- ---------------------------------------------------------------------
 
