diff --git a/kdevplatform/KDevPlatformConfig.cmake.in b/kdevplatform/KDevPlatformConfig.cmake.in
index caf641fcbf..a6732b61a8 100644
--- a/kdevplatform/KDevPlatformConfig.cmake.in
+++ b/kdevplatform/KDevPlatformConfig.cmake.in
@@ -3,6 +3,7 @@
 set(KDEV_PLUGIN_VERSION @KDEV_PLUGIN_VERSION@)
 
 include("${CMAKE_CURRENT_LIST_DIR}/KDevPlatformMacros.cmake")
+include(CMakeFindDependencyMacro)
 
 find_dependency(Qt5Concurrent "@QT_MIN_VERSION@")
 find_dependency(Qt5Core "@QT_MIN_VERSION@")
