diff --git a/lib/AST/ASTScope.cpp b/lib/AST/ASTScope.cpp
index 452bcc5..287f027 100644
--- a/lib/AST/ASTScope.cpp
+++ b/lib/AST/ASTScope.cpp
@@ -152,6 +152,8 @@ ASTScope::ASTScope(const ASTScope *parent, ArrayRef<ASTScope *> children)
   }));
 }
 
+ASTScope::~ASTScope() {}
+
 /// Determine whether we should completely skip the given element in a
 /// \c BraceStmt.
 static bool shouldSkipBraceStmtElement(ASTNode element) {
diff --git a/lib/IRGen/IRGenSIL.cpp b/lib/IRGen/IRGenSIL.cpp
index bba2db1..331f0ee 100644
--- a/lib/IRGen/IRGenSIL.cpp
+++ b/lib/IRGen/IRGenSIL.cpp
@@ -181,7 +181,8 @@ public:
   {}
   
   LoweredValue(Explosion &e)
-    : kind(Kind::Explosion), explosion{{}} {
+    : kind(Kind::Explosion), explosion{} {
+    explosion = {};
     auto Elts = e.claimAll();
     explosion.values.append(Elts.begin(), Elts.end());
   }
