diff --git a/lib/ClangImporter/ImportName.h b/lib/ClangImporter/ImportName.h
index 7a2e82c..ad25ecb 100644
--- a/lib/ClangImporter/ImportName.h
+++ b/lib/ClangImporter/ImportName.h
@@ -91,7 +91,7 @@ struct ImportedName {
 
   /// For names that map Objective-C error handling conventions into
   /// throwing Swift methods, describes how the mapping is performed.
-  Optional<ImportedErrorInfo> ErrorInfo;
+  Optional<ImportedErrorInfo> ErrorInfo = Optional<ImportedErrorInfo>();
 
   /// For a declaration name that makes the declaration into an
   /// instance member, the index of the "Self" parameter.
