diff --git a/lib/ClangImporter/ClangImporter.cpp b/lib/ClangImporter/ClangImporter.cpp
index 241f9ce9c7..709c2e04a2 100644
--- a/lib/ClangImporter/ClangImporter.cpp
+++ b/lib/ClangImporter/ClangImporter.cpp
@@ -472,7 +472,7 @@ getNormalInvocationArguments(std::vector<std::string> &invocationArgStrs,
       invocationArgStrs.insert(invocationArgStrs.end(),
                                {"-fobjc-runtime=ios-7.0"});
   } else {
-    invocationArgStrs.insert(invocationArgStrs.end(), {"-x", "c", "-std=gnu11"});
+    invocationArgStrs.insert(invocationArgStrs.end(), {"-x", "c", "-std=gnu11", "-D_GNU_SOURCE"});
   }
 
   // Set C language options.
