diff --git a/lib/ClangImporter/ClangImporter.cpp b/lib/ClangImporter/ClangImporter.cpp
index fae21e9369..1b5e69bc3c 100644
--- a/lib/ClangImporter/ClangImporter.cpp
+++ b/lib/ClangImporter/ClangImporter.cpp
@@ -533,7 +533,7 @@ getNormalInvocationArguments(std::vector<std::string> &invocationArgStrs,
       // not import Objective-C modules.
       //
       // Just use the most feature-rich C language mode.
-      "-x", "c", "-std=gnu11",
+      "-x", "c", "-std=gnu11", "-D_GNU_SOURCE",
     });
 
     // The module map used for Glibc depends on the target we're compiling for,
