diff --git a/cmake/modules/AddSwift.cmake b/cmake/modules/AddSwift.cmake
index 64902335dd..ce8fae4d4d 100644
--- a/cmake/modules/AddSwift.cmake
+++ b/cmake/modules/AddSwift.cmake
@@ -198,8 +198,6 @@ function(_add_variant_c_compile_flags)
 
   is_build_type_optimized("${CFLAGS_BUILD_TYPE}" optimized)
   if(optimized)
-    list(APPEND result "-O2")
-
     # Add -momit-leaf-frame-pointer on x86.
     if("${CFLAGS_ARCH}" STREQUAL "i386" OR "${CFLAGS_ARCH}" STREQUAL "x86_64")
       list(APPEND result "-momit-leaf-frame-pointer")
