--- a/include/swift/Basic/NullablePtr.h	1496863235.690130532
+++ b/include/swift/Basic/NullablePtr.h	1496863257.700130543
@@ -37,7 +37,7 @@
   template<typename OtherT>
   NullablePtr(NullablePtr<OtherT> Other,
               typename std::enable_if<
-                std::is_convertible<OtherT, T>::value,
+                __is_convertible_to(OtherT, T),
                 PlaceHolder
               >::type = PlaceHolder()) : Ptr(Other.getPtrOrNull()) {}
   
