From 56fc0d12d6881a5165b1e61d5f56b08c24e8478a Mon Sep 17 00:00:00 2001
From: Mark Lacey <mark.lacey@apple.com>
Date: Tue, 13 Dec 2016 11:30:20 -0800
Subject: [PATCH] Fix -DNDEBUG build.

---
 lib/AST/Expr.cpp | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/lib/AST/Expr.cpp b/lib/AST/Expr.cpp
index d5299ca..0bc6542 100644
--- a/lib/AST/Expr.cpp
+++ b/lib/AST/Expr.cpp
@@ -218,15 +218,15 @@ void Expr::propagateLValueAccessKind(AccessKind accessKind,
   /// A visitor class which walks an entire l-value expression.
   class PropagateAccessKind
        : public ExprVisitor<PropagateAccessKind, void, AccessKind> {
-#ifndef NDEBUG
     std::function<Type(Expr *)> GetType;
+#ifndef NDEBUG
     bool AllowOverwrite;
 #endif
   public:
     PropagateAccessKind(std::function<Type(Expr *)> getType,
-                        bool allowOverwrite)
+                        bool allowOverwrite) : GetType(getType)
 #ifndef NDEBUG
-      : GetType(getType), AllowOverwrite(allowOverwrite)
+                                               , AllowOverwrite(allowOverwrite)
 #endif
     {}
 
