diff --git a/cmake/modules/AddSwift.cmake b/cmake/modules/AddSwift.cmake
index 64902335dd..8b85929a98 100644
--- a/cmake/modules/AddSwift.cmake
+++ b/cmake/modules/AddSwift.cmake
@@ -1690,12 +1682,9 @@ function(add_swift_library name)
               add_dependencies("swift-stdlib${VARIANT_SUFFIX}"
                   ${lipo_target}
                   ${lipo_target_static})
-              if((NOT "${name}" STREQUAL "swiftStdlibCollectionUnittest") AND
-                 (NOT "${name}" STREQUAL "swiftStdlibUnicodeUnittest"))
-                add_dependencies("swift-test-stdlib${VARIANT_SUFFIX}"
-                    ${lipo_target}
-                    ${lipo_target_static})
-              endif()
+              add_dependencies("swift-test-stdlib${VARIANT_SUFFIX}"
+                  ${lipo_target}
+                  ${lipo_target_static})
             endif()
           endforeach()
         endif()
