diff --git a/lib/Driver/ToolChains.cpp b/lib/Driver/ToolChains.cpp
index ee44637fe0..10e6500c92 100644
--- a/lib/Driver/ToolChains.cpp
+++ b/lib/Driver/ToolChains.cpp
@@ -1291,15 +1291,8 @@ toolchains::GenericUnix::constructInvocation(const LinkJobAction &job,
   }
 
   // Select the linker to use.
-  std::string Linker;
-  if (const Arg *A = context.Args.getLastArg(options::OPT_use_ld)) {
-    Linker = A->getValue();
-  } else {
-    Linker = getDefaultLinker();
-  }
-  if (!Linker.empty()) {
-    Arguments.push_back(context.Args.MakeArgString("-fuse-ld=" + Linker));
-  }
+  if (const Arg *A = context.Args.getLastArg(options::OPT_use_ld))
+    Arguments.push_back(context.Args.MakeArgString("-fuse-ld=" + std::string(A->getValue())));
 
   // Configure the toolchain.
   // By default, use the system clang++ to link.
@@ -1419,6 +1402,12 @@ toolchains::GenericUnix::constructInvocation(const LinkJobAction &job,
     Arguments.push_back(context.Args.MakeArgString(PostInputObjectPath));
   }
 
+  if (context.Args.getLastArg(options::OPT_g))
+    Arguments.push_back("-Wl,-nostrip");
+
+  // Needed when using lld
+    Arguments.push_back("-Wl,-export-dynamic");
+
   // This should be the last option, for convenience in checking output.
   Arguments.push_back("-o");
   Arguments.push_back(context.Output.getPrimaryOutputFilename().c_str());
