diff --git a/lib/Immediate/CMakeLists.txt b/lib/Immediate/CMakeLists.txt
index f6919cddf0..ebaf85a0e8 100644
--- a/lib/Immediate/CMakeLists.txt
+++ b/lib/Immediate/CMakeLists.txt
@@ -8,5 +8,9 @@ add_swift_library(swiftImmediate STATIC
     swiftSILOptimizer
     swiftIRGen
   LLVM_COMPONENT_DEPENDS
-    linker mcjit)
+    linker
+    mcjit
+    ExecutionEngine
+    TransformUtils
+)
 
diff --git a/lib/IRGen/CMakeLists.txt b/lib/IRGen/CMakeLists.txt
index 3749f7cd32..b080c03450 100644
--- a/lib/IRGen/CMakeLists.txt
+++ b/lib/IRGen/CMakeLists.txt
@@ -52,4 +52,6 @@ add_swift_library(swiftIRGen STATIC
     # FIXME: Clang should really export these in some reasonable manner.
     clangCodeGen
     clangAST
+  LLVM_COMPONENT_DEPENDS
+    TransformUtils
 )
diff --git a/tools/swift-llvm-opt/CMakeLists.txt b/tools/swift-llvm-opt/CMakeLists.txt
index 628d7bddb8..ce60754756 100644
--- a/tools/swift-llvm-opt/CMakeLists.txt
+++ b/tools/swift-llvm-opt/CMakeLists.txt
@@ -13,6 +13,11 @@ add_swift_host_tool(swift-llvm-opt
 
   LLVM_COMPONENT_DEPENDS
     DebugInfoCodeView
+    ScalarOpts
+    Vectorize
+    InstCombine
+    Target
+    CodeGen
 
   SWIFT_COMPONENT tools
 )
diff --git a/stdlib/public/SwiftRemoteMirror/CMakeLists.txt b/stdlib/public/SwiftRemoteMirror/CMakeLists.txt
index 260d04e0f9..0d7f8cf1f3 100644
--- a/stdlib/public/SwiftRemoteMirror/CMakeLists.txt
+++ b/stdlib/public/SwiftRemoteMirror/CMakeLists.txt
@@ -5,6 +5,6 @@ if(SWIFT_BUILD_DYNAMIC_STDLIB OR SWIFT_BUILD_REMOTE_MIRROR)
       SwiftRemoteMirror.cpp
       LINK_LIBRARIES swiftReflection
       C_COMPILE_FLAGS ${SWIFT_RUNTIME_CXX_FLAGS}
-      LINK_FLAGS ${SWIFT_RUNTIME_LINK_FLAGS}
+      LINK_FLAGS ${SWIFT_RUNTIME_LINK_FLAGS} "-lLLVMSupport"
       INSTALL_IN_COMPONENT swift-remote-mirror)
 endif()
diff --git a/stdlib/public/core/CMakeLists.txt b/stdlib/public/core/CMakeLists.txt
index fc49842db8..e5210337be 100644
--- a/stdlib/public/core/CMakeLists.txt
+++ b/stdlib/public/core/CMakeLists.txt
@@ -222,7 +222,7 @@ add_swift_library(swiftCore ${SWIFT_STDLIB_LIBRARY_BUILD_TYPES} IS_STDLIB IS_STD
     ${GROUP_INFO_JSON_FILE}
   SWIFT_COMPILE_FLAGS ${swift_stdlib_compile_flags}
   LINK_FLAGS ${swift_core_link_flags}
-  PRIVATE_LINK_LIBRARIES ${swift_core_private_link_libraries}
+  PRIVATE_LINK_LIBRARIES ${swift_core_private_link_libraries} "-lLLVMSupport"
   INCORPORATE_OBJECT_LIBRARIES swiftRuntime swiftStdlibStubs
   FRAMEWORK_DEPENDS ${swift_core_framework_depends}
   INSTALL_IN_COMPONENT stdlib)
diff --git a/cmake/modules/AddSwift.cmake b/cmake/modules/AddSwift.cmake
index 5dd80da545..ab7b6570aa 100644
--- a/cmake/modules/AddSwift.cmake
+++ b/cmake/modules/AddSwift.cmake
@@ -977,23 +977,6 @@ function(_add_swift_library_single target name)
         "${swift_module_dependency_target}"
         ${LLVM_COMMON_DEPENDS})
 
-  # HACK: On some systems or build directory setups, CMake will not find static
-  # archives of Clang libraries in the Clang build directory, and it will pass
-  # them as '-lclangFoo'.  Some other logic in CMake would reorder libraries
-  # specified with this syntax, which breaks linking.
-  set(prefixed_link_libraries)
-  foreach(dep ${SWIFTLIB_SINGLE_LINK_LIBRARIES})
-    if("${dep}" MATCHES "^clang")
-      if("${SWIFT_HOST_VARIANT_SDK}" STREQUAL "WINDOWS")
-        set(dep "${LLVM_LIBRARY_OUTPUT_INTDIR}/${dep}.lib")
-      else()
-        set(dep "${LLVM_LIBRARY_OUTPUT_INTDIR}/lib${dep}.a")
-      endif()
-    endif()
-    list(APPEND prefixed_link_libraries "${dep}")
-  endforeach()
-  set(SWIFTLIB_SINGLE_LINK_LIBRARIES "${prefixed_link_libraries}")
-
   if("${libkind}" STREQUAL "SHARED")
     target_link_libraries("${target}" PRIVATE ${SWIFTLIB_SINGLE_LINK_LIBRARIES})
   elseif("${libkind}" STREQUAL "OBJECT")
--- a/tools/SourceKit/lib/SwiftLang/CMakeLists.txt      2017-06-21 13:12:15.138995828 +0200
+++ b/tools/SourceKit/lib/SwiftLang/CMakeLists.txt      2017-06-21 13:11:42.028996595 +0200
@@ -10,7 +10,7 @@
   SwiftSourceDocInfo.cpp
   DEPENDS SourceKitCore swiftDriver swiftFrontend swiftClangImporter swiftIDE
           swiftAST swiftMarkup swiftParse swiftSIL swiftSILGen swiftSILOptimizer
-          swiftIRGen swiftSema swiftBasic swiftSerialization swiftOption libcmark_static
+          swiftIRGen swiftSema swiftBasic swiftSerialization swiftOption cmark
     # Clang dependencies.
       clangIndex
       clangFormat
--- a/lib/Markup/CMakeLists.txt 2017-06-21 13:12:12.516995889 +0200
+++ b/lib/Markup/CMakeLists.txt 2017-06-21 13:11:40.437996632 +0200
@@ -4,5 +4,5 @@
   Markup.cpp
   
   LINK_LIBRARIES
-    libcmark_static)
+    cmark)
 
--- a/cmake/modules/AddSwift.cmake	2017-06-21 15:55:08.235951427 +0200
+++ b/cmake/modules/AddSwift.cmake	2017-06-21 15:54:20.768951674 +0200
@@ -343,9 +343,7 @@
     DEPLOYMENT_VERSION_WATCHOS "${LFLAGS_DEPLOYMENT_VERSION_WATCHOS}"
     RESULT_VAR_NAME result)
 
-  if("${LFLAGS_SDK}" STREQUAL "LINUX")
-    list(APPEND result "-lpthread" "-latomic" "-ldl")
-  elseif("${LFLAGS_SDK}" STREQUAL "FREEBSD")
+  if("${LFLAGS_SDK}" STREQUAL "FREEBSD")
     list(APPEND result "-lpthread")
   elseif("${LFLAGS_SDK}" STREQUAL "CYGWIN")
     # No extra libraries required.
