diff --git a/lib/Driver/ToolChains.cpp b/lib/Driver/ToolChains.cpp
index f0e03d0570..3a5a262226 100644
--- a/lib/Driver/ToolChains.cpp
+++ b/lib/Driver/ToolChains.cpp
@@ -1488,17 +1488,11 @@ std::string toolchains::GenericUnix::getDefaultLinker() const {
   case llvm::Triple::armeb:
   case llvm::Triple::thumb:
   case llvm::Triple::thumbeb:
-    // BFD linker has issues wrt relocation of the protocol conformance
-    // section on these targets, it also generates COPY relocations for
-    // final executables, as such, unless specified, we default to gold
-    // linker.
-    return "gold";
   case llvm::Triple::x86_64:
   case llvm::Triple::ppc64:
   case llvm::Triple::ppc64le:
   case llvm::Triple::systemz:
-    // BFD linker has issues wrt relocations against protected symbols.
-    return "gold";
+    return "lld";
   default:
     // Otherwise, use the default BFD linker.
     return "";
@@ -1743,6 +1737,9 @@ toolchains::GenericUnix::constructInvocation(const LinkJobAction &job,
     Arguments.push_back(context.Args.MakeArgString(PostInputObjectPath));
   }
 
+  // Needed when using lld
+    Arguments.push_back("-Wl,-export-dynamic");
+
   // This should be the last option, for convenience in checking output.
   Arguments.push_back("-o");
   Arguments.push_back(context.Output.getPrimaryOutputFilename().c_str());
