diff --git a/lib/SILOptimizer/Utils/SILSSAUpdater.cpp b/lib/SILOptimizer/Utils/SILSSAUpdater.cpp
index 75fe84943b..241e342368 100644
--- a/lib/SILOptimizer/Utils/SILSSAUpdater.cpp
+++ b/lib/SILOptimizer/Utils/SILSSAUpdater.cpp
@@ -340,7 +340,7 @@ public:
 
   /// InstrIsPHI - Check if an instruction is a PHI.
   ///
-  static SILPHIArgument *InstrIsPHI(ValueBase *I) {
+  static SILPHIArgument *InstrIsPHI(SILNode *I) {
     auto *Res = dyn_cast<SILPHIArgument>(I);
     return Res;
   }
