diff --git a/stdlib/public/Platform/glibc.modulemap.gyb b/stdlib/public/Platform/glibc.modulemap.gyb
index a2f325e..67a48f6 100644
--- a/stdlib/public/Platform/glibc.modulemap.gyb
+++ b/stdlib/public/Platform/glibc.modulemap.gyb
@@ -19,17 +19,13 @@
 /// It's not named just Glibc so that it doesn't conflict in the event of a
 /// future official glibc modulemap.
 module SwiftGlibc [system] {
-% if CMAKE_SDK in ["LINUX", "FREEBSD", "CYGWIN"]:
-  link "pthread"
-  // FIXME: util contains rarely used functions and not usually needed. Unfortunately
-  // link directive doesn't work in the submodule yet.
-  link "util"
-% end
-
-  link "dl"
-
+  config_macros [exhaustive] _ALL_SOURCE
   // C standard library
   module C {
+    module types {
+      header "${GLIBC_ARCH_INCLUDE_PATH}/sys/types.h"
+      export *
+    }
 % if CMAKE_SDK in ["LINUX", "FREEBSD", "CYGWIN"]:
     module complex {
       header "${GLIBC_INCLUDE_PATH}/complex.h"
@@ -82,10 +78,10 @@ module SwiftGlibc [system] {
     //   header "${GLIBC_INCLUDE_PATH}/iso646.h"
     //   export *
     // }
-    // module limits {
-    //   header "${GLIBC_INCLUDE_PATH}/limits.h"
-    //   export *
-    // }
+    module limits {
+      header "${GLIBC_INCLUDE_PATH}/limits.h"
+      export *
+    }
 
     module locale {
       header "${GLIBC_INCLUDE_PATH}/locale.h"
@@ -375,10 +371,6 @@ module SwiftGlibc [system] {
         header "${GLIBC_ARCH_INCLUDE_PATH}/sys/times.h"
         export *
       }
-      module types {
-        header "${GLIBC_ARCH_INCLUDE_PATH}/sys/types.h"
-        export *
-      }
       module uio {
         header "${GLIBC_ARCH_INCLUDE_PATH}/sys/uio.h"
         export *
