diff --git a/lib/AST/ASTScope.cpp b/lib/AST/ASTScope.cpp
index 452bcc5..287f027 100644
--- a/lib/AST/ASTScope.cpp
+++ b/lib/AST/ASTScope.cpp
@@ -152,6 +152,8 @@ ASTScope::ASTScope(const ASTScope *parent, ArrayRef<ASTScope *> children)
   }));
 }
 
+ASTScope::~ASTScope() {}
+
 /// Determine whether we should completely skip the given element in a
 /// \c BraceStmt.
 static bool shouldSkipBraceStmtElement(ASTNode element) {
