diff --git a/stdlib/public/Platform/glibc.modulemap.gyb b/stdlib/public/Platform/glibc.modulemap.gyb
index 59c6bfd50c..4a2204670e 100644
--- a/stdlib/public/Platform/glibc.modulemap.gyb
+++ b/stdlib/public/Platform/glibc.modulemap.gyb
@@ -19,17 +19,7 @@
 /// It's not named just Glibc so that it doesn't conflict in the event of a
 /// future official glibc modulemap.
 module SwiftGlibc [system] {
-% if CMAKE_SDK in ["LINUX", "FREEBSD", "CYGWIN"]:
-  link "pthread"
-  // FIXME: util contains rarely used functions and not usually needed. Unfortunately
-  // link directive doesn't work in the submodule yet.
-  link "util"
-% end
-
-% if CMAKE_SDK != "FREEBSD":
-  link "dl"
-% end
-
+  header "/usr/include/swift/Macros.h"
   // C standard library
   module C {
 % if CMAKE_SDK in ["LINUX", "FREEBSD", "CYGWIN"]:
