diff --git a/stdlib/public/Platform/glibc.modulemap.gyb b/stdlib/public/Platform/glibc.modulemap.gyb
index bedfa531e9..41d7e68227 100644
--- a/stdlib/public/Platform/glibc.modulemap.gyb
+++ b/stdlib/public/Platform/glibc.modulemap.gyb
@@ -19,23 +19,17 @@
 /// It's not named just Glibc so that it doesn't conflict in the event of a
 /// future official glibc modulemap.
 module SwiftGlibc [system] {
-% if CMAKE_SDK in ["LINUX", "FREEBSD", "CYGWIN"]:
-  link "pthread"
-  // FIXME: util contains rarely used functions and not usually needed. Unfortunately
-  // link directive doesn't work in the submodule yet.
-  link "util"
-% end
-
-% if CMAKE_SDK != "FREEBSD" and CMAKE_SDK != "HAIKU":
-  link "dl"
-% end
-
 % if CMAKE_SDK == "HAIKU":
   link "network"
   link "bsd"
   link "execinfo"
 % end
 
+  module types {
+    header "${GLIBC_ARCH_INCLUDE_PATH}/sys/types.h"
+    export *
+  }
+
   // C standard library
   module C {
 % if CMAKE_SDK in ["LINUX", "FREEBSD", "CYGWIN", "HAIKU"]:
@@ -176,13 +171,6 @@ module SwiftGlibc [system] {
 
   // POSIX
   module POSIX {
-% if CMAKE_SDK in ["LINUX", "CYGWIN"]:
-    module wait {
-      header "${GLIBC_INCLUDE_PATH}/wait.h"
-      export *
-    }
-% end
-
 % if CMAKE_SDK in ["LINUX", "FREEBSD"]:
     module aio {
       header "${GLIBC_INCLUDE_PATH}/aio.h"
@@ -451,10 +439,6 @@ module SwiftGlibc [system] {
         header "${GLIBC_ARCH_INCLUDE_PATH}/sys/times.h"
         export *
       }
-      module types {
-        header "${GLIBC_ARCH_INCLUDE_PATH}/sys/types.h"
-        export *
-      }
 % if CMAKE_SDK in ["FREEBSD"]:
       module event {
         header "${GLIBC_ARCH_INCLUDE_PATH}/sys/event.h"
