diff --git a/CMakeLists.txt b/CMakeLists.txt
index 70823a5515..68c0735995 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -897,3 +894,24 @@ if(XCODE)
   add_custom_target(Miscellaneous
       SOURCES ${SWIFT_TOPLEVEL_HEADERS})
 endif()
+
+install(DIRECTORY include/
+  DESTINATION include
+  FILES_MATCHING
+  PATTERN "*.h"
+  PATTERN "*.def"
+  PATTERN ".svn" EXCLUDE
+  )
+
+install(DIRECTORY lib/IRGen
+  DESTINATION include/swift
+  FILES_MATCHING
+  PATTERN "*.h"
+  PATTERN "*.def"
+  PATTERN ".svn" EXCLUDE
+  )
+
+install(FILES cmake/modules/SwiftUtils.cmake DESTINATION lib/cmake/swift)
+install(FILES cmake/modules/SwiftAddCustomCommandTarget.cmake DESTINATION lib/cmake/swift)
+install(FILES ${CMAKE_BINARY_DIR}/include/swift/Config.h DESTINATION include/swift)
+install(FILES ${CMAKE_BINARY_DIR}/include/swift/Syntax/SyntaxKind.h DESTINATION include/swift/Syntax)
diff --git a/cmake/modules/AddSwift.cmake b/cmake/modules/AddSwift.cmake
index fd6db69b0d..a748269c69 100644
--- a/cmake/modules/AddSwift.cmake
+++ b/cmake/modules/AddSwift.cmake
@@ -801,6 +799,15 @@ function(_add_swift_library_single target name)
       ${SWIFT_SECTIONS_OBJECT_END})
   _set_target_prefix_and_suffix("${target}" "${libkind}" "${SWIFTLIB_SINGLE_SDK}")
 
+  if(${SWIFTLIB_INSTALL_IN_COMPONENT} STREQUAL "dev" AND (NOT SWIFTLIB_FORCE_BUILD_FOR_HOST_SDK))
+    install(TARGETS ${target}
+      COMPONENT ${target}
+      ${export_to_swifttargets}
+      LIBRARY DESTINATION lib${LLVM_LIBDIR_SUFFIX}
+      ARCHIVE DESTINATION lib${LLVM_LIBDIR_SUFFIX}
+      RUNTIME DESTINATION bin)
+  endif()
+
   if(SWIFTLIB_SINGLE_TARGET_LIBRARY)
     if(NOT "${SWIFT_${SWIFTLIB_SINGLE_SDK}_ICU_UC_INCLUDE}" STREQUAL "")
       target_include_directories("${target}" SYSTEM PRIVATE "${SWIFT_${SWIFTLIB_SINGLE_SDK}_ICU_UC_INCLUDE}")
