--- a/stdlib/public/runtime/CMakeLists.txt	2017-06-07 17:44:00.623701441 +0200
+++ b/stdlib/public/runtime/CMakeLists.txt	2017-06-07 17:43:16.255699257 +0200
@@ -36,6 +36,9 @@ set(swift_runtime_objc_sources
     "${SWIFT_SOURCE_DIR}/lib/Demangling/OldRemangler.cpp"
     "${SWIFT_SOURCE_DIR}/lib/Demangling/Remangler.cpp")
 
+SET_SOURCE_FILES_PROPERTIES(Reflection.mm PROPERTIES COMPILE_FLAGS -fno-exceptions)
+SET_SOURCE_FILES_PROPERTIES(SwiftObject.mm PROPERTIES COMPILE_FLAGS -fno-exceptions)
+
 set(swift_runtime_sources
     AnyHashableSupport.cpp
     Casting.cpp
