diff --git a/lib/Driver/Compilation.cpp b/lib/Driver/Compilation.cpp
index d1fb5f4..a745f34 100644
--- a/lib/Driver/Compilation.cpp
+++ b/lib/Driver/Compilation.cpp
@@ -433,7 +433,7 @@ int Compilation::performJobsImpl() {
   }
 
   int Result = EXIT_SUCCESS;
-  llvm::TimerGroup DriverTimerGroup("Driver Time Compilation");
+  llvm::TimerGroup DriverTimerGroup("Compilation Drive Time", "");
   llvm::SmallDenseMap<const Job *, std::unique_ptr<llvm::Timer>, 16>
     DriverTimers;
 
@@ -465,7 +465,7 @@ int Compilation::performJobsImpl() {
       DriverTimers.insert({
         BeganCmd,
         std::unique_ptr<llvm::Timer>(
-          new llvm::Timer(OS.str(), DriverTimerGroup))
+          new llvm::Timer(OS.str(), "", DriverTimerGroup))
       });
       DriverTimers[BeganCmd]->startTimer();
     }
diff --git a/include/swift/Basic/Timer.h b/include/swift/Basic/Timer.h
index 681893b..5584323 100644
--- a/include/swift/Basic/Timer.h
+++ b/include/swift/Basic/Timer.h
@@ -33,7 +33,7 @@ namespace swift {
   public:
     explicit SharedTimer(StringRef name) {
       if (CompilationTimersEnabled == State::Enabled)
-        Timer.emplace(name, StringRef("Swift compilation"));
+        Timer.emplace(name, StringRef("Swift compilation"), StringRef(""), StringRef(""));
       else
         CompilationTimersEnabled = State::Skipped;
     }
