From d01a854a9e02705e183628c7c1f9167bb4b9cba0 Mon Sep 17 00:00:00 2001
From: "Martin T. H. Sandsmark" <martin.sandsmark@kde.org>
Date: Sat, 13 Aug 2016 13:01:07 +0200
Subject: [PATCH] Fix crash when closing session

---
 src/TerminalDisplay.cpp | 8 ++------
 1 file changed, 2 insertions(+), 6 deletions(-)

diff --git a/src/TerminalDisplay.cpp b/src/TerminalDisplay.cpp
index 4f28837..555a991 100644
--- a/src/TerminalDisplay.cpp
+++ b/src/TerminalDisplay.cpp
@@ -429,7 +429,7 @@ TerminalDisplay::TerminalDisplay(QWidget* parent)
     // that TerminalDisplay will handle repainting its entire area.
     setAttribute(Qt::WA_OpaquePaintEvent);
 
-    _gridLayout = new QGridLayout(this);
+    _gridLayout = new QGridLayout;
     _gridLayout->setContentsMargins(0, 0, 0, 0);
 
     setLayout(_gridLayout);
@@ -448,9 +448,6 @@ TerminalDisplay::~TerminalDisplay()
     disconnect(_blinkCursorTimer);
 
     delete[] _image;
-
-    delete _gridLayout;
-    delete _outputSuspendedLabel;
     delete _filterChain;
 }
 
@@ -3067,8 +3064,7 @@ void TerminalDisplay::outputSuspended(bool suspended)
                                                 "<a href=\"http://en.wikipedia.org/wiki/Software_flow_control\">suspended</a>"
                                                 " by pressing Ctrl+S."
                                                 "  Press <b>Ctrl+Q</b> to resume."
-                                                "  Click <a href=\"#close\">here</a> to dismiss this message.</qt>"),
-                                           this);
+                                                "  Click <a href=\"#close\">here</a> to dismiss this message.</qt>"));
 
         QPalette palette(_outputSuspendedLabel->palette());
         KColorScheme::adjustBackground(palette, KColorScheme::NeutralBackground);
-- 
2.9.2

