--- a/Configurations/10-main.conf	2022-11-01 15:14:36.000000000 +0100
+++ b/Configurations/10-main.conf	2022-12-16 21:49:35.677654187 +0100
@@ -764,7 +764,7 @@
         #
         # ./Configure linux-armv4 -march=armv6 -D__ARM_MAX_ARCH__=8
         #
-        inherit_from     => [ "linux-latomic" ],
+        inherit_from     => [ "linux-generic32" ],
         asm_arch         => 'armv4',
         perlasm_scheme   => "linux32",
     },
