--- a/src/lib/utils/buffer_stuffer.h	2026-08-13 12:11:32.000000000 +0200
+++ b/src/lib/utils/buffer_stuffer.h	2026-08-15 14:25:53.978681325 +0200
@@ -10,6 +10,7 @@
 #include <botan/assert.h>
 #include <botan/strong_type.h>
 #include <botan/types.h>
+#include <algorithm>
 #include <span>
 
 namespace Botan {
--- a/src/lib/utils/mem_pool/mem_pool.cpp	2026-08-15 14:35:14.153624181 +0200
+++ b/src/lib/utils/mem_pool/mem_pool.cpp	2026-08-15 14:32:54.156544069 +0200
@@ -10,6 +10,7 @@
 #include <botan/internal/target_info.h>
 #include <algorithm>
 #include <optional>
+#include <exception>
 
 #if defined(BOTAN_HAS_VALGRIND) || defined(BOTAN_ENABLE_DEBUG_ASSERTS)
    /**
--- a/src/lib/pubkey/xmss/xmss_address.h	2026-08-15 14:30:25.991870422 +0200
+++ b/src/lib/pubkey/xmss/xmss_address.h	2026-08-15 14:30:39.607929550 +0200
@@ -13,6 +13,7 @@
 #include <botan/types.h>
 #include <array>
 #include <span>
+#include <algorithm>
 
 namespace Botan {
 
