--- a/ormolu.cabal	2022-09-12 12:28:44.275500737 +0200
+++ b/ormolu.cabal	2022-09-12 12:29:23.946045417 +0200
@@ -36,6 +36,13 @@
     description: Parse the default fixity information via Template Haskell
     manual:      True
 
+flag ghc-lib
+    description:
+        Force dependency on ghc-lib-parser even if GHC API in the ghc package is supported
+
+    default:     False
+    manual:      True
+
 library
     exposed-modules:
         Ormolu
@@ -105,7 +112,6 @@
         dlist >=0.8 && <2.0,
         exceptions >=0.6 && <0.11,
         filepath >=1.2 && <1.5,
-        ghc-lib-parser >=9.2 && <9.3,
         megaparsec >=9.0,
         mtl >=2.0 && <3.0,
         syb >=0.7 && <0.8,
@@ -113,7 +119,8 @@
         text >=0.2 && <3.0,
         th-lift-instances >=0.1 && <0.2
 
-    mixins:           ghc-lib-parser hiding (Language.Haskell.TH.Syntax)
+    if flag(ghc-lib)
+        mixins:           ghc-lib-parser hiding (Language.Haskell.TH.Syntax)
 
     if flag(fixity-th)
         cpp-options: -DFIXITY_TH
@@ -121,6 +128,14 @@
     else
         build-depends: file-embed >=0.0.15 && <0.1
 
+    if ((!flag(ghc-lib) && impl(ghc >=9.2)) && impl(ghc <9.3))
+        build-depends:
+            ghc ==9.2.*,
+            ghc-boot-th -any,
+            ghc-boot -any
+    else
+        build-depends: ghc-lib-parser ==9.0.*
+
     if flag(dev)
         ghc-options:
             -Wall -Werror -Wcompat -Wincomplete-record-updates
