--- a/Distribution/Gentoo/GHC.hs	2020-01-28 01:21:10.756990380 +0100
+++ b/Distribution/Gentoo/GHC.hs	2020-01-28 01:20:29.993642135 +0100
@@ -67,7 +67,9 @@
 #else
                                                         False   -- is output in binary mode
 #endif /* MIN_VERSION_Cabal(2,1,0) */
-#if MIN_VERSION_Cabal(2,1,0)
+#if MIN_VERSION_Cabal(3,1,0)
+                         return (BL8.unpack out)
+#elif MIN_VERSION_Cabal(2,1,0)
                          case out of
                              ~(DSU.IODataBinary bs) -> return (BL8.unpack bs)
 #else
