diff --git a/IGC/AdaptorOCL/SPIRV/libSPIRV/SPIRVInstruction.h b/IGC/AdaptorOCL/SPIRV/libSPIRV/SPIRVInstruction.h
index ca50e231..a2789c4e 100644
--- a/IGC/AdaptorOCL/SPIRV/libSPIRV/SPIRVInstruction.h
+++ b/IGC/AdaptorOCL/SPIRV/libSPIRV/SPIRVInstruction.h
@@ -1118,10 +1118,10 @@ public:
     assert(BB && "Invalid BB");
   }
 
-  SPIRVFunctionCallGeneric(SPIRVModule *BM, SPIRVWord ResId, SPIRVType *TheType,
+  SPIRVFunctionCallGeneric(SPIRVBasicBlock *BB, SPIRVWord ResId, SPIRVType *TheType,
     const std::vector<SPIRVWord> &TheArgs)
     : SPIRVInstruction(TheArgs.size() + FixedWordCount, OC, TheType, ResId,
-      BM),
+      BB),
     Args(TheArgs) {}
   SPIRVFunctionCallGeneric():SPIRVInstruction(OC) {}
   const std::vector<SPIRVWord> &getArguments() {
