diff --git a/IGC/common/SysUtils.hpp b/IGC/common/SysUtils.hpp
index b937a98e..92a8f6b6 100644
--- a/IGC/common/SysUtils.hpp
+++ b/IGC/common/SysUtils.hpp
@@ -31,6 +31,7 @@ SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 #if defined _WIN32
 #include <Windows.h>
 #else
+#undef __unused
 #include <sys/stat.h>
 #include <unistd.h>
 #endif
diff --git a/3d/common/iStdLib/File.h b/3d/common/iStdLib/File.h
index 9635c21c..96b9beb0 100644
--- a/3d/common/iStdLib/File.h
+++ b/3d/common/iStdLib/File.h
@@ -32,6 +32,7 @@ SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 #include "MemCopy.h"
 
 #include "Debug.h"
+#undef __unused
 
 #if defined _WIN32
 #   include <direct.h>
