--- a/gn/BUILD.gn	2023-11-17 23:52:19.031936414 +0100
+++ b/gn/BUILD.gn	2023-11-17 23:52:45.987608838 +0100
@@ -716,7 +716,7 @@
     if (is_mac || is_ios) {
       ldflags = [ "-dead_strip" ]
     } else {
-      ldflags = [ "-s", "-Wl,--gc-sections" ]
+      ldflags = [ "-Wl,--gc-sections" ]
     }
     if (target_cpu == "wasm") {
       # The compiler asks us to add an optimization flag to both cflags
