diff --git a/make-linux.mk b/make-linux.mk
index cd946491..ec12ac1d 100644
--- a/make-linux.mk
+++ b/make-linux.mk
@@ -74,8 +74,7 @@ else
 	override CFLAGS+=-Wall -Wno-deprecated -pthread $(INCLUDES) -DNDEBUG $(DEFS)
 	CXXFLAGS?=-O3 -fstack-protector
 	override CXXFLAGS+=-Wall -Wno-deprecated -std=c++11 -pthread $(INCLUDES) -DNDEBUG $(DEFS)
-	STRIP?=strip
-	STRIP+=--strip-all
+	STRIP?=echo
 endif
 
 ifeq ($(ZT_TRACE),1)
@@ -97,11 +96,11 @@ CC_MACH=$(shell $(CC) -dumpmachine | cut -d '-' -f 1)
 ZT_ARCHITECTURE=999
 ifeq ($(CC_MACH),x86_64)
 	ZT_ARCHITECTURE=2
-	ZT_USE_X64_ASM_CRYPTO=1
+	ZT_USE_X64_ASM_CRYPTO=0
 endif
 ifeq ($(CC_MACH),amd64)
 	ZT_ARCHITECTURE=2
-	ZT_USE_X64_ASM_CRYPTO=1
+	ZT_USE_X64_ASM_CRYPTO=0
 endif
 ifeq ($(CC_MACH),powerpc64le)
 	ZT_ARCHITECTURE=8
