--- a/make-linux.mk	2018-04-27 22:52:33.000000000 +0200
+++ b/make-linux.mk	2018-05-16 13:47:32.180195747 +0200
@@ -76,8 +76,7 @@
 	CXXFLAGS?=-O3 -fstack-protector -fPIE
 	override CXXFLAGS+=-Wall -Wno-deprecated -std=c++11 -pthread $(INCLUDES) -DNDEBUG $(DEFS)
 	LDFLAGS=-pie -Wl,-z,relro,-z,now
-	STRIP?=strip
-	STRIP+=--strip-all
+	STRIP?=echo
 endif
 
 ifeq ($(ZT_TRACE),1)
@@ -162,6 +161,11 @@
 	ZT_ARCHITECTURE=3
 	override DEFS+=-DZT_NO_TYPE_PUNNING
 	ZT_USE_ARM32_NEON_ASM_CRYPTO=1
+endif
+ifeq ($(CC_MACH),armv7a)
+	ZT_ARCHITECTURE=3
+	override DEFS+=-DZT_NO_TYPE_PUNNING
+	ZT_USE_ARM32_NEON_ASM_CRYPTO=0
 endif
 ifeq ($(CC_MACH),armv7l)
 	ZT_ARCHITECTURE=3
