diff --git a/make-linux.mk b/make-linux.mk
index 87d29afe..dc411b57 100644
--- a/make-linux.mk
+++ b/make-linux.mk
@@ -193,7 +193,6 @@ all:	one
 
 one:	$(OBJS) service/OneService.o one.o osdep/LinuxEthernetTap.o
 	$(CXX) $(CXXFLAGS) $(LDFLAGS) -o zerotier-one $(OBJS) service/OneService.o one.o osdep/LinuxEthernetTap.o $(LDLIBS)
-	$(STRIP) zerotier-one
 	ln -sf zerotier-one zerotier-idtool
 	ln -sf zerotier-one zerotier-cli
 
