--- a/src/gui/opengl/qopengles2ext.h	2019-09-15 14:46:08.206538582 +0200
+++ b/src/gui/opengl/qopengles2ext.h	2019-09-15 14:46:40.680320130 +0200
@@ -1925,9 +1925,9 @@
 #define GL_EXT_texture_filter_minmax 1
 #endif /* GL_EXT_texture_filter_minmax */
 
-#ifndef GL_EXT_texture_format_BGRA8888
-#define GL_EXT_texture_format_BGRA8888 1
-#endif /* GL_EXT_texture_format_BGRA8888 */
+//#ifndef GL_EXT_texture_format_BGRA8888
+//#define GL_EXT_texture_format_BGRA8888 1
+//#endif /* GL_EXT_texture_format_BGRA8888 */
 
 #ifndef GL_EXT_texture_norm16
 #define GL_EXT_texture_norm16 1
