--- a/src/coreclr/vm/CMakeLists.txt	2021-07-02 23:28:19.000000000 +0200
+++ b/src/coreclr/vm/CMakeLists.txt	2021-08-01 12:45:52.662224364 +0200
@@ -854,8 +854,6 @@
 elseif(CLR_CMAKE_TARGET_ARCH_ARM64)
     set(VM_SOURCES_DAC_AND_WKS_ARCH
         ${ARCH_SOURCES_DIR}/stubs.cpp
-        exceptionhandling.cpp
-        gcinfodecoder.cpp
     )
 
     set(VM_HEADERS_DAC_AND_WKS_ARCH
@@ -865,6 +863,8 @@
 
     set(VM_SOURCES_WKS_ARCH
         ${ARCH_SOURCES_DIR}/profiler.cpp
+        exceptionhandling.cpp
+        gcinfodecoder.cpp
     )
 
     if(CLR_CMAKE_HOST_UNIX)
