--- a/src/coreclr/utilcode/CMakeLists.txt	2021-09-01 19:08:38.000000000 +0200
+++ b/src/coreclr/utilcode//CMakeLists.txt	2021-10-14 15:07:11.950723392 +0200
@@ -54,6 +54,9 @@
   yieldprocessornormalized.cpp
 )
 
+# exclude from optimizations, causes hang
+set_source_files_properties(clrhelpers.cpp PROPERTIES COMPILE_OPTIONS -O0 SKIP_PRECOMPILE_HEADERS ON)
+
 # These source file do not yet compile on Linux.
 # They should be moved out from here into the declaration
 # of UTILCODE_SOURCES above after fixing compiler errors.
