--- a/src/coreclr/debug/dbgutil/elfreader.cpp	2023-03-24 00:45:48.000000000 +0100
+++ b/src/coreclr/debug/dbgutil/elfreader.cpp	2023-05-13 14:11:54.859037311 +0200
@@ -115,9 +115,9 @@
     // Enumerate program headers searching for the PT_DYNAMIC header, etc.
     if (!EnumerateProgramHeaders(
         baseAddress,
-#ifdef TARGET_ALPINE_LINUX
-        // On Alpine, the below dynamic entries for hash, string table, etc. are
-        // RVAs instead of absolute address like on all other Linux distros. Get
+#ifdef TARGET_LINUX_MUSL
+        // On musl libc, the below dynamic entries for hash, string table, etc. are
+        // RVAs instead of absolute address like on other libc's. Get
         // the "loadbias" (basically the base address of the module) and add to
         // these RVAs.
         &loadbias,
