--- a/src/coreclr/debug/createdump/crashinfounix.cpp	2022-11-02 18:13:46.762748373 +0100
+++ b/src/coreclr/debug/createdump/crashinfounix.cpp	2022-11-02 18:14:17.700155592 +0100
@@ -395,7 +395,7 @@
         // performance optimization.
         m_canUseProcVmReadSyscall = false;
         assert(m_fd != -1);
-        *read = pread64(m_fd, buffer, size, (off64_t)address);
+        *read = pread(m_fd, buffer, size, (off_t)address);
     }
 
     if (*read == (size_t)-1)
