--- a/src/CMakeLists.txt	2017-12-04 17:40:58.971683343 +0100
+++ b/src/CMakeLists.txt	2017-12-04 17:41:51.020076262 +0100
@@ -1,6 +1,8 @@
 include_directories(${LIBINTL_INCLUDE_DIR})
 include_directories(${LIBICONV_INCLUDE_DIR})
-include_directories(${LIBEXECINFO_INCLUDE_DIR})
+if (ENABLE_BACKTRACE)
+  include_directories(${LIBEXECINFO_INCLUDE_DIR})
+endif()
 include_directories(${DL_INCLUDE_DIR})
 include_directories(${PTHREAD_INCLUDE_DIR})
 
