diff --git a/Package.swift b/Package.swift
index 55270ed..2d521ce 100644
--- a/Package.swift
+++ b/Package.swift
@@ -3,9 +3,5 @@
 import PackageDescription
 
 let package = Package(
-    name: "BCrypt",
-    dependencies: [
-        // Module for generating random bytes and numbers.
-        .Package(url: "https://github.com/vapor/random.git", Version(1,0,0, prereleaseIdentifiers: ["beta"]))
-    ]
+    name: "BCrypt"
 )
