diff --git a/Package.swift b/Package.swift
index 54252b1..7ed1a92 100644
--- a/Package.swift
+++ b/Package.swift
@@ -7,12 +7,6 @@ let package = Package(
         Target(name: "SourceKitter"),
         Target(name: "LanguageServerProtocol", dependencies: ["BaseProtocol", "SourceKitter"]),
         Target(name: "LanguageServer", dependencies: ["LanguageServerProtocol", "BaseProtocol"])
-    ],
-    dependencies: [
-        .Package(url: "https://github.com/RLovelett/SourceKit.git", majorVersion: 1),
-        .Package(url: "https://github.com/thoughtbot/Argo.git", majorVersion: 4),
-        .Package(url: "https://github.com/edwardaux/Ogra.git", majorVersion: 4),
-        .Package(url: "https://github.com/thoughtbot/Curry.git", majorVersion: 3)
     ]
 )
 
