diff --git a/Makefile b/Makefile
index 6a7c1a4..40bdeb9 100644
--- a/Makefile
+++ b/Makefile
@@ -32,8 +32,7 @@ ALL_TOOLS=msgfmt msgmerge xgettext autopoint
 ALL_M4S=$(sort $(wildcard m4/*.m4))
 ALL_DATA=$(sort $(wildcard data/*))
 
-CFLAGS=-O0 -fPIC
-LDLIBS=-liconv
+CFLAGS+=-fPIC
 
 AR      ?= $(CROSS_COMPILE)ar
 RANLIB  ?= $(CROSS_COMPILE)ranlib
diff --git a/install.sh b/install.sh
old mode 100644
new mode 100755
