--- a/include/CMakeLists.txt    2017-07-19 19:57:20.795909023 +0200
+++ b/include/CMakeLists.txt    2017-07-19 19:58:38.642776174 +0200
@@ -1,3 +1,7 @@
+set(BLOCKS_HEADERS
+  ../lib/BlocksRuntime/Block_private.h
+  ../lib/BlocksRuntime/Block.h)
+
 if (COMPILER_RT_BUILD_SANITIZERS)
   set(SANITIZER_HEADERS
     sanitizer/allocator_interface.h
@@ -20,6 +24,7 @@
 endif(COMPILER_RT_BUILD_XRAY)
 
 set(COMPILER_RT_HEADERS
+  ${BLOCK_HEADERS}
   ${SANITIZER_HEADERS}
        ${XRAY_HEADERS})
 
@@ -49,3 +54,8 @@
 install(FILES ${XRAY_HEADERS}
   PERMISSIONS OWNER_READ OWNER_WRITE GROUP_READ WORLD_READ
   DESTINATION ${COMPILER_RT_INSTALL_PATH}/include/xray)
+
+# Install block headers
+install(FILES ${BLOCKS_HEADERS}
+  PERMISSIONS OWNER_READ OWNER_WRITE GROUP_READ WORLD_READ
+  DESTINATION ${COMPILER_RT_INSTALL_PATH}/include)
--- a/lib/builtins/CMakeLists.txt       2015-01-14 16:55:17.000000000 +0100
+++ b/lib/builtins/CMakeLists.txt       2015-04-08 19:40:41.143179536 +0200
@@ -134,7 +134,9 @@
   udivti3.c
   umoddi3.c
   umodsi3.c
-  umodti3.c)
+  umodti3.c
+  ../BlocksRuntime/runtime.c
+  ../BlocksRuntime/data.c)
 
 if (HAVE_UNWIND_H)
   set(GENERIC_SOURCES
--- /dev/null   2015-04-08 17:57:08.408096785 +0200
+++ b/lib/BlocksRuntime/config.h        2015-04-08 19:40:14.066352552 +0200
@@ -0,0 +1,7 @@
+#ifndef _CONFIG_H_
+#define _CONFIG_H_
+
+#define HAVE_SYNC_BOOL_COMPARE_AND_SWAP_INT 1
+#define HAVE_SYNC_BOOL_COMPARE_AND_SWAP_LONG 1
+
+#endif /* _CONFIG_H_ */

