--- a/lib/builtins/CMakeLists.txt	2015-01-14 16:55:17.000000000 +0100
+++ b/lib/builtins/CMakeLists.txt	2015-04-08 19:40:41.143179536 +0200
@@ -134,7 +134,9 @@
   udivti3.c
   umoddi3.c
   umodsi3.c
-  umodti3.c)
+  umodti3.c
+  ../BlocksRuntime/runtime.c
+  ../BlocksRuntime/data.c)
 
 if (HAVE_UNWIND_H)
   set(GENERIC_SOURCES
--- a/include/CMakeLists.txt	2014-07-07 19:39:31.000000000 +0200
+++ b/include/CMakeLists.txt	2015-04-08 19:40:41.117179700 +0200
@@ -8,6 +8,10 @@
   sanitizer/msan_interface.h
   sanitizer/tsan_interface_atomic.h)
 
+set(BLOCKSRUNTIME_HEADERS
+  ../lib/BlocksRuntime/Block_private.h
+  ../lib/BlocksRuntime/Block.h)
+
 set(XRAY_HEADERS
   xray/xray_interface.h)
 
@@ -29,3 +34,7 @@
 install(FILES ${XRAY_HEADERS}
   PERMISSIONS OWNER_READ OWNER_WRITE GROUP_READ WORLD_READ
   DESTINATION ${COMPILER_RT_INSTALL_PATH}/include/xray)
+
+install(FILES ${BLOCKSRUNTIME_HEADERS}
+  PERMISSIONS OWNER_READ OWNER_WRITE GROUP_READ WORLD_READ
+  DESTINATION /usr/include)
--- /dev/null	2015-04-08 17:57:08.408096785 +0200
+++ b/lib/BlocksRuntime/config.h	2015-04-08 19:40:14.066352552 +0200
@@ -0,0 +1,7 @@
+#ifndef _CONFIG_H_
+#define _CONFIG_H_
+
+#define HAVE_SYNC_BOOL_COMPARE_AND_SWAP_INT 1
+#define HAVE_SYNC_BOOL_COMPARE_AND_SWAP_LONG 1
+
+#endif /* _CONFIG_H_ */
