--- a/Package@swift-4.swift	2017-09-22 10:30:26.039958236 +0200
+++ b/Package@swift-4.swift	2017-09-22 10:29:51.790959029 +0200
@@ -4,8 +4,8 @@
 let package = Package(
     name: "Core",
     products: [
-        .library(name: "libc", targets: ["libc"]),
-        .library(name: "Core", targets: ["Core"]),
+        .library(name: "libc", type: .dynamic, targets: ["libc"]),
+        .library(name: "Core", type: .dynamic, targets: ["Core"]),
     ],
     targets: [
         .target(name: "libc"),
