diff --git a/Package.swift b/Package.swift
index 7fa5d45d..7fe39838 100644
--- a/Package.swift
+++ b/Package.swift
@@ -7,9 +7,5 @@ let package = Package(
     targets: [
         Target(name: "Core", dependencies: ["libc"]),
         Target(name: "libc")
-    ],
-    dependencies: [
-        .Package(url: "https://github.com/vapor/bits.git", beta),
-        .Package(url: "https://github.com/vapor/debugging.git", beta),
     ]
 )
