diff --git a/Package@swift-4.swift b/Package@swift-4.swift
index 86ab3835..d523829f 100644
--- a/Package@swift-4.swift
+++ b/Package@swift-4.swift
@@ -7,13 +7,9 @@ let package = Package(
         .library(name: "libc", targets: ["libc"]),
         .library(name: "Core", targets: ["Core"]),
     ],
-    dependencies: [
-        .package(url: "https://github.com/vapor/bits.git", .upToNextMajor(from: "1.1.0")),
-        .package(url: "https://github.com/vapor/debugging.git", .upToNextMajor(from: "1.1.0")),
-    ],
     targets: [
         .target(name: "libc"),
-        .target(name: "Core", dependencies: ["libc", "Bits", "Debugging"]),
+        .target(name: "Core", dependencies: ["libc"]),
         .testTarget(name: "CoreTests", dependencies: ["Core"]),
     ]
 )
