--- a/rts/Linker.c	2017-07-04 16:15:15.000000000 +0200
+++ b/rts/Linker.c	2018-02-07 19:20:56.326745728 +0100
@@ -479,7 +479,7 @@
 #   endif /* RTLD_DEFAULT */
 
     compileResult = regcomp(&re_invalid,
-           "(([^ \t()])+\\.so([^ \t:()])*):([ \t])*(invalid ELF header|file too short|invalid file format)",
+           "(([^ \t()])+\\.so([^ \t:()])*):([ \t])*(invalid ELF header|file too short|invalid file format|Exec format error)",
            REG_EXTENDED);
     if (compileResult != 0) {
         barf("Compiling re_invalid failed");
