--- a/compiler/main/DynFlags.hs   2021-02-17 18:33:53.436949901 +0100
+++ b/compiler/main/DynFlags.hs   2021-02-17 18:35:34.516806445 +0100
@@ -4646,6 +4646,8 @@
 default_PIC :: Platform -> [GeneralFlag]
 default_PIC platform =
   case (platformOS platform, platformArch platform) of
+    (OSLinux, ArchX86_64) -> [Opt_PIC, Opt_ExternalDynamicRefs]
+    (OSLinux, ArchARM64) -> [Opt_PIC, Opt_ExternalDynamicRefs]
     (OSDarwin, ArchX86_64) -> [Opt_PIC]
     (OSOpenBSD, ArchX86_64) -> [Opt_PIC] -- Due to PIE support in
                                          -- OpenBSD since 5.3 release
--- a/includes/rts/Flags.h	2021-02-17 15:59:34.640783683 +0100
+++ b/includes/rts/Flags.h	2021-02-17 16:00:07.766190465 +0100
@@ -202,7 +202,7 @@
  * files were compiled with -fPIC -fexternal-dynamic-refs and load them
  * anywhere in the address space.
  */
-#if defined(x86_64_HOST_ARCH) && defined(darwin_HOST_OS)
+#if defined(darwin_HOST_OS) || defined(aarch64_HOST_ARCH) || defined(x86_64_HOST_ARCH)
 #define DEFAULT_LINKER_ALWAYS_PIC true
 #else
 #define DEFAULT_LINKER_ALWAYS_PIC false
