--- a/compiler/main/DynFlags.hs	2019-10-06 21:22:02.149248746 +0200
+++ b/compiler/main/DynFlags.hs	2019-10-06 21:22:13.846286195 +0200
@@ -4750,7 +4750,6 @@
         Opt_WarnDodgyForeignImports,
         Opt_WarnInlineRuleShadowing,
         Opt_WarnAlternativeLayoutRuleTransitional,
-        Opt_WarnUnsupportedLlvmVersion,
         Opt_WarnMissedExtraSharedLib,
         Opt_WarnTabs,
         Opt_WarnUnrecognisedWarningFlags,
