--- a/configure	2016-09-30 09:55:09.078030370 +0200
+++ b/configure	2016-09-30 09:56:06.605025637 +0200
@@ -8203,9 +8203,9 @@
     # If gcc knows about the stack protector, turn it off.
     # Otherwise the stack-smash handler gets triggered.
     echo 'int main(void) {return 0;}' > conftest.c
-    if $CC -c conftest.c -fno-stack-protector > /dev/null 2>&1
+    if $CC -c conftest.c > /dev/null 2>&1
     then
-        CFLAGS="$CFLAGS -fno-stack-protector"
+        CFLAGS="$CFLAGS"
     fi
 
     rm -f conftest.c conftest.o
@@ -8292,9 +8292,9 @@
     # If gcc knows about the stack protector, turn it off.
     # Otherwise the stack-smash handler gets triggered.
     echo 'int main(void) {return 0;}' > conftest.c
-    if $CC -c conftest.c -fno-stack-protector > /dev/null 2>&1
+    if $CC -c conftest.c > /dev/null 2>&1
     then
-        CONF_CC_OPTS_STAGE0="$CONF_CC_OPTS_STAGE0 -fno-stack-protector"
+        CONF_CC_OPTS_STAGE0="$CONF_CC_OPTS_STAGE0"
     fi
 
     rm -f conftest.c conftest.o
@@ -8381,9 +8381,9 @@
     # If gcc knows about the stack protector, turn it off.
     # Otherwise the stack-smash handler gets triggered.
     echo 'int main(void) {return 0;}' > conftest.c
-    if $CC -c conftest.c -fno-stack-protector > /dev/null 2>&1
+    if $CC -c conftest.c > /dev/null 2>&1
     then
-        CONF_CC_OPTS_STAGE1="$CONF_CC_OPTS_STAGE1 -fno-stack-protector"
+        CONF_CC_OPTS_STAGE1="$CONF_CC_OPTS_STAGE1"
     fi
 
     rm -f conftest.c conftest.o
@@ -8470,9 +8470,9 @@
     # If gcc knows about the stack protector, turn it off.
     # Otherwise the stack-smash handler gets triggered.
     echo 'int main(void) {return 0;}' > conftest.c
-    if $CC -c conftest.c -fno-stack-protector > /dev/null 2>&1
+    if $CC -c conftest.c > /dev/null 2>&1
     then
-        CONF_CC_OPTS_STAGE2="$CONF_CC_OPTS_STAGE2 -fno-stack-protector"
+        CONF_CC_OPTS_STAGE2="$CONF_CC_OPTS_STAGE2"
     fi
 
     rm -f conftest.c conftest.o
--- a/aclocal.m4	2016-09-30 09:56:11.279025252 +0200
+++ b/aclocal.m4	2016-09-30 09:56:33.229023446 +0200
@@ -606,9 +606,9 @@
     # If gcc knows about the stack protector, turn it off.
     # Otherwise the stack-smash handler gets triggered.
     echo 'int main(void) {return 0;}' > conftest.c
-    if $CC -c conftest.c -fno-stack-protector > /dev/null 2>&1
+    if $CC -c conftest.c > /dev/null 2>&1
     then
-        $2="$$2 -fno-stack-protector"
+        $2="$$2"
     fi
 
     rm -f conftest.c conftest.o
