diff --git a/src/gallium/drivers/lima/lima_draw.c b/src/gallium/drivers/lima/lima_draw.c
index 3b8f7526f3216889189fd87ba49023e6b0c7540c..d604807038934fce3d4c47b73c70131b79447bbe 100644
--- a/src/gallium/drivers/lima/lima_draw.c
+++ b/src/gallium/drivers/lima/lima_draw.c
@@ -335,6 +335,7 @@ lima_pack_clear_plbu_cmd(struct lima_context *ctx)
 
    void *cpu;
    unsigned offset;
+   struct lima_context_framebuffer *fb = &ctx->framebuffer;
    struct pipe_resource *pres = NULL;
    u_upload_alloc(ctx->uploader, 0, lima_clear_buffer_size,
                   0x40, &offset, &pres, &cpu);
@@ -379,8 +380,12 @@ lima_pack_clear_plbu_cmd(struct lima_context *ctx)
    PLBU_CMD_VIEWPORT_Y(0);
    PLBU_CMD_VIEWPORT_H(0x45800000);
 
-   struct pipe_scissor_state *scissor = &ctx->scissor;
-   PLBU_CMD_SCISSORS(scissor->minx, scissor->maxx, scissor->miny, scissor->maxy);
+   if (ctx->rasterizer->base.scissor) {
+      struct pipe_scissor_state *scissor = &ctx->scissor;
+      PLBU_CMD_SCISSORS(scissor->minx, scissor->maxx, scissor->miny, scissor->maxy);
+   } else {
+      PLBU_CMD_SCISSORS(0, fb->base.width, 0, fb->base.height);
+   }
 
    PLBU_CMD_RSW_VERTEX_ARRAY(va + lima_clear_render_state_offset, gl_pos_va);
 
@@ -848,6 +853,8 @@ lima_pack_vs_cmd(struct lima_context *ctx, const struct pipe_draw_info *info)
 static void
 lima_pack_plbu_cmd(struct lima_context *ctx, const struct pipe_draw_info *info)
 {
+   struct lima_context_framebuffer *fb = &ctx->framebuffer;
+
    lima_pack_head_plbu_cmd(ctx);
 
    /* If it's zero scissor, we skip adding all other commands */
@@ -890,6 +897,8 @@ lima_pack_plbu_cmd(struct lima_context *ctx, const struct pipe_draw_info *info)
    if (ctx->rasterizer->base.scissor) {
       struct pipe_scissor_state *scissor = &ctx->scissor;
       PLBU_CMD_SCISSORS(scissor->minx, scissor->maxx, scissor->miny, scissor->maxy);
+   } else {
+      PLBU_CMD_SCISSORS(0, fb->base.width, 0, fb->base.height);
    }
 
    PLBU_CMD_UNKNOWN1();
