--- a/src/gallium/auxiliary/gallivm/lp_bld_init.c	2023-07-30 21:34:27.764793082 +0200
+++ b/src/gallium/auxiliary/gallivm/lp_bld_init.c	2023-07-30 21:35:20.589518763 +0200
@@ -42,8 +42,10 @@
 
 #include <llvm/Config/llvm-config.h>
 #include <llvm-c/Analysis.h>
+#if LLVM_VERSION_MAJOR < 17
 #include <llvm-c/Transforms/Scalar.h>
-#if LLVM_VERSION_MAJOR >= 7
+#endif
+#if LLVM_VERSION_MAJOR >= 7 && LLVM_VERSION_MAJOR < 17
 #include <llvm-c/Transforms/Utils.h>
 #endif
 #include <llvm-c/BitWriter.h>
