--- a/src/util/u_process.c	2018-11-01 09:56:42.445027162 +0100
+++ b/src/util/u_process.c	2018-11-01 09:55:11.725408695 +0100
@@ -54,7 +54,7 @@
    return program_invocation_name;
 }
 #    define GET_PROGRAM_NAME() __getProgramName()
-#elif defined(__CYGWIN__)
+#elif defined(__CYGWIN__) || defined(__linux__) && !defined(__UCLIBC__)
 #    define GET_PROGRAM_NAME() program_invocation_short_name
 #elif defined(__FreeBSD__) && (__FreeBSD__ >= 2)
 #    include <osreldate.h>
