--- a/messageviewer/src/viewer/viewer_p.cpp	2018-04-20 14:36:08.324271088 +0200
+++ b/messageviewer/src/viewer/viewer_p.cpp	2018-04-20 14:37:01.649185382 +0200
@@ -2248,7 +2248,7 @@
     htmlWriter()->reset();
     //TODO: if the item doesn't have the payload fetched, try to fetch it? Maybe not here, but in setMessageItem.
     if (mMessage) {
-        mColorBar->show();
+        mColorBar->hide();
         displayMessage();
     } else if (mMessagePartNode) {
         setMessagePart(mMessagePartNode);
