--- a/messageviewer/src/settings/messageviewer.kcfg.cmake
+++ b/messageviewer/src/settings/messageviewer.kcfg.cmake
@@ -58,6 +58,10 @@
   <entry name="AutoImportKeys" type="Bool">
       <default>false</default>
     </entry>
+    <entry name="showColorBar" type="Bool">
+      <default>false</default>
+      <label>Show HTML Side Bar</label>
+    </entry>
     <entry name="ShowExpandQuotesMark" type="Bool">
       <default>false</default>
       <label>Show expand/collapse quote marks</label>
--- a/messageviewer/src/ui/settings.ui
+++ b/messageviewer/src/ui/settings.ui
@@ -92,6 +92,13 @@
        </property>
       </widget>
      </item>
+     <item>
+      <widget class="QCheckBox" name="kcfg_showColorBar">
+       <property name="text">
+        <string>Show &amp;HTML Side Bar</string>
+       </property>
+      </widget>
+     </item>
      <item>
       <widget class="QCheckBox" name="kcfg_ShrinkQuotes">
        <property name="text">
@@ -217,6 +224,7 @@
  </customwidgets>
  <tabstops>
   <tabstop>kcfg_AccessKeyEnabled</tabstop>
+  <tabstop>kcfg_showColorBar</tabstop>
   <tabstop>kcfg_ShrinkQuotes</tabstop>
   <tabstop>kcfg_ShowExpandQuotesMark</tabstop>
   <tabstop>kcfg_CollapseQuoteLevelSpin</tabstop>
--- a/messageviewer/src/viewer/viewer_p.cpp
+++ b/messageviewer/src/viewer/viewer_p.cpp
@@ -1439,7 +1439,7 @@
     }
 #endif
 
+    if (MessageViewer::MessageViewerSettings::self()->showColorBar() && mMsgDisplay) {
-    if (mMsgDisplay) {
         mColorBar->show();
     } else {
         mColorBar->hide();
@@ -2278,7 +2278,11 @@
     htmlWriter()->reset();
     //TODO: if the item doesn't have the payload fetched, try to fetch it? Maybe not here, but in setMessageItem.
     if (mMessage) {
+        if (MessageViewer::MessageViewerSettings::self()->showColorBar()) {
+            mColorBar->show();
+        } else {
+            mColorBar->hide();
+        }
-        mColorBar->show();
         displayMessage();
     } else if (mMessagePartNode) {
         setMessagePart(mMessagePartNode);
