--- a/CMakeLists.txt	2016-10-04 11:17:50.600267767 +0200
+++ b/CMakeLists.txt	2016-10-04 11:18:49.547270669 +0200
@@ -389,7 +389,7 @@
 # soname, etc...
 add_subdirectory(src)
 
-if (LIBCXXABI_BUILT_STANDALONE AND NOT LIBCXXABI_ENABLE_SHARED)
+if (LIBCXXABI_BUILT_STANDALONE)
   # We can't reasonably test the system C++ library with a static libc++abi.
   # We either need to be able to replace libc++abi at run time (with a shared
   # libc++abi), or we need to be able to replace the C++ runtime (with a non-
