--- a/src/settings/dolphin_directoryviewpropertysettings.kcfg	2018-05-04 11:38:59.465947417 +0200
+++ b/src/settings/dolphin_directoryviewpropertysettings.kcfg	2018-05-04 11:39:14.016922568 +0200
@@ -30,7 +30,7 @@
         <entry name="PreviewsShown" type="Bool" >
             <label context="@label">Previews shown</label>
             <whatsthis context="@info:whatsthis">When this option is enabled, a preview of the file content is shown as an icon.</whatsthis>
-            <default>true</default>
+            <default>false</default>
         </entry>
 
         <entry name="GroupedSorting" type="Bool" >
