--- a/rls-analysis-0.12.0/src/loader.rs
+++ b/rls-analysis-0.12.0/src/loader.rs
@@ -88,11 +88,9 @@
             .join("save-analysis");
         // FIXME sys_root_path allows to break out of 'sandbox' - is that Ok?
         let sys_root_path = sys_root_path();
-        let target_triple = extract_target_triple(sys_root_path.as_path());
         let libs_path = sys_root_path
             .join("lib")
             .join("rustlib")
-            .join(&target_triple)
             .join("analysis");
 
         let src_path = sys_root_path
@@ -109,25 +107,6 @@
     }
 }
 
-// FIXME: This can fail when using a custom toolchain in rustup (often linked to
-// `/$rust_repo/build/$target/stage2`)
-fn extract_target_triple(sys_root_path: &Path) -> String {
-    // Extracts nightly-x86_64-pc-windows-msvc from
-    // $HOME/.rustup/toolchains/nightly-x86_64-pc-windows-msvc
-    let toolchain = sys_root_path
-        .iter()
-        .last()
-        .and_then(OsStr::to_str)
-        .expect("extracting toolchain failed");
-    // Extracts x86_64-pc-windows-msvc from nightly-x86_64-pc-windows-pc
-    let triple = toolchain
-        .splitn(2, '-')
-        .last()
-        .map(String::from)
-        .expect("extracting triple failed");
-    triple
-}
-
 fn sys_root_path() -> PathBuf {
     env::var("SYSROOT")
         .ok()
