diff --git a/CMakeLists.txt b/CMakeLists.txt
index 2e179bf5..0293abfb 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -74,21 +74,23 @@ set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} "${CMAKE_SOURCE_DIR}/config/")

 if( MINGW OR UNIX )
     if (${CMAKE_CXX_COMPILER_ID} STREQUAL GNU)
-        ADD_DEFINITIONS(-Wall
-                    -Wcast-align
-                    -Wpointer-arith
-                    -Wformat-security
-                    -Wmissing-format-attribute
-                    -Woverloaded-virtual
-                    -W
-                   )
+        set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} "
+                    "-Wall "
+                    "-Wcast-align "
+                    "-Wpointer-arith "
+                    "-Wformat-security "
+                    "-Wmissing-format-attribute "
+                    "-Woverloaded-virtual "
+                    "-W "
+                    )
+        set(CMAKE_C_FLAGS "${CMAKE_CXX_FLAGS}")
     ENDIF()

     message(STATUS "Compiler info: ${CMAKE_CXX_COMPILER_ID} (${CMAKE_CXX_COMPILER}) ; version: ${CMAKE_CXX_COMPILER_VERSION}")
     IF ( CYGWIN OR (CMAKE_CXX_COMPILER_VERSION VERSION_GREATER 5.0))
-        ADD_DEFINITIONS( -std=gnu++98 ) # to support snprintf
+        set (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=gnu++98" ) # to support snprintf
     ELSE()
-        ADD_DEFINITIONS( -std=c++98 )
+        set (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++98" )
     ENDIF()

 ENDIF( MINGW OR UNIX )
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index dceee236..ba4c1b81 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -219,7 +219,6 @@ IF(NOT HAVE_TIMEGM )
 ENDIF( NOT HAVE_TIMEGM )

 SET( EXIV2_SRC        ${EXIV2_SRC}    getopt_win32.c )
-SET( LIBEXIV2_SRC     ${LIBEXIV2_SRC} getopt_win32.c )

 ##
 # msvn tuning
