--- a/lldb/source/Utility/ArchSpec.cpp
+++ b/lldb/source/Utility/ArchSpec.cpp
@@ -1114,8 +1118,12 @@ static bool isCompatibleEnvironment(llvm::Triple::EnvironmentType lhs,
       (rhs == llvm::Triple::Android && lhs == llvm::Triple::EABI) ||
       (lhs == llvm::Triple::GNUEABI && rhs == llvm::Triple::EABI) ||
       (rhs == llvm::Triple::GNUEABI && lhs == llvm::Triple::EABI) ||
+      (lhs == llvm::Triple::MuslEABI && rhs == llvm::Triple::EABI) ||
+      (rhs == llvm::Triple::MuslEABI && lhs == llvm::Triple::EABI) ||
       (lhs == llvm::Triple::GNUEABIHF && rhs == llvm::Triple::EABIHF) ||
-      (rhs == llvm::Triple::GNUEABIHF && lhs == llvm::Triple::EABIHF))
+      (rhs == llvm::Triple::GNUEABIHF && lhs == llvm::Triple::EABIHF) ||
+      (lhs == llvm::Triple::MuslEABIHF && rhs == llvm::Triple::EABIHF) ||
+      (rhs == llvm::Triple::MuslEABIHF && lhs == llvm::Triple::EABIHF))
     return true;
 
   return false;
--- a/lldb/source/Host/common/HostInfoBase.cpp
+++ b/lldb/source/Host/common/HostInfoBase.cpp
@@ -373,6 +373,18 @@
   arch_32.Clear();
   arch_64.Clear();
 
+  // Canonicalize armv7l and armv6l to their alias targets.  The 'l' is a linux addition.
+  if (triple.getOS() == llvm::Triple::Linux && triple.getArch() == llvm::Triple::arm) {
+    switch(triple.getSubArch()) {
+    default:
+      break;
+    case llvm::Triple::SubArchType::ARMSubArch_v6:
+    case llvm::Triple::SubArchType::ARMSubArch_v7:
+      triple.setArchName("arm");
+       break;
+    }
+  }
+
   switch (triple.getArch()) {
   default:
     arch_32.SetTriple(triple);
