--- a/scripts/Makefile.clang	2023-07-20 21:25:47.582934534 +0200
+++ b/scripts/Makefile.clang	2023-07-20 21:25:53.743020922 +0200
@@ -30,9 +30,5 @@
 # certain optimization flags it knows it has not implemented.
 # Make it behave more like gcc by erroring when these flags are encountered
 # so they can be implemented or wrapped in cc-option.
-CLANG_FLAGS	+= -Werror=unknown-warning-option
-CLANG_FLAGS	+= -Werror=ignored-optimization-argument
-CLANG_FLAGS	+= -Werror=option-ignored
-CLANG_FLAGS	+= -Werror=unused-command-line-argument
 KBUILD_CPPFLAGS	+= $(CLANG_FLAGS)
 export CLANG_FLAGS
