--- a/drivers/bluetooth/btusb.c	2025-01-09 13:33:55.000000000 +0100
+++ b/drivers/bluetooth/btusb.c	2025-01-10 15:58:01.314301326 +0100
@@ -2866,6 +2866,7 @@
 	return 0;
 }
 
+#define QCA_DEBUG_ACL_HANDLE 0x2EDC
 #define QCA_MEMDUMP_ACL_HANDLE 0x2EDD
 #define QCA_MEMDUMP_SIZE_MAX  0x100000
 #define QCA_MEMDUMP_VSE_CLASS 0x01
@@ -3050,6 +3051,13 @@
 {
 	if (handle_dump_pkt_qca(hdev, skb))
 		return 0;
+	/* We receive debug logs from chip as an ACL packets.
+	 * Instead of sending the data to ACL to decode the
+	 * received data, we are pushing them to the above layers
+	 * as a diagnostic packet.
+	 */
+	if (get_unaligned_le16(skb->data) == QCA_DEBUG_ACL_HANDLE)
+		return hci_recv_diag(hdev, skb);
 	return hci_recv_frame(hdev, skb);
 }
 
