--- a/arch/arm64/boot/dts/rockchip/rk3588-orangepi-5-plus.dts	2024-03-16 13:10:46.750621038 +0100
+++ b/arch/arm64/boot/dts/rockchip/rk3588-orangepi-5-plus.dts	2024-03-16 13:11:01.406450489 +0100
@@ -515,6 +515,7 @@
 		pinctrl-0 = <&pmic_pins>, <&rk806_dvs1_null>,
 			    <&rk806_dvs2_null>, <&rk806_dvs3_null>;
 		spi-max-frequency = <1000000>;
+		system-power-controller;
 
 		vcc1-supply = <&vcc5v0_sys>;
 		vcc2-supply = <&vcc5v0_sys>;
