--- a/arch/arm64/boot/dts/rockchip/rk3588-turing-rk1.dtsi	2025-04-27 22:39:28.350022364 +0200
+++ b/arch/arm64/boot/dts/rockchip/rk3588-turing-rk1.dtsi	2025-04-27 22:38:53.611614119 +0200
@@ -214,14 +214,16 @@
 };
 
 &package_thermal {
+	polling-delay = <1000>;
+
 	trips {
 		package_active1: trip-active1 {
-			temperature = <45000>;
+			temperature = <50000>;
 			hysteresis = <5000>;
 			type = "active";
 		};
 		package_active2: trip-active2 {
-			temperature = <50000>;
+			temperature = <55000>;
 			hysteresis = <5000>;
 			type = "active";
 		};
