--- a/drivers/net/wireless/ath/ath11k/ce.c	2025-02-19 20:49:01.678043121 +0100
+++ b/drivers/net/wireless/ath/ath11k/ce.c	2025-02-19 20:49:44.661588715 +0100
@@ -352,7 +352,7 @@
 		ret = ath11k_ce_rx_buf_enqueue_pipe(pipe, skb, paddr);
 
 		if (ret) {
-			ath11k_warn(ab, "failed to enqueue rx buf: %d\n", ret);
+			ath11k_dbg(ab, ATH11K_DBG_CE, "failed to enqueue rx buf: %d\n", ret);
 			dma_unmap_single(ab->dev, paddr,
 					 skb->len + skb_tailroom(skb),
 					 DMA_FROM_DEVICE);
