--- a/drivers/gpu/drm/panfrost/panfrost_gem_shrinker.c	2020-02-21 12:24:07.189824252 +0100
+++ b/drivers/gpu/drm/panfrost/panfrost_gem_shrinker.c	2020-02-21 12:24:19.797884935 +0100
@@ -78,7 +78,7 @@
 	mutex_unlock(&pfdev->shrinker_lock);
 
 	if (freed > 0)
-		pr_info_ratelimited("Purging %lu bytes\n", freed << PAGE_SHIFT);
+		pr_debug_ratelimited("Purging %lu bytes\n", freed << PAGE_SHIFT);
 
 	return freed;
 }
