diff --git a/Package@swift-4.swift b/Package@swift-4.swift
index e987d26..ee10cb6 100644
--- a/Package@swift-4.swift
+++ b/Package@swift-4.swift
@@ -5,7 +5,7 @@ import PackageDescription
 let package = Package(
     name: "Redis",
     products: [
-        .library(name: "Redis", targets: ["Redis"]),
+        .library(name: "Redis", type: .dynamic, targets: ["Redis"]),
     ],
     targets: [
         .target(name: "Redis"),
