--- a/CMakeLists.txt	2016-08-18 21:53:42.386865427 +0200
+++ b/CMakeLists.txt	2016-08-18 21:54:29.886816724 +0200
@@ -77,7 +77,7 @@
 pkg_check_modules(GSTREAMER_TAG REQUIRED gstreamer-tag-1.0)
 pkg_check_modules(LIBGPOD libgpod-1.0>=0.7.92)
 pkg_check_modules(LIBMTP libmtp>=1.0)
-pkg_check_modules(LIBMYGPO_QT5 libmygpo-qt5>=1.0.9)
+pkg_check_modules(LIBMYGPO_QT5 libmygpo-qt5>=1.0.8)
 pkg_check_modules(LIBPULSE libpulse)
 pkg_check_modules(LIBXML libxml-2.0)
 pkg_check_modules(SPOTIFY libspotify>=12.1.45)
