--- a/CMakeLists.txt	2016-08-25 18:53:45.237154134 +0200
+++ b/CMakeLists.txt	2016-08-25 18:56:27.137016771 +0200
@@ -371,18 +371,21 @@
   endif (NOT APPLE)
 endif (USE_SYSTEM_QXT)
 
-# Use system gmock if it's available
-# We need to look for both gmock and gtest
-find_path(GMOCK_INCLUDE_DIRS gmock/gmock.h)
-find_library(GMOCK_LIBRARIES gmock)
-if(GMOCK_INCLUDE_DIRS)
-  find_path(GTEST_INCLUDE_DIRS gtest/gtest.h)
-  find_library(GTEST_LIBRARIES gtest)
-  if(GTEST_INCLUDE_DIRS)
-    set(USE_SYSTEM_GMOCK 1)
-    set(GMOCK_LIBRARIES ${GMOCK_LIBRARIES} ${GTEST_LIBRARIES})
-  endif(GTEST_INCLUDE_DIRS)
-endif(GMOCK_INCLUDE_DIRS)
+option(ENABLE_TESTS "If you want to compile the tests" OFF)
+if (ENABLE_TESTS)
+  # Use system gmock if it's available
+  # We need to look for both gmock and gtest
+  find_path(GMOCK_INCLUDE_DIRS gmock/gmock.h)
+  find_library(GMOCK_LIBRARIES gmock)
+  if(GMOCK_INCLUDE_DIRS)
+    find_path(GTEST_INCLUDE_DIRS gtest/gtest.h)
+    find_library(GTEST_LIBRARIES gtest)
+    if(GTEST_INCLUDE_DIRS)
+      set(USE_SYSTEM_GMOCK 1)
+      set(GMOCK_LIBRARIES ${GMOCK_LIBRARIES} ${GTEST_LIBRARIES})
+    endif(GTEST_INCLUDE_DIRS)
+  endif(GMOCK_INCLUDE_DIRS)
+endif (ENABLE_TESTS)
 
 # Use the system's sha2 if it's available.
 find_path(SHA2_INCLUDE_DIRS sha2.h)
@@ -417,7 +420,9 @@
   add_subdirectory(3rdparty/qtwin)
   add_subdirectory(3rdparty/tinysvcmdns)
 endif (WIN32)
-add_subdirectory(tests)
+if (ENABLE_TESTS)
+  add_subdirectory(tests)
+endif (ENABLE_TESTS)
 add_subdirectory(dist)
 add_subdirectory(tools/ultimate_lyrics_parser)
 add_subdirectory(ext/libclementine-common)
