--- a/src/corelib/tools/qlist.h	2017-07-10 14:41:56.039610564 +0200
+++ b/src/corelib/tools/qlist.h	2017-07-10 14:42:23.782608281 +0200
@@ -111,7 +111,7 @@
     void remove(int i, int n);
     void move(int from, int to);
     inline int size() const Q_DECL_NOTHROW { return d->end - d->begin; }
-    inline bool isEmpty() const Q_DECL_NOTHROW { return d->end  == d->begin; }
+    inline bool isEmpty() const Q_DECL_NOTHROW { return d ? d->end  == d->begin : true; }
     inline void **at(int i) const Q_DECL_NOTHROW { return d->array + d->begin + i; }
     inline void **begin() const Q_DECL_NOTHROW { return d->array + d->begin; }
     inline void **end() const Q_DECL_NOTHROW { return d->array + d->end; }
