--- a/deps/spdlog/include/spdlog/details/os-inl.h	2022-11-02 15:47:02.817230055 +0100
+++ b/deps/spdlog/include/spdlog/details/os-inl.h	2022-11-02 15:47:53.533914829 +0100
@@ -237,7 +237,7 @@
     int fd = ::fileno(f);
 #    endif
 // 64 bits(but not in osx or cygwin, where fstat64 is deprecated)
-#    if (defined(__linux__) || defined(__sun) || defined(_AIX)) && (defined(__LP64__) || defined(_LP64))
+#    if (defined(__GLIBC__) || defined(__sun) || defined(_AIX)) && (defined(__LP64__) || defined(_LP64))
     struct stat64 st;
     if (::fstat64(fd, &st) == 0)
     {
