--- a/mount.h
+++ b/mount.h
@@ -20,6 +20,8 @@
 #ifndef _MOUNT_H_
 #define _MOUNT_H_
 
+#include <paths.h>
+
 /* exit status - bits below are ORed */
 #define EX_USAGE	1	/* incorrect invocation or permission */
 #define EX_SYSERR	2	/* out of memory, cannot fork, ... */
--- a/getcifsacl.c	2014-07-11 17:14:06.000000000 +0200
+++ b/getcifsacl.c	2015-10-06 22:54:41.616738999 +0200
@@ -32,8 +32,10 @@
 #include <stdlib.h>
 #include <stddef.h>
 #include <errno.h>
+#include <linux/limits.h>
 #include <limits.h>
 #include <ctype.h>
+#include <endian.h>
 #include <sys/xattr.h>
 #include "cifsacl.h"
 #include "idmap_plugin.h"
--- a/setcifsacl.c	2015-10-06 22:53:37.367735836 +0200
+++ b/setcifsacl.c	2015-10-06 22:55:00.090739909 +0200
@@ -32,7 +32,9 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <errno.h>
+#include <linux/limits.h>
 #include <limits.h>
+#include <endian.h>
 #include <ctype.h>
 #include <sys/xattr.h>
 
