--- glib/gbitlock.c	2015-01-29 19:26:55.060552703 +0100
+++ glib/gbitlock.c	2015-01-29 19:26:07.834522798 +0100
@@ -174,7 +174,7 @@
 static volatile gint g_bit_lock_contended[CONTENTION_CLASSES];
 
 #if (defined (i386) || defined (__amd64__))
-  #if __GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 5)
+  #if __GNUC__ > 5 || (__GNUC__ == 5 && __GNUC_MINOR__ >= 5)
     #define USE_ASM_GOTO 1
   #endif
 #endif
