--- a/frontend/OBSApp.cpp	2026-08-17 19:21:07.083508819 +0200
+++ b/frontend/OBSApp.cpp	2026-08-17 19:21:24.996758397 +0200
@@ -48,9 +48,9 @@
 #endif
 
 #include <chrono>
+#include <sstream>
 
 #ifdef _WIN32
-#include <sstream>
 #define WIN32_LEAN_AND_MEAN
 #include <windows.h>
 #else
--- a/frontend/widgets/OBSBasic.cpp	2026-08-17 19:19:53.342484301 +0200
+++ b/frontend/widgets/OBSBasic.cpp	2026-08-17 19:20:38.361109196 +0200
@@ -61,9 +61,7 @@
 #include <QWidgetAction>
 
 #include <mutex>
-#ifdef _WIN32
 #include <sstream>
-#endif
 #include <string>
 #include <unordered_set>
 
--- a/frontend/widgets/OBSBasic_MainControls.cpp	2026-08-17 19:21:50.045107816 +0200
+++ b/frontend/widgets/OBSBasic_MainControls.cpp	2026-08-17 19:21:57.016205144 +0200
@@ -50,9 +50,7 @@
 #include <nlohmann/json.hpp>
 #include <QDesktopServices>
 
-#ifdef _WIN32
 #include <sstream>
-#endif
 
 extern bool restart;
 extern bool restart_safe;
