--- a/configure.ac	2015-09-27 08:30:19.000000000 +0200
+++ b/configure.ac	2016-09-14 22:26:02.205599949 +0200
@@ -327,42 +327,7 @@
   AC_DEFINE([HAVE_RTLD_NEXT], [1], [Have RTLD_NEXT enabled libc])
 fi
 
-dnl we need to handle symbols differently based upon their version, 
-dnl but we have to know which symbols the libc supports first
-AC_MSG_CHECKING([libc path])
-echo "int main(void) { return 0; }" > libctest.c
-try_libc_path="$CC $CFLAGS $CPPFLAGS $LDFLAGS -o libctest libctest.c > libctest.log 2>&1"
-try_link() {
-	(
-		echo "try_link: trying: ${try_libc_path} $1"
-		eval ${try_libc_path} $1
-		ret=$?
-		cat libctest.log
-		if test ${ret} -eq 0 ; then
-			echo "try_link: SUCCESS"
-		else
-			echo "try_link: FAILURE (ret = ${ret})"
-		fi
-		exit ${ret}
-	) 1>&AS_MESSAGE_LOG_FD
-}
-LIBC_PATH=$(AS_IF(
-	dnl GNU linker (bfd & gold)
-	[try_link -Wl,--verbose],
-		[$AWK '/ttempt to open/ { if (($(NF-1) ~ /\/libc\.so/) && ($NF == "succeeded")) LIBC = $(NF-1); }; END {print LIBC}' libctest.log],
-	dnl Solaris linker
-	[try_link -Wl,-m],
-		[set -- `$EGREP -o '/[[^ ]]*/libc.so' libctest.log`; echo $1]
-))
-rm -f libctest*
-if test x"$LIBC_PATH" = x || ! test -r "$LIBC_PATH" ; then
-  AC_MSG_ERROR([Unable to determine LIBC PATH ($LIBC_PATH)])
-fi
-AC_DEFINE_UNQUOTED([LIBC_PATH], ["$LIBC_PATH"], [Full path to the libc])
-AC_MSG_RESULT([$LIBC_PATH])
-AC_SUBST([LIBC_PATH])
-
-dnl when intercepting libc calls, we have to know the name of the 
+dnl when intercepting libc calls, we have to know the name of the
 dnl libc to load and search with dl*() calls
 AC_MSG_CHECKING([libc version])
 dnl the sed script at the end here looks funny but it's ok ...
@@ -370,8 +335,8 @@
 $CC $CFLAGS $CPPFLAGS $LDFLAGS -o libctest libctest.c
 LIBC_VERSION=$(
 	$READELF -d libctest | \
-	$EGREP 'NEEDED.* \@<:@libc\.so' | \
-	$AWK '{print $NF}' | [sed -e 's:\[::' -e 's:\]::']
+	$EGREP 'NEEDED.* \@<:@.*ld-musl.*\.so.1' | \
+	$AWK '{print $5}' | [sed -e 's:\[::' -e 's:\]::']
 )
 rm -f libctest*
 if test "$LIBC_VERSION"x = x ; then
@@ -379,6 +344,9 @@
 fi
 AC_MSG_RESULT([$LIBC_VERSION])
 AC_DEFINE_UNQUOTED([LIBC_VERSION], ["$LIBC_VERSION"], [Name of libc to hook into])
+AC_DEFINE_UNQUOTED([LIBC_PATH], ["$LIBC_VERSION"], [Name of libc to hook into])
+LIBC_PATH="$LIBC_VERSION"
+AC_SUBST([LIBC_PATH])
 
 dnl Check for /proc/ features
 AC_DEFUN([SB_CHECK_PATH],[dnl
