--- a/external/libgit2/src/util.c	2016-10-27 17:22:27.236232686 +0200
+++ b/external/libgit2/src/util.c	2016-10-27 17:24:17.761229606 +0200
@@ -633,7 +633,7 @@
 	defined(__gnu_hurd__) || defined(__ANDROID_API__) || \
 	defined(__sun) || defined(__CYGWIN__) || \
 	(__GLIBC__ == 2 && __GLIBC_MINOR__ < 8) || \
-	(defined(_MSC_VER) && _MSC_VER < 1500)
+	(defined(_MSC_VER) && _MSC_VER < 1500) || !defined(__GLIBC__)
 	git__insertsort_r(els, nel, elsize, NULL, cmp, payload);
 #elif defined(GIT_WIN32)
 	git__qsort_r_glue glue = { cmp, payload };
