--- flang/tools/flang-driver/CMakeLists.txt.bak	2025-02-22 18:26:25.796302118 +0100
+++ flang/tools/flang-driver/CMakeLists.txt	2025-02-22 18:26:34.204211478 +0100
@@ -43,7 +43,3 @@
 endif()
 
 install(TARGETS flang DESTINATION "${CMAKE_INSTALL_BINDIR}")
-
-# Keep "flang-new" as a symlink for backwards compatiblity. Remove once "flang"
-# is a widely adopted name.
-add_flang_symlink(flang-new flang)
