--- a/flang/runtime/CMakeLists.txt	2025-03-05 21:34:30.350898328 +0100
+++ b/flang/runtime/CMakeLists.txt	2025-03-05 21:33:45.083369923 +0100
@@ -315,6 +315,10 @@
 endif()
 set_target_properties(FortranRuntime PROPERTIES FOLDER "Flang/Runtime Libraries")
 
+target_compile_options(FortranRuntime PRIVATE
+  $<$<COMPILE_LANGUAGE:CXX>:-fno-exceptions -fno-rtti -fno-unwind-tables -fno-asynchronous-unwind-tables>
+)
+
 # If FortranRuntime is part of a Flang build (and not a separate build) then
 # add dependency to make sure that Fortran runtime library is being built after
 # we have the Flang compiler available.  This also includes the MODULE files
