--- a/vendor/github.com/seccomp/libseccomp-golang/seccomp_internal.go	2021-04-25 16:04:23.069169154 +0200
+++ b/vendor/github.com/seccomp/libseccomp-golang/seccomp_internal.go	2021-04-25 16:05:51.348833375 +0200
@@ -532,8 +532,6 @@
 func actionFromNative(a C.uint32_t) (ScmpAction, error) {
 	aTmp := a & 0xFFFF
 	switch a & 0xFFFF0000 {
-	case C.C_ACT_KILL:
-		return ActKill, nil
 	case C.C_ACT_KILL_PROCESS:
 		return ActKillProcess, nil
 	case C.C_ACT_KILL_THREAD:
