diff --git a/Source/SwiftLintFramework/Extensions/NSRegularExpression+SwiftLint.swift b/Source/SwiftLintFramework/Extensions/NSRegularExpression+SwiftLint.swift
index a7033ed..aab9055 100644
--- a/Source/SwiftLintFramework/Extensions/NSRegularExpression+SwiftLint.swift
+++ b/Source/SwiftLintFramework/Extensions/NSRegularExpression+SwiftLint.swift
@@ -8,11 +8,6 @@
 
 import Foundation
 
-#if os(Linux)
-public typealias NSRegularExpression = RegularExpression
-public typealias NSTextCheckingResult = TextCheckingResult
-#endif
-
 private var regexCache = [RegexCacheKey: NSRegularExpression]()
 
 private struct RegexCacheKey: Hashable {
