diff --git a/Source/SwiftLintFramework/Extensions/NSRegularExpression+SwiftLint.swift b/Source/SwiftLintFramework/Extensions/NSRegularExpression+SwiftLint.swift
index aa6ea29..9adc9e0 100644
--- a/Source/SwiftLintFramework/Extensions/NSRegularExpression+SwiftLint.swift
+++ b/Source/SwiftLintFramework/Extensions/NSRegularExpression+SwiftLint.swift
@@ -9,7 +9,6 @@
 import Foundation
 
 #if os(Linux)
-public typealias NSRegularExpression = RegularExpression
 public typealias NSTextCheckingResult = TextCheckingResult
 #endif
 
