diff --git a/Source/SwiftLintFramework/Extensions/QueuedPrint.swift b/Source/SwiftLintFramework/Extensions/QueuedPrint.swift
index ab4d001..96aeadb 100644
--- a/Source/SwiftLintFramework/Extensions/QueuedPrint.swift
+++ b/Source/SwiftLintFramework/Extensions/QueuedPrint.swift
@@ -16,11 +16,9 @@ private let outputQueue: DispatchQueue = {
         target: .global(qos: .userInteractive)
     )
 
-    #if !os(Linux)
     atexit_b {
         queue.sync(flags: .barrier) {}
     }
-    #endif
 
     return queue
 }()
