diff --git a/Package.swift b/Package.swift
index 4468bb99..4c85f602 100644
--- a/Package.swift
+++ b/Package.swift
@@ -7,27 +7,15 @@ let package = Package(
         .executable(name: "swiftlint", targets: ["swiftlint"]),
         .library(name: "SwiftLintFramework", targets: ["SwiftLintFramework"])
     ],
-    dependencies: [
-        .package(url: "https://github.com/Carthage/Commandant.git", from: "0.13.0"),
-        .package(url: "https://github.com/jpsim/SourceKitten.git", from: "0.19.1"),
-        .package(url: "https://github.com/jpsim/Yams.git", from: "0.5.0"),
-        .package(url: "https://github.com/scottrhoyt/SwiftyTextTable.git", from: "0.8.0"),
-    ],
     targets: [
         .target(
             name: "swiftlint",
             dependencies: [
-                "Commandant",
                 "SwiftLintFramework",
-                "SwiftyTextTable",
             ]
         ),
         .target(
-            name: "SwiftLintFramework",
-            dependencies: [
-                "SourceKittenFramework",
-                "Yams",
-            ]
+            name: "SwiftLintFramework"
         ),
         .testTarget(
             name: "SwiftLintFrameworkTests",
