diff --git a/Package.swift b/Package.swift
index 1f5e2d3..dc4b41d 100644
--- a/Package.swift
+++ b/Package.swift
@@ -8,10 +8,5 @@ let package = Package(
       dependencies: [
         .Target(name: "SwiftLintFramework")
       ]),
-  ],
-  dependencies: [
-    .Package(url: "https://github.com/jpsim/SourceKitten.git", majorVersion: 0, minor: 17),
-    .Package(url: "https://github.com/behrang/YamlSwift.git", majorVersion: 3),
-    .Package(url: "https://github.com/scottrhoyt/SwiftyTextTable.git", majorVersion: 0, minor: 5),
   ]
 )
