diff --git a/Source/swiftlint/Commands/AutoCorrectCommand.swift b/Source/swiftlint/Commands/AutoCorrectCommand.swift
index 45182302..b133205e 100644
--- a/Source/swiftlint/Commands/AutoCorrectCommand.swift
+++ b/Source/swiftlint/Commands/AutoCorrectCommand.swift
@@ -45,7 +45,7 @@ struct AutoCorrectCommand: CommandProtocol {
                 queuedPrint(correctionLogs.joined(separator: "\n"))
             }
             if options.format {
-                let formattedContents = linter.file.format(trimmingTrailingWhitespace: true,
+                let formattedContents = try! linter.file.format(trimmingTrailingWhitespace: true,
                                                            useTabs: useTabs,
                                                            indentWidth: indentWidth)
                 _ = try? formattedContents
