diff --git a/Source/SwiftLintFramework/Models/Command.swift b/Source/SwiftLintFramework/Models/Command.swift
index 3de832ad..0e2899f6 100644
--- a/Source/SwiftLintFramework/Models/Command.swift
+++ b/Source/SwiftLintFramework/Models/Command.swift
@@ -66,7 +66,7 @@ public struct Command: Equatable {
 
     public init?(string: NSString, range: NSRange) {
         let scanner = Scanner(string: string.substring(with: range))
-        _ = scanner.scanString(string: "swiftlint:")
+        _ = scanner.scanString("swiftlint:")
         guard let actionAndModifierString = scanner.scanUpToString(" ") else {
             return nil
         }
@@ -84,7 +84,7 @@ public struct Command: Equatable {
         line = lineAndCharacter.line
         character = lineAndCharacter.character
 
-        let hasModifier = actionAndModifierScanner.scanString(string: ":") != nil
+        let hasModifier = actionAndModifierScanner.scanString(":") != nil
 
         // Modifier
         if hasModifier {
diff --git a/Source/SwiftLintFramework/Rules/IdentifierNameRule.swift b/Source/SwiftLintFramework/Rules/IdentifierNameRule.swift
index d038b7e9..4648d524 100644
--- a/Source/SwiftLintFramework/Rules/IdentifierNameRule.swift
+++ b/Source/SwiftLintFramework/Rules/IdentifierNameRule.swift
@@ -126,7 +126,7 @@ fileprivate extension String {
         guard characters.count > 1 else {
             return true
         }
-        let range = secondIndex..<characters.index(after: secondIndex)
+        let range: Range<Index> = secondIndex..<characters.index(after: secondIndex)
         let secondCharacter = substring(with: range)
         return secondCharacter.isLowercase()
     }
diff --git a/Source/SwiftLintFramework/Extensions/NSTextCheckingResult+SwiftLint.swift b/Source/SwiftLintFramework/Extensions/NSTextCheckingResult+SwiftLint.swift
index e69de29b..8044abc6 100644
--- a/Source/SwiftLintFramework/Extensions/NSTextCheckingResult+SwiftLint.swift
+++ b/Source/SwiftLintFramework/Extensions/NSTextCheckingResult+SwiftLint.swift
@@ -0,0 +1,9 @@
+import Foundation
+
+#if swift(>=4.0)
+extension NSTextCheckingResult {
+    public func rangeAt(_ index: Int) -> NSRange {
+        return range(at: index)
+    }
+}
+#endif
