--- a/flang/CMakeLists.txt	2026-03-27 17:08:57.738519888 +0100
+++ b/flang/CMakeLists.txt	2026-03-27 17:13:13.139967546 +0100
@@ -497,7 +497,7 @@
 include(FlangCommon)
 include(GetClangResourceDir)
 
-get_clang_resource_dir(HEADER_BINARY_DIR PREFIX ${LLVM_LIBRARY_OUTPUT_INTDIR}/.. SUBDIR include)
+get_clang_resource_dir(HEADER_BINARY_DIR PREFIX ${LLVM_LIBRARY_OUTPUT_INTDIR} SUBDIR include)
 
 if (FLANG_INCLUDE_TESTS)
   add_compile_definitions(FLANG_INCLUDE_TESTS=1)
