--- a/src/plugins/part.c	2017-11-09 22:43:30.664929262 +0000
+++ b/src/plugins/part.c	2017-11-09 22:49:10.879975116 +0000
@@ -1354,7 +1355,7 @@
         real_flags |= 0x4000000000000000; /* 1 << 62 */
     if (flags & BD_PART_FLAG_GPT_NO_AUTOMOUNT)
         real_flags |= 0x8000000000000000; /* 1 << 63 */
-    mask_str = g_strdup_printf ("%.16"__PRI64_PREFIX"x", real_flags);
+    mask_str = g_strdup_printf ("%.16"__PRI64"x", real_flags);
 
     args[2] = g_strdup_printf ("%d:=:%s", part_num, mask_str);
     g_free (mask_str);
