diff --git a/Sources/Yams/Resolver.swift b/Sources/Yams/Resolver.swift
index 4616083..106fc0b 100644
--- a/Sources/Yams/Resolver.swift
+++ b/Sources/Yams/Resolver.swift
@@ -72,10 +72,6 @@ extension Resolver {
     ])
 }
 
-#if os(Linux)
-    typealias NSRegularExpression = RegularExpression
-#endif
-
 func pattern(_ string: String) -> NSRegularExpression {
     do {
         return try .init(pattern: string, options: [])
