diff --git a/Package@swift-4.swift b/Package@swift-4.swift
index 9441e85..e1eaeac 100644
--- a/Package@swift-4.swift
+++ b/Package@swift-4.swift
@@ -4,7 +4,7 @@ import PackageDescription
 let package = Package(
     name: "Yams",
     products: [
-      .library(name: "Yams", targets: ["Yams"])
+      .library(name: "Yams", type: .dynamic, targets: ["Yams"])
     ],
     targets: [
         .target(name: "CYaml"),
