diff --git a/Package.swift b/Package.swift
index 70f18ac..d8fa3bc 100644
--- a/Package.swift
+++ b/Package.swift
@@ -3,9 +3,5 @@ import PackageDescription
 let beta = Version(2,0,0, prereleaseIdentifiers: ["beta"])
 
 let package = Package(
-    name: "Leaf",
-    dependencies: [
-        .Package(url: "https://github.com/vapor/core.git", majorVersion: 2),
-        .Package(url: "https://github.com/vapor/node.git", majorVersion: 2),
-    ]
+    name: "Leaf"
 )
