--- a/configure.ac	2024-03-13 23:27:04.623854830 +0100
+++ b/configure.ac	2024-03-13 23:28:00.839416637 +0100
@@ -754,7 +754,7 @@
 dnl --------------------------------------------------
 save_cflags=$CFLAGS
 AC_MSG_CHECKING([neon support])
-CFLAGS="$save_cflags $OPT_CFLAGS -mfpu=neon -mcpu=cortex-a53"
+CFLAGS="$save_cflags $OPT_CFLAGS"
 HAVE_NEON=""
 AC_LINK_IFELSE(
   [AC_LANG_PROGRAM([#include "arm_neon.h"], [
@@ -988,14 +988,14 @@
         fi
 
         AC_MSG_CHECKING([neon support])
-        CXXFLAGS="$save_cxxflags -mfpu=neon -mcpu=cortex-a53"
+        CXXFLAGS="$save_cxxflags"
 
         TESS_NEON=""
         AC_LINK_IFELSE(
          [AC_LANG_PROGRAM([#include "arm_neon.h"],
                           [int32x4_t round = vdupq_n_s32(10);
                            return(0);])],
-         [TESS_NEON="-mfpu=neon -mcpu=cortex-a53 -D__ARM_NEON__"],
+         [TESS_NEON="-D__ARM_NEON__"],
          [TESS_NEON=""])
 
         if test "x$TESS_NEON" != x && test "x$enable_neon" != xno; then
@@ -2432,7 +2432,7 @@
     CAL_CFLAGS="$CAL_CFLAGS $HAVE_AVX2"
   fi
   if test "x$HAVE_NEON" != x; then
-    CAL_NEON_CFLAGS="-mfpu=neon -mcpu=cortex-a53"
+    CAL_NEON_CFLAGS=""
     CAL_CFLAGS="$CAL_CFLAGS $HAVE_NEON"
   fi
 fi
