--- a/configure.ac	2021-04-03 19:41:26.423041385 +0200
+++ b/configure.ac	2021-04-03 19:41:56.425246509 +0200
@@ -857,7 +857,7 @@
     else
 
       save_cxxflags="$CXXFLAGS"
-      cxxflags_to_try="-std=c++17 -stdlib=libstdc++"
+      cxxflags_to_try="-std=c++17"
       CXXFLAGS_TO_USE=""
 
       AC_MSG_CHECKING([supported C++ compiler flags])
@@ -984,7 +984,7 @@
       OCR_DEVS="\$(DD)ocr.dev \$(DD)hocr.dev \$(DD)pdfocr8.dev \$(DD)pdfocr24.dev \$(DD)pdfocr32.dev"
       LEPTONICAINCLUDE="include base/leptonica.mak"
       TESSERACTINCLUDE="include base/tesseract.mak"
-      TESSERACT_LIBS="-lstdc++"
+      TESSERACT_LIBS="-lc++"
     fi
     AC_LANG_POP()
   else 
