--- a/camel/camel-iconv.c
+++ b/camel/camel-iconv.c
@@ -38,6 +38,10 @@
 #include "camel-iconv.h"
 #include "iconv-detect.h"
 
+#ifndef ICONV_10646
+#define ICONV_10646 "iso-10646"
+#endif /*ICONV_10646*/
+
 #define cd(x)
 
 G_LOCK_DEFINE_STATIC (iconv);
--- a/configure
+++ b/configure
@@ -22511,11 +22511,11 @@
 
 			cd = iconv_open ("UTF-8", "ISO-2022-JP");
 			if (cd == (iconv_t) -1)
-				exit (1);
+				exit (0);
 			if (iconv (cd, &jp, &jp_len, &trans, &utf8_len) == -1 || jp_len != 0)
-				exit (1);
+				exit (0);
 			if (memcmp (utf8, transbuf, utf8_real_len) != 0)
-				exit (1);
+				exit (0);
 			return (0);}
 
 
