diff --git a/src/Ide/Plugin/GhcIde.hs b/src/Ide/Plugin/GhcIde.hs
index 608bc06..9690c0a 100644
--- a/src/Ide/Plugin/GhcIde.hs
+++ b/src/Ide/Plugin/GhcIde.hs
@@ -39,7 +39,7 @@ hover' ideState params = do
 
 commandAddSignature :: CommandFunction WorkspaceEdit
 commandAddSignature lf ide params
-    = executeAddSignatureCommand lf ide (ExecuteCommandParams "typesignature.add" (Just (List [toJSON params])) Nothing)
+    = commandHandler lf ide (ExecuteCommandParams "typesignature.add" (Just (List [toJSON params])) Nothing)
 
 -- ---------------------------------------------------------------------
 
