diff --git a/hls-plugin-api.cabal b/hls-plugin-api.cabal
index fee9ceed11..feaece622f 100644
--- a/hls-plugin-api.cabal
+++ b/hls-plugin-api.cabal
@@ -56,22 +56,6 @@ library
     , text
     , unordered-containers
 
-  if impl(ghc < 8.10.5)
-    build-depends:
-      ghc-api-compat ==8.6
-  elif impl(ghc == 8.10.5)
-    build-depends:
-      ghc-api-compat ==8.10.5
-  elif impl(ghc == 8.10.6)
-    build-depends:
-      ghc-api-compat ==8.10.6
-  elif impl(ghc == 8.10.7)
-    build-depends:
-      ghc-api-compat ==8.10.7
-  elif impl(ghc == 9.0.1)
-    build-depends:
-      ghc-api-compat ==9.0.1
-
   if os(windows)
     build-depends: Win32
 
diff --git a/src/Ide/Types.hs b/src/Ide/Types.hs
index ae3253f6f5..c83d2c6e89 100644
--- a/src/Ide/Types.hs
+++ b/src/Ide/Types.hs
@@ -41,7 +41,7 @@ import           Data.String
 import qualified Data.Text                       as T
 import           Data.Text.Encoding              (encodeUtf8)
 import           Development.IDE.Graph
-import           DynFlags                        (DynFlags)
+import           GHC                             (DynFlags)
 import           GHC.Generics
 import           Ide.Plugin.Config
 import           Ide.Plugin.Properties
