--- a/third_party/cmocka/cmocka.h	2025-02-28 15:26:25.455111376 +0100
+++ b/third_party/cmocka/cmocka.h	2025-02-28 15:26:41.698942457 +0100
@@ -111,7 +111,7 @@
     ((LargestIntegralType)(value))
 
 /* Smallest integral type capable of holding a pointer. */
-#if !defined(_UINTPTR_T) && !defined(_UINTPTR_T_DEFINED)
+#if !defined(_UINTPTR_T) && !defined(_UINTPTR_T_DEFINED) && !defined(HAVE_UINTPTR_T) && !defined(__UINTPTR_TYPE__)
 # if defined(_WIN32)
     /* WIN32 is an ILP32 platform */
     typedef unsigned int uintptr_t;
