--- wscript.bak	2016-09-02 11:40:25.991244228 +0200
+++ wscript	2016-09-02 11:40:51.815663828 +0200
@@ -116,9 +116,9 @@
 def options(ctx):
 	ctx.load('compiler_c vala')
 	ctx.add_option('--noopt', action='store_true', default=False, dest='noopt', help="Turn off compiler optimizations")
-	ctx.add_option('--debug', action='store_true', default=True, dest='debug', help="Turn on debugging symbols")
+	ctx.add_option('--debug', action='store_true', default=False, dest='debug', help="Turn on debugging symbols")
 	ctx.add_option('--no-debug', action='store_false', dest='debug', help="Turn off debugging symbols")
-	ctx.add_option('--no-ldconfig', action='store_false', default=True, dest='ldconfig', help="Don't run ldconfig after installation")
+	ctx.add_option('--no-ldconfig', action='store_false', default=False, dest='ldconfig', help="Don't run ldconfig after installation")
 	ctx.add_option('--platform', default=_PLATFORM, help="Target platform")
 	ctx.add_option('--with-experimental-api', action='store_true', default=False, dest='experimental', help="Include experimental API.")
 
