--- a/include/my_global.h	2017-07-08 15:08:17.251720719 +0200
+++ b/include/my_global.h	2017-07-08 15:08:32.074721449 +0200
@@ -805,17 +805,6 @@
 #define SIZE_T_MAX      (~((size_t) 0))
 #endif
 
-#ifndef isfinite
-#ifdef HAVE_FINITE
-#define isfinite(x) finite(x)
-#else
-#define finite(x) (1.0 / fabs(x) > 0.0)
-#endif /* HAVE_FINITE */
-#elif (__cplusplus >= 201103L)
-#include <cmath>
-static inline bool isfinite(double x) { return std::isfinite(x); }
-#endif /* isfinite */
-
 #ifndef HAVE_ISNAN
 #define isnan(x) ((x) != (x))
 #endif
