--- a/src/libqhullcpp/QhullIterator.h	2016-09-13 20:01:15.220326000 +0200
+++ b/src/libqhullcpp/QhullIterator.h	2016-09-13 20:01:25.569662283 +0200
@@ -16,8 +16,6 @@
 #include <assert.h>
 #include <string>
 #include <vector>
-//! Avoid dependence on <iterator>
-namespace std { struct bidirectional_iterator_tag; struct random_access_iterator_tag; }
 
 namespace orgQhull {
 
--- a/src/libqhullcpp/QhullLinkedList.h	2016-09-13 20:01:44.296257417 +0200
+++ b/src/libqhullcpp/QhullLinkedList.h	2016-09-13 20:01:51.020467015 +0200
@@ -9,8 +9,6 @@
 #ifndef QHULLLINKEDLIST_H
 #define QHULLLINKEDLIST_H
 
-namespace std { struct bidirectional_iterator_tag; struct random_access_iterator_tag; }
-
 #include "QhullError.h"
 
 #ifdef QHULL_USES_QT
