diff --git a/Sources/XCTest/Public/XCAbstractTest.swift b/Sources/XCTest/Public/XCAbstractTest.swift
index 464365c..104b2a4 100644
--- a/Sources/XCTest/Public/XCAbstractTest.swift
+++ b/Sources/XCTest/Public/XCAbstractTest.swift
@@ -34,7 +34,7 @@ open class XCTest {
 
     /// The test run object that executed the test, an instance of
     /// testRunClass. If the test has not yet been run, this will be nil.
-    open private(set) var testRun: XCTestRun? = nil
+    private(set) var testRun: XCTestRun? = nil
 
     /// The method through which tests are executed. Must be overridden by
     /// subclasses.
diff --git a/Sources/XCTest/Public/XCTestCase.swift b/Sources/XCTest/Public/XCTestCase.swift
index c6fc19b..cd7ec2b 100644
--- a/Sources/XCTest/Public/XCTestCase.swift
+++ b/Sources/XCTest/Public/XCTestCase.swift
@@ -127,7 +127,7 @@ open class XCTestCase: XCTest {
     /// class.
     open class func tearDown() {}
 
-    open var continueAfterFailure: Bool {
+    var continueAfterFailure: Bool {
         get {
             return true
         }
