diff --git a/src/Ide/Plugin/ModuleName.hs b/src/Ide/Plugin/ModuleName.hs
index 0348690b8..48bf577fb 100644
--- a/src/Ide/Plugin/ModuleName.hs
+++ b/src/Ide/Plugin/ModuleName.hs
@@ -36,7 +36,10 @@ import           Development.IDE.GHC.Compat (GenLocated (L), getSessionDynFlags,
                                              pm_parsed_source, unLoc)
 import           Ide.Types
 import           Language.LSP.Server
-import           Language.LSP.Types
+import           Language.LSP.Types         hiding
+                                            (SemanticTokenAbsolute (length, line),
+                                             SemanticTokenRelative (length),
+                                             SemanticTokensEdit (_start))
 import           Language.LSP.VFS           (virtualFileText)
 import           System.Directory           (canonicalizePath)
 import           System.FilePath            (dropExtension, splitDirectories,
