--- core/core_debug.cpp.bak	2017-11-10 10:21:39.753580137 +0000
+++ core/core_debug.cpp	2017-11-10 10:22:04.226520295 +0000
@@ -22,7 +22,7 @@
 
 Q_LOGGING_CATEGORY(KDECONNECT_CORE, "kdeconnect.core")
 
-#ifdef Q_OS_LINUX
+#if defined(__GLIBC__)
 #include <execinfo.h>
 #include <stdlib.h>
 #include <unistd.h>
@@ -30,7 +30,7 @@
 
 void logBacktrace()
 {
-#ifdef Q_OS_LINUX
+#if defined(__GLIBC__)
     void* array[32];
     size_t size = backtrace (array, 32);
     char** strings = backtrace_symbols (array, size);
