--- a/m4/id3_cxx.m4	2016-02-04 21:40:02.928232669 +0100
+++ b/m4/id3_cxx.m4	2016-02-04 21:40:39.353131738 +0100
@@ -81,7 +81,7 @@
 
   dnl Check whether we have bool
   AC_MSG_CHECKING(whether C++ has bool)
-  AC_TRY_RUN([main() { bool b1=true; bool b2=false; }],
+  AC_TRY_RUN([int main() { bool b1=true; bool b2=false; return 0; }],
              [ AC_MSG_RESULT(yes) ],
              [ AC_MSG_RESULT(no)
                AC_DEFINE(CXX_HAS_NO_BOOL) ],
