--- a/cbits/sse-42.c	2018-03-10 14:54:46.425893805 +0100
+++ b/cbits/sse-42.c	2018-03-10 14:55:13.696840606 +0100
@@ -49,8 +49,7 @@
 #define _MODE (SIDD_UWORD_OPS | SIDD_CMP_EQUAL_EACH)
 
 static inline __m128i cmp_mask(__m128i a, __m128i b) {
-    const int mode = SIDD_UWORD_OPS | SIDD_CMP_EQUAL_EACH | SIDD_BIT_MASK;
-    return _mm_cmpistrm(a, b, mode);
+    return _mm_cmpistrm(a, b, SIDD_UWORD_OPS | SIDD_CMP_EQUAL_EACH | SIDD_BIT_MASK);
 }
 
 static inline int32_t line_result(uint32_t m, int start) {
