--- a/src/libunwind/build.rs
+++ b/src/libunwind/build.rs
@@ -12,7 +12,6 @@
     } else if target.contains("linux") {
         if target.contains("musl") {
             // linking for musl is handled in lib.rs
-            llvm_libunwind::compile();
         } else if !target.contains("android") {
             println!("cargo:rustc-link-lib=gcc_s");
         }
--- a/src/libunwind/lib.rs
+++ b/src/libunwind/lib.rs
@@ -36,6 +36,5 @@
 }
 
 #[cfg(target_env = "musl")]
-#[link(name = "unwind", kind = "static", cfg(target_feature = "crt-static"))]
-#[link(name = "gcc_s", cfg(not(target_feature = "crt-static")))]
+#[link(name = "unwind")]
 extern {}
