--- a/src/librustc_codegen_llvm/back/link.rs
+++ b/src/librustc_codegen_llvm/back/link.rs
@@ -1259,7 +1259,11 @@
             None => continue,
         };
         match lib.kind {
-            NativeLibraryKind::NativeUnknown => cmd.link_dylib(&name.as_str()),
+            NativeLibraryKind::NativeUnknown => if sess.crt_static() {
+                                                    cmd.link_staticlib(&name.as_str())
+                                                } else {
+                                                    cmd.link_dylib(&name.as_str())
+                                                },
             NativeLibraryKind::NativeFramework => cmd.link_framework(&name.as_str()),
             NativeLibraryKind::NativeStaticNobundle => cmd.link_staticlib(&name.as_str()),
             NativeLibraryKind::NativeStatic => cmd.link_whole_staticlib(&name.as_str(),
@@ -1633,7 +1637,11 @@
                 continue
             }
             match lib.kind {
-                NativeLibraryKind::NativeUnknown => cmd.link_dylib(&name.as_str()),
+                NativeLibraryKind::NativeUnknown => if sess.crt_static() {
+                                                        cmd.link_staticlib(&name.as_str())
+                                                    } else {
+                                                        cmd.link_dylib(&name.as_str())
+                                                    },
                 NativeLibraryKind::NativeFramework => cmd.link_framework(&name.as_str()),
                 NativeLibraryKind::NativeStaticNobundle => {
                     // Link "static-nobundle" native libs only if the crate they originate from
