--- a/src/tools/rls/rls-analysis/src/loader.rs	2019-10-03 00:02:56.309997506 +0200
+++ b/src/tools/rls/rls-analysis/src/loader.rs	2019-10-03 00:03:32.776253352 +0200
@@ -77,7 +77,7 @@
         let libs_path =
             sys_root_path.join("lib").join("rustlib").join(&target_triple).join("analysis");
 
-        let src_path = sys_root_path.join("lib").join("rustlib").join("src").join("rust");
+        let src_path = sys_root_path.join("lib").join("rustlib");
 
         vec![SearchDirectory::new(libs_path, Some(src_path)), SearchDirectory::new(deps_path, None)]
     }
