--- a/library/unwind/build.rs
+++ b/library/unwind/build.rs
@@ -14,7 +14,6 @@
     } else if target.contains("linux") {
         // linking for Linux is handled in lib.rs
         if target.contains("musl") {
-            llvm_libunwind::compile();
         }
     } else if target.contains("freebsd") {
         println!("cargo:rustc-link-lib=gcc_s");
--- a/library/unwind/src/lib.rs
+++ b/library/unwind/src/lib.rs
@@ -19,8 +19,7 @@
 }
 
 #[cfg(target_env = "musl")]
-#[link(name = "unwind", kind = "static", cfg(target_feature = "crt-static"))]
-#[link(name = "gcc_s", cfg(not(target_feature = "crt-static")))]
+#[link(name = "unwind")]
 extern "C" {}
 
 #[cfg(target_os = "redox")]
