--- a/src/etc/mklldeps.py	2016-03-23 00:30:38.000000000 +0100
+++ b/src/etc/mklldeps.py	2016-04-12 11:40:09.465964118 +0200
@@ -99,6 +99,7 @@
     # right standard library.
     if use_libcpp != "0" or 'stdlib=libc++' in out:
         f.write("#[cfg_attr(not(target_env = \"msvc\"), link(name = \"c++\"))]\n")
+        f.write("#[cfg_attr(not(target_env = \"msvc\"), link(name = \"c++abi\"))]\n")
     else:
         f.write("#[cfg_attr(not(target_env = \"msvc\"), link(name = \"" + stdcpp_name + "\"))]\n")
 
