From 066e96f3a3845b778230d866eeef3eb333107feb Mon Sep 17 00:00:00 2001
From: Jake Goulding <jake.goulding@gmail.com>
Date: Sat, 24 Sep 2016 10:44:35 -0400
Subject: [PATCH] [4.0] Configure allows LLVM 4.0

---
 configure | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/configure b/configure
index 56414bb..5cfbeb8 100755
--- a/configure
+++ b/configure
@@ -1062,6 +1062,9 @@ then
         (3.[7-9]*)
             msg "found ok version of LLVM: $LLVM_VERSION"
             ;;
+        (4.[0]*)
+            msg "found ok version of LLVM: $LLVM_VERSION"
+            ;;
         (*)
             err "bad LLVM version: $LLVM_VERSION, need >=3.7"
             ;;
