From 8678624ecb9ed58c3ab6c76cc662f8f5282e5457 Mon Sep 17 00:00:00 2001
From: Jake Goulding <jake.goulding@gmail.com>
Date: Sun, 6 Nov 2016 10:03:44 -0500
Subject: [PATCH] [4.0][Lost Functionality] Disable target and feature listing
 support

---
 src/rustllvm/PassWrapper.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/rustllvm/PassWrapper.cpp b/src/rustllvm/PassWrapper.cpp
index 87fcfc2..9ff7330 100644
--- a/src/rustllvm/PassWrapper.cpp
+++ b/src/rustllvm/PassWrapper.cpp
@@ -243,7 +243,7 @@ from_rust(LLVMRustCodeGenOptLevel level)
   }
 }
 
-#if LLVM_RUSTLLVM
+#if LLVM_RUSTLLVM && false
 /// getLongestEntryLength - Return the length of the longest entry in the table.
 ///
 static size_t getLongestEntryLength(ArrayRef<SubtargetFeatureKV> Table) {
