--- a/src/librustc_trans/back/linker.rs	2016-11-11 14:41:58.468065080 +0100
+++ b/src/librustc_trans/back/linker.rs	2016-11-11 14:42:41.777985347 +0100
@@ -194,7 +194,7 @@
     }
 
     fn debuginfo(&mut self) {
-        // Don't do anything special here for GNU-style linkers.
+        self.cmd.arg("-Wl,-nostrip");
     }
 
     fn no_default_libraries(&mut self) {
--- a/src/librustc_trans/back/link.rs	2016-11-11 16:04:06.436523378 +0100
+++ b/src/librustc_trans/back/link.rs	2016-11-11 16:10:08.212914205 +0100
@@ -767,8 +767,11 @@
     // Pass optimization flags down to the linker.
     cmd.optimize();
 
-    // Pass debuginfo flags down to the linker.
-    cmd.debuginfo();
+    // do not strip when debugging
+    if sess.opts.debuginfo != config::NoDebugInfo {
+        // Pass debuginfo flags down to the linker.
+        cmd.debuginfo();
+    }
 
     // We want to prevent the compiler from accidentally leaking in any system
     // libraries, so we explicitly ask gcc to not link to any libraries by
