--- a/src/bootstrap/compile.rs	2016-10-08 14:32:50.900901529 +0200
+++ b/src/bootstrap/compile.rs	2016-10-08 14:33:04.700902208 +0200
@@ -91,10 +91,6 @@
         t!(fs::create_dir_all(&libdir));
     }
     add_to_sysroot(&out_dir, &libdir);
-
-    if target.contains("musl") && !target.contains("mips") {
-        copy_musl_third_party_objects(build, &libdir);
-    }
 }
 
 /// Copies the crt(1,i,n).o startup objects
