--- a/src/bootstrap/compile.rs	2016-10-08 14:32:50.900901529 +0200
+++ b/src/bootstrap/compile.rs	2016-10-08 14:33:04.700902208 +0200
@@ -89,19 +89,6 @@
         t!(fs::create_dir_all(&libdir));
     }
     add_to_sysroot(&out_dir, &libdir);
-
-    if target.contains("musl") && !target.contains("mips") {
-        copy_musl_third_party_objects(build, target, &libdir);
-    }
-}
-
-/// Copies the crt(1,i,n).o startup objects
-///
-/// Only required for musl targets that statically link to libc
-fn copy_musl_third_party_objects(build: &Build, target: &str, into: &Path) {
-    for &obj in &["crt1.o", "crti.o", "crtn.o"] {
-        copy(&build.musl_root(target).unwrap().join("lib").join(obj), &into.join(obj));
-    }
 }
 
 /// Build and prepare startup objects like rsbegin.o and rsend.o
