--- a/src/bootstrap/builder.rs	2018-07-13 16:07:32.027885858 +0200
+++ b/src/bootstrap/builder.rs	2018-07-13 16:07:52.857860330 +0200
@@ -768,8 +768,7 @@
             }
         }
 
-        if cmd == "build" && mode == Mode::Libstd
-            && self.config.extended && compiler.is_final_stage(self)
+        if cmd == "build" && mode == Mode::Libstd && compiler.is_final_stage(self)
         {
             cargo.env("RUSTC_SAVE_ANALYSIS", "api".to_string());
         }
