--- a/src/compiler-rt/CMakeLists.txt	2016-08-23 14:21:49.488904682 +0200
+++ b/src/compiler-rt/CMakeLists.txt	2016-08-23 14:24:52.712397348 +0200
@@ -107,7 +107,7 @@
 
   # Make use of LLVM CMake modules.
   file(TO_CMAKE_PATH ${LLVM_BINARY_DIR} LLVM_BINARY_DIR_CMAKE_STYLE)
-  set(LLVM_CMAKE_PATH "${LLVM_BINARY_DIR_CMAKE_STYLE}/share/llvm/cmake")
+  set(LLVM_CMAKE_PATH "/usr/lib/cmake/llvm")
   list(APPEND CMAKE_MODULE_PATH "${LLVM_CMAKE_PATH}")
   # Get some LLVM variables from LLVMConfig.
   include("${LLVM_CMAKE_PATH}/LLVMConfig.cmake")
--- a/configure	2016-08-23 14:20:59.412047896 +0200
+++ b/configure	2016-08-23 14:21:20.311987852 +0200
@@ -991,7 +991,7 @@
     LLVM_VERSION=$($LLVM_CONFIG --version)
 
     case $LLVM_VERSION in
-        (3.[7-8]*)
+        (3.[7-9]*)
             msg "found ok version of LLVM: $LLVM_VERSION"
             ;;
         (*)
