--- a/src/librustc/util/common.rs	2016-08-23 17:28:57.292060998 +0200
+++ b/src/librustc/util/common.rs	2016-08-23 17:29:06.715031657 +0200
@@ -248,13 +248,6 @@
     CString::new(p.to_str().unwrap()).unwrap()
 }
 
-// FIXME(stage0): remove this
-// HACK: this is needed because the interpretation of slice
-// patterns changed between stage0 and now.
-#[cfg(stage0)]
-pub fn slice_pat<'a, 'b, T>(t: &'a &'b [T]) -> &'a &'b [T] {
-    t
-}
 #[cfg(not(stage0))]
 pub fn slice_pat<'a, 'b, T>(t: &'a &'b [T]) -> &'b [T] {
     *t
--- a/src/libstd/lib.rs	2016-08-23 17:16:50.038496188 +0200
+++ b/src/libstd/lib.rs	2016-08-23 17:28:35.515129627 +0200
@@ -469,14 +469,6 @@
 // because rustdoc only looks for these modules at the crate level.
 include!("primitive_docs.rs");
 
-// FIXME(stage0): remove this after a snapshot
-// HACK: this is needed because the interpretation of slice
-// patterns changed between stage0 and now.
-#[cfg(stage0)]
-fn slice_pat<'a, 'b, T>(t: &'a &'b [T]) -> &'a &'b [T] {
-    t
-}
-#[cfg(not(stage0))]
 fn slice_pat<'a, 'b, T>(t: &'a &'b [T]) -> &'b [T] {
     *t
 }
