--- a/src/librustc_llvm/build.rs	2016-12-16 16:52:55.131616336 +0100
+++ b/src/librustc_llvm/build.rs	2016-12-16 16:54:28.956361505 +0100
@@ -132,21 +132,10 @@
     let mut cmd = Command::new(&llvm_config);
     cmd.arg("--libs");
 
-    // Force static linking with "--link-static" if available.
-    let mut version_cmd = Command::new(&llvm_config);
-    version_cmd.arg("--version");
-    let version_output = output(&mut version_cmd);
-    let mut parts = version_output.split('.');
-    if let (Some(major), Some(minor)) = (parts.next().and_then(|s| s.parse::<u32>().ok()),
-                                         parts.next().and_then(|s| s.parse::<u32>().ok())) {
-        if major > 3 || (major == 3 && minor >= 8) {
-            cmd.arg("--link-static");
-        }
-    }
-
     if !is_crossed {
         cmd.arg("--system-libs");
     }
+
     cmd.args(&components[..]);
 
     for lib in output(&mut cmd).split_whitespace() {
@@ -163,12 +162,7 @@
             continue;
         }
 
-        let kind = if name.starts_with("LLVM") {
-            "static"
-        } else {
-            "dylib"
-        };
-        println!("cargo:rustc-link-lib={}={}", kind, name);
+        println!("cargo:rustc-link-lib={}={}", "dylib", name);
     }
 
     // LLVM ldflags
