--- src/librustc_llvm/build.rs.bak	2016-12-16 16:52:55.131616336 +0100
+++ src/librustc_llvm/build.rs	2016-12-16 16:54:28.956361505 +0100
@@ -130,7 +130,6 @@
     // of llvm-config, not the target that we're attempting to link.
     let mut cmd = Command::new(&llvm_config);
     cmd.arg("--libs");
-    cmd.arg("--link-static");
     if !is_crossed {
         cmd.arg("--system-libs");
     }
@@ -163,12 +162,7 @@
             continue;
         }
 
-        let kind = if name.starts_with("LLVM") {
-            "static"
-        } else {
-            "dylib"
-        };
-        println!("cargo:rustc-link-lib={}={}", kind, name);
+        println!("cargo:rustc-link-lib={}={}", "dylib", name);
     }
 
     // LLVM ldflags
