--- a/src/librustc_llvm/build.rs	2016-12-16 16:52:55.131616336 +0100
+++ b/src/librustc_llvm/build.rs	2016-12-16 16:54:28.956361505 +0100
@@ -163,12 +162,7 @@
             continue;
         }
 
-        let kind = if name.starts_with("LLVM") {
-            "static"
-        } else {
-            "dylib"
-        };
-        println!("cargo:rustc-link-lib={}={}", kind, name);
+        println!("cargo:rustc-link-lib={}={}", "dylib", name);
     }
 
     // LLVM ldflags
