--- a/src/rustllvm/RustWrapper.cpp	2016-12-16 16:49:46.181140277 +0100
+++ b/src/rustllvm/RustWrapper.cpp	2016-12-16 16:50:21.006042440 +0100
@@ -540,8 +540,8 @@
     unsigned RuntimeVer,
     const char* SplitName) {
     return wrap(Builder->createCompileUnit(Lang,
-                                           File,
-                                           Dir,
+                                           Builder->createFile(File,
+                                           Dir),
                                            Producer,
                                            isOptimized,
                                            Flags,
