--- a/src/librustc_back/target/linux_base.rs	2016-08-12 13:51:24.073585071 +0200
+++ b/src/librustc_back/target/linux_base.rs	2016-08-12 13:52:19.097583538 +0200
@@ -17,6 +17,7 @@
         executables: true,
         linker_is_gnu: true,
         has_rpath: true,
+        no_compiler_rt: true,
         pre_link_args: vec![
             // We want to be able to strip as much executable code as possible
             // from the linker command line, and this flag indicates to the
