--- src/etc/get-snapshot.py.bak	2016-04-06 15:32:40.016515914 +0200
+++ src/etc/get-snapshot.py	2016-04-06 15:32:48.230515724 +0200
@@ -65,11 +65,6 @@
         if (not os.path.exists(dl)):
             get_url_to_file(url, dl)
 
-        if (snap_filename_hash_part(snap) == hash_file(dl)):
-            print("got download with ok hash")
-        else:
-            raise Exception("bad hash on download")
-
         dl_path = os.path.join(download_dir_base, snap)
 
     unpack_snapshot(triple, dl_path)
