--- a/src/rustllvm/RustWrapper.cpp	2016-11-10 13:37:00.961622332 +0100
+++ b/src/rustllvm/RustWrapper.cpp	2016-11-10 13:40:58.618033958 +0100
@@ -911,10 +911,10 @@
 LLVMRustLinkInExternalBitcode(LLVMModuleRef dst, char *bc, size_t len) {
     Module *Dst = unwrap(dst);
     std::unique_ptr<MemoryBuffer> buf = MemoryBuffer::getMemBufferCopy(StringRef(bc, len));
-    ErrorOr<std::unique_ptr<Module>> Src =
-        llvm::getLazyBitcodeModule(std::move(buf), Dst->getContext());
+    Expected<std::unique_ptr<Module>> Src =
+        llvm::getOwningLazyBitcodeModule(std::move(buf), Dst->getContext());
     if (!Src) {
-        LLVMRustSetLastError(Src.getError().message().c_str());
+        LLVMRustSetLastError(toString(Src.takeError()).c_str());
         return false;
     }
 
