--- a/src/libunwind/libunwind.rs	2016-08-07 12:55:17.209674595 +0200
+++ b/src/libunwind/libunwind.rs	2016-08-07 12:55:40.933673934 +0200
@@ -108,7 +108,7 @@
                    target_env = "musl",
                    not(target_arch = "x86"),
                    not(target_arch = "x86_64"))),
-           link(name = "gcc_s"))]
+           link(name = "unwind"))]
 #[cfg_attr(all(target_os = "linux",
                target_env = "musl",
                any(target_arch = "x86", target_arch = "x86_64"),
