--- a/src/libunwind/libunwind.rs	2016-08-07 12:55:17.209674595 +0200
+++ b/src/libunwind/libunwind.rs	2016-08-07 12:55:40.933673934 +0200
@@ -108,7 +108,7 @@
                    target_env = "musl",
                    not(target_arch = "x86"),
                    not(target_arch = "x86_64"))),
-           link(name = "gcc_s"))]
+           link(name = "unwind"))]
 #[cfg_attr(all(target_os = "linux",
                target_env = "musl",
                any(target_arch = "x86", target_arch = "x86_64"),
--- a/src/librustc_back/target/x86_64_unknown_linux_gnu.rs	2016-08-12 13:50:06.698587227 +0200
+++ b/src/librustc_back/target/x86_64_unknown_linux_gnu.rs	2016-08-12 13:50:25.171586712 +0200
@@ -15,6 +15,7 @@
     base.cpu = "x86-64".to_string();
     base.max_atomic_width = 64;
     base.pre_link_args.push("-m64".to_string());
+    base.no_default_libraries = false;
 
     Target {
         llvm_target: "x86_64-unknown-linux-gnu".to_string(),
