--- a/src/widgets/kpropertiesdialog.cpp	2018-10-16 11:14:36.500153909 +0200
+++ b/src/widgets/kpropertiesdialog.cpp	2018-10-16 11:15:27.475188431 +0200
@@ -1166,9 +1166,9 @@
             l = new QLabel(d->m_frame);
             grid->addWidget(l, curRow++, 2);
             l->setTextInteractionFlags(Qt::TextSelectableByMouse | Qt::TextSelectableByKeyboard);
-            l->setText(mp->mountType());
 
             if (mp) {
+                l->setText(mp->mountType());
                 l = new QLabel(i18n("Mounted on:"), d->m_frame);
                 grid->addWidget(l, curRow, 0, Qt::AlignRight);
 
