--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -118,16 +118,19 @@
                       )
 
 if (NOT APPLE AND NOT WIN32)
-    find_package(X11)
-endif()
-
-set(HAVE_X11 ${X11_FOUND})
-if (HAVE_X11)
+    option(WITH_X11 "Build with X11 integration" ON)
+    if (WITH_X11)
+        find_package(X11 REQUIRED)
     if (QT_MAJOR_VERSION STREQUAL "5")
         find_package(Qt5 ${REQUIRED_QT_VERSION} CONFIG REQUIRED X11Extras)
     else()
         # qtx11extras_p.h is in Qt6Gui, which is implied by the find_package(QtWidgets) above
     endif()
+else()
+    set(WITH_X11 OFF)
+endif()
+
+
 endif()
 
 find_package(ACL) 
