--- src/widgets/kpropertiesdialog.cpp.bak	2021-07-12 14:47:25.200367001 +0200
+++ src/widgets/kpropertiesdialog.cpp	2021-07-12 14:47:43.167662807 +0200
@@ -2272,7 +2272,7 @@
         d->fileSystemSupportsACLs = fileSystemSupportsACL(path);
     }
     if (d->fileSystemSupportsACLs) {
-        std::for_each(theNotSpecials.begin(), theNotSpecials.end(), std::mem_fun(&QWidget::hide));
+        std::for_each(theNotSpecials.begin(), theNotSpecials.end(), std::mem_fn(&QWidget::hide));
         extendedACLs = new KACLEditWidget(&dlg);
         extendedACLs->setEnabled(d->canChangePermissions);
         vbox->addWidget(extendedACLs);
