diff --git a/libavformat/hls.c b/libavformat/hls.c
index f2b98a6d72da8..84a5e66149b17 100644
--- a/libavformat/hls.c
+++ b/libavformat/hls.c
@@ -753,7 +753,8 @@ static int test_segment(AVFormatContext *s, const AVInputFormat *in_fmt, struct
         if (in_fmt->extensions) {
             matchF =      av_match_ext(    seg->url, in_fmt->extensions)
                      + 2*(ff_match_url_ext(seg->url, in_fmt->extensions) > 0);
-            if(av_match_name("mp4", in_fmt->name)) {
+            // Youtube uses aac files with .ts extension
+            if(av_match_name("mp4", in_fmt->name) || av_match_name("aac", in_fmt->name)) {
                 matchF |=      av_match_ext(    seg->url, "ts,m2t,m2ts,mts,mpg,m4s,mpeg,mpegts")
                           + 2*(ff_match_url_ext(seg->url, "ts,m2t,m2ts,mts,mpg,m4s,mpeg,mpegts") > 0);
             }
