--- src/Language/Haskell/LSP/Types/DataTypesJSON.hs.baj	2020-03-23 21:41:21.811574101 +0100
+++ src/Language/Haskell/LSP/Types/DataTypesJSON.hs	2020-03-23 21:46:08.986999370 +0100
@@ -1152,12 +1152,6 @@
     _watchers :: List FileSystemWatcher
   } deriving (Show, Read, Eq)
 
-data FileSystemWatcher =
-  FileSystemWatcher {
-    _globPattern :: String,
-    _kind :: Maybe WatchKind
-  } deriving (Show, Read, Eq)
-
 data WatchKind =
   WatchKind {
     -- | Watch for create events
@@ -1168,6 +1162,12 @@
     _watchDelete :: Bool
   } deriving (Show, Read, Eq)
 
+data FileSystemWatcher =
+  FileSystemWatcher {
+    _globPattern :: String,
+    _kind :: Maybe WatchKind
+  } deriving (Show, Read, Eq)
+
 instance A.ToJSON WatchKind where
   toJSON wk = A.Number (createNum + changeNum + deleteNum)
     where
@@ -1183,8 +1183,8 @@
     | otherwise = mempty
   parseJSON _            = mempty
 
-deriveJSON lspOptions ''DidChangeWatchedFilesRegistrationOptions
 deriveJSON lspOptions ''FileSystemWatcher
+deriveJSON lspOptions ''DidChangeWatchedFilesRegistrationOptions
 
 -- ---------------------------------------------------------------------
 {-
