diff --git a/src/Language/Haskell/LSP/Types/WorkspaceEdit.hs b/src/Language/Haskell/LSP/Types/WorkspaceEdit.hs
index e80dca4..5c21bc2 100644
--- a/src/Language/Haskell/LSP/Types/WorkspaceEdit.hs
+++ b/src/Language/Haskell/LSP/Types/WorkspaceEdit.hs
@@ -153,13 +153,13 @@ instance Monoid WorkspaceEdit where
   mempty = WorkspaceEdit Nothing Nothing
   mappend (WorkspaceEdit a b) (WorkspaceEdit c d) = WorkspaceEdit (a <> c) (b <> d)
 
-deriveJSON lspOptions ''WorkspaceEdit
-
 #if __GLASGOW_HASKELL__ >= 804
 instance Semigroup WorkspaceEdit where
   (<>) = mappend
 #endif
 
+deriveJSON lspOptions ''WorkspaceEdit
+
 -- ---------------------------------------------------------------------
 
 -- | Applies a 'TextEdit' to some 'Text'.
