diff --git a/Package.swift b/Package.swift
index 2638a2f..ed1cbcb 100644
--- a/Package.swift
+++ b/Package.swift
@@ -1,9 +1,5 @@
 import PackageDescription
 
 let package = Package(
-    name: "LeafProvider",
-    dependencies: [
-        .Package(url: "https://github.com/vapor/vapor.git", majorVersion: 2),
-        .Package(url: "https://github.com/vapor/leaf.git", majorVersion: 2),
-    ]
+    name: "LeafProvider"
 )
