--- a/src/shared/path-lookup.c
+++ b/src/shared/path-lookup.c
@@ -888,6 +888,8 @@
                 return strv_new("/run/systemd/system-generators",
                                 "/etc/systemd/system-generators",
                                 "/usr/local/lib/systemd/system-generators",
+                                "/usr/lib/systemd/system-generators",
+                                "/lib/systemd/system-generators",
                                 SYSTEM_GENERATOR_PATH);
 
         case UNIT_FILE_GLOBAL:
