--- a/include/utils.h	2019-01-22 20:45:07.296750831 +0100
+++ b/include/utils.h	2019-01-22 20:45:30.369536179 +0100
@@ -8,12 +8,7 @@
 #include <libnftnl/common.h>
 
 #include "config.h"
-#ifdef HAVE_VISIBILITY_HIDDEN
-#	define __visible	__attribute__((visibility("default")))
-#	define EXPORT_SYMBOL(x)	typeof(x) (x) __visible;
-#else
-#	define EXPORT_SYMBOL
-#endif
+#define EXPORT_SYMBOL
 
 #define __noreturn	__attribute__((__noreturn__))
 
