--- a/spice-common/common/quic.c	2017-11-12 10:43:42.174433917 +0100
+++ b/spice-common/common/quic.c	2017-11-12 10:43:59.853400467 +0100
@@ -878,7 +878,7 @@
     /* The only valid values are 1, 3 and 5.
        0, 2 and 4 are obsolete and the rest of the
        values are considered out of the range. */
-    SPICE_VERIFY(evol == 1 || evol == 3 || evol == 5);
+    //SPICE_VERIFY(evol == 1 || evol == 3 || evol == 5);
     spice_assert(bpc <= 8 && bpc > 0);
 
     *ncounters = 8;
