--- a/project/cmake/scripts/linux/Macros.cmake	2017-02-05 14:57:02.489779420 +0100
+++ b/project/cmake/scripts/linux/Macros.cmake	2017-02-05 14:57:39.304758224 +0100
@@ -73,16 +73,16 @@
   string(STRIP "${${lib}_FILENAME}" ${lib}_FILENAME)
   if(NOT ${lib}_FILENAME)
     execute_process(COMMAND ${CMAKE_C_COMPILER} -nostdlib -o /dev/null -Wl,-t ${link_lib}
-                    OUTPUT_QUIET
-                    ERROR_VARIABLE _TMP_FILENAME)
+                    ERROR_QUIET
+                    OUTPUT_VARIABLE _TMP_FILENAME)
     string(REGEX MATCH ".*lib${liblow}.so" ${lib}_FILENAME ${_TMP_FILENAME})
   endif()
   if(${lib}_FILENAME)
-    execute_process(COMMAND objdump -p ${${lib}_FILENAME}
+    execute_process(COMMAND readelf -d ${${lib}_FILENAME}
                     COMMAND grep SONAME.*${liblow}
+                    COMMAND grep -oP "\\[\\K[^]]+"
                     ERROR_QUIET
                     OUTPUT_VARIABLE ${lib}_SONAME)
-    string(REPLACE "SONAME " "" ${lib}_SONAME ${${lib}_SONAME})
     string(STRIP ${${lib}_SONAME} ${lib}_SONAME)
     if(VERBOSE)
       message(STATUS "${lib} soname: ${${lib}_SONAME}")
