--- a/kstyle/breezeshadowhelper.cpp	2016-08-23 14:28:52.000000000 +0200
+++ b/kstyle/breezeshadowhelper.cpp	2016-09-14 12:11:53.986138791 +0200
@@ -72,8 +72,13 @@
     {
 
         #if BREEZE_HAVE_X11
-        if( Helper::isX11() )
-        { foreach( const quint32& value, _pixmaps  ) xcb_free_pixmap( Helper::connection(), value ); }
+        if( Helper::isX11() ) {
+            foreach( const quint32& value, _pixmaps  ) {
+                xcb_connection_t *xcbconnection = Helper::connection();
+		if(xcbconnection)
+                    xcb_free_pixmap( xcbconnection, value );
+            }
+        }
         #endif
 
     }
