--- a/shell/main.cpp	2017-06-27 12:00:38.000000000 +0200
+++ b/shell/main.cpp	2017-07-08 13:13:46.464382440 +0200
@@ -52,6 +52,11 @@
 
     QQuickWindow::setDefaultAlphaBuffer(true);
 
+    const QByteArray loopName = qgetenv("QSG_RENDER_LOOP");
+    if (loopName.isEmpty()) {
+        setenv("QSG_RENDER_LOOP", "basic", 1);
+    }
+
     KLocalizedString::setApplicationDomain("plasmashell");
 
     QApplication app(argc, argv);
