--- a/drivers/gpu/drm/rockchip/rockchip_drm_fb.c	2020-04-05 16:44:24.455973964 +0200
+++ b/drivers/gpu/drm/rockchip/rockchip_drm_fb.c	2020-04-05 16:45:08.115463022 +0200
@@ -88,8 +88,8 @@
 	 * this value would be used to check framebuffer size limitation
 	 * at drm_mode_addfb().
 	 */
-	dev->mode_config.max_width = 4096;
-	dev->mode_config.max_height = 4096;
+	dev->mode_config.max_width = 8192;
+	dev->mode_config.max_height = 8192;
 
 	dev->mode_config.funcs = &rockchip_drm_mode_config_funcs;
 	dev->mode_config.helper_private = &rockchip_mode_config_helpers;
