diff --git a/Package.swift b/Package.swift
index 5f78f8b..01c0c44 100644
--- a/Package.swift
+++ b/Package.swift
@@ -15,3 +15,16 @@ let package = Package(
         .Package(url: "https://github.com/vapor/clibressl.git", majorVersion: 1)
     ]
 )
+
+let elib = Product(name: "Essentials", type: .Library(.Dynamic), modules: "Essentials")
+let blib = Product(name: "BCrypt", type: .Library(.Dynamic), modules: "BCrypt")
+let halib = Product(name: "Hash", type: .Library(.Dynamic), modules: "Hash")
+let rlib = Product(name: "Random", type: .Library(.Dynamic), modules: "Random")
+let hclib = Product(name: "HMAC", type: .Library(.Dynamic), modules: "HMAC")
+let clib = Product(name: "Cipher", type: .Library(.Dynamic), modules: "Cipher")
+products.append(elib)
+products.append(blib)
+products.append(halib)
+products.append(rlib)
+products.append(hclib)
+products.append(clib)
