diff --git a/Package.swift b/Package.swift
index 5f78f8b..8a9bced 100644
--- a/Package.swift
+++ b/Package.swift
@@ -10,8 +10,4 @@ let package = Package(
         Target(name: "Random", dependencies: ["Essentials"]),
         Target(name: "HMAC", dependencies: ["Essentials", "Random"]),
         Target(name: "Cipher", dependencies: ["Essentials"]),
-    ],
-    dependencies: [
-        .Package(url: "https://github.com/vapor/core.git", majorVersion: 1),
-        .Package(url: "https://github.com/vapor/clibressl.git", majorVersion: 1)
     ]
 )
