--- a/xbmc/utils/RssReader.cpp	2016-11-27 19:26:10.347850005 +0100
+++ b/xbmc/utils/RssReader.cpp	2016-11-27 19:26:38.297751744 +0100
@@ -269,11 +269,11 @@
   if (m_tagSet.empty())
     AddTag("title");
 
-  while (itemNode > 0)
+  while (itemNode)
   {
     TiXmlNode* childNode = itemNode->FirstChild();
     mTagElements.clear();
-    while (childNode > 0)
+    while (childNode)
     {
       std::string strName = childNode->ValueStr();
 
