--- a/xbmc/guilib/TextureBundleXBT.h	2016-09-19 19:41:10.185125803 +0200
+++ b/xbmc/guilib/TextureBundleXBT.h	2016-09-19 19:41:19.334079785 +0200
@@ -24,6 +24,7 @@
 #include <string>
 #include <stdint.h>
 #include <vector>
+#include <ctime>
 
 class CAutoTexBuffer;
 class CBaseTexture;
--- a/xbmc/system.h.bak	2016-09-19 19:14:19.977877846 +0200
+++ b/xbmc/system.h	2016-09-19 19:32:32.082352652 +0200
@@ -219,6 +219,7 @@
 
 #if defined(TARGET_POSIX)
 #include <time.h>
+#include <stdlib.h>
 #include <sys/time.h>
 #include <sys/types.h>
 #include <errno.h>
--- a/xbmc/guilib/GUIWindowManager.h	2016-09-19 19:25:20.456126201 +0200
+++ b/xbmc/guilib/GUIWindowManager.h	2016-09-19 19:26:46.030982885 +0200
@@ -29,6 +29,7 @@
  */
 
 #include <list>
+#include <cstdlib>
 #include <utility>
 
 #include "DirtyRegionTracker.h"
--- a/xbmc/guilib/XBTF.h	2016-09-19 19:15:58.528825390 +0200
+++ b/xbmc/guilib/XBTF.h	2016-09-19 19:16:08.327818594 +0200
@@ -22,6 +22,7 @@
 #include <map>
 #include <string>
 #include <vector>
+#include <ctime>
 
 #include <stdint.h>
 
