--- a/configure.ac	2016-10-24 20:51:06.000000000 +0200
+++ b/configure.ac	2016-10-25 22:19:13.210932049 +0200
@@ -64,13 +64,9 @@
     fi
   elif [[ "$host_vendor" != "apple" ]]; then
     AC_MSG_CHECKING([for lib$2 soname])
-    $1_FILENAME=$($CC -nostdlib -o /dev/null $CFLAGS $LDFLAGS $4 -l$2 -Wl,-M 2>/dev/null | grep "^LOAD.*$2" | awk '{V=2; print $V}')
-    if [[ -z $$1_FILENAME ]]; then
-      #try gold linker syntax
-      $1_FILENAME=$($CC -nostdlib -o /dev/null $CFLAGS $LDFLAGS $4 -l$2 -Wl,-t 3>&1 1>&2 2>&3 | grep "lib$2")
-    fi
+    $1_FILENAME=$($CC -nostdlib -o /dev/null $CFLAGS $LDFLAGS $4 -l$2 -Wl,-t 2>&1 | grep "lib$2")
     if [[ ! -z $$1_FILENAME ]]; then
-      $1_SONAME=$($OBJDUMP -p $$1_FILENAME | grep "SONAME.*$2" | awk '{V=2; print $V}')
+      $1_SONAME=$($READELF -d $$1_FILENAME | grep "SONAME.*$2" | grep -oP '\(\K[[^)]]+')
     fi
   else
     AC_MSG_CHECKING([for lib$2 dylib])
