--- a/project/cmake/scripts/linux/Macros.cmake	2017-01-19 19:27:14.830368107 +0100
+++ b/project/cmake/scripts/linux/Macros.cmake	2017-01-19 19:28:24.374098597 +0100
@@ -73,12 +73,12 @@
   string(STRIP "${${lib}_FILENAME}" ${lib}_FILENAME)
   if(NOT ${lib}_FILENAME)
     execute_process(COMMAND ${CMAKE_C_COMPILER} -nostdlib -o /dev/null -Wl,-t ${link_lib}
-                    OUTPUT_QUIET
-                    ERROR_VARIABLE _TMP_FILENAME)
+                    ERROR_QUIET
+                    OUTPUT_VARIABLE _TMP_FILENAME)
     string(REGEX MATCH ".*lib${liblow}.so" ${lib}_FILENAME ${_TMP_FILENAME})
   endif()
   if(${lib}_FILENAME)
-    execute_process(COMMAND objdump -p ${${lib}_FILENAME}
+    execute_process(COMMAND readelf -d ${${lib}_FILENAME}
                     COMMAND grep SONAME.*${liblow}
                     ERROR_QUIET
                     OUTPUT_VARIABLE ${lib}_SONAME)
