--- a/src/helper/UserSession.cpp.bak	2016-02-17 23:30:24.896244709 +0100
+++ b/src/helper/UserSession.cpp	2016-02-17 23:30:35.820244058 +0100
@@ -26,6 +26,7 @@
 #include <sys/types.h>
 #include <sys/ioctl.h>
 #include <unistd.h>
+#include <cerrno>
 #include <pwd.h>
 #include <grp.h>
 #include <unistd.h>
--- a/src/daemon/VirtualTerminal.cpp.bak	2016-02-17 23:29:56.496246402 +0100
+++ b/src/daemon/VirtualTerminal.cpp	2016-02-17 23:30:07.745245731 +0100
@@ -23,6 +23,7 @@
 #include "VirtualTerminal.h"
 
 #include <unistd.h>
+#include <cerrno>
 #include <fcntl.h>
 #include <signal.h>
 #include <linux/vt.h>
