--- a/src/helper/UserSession.cpp	2016-02-17 23:30:24.896244709 +0100
+++ b/src/helper/UserSession.cpp	2016-02-17 23:30:35.820244058 +0100
@@ -26,6 +26,7 @@
 #include <sys/types.h>
 #include <sys/ioctl.h>
 #include <unistd.h>
+#include <cerrno>
 #include <pwd.h>
 #include <grp.h>
 #include <unistd.h>
--- a/src/daemon/VirtualTerminal.cpp	2016-02-17 23:29:56.496246402 +0100
+++ b/src/daemon/VirtualTerminal.cpp	2016-02-17 23:30:07.745245731 +0100
@@ -23,6 +23,7 @@
 #include "VirtualTerminal.h"
 
 #include <unistd.h>
+#include <cerrno>
 #include <fcntl.h>
 #include <signal.h>
 #include <linux/vt.h>
--- a/src/helper/HelperApp.cpp	2017-12-08 12:36:12.154294996 +0100
+++ b/src/helper/HelperApp.cpp	2017-12-08 12:36:32.487250657 +0100
@@ -33,6 +33,7 @@
 #include <iostream>
 #include <unistd.h>
 #include <sys/socket.h>
+#include <sys/time.h>
 
 #include <utmp.h>
 #include <utmpx.h>
