--- src/Language/Haskell/TH/Lens.hs.bak	2020-01-27 23:59:30.637933018 +0100
+++ src/Language/Haskell/TH/Lens.hs	2020-01-28 00:02:02.966514892 +0100
@@ -2062,7 +2062,7 @@
       remitter _ = Nothing
 #endif
 
-_TupE :: Prism' Exp [Exp]
+_TupE :: Prism' Exp [Maybe Exp]
 _TupE
   = prism' reviewer remitter
   where
@@ -2070,7 +2070,7 @@
       remitter (TupE x) = Just x
       remitter _ = Nothing
 
-_UnboxedTupE :: Prism' Exp [Exp]
+_UnboxedTupE :: Prism' Exp [Maybe Exp]
 _UnboxedTupE
   = prism' reviewer remitter
   where
