--- Data/ByteArray/Sized.hs.bak	2018-09-22 18:58:12.399968766 +0200
+++ Data/ByteArray/Sized.hs	2018-09-22 18:59:59.812525800 +0200
@@ -77,6 +77,7 @@
 import qualified Data.ByteArray.Types as ByteArray (allocRet)
 
 #if MIN_VERSION_basement(0,0,7)
+import           GHC.TypeLits
 import           Basement.BlockN (BlockN)
 import qualified Basement.BlockN as BlockN
 import qualified Basement.PrimType as Base
@@ -132,7 +133,7 @@
          , KnownNat n
          , Countable ty n
          , KnownNat nbytes
-         , nbytes ~ (Base.PrimSize ty (*) n)
+         , nbytes ~ (Base.PrimSize ty GHC.TypeLits.* n)
          ) => ByteArrayN nbytes (BlockN n ty) where
     allocRet _ f = do
         mba <- BlockN.new @n
