--- a/src/plugins/multimedia/ffmpeg/playbackengine/qffmpegmediadataholder.cpp	2025-05-04 15:16:35.000000000 +0200
+++ b/src/plugins/multimedia/ffmpeg/playbackengine/qffmpegmediadataholder.cpp	2025-08-21 17:21:19.754345439 +0200
@@ -215,6 +215,9 @@
     AVDictionaryHolder dict;
     constexpr auto NetworkTimeoutUs = "5000000";
     av_dict_set(dict, "timeout", NetworkTimeoutUs, 0);
+    av_dict_set(dict, "fflags", "discardcorrupt", 0 );
+    av_dict_set(dict, "reconnect_on_network_error", "1", 0);
+    av_dict_set(dict, "reconnect", "1", 0);
 
     const QByteArray protocolWhitelist = qgetenv("QT_FFMPEG_PROTOCOL_WHITELIST");
     if (!protocolWhitelist.isNull())
