--- a/src/plugins/multimedia/ffmpeg/playbackengine/qffmpegmediadataholder.cpp	2025-11-22 17:31:49.016845638 +0100
+++ b/src/plugins/multimedia/ffmpeg/playbackengine/qffmpegmediadataholder.cpp	2025-11-22 17:33:26.315407926 +0100
@@ -225,6 +225,10 @@
     {
         const milliseconds timeout = playbackOptions.networkTimeout();
         av_dict_set_int(dict, "timeout", duration_cast<microseconds>(timeout).count(), 0);
+        av_dict_set(dict, "fflags", "discardcorrupt", 0 );
+        av_dict_set(dict, "reconnect_on_network_error", "1", 0);
+        av_dict_set(dict, "reconnect", "1", 0);
+
         qCDebug(qLcMediaDataHolder) << "Using custom network timeout:" << timeout;
     }
 
