--- a/src/plugins/multimedia/ffmpeg/qffmpegmediaintegration.cpp	2025-08-22 19:29:14.583155560 +0200
+++ b/src/plugins/multimedia/ffmpeg/qffmpegmediaintegration.cpp	2025-08-22 19:29:00.922188222 +0200
@@ -110,6 +110,8 @@
     if (qEnvironmentVariableIsSet("QT_FFMPEG_DEBUG")) {
         av_log_set_level(AV_LOG_DEBUG);
         UseCustomFFmpegLogger = true;
+    } else {
+        av_log_set_level(AV_LOG_FATAL);
     }
 
     av_log_set_callback(&qffmpegLogCallback);
