--- a/src/CMakeLists.txt  2018-09-07 13:09:42.439702458 +0200
+++ b/src/CMakeLists.txt  2018-09-07 13:10:27.054656713 +0200
@@ -123,7 +123,7 @@
         if (GRAPHITE2_SANITIZERS)
             target_link_libraries(graphite2 c gcc_s)
         else ()
-            target_link_libraries(graphite2 c gcc)
+            target_link_libraries(graphite2 c)
         endif ()
     endif (${CMAKE_CXX_COMPILER} MATCHES  ".*mingw.*")
     include(Graphite)
