--- a/build/generic/configure	2016-10-16 13:47:28.236627181 +0200
+++ b/build/generic/configure	2016-10-16 13:47:46.661625664 +0200
@@ -4369,7 +4369,7 @@
 	SHARED_LIB="libxvidcore.\$(SHARED_EXTENSION).\$(API_MAJOR).\$(API_MINOR)"
 	SO_API_MAJOR_LINK="libxvidcore.\$(SHARED_EXTENSION).\$(API_MAJOR)"
 	SO_LINK="libxvidcore.\$(SHARED_EXTENSION)"
-	SPECIFIC_LDFLAGS="-Wl,-soname,libxvidcore.\$(SHARED_EXTENSION).\$(API_MAJOR) -shared -Wl,--version-script=libxvidcore.ld -lc -lm"
+	SPECIFIC_LDFLAGS="-Wl,-soname,libxvidcore.\$(SHARED_EXTENSION).\$(API_MAJOR) -shared -Wl,--sandbox-version-script=libxvidcore.ld -lc -lm"
 	SPECIFIC_CFLAGS="-fPIC"
 	;;
      *bsd*|irix*)
